library(shiny)
library(shinyWidgets)
library(plyr)
library(ggplot2)
library(umap)
library(tidyr)
library(dbscan)
library(readr)
library(dplyr)
library(stringr)
library(RColorBrewer)
library(gt)
library(gtsummary)
library(flextable)
library(Rediscover)
library(survival)
library(gridExtra)
library(survminer)
library(tranSurv)
library(DT)
library(ComplexHeatmap)
library(ggsci)
library(scales)
library(patchwork)
library(sjPlot)
library(sjlabelled)
library(forcats)
library(markdown)
library(PropCIs)
library(shinythemes)
library(httr)
library(drawProteins)
library(ggrepel)
library(rms)
library(blorr)
library(dcurves)
library(Matching)
library(survRM2)
library(shinydashboard)
library(pROC)
library(tidymodels)
library(withr)
library(rpart)
library(ranger)
library(bonsai)
library(probably)
library(discrim)
library(flexsurv)
library(data.table)
library(qs2)
library(foreach)
library(doParallel)
library(mltools)
library(reshape2)
library(bigstep)
library(furrr)
library(broom)
library(twang)
library(shinyjs)


if(file.exists("source/Data_mutation_cord.qs")){
  file.remove("source/Data_mutation_cord.qs")
}

registerDoParallel(cores = max(1, parallel::detectCores() - 1, na.rm = TRUE))

tidymodels_prefer()
options(pillar.advice = FALSE, pillar.min_title_chars = Inf)

nietzsche = read.csv(header = TRUE,
                     file("source/nietzsche.csv",
                          encoding='UTF-8-BOM'))$text
tmp <- reactiveValues()

# Rstan options
#rstan_options(auto_write = TRUE)
options(mc.cores = parallel::detectCores())
options(expressions = 500000) 

# Specify the application port
options(shiny.host = "0.0.0.0")
options(shiny.port = 3838)

ht_opt$message = FALSE

Penal = 1
Toler = 10^-14
GENE_NO_THRESHOLD = 30

get_cpu_architecture <- function() {
  arch <- R.version$arch
  if (arch == "aarch64") {
    return("arm64")
  } else if (arch == "x86_64") {
    return("amd64")
  } else {
    return("unknown")
  }
}

QS_READ <- function(nthreads, file, ...) {
  cpu_arch <- get_cpu_architecture()
  if (cpu_arch == "arm64") {
    # Arm64の場合、nthreadsをデフォルト設定
    qs2::qs_read(file=file, nthreads=nthreads, ...)
  } else if(!file.exists("/.dockerenv")){
    # その他のアーキテクチャの場合、nthreadsを1に設定
    # readRDS(file=file, ...)
    qs2::qs_read(file=file, nthreads = 1, ...)
  } else {
    qs2::qs_read(file=file, nthreads = 1, ...)
    # print(1)
  }
}
QS_SAVE <- function(nthreads, object, file, ...) {
  cpu_arch <- get_cpu_architecture()
  if (cpu_arch == "arm64") {
    # Arm64の場合、nthreadsをデフォルト設定
    qs2::qs_save(nthreads=nthreads, object=object, file=file, ...)
  } else if(!file.exists("/.dockerenv")){
    # その他のアーキテクチャの場合、nthreadsを1に設定
    # saveRDS(object=object, file=file, ...)
    qs2::qs_save(nthreads = 1, object=object, file=file, ...)
  } else {
    qs2::qs_save(nthreads = 1, object=object, file=file, ...)
    # print(1)
  }
}


options(shiny.maxRequestSize=16*1024^3)
conflicted::conflicts_prefer(.quiet = T, gtsummary::all_double)
conflicted::conflicts_prefer(.quiet = T, gtsummary::all_factor)
conflicted::conflicts_prefer(.quiet = T, gtsummary::all_integer)
conflicted::conflicts_prefer(.quiet = T, gtsummary::all_logical)
conflicted::conflicts_prefer(.quiet = T, gtsummary::all_numeric)
conflicted::conflicts_prefer(.quiet = T, patchwork::area)
conflicted::conflicts_prefer(.quiet = T, dplyr::arrange)
conflicted::conflicts_prefer(.quiet = T, sjlabelled::as_factor)
conflicted::conflicts_prefer(.quiet = T, gtsummary::as_flextable)
conflicted::conflicts_prefer(.quiet = T, sjlabelled::as_label)
conflicted::conflicts_prefer(.quiet = T, flextable::border)
conflicted::conflicts_prefer(.quiet = T, shinydashboard::box)
conflicted::conflicts_prefer(.quiet = T, readr::col_factor)
conflicted::conflicts_prefer(.quiet = T, purrr::compact)
conflicted::conflicts_prefer(.quiet = T, flextable::compose)
conflicted::conflicts_prefer(.quiet = T, gtsummary::continuous_summary)
conflicted::conflicts_prefer(.quiet = T, dplyr::count)
conflicted::conflicts_prefer(.quiet = T, DT::dataTableOutput)
conflicted::conflicts_prefer(.quiet = T, dplyr::desc)
conflicted::conflicts_prefer(.quiet = T, purrr::discard)
conflicted::conflicts_prefer(.quiet = T, dplyr::failwith)
conflicted::conflicts_prefer(.quiet = T, stringr::fixed)
conflicted::conflicts_prefer(.quiet = T, flextable::font)
conflicted::conflicts_prefer(.quiet = T, gt::html)
conflicted::conflicts_prefer(.quiet = T, dplyr::id)
conflicted::conflicts_prefer(.quiet = T, plyr::is.discrete)
conflicted::conflicts_prefer(.quiet = T, dplyr::lag)
conflicted::conflicts_prefer(.quiet = T, dplyr::mutate)
conflicted::conflicts_prefer(.quiet = T, survival::myeloma)
conflicted::conflicts_prefer(.quiet = T, shiny::observe)
conflicted::conflicts_prefer(.quiet = T, tidyr::pack)
conflicted::conflicts_prefer(.quiet = T, recipes::prepare)
conflicted::conflicts_prefer(.quiet = T, dials::prune)
conflicted::conflicts_prefer(.quiet = T, dplyr::rename)
conflicted::conflicts_prefer(.quiet = T, DT::renderDataTable)
conflicted::conflicts_prefer(.quiet = T, flextable::rotate)
conflicted::conflicts_prefer(.quiet = T, dplyr::select)
conflicted::conflicts_prefer(.quiet = T, dials::smoothness)
conflicted::conflicts_prefer(.quiet = T, yardstick::spec)
conflicted::conflicts_prefer(.quiet = T, dplyr::src)
conflicted::conflicts_prefer(.quiet = T, recipes::step)
conflicted::conflicts_prefer(.quiet = T, dplyr::summarise)
conflicted::conflicts_prefer(.quiet = T, dplyr::summarize)
conflicted::conflicts_prefer(.quiet = T, parsnip::translate)
conflicted::conflicts_prefer(.quiet = T, tidyr::unpack)
conflicted::conflicts_prefer(.quiet = T, recipes::update)
conflicted::conflicts_prefer(.quiet = T, rms::validate)
conflicted::conflicts_prefer(.quiet = T, dplyr::between)
conflicted::conflicts_prefer(.quiet = T, dplyr::first)
conflicted::conflicts_prefer(.quiet = T, dplyr::last)
conflicted::conflicts_prefer(.quiet = T, purrr::transpose)
conflicted::conflicts_prefer(.quiet = T, foreach::accumulate)
conflicted::conflicts_prefer(.quiet = T, fastshap::explain)
conflicted::conflicts_prefer(.quiet = T, tidyr::replace_na)
conflicted::conflicts_prefer(.quiet = T, bigstep::prepare_data)
conflicted::conflicts_prefer(.quiet = T, survival::cluster)
conflicted::conflicts_prefer(.quiet = T, data.table::dcast)
conflicted::conflicts_prefer(.quiet = T, yardstick::mcc)
conflicted::conflicts_prefer(.quiet = T, yardstick::rmse)
conflicted::conflicts_prefer(.quiet = T, yardstick::sensitivity)
conflicted::conflicts_prefer(.quiet = T, tidyr::smiths)
conflicted::conflicts_prefer(.quiet = T, purrr::when)
conflicted::conflicts_prefer(.quiet = T, shinyWidgets::alert)
conflicted::conflicts_prefer(.quiet = T, methods::removeClass)
conflicted::conflicts_prefer(.quiet = T, shiny::runExample)
conflicted::conflicts_prefer(.quiet = T, rstan::show)
Abs = function(x){
  x = as.numeric(x)
  x[is.na(x)]=1
  return(x)
}

fun_zero <- function(a, b){
  if(length(a) != length(b)){
    if(length(a) / length(b) == as.integer(length(a) / length(b))){
      b = rep(b,length(a) / length(b))
    }
  }
  return(ifelse(b == 0, 0, a / b))
}

gg_empty = function(){
  g = ggplot()
  g = g + geom_blank()
  g = g + ggtitle("")
  g = g + theme_void()
  return(g)
}

ggsurvplot_empty = function(){
  g = ggsurvplot(survfit(Surv(time, status) ~ 1, data = lung),
                 title = "",
                 palette = "white",
                 ggtheme=theme_void())
  g$plot <- g$plot + theme(axis.text.x = element_blank(),
                           axis.text.y = element_blank(),
                           plot.title = element_blank(),
                           plot.subtitle = element_blank(),
                           legend.title=element_blank(),
                           legend.text=element_blank(),
                           axis.ticks=element_blank(),
                           axis.ticks.x = element_blank())
  return(g)
}

gg_drug_plot = function(data, x_name, y_name, x_lab, y_lab, Total_pts, x_max_patients){
  Cor = round(cor.test(data[,colnames(data) == x_name], data[,colnames(data) == y_name])$estimate,3)
  P = round(cor.test(data[,colnames(data) == x_name], data[,colnames(data) == y_name])$p.value,3)
  cor = data.frame(Cor, P)
  Max_x = max(data[,colnames(data) == x_name],na.rm = T) * 1.05
  Max_y = max(data[,colnames(data) == y_name],na.rm = T)
  g <- ggplot(data, aes_(x = as.name(x_name), y = as.name(y_name), color = as.name("Cancers")))
  g <- g + geom_point(size = Total_pts/x_max_patients*10, alpha = .5)
  g <- g + scale_fill_nejm()
  g <- g + coord_cartesian(xlim = c(0, Max_x), ylim = c(0, Max_y*1.05))
  g <- g + theme_bw()
  g <- g + labs(x = x_lab,y = y_lab, color = "Cancer type")
  g <- g + geom_smooth(method = lm, se = FALSE, na.rm = T, colour = "red")
  g <- g + geom_text(data=cor, colour = "red", mapping = aes(x = 0.3*Max_x, y = Max_y*1.03, label = paste("r =",Cor, ", p =",P)))
}

shannon.entropy = function(x, x_length){ 
  x = x[(!is.na(x))]
  x = x[(x != 0)]
  if(sum(x) <= 0) {
    return(log(x_length, base=2)) 
  } else {
    p = x/sum(x)
    score = -sum(p * log(p, base=2))
    return(score)
  }
} 

odds.ratio <- function(a, b, c, d, correct=FALSE){
  cl <- function(x){
    or*exp(c(1,-1)*qnorm(x)*sqrt(1/a+1/b+1/c+1/d))
  }
  if (correct || a*b*c*d==0) {
    a <- a+0.5
    b <- b+0.5
    c <- c+0.5
    d <- d+0.5
  }
  or <- a*d/(b*c)
  conf <- rbind(cl90=cl(0.05), cl95=cl(0.025), cl99=cl(0.005), cl999=cl(0.0005))
  conf <- data.frame(conf)
  colnames(conf) <- paste(c("lower","upper"), " limit" , sep="")
  rownames(conf) <- paste(c(90, 95, 99, 99.9), "%CI" , sep="")
  list(or=or, conf=conf)
}

save_table_docx <- function(Data_summary, separate_var, title, filename){
  tmp = Data_summary %>%
    tbl_summary(
      by = separate_var,
      statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                            "{mean} ({sd})",
                                            "{median} ({p25}, {p75})", 
                                            "{min}, {max}"),
                       all_categorical() ~ "{n} / {N} ({p}%)"),
      type = list(all_continuous() ~ "continuous2",
                  all_dichotomous() ~ "categorical"),
      digits = all_continuous() ~ 2,
      missing = "ifany")
  if(!is.null(separate_var)){
    tmp = tmp %>%
      add_p(pvalue_fun = ~style_pvalue(.x, digits = 2)) %>%
      add_overall()
  }
  tmp %>%
    add_n() %>%
    modify_header(label ~ "**Variable**") %>%
    modify_caption(title) %>%
    bold_labels() %>% as_flex_table() %>% save_as_docx(path = filename)
}

surv_curv_entry <- function(fit, data, title, legend_, diff_0, diff_1){
  g = ggsurvplot(
    fit = fit,
    combine = TRUE,
    data = data,
    xlab = "Time from enrollment (months)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    surv.scale = "percent",
    font.title = 8,
    font.subtitle = 8,
    font.main = 8,
    font.submain = 8,
    font.caption = 8,
    font.legend = 8,
    pval = FALSE,
    surv.median.line = "v",
    palette = "Dark2",
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    cumevents = FALSE,
    cumcensor = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, min(365.25*10, max(data$time_enroll_final)) * 1.05),
    xscale = "d_m",
    break.x.by = 6 * 365.25 / 12,
    # break.x.by = ceiling(max(data$time_enroll_final) / 365.25 / 6,
    legend.labs = legend_
  )
  if(is.null(diff_0)){
    tmp = summary(fit)$table
    legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
    g = g +
      labs(title = title,
           subtitle = paste0("Median OS, ", legends, " months"))
  } else{
    tmp = data.frame(summary(fit)$table)
    legends = paste0(round(tmp$median[1] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1),")")
    for(i in 2:length(tmp$median)){
      legends = paste(legends, paste0(round(tmp$median[i] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[i] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[i] / 365.25 * 12, digits = 1),")"), sep=", ")
    }
    g = g +
      labs(title = paste0(title, ": log-rank, p=", format(
        1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
        "/Wilcoxon, p=", format(
          1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3)),
        subtitle = paste0("Median OS, ", legends, " months"))
  }
  g$table <- g$table + theme(plot.title = element_blank(),
                             plot.subtitle = element_blank())
  return(g)
}

surv_curv_CTx <- function(fit, data, title, legend_, diff_0){
  g = ggsurvplot(
    fit = fit,
    combine = TRUE,
    data = data,
    xlab = "Time from chemotherapy induction, risk-set adjusted (months)",
    ylab = "Survival Probability",
    censor = TRUE,
    conf.int = FALSE,
    surv.scale = "percent",
    font.title = 8,
    font.subtitle = 8,
    font.main = 8,
    font.submain = 8,
    font.caption = 8,
    font.legend = 8,
    pval = FALSE,
    surv.median.line = "v",
    palette = "Dark2",
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    cumevents = FALSE,
    cumcensor = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xlim = c(0, min(365.25*10, max(data$time_palliative_final)) * 1.05),
    xscale = "d_m",
    break.x.by = 12 * 365.25 / 12,
    # break.x.by = ceiling(max(data$time_enroll_final) / 365.25 / 6,
    legend.labs = legend_
  )
  if(is.null(diff_0)){
    tmp = summary(fit)$table
    legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
    g = g +
      labs(title = title,
           subtitle = paste0("Median OS, ", legends, " months"))
  } else{
    tmp = data.frame(summary(fit)$table)
    legends = paste0(round(tmp$median[1] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1),")")
    for(i in 2:length(tmp$median)){
      legends = paste(legends, paste0(round(tmp$median[i] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[i] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[i] / 365.25 * 12, digits = 1),")"), sep=", ")
    }
    data_tmp = tidy(diff_0, exponentiate=TRUE, conf.int=TRUE)
    title_HR = NULL
    if(length(data_tmp$term) > 0){
      for(i in 1:length(data_tmp$term)){
        title_HR = c(title_HR,
                     paste0(format(data_tmp$estimate[i],digits=3), " (",
                            format(data_tmp$conf.low[i],digits=3), "-",
                            format(data_tmp$conf.high[i],digits=3), ") ",
                            "p=", format(data_tmp$p.value[i],digits=3)))
      }
    }
    g = g +
      labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
        subtitle = paste0("Median OS, ", legends, " months"))
  }
  g$table <- g$table + theme(plot.title = element_blank(),
                             plot.subtitle = element_blank())
  return(g)
}

surv_curv_drug <- function(fit, data, title, diff_0, diff_1, diff_2){
  g = ggsurvplot(
    fit = fit,
    combine = TRUE,
    data = data,
    xlab = "Time from drug initiation (months)",
    ylab = "Survival Probability",
    censor = TRUE,
    surv.scale = "percent",
    conf.int = FALSE,
    font.title = 8,
    font.subtitle = 8,
    font.main = 8,
    font.submain = 8,
    font.caption = 8,
    font.legend = 8,
    pval = FALSE,
    surv.median.line = "v",
    palette = "Dark2",
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    cumevents = FALSE,
    cumcensor = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xscale = "d_m",
    xlim = c(0, 3*365.35),
    break.x.by = 3 * 365.25 / 12,
    # break.x.by = ceiling(max(data$time_enroll_final) / 365.25 / 6,
  )
  if(is.null(diff_0)){
    if(is.null(diff_2)){
      tmp = summary(fit)$table
      legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
      g = g +
        labs(title = title,
             subtitle = paste0("Median OS, ", legends, " months"))
      g$table <- g$table + theme(plot.title = element_blank(),
                                 plot.subtitle = element_blank())
    } else {
      tmp = summary(fit)$table
      legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
      data_tmp = tidy(diff_2, exponentiate=TRUE, conf.int=TRUE)
      title_HR = NULL
      if(length(data_tmp$term) > 0){
        for(i in 1:length(data_tmp$term)){
          title_HR = c(title_HR,
                       paste0(format(data_tmp$estimate[i],digits=3), " (",
                              format(data_tmp$conf.low[i],digits=3), "-",
                              format(data_tmp$conf.high[i],digits=3), ") ",
                              "p=", format(data_tmp$p.value[i],digits=3)))
        }
      }
      g = g +
        labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
             subtitle = paste0("Median OS, ", legends, " months"))
      g$table <- g$table + theme(plot.title = element_blank(),
                                 plot.subtitle = element_blank())
    }
  } else{
    tmp = data.frame(summary(fit)$table)
    legends = paste0(round(tmp$median[1] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1),")")
    for(i in 2:length(tmp$median)){
      legends = paste(legends, paste0(round(tmp$median[i] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[i] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[i] / 365.25 * 12, digits = 1),")"), sep=", ")
    }
    data_tmp = tidy(diff_2, exponentiate=TRUE, conf.int=TRUE)
    title_HR = NULL
    if(length(data_tmp$term) > 0){
      for(i in 1:length(data_tmp$term)){
        title_HR = c(title_HR,
                     paste0(format(data_tmp$estimate[i],digits=3), " (",
                            format(data_tmp$conf.low[i],digits=3), "-",
                            format(data_tmp$conf.high[i],digits=3), ") ",
                            "p=", format(data_tmp$p.value[i],digits=3)))
      }
    }
    g = g +
      labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
           subtitle = paste0("Median OS, ", legends, " months"))
    g = g +
      labs(title = paste0(title, ": log-rank, p=", format(
        1 - pchisq(diff_0$chisq, length(diff_0$n)-1, lower.tail = TRUE),digits=3),
        "/Wilcoxon, p=", format(
          1 - pchisq(diff_1$chisq, length(diff_1$n)-1, lower.tail = TRUE),digits=3)),
        subtitle = paste0("Median OS, ", legends," months, hazard ratio=", paste(title_HR, collapse = "/")))
    g$table <- g$table + theme(plot.title = element_blank(),
                               plot.subtitle = element_blank())
  }
  return(g)
}

surv_curv_drug_IPTW <- function(fit, data, title, diff_2){
  g = ggsurvplot(
    fit = fit,
    combine = TRUE,
    data = data,
    xlab = "Time from drug initiation (months)",
    ylab = "Survival Probability, IPTW corrected",
    censor = TRUE,
    surv.scale = "percent",
    conf.int = FALSE,
    font.title = 8,
    font.subtitle = 8,
    font.main = 8,
    font.submain = 8,
    font.caption = 8,
    font.legend = 8,
    pval = FALSE,
    surv.median.line = "v",
    palette = "Dark2",
    risk.table = TRUE,
    risk.table.y.text = FALSE,
    cumevents = FALSE,
    cumcensor = FALSE,
    tables.theme = clean_theme(), 
    legend = c(0.8,0.8),
    xscale = "d_m",
    xlim = c(0, 3*365.35),
    break.x.by = 3 * 365.25 / 12,
    # break.x.by = ceiling(max(data$time_enroll_final) / 365.25 / 6,
  )
  tmp = summary(fit)$table
  legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
  tmp = data.frame(summary(fit)$table)
  legends = paste0(round(tmp$median[1] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1),")")
  for(i in 2:length(tmp$median)){
    legends = paste(legends, paste0(round(tmp$median[i] / 365.25 * 12, digits = 1), " (", round(tmp$X0.95LCL[i] / 365.25 * 12, digits = 1), "-", round(tmp$X0.95UCL[i] / 365.25 * 12, digits = 1),")"), sep=", ")
  }
  data_tmp = tidy(diff_2, exponentiate=TRUE, conf.int=TRUE)
  title_HR = NULL
  if(length(data_tmp$term) > 0){
    for(i in 1:length(data_tmp$term)){
      title_HR = c(title_HR,
                   paste0(format(data_tmp$estimate[i],digits=3), " (",
                          format(data_tmp$conf.low[i],digits=3), "-",
                          format(data_tmp$conf.high[i],digits=3), ") ",
                          "p=", format(data_tmp$p.value[i],digits=3)))
    }
  }
  g = g +
    labs(title = paste0(title, ": hazard ratio=", paste(title_HR, collapse = "/")),
         subtitle = paste0("Median OS, ", legends, " months. Risk table is not valid due to IPTW correction."))
  g$table <- g$table + theme(plot.title = element_blank(),
                             plot.subtitle = element_blank())
  return(g)
}

# Define UI for application that draws a histogram
ui <- dashboardPage(
  skin = "black",
  title = "FELIS",
  dashboardHeader(
    title = HTML(
      "<div style='display: flex; align-items: center; background-color: #FFFFFF;'>
         <a href='https://github.com/MANO-B/FELIS'>
           <img src='FELIS.png' height='30px' style='margin-right: 10px;'>
         </a>
         <h6 style='margin: 0;'>FELIS version 2.0.0</h6>
       </div>"
    ),
    tags$li(class = "dropdown", 
            style = "height: 30px; padding: 0px;",
            actionButton(
              inputId = "logout_button",
              label = "Logout",
              icon = icon("sign-out"),
              style = "height: 28px; line-height: 28px; color: white; background-color: #d9534f; border-color: #ac2925; padding: 0px 10px;"
            )
    )
  ),
  dashboardSidebar(
    sidebarMenu(
      h3("Settings"),
      menuItem("Input C-CAT files", tabName = "InputC-CATfiles", icon = icon("dashboard")),
      menuItem("Setting", tabName = "Setting", icon = icon("dashboard")),
      menuItem("Analysis", tabName = "Analysis", icon = icon("th")),
      hr(),
      h3("Results"),
      menuItem("Case summary", icon = icon("th"),
               hr(),
               p("Tables"),
               menuSubItem("Summarized by mutation pattern", tabName = "Summarizedbymutationpattern", icon = icon("angle-right")),
               menuSubItem("Summarized by histology", tabName = "Summarizedbyhistology", icon = icon("angle-right")),
               hr()
      ),
      menuItem("Inter-panel comparison", icon = icon("th"),
               hr(),
               p("Figures"),
               menuSubItem("Comparison figure", tabName = "Inter_panel", icon = icon("angle-right")),
               p("Downloadable table"),
               menuSubItem("Table of comparison data", tabName = "Table_inter_panel", icon = icon("angle-right")),
               hr()
      ),
      menuItem("Oncoprint", icon = icon("th"),
               hr(),
               p("Figures"),
               menuSubItem("Oncoprint", tabName = "Oncoprint", icon = icon("angle-right")),
               menuSubItem("Lolliplot for the selected gene", tabName = "Lolliplotfortheselectedgene", icon = icon("angle-right")),
               hr(),
               p("Downloadable table"),
               menuSubItem("Table of clinical and mutation information per patient", tabName = "Tableofclinicalandmutationinformationperpatient", icon = icon("angle-right")),
               hr()
      ),
      menuItem("Mutual exclusivity", tabName = "Mutualexclusivity", icon = icon("th")),
      menuItem("Variation by histology", tabName = "Variationbyhistology", icon = icon("th")),
      menuItem("Clustering analysis", icon = icon("th"),
               hr(),
               p("Mutation-based clustering"),
               menuSubItem("Basic data", tabName = "Basicdata", icon = icon("angle-right")),
               menuSubItem("Table of basic data", tabName = "Table_basic_data", icon = icon("angle-right")),
               menuSubItem("UMAP clustering based on mutations", tabName = "UMAPclusteringbasedonmutations", icon = icon("angle-right")),
               menuSubItem("Cluster and age relationship", tabName = "Clusterandagerelationship", icon = icon("angle-right")),
               menuSubItem("Cluster and histology relationship", tabName = "Clusterandhistologyrelationship", icon = icon("angle-right")),
               menuSubItem("Heterogeneity within histologic types", tabName = "Heterogeneitywithinhistologictypes", icon = icon("angle-right")),
               menuSubItem("Table of clusters and histologies", tabName = "Tableofclustersandhistologies", icon = icon("angle-right")),
               menuSubItem("Table of clusters and genetic variants", tabName = "Tableofclustersandgeneticvariants", icon = icon("angle-right")),
               hr(),
               p("Factors leading to treatment"),
               menuSubItem("Frequency of patients with targeted therapy", tabName = "Frequencyofpatientswithtargetedtherapy", icon = icon("angle-right")),
               menuSubItem("Relationship between pre-CGP and post-CGP periods", tabName = "Relationshipbetweenpre-CGPandpost-CGPperiods", icon = icon("angle-right")),
               menuSubItem("Patients per histology and treatment reach rate", tabName = "Patientsperhistologyandtreatmentreachrate", icon = icon("angle-right")),
               menuSubItem("Pre-CGP period and treatment reach rate", tabName = "Pre-CGPperiodandtreatmentreachrate", icon = icon("angle-right")),
               menuSubItem("Post-CGP period and treatment reach rate", tabName = "Post-CGPperiodandtreatmentreachrate", icon = icon("angle-right")),
               menuSubItem("Age and treatment reach rate", tabName = "Ageandtreatmentreachrate", icon = icon("angle-right")),
               hr()
      ),
      menuItem("Survival after CGP", icon = icon("th"),
               hr(),
               p("Survival analysis"),
               menuSubItem("Survival and treatment after CGP", tabName = "SurvivalandtreatmentafterCGP", icon = icon("angle-right")),
               menuSubItem("Survival after CGP and performance status", tabName = "SurvivalafterCGPandperformancestatus", icon = icon("angle-right")),
               menuSubItem("Survival after CGP and previous treatment", tabName = "SurvivalafterCGPandprevioustreatment", icon = icon("angle-right")),
               menuSubItem("Survival after CGP and mutations, forest plot", tabName = "SurvivalafterCGPandmutationsforestplot", icon = icon("angle-right")),
               menuSubItem("Survival after CGP and mutations, KM-curve", tabName = "SurvivalafterCGPandmutationsKM-curve", icon = icon("angle-right")),
               menuSubItem("Hazard ratio for survival after CGP - genes", tabName = "HazardratioforsurvivalafterCGP_1", icon = icon("angle-right")),
               menuSubItem("Hazard ratio for survival after CGP - clusters", tabName = "HazardratioforsurvivalafterCGP_2", icon = icon("angle-right")),
               menuSubItem("Survival period and treatment reach rate", tabName = "Survival_treatment_reach", icon = icon("angle-right")),
               hr()
      ),
      menuItem("CGP benefit prediction", icon = icon("th"),
               hr(),
               p("Factors leading to treatment"),
               menuSubItem("Factors leading to treatment, pre-CGP, Nomogram", tabName = "FactorsleadingtoTreatmentpre-CGPNomogram", icon = icon("angle-right")),
               menuSubItem("Factors for recommending treatment, pre-CGP, Nomogram", tabName = "FactorsrecommendingTreatmentpre-CGPNomogram", icon = icon("angle-right")),
               menuSubItem("Factors for evidence level A/B, pre-CGP, Nomogram", tabName = "FactorsrecommendingTreatmentABpre-CGPNomogram", icon = icon("angle-right")),
               menuSubItem("Factors leading to treatment, post-CGP, Nomogram", tabName = "FactorsleadingtoTreatmentpost-CGPNomogram", icon = icon("angle-right")),
               menuSubItem("Factors leading to treatment, pre-CGP, Odds ratio", tabName = "FactorsleadingtoTreatmentpre-CGPOddsratio", icon = icon("angle-right")),
               menuSubItem("Factors for recommending treatment, pre-CGP, Odds ratio", tabName = "FactorsrecommendingTreatmentpre-CGPOddsratio", icon = icon("angle-right")),
               menuSubItem("Factors for evidence level A/B, pre-CGP, Odds ratio", tabName = "FactorsrecommendingTreatmentABpre-CGPOddsratio", icon = icon("angle-right")),
               menuSubItem("Factors leading to treatment, post-CGP, Odds ratio", tabName = "FactorsleadingtoTreatmentpost-CGPOddsratio", icon = icon("angle-right")),
               menuSubItem("Factors leading to treatment, decision curve", tabName = "FactorsleadingtoTreatmentdecisioncurve", icon = icon("angle-right")),
               menuSubItem("Factors for recommending treatment, decision curve", tabName = "FactorsrecommendingTreatmentdecisioncurve", icon = icon("angle-right")),
               menuSubItem("Factors for evidence level A/B, decision curve", tabName = "FactorsrecommendingTreatmentABdecisioncurve", icon = icon("angle-right")),
               menuSubItem("ROC curve of nomogram, leading to treatment", tabName = "ROCnomogram", icon = icon("angle-right")),
               menuSubItem("ROC curve of nomogram, recommending to treatment", tabName = "ROCnomogram_rec", icon = icon("angle-right")),
               menuSubItem("ROC curve of nomogram, for evidence level A/B", tabName = "ROCnomogram_GMT", icon = icon("angle-right")),
               menuSubItem("Machine learning prediction, leading to treatment", tabName = "preCGPMachinelearning", icon = icon("angle-right")),
               menuSubItem("Machine learning prediction, recommending treatment", tabName = "preCGPMachinelearning_rec", icon = icon("angle-right")),
               menuSubItem("Machine learning prediction, evidence level A/B", tabName = "preCGPMachinelearning_GMT", icon = icon("angle-right")),
               menuSubItem("Factors leading to treatment, table", tabName = "FactorsleadingtoTreatmenttable", icon = icon("angle-right")),
               menuSubItem("Factors for recommending treatment, table", tabName = "FactorsleadingtoTreatmenttable_rec", icon = icon("angle-right")),
               menuSubItem("Factors for evidence level A/B, table", tabName = "FactorsleadingtoTreatmenttable_GMT", icon = icon("angle-right")),
               menuSubItem("Prediction result table, leading to treatment", tabName = "Table_prediction", icon = icon("angle-right")),
               menuSubItem("Prediction result table, recommending treatment", tabName = "Table_prediction_rec", icon = icon("angle-right")),
               menuSubItem("Prediction result table, evidence level A/B", tabName = "Table_prediction_GMT", icon = icon("angle-right")),
               #menuSubItem("Analyze your data", tabName = "Input_data", icon = icon("th")),
               hr()
      ),
      menuItem("Survival after CTx with risk-set adjustment", icon = icon("th"),
               hr(),
               p("Survival analysis"),
               menuSubItem("Survival and treatment after CTx", tabName = "SurvivalandtreatmentafterCTx", icon = icon("angle-right")),
               menuSubItem("Genetic variants and survival, forest plot", tabName = "Geneticvariantsandsurvivalforestplot2", icon = icon("angle-right")),
               menuSubItem("Genetic variants and survival, KM-curve", tabName = "GeneticvariantsandsurvivalKM-curve2", icon = icon("angle-right")),
               menuSubItem("Diagnosis and survival, forest plot", tabName = "Diagnosisandsurvivalforestplot2", icon = icon("angle-right")),
               menuSubItem("Mutational cluster and survival, forest plot", tabName = "DiagnosisandsurvivalKM-curve2", icon = icon("angle-right")),
               menuSubItem("Hazard ratio for survival after CTx", tabName = "HazardratioforsurvivalafterCTx_1", icon = icon("angle-right")),
               hr()
      ),
      menuItem("Survival after CTx with Bayesian inference", icon = icon("th"),
               hr(),
               menuSubItem("Survival corrected for left-truncation bias", tabName = "Survivalcorrectedforleft-truncationbias", icon = icon("angle-right")),
               menuSubItem("Genetic variants and survival, forest plot", tabName = "Geneticvariantsandsurvivalforestplot", icon = icon("angle-right")),
               menuSubItem("Genetic variants and survival, KM-curve", tabName = "GeneticvariantsandsurvivalKM-curve", icon = icon("angle-right")),
               menuSubItem("Diagnosis and survival, forest plot", tabName = "Diagnosisandsurvivalforestplot", icon = icon("angle-right")),
               menuSubItem("Diagnosis and survival, KM-curve", tabName = "DiagnosisandsurvivalKM-curve", icon = icon("angle-right")),
               hr()
      ),
      menuItem("Bias correction simulation", icon = icon("th"),
               hr(),
               menuSubItem("An example of bias adjustment", tabName = "SurvivalSimurationKMCurve", icon = icon("angle-right")),
               hr()
      ),
      menuItem("Drug response", icon = icon("th"),
               hr(),
               p("Tables"),
               menuSubItem("Drug use, by line of treatment", tabName = "Drugusebylineoftreatment", icon = icon("angle-right")),
               menuSubItem("Drug use, by treatment effect", tabName = "Drugusebytreatmenteffect", icon = icon("angle-right")),
               menuSubItem("Use of designated line agents, by mutation pattern", tabName = "Useofdesignatedlineagentsbymutationpattern", icon = icon("angle-right")),
               menuSubItem("Use of designated line agents, by histology", tabName = "Useofdesignatedlineagentsbyhistology", icon = icon("angle-right")),
               menuSubItem("Use of designated line agents, by mutated genes", tabName = "Useofdesignatedlineagentsbymutatedgenes", icon = icon("angle-right")),
               menuSubItem("Use of designated lines and drugs with ToT information, by mutation pattern", tabName = "UseofdesignatedlinesanddrugswithToTinformationbymutationpattern", icon = icon("angle-right")),
               menuSubItem("Use of designated lines and drugs with ToT information, by histology", tabName = "UseofdesignatedlinesanddrugswithToTinformationbyhistology", icon = icon("angle-right")),
               menuSubItem("Use of designated lines and drugs with ToT information, by mutated genes", tabName = "UseofdesignatedlinesanddrugswithToTinformationbymutatedgenes", icon = icon("angle-right")),
               menuSubItem("Use of designated lines and drugs with RECIST information, by mutation pattern", tabName = "UseofdesignatedlinesanddrugswithRECISTinformationbymutationpattern", icon = icon("angle-right")),
               menuSubItem("Use of designated lines and drugs with RECIST information, by histology", tabName = "UseofdesignatedlinesanddrugswithRECISTinformationbyhistology", icon = icon("angle-right")),
               menuSubItem("Use of designated lines and drugs with RECIST information, by mutated genes", tabName = "UseofdesignatedlinesanddrugswithRECISTinformationbymutatedgenes", icon = icon("angle-right")),
               hr(),
               p("Time on treatment"),
               menuSubItem("Time on treatment and pre-treatment for the specified treatment, scatter plot", tabName = "Timeontreatmentandpre-treatmentforthespecifiedtreatmentscatterplot", icon = icon("angle-right")),
               menuSubItem("Time on treatment and pre-treatment for the specified treatment, KM-curve", tabName = "Timeontreatmentandpre-treatmentforthespecifiedtreatmentKM-curve", icon = icon("angle-right")),
               menuSubItem("Time on treatment by tissue type, KM-curve", tabName = "TimeontreatmentbytissuetypeKM-curve", icon = icon("angle-right")),
               menuSubItem("Time on treatment by tissue type, forest plot", tabName = "Timeontreatmentbytissuetypeforestplot", icon = icon("angle-right")),
               menuSubItem("Time on treatment by gene mutation cluster, KM-curve", tabName = "TimeontreatmentbygenemutationclusterKM-curve", icon = icon("angle-right")),
               menuSubItem("Time on treatment by gene mutation cluster, forest plot", tabName = "Timeontreatmentbymutatedclusterforestplot", icon = icon("angle-right")),
               menuSubItem("Time on treatment by mutated genes, forest plot", tabName = "Timeontreatmentbymutatedgenesforestplot", icon = icon("angle-right")),
               menuSubItem("Time on treatment by mutated genes, KM-curve", tabName = "TimeontreatmentbymutatedgenesKM-curve", icon = icon("angle-right")),
               menuSubItem("Time on treatment and mutations of interest, KM-curve", tabName = "TimeontreatmentandmutationsofinterestKM-curve", icon = icon("angle-right")),
               menuSubItem("Hazard ratio on time on treatment - genes", tabName = "Hazardratioontimeontreatment_1", icon = icon("angle-right")),
               menuSubItem("Hazard ratio on time on treatment - clusters", tabName = "Hazardratioontimeontreatment_2", icon = icon("angle-right")),
               menuSubItem("Volcano plot for ToT Hazard ratio, No.1", tabName = "VolcanoPlot_ToT_1", icon = icon("angle-right")),
               menuSubItem("Volcano plot for ToT Hazard ratio, No.2", tabName = "VolcanoPlot_ToT_2", icon = icon("angle-right")),
               menuSubItem("Volcano plot for ToT Hazard ratio, No.3", tabName = "VolcanoPlot_ToT_3", icon = icon("angle-right")),
               menuSubItem("Volcano plot for ToT Hazard ratio, No.4", tabName = "VolcanoPlot_ToT_4", icon = icon("angle-right")),
               menuSubItem("Raw data of volcano plot", tabName = "VolcanoTable_ToT", icon = icon("angle-right")),
               hr(),
               p("Response rate"),
               menuSubItem("Volcano plot for objective response rate, No.1", tabName = "VolcanoPlotORR_1", icon = icon("angle-right")),
               menuSubItem("Volcano plot for objective response rate, No.2", tabName = "VolcanoPlotORR_2", icon = icon("angle-right")),
               menuSubItem("Volcano plot for objective response rate, No.3", tabName = "VolcanoPlotORR_3", icon = icon("angle-right")),
               menuSubItem("Volcano plot for objective response rate, No.4", tabName = "VolcanoPlotORR_4", icon = icon("angle-right")),
               menuSubItem("Raw data of volcano plot", tabName = "VolcanoTableORR", icon = icon("angle-right")),
               menuSubItem("Odds ratio on objective response rate - genes", tabName = "Oddsratioonobjectiveresponserate_1", icon = icon("angle-right")),
               menuSubItem("Odds ratio on objective response rate - clusters", tabName = "Oddsratioonobjectiveresponserate_2", icon = icon("angle-right")),
               menuSubItem("Odds ratio on disease control rate", tabName = "Oddsratioondiseasecontrolrate", icon = icon("angle-right")),
               menuSubItem("Mutation clustering and RECIST", tabName = "MutationclusteringandRECIST", icon = icon("angle-right")),
               menuSubItem("Mutation pattern and RECIST", tabName = "MutationpatternandRECIST", icon = icon("angle-right")),
               menuSubItem("Histology and RECIST", tabName = "HistologyandRECIST", icon = icon("angle-right")),
               menuSubItem("Mutated genes and RECIST", tabName = "MutatedgenesandRECIST", icon = icon("angle-right")),
               hr(),
               p("Survival after CGP"),
               menuSubItem("Survival and drug", tabName = "Survival_drug", icon = icon("angle-right"))
      ),
      hr(),
      h3("Instruction"),
      menuItem("About FELIS", tabName = "Instruction", icon = icon("th")),
      menuItem("Tips", tabName = "Tips", icon = icon("th")),
      hr()
    )
  ),
  ## Body content
  dashboardBody(
    hr(),
    useShinyjs(),
    tags$head(
      tags$style(HTML("
        /* ヘッダー */
        .main-header {
          height: 30px;
          line-height: 30px;
          z-index: 1000;
        }
        .main-header .logo { 
          height: 30px; 
          line-height: 30px; 
        }
        .navbar { 
          min-height: 30px; 
        }
    
        /* ログアウトボタン周り */
        #logout_button {
          transition: background-color 0.3s;
          margin: 0;
          padding: 2px 8px;
          height: 22px;
          font-size: 12px;
          margin-bottom: 0 !important; /* 下の余白をゼロに */
        }
        /* サイドバーの表示/非表示ボタン */
        .sidebar-toggle {
          height: 30px;
          padding-top: 1px !important;
          margin-bottom: 0 !important; /* 下の余白をゼロに */
          padding-bottom: 0 !important; /* ボタンのパディングをゼロに */
          margin-top: 0 !important; /* 上の余白をゼロに */
        }
       /* サイドバー内の文字列を折り返す */
        .sidebar-menu li a {
          white-space: normal;  /* 改行を許可 */
          word-wrap: break-word;  /* 長い単語を折り返す */
          overflow-wrap: break-word; /* 単語が溢れる場合に折り返す */
          line-height: 1.5;  /* 行間を広げて見やすく */
        }
    
        /* サイドバーのメニューの幅が広がったとき */
        .sidebar-menu {
          word-wrap: break-word; /* メニュー全体の折り返し */
        }
    
        /* サイドバー */
        .main-sidebar {
          position: fixed;
          margin-top: 0px;
          left: 0;
          height: calc(100vh - 20px); /* 高さ調整 */
          overflow-y: auto;
          overflow-x: hidden;
          scrollbar-width: none;
          scroll-behavior: smooth;
        }
        .sidebar {
          height: 100%;
          overflow-y: auto;
          overflow-x: hidden;
        }
        .sidebar-menu {
          margin-top: 0px;
        }
    
        /* コンテンツ本体 */
        .content-wrapper, .right-side, .content {
          position: fixed;
          top: 0px;
          margin-top: 30px;
          padding-top: 0px;
          overflow-y: auto;
          height: calc(100vh - 0px);
          width: calc(100% - 230px);
          background-color: #ebf1fc;
          #background-color: #d5e0f5;
        }
    
        /* Bodyのスクロールバー */
        .content-wrapper:hover::-webkit-scrollbar {
          width: 8px;
        }
        .content-wrapper:hover::-webkit-scrollbar-thumb {
          background-color: #888;
          border-radius: 4px;
          border: 2px solid white;
        }
    
        /* ヘッダー部の余白を消す */
        .main-header .navbar .nav > li {
          margin: 0 !important;
          padding: 0 !important;
        }
    
        /* サイドバーのトグルアイコン */
        .main-header .sidebar-toggle:before {
          content: '\\e068';
        }
    
        /* 追加で、ログアウトボタンとサイドバー表示ボタンの余白を完全にゼロに */
        .main-header .navbar-nav {
          margin: 0 !important;
        }
        .main-header .navbar-nav > li {
          margin: 0 !important;
          padding: 0 !important;
        }
      "))
    ),
    tabItems(
      tabItem(tabName = "InputC-CATfiles",
              fluidRow(
                column(4,
                       strong("Required: C-CAT data files (zipped files are acceptable)"),
                       hr(),
                       fileInput(inputId = "clinical_files",
                                 label = "Choose case CSV Files",
                                 multiple = TRUE
                       ),
                       hr(),
                       fileInput(inputId = "report_files",
                                 label = "Choose report CSV Files",
                                 multiple = TRUE
                       ),
                       br(),
                       htmlOutput("select_new_analysis"),
                       h6("If you select No, csv files are not necessary."),
                       htmlOutput("select_intermediate_file"),
                       h6("If you select Yes, faster when performing the same analysis repeatedly."),
                       br(),
                       br(),
                       downloadButton('download_test_clinical_data', 'Download sample clinical data'),
                       br(),
                       br(),
                       downloadButton('download_test_report_data', 'Download sample report data'),
                       br()
                ),
                column(4,
                       strong("Option files"),
                       hr(),
                       fileInput(inputId = "ID_histology",
                                 label = "Correspondence table between ID and histology (CSV)",
                                 multiple = TRUE,
                                 accept = c(
                                   "text/csv",
                                   "text/comma-separated-values,text/plain",
                                   ".csv")
                       ),
                       "Specify the ID of the patient whose diagnosis you want to correct and the modified histology.",
                       br(),
                       br(),
                       downloadButton('download_ID_histology', 'Download CSV file template'),
                       br(),
                       br(),
                       hr(),
                       fileInput(inputId = "ID_drug",
                                 label = "Correspondence table between ID and drug information (CSV)",
                                 multiple = TRUE,
                                 accept = c(
                                   "text/csv",
                                   "text/comma-separated-values,text/plain",
                                   ".csv")
                       ),
                       "To analyze only the drug after curation",
                       br(),
                       br(),
                       downloadButton('download_ID_drug_pre_CGP', 'Download CSV template (before CGP)'),
                       br(),
                       br(),
                       downloadButton('download_ID_drug_post_CGP', 'Download CSV template (after CGP)'),
                       br(),
                       br(),
                       hr(),
                       fileInput(inputId = "drug_rename",
                                 label = "Correspondence table for drug renaming (CSV)",
                                 multiple = TRUE,
                                 accept = c(
                                   "text/csv",
                                   "text/comma-separated-values,text/plain",
                                   ".csv")
                       ),
                       "To analyze similar drugs together by renaming the drugs",
                       br(),
                       "Reclassified as molecular targeted therapies, immune checkpoint inhibitors, etc.",
                       br(),
                       "Drugs are listed in ABC order, separated by commas",
                       br(),
                       br(),
                       downloadButton('download_drug_rename', 'Download CSV template'),
                       br(),
                       br(),
                       fileInput(inputId = "drug_combination_rename",
                                 label = "Correspondence table for drug combination renaming (CSV)",
                                 multiple = TRUE,
                                 accept = c(
                                   "text/csv",
                                   "text/comma-separated-values,text/plain",
                                   ".csv")
                       ),
                       "To rename drug combinations to groups",
                       br(),
                       "Reclassified as molecular targeted therapies, immune checkpoint inhibitors, etc.",
                       br(),
                       "Drugs are listed in ABC order, separated by commas",
                       br(),
                       br(),
                       downloadButton('download_drug_combination_rename', 'Download CSV template'),
                       br(),
                       br()
                ),
                column(4,
                       strong("Option files"),
                       hr(),
                       fileInput(inputId = "mutation_rename",
                                 label = "Correspondence table for mutation renaming (CSV)",
                                 multiple = TRUE,
                                 accept = c(
                                   "text/csv",
                                   "text/comma-separated-values,text/plain",
                                   ".csv")
                       ),
                       "To analyze similar mutations together by renaming mutations",
                       br(),
                       "Reclassified as Exon 19 mutation, Exon 20 mutation, gene amplification, etc.",
                       br(),
                       "'Other' if there is an unspecified mutation in the designated gene",
                       br(),
                       br(),
                       downloadButton('download_mutation_rename', 'Download CSV template'),
                       br(),
                       br(),
                       hr(),
                       fileInput(inputId = "reaanotation",
                                 label = "Correspondence table for mutation reannotation (CSV)",
                                 multiple = TRUE,
                                 accept = c(
                                   "text/csv",
                                   "text/comma-separated-values,text/plain",
                                   ".csv")
                       ),
                       "To reannotate variants",
                       br(),
                       "F: pathogenic variants, G: neutral variants",
                       br(),
                       br(),
                       downloadButton('download_reannotation', 'Download CSV template'),
                       br(),
                       br(),
                       hr(),
                       fileInput(inputId = "histology_rename",
                                 label = "Correspondence table of histological type renaming (CSV)",
                                 multiple = TRUE,
                                 accept = c(
                                   "text/csv",
                                   "text/comma-separated-values,text/plain",
                                   ".csv")
                       ),
                       "To analyze similar tissue types together by renaming them",
                       br(),
                       "Reclassified as differentiated gastric cancer, undifferentiated gastric cancer, etc.",
                       br(),
                       br(),
                       downloadButton('download_histology_rename', 'Download CSV template'),
                       br(),
                       br(),
                       hr(),
                       fileInput(inputId = "regimen_rename",
                                 label = "Correspondence table for drug renaming based on regimen (CSV)",
                                 multiple = TRUE,
                                 accept = c(
                                   "text/csv",
                                   "text/comma-separated-values,text/plain",
                                   ".csv")
                       ),
                       "Provide the drug name if the drug name is unknown and the regimen is known.",
                       br(),
                       "Conversion from MAP therapy to 'Cisplatin,Doxorubicin,Methotrexate'",
                       br(),
                       br(),
                       downloadButton('download_regimen_rename', 'Download CSV template'),
                       br(),
                       br()
                ),
              ),
              hr(),
              h6("FELIS; Functions Especially for LIquid and Solid tumor clinical sequencing."),
              a(href="https://github.com/MANO-B/FELIS",h6("https://github.com/MANO-B/FELIS"))
      ),
      tabItem(tabName = "Setting",
              actionButton("start_setting", "Start file loading/analysis settings"),
              br(),
              hr(),
              fluidRow(
                column(3,strong("Filter on histology"),
                       hr(),
                       htmlOutput("select_organ"),
                       htmlOutput("select_histology"),
                       htmlOutput("select_histology_group_1"),
                       htmlOutput("select_histology_group_1_name"),
                       htmlOutput("select_histology_group_2"),
                       htmlOutput("select_histology_group_2_name"),
                       htmlOutput("select_histology_group_3"),
                       htmlOutput("select_histology_group_3_name"),
                       htmlOutput("select_histology_group_4"),
                       htmlOutput("select_histology_group_4_name"),
                       htmlOutput("select_minimum_pts"),
                       htmlOutput("select_histology_detail"),
                       br()
                ),
                column(3,strong("Filters for clinical information"),
                       hr(),
                       htmlOutput("select_sex"),
                       h5("女: Female, 男: Male, 未入力・不明: Unknown"),
                       br(),
                       htmlOutput("select_panel"),
                       htmlOutput("select_age"),
                       htmlOutput("select_mid_age"),
                       htmlOutput("select_PS"),
                       h5("不明: Unknown"),
                       br(),
                       htmlOutput("select_smoking"),
                       h5("あり: Yes, なし: No, 不明: Unknown"),
                       br(),
                       htmlOutput("select_year"),
                       br()
                ),
                column(3,strong("Filters on genes"),
                       hr(),
                       htmlOutput("select_TMB_threshold"),
                       htmlOutput("select_gene"),
                       htmlOutput("select_gene_group_1"),
                       htmlOutput("select_gene_group_2"),
                       htmlOutput("select_gene_group_analysis"),
                       htmlOutput("select_lolliplot_gene"),
                       htmlOutput("select_lolliplot_no"),
                       br(),
                       br(),
                       strong("Filter on mutation types"),
                       hr(),
                       strong("For detailed study of mutations of a gene"),
                       br(),
                       htmlOutput("select_special_gene"),
                       htmlOutput("select_special_gene_mutation_1"),
                       htmlOutput("select_special_gene_mutation_1_name"),
                       htmlOutput("select_special_gene_mutation_2"),
                       htmlOutput("select_special_gene_mutation_2_name"),
                       htmlOutput("select_special_gene_annotation"),
                       htmlOutput("select_special_gene_independent"),
                       br()
                ),
                column(3,strong("Other Settings"),
                       hr(),
                       htmlOutput("select_gene_no"),
                       htmlOutput("select_oncoprint_option"),
                       htmlOutput("select_patho"),
                       htmlOutput("select_T_N"),
                       htmlOutput("select_fusion"),
                       htmlOutput("select_amplification"),
                       htmlOutput("select_eps"),
                       htmlOutput("select_RMST_CGP"),
                       htmlOutput("select_RMST_CTx"),
                       htmlOutput("select_RMST_drug"),
                       htmlOutput("select_target_line"),
                       htmlOutput("select_MSI"),
                       htmlOutput("select_MMR"),
                       htmlOutput("select_HER2"),
                       htmlOutput("select_machine_learning"),
                       htmlOutput("select_importance"),
                       htmlOutput("select_clustering")
                ),
              )
      ),
      tabItem(tabName = "Analysis",
              fluidRow(
                column(6,
                       strong("Standard Analysis"),
                       hr(),
                       actionButton("summary_base", "Case summary"),
                       br(),
                       br(),
                       actionButton("inter_panel", "Inter-panel comparison"),
                       br(),
                       br(),
                       actionButton("oncoprint", "Oncoprint"),
                       br(),
                       br(),
                       actionButton("mutually_exclusive", "Mutually exclusive or co-occurring mutation"),
                       br(),
                       br(),
                       actionButton("mut_subtype", "Mutation rate of each gene for each histology"),
                       br(),
                       br(),
                       actionButton("custering_analysis", "Clustering based on variants"),
                       br(),
                       br(),
                       actionButton("survival_CGP_analysis", "Survival after CGP test"),
                       br(),
                       br(),
                       actionButton("CGP_benefit_analysis", "CGP benefit prediction analysis"),
                       br(),
                       br(),
                       actionButton("survival_CTx_analysis2", "Survival after CTx induction, risk-set adjustment"),
                       br(),
                       h6("  Take care of left-truncation bias."),
                       br(),
                       actionButton("survival_CTx_analysis", "Survival after CTx induction, Bayesian inference"),
                       br(),
                       h6("  It takes minutes."),
                       br()
                ),
                  column(6,
                       strong("Drug response analysis"),
                       hr(),
                       actionButton("drug_table", "List of drugs used in Palliative CTx (1st-4th line)"),
                       br(),
                       br(),
                       hr(),
                       htmlOutput("select_drug"),
                       br(),
                       htmlOutput("select_drug_other"),
                       br(),
                       htmlOutput("select_drug_group_1"),
                       br(),
                       htmlOutput("select_drug_group_1_name"),
                       br(),
                       htmlOutput("select_drug_group_2"),
                       br(),
                       htmlOutput("select_drug_group_2_name"),
                       br(),
                       htmlOutput("select_drug_group_3"),
                       br(),
                       htmlOutput("select_drug_group_3_name"),
                       br(),
                       htmlOutput("select_drug_group_4"),
                       br(),
                       htmlOutput("select_drug_group_4_name"),
                       br(),
                       hr(),
                       actionButton("drug_analysis", "Analyze with the setting selected above"),
                       br(),
                       htmlOutput("select_drug_volcano"),
                       br(),
                       htmlOutput("select_volcano_pathology"),
                       br(),
                       htmlOutput("select_drug_survuval_CTx"),
                       br()
                )
              ),
              hr(),
              h5("Figures in results are downloadable as png files.")
      ),
      tabItem(tabName = "Summarizedbymutationpattern",
              gt_output("table_summary_1"),
                       hr(),
                       h6(paste("Table. Characteristics for selected patients.",
                                "The present, retrospective cohort study was performed with clinicogenomic,",
                                "real-world data on the patients who were registered in the C-CAT database",
                                "from June 1, 2019. The patients were registered by hospitals throughout Japan",
                                "and provided written informed consent to the secondary use of their clinicogenomic",
                                "data for research."))
      ),
      tabItem(tabName = "Summarizedbyhistology",
              gt_output("table_summary_2"),
              hr(),
              DT::dataTableOutput("table_summary_3"),
              hr(),
              h6("Table. Characteristics for selected patients.")
      ),
      tabItem(tabName = "Inter_panel",
              plotOutput('figure_inter_panel', 
                         height = "2000px",
                         width = "1000px"),
              hr(),
              h6("Figure. Comparison of panels. (Upper) Distribution of variant allele frequency for each panel. (Lower) No figures to draw yet.")
      ),
      tabItem(tabName = "Table_inter_panel",
              DT::dataTableOutput("table_inter_panel"),
              hr(),
              h6("Pts: patients, TMB: tumor mutational burden (muts/Mb), AAC: amino-acid change (muts/patients), PG: pathogenic mutations (muts/patients)."),
              h6("F1: FoundationOne, NOP: NCC OncoPanel, F1L: FoundationOne Liquid, GMT: GenMine TOP panel")
      ),
      tabItem(tabName = "Oncoprint",
              plotOutput('figure_oncoprint', 
                         height = "1000px",
                         width = "1000px"),
              hr(),
              h6("Figure. Recurrent oncogenic mutations in selected cases. The 30 genes with the highest frequency of oncogenic mutations are shown. Mutational landscapes were created using ComplexHeatmap package for R.")
      ),
      tabItem(tabName = "Lolliplotfortheselectedgene",
              plotOutput('figure_lolliplot2', 
                                  height = "500px",
                                  width = "1500px"),
                       hr(),
                       h6("Figure. Frequency of oncogenic mutations in the selected gene. The most frequent oncogenic mutations are shown with amino acid change."),
                       h6("Mutplot by Zhang W, PMID:31091262. If error occurs, correct 'source/UniPlot.txt'."),
                       h6("Protein structure source: Uniprot"),
                       HTML("<p>Github for Mutplot. <a href='https://github.com/VivianBailey/Mutplot'>Link for the website</a></p>")
      ),
      tabItem(tabName = "Tableofclinicalandmutationinformationperpatient",
              DT::dataTableOutput("table_patient")
      ),
      tabItem(tabName = "Mutualexclusivity",
              plotOutput('figure_mutually_exclusive', 
                                  height = "2000px",
                                  width = "1000px"),
                       hr(),
              DT::dataTableOutput("table_mutually_exclusive"),
              h6("Figure. Alterations among mutually exclusive or co-occurring pairs. The 30 genes with the highest frequency of oncogenic mutations were selected to determine whether oncogenic mutations are likely to occur simultaneously between the two genes. Blue boxes indicates mutually exclusivity and red boxes indicates co-occurrence. An asterisk shows a significant correlation (p < 0.001). Analysis was performed with Rediscover package in R language. Odds ratios were estimated by Fisher exact test.")
      ),
      tabItem(tabName = "Variationbyhistology",
              plotOutput('figure_mut_subtype', 
                                  height = "1000px",
                                  width = "1000px"),
                       hr(),
                       h6("Figure. Recurrent oncogenic mutations across subtypes. The 30 genes with the highest frequency of oncogenic mutations were displayed.")
      ),
      tabItem("Basicdata",
              fluidRow(plotOutput('figure_base', 
                                  height = "1500px",
                                  width = "1000px"),
                       hr(),
                       h6("Figure. Distribution of age, sex, detected oncogenic mutations, tumor mutation burden (TMB), metastasis pattern, patients with treatment option recommended by the expert panel, patients received recommended chemotherapy, mediantime from the initiation date of the first palliative chemotherapy to CGP, and median time from CGP to final observation. In the boxplots of age and TMB, the box borders indicate the 25th and 75th percentiles, the inner line the median, and the whiskers 1.5× the interquartile range.")
              )
      ),
      tabItem("UMAPclusteringbasedonmutations",
              plotOutput('figure_cluster', 
                         height = "1000px",
                         width = "1200px"),
              hr(),
              # plotOutput('figure_cluster_2', 
              #            height = "1000px",
              #            width = "4000px"),
              h6("Figure. Unsupervised clustering of the patients based on the detected oncogenic mutations. Two-dimensional mutational pattern mapping was generated using Uniform Manifold Approximation and Projection (UMAP). The three variants and histotypes with the highest odds ratios that were more common than the other clusters at p<0.05. Clustering analysis was performed as follows. All pathogenic mutations detected by the cancer-related genes were assembled into a binary matrix format per patient. The dimension of this input matrix was reduced using Uniform Manifold Approximation and Projection (UMAP) via the umap package for R (with default hyperparameters). Clustering analysis was performed using the Hierarchical Density-Based Spatial Clustering of Applications with Noise (HDBSCAN) via the dbscan package for R (EPS: 1.0; minimum points: 3)."),
              HTML("<p>Mochizuki T, et al., Factors predictive of second-line chemotherapy in soft tissue sarcoma: An analysis of the National Genomic Profiling Database. Cancer Science, 2023. <a href='https://doi.org/10.1111/cas.16050'>Link for the paper</a></p>")
      ),
      tabItem("Clusterandagerelationship",
              plotOutput('figure_cluster_age', 
                         height = "1000px",
                         width = "1200px")
      ),
      tabItem("Clusterandhistologyrelationship",
              plotOutput('figure_cluster_subtype', 
                         height = "1000px",
                         width = "1200px")
      ),
      tabItem("Heterogeneitywithinhistologictypes",
              plotOutput('figure_entropy', 
                         height = "1000px",
                         width = "1500px"),
              hr(),
              h6("Figure. Unsupervised clustering based on oncogenic mutations. For each histology, the percentage of cases belonging to one of the clusters is shown as a bar graph. Characteristic oncogenic mutations were found in each cluster. There was a tendency for each histologic type to cluster in specific clusters. Heterogeneity of genetic variation within histologic types was assessed by Shannon's entropy.")
      ),
      tabItem("Table_basic_data",
              DT::dataTableOutput("table_basic_data")
      ),
      tabItem("Tableofclustersandhistologies",
              DT::dataTableOutput("table_disease")
      ),
      tabItem("Tableofclustersandgeneticvariants",
              DT::dataTableOutput("table_mutation")
      ),
      tabItem("Frequencyofpatientswithtargetedtherapy",
              plotOutput('figure_drug_evidence', 
                         height = "1000px",
                         width = "1000px"),
              hr(),
              DT::dataTableOutput("table_drug_evidence"),
              DT::dataTableOutput("table_evidence"),
              h6("Figure. Level of evidence for targeted therapy for detected gene mutations. The highest level of evidence was extracted for each patient. Evidence levels of C-CAT are defined as A for biomarkers that predict a response to Japanese Pharmaceuticals and Medical Devices Agency (PMDA)– or FDA-approved therapies or are described in professional guidelines, B for biomarkers that predict a response based on well-powered studies with consensus of experts in the field, C for biomarkers that predict a response to therapies approved by the PMDA or FDA in another type of tumor or that predict a response based on clinical studies, D for biomarkers that predict a response based on case reports, E for biomarkers that show plausible therapeutic significance based on preclinical studies.")
      ),
      tabItem("Relationshipbetweenpre-CGPandpost-CGPperiods",
              plotOutput('figure_CGP_pre_post', 
                         height = "1000px",
                         width = "1000px"),
              hr(),
              h6("Figure. Relationship between pre-CGP and post-CGP periods. Pearson's product-rate correlation coefficients, P-values, and regression lines were calculated in the R language.")
      ),
      tabItem("Patientsperhistologyandtreatmentreachrate",
              plotOutput('figure_patient_treatment', 
                         height = "2000px",
                         width = "2000px"),
              hr(),
              h6("Figure. Patients per histology and treatment reach rate. Pearson's product-rate correlation coefficients, P-values, and regression lines were calculated in the R language.")
              
      ),
      tabItem("Pre-CGPperiodandtreatmentreachrate",
              plotOutput('figure_Pre_CGP_treatment', 
                         height = "2000px",
                         width = "2000px"),
              hr(),
              h6("Figure. Pre-CGP period and treatment reach rate. Pearson's product-rate correlation coefficients, P-values, and regression lines were calculated in the R language.")
              
      ),
      tabItem("Post-CGPperiodandtreatmentreachrate",
              plotOutput('figure_Post_CGP_treatment', 
                         height = "2000px",
                         width = "2000px"),
              hr(),
              h6("Figure. Post-CGP period and treatment reach rate. Pearson's product-rate correlation coefficients, P-values, and regression lines were calculated in the R language.")
              
      ),
      tabItem("Ageandtreatmentreachrate",
              plotOutput('figure_Age_and_treatment', 
                         height = "2000px",
                         width = "2000px"),
              hr(),
              h6("Figure. Age and treatment reach rate. Pearson's product-rate correlation coefficients, P-values, and regression lines were calculated in the R language.")
      ),
      tabItem("SurvivalandtreatmentafterCGP",
              plotOutput('figure_survival_CGP_1', 
                         height = "2500px",
                         width = "1500px"),
              hr(),
              h6("Figure. Survival analysis using the conventional Kaplan–Meier estimator, log–rank test, and the Kaplan–Meier estimator with number-at-risk adjustment were undertaken with survival package for R. EP: expert panel.")
      ),
      tabItem("SurvivalafterCGPandperformancestatus",
              plotOutput('figure_survival_CGP_2', 
                         height = "3200px",
                         width = "1500px")
      ),
      tabItem("SurvivalafterCGPandprevioustreatment",
              plotOutput('figure_survival_CGP_CTx', 
                         height = "2500px",
                         width = "1500px")
      ),
      tabItem("SurvivalafterCGPandmutationsforestplot",
              plotOutput('figure_survival_CGP_3', 
                         height = "1000px",
                         width = "1200px"),
              hr(),
              h6("Figure. Suvival periods after CGP and gene mutations estimated with conventional Kaplan-Meier estimator. Restricted mean survival time in two years were estimated with survRM2 package in R."),
              DT::dataTableOutput("figure_survival_CGP_3_data")
      ),
      tabItem("SurvivalafterCGPandmutationsKM-curve",
              plotOutput('figure_survival_CGP_4', 
                         height = "3000px",
                         width = "1500px")
      ),
      tabItem("HazardratioforsurvivalafterCGP_1",
              gt_output('figure_survival_CGP_5_1'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity.")
      ),
      tabItem("HazardratioforsurvivalafterCGP_2",
              gt_output('figure_survival_CGP_5_2'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity.")
      ),
      tabItem("Survival_treatment_reach",
              plotOutput('figure_survival_treatment_reach', 
                         height = "2000px",
                         width = "1000px"),
              hr(),
              h6("Figure. Treatment reach rate and survival period after CGP. Only deceased (not censored) patients were analyzed. The moving average curve was drawn using the geom_smooth function, with the following settings; method = 'loess', span = 0.75, and se = F."),
              DT::dataTableOutput("table_survival_treatment_reach"),
              downloadButton('download_treatment_reach_data', 'Download treatment reach data'),
      ),
      tabItem("FactorsleadingtoTreatmentpre-CGPNomogram",
              plotOutput('figure_nomogram_treat_pre_CGP', 
                         height = "2000px",
                         width = "1000px"),
              hr(),
              verbatimTextOutput("histology_nomogram_pre_CGP"),
              hr(),
              h6("Figure. Based on clinical information obtained prior to CGP testing, a nomogram was developed to predict the likelihood of actually receiving treatment based on the genetic mutations after CGP. The nomogram was created with the lrm function of the rms package for R with a setting of penalty=1. Nagelkerke R2 was calculated with blorr package for R. Possible sampling bias was corrected with 500-time bootstrap sampling and then concordance index was estimated. Best_effect: the best treatment effect of CTx before CGP.")
      ),
      tabItem("FactorsrecommendingTreatmentpre-CGPNomogram",
              plotOutput('figure_nomogram_treat_pre_CGP_rec', 
                         height = "2000px",
                         width = "1000px"),
              hr(),
              verbatimTextOutput("histology_nomogram_pre_CGP_rec"),
              hr(),
              h6("Figure. Based on clinical information obtained prior to CGP testing, a nomogram was developed to predict the likelihood of actually receiving treatment based on the genetic mutations after CGP. The nomogram was created with the lrm function of the rms package for R with a setting of penalty=1. Nagelkerke R2 was calculated with blorr package for R. Possible sampling bias was corrected with 500-time bootstrap sampling and then concordance index was estimated. Best_effect: the best treatment effect of CTx before CGP.")
      ),
      tabItem("FactorsrecommendingTreatmentABpre-CGPNomogram",
              plotOutput('figure_nomogram_treat_pre_CGP_GMT', 
                         height = "2000px",
                         width = "1000px"),
              hr(),
              verbatimTextOutput("histology_nomogram_pre_CGP_GMT"),
              hr(),
              h6("Figure. Based on clinical information obtained prior to CGP testing, a nomogram was developed to predict the likelihood of actually receiving treatment based on the genetic mutations after CGP. The nomogram was created with the lrm function of the rms package for R with a setting of penalty=1. Nagelkerke R2 was calculated with blorr package for R. Possible sampling bias was corrected with 500-time bootstrap sampling and then concordance index was estimated. Best_effect: the best treatment effect of CTx before CGP.")
      ),
      tabItem("FactorsleadingtoTreatmentpost-CGPNomogram",
              plotOutput('figure_nomogram_treat_post_CGP', 
                         height = "2000px",
                         width = "1000px"),
              hr(),
              verbatimTextOutput("histology_nomogram_post_CGP"),
              hr(),
              h6("Figure. Based on clinical and genetic information, a nomogram was developed to predict the likelihood of actually receiving treatment for patients with treatment options recommended by the expert panel. The nomogram was created with the lrm function of the rms package for R with a setting of penalty=1. Nagelkerke R2 was calculated with blorr package for R. Possible sampling bias was corrected with 500-time bootstrap sampling and then concordance index was estimated. Best_effect: the best treatment effect of CTx before CGP.")
      ),
      tabItem("FactorsleadingtoTreatmentpre-CGPOddsratio",
              gt_output('figure_survival_CGP_7'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors selected in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity.")
      ),
      tabItem("FactorsrecommendingTreatmentpre-CGPOddsratio",
              gt_output('figure_survival_CGP_7_rec'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors selected in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity.")
      ),
      tabItem("FactorsrecommendingTreatmentABpre-CGPOddsratio",
              gt_output('figure_survival_CGP_7_GMT'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors selected in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity.")
      ),
      tabItem("FactorsleadingtoTreatmentpost-CGPOddsratio",
              gt_output('figure_survival_CGP_6'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors selected in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity.")
      ),
      tabItem("FactorsleadingtoTreatmentdecisioncurve",
              plotOutput('figure_DCA_pre_CGP_1', 
                         height = "1000px",
                         width = "1300px"),
              hr(),
              h6("Figure. Decision curve analysis was performed to verify the usefulness of the nomogram with dcurves package for R. Ten-fold cross-validation was performed to prevent overfitting. If the blue line is located above the other lines, then the nomogram-based decision to perform CGP testing may be worthwhile.")
      ),
      tabItem("FactorsrecommendingTreatmentdecisioncurve",
              plotOutput('figure_DCA_pre_CGP_1_rec', 
                         height = "1000px",
                         width = "1300px"),
              hr(),
              h6("Figure. Decision curve analysis was performed to verify the usefulness of the nomogram with dcurves package for R. Ten-fold cross-validation was performed to prevent overfitting. If the blue line is located above the other lines, then the nomogram-based decision to perform CGP testing may be worthwhile.")
      ),
      tabItem("FactorsrecommendingTreatmentABdecisioncurve",
              plotOutput('figure_DCA_pre_CGP_1_GMT', 
                         height = "1000px",
                         width = "1300px"),
              hr(),
              h6("Figure. Decision curve analysis was performed to verify the usefulness of the nomogram with dcurves package for R. Ten-fold cross-validation was performed to prevent overfitting. If the blue line is located above the other lines, then the nomogram-based decision to perform CGP testing may be worthwhile.")
      ),
      tabItem("ROCnomogram",
              plotOutput('figure_DCA_pre_CGP_ROC', 
                         height = "3000px",
                         width = "1000px"),
              hr(),
              h6("Figure. Predicted treatment reach rate and Receiver Operatorating Characteristic curve of the nomogram using pre-CGP information by pROC package for R. The nomogram was based on logistic regression analysis, random forest model, and LightGBM model, all of which calculated sensitivity and specificity using prediction results from 5-fold cross-validation and plotted ROC curves. The random forest model and LightGBM model used 5-fold cross-validation on a single training set and performed a grid search with n=8 for each parameter to determine the optimal parameters.")
      ),
      tabItem("ROCnomogram_rec",
              plotOutput('figure_DCA_pre_CGP_ROC_rec', 
                         height = "3000px",
                         width = "1000px"),
              hr(),
              h6("Figure. Predicted treatment recommendation rate and Receiver Operatorating Characteristic curve of the nomogram using pre-CGP information by pROC package for R. The nomogram was based on logistic regression analysis, random forest model, and LightGBM model, all of which calculated sensitivity and specificity using prediction results from 5-fold cross-validation and plotted ROC curves. The random forest model and LightGBM model used 5-fold cross-validation on a single training set and performed a grid search with n=8 for each parameter to determine the optimal parameters.")
      ),
      tabItem("ROCnomogram_GMT",
              plotOutput('figure_DCA_pre_CGP_ROC_GMT', 
                         height = "3000px",
                         width = "1000px"),
              hr(),
              h6("Figure. Predicted treatment recommendation rate and Receiver Operatorating Characteristic curve of the nomogram using pre-CGP information by pROC package for R. The nomogram was based on logistic regression analysis, random forest model, and LightGBM model, all of which calculated sensitivity and specificity using prediction results from 5-fold cross-validation and plotted ROC curves. The random forest model and LightGBM model used 5-fold cross-validation on a single training set and performed a grid search with n=8 for each parameter to determine the optimal parameters.")
      ),
      tabItem("preCGPMachinelearning",
              plotOutput('figure_DCA_pre_CGP_Machine_learning', 
                         height = "2000px",
                         width = "1500px"),
              hr(),
              h6("Figure. Predicted treatment reach rate with Random forest and LightGBM models using pre-CGP information by tidymodels package for R. Generalized additive model was used for calibration. 5-fold cross validation was performed. SHAP summary barplot represents the mean absolute SHAP values across the cases in the holdout test set, while boxplot represent the distribution of individual feature values. SHAP values were calculated with fastshap package with 5-time Monte Carlo simulation.")
      ),
      tabItem("preCGPMachinelearning_rec",
              plotOutput('figure_DCA_pre_CGP_Machine_learning_rec', 
                         height = "2000px",
                         width = "1500px"),
              hr(),
              h6("Figure. Predicted treatment recommendation rate with Random forest and LightGBM models using pre-CGP information by tidymodels package for R. Generalized additive model was used for calibration. 5-fold cross validation was performed. SHAP summary barplot represents the mean absolute SHAP values across the cases in the holdout test set, while boxplot represent the distribution of individual feature values. SHAP values were calculated with fastshap package with 5-time Monte Carlo simulation.")
      ),
      tabItem("preCGPMachinelearning_GMT",
              plotOutput('figure_DCA_pre_CGP_Machine_learning_GMT', 
                         height = "2000px",
                         width = "1500px"),
              hr(),
              h6("Figure. Predicted treatment recommendation rate (variants annotated as evidence level A or B) with Random forest and LightGBM models using pre-CGP information by tidymodels package for R. Generalized additive model was used for calibration. 5-fold cross validation was performed. SHAP summary barplot represents the mean absolute SHAP values across the cases in the holdout test set, while boxplot represent the distribution of individual feature values. SHAP values were calculated with fastshap package with 5-time Monte Carlo simulation.")
      ),
      tabItem("FactorsleadingtoTreatmenttable",
              gt_output('table_DCA_pre_CGP_2')
      ),
      tabItem("FactorsleadingtoTreatmenttable_rec",
              gt_output('table_DCA_pre_CGP_2_rec')
      ),
      tabItem("FactorsleadingtoTreatmenttable_GMT",
              gt_output('table_DCA_pre_CGP_2_GMT')
      ),
      tabItem(tabName = "Table_prediction",
              DT::dataTableOutput("table_prediction")
      ),
      tabItem(tabName = "Table_prediction_rec",
              DT::dataTableOutput("table_prediction_rec")
      ),
      tabItem(tabName = "Table_prediction_GMT",
              DT::dataTableOutput("table_prediction_GMT")
      ),
      tabItem(tabName = "Input_data",
              actionButton("predict_nomogram_setting", "Start prediction setting"),
              br(),
              hr(),
              fluidRow(
                column(4,strong("Clinical information"),
                       hr(),
                       htmlOutput("input_age"),
                       htmlOutput("input_sex"),
                       htmlOutput("input_PS"),
                       htmlOutput("input_Lines"),
                       htmlOutput("input_Best_effect"),
                       htmlOutput("input_organ"),
                       htmlOutput("input_Panel"),
                       br(),
                       hr()
                ),
                column(4,strong("Metastasis information"),
                       hr(),
                       htmlOutput("input_Lymph_met"),
                       htmlOutput("input_Brain_met"),
                       htmlOutput("input_Lung_met"),
                       htmlOutput("input_Bone_met"),
                       htmlOutput("input_Liver_met")
                ),
              ),
              br(),
              hr(),
              br(),
              actionButton("start_prediction", "Start prediction"),
              hr(),
              verbatimTextOutput("prediction_nomogram"),
              br(),
              br(),
              hr()
      ),
      tabItem("SurvivalandtreatmentafterCTx",
              plotOutput('figure_survival_CTx_1_2', 
                         height = "4000px",
                         width = "4000px"),
              hr(),
              h6("Figure. Overall survival after the first survival-prolonging chemotherapy after adjusting for left-truncation bias. To evaluate the association between prognostic factors and survival, a risk-set adjustment model was applied to adjust for left-truncation bias with survival package. Note that the analysis assumes independent left-truncation (conditional Kendall tau = 0).")
      ),
      tabItem("Geneticvariantsandsurvivalforestplot2",
              plotOutput('figure_survival_CTx_2_2', 
                         height = "1000px",
                         width = "1000px"),
              hr(),
              DT::dataTableOutput("figure_survival_CTx_2_data_2")
      ),
      tabItem("GeneticvariantsandsurvivalKM-curve2",
              plotOutput('figure_survival_CTx_3_2', 
                         height = "3600px",
                         width = "1500px"),
              hr(),
              h6("Figure. Overall survival after the first survival-prolonging chemotherapy after adjusting for left-truncation bias. To evaluate the association between oncogenic mutations and survival, a risk-set adjustment model was performed to adjust for left-truncation bias with survival package.")
      ),
      tabItem("Diagnosisandsurvivalforestplot2",
              plotOutput('figure_survival_CTx_4_2', 
                         height = "1200px",
                         width = "1000px"),
              hr()
      ),
      tabItem("DiagnosisandsurvivalKM-curve2",
              plotOutput('figure_survival_CTx_5_2', 
                         height = "1200px",
                         width = "1000px"),
              hr(),
              h6("Figure. Overall survival after the first survival-prolonging chemotherapy after adjusting for left-truncation bias. To evaluate the association between oncogenic mutations and survival, a risk-set adjustment model was performed to adjust for left-truncation bias with survival package.")
      ),
      tabItem("HazardratioforsurvivalafterCTx_1",
              plotOutput('figure_survival_CTx_5_1_2', 
                         height = "1200px",
                         width = "1000px"),
              hr(),
              h6("Figure. Hazard ratio estimated by cox model with survival package.")
      ),
      tabItem("Survivalcorrectedforleft-truncationbias",
              plotOutput('figure_survival_CTx_1', 
                         height = "2500px",
                         width = "1000px"),
              hr(),
              h6(paste0("Figure. Overall survival after the first survival-prolonging chemotherapy after adjusting for left-truncation bias. To evaluate the association between oncogenic mutations and survival, a Bayesian survival simulation based on a semi-independent, two-hit model was performed to adjust for left-truncation bias. ",
                        "Two survival curves from the date of commencement of chemotherapy for prolonging survival to the date of CGP and from the CGP testing date to the date of death were fitted with Weibull distribution and log-logistic distribution, respectively. The overall survival curve from the first chemotherapy induction was approximated by merging these survival curves. ",
                        "Survival curves were obtained from each of the 8000 iterations of inference, and the median survival and 95% equal-tailed CIs were calculated. Bayesian inference was performed with the rstan package for R. P value of conditional Kendall tau statistics was calculated, and the survival curves were adjusted for length bias, using a structural transformation method with tranSurv package for R.")),
              HTML("<p>Tamura T, et al., Selection bias due to delayed comprehensive genomic profiling in Japan. Cancer Science, 2022. <a href='https://doi.org/10.1111/cas.15651'>Link for the paper</a></p>")
      ),
      tabItem("Geneticvariantsandsurvivalforestplot",
              plotOutput('figure_survival_CTx_2', 
                         height = "1000px",
                         width = "1000px"),
              hr(),
              DT::dataTableOutput("figure_survival_CTx_2_data")
      ),
      tabItem("GeneticvariantsandsurvivalKM-curve",
              plotOutput('figure_survival_CTx_3', 
                         height = "3600px",
                         width = "1500px"),
              hr(),
              h6("Figure. Overall survival after the first survival-prolonging chemotherapy after adjusting for left-truncation bias. To evaluate the association between oncogenic mutations and survival, a Bayesian survival simulation based on a semi-independent, two-hit model was performed to adjust for left-truncation bias. Two survival curves from the date of commencement of chemotherapy for prolonging survival to the date of CGP and from the CGP testing date to the date of death were fitted with Weibull distribution and log-logistic distribution, respectively. The overall survival curve from the first chemotherapy induction was approximated by merging these survival curves. Survival curves were obtained from each of the 8000 iterations of inference, and the median survival and 95% equal-tailed CIs were calculated. Bayesian inference was performed with the rstan package for R."),
              HTML("<p>Tamura T, et al., Selection bias due to delayed comprehensive genomic profiling in Japan. Cancer Science, 2022. <a href='https://doi.org/10.1111/cas.15651'>Link for the paper</a></p>")
      ),
      tabItem("Diagnosisandsurvivalforestplot",
              plotOutput('figure_survival_CTx_4', 
                         height = "600px",
                         width = "1000px"),
              hr(),
              DT::dataTableOutput("figure_survival_CTx_4_data")
      ),
      tabItem("DiagnosisandsurvivalKM-curve",
              plotOutput('figure_survival_CTx_5', 
                         height = "3200px",
                         width = "3000px"),
              hr(),
              DT::dataTableOutput("figure_survival_CTx_5_table"),
              h6("Figure. Overall survival after the first survival-prolonging chemotherapy after adjusting for left-truncation bias. To evaluate the association between oncogenic mutations and survival, a Bayesian survival simulation based on a semi-independent, two-hit model was performed to adjust for left-truncation bias. Two survival curves from the date of commencement of chemotherapy for prolonging survival to the date of CGP and from the CGP testing date to the date of death were fitted with Weibull distribution and log-logistic distribution, respectively. The overall survival curve from the first chemotherapy induction was approximated by merging these survival curves. Survival curves were obtained from each of the 8000 iterations of inference, and the median survival and 95% equal-tailed CIs were calculated. Bayesian inference was performed with the rstan package for R."),
              HTML("<p>Tamura T, et al., Selection bias due to delayed comprehensive genomic profiling in Japan. Cancer Science, 2022. <a href='https://doi.org/10.1111/cas.15651'>Link for the paper</a></p>")
      ),
      tabItem("SurvivalSimurationKMCurve",
              fluidRow(
                column(3,strong("Simulation settings"),
                       hr(),
                       htmlOutput("input_shape_llogis"),
                       htmlOutput("input_scale_llogis"),
                       br(),
                       hr()
                ),
                column(3,strong(" "),
                       hr(),
                       htmlOutput("input_shape_CGP"),
                       htmlOutput("input_scale_CGP"),
                       br(),
                       hr()
                ),
                column(3,strong(" "),
                       hr(),
                       htmlOutput("input_shape_weibull"),
                       htmlOutput("input_scale_weibull"),
                       br(),
                       hr()
                ),
                column(3,strong(" "),
                       hr(),
                       htmlOutput("input_censor_rate"),
                       br(),
                       hr()
                )
              ),
              br(),
              hr(),
              actionButton("survival_simuration", "Left-truncation bias adjustment simulation"),
              br(),
              hr(),
              plotOutput('figure_bias_correction_simulation', 
                         height = "1000px",
                         width = "1000px"),
              hr(),
              h6("Figure. Overall survival after the first survival-prolonging chemotherapy. "),
              HTML("<p>Tamura T, et al., Selection bias due to delayed comprehensive genomic profiling in Japan. Cancer Science, 2022. <a href='https://doi.org/10.1111/cas.15651'>Link for the paper</a></p>"),
              br(),
              verbatimTextOutput("text_bias_correction_simulation"),
              br(),
              br(),
              hr()
      ),
      tabItem("Drugusebylineoftreatment",
              gt_output("table_drug_all_1")
      ),
      tabItem("Drugusebytreatmenteffect",
              gt_output("table_drug_all_2")
      ),
      tabItem("Useofdesignatedlineagentsbymutationpattern",
              gt_output("table_drug_line_1")
      ),
      tabItem("Useofdesignatedlineagentsbyhistology",
              gt_output("table_drug_line_2")
      ),
      tabItem("Useofdesignatedlineagentsbymutatedgenes",
              gt_output("table_drug_line_3")
      ),
      tabItem("UseofdesignatedlinesanddrugswithToTinformationbymutationpattern",
              gt_output("table_drug_ToT_1")
      ),
      tabItem("UseofdesignatedlinesanddrugswithToTinformationbyhistology",
              gt_output("table_drug_ToT_2")
      ),
      tabItem("UseofdesignatedlinesanddrugswithToTinformationbymutatedgenes",
              gt_output("table_drug_ToT_3")
      ),
      tabItem("UseofdesignatedlinesanddrugswithRECISTinformationbymutationpattern",
              gt_output("table_drug_RECIST_1")
      ),
      tabItem("UseofdesignatedlinesanddrugswithRECISTinformationbyhistology",
              gt_output("table_drug_RECIST_2")
      ),
      tabItem("UseofdesignatedlinesanddrugswithRECISTinformationbymutatedgenes",
              gt_output("table_drug_RECIST_3")
      ),
      tabItem("Timeontreatmentandpre-treatmentforthespecifiedtreatmentscatterplot",
              plotOutput('figure_drug_1', 
                         height = "2500px",
                         width = "1500px"),
              hr(),
              h6(paste0("Figure. Time on treatment analysis for the survival-prolonging chemotherapy. ",
                        "Time on treatment represents the period from the start date of chemotherapy to the end date; ",
                        "if the patient was on medication at the time of CGP testing, the patient was censored; ",
                        "otherwise, the patient was terminated, and a survival curve was generated using the Kaplan-Meier method.\n",
                        "(Upper) Swimmer plot of the designated treatment and the previous treatment.\n",
                        "(Lower) Scatter plot of the time on treatment of the designated treatment and the previous treatment. Only deceased patients were analyzed.\n"
              ))
      ),
      tabItem("Timeontreatmentandpre-treatmentforthespecifiedtreatmentKM-curve",
              plotOutput('figure_drug_2', 
                         height = "3200px",
                         width = "1500px"),
              hr(),
              h6(paste0("Figure. Time on treatment analysis for the survival-prolonging chemotherapy. ",
                        "Time on treatment represents the period from the start date of chemotherapy to the end date; ",
                        "if the patient was on medication at the time of CGP testing, the patient was censored; ",
                        "otherwise, the patient was terminated, and a survival curve was generated using the Kaplan-Meier method.\n",
                        "(Left-1) Comparison of time on treatment between the indicated treatment and its previous course of treatment in the same patient group.\n",
                        "(Left-2) Time on treatment of all cources.\n",
                        "(Left-3) Time-on-treatment comparison of the all treatments based on the presence or absence of mutations in any of the specified genes.\n",
                        "(Left-4) Comparison of time on treatment between designated treatment and the other treatments.\n",
                        "(Right-1) Comparison of time on treatment grouped by presence or absence of mutations in the specified genes and by specified treatment and other treatments.\n",
                        "(Right-2) Comparison of time on treatment of the specified treatment grouped by the presence or absence of mutations in the specified genes.\n",
                        "(Right-3) Comparison of time on treatment of the specified treatment grouped by the presence or absence of lymph node metastasis.\n"
              ))
      ),
      tabItem("TimeontreatmentbytissuetypeKM-curve",
              plotOutput('figure_drug_3', 
                         height = "3200px",
                         width = "1500px"),
              hr(),
              h6(paste0("Figure. Time on treatment analysis for the survival-prolonging chemotherapy. ",
                        "Time on treatment represents the period from the start date of chemotherapy to the end date; ",
                        "if the patient was on medication at the time of CGP testing, the patient was censored; ",
                        "otherwise, the patient was terminated, and a survival curve was generated using the Kaplan-Meier method.\n",
                        "(Upper) Comparison of time on treatment of the all treatments among the major histology.\n",
                        "(Lower) Comparison of time on treatment of the specified treatment among the major histology.\n"
              )),
              hr(),
              DT::dataTableOutput("figure_drug_3_table")
      ),
      tabItem("Timeontreatmentbytissuetypeforestplot",
              plotOutput('figure_drug_3_forest', 
                         height = "600px",
                         width = "1000px"),
              hr(),
              DT::dataTableOutput("figure_drug_3_forest_table")
      ),
      tabItem("TimeontreatmentbygenemutationclusterKM-curve",
              plotOutput('figure_drug_cluster', 
                         height = "2500px",
                         width = "1500px")
      ),
      tabItem("Timeontreatmentbymutatedclusterforestplot",
              plotOutput('figure_drug_cluster_2', 
                         height = "600px",
                         width = "1000px"),
              hr(),
              DT::dataTableOutput("figure_drug_cluster_table")
      ),
      tabItem("Timeontreatmentbymutatedgenesforestplot",
              plotOutput('figure_drug_4', 
                         height = "1500px",
                         width = "1500px"),
              hr(),
              DT::dataTableOutput("figure_drug_4_table")
      ),
      tabItem("TimeontreatmentbymutatedgenesKM-curve",
              plotOutput('figure_drug_5', 
                         height = "2500px",
                         width = "1500px")
      ),
      tabItem("TimeontreatmentandmutationsofinterestKM-curve",
              plotOutput('figure_drug_pattern', 
                         height = "2500px",
                         width = "1500px")
      ),
      tabItem("Hazardratioontimeontreatment_1",
              gt_output('figure_drug_6_1'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors selected in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity.")
      ),
      tabItem("Hazardratioontimeontreatment_2",
              gt_output('figure_drug_6_2'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors selected in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity.")
      ),
      tabItem("VolcanoPlot_ToT_1",
              h6("Volcano plots for frequent 1-50 regimens"),
              hr(),
              plotOutput('figure_volcano_ToT_1', 
                         height = "6000px",
                         width = "4000px"),
              hr()
      ),
      tabItem("VolcanoPlot_ToT_2",
              h6("Volcano plots for frequent 51-100 regimens"),
              hr(),
              plotOutput('figure_volcano_ToT_2', 
                         height = "6000px",
                         width = "4000px"),
              hr()
      ),
      tabItem("VolcanoPlot_ToT_3",
              h6("Volcano plots for frequent 101-150 regimens"),
              hr(),
              plotOutput('figure_volcano_ToT_3', 
                         height = "6000px",
                         width = "4000px"),
              hr()
      ),
      tabItem("VolcanoPlot_ToT_4",
              h6("Volcano plots for frequent 151-200 regimens"),
              hr(),
              plotOutput('figure_volcano_ToT_4', 
                         height = "6000px",
                         width = "4000px"),
              hr()
      ),
      tabItem("VolcanoTable_ToT",
              h6("Raw data for volcano plots of all regimens"),
              h6("Hazard ratio and p-value were calculated by cox regression model concering pathology."),
              h6("All regimens, and genes in which more than or equal to 8 of the treated patients had mutations were included in the analysis."),
              hr(),
              DT::dataTableOutput("table_volcano_ToT")
      ),
      tabItem("VolcanoPlotORR_1",
              h6("Volcano plots for frequent 1-50 regimens"),
              hr(),
              plotOutput('figure_volcano_1', 
                         height = "6000px",
                         width = "4000px"),
              hr(),
              h6("Figure. Patients with objective response data treated with the specified drugs were divided into two groups: those who obtained an Objective response and those who did not, and their odds ratios and p-values were calculated and a volcano plot was plotted."),
              h6("Odds ratio and p-value were calculated by logistic regression model concering pathology."),
              h6("All regimens, and genes in which more than or equal to 8 of the treated patients had mutations were included in the analysis."),
              h6("Objective response: CR or PR.")
      ),
      tabItem("VolcanoPlotORR_2",
              h6("Volcano plots for frequent 51-100 regimens"),
              hr(),
              plotOutput('figure_volcano_2', 
                         height = "6000px",
                         width = "4000px"),
              hr(),
              h6("Figure. Patients with objective response data treated with the specified drugs were divided into two groups: those who obtained an Objective response and those who did not, and their odds ratios and p-values were calculated and a volcano plot was plotted."),
              h6("Odds ratio and p-value were calculated by logistic regression model concering pathology."),
              h6("All regimens, and genes in which more than or equal to 8 of the treated patients had mutations were included in the analysis."),
              h6("Objective response: CR or PR.")
      ),
      tabItem("VolcanoPlotORR_3",
              h6("Volcano plots for frequent 101-150 regimens"),
              hr(),
              plotOutput('figure_volcano_3', 
                         height = "6000px",
                         width = "4000px"),
              hr(),
              h6("Figure. Patients with objective response data treated with the specified drugs were divided into two groups: those who obtained an Objective response and those who did not, and their odds ratios and p-values were calculated and a volcano plot was plotted."),
              h6("Odds ratio and p-value were calculated by logistic regression model concering pathology."),
              h6("All regimens, and genes in which more than or equal to 8 of the treated patients had mutations were included in the analysis."),
              h6("Objective response: CR or PR.")
      ),
      tabItem("VolcanoPlotORR_4",
              h6("Volcano plots for frequent 151-200 regimens"),
              hr(),
              plotOutput('figure_volcano_4', 
                         height = "6000px",
                         width = "4000px"),
              hr(),
              h6("Figure. Patients with objective response data treated with the specified drugs were divided into two groups: those who obtained an Objective response and those who did not, and their odds ratios and p-values were calculated and a volcano plot was plotted."),
              h6("Odds ratio and p-value were calculated by logistic regression model concering pathology."),
              h6("All regimens, and genes in which more than or equal to 8 of the treated patients had mutations were included in the analysis."),
              h6("Objective response: CR or PR.")
      ),
      tabItem("VolcanoTableORR",
              h6("Raw data for volcano plots of all regimens"),
              h6("Odds ratio and p-value were calculated by logistic regression model concering pathology."),
              h6("All regimens, and genes in which more than or equal to 8 of the treated patients had mutations were included in the analysis."),
              hr(),
              DT::dataTableOutput("table_volcano")
      ),
      tabItem("Oddsratioonobjectiveresponserate_1",
              gt_output('figure_drug_7_1'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors selected in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity."),
              h6("Candidate genes: arbitrary selected genes, the five most frequently mutated genes, genes with significance in odds ratio of objective response rate or time on treatment."),
              h6("Objective response: CR or PR.")
      ),
      tabItem("Oddsratioonobjectiveresponserate_2",
              gt_output('figure_drug_7_2'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors selected in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity."),
              h6("Candidate genes: arbitrary selected genes, the five most frequently mutated genes, genes with significance in odds ratio of objective response rate or time on treatment."),
              h6("Objective response: CR or PR.")
      ),
      tabItem("Oddsratioondiseasecontrolrate",
              gt_output('figure_drug_8'),
              hr(),
              h6("Table. Univariable and multiple variable regression analysis were performed with gtsummary package for R. Factors selected in the multivariable analysis were selected by variable decreasing method based on the Akaike information criterion with stat package for R. Variables with variance inflation factor > 10 were excluded to eliminate multicollinearity."),
              h6("Candidate genes: arbitrary selected genes, the five most frequently mutated genes, genes with significance in odds ratio of objective response rate or time on treatment."),
              h6("Disease control: CR, PR, or SD.")
      ),
      tabItem("MutationclusteringandRECIST",
              DT::dataTableOutput("table_outcome_cluster"),
              hr(),
              h6("OR: Objective response (CR or PR), ORR: Objective response rate, DC: Disease control (CR, PR, or SD), DCR: Disease control rate"),
              h6("95% confidence intervals were calculated using the Clopper-Pearson method.")
      ),
      tabItem("MutationpatternandRECIST",
              DT::dataTableOutput("table_outcome_1"),
              hr(),
              h6("OR: Objective response (CR or PR), ORR: Objective response rate, DC: Disease control (CR, PR, or SD), DCR: Disease control rate"),
              h6("95% confidence intervals were calculated using the Clopper-Pearson method.")
      ),
      tabItem("HistologyandRECIST",
              DT::dataTableOutput("table_outcome_2"),
              hr(),
              h6("OR: Objective response (CR or PR), ORR: Objective response rate, DC: Disease control (CR, PR, or SD), DCR: Disease control rate"),
              h6("95% confidence intervals were calculated using the Clopper-Pearson method.")
      ),
      tabItem("MutatedgenesandRECIST",
              DT::dataTableOutput("table_outcome_3"),
              hr(),
              h6("OR: Objective response (CR or PR), ORR: Objective response rate, DC: Disease control (CR, PR, or SD), DCR: Disease control rate"),
              h6("95% confidence intervals were calculated using the Clopper-Pearson method.")
      ),
      tabItem("Survival_drug",
              plotOutput('figure_survival_drug', 
                         height = "4000px",
                         width = "3000px")
      ),
      tabItem("Instruction",
              includeMarkdown("www/README.md")
      ),
      tabItem("Tips",
              includeMarkdown("www/Tips.md")
      )
    ),
    hr(),
    br()
  )
)

  
# Define server logic required to draw a histogram
server <- function(input, output, session) {
  Data_case_raw =  reactive({
    withProgress(message = "Clinical data loading.", {
      if(input$new_analysis =="No, use the previous dataset" &
         file.exists("source/clinical_data.qs")){
        clin_tmp = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/clinical_data.qs")
      } else {
        Encode = guess_encoding(input$clinical_files[[1, 'datapath']])[[1]][[1]]
        if(Encode == "Shift_JIS"){
          Encode = "CP932"
        }
        clin_tmp <- read_csv(col_names = TRUE,
                             file = input$clinical_files$datapath,
                             locale = locale(encoding=Encode),
                             num_threads=max(1, parallel::detectCores() - 1, na.rm = TRUE),
                             progress=FALSE,
                             show_col_types=FALSE,
                             name_repair=make.names,
                             col_select = c(
                               C.CAT調査結果.基本項目.ハッシュID,
                               症例.基本情報.年齢,
                               症例.背景情報.病理診断名,
                               症例.背景情報.臨床診断名,
                               症例.検体情報.病理診断名,
                               症例.基本情報.性別.名称.,
                               症例.基本情報.がん種.OncoTree.,
                               症例.基本情報.がん種.OncoTree..名称.,
                               症例.基本情報.がん種.OncoTree.LEVEL1.,
                               症例.検体情報.パネル.名称.,
                               症例.背景情報.ECOG.PS.名称.,
                               症例.背景情報.家族歴有無.名称.,
                               症例.背景情報.喫煙歴有無.名称.,
                               症例.背景情報.喫煙年数,
                               #症例.背景情報.喫煙１日本数,
                               症例.背景情報.アルコール多飲有無.名称.,
                               症例.背景情報.重複がん有無.異なる臓器..名称.,
                               症例.背景情報.多発がん有無.同一臓器..名称.,
                               症例.がん種情報.登録時転移有無.名称.,
                               症例.がん種情報.登録時転移部位.名称.,
                               症例.検体情報.腫瘍細胞含有割合,
                               症例.検体情報.検体採取部位.名称.,
                               症例.EP前レジメン情報.治療ライン.名称.,
                               症例.EP前レジメン情報.実施目的.名称.,
                               症例.EP前レジメン情報.化学療法レジメン名称,
                               症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                               症例.EP前レジメン情報.投与開始日,
                               症例.EP前レジメン情報.投与終了日,
                               症例.EP前レジメン情報.終了理由.名称.,
                               症例.EP前レジメン情報.レジメン継続区分.名称.,
                               症例.EP前レジメン情報.最良総合効果.名称.,
                               症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                               症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                               症例.EP後レジメン情報.治療ライン,
                               症例.EP後レジメン情報.治療ライン.名称.,
                               症例.EP後レジメン情報.化学療法レジメン名称,
                               症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                               症例.EP後レジメン情報.投与開始日,
                               症例.EP後レジメン情報.投与終了日,
                               症例.EP後レジメン情報.エキスパートパネル開催日,
                               症例.EP後レジメン情報.終了理由.名称.,
                               症例.EP後レジメン情報.レジメン継続区分.名称.,
                               症例.EP後レジメン情報.最良総合効果.名称.,
                               症例.転帰情報.転帰.名称.,
                               症例.背景情報.診断日,
                               症例.管理情報.登録日,
                               症例.転帰情報.最終生存確認日,
                               症例.転帰情報.死亡日,
                               症例.がん種情報.固形がん.マイクロサテライト不安定性.名称.,
                               症例.がん種情報.固形がん.ミスマッチ修復機能欠損.名称.,
                               症例.がん種情報.肺.EGFR.名称.,
                               症例.がん種情報.肺.ALK融合.名称.,
                               症例.がん種情報.肺.ROS1.名称.,
                               症例.がん種情報.肺.BRAF.V600..名称.,
                               症例.がん種情報.肺.MET遺伝子エクソン14スキッピング変異.名称.,
                               症例.がん種情報.肺.KRAS.G12C遺伝子変異.名称.,
                               症例.がん種情報.肺.RET融合遺伝子.名称.,
                               症例.がん種情報.肺.PD.L1.IHC..名称.,
                               症例.がん種情報.乳.HER2.IHC..名称.,
                               症例.がん種情報.乳.HER2.FISH..名称.,
                               症例.がん種情報.乳.ER.名称.,
                               症例.がん種情報.乳.PgR.名称.,
                               症例.がん種情報.食道.胃.腸.HER2.名称.,
                               症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称.
                             ),
                             col_types = cols(
                               C.CAT調査結果.基本項目.ハッシュID = col_character(),
                               症例.基本情報.年齢 = col_integer(),
                               症例.背景情報.病理診断名 = col_character(),
                               症例.背景情報.臨床診断名 = col_character(),
                               症例.検体情報.病理診断名 = col_character(),
                               症例.基本情報.性別.名称. = col_character(),
                               症例.基本情報.がん種.OncoTree. = col_character(),
                               症例.基本情報.がん種.OncoTree..名称. = col_character(),
                               症例.基本情報.がん種.OncoTree.LEVEL1. = col_character(),
                               症例.検体情報.パネル.名称. = col_character(),
                               症例.背景情報.ECOG.PS.名称. = col_character(),
                               症例.背景情報.家族歴有無.名称. = col_character(),
                               症例.背景情報.喫煙歴有無.名称. = col_character(),
                               症例.背景情報.喫煙年数 = col_integer(),
                               #症例.背景情報.喫煙１日本数 = col_integer(),
                               症例.背景情報.アルコール多飲有無.名称. = col_character(),
                               症例.背景情報.重複がん有無.異なる臓器..名称. = col_character(),
                               症例.背景情報.多発がん有無.同一臓器..名称. = col_character(),
                               症例.がん種情報.登録時転移有無.名称. = col_character(),
                               症例.がん種情報.登録時転移部位.名称. = col_character(),
                               症例.検体情報.腫瘍細胞含有割合 = col_integer(),
                               症例.検体情報.検体採取部位.名称. = col_character(),
                               症例.EP前レジメン情報.治療ライン.名称. = col_character(),
                               症例.EP前レジメン情報.実施目的.名称. = col_character(),
                               症例.EP前レジメン情報.化学療法レジメン名称 = col_character(),
                               症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = col_character(),
                               症例.EP前レジメン情報.投与開始日 = col_character(),
                               症例.EP前レジメン情報.投与終了日 = col_character(),
                               症例.EP前レジメン情報.終了理由.名称. = col_character(),
                               症例.EP前レジメン情報.レジメン継続区分.名称. = col_character(),
                               症例.EP前レジメン情報.最良総合効果.名称. = col_character(),
                               症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. = col_character(),
                               症例.EP後レジメン情報.提示された治療薬を投与した.名称. = col_character(),
                               症例.EP後レジメン情報.治療ライン = col_integer(),
                               症例.EP後レジメン情報.治療ライン.名称. = col_character(),
                               症例.EP後レジメン情報.化学療法レジメン名称 = col_character(),
                               症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = col_character(),
                               症例.EP後レジメン情報.投与開始日 = col_character(),
                               症例.EP後レジメン情報.投与終了日 = col_character(),
                               症例.EP後レジメン情報.エキスパートパネル開催日 = col_character(),
                               症例.EP後レジメン情報.終了理由.名称. = col_character(),
                               症例.EP後レジメン情報.レジメン継続区分.名称. = col_character(),
                               症例.EP後レジメン情報.最良総合効果.名称. = col_character(),
                               症例.転帰情報.転帰.名称. = col_character(),
                               症例.背景情報.診断日 = col_character(),
                               症例.管理情報.登録日 = col_character(),
                               症例.転帰情報.最終生存確認日 = col_character(),
                               症例.転帰情報.死亡日 = col_character(),
                               症例.がん種情報.固形がん.マイクロサテライト不安定性.名称. = col_character(),
                               症例.がん種情報.固形がん.ミスマッチ修復機能欠損.名称. = col_character(),
                               症例.がん種情報.肺.EGFR.名称. = col_character(),
                               症例.がん種情報.肺.ALK融合.名称. = col_character(),
                               症例.がん種情報.肺.ROS1.名称. = col_character(),
                               症例.がん種情報.肺.BRAF.V600..名称. = col_character(),
                               症例.がん種情報.肺.MET遺伝子エクソン14スキッピング変異.名称. = col_character(),
                               症例.がん種情報.肺.KRAS.G12C遺伝子変異.名称. = col_character(),
                               症例.がん種情報.肺.RET融合遺伝子.名称. = col_character(),
                               症例.がん種情報.肺.PD.L1.IHC..名称. = col_character(),
                               症例.がん種情報.乳.HER2.IHC..名称. = col_character(),
                               症例.がん種情報.乳.HER2.FISH..名称. = col_character(),
                               症例.がん種情報.乳.ER.名称. = col_character(),
                               症例.がん種情報.乳.PgR.名称. = col_character(),
                               症例.がん種情報.食道.胃.腸.HER2.名称. = col_character(),
                               症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称. = col_character(),
                               .default = col_guess()
                             ))
        incProgress(1 / 4)
        #clin_tmp = as.data.frame(clin_tmp)
        clin_tmp = data.table(clin_tmp)
        clin_tmp$症例.背景情報.ECOG.PS.名称. = as.character(clin_tmp$症例.背景情報.ECOG.PS.名称.)
        clin_tmp$症例.基本情報.がん種.OncoTree..名称. = 
          str_split(str_replace(clin_tmp$症例.基本情報.がん種.OncoTree..名称.,
                                "\\)_", "\\)__"), "__", simplify = TRUE)[,1]
        clin_tmp = clin_tmp %>% dplyr::mutate(
          症例.基本情報.がん種.OncoTree..名称. = case_when(
            str_detect(症例.基本情報.がん種.OncoTree..名称., paste0(" \\(", 症例.基本情報.がん種.OncoTree.LEVEL1., "\\)")) ~ 症例.基本情報.がん種.OncoTree.LEVEL1.,
            TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
          ),
          症例.がん種情報.登録時転移有無.名称. = case_when(
            is.na(症例.がん種情報.登録時転移有無.名称.) ~ "不明",
            症例.がん種情報.登録時転移有無.名称. == "" ~ "不明",
            TRUE ~ 症例.がん種情報.登録時転移有無.名称.
          ),
          症例.背景情報.家族歴有無.名称. = case_when(
            is.na(症例.背景情報.家族歴有無.名称.) ~ "不明",
            症例.背景情報.家族歴有無.名称. == "" ~ "不明",
            TRUE ~ 症例.背景情報.家族歴有無.名称.
          ),
          症例.背景情報.ECOG.PS.名称. = case_when(
            is.na(症例.背景情報.ECOG.PS.名称.) ~ "不明",
            症例.背景情報.ECOG.PS.名称. == "" ~ "不明",
            TRUE ~ 症例.背景情報.ECOG.PS.名称.
          ),
          症例.基本情報.性別.名称. = case_when(
            is.na(症例.基本情報.性別.名称.) ~ "未入力・不明",
            症例.基本情報.性別.名称. == "" ~ "未入力・不明",
            TRUE ~ 症例.基本情報.性別.名称.
          ),
          症例.基本情報.がん種.OncoTree.LEVEL1. = case_when(
            症例.基本情報.がん種.OncoTree.LEVEL1. == "WHO_BRAIN" ~ "BRAIN",
            TRUE ~ 症例.基本情報.がん種.OncoTree.LEVEL1.
          )
        )
        
        if(input$histology_detail == "Yes, use oncotree 1st level"){
          clin_tmp$症例.基本情報.がん種.OncoTree..名称. = clin_tmp$症例.基本情報.がん種.OncoTree.LEVEL1.
        }
        if(!is.null(input$ID_histology)){
          ID_histology_list = data.frame(NULL)
          for(i in 1:length(input$ID_histology[,1])){
            ID_histology_list <- rbind(ID_histology_list,
                              read.csv(header = TRUE,
                                file(input$ID_histology[[i, 'datapath']],
                                     encoding='UTF-8-BOM')))
          }
          clin_tmp$症例.基本情報.がん種.OncoTree..名称._tmp = unlist(lapply(list(clin_tmp$C.CAT調査結果.基本項目.ハッシュID), function(x) {
            as.vector(ID_histology_list$Histology[match(x, ID_histology_list$ID)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            症例.基本情報.がん種.OncoTree..名称. = case_when(
              !is.na(症例.基本情報.がん種.OncoTree..名称._tmp) ~ 症例.基本情報.がん種.OncoTree..名称._tmp,
              TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
            )
          )
        }
        if(!is.null(input$regimen_rename)){
          RenList = data.frame(NULL)
          for(i in 1:length(input$regimen_rename[,1])){
            RenList <- rbind(RenList,
                             read.csv(header = TRUE,
                                      file(input$regimen_rename[[i, 'datapath']],
                                           encoding='UTF-8-BOM')))
          }
          for(i in 1:length(RenList$P1)){
            clin_tmp = clin_tmp %>% dplyr::mutate(
              症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = case_when(
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "" &
                  症例.EP前レジメン情報.化学療法レジメン名称 == RenList$P1[i] ~ RenList$P2[i],
                TRUE ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
              ))
          }
        }
        if(!is.null(input$drug_rename)){
          drug_rename_list = data.frame(NULL)
          for(i in 1:length(input$drug_rename[,1])){
            drug_rename_list <- rbind(drug_rename_list,
                                       read.csv(header = TRUE,
                                         file(input$drug_rename[[i, 'datapath']],
                                              encoding='UTF-8-BOM')))
          }
          clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp = unlist(lapply(list(clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.), function(x) {
            as.vector(drug_rename_list$Rename[match(x, drug_rename_list$Drug)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = case_when(
              !is.na(症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp) ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp,
              TRUE ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
            )
          )
        }
        clin_tmp = clin_tmp %>% dplyr::mutate(
          症例.基本情報.がん種.OncoTree..名称. = case_when(
            is.na(症例.基本情報.がん種.OncoTree..名称.) ~ 症例.基本情報.がん種.OncoTree.LEVEL1.,
            症例.基本情報.がん種.OncoTree..名称. == "" ~ 症例.基本情報.がん種.OncoTree.LEVEL1.,
            TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
          )
        )
        if(!is.null(input$histology_rename)){
          histology_rename_list = data.frame(NULL)
          for(i in 1:length(input$histology_rename[,1])){
            histology_rename_list <- rbind(histology_rename_list,
                                           read.csv(header = TRUE,
                                             file(input$histology_rename[[i, 'datapath']],
                                                  encoding='UTF-8-BOM')))
          }
          clin_tmp$症例.基本情報.がん種.OncoTree..名称._tmp = unlist(lapply(list(clin_tmp$症例.基本情報.がん種.OncoTree..名称.), function(x) {
            as.vector(histology_rename_list$Rename[match(x, histology_rename_list$Histology)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            症例.基本情報.がん種.OncoTree..名称. = case_when(
              !is.na(症例.基本情報.がん種.OncoTree..名称._tmp) ~ 症例.基本情報.がん種.OncoTree..名称._tmp,
              TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
            )
          )
        }
        
        # NA の処理
        clin_tmp <- clin_tmp %>%
          mutate(across(
            c(
              症例.検体情報.検体採取部位.名称.,
              症例.背景情報.ECOG.PS.名称.,
              症例.背景情報.重複がん有無.異なる臓器..名称.,
              症例.背景情報.多発がん有無.同一臓器..名称.
            ),
            ~ replace_na(.x, "不明")
          )) %>%
          mutate(across(
            c(
              症例.背景情報.喫煙歴有無.名称.,
              症例.検体情報.パネル.名称.,
              症例.がん種情報.登録時転移部位.名称.,
              症例.転帰情報.死亡日,
              症例.転帰情報.最終生存確認日,
              症例.管理情報.登録日,
              症例.EP前レジメン情報.投与開始日,
              症例.EP前レジメン情報.投与終了日,
              症例.EP後レジメン情報.投与開始日,
              症例.EP後レジメン情報.投与終了日,
              症例.EP後レジメン情報.エキスパートパネル開催日,
              症例.背景情報.診断日
            ),
            ~ replace_na(.x, "")
          )) %>%
          mutate(across(
            c(
              症例.管理情報.登録日,
              症例.転帰情報.最終生存確認日,
              症例.転帰情報.死亡日,
              症例.EP前レジメン情報.投与開始日,
              症例.EP前レジメン情報.投与終了日,
              症例.EP後レジメン情報.投与開始日,
              症例.EP後レジメン情報.投与終了日,
              症例.EP後レジメン情報.エキスパートパネル開催日,
              症例.背景情報.診断日
            ),
            ~ str_replace_all(.x, "/", "-")
          )) %>%
          mutate(症例.基本情報.年齢 = as.integer(症例.基本情報.年齢))
      
        incProgress(1 / 4)
        
        clin_tmp$症例.基本情報.がん種.OncoTree..名称. = 
          str_split(str_replace(clin_tmp$症例.基本情報.がん種.OncoTree..名称.,
                                "\\)_", "\\)__"), "__", simplify = TRUE)[,1]
        if(!is.null(input$ID_histology)){
          clin_tmp$症例.基本情報.がん種.OncoTree..名称._tmp = unlist(lapply(list(clin_tmp$症例.基本情報.がん種.OncoTree..名称.), function(x) {
            as.vector(ID_histology_list$Histology[match(x, ID_histology_list$ID)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            症例.基本情報.がん種.OncoTree..名称. = case_when(
              !is.na(症例.基本情報.がん種.OncoTree..名称._tmp) ~ 症例.基本情報.がん種.OncoTree..名称._tmp,
              TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
            )
          )
        }
        clin_tmp = clin_tmp %>% dplyr::mutate(
          症例.基本情報.がん種.OncoTree..名称. = case_when(
            症例.基本情報.がん種.OncoTree..名称. == "" ~ 症例.基本情報.がん種.OncoTree.LEVEL1.,
            TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
          ),
          症例.基本情報.がん種.OncoTree. = case_when(
            症例.基本情報.がん種.OncoTree. == "" ~ 症例.基本情報.がん種.OncoTree.LEVEL1.,
            TRUE ~ 症例.基本情報.がん種.OncoTree.
          )
        )
        if(!is.null(input$histology_rename)){
          clin_tmp$症例.基本情報.がん種.OncoTree..名称._tmp = unlist(lapply(list(clin_tmp$症例.基本情報.がん種.OncoTree..名称.), function(x) {
            as.vector(histology_rename_list$Rename[match(x, histology_rename_list$Histology)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            症例.基本情報.がん種.OncoTree..名称. = case_when(
              !is.na(症例.基本情報.がん種.OncoTree..名称._tmp) ~ 症例.基本情報.がん種.OncoTree..名称._tmp,
              TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
            )
          )
        }
        clin_tmp = clin_tmp %>% dplyr::mutate(
          症例.基本情報.がん種.OncoTree. = str_replace(str_replace(症例.基本情報.がん種.OncoTree..名称., "\\)", ""), ".*\\(",""),
          症例.背景情報.アルコール多飲有無.名称. = case_when(
            is.na(症例.背景情報.アルコール多飲有無.名称.) ~ "不明",
            症例.背景情報.アルコール多飲有無.名称. == "" ~ "不明",
            TRUE ~ 症例.背景情報.アルコール多飲有無.名称.
          ),
          症例.背景情報.重複がん有無.異なる臓器..名称. = case_when(
            症例.背景情報.重複がん有無.異なる臓器..名称. == "" ~ "不明",
            TRUE ~ 症例.背景情報.重複がん有無.異なる臓器..名称.
          ),
          症例.背景情報.多発がん有無.同一臓器..名称. = case_when(
            症例.背景情報.多発がん有無.同一臓器..名称. == "" ~ "不明",
            TRUE ~ 症例.背景情報.多発がん有無.同一臓器..名称.
          ),
          症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. = case_when(
            is.na(症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.) ~ "不明",
            症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. == "" ~ "いいえ",
            TRUE ~ 症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.
          ),
          症例.EP後レジメン情報.提示された治療薬を投与した.名称. = case_when(
            is.na(症例.EP後レジメン情報.提示された治療薬を投与した.名称.) ~ "不明",
            症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "不明" ~ "不明",
            症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "" ~ "投与しなかった",
            TRUE ~ 症例.EP後レジメン情報.提示された治療薬を投与した.名称.
          ),
          症例.背景情報.喫煙歴有無.名称. = case_when(
            症例.背景情報.喫煙歴有無.名称. == "" ~ "不明",
            TRUE ~ 症例.背景情報.喫煙歴有無.名称.
          )
        )
        convert_date <- function(x) {
          case_when(
            x < "1924-02-19" & x > "1901-01-01" ~ paste0("20", str_sub(x, 3, 10)),
            TRUE ~ x
          )
        }
        
        # 対象の5列に一括で適用
        clin_tmp <- clin_tmp %>%
          mutate(across(
            c(
              症例.EP前レジメン情報.投与開始日,
              症例.EP後レジメン情報.投与開始日,
              症例.EP前レジメン情報.投与終了日,
              症例.EP後レジメン情報.投与終了日,
              症例.EP後レジメン情報.エキスパートパネル開催日
            ),
            convert_date
          ),
            Lymph_met = if_else(
              症例.がん種情報.登録時転移部位.名称. %in% c("リンパ節", "リンパ節/リンパ管"),
              "Yes", "No"
            ),
            Brain_met = if_else(
              症例.がん種情報.登録時転移部位.名称. %in% c("脳", "中枢神経系"),
              "Yes", "No"
            ),
            Lung_met = if_else(
              症例.がん種情報.登録時転移部位.名称. == "肺",
              "Yes", "No"
            ),
            Bone_met = if_else(
              症例.がん種情報.登録時転移部位.名称. %in% c("骨髄", "骨"),
              "Yes", "No"
            ),
            Liver_met = if_else(
              症例.がん種情報.登録時転移部位.名称. == "肝",
              "Yes", "No"
            ),
            Other_met = if_else(
              !症例.がん種情報.登録時転移部位.名称. %in%
                c("リンパ節", "リンパ節/リンパ管", "脳", "中枢神経系", "肺", "骨髄", "骨", "肝", ""),
              "Yes", "No"
            ),
            Not_brain_bone_liver_met = if_else(
              !症例.がん種情報.登録時転移部位.名称. %in%
                c("脳", "中枢神経系", "骨髄", "骨", "肝", ""),
              "Yes", "No"
            )
          )
        
        RenList = read.csv(file("source/drug_rename.csv",
                                encoding='UTF-8-BOM'), header = T)
        for(i in 1:length(RenList$P1)){
          clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN., RenList$P1[i], RenList$P2[i])
          clin_tmp$症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = str_replace_all(clin_tmp$症例.EP後レジメン情報.薬剤名.YJ一般名.EN., RenList$P1[i], RenList$P2[i])
        }
        
        if(!is.null(input$regimen_rename)){
          RenList = data.frame(NULL)
          for(i in 1:length(input$regimen_rename[,1])){
            RenList <- rbind(RenList,
                             read.csv(header = TRUE,
                                      file(input$regimen_rename[[i, 'datapath']],
                                           encoding='UTF-8-BOM')))
          }
          for(i in 1:length(RenList$P1)){
            clin_tmp = clin_tmp %>% dplyr::mutate(
              症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = case_when(
                症例.EP前レジメン情報.薬剤名.YJ一般名.EN. == "" &
                  症例.EP前レジメン情報.化学療法レジメン名称 == RenList$P1[i] ~ RenList$P2[i],
                TRUE ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
              ))
          }
        }
        if(!is.null(input$drug_rename)){
          drug_rename_list = data.frame(NULL)
          for(i in 1:length(input$drug_rename[,1])){
            drug_rename_list <- rbind(drug_rename_list,
                                      read.csv(header = TRUE,
                                               file(input$drug_rename[[i, 'datapath']],
                                                    encoding='UTF-8-BOM')))
          }
          clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp = unlist(lapply(list(clin_tmp$症例.EP前レジメン情報.薬剤名.YJ一般名.EN.), function(x) {
            as.vector(drug_rename_list$Rename[match(x, drug_rename_list$Drug)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            症例.EP前レジメン情報.薬剤名.YJ一般名.EN. = case_when(
              !is.na(症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp) ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN._tmp,
              TRUE ~ 症例.EP前レジメン情報.薬剤名.YJ一般名.EN.
            )
          )
        }
        incProgress(1 / 4)
        clin_tmp = clin_tmp %>%
          dplyr::mutate(HER2_IHC = case_when(
            症例.がん種情報.乳.HER2.IHC..名称. == "陽性（3+）" |
              症例.がん種情報.食道.胃.腸.HER2.名称. == "陽性（3+）" |
              症例.がん種情報.乳.HER2.IHC..名称. == "境界域（2+）" & 症例.がん種情報.乳.HER2.FISH..名称. == "陽性" |
              症例.がん種情報.食道.胃.腸.HER2.名称. == "境界域（2+）" & 症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称. == "陽性" ~ "Positive",
            症例.がん種情報.乳.HER2.IHC..名称. %in% c("陰性",  "陰性（1+）") |
              症例.がん種情報.食道.胃.腸.HER2.名称. %in% c("陰性",  "陰性（1+）") |
              症例.がん種情報.乳.HER2.IHC..名称. == "境界域（2+）" & 症例.がん種情報.乳.HER2.FISH..名称. %in% c("equivocal", "陰性") |
              症例.がん種情報.食道.胃.腸.HER2.名称. == "境界域（2+）" & 症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称. %in% c("equivocal", "陰性") ~ "Negative",
            TRUE ~ "Unknown"
          ),
          MSI_PCR = case_when(
            症例.がん種情報.固形がん.マイクロサテライト不安定性.名称. == "陽性" ~ "Positive",
            症例.がん種情報.固形がん.マイクロサテライト不安定性.名称. == "陰性" ~ "Negative",
            TRUE ~ "Unknown"
          ),
          MMR_IHC = case_when(
            症例.がん種情報.固形がん.ミスマッチ修復機能欠損.名称. == "dMMR(欠損)" ~ "dMMR",
            症例.がん種情報.固形がん.ミスマッチ修復機能欠損.名称. == "pMMR(正常)" ~ "pMMR",
            TRUE ~ "Unknown"
          )
        )
  
        clin_tmp = clin_tmp %>%
          dplyr::mutate(
            EP_option = case_when(
            症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. ==
              "はい" ~ 1,
            TRUE ~ 0),
            EP_treat = case_when(
              症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
              "投与した" ~ 1,
            TRUE ~ 0)) %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                        症例.基本情報.年齢,
                        Lymph_met,
                        Brain_met,
                        Lung_met,
                        Bone_met,
                        Liver_met,
                        Other_met,
                        Not_brain_bone_liver_met,
                        EP_option,
                        EP_treat,
                        症例.背景情報.病理診断名,
                        症例.背景情報.臨床診断名,
                        症例.検体情報.病理診断名,
                        症例.基本情報.性別.名称.,
                        症例.基本情報.がん種.OncoTree.,
                        症例.基本情報.がん種.OncoTree..名称.,
                        症例.基本情報.がん種.OncoTree.LEVEL1.,
                        症例.検体情報.パネル.名称.,
                        症例.背景情報.ECOG.PS.名称.,
                        症例.背景情報.家族歴有無.名称.,
                        症例.背景情報.喫煙歴有無.名称.,
                        症例.背景情報.喫煙年数,
                        #症例.背景情報.喫煙１日本数,
                        症例.背景情報.アルコール多飲有無.名称.,
                        症例.背景情報.重複がん有無.異なる臓器..名称.,
                        症例.背景情報.多発がん有無.同一臓器..名称.,
                        症例.がん種情報.登録時転移有無.名称.,
                        症例.がん種情報.登録時転移部位.名称.,
                        症例.検体情報.腫瘍細胞含有割合,
                        症例.検体情報.検体採取部位.名称.,
                        症例.EP前レジメン情報.治療ライン.名称.,
                        症例.EP前レジメン情報.実施目的.名称.,
                        症例.EP前レジメン情報.化学療法レジメン名称,
                        症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                        症例.EP前レジメン情報.投与開始日,
                        症例.EP前レジメン情報.投与終了日,
                        症例.EP前レジメン情報.終了理由.名称.,
                        症例.EP前レジメン情報.レジメン継続区分.名称.,
                        症例.EP前レジメン情報.最良総合効果.名称.,
                        症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                        症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                        症例.EP後レジメン情報.治療ライン,
                        症例.EP後レジメン情報.治療ライン.名称.,
                        症例.EP後レジメン情報.化学療法レジメン名称,
                        症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                        症例.EP後レジメン情報.投与開始日,
                        症例.EP後レジメン情報.投与終了日,
                        症例.EP後レジメン情報.終了理由.名称.,
                        症例.EP後レジメン情報.レジメン継続区分.名称.,
                        症例.EP後レジメン情報.最良総合効果.名称.,
                        症例.転帰情報.転帰.名称.,
                        症例.背景情報.診断日,
                        症例.管理情報.登録日,
                        症例.転帰情報.最終生存確認日,
                        症例.転帰情報.死亡日,
                        症例.がん種情報.固形がん.マイクロサテライト不安定性.名称.,
                        症例.がん種情報.固形がん.ミスマッチ修復機能欠損.名称.,
                        症例.がん種情報.肺.EGFR.名称.,
                        症例.がん種情報.肺.ALK融合.名称.,
                        症例.がん種情報.肺.ROS1.名称.,
                        症例.がん種情報.肺.BRAF.V600..名称.,
                        症例.がん種情報.肺.MET遺伝子エクソン14スキッピング変異.名称.,
                        症例.がん種情報.肺.KRAS.G12C遺伝子変異.名称.,
                        症例.がん種情報.肺.RET融合遺伝子.名称.,
                        症例.がん種情報.肺.PD.L1.IHC..名称.,
                        症例.がん種情報.乳.HER2.IHC..名称.,
                        症例.がん種情報.乳.HER2.FISH..名称.,
                        症例.がん種情報.乳.ER.名称.,
                        症例.がん種情報.乳.PgR.名称.,
                        症例.がん種情報.食道.胃.腸.HER2.名称.,
                        症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称.,
                        HER2_IHC,
                        MSI_PCR,
                        MMR_IHC
          )
        incProgress(1 / 4)
        QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), clin_tmp, file="source/clinical_data.qs")
      }
    })  
    return(clin_tmp)
  })

  Data_report_raw =  reactive({
    withProgress(message = "Mutation data loading.", {
      if(input$new_analysis =="No, use the previous dataset" &
         file.exists("source/variant_data.qs")){
        Data_MAF = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/variant_data.qs")
      } else {
        Encode = guess_encoding(input$report_files[[1, 'datapath']])[[1]][[1]]
        if(Encode == "Shift_JIS"){
          Encode = "CP932"
        }
        clin_tmp <- read_csv(col_names = TRUE,
                             file = input$report_files$datapath,
                             locale = locale(encoding=Encode),
                             num_threads=max(1, parallel::detectCores() - 1, na.rm = TRUE),
                             progress=FALSE,
                             show_col_types=FALSE,
                             name_repair=make.names,
                             col_select = c(
                               C.CAT調査結果.基本項目.ハッシュID,
                               C.CAT調査結果.変異情報.変異種類,
                               C.CAT調査結果.変異情報.変異由来,
                               C.CAT調査結果.変異情報.マーカー,
                               C.CAT調査結果.変異情報.変異内容,
                               C.CAT調査結果.変異情報.マーカー詳細,
                               C.CAT調査結果.変異情報.変異アレル頻度,
                               C.CAT調査結果.変異情報.コピー数変化,
                               C.CAT調査結果.変異情報.クロモソーム番号,
                               C.CAT調査結果.変異情報.物理位置,
                               C.CAT調査結果.変異情報.リファレンス塩基,
                               C.CAT調査結果.変異情報.塩基変化,
                               C.CAT調査結果.変異情報.変異種類,
                               C.CAT調査結果.変異情報.変異アレル頻度,
                               C.CAT調査結果.変異情報.変異内容,
                               C.CAT調査結果.変異情報TMB.TMB,
                               C.CAT調査結果.エビデンス.エビデンスレベル,
                               C.CAT調査結果.エビデンス.薬剤
                             ),
                             col_types = cols(
                               C.CAT調査結果.基本項目.ハッシュID = col_character(),
                               C.CAT調査結果.変異情報.変異種類 = col_character(),
                               C.CAT調査結果.変異情報.変異由来 = col_character(),
                               C.CAT調査結果.変異情報.マーカー = col_character(),
                               C.CAT調査結果.変異情報.変異内容 = col_character(),
                               C.CAT調査結果.変異情報.マーカー詳細 = col_character(),
                               C.CAT調査結果.変異情報.変異アレル頻度 = col_character(),
                               C.CAT調査結果.変異情報.コピー数変化 = col_character(),
                               C.CAT調査結果.変異情報.クロモソーム番号 = col_character(),
                               C.CAT調査結果.変異情報.物理位置 = col_character(),
                               C.CAT調査結果.変異情報.リファレンス塩基 = col_character(),
                               C.CAT調査結果.変異情報.塩基変化 = col_character(),
                               C.CAT調査結果.変異情報.変異種類 = col_character(),
                               C.CAT調査結果.変異情報.変異アレル頻度 = col_character(),
                               C.CAT調査結果.変異情報.変異内容 = col_character(),
                               C.CAT調査結果.変異情報TMB.TMB = col_character(),
                               C.CAT調査結果.エビデンス.エビデンスレベル = col_character(),
                               C.CAT調査結果.エビデンス.薬剤 = col_character(),
                              .default = col_guess()
                             ))
        incProgress(1 / 4)
        #clin_tmp = as.data.frame(clin_tmp)
        clin_tmp = data.table(clin_tmp) %>%
          dplyr::filter(
            !str_detect(C.CAT調査結果.変異情報.マーカー, ",") &
              C.CAT調査結果.変異情報.マーカー != "" &
              C.CAT調査結果.変異情報.変異種類 != "expression"
          ) %>% dplyr::mutate(
          C.CAT調査結果.変異情報.変異種類 = case_when(
            C.CAT調査結果.変異情報.変異種類 == "rearrangement" &
              C.CAT調査結果.変異情報.変異内容 == "rearrangements" ~ "rearrangement",
            str_detect(C.CAT調査結果.変異情報.変異種類, "exon skipping") ~ "exon_skipping",
            str_detect(C.CAT調査結果.変異情報.変異内容, "exon skipping") ~ "exon_skipping",
            C.CAT調査結果.変異情報.変異種類 == "rearrangement" &
              C.CAT調査結果.変異情報.変異内容 == "long deletion" ~ "deletion",
            C.CAT調査結果.変異情報.変異種類 == "rearrangement" ~ C.CAT調査結果.変異情報.変異内容,
            C.CAT調査結果.変異情報.変異種類 == "copy_number_alteration" ~ C.CAT調査結果.変異情報.変異内容,
            C.CAT調査結果.変異情報.変異種類 == "small_scale_variant" &
             str_detect(C.CAT調査結果.変異情報.変異内容, "\\*") ~ "nonsense_frameshift",
            C.CAT調査結果.変異情報.変異種類 == "small_scale_variant"  ~ "small_scale_variant",
            C.CAT調査結果.変異情報.変異種類 == "other_biomarker" ~ "TMB_MSI_high",
            TRUE ~ C.CAT調査結果.変異情報.変異種類
          ))
        if(!is.null(input$reaanotation)){
          reaanotation_list = data.frame(NULL)
          for(i in 1:length(input$reaanotation[,1])){
            reaanotation_list <- rbind(reaanotation_list,
                                          read.csv(header = TRUE,
                                                   file(input$reaanotation[[i, 'datapath']],
                                                        encoding='UTF-8-BOM')))
          }
          clin_tmp$Gene_alt = paste0(clin_tmp$C.CAT調査結果.変異情報.マーカー, "_", clin_tmp$C.CAT調査結果.変異情報.変異内容)
          clin_tmp$C.CAT調査結果.エビデンス.エビデンスレベル_tmp = unlist(lapply(list(clin_tmp$Gene_alt), function(x) {
            as.vector(reaanotation_list$Level[match(x, reaanotation_list$Gene_alt)])}))
          clin_tmp = clin_tmp %>% dplyr::mutate(
            C.CAT調査結果.エビデンス.エビデンスレベル = case_when(
              !is.na(C.CAT調査結果.エビデンス.エビデンスレベル_tmp) ~ C.CAT調査結果.エビデンス.エビデンスレベル_tmp,
              TRUE ~ C.CAT調査結果.エビデンス.エビデンスレベル
            ))
          clin_tmp = clin_tmp %>% dplyr::select(-Gene_alt, -C.CAT調査結果.エビデンス.エビデンスレベル_tmp)
        }
        clin_tmp = clin_tmp %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "NKX2\\-1", "NKX2XXXX1"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "NKX3\\-1", "NKX3XXXX1"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H1\\-2", "H1XXXX2"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3\\-4", "H3XXXX4"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3\\-5", "H3XXXX5"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3\\-3B", "H3XXXX3B"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "HLA\\-DRB1", "HLAXXXXDRB1"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3\\-3A", "H3XXXX3A")
        ) %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "-", "_")
          )
        
        if(input$fusion == "Split into two and treat as mutation in each gene"){
          if(any(str_detect(clin_tmp$C.CAT調査結果.変異情報.マーカー, "_"))){
            clin_tmp_base = clin_tmp %>%
              dplyr::filter(!str_detect(C.CAT調査結果.変異情報.マーカー, "_"))
            clin_tmp_fusion = clin_tmp %>%
              dplyr::filter(str_detect(C.CAT調査結果.変異情報.マーカー, "_"))
            clin_tmp_fusion_1 = clin_tmp_fusion %>%
              dplyr::mutate(C.CAT調査結果.変異情報.変異内容 = C.CAT調査結果.変異情報.マーカー,
                            C.CAT調査結果.変異情報.マーカー = str_split(C.CAT調査結果.変異情報.マーカー, "_",simplify = T)[,1])
            clin_tmp_fusion_2 = clin_tmp_fusion %>%
              dplyr::mutate(C.CAT調査結果.変異情報.変異内容 = C.CAT調査結果.変異情報.マーカー,
                            C.CAT調査結果.変異情報.マーカー = str_split(C.CAT調査結果.変異情報.マーカー, "_",simplify = T)[,2])
            clin_tmp = rbind(clin_tmp_base, clin_tmp_fusion_1, clin_tmp_fusion_2)
          } else if(input$fusion == "All fusion genes are named as 'fusion_genes'"){
            clin_tmp_base = clin_tmp %>%
              dplyr::filter(!str_detect(C.CAT調査結果.変異情報.マーカー, "_"))
            clin_tmp_fusion = clin_tmp %>%
              dplyr::filter(str_detect(C.CAT調査結果.変異情報.マーカー, "_"))
            clin_tmp_fusion = clin_tmp_fusion %>%
              dplyr::mutate(C.CAT調査結果.変異情報.変異内容 = C.CAT調査結果.変異情報.マーカー,
                            C.CAT調査結果.変異情報.マーカー = "fusion_genes")
            clin_tmp = rbind(clin_tmp_base, clin_tmp_fusion)
          }
        }
        if(input$amplification == "Treat as indipendent gene"){
          clin_tmp_base = clin_tmp %>%
            dplyr::filter(C.CAT調査結果.変異情報.変異内容 != "amplification" &
                            C.CAT調査結果.変異情報.変異内容 != "loss")
          clin_tmp_amplification = clin_tmp %>%
            dplyr::filter(C.CAT調査結果.変異情報.変異内容 == "amplification") %>%
            dplyr::mutate(C.CAT調査結果.変異情報.マーカー = paste0(C.CAT調査結果.変異情報.マーカー, "_amp"))
          clin_tmp_loss = clin_tmp %>%
            dplyr::filter(C.CAT調査結果.変異情報.変異内容 == "loss") %>%
            dplyr::mutate(C.CAT調査結果.変異情報.マーカー = paste0(C.CAT調査結果.変異情報.マーカー, "_loss"))
          clin_tmp = rbind(clin_tmp_base, clin_tmp_amplification, clin_tmp_loss)
        }
        if(input$T_N == "Only somatic for T/N panel"){
          clin_tmp = clin_tmp %>%
            dplyr::filter(C.CAT調査結果.変異情報.変異由来 != "Germline")
        }
        clin_tmp = clin_tmp %>%
          dplyr::mutate(
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "NKX2XXXX1", "NKX2_1"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "NKX3XXXX1", "NKX3_1"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H1XXXX2", "H1_2"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3XXXX4", "H3_4"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3XXXX5", "H3_5"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3XXXX3B", "H3_3B"),
            C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "HLAXXXXDRB1", "HLA_DRB1"),
          C.CAT調査結果.変異情報.マーカー = str_replace(C.CAT調査結果.変異情報.マーカー, "H3XXXX3A", "H3_3A")
        )
        incProgress(1 / 4)
        if(!is.null(input$mutation_rename)){
          mutation_rename_list = data.frame(NULL)
          for(i in 1:length(input$mutation_rename[,1])){
            mutation_rename_list <- rbind(mutation_rename_list,
                                      read.csv(header = TRUE,
                                        file(input$mutation_rename[[i, 'datapath']],
                                             encoding='UTF-8-BOM')))
          }
          Gene_list = sort(unique(mutation_rename_list$Gene))
          clin = clin_tmp %>% dplyr::filter(!C.CAT調査結果.変異情報.マーカー %in% Gene_list)
          for(i in 1:length(Gene_list)){
            mutation_rename_list_tmp = mutation_rename_list[mutation_rename_list$Gene == Gene_list[i],]
            clin_tmp2 = clin_tmp %>% dplyr::filter(C.CAT調査結果.変異情報.マーカー == Gene_list[i])
            clin_tmp2$C.CAT調査結果.変異情報.変異内容_tmp = unlist(lapply(list(clin_tmp2$C.CAT調査結果.変異情報.変異内容), function(x) {
              as.vector(mutation_rename_list_tmp$Rename[match(x, mutation_rename_list_tmp$Mutation)])}))
            clin_tmp2 = clin_tmp2 %>% dplyr::mutate(
              C.CAT調査結果.変異情報.変異内容 = case_when(
                !is.na(C.CAT調査結果.変異情報.変異内容_tmp) ~ C.CAT調査結果.変異情報.変異内容_tmp,
                TRUE ~ "Other"
              )
            )
            clin_tmp2 = clin_tmp2 %>% dplyr::select(-C.CAT調査結果.変異情報.変異内容_tmp)
            clin = rbind(clin, clin_tmp2)
          }
          clin_tmp = clin
        }
        Data_MAF = clin_tmp
        Data_MAF$C.CAT調査結果.変異情報.物理位置2 =
          Data_MAF$C.CAT調査結果.変異情報.物理位置
        Data_MAF = Data_MAF %>%
          dplyr::select(C.CAT調査結果.変異情報.マーカー,
                        C.CAT調査結果.変異情報.クロモソーム番号,
                        C.CAT調査結果.変異情報.物理位置,
                        C.CAT調査結果.変異情報.物理位置2,
                        C.CAT調査結果.変異情報.リファレンス塩基,
                        C.CAT調査結果.変異情報.塩基変化,
                        C.CAT調査結果.変異情報.変異種類,
                        C.CAT調査結果.エビデンス.エビデンスレベル,
                        C.CAT調査結果.エビデンス.薬剤,
                        C.CAT調査結果.基本項目.ハッシュID,
                        C.CAT調査結果.変異情報.変異アレル頻度,
                        C.CAT調査結果.変異情報.変異内容,
                        C.CAT調査結果.変異情報TMB.TMB)
        colnames(Data_MAF) = c("Hugo_Symbol",
                               "Chromosome",
                               "Start_Position",
                               "End_Position",
                               "Reference_Allele",
                               "Tumor_Seq_Allele2",
                               "Variant_Classification",
                               "Evidence_level",
                               "Drug",
                               "Tumor_Sample_Barcode",
                               "VAF",
                               "amino.acid.change",
                               "TMB")
        Data_MAF = Data_MAF %>% dplyr::mutate(
          Variant_Type = case_when(
            Hugo_Symbol == "MSI" ~ "MSI",
            Hugo_Symbol == "TMB" ~ "TMB",
            nchar(Reference_Allele) == 1 &
              nchar(Tumor_Seq_Allele2) == 1 ~ "SNP",
            nchar(Reference_Allele) == 2 &
              nchar(Tumor_Seq_Allele2) == 2 ~ "DNP",
            nchar(Reference_Allele) == 3 &
              nchar(Tumor_Seq_Allele2) == 3 ~ "TNP",
            nchar(Reference_Allele) == 1 &
              Reference_Allele != "-" &
              nchar(Tumor_Seq_Allele2) > 1 ~ "INS",
            nchar(Reference_Allele) > 1 &
              Tumor_Seq_Allele2 != "-" ~ Variant_Classification,
            TRUE ~ "Other"
          ))
        incProgress(1 / 4)
        TMB_tmp = str_split(Data_MAF$TMB, "Muts", simplify = TRUE)[,1]
        TMB_tmp = str_split(TMB_tmp, "mutations", simplify = TRUE)[,1]
        Data_MAF$TMB = as.numeric(TMB_tmp)
        if(length(Data_MAF[is.na(Data_MAF$TMB),]$TMB) > 0){
          Data_MAF[is.na(Data_MAF$TMB),]$TMB = 0
        }
        if(length(Data_MAF[is.na(Data_MAF$Evidence_level),]$Evidence_level) > 0){
          Data_MAF[is.na(Data_MAF$Evidence_level),]$Evidence_level = ""
        }
        incProgress(1 / 4)
        QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), Data_MAF, file="source/variant_data.qs")
      }
    })
    return(Data_MAF)
  })

  Data_case =  reactive({
    clin_tmp = Data_case_raw()
    if(!is.null(input$histology_group_1)){
      clin_tmp = clin_tmp %>% dplyr::mutate(
        症例.基本情報.がん種.OncoTree. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_1 ~ str_replace(str_replace(input$histology_group_1_name, "\\)", ""), ".*\\(",""),
          TRUE ~ 症例.基本情報.がん種.OncoTree.
        ),
        症例.基本情報.がん種.OncoTree..名称. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_1 ~ input$histology_group_1_name,
          TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
        )
        
      )
    }
    if(!is.null(input$histology_group_2)){
      clin_tmp = clin_tmp %>% dplyr::mutate(
        症例.基本情報.がん種.OncoTree. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_2 ~ str_replace(str_replace(input$histology_group_2_name, "\\)", ""), ".*\\(",""),
          TRUE ~ 症例.基本情報.がん種.OncoTree.
        ),
        症例.基本情報.がん種.OncoTree..名称. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_2 ~ input$histology_group_2_name,
          TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
        )
      )
    }
    if(!is.null(input$histology_group_3)){
      clin_tmp = clin_tmp %>% dplyr::mutate(
        症例.基本情報.がん種.OncoTree. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_3 ~ str_replace(str_replace(input$histology_group_3_name, "\\)", ""), ".*\\(",""),
          TRUE ~ 症例.基本情報.がん種.OncoTree.
        ),
        症例.基本情報.がん種.OncoTree..名称. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_3 ~ input$histology_group_3_name,
          TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
        )
      )
    }
    if(!is.null(input$histology_group_4)){
      clin_tmp = clin_tmp %>% dplyr::mutate(
        症例.基本情報.がん種.OncoTree. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_4 ~ str_replace(str_replace(input$histology_group_4_name, "\\)", ""), ".*\\(",""),
          TRUE ~ 症例.基本情報.がん種.OncoTree.
        ),
        症例.基本情報.がん種.OncoTree..名称. = case_when(
          症例.基本情報.がん種.OncoTree..名称. %in% input$histology_group_4 ~ input$histology_group_4_name,
          TRUE ~ 症例.基本情報.がん種.OncoTree..名称.
        )
      )
    }
    clin_tmp <- clin_tmp %>%
      # ① cancer name によるがん種名称の再設定
      { 
        if (!is.null(input$minimum_pts) & input$minimum_pts != 1) {
          Cancername <- clin_tmp %>% 
            distinct(C.CAT調査結果.基本項目.ハッシュID, 症例.基本情報.がん種.OncoTree.) %>% 
            pull(症例.基本情報.がん種.OncoTree.)
          counts <- table(Cancername)
          Cancername <- names(counts[counts >= input$minimum_pts])
          mutate(., 
                 症例.基本情報.がん種.OncoTree..名称. = if_else(症例.基本情報.がん種.OncoTree. %in% Cancername,
                                                     症例.基本情報.がん種.OncoTree..名称.,
                                                     症例.基本情報.がん種.OncoTree.LEVEL1.),
                 症例.基本情報.がん種.OncoTree. = if_else(症例.基本情報.がん種.OncoTree. %in% Cancername,
                                                 症例.基本情報.がん種.OncoTree.,
                                                 症例.基本情報.がん種.OncoTree.LEVEL1.)
          )
        } else .
      } %>%
      # ② 各フィルター処理（NULL チェック付き）
      { if (!is.null(input$histology)) 
        filter(., 症例.基本情報.がん種.OncoTree..名称. %in% input$histology) 
        else . } %>%
      { if (!is.null(input$panel)) 
        filter(., 症例.検体情報.パネル.名称. %in% input$panel) 
        else . } %>%
      { if (!is.null(input$PS)) 
        filter(., 症例.背景情報.ECOG.PS.名称. %in% input$PS) 
        else . } %>%
      { if (!is.null(input$sex)) 
        filter(., 症例.基本情報.性別.名称. %in% input$sex) 
        else . } %>%
      { if (!is.null(input$smoking)) 
        filter(., 症例.背景情報.喫煙歴有無.名称. %in% input$smoking) 
        else . } %>%
      # ③ 登録日のフィルタリング：まず日付変換し、年フィルタ適用
      { 
        if (!is.null(input$year)) {
          . <- mutate(., 症例.管理情報.登録日 = as.Date(症例.管理情報.登録日))
          filter(., (as.POSIXlt(症例.管理情報.登録日)$year + 1900) %in% as.integer(input$year) ) %>%
            mutate(., 症例.管理情報.登録日 = as.character(症例.管理情報.登録日))
        } else .
      } %>%
      # ④ 年齢フィルター（Data_case_raw() の最小・最大と比較して、入力範囲が変化している場合のみ）
      {
        if (!is.null(input$age)) {
          min_age <- min(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE)
          max_age <- max(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE)
          if (any(input$age != c(min_age, max_age))) {
            mutate(., 症例.基本情報.年齢 = replace_na(症例.基本情報.年齢, -1)) %>%
              filter(症例.基本情報.年齢 >= input$age[1] & 症例.基本情報.年齢 <= input$age[2])
          } else .
        } else .
      } %>%
      # ⑤ 年齢を基に若年/高齢フラグの追加
      { if (!is.null(input$mid_age)) 
        mutate(., YoungOld = if_else(症例.基本情報.年齢 <= input$mid_age, "Younger", "Older"))
        else .
      } %>% dplyr::distinct() %>% dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
    return(clin_tmp)
  })

  Data_report =  reactive({
    Data_MAF = Data_report_raw()
    Data_report_TMB = Data_MAF %>%
      dplyr::filter(Hugo_Symbol == "TMB") %>%
      dplyr::distinct(Tumor_Sample_Barcode, TMB, .keep_all = T) %>%
      dplyr::arrange(Tumor_Sample_Barcode) %>%
      dplyr::mutate(
        Evidence_level = case_when(
          TMB < input$TMB_threshold ~ "",
          TMB >= input$TMB_threshold ~ "F",
          TRUE ~ ""
        ),
        amino.acid.change = case_when(
          TMB < input$TMB_threshold ~ "",
          TMB >= input$TMB_threshold ~ "high",
          TRUE ~ ""
        )
      )
    Data_MAF = Data_MAF %>%
      dplyr::arrange(Tumor_Sample_Barcode) %>%
      dplyr::filter(Hugo_Symbol != "TMB") %>%
      dplyr::mutate(Evidence_level = case_when(
        Hugo_Symbol == "MSI" & amino.acid.change == "high" & Evidence_level == "" ~ "F",
        TRUE ~ Evidence_level
      ))
    Data_MAF = rbind(Data_MAF, Data_report_TMB)
    if(!is.null(input$special_gene_annotation)){
      if(input$special_gene_annotation == "This gene is also analyzed for VUS and Benign" &
         !is.null(input$special_gene)){
        Data_MAF = Data_MAF %>% dplyr::mutate(
          Evidence_level = case_when(
            Hugo_Symbol == input$special_gene &
              Evidence_level == "" ~ "F",
            TRUE ~ Evidence_level
          )
        )
      }
    }
    if(!is.null(input$special_gene_independent)){
      if(input$special_gene_independent == "Treat as independent genes in analyses" &
       !is.null(input$special_gene)){
        Data_MAF = Data_MAF %>% dplyr::mutate(
          Hugo_Symbol = case_when(
            Hugo_Symbol == input$special_gene &
              amino.acid.change %in% input$special_gene_mutation_1 ~ paste0(Hugo_Symbol, "_", input$special_gene_mutation_1_name),
            Hugo_Symbol == input$special_gene &
              amino.acid.change %in% input$special_gene_mutation_2 ~ paste0(Hugo_Symbol, "_", input$special_gene_mutation_2_name),
            Hugo_Symbol == input$special_gene ~ paste0(Hugo_Symbol, "_NOS"),
            TRUE ~ Hugo_Symbol
          )
        )
      }
    }
    if(!is.null(input$gene_group_analysis)){
      if(input$gene_group_analysis == "Only cases with mutations in the gene set are analyzed" &
       !is.null(input$gene)){
        if(input$patho == "Only pathogenic muts"){
          ID_geneset = unique((Data_MAF %>% dplyr::filter(
            Hugo_Symbol %in% input$gene & Evidence_level == "F"
          ))$Tumor_Sample_Barcode)
        } else if(input$patho == "All muts"){
          ID_geneset = unique((Data_MAF %>% dplyr::filter(
            Hugo_Symbol %in% input$gene
          ))$Tumor_Sample_Barcode)
        }
        Data_MAF = Data_MAF %>%
          dplyr::filter(Tumor_Sample_Barcode %in%
                          ID_geneset)
      }
      if(input$gene_group_analysis == "Only cases without mutations in the gene set are analyzed" &
         !is.null(input$gene)){
        if(input$patho == "Only pathogenic muts"){
          ID_geneset = unique((Data_MAF %>% dplyr::filter(
            Hugo_Symbol %in% input$gene & Evidence_level == "F"
          ))$Tumor_Sample_Barcode)
        } else if(input$patho == "All muts"){
          ID_geneset = unique((Data_MAF %>% dplyr::filter(
            Hugo_Symbol %in% input$gene
          ))$Tumor_Sample_Barcode)
        }
        Data_MAF = Data_MAF %>%
          dplyr::filter(!Tumor_Sample_Barcode %in% ID_geneset)
        ALL_ID =  (Data_case_raw() %>% dplyr::filter(
          !C.CAT調査結果.基本項目.ハッシュID %in% ID_geneset
        ))$C.CAT調査結果.基本項目.ハッシュID
        DEFF_ID = ALL_ID[!ALL_ID %in% Data_MAF$Tumor_Sample_Barcode]
        if(length(DEFF_ID)>0){
          TEST_ALL = NULL
          TEST = (Data_MAF %>% dplyr::filter(
            Hugo_Symbol == "TMB"
          ))[1,]
          TEST$TMB=0
          TEST$Evidence_level = ""
          TEST$Drug = ""
          for(i in 1:length(DEFF_ID)){
            TEST$Tumor_Sample_Barcode = DEFF_ID[i]
            TEST_ALL = rbind(TEST_ALL, TEST)
          }
          Data_MAF = rbind(Data_MAF, TEST_ALL)
        }
      }
    }
    Data_MAF = Data_MAF %>% dplyr::distinct() %>% dplyr::arrange(Tumor_Sample_Barcode)
    return(Data_MAF)}
  )
  
  observeEvent(input$start_setting, {
    withProgress(message = sample(nietzsche)[1], {
      output$select_organ = renderUI({ 
        pickerInput("organ", "Filter by organ",
                    choices = sort(unique(Data_case_raw()$症例.基本情報.がん種.OncoTree.LEVEL1.)), 
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    selected = sort(unique(Data_case_raw()$症例.基本情報.がん種.OncoTree.LEVEL1.)), multiple = TRUE)
      })
      output$select_histology = renderUI({ 
        pickerInput("histology", "Filter by histology",
                    choices = sort(unique((Data_case_raw() %>% filter(症例.基本情報.がん種.OncoTree.LEVEL1. %in% input$organ))$症例.基本情報.がん種.OncoTree..名称.)), 
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    selected = sort(unique((Data_case_raw() %>% filter(症例.基本情報.がん種.OncoTree.LEVEL1. %in% input$organ))$症例.基本情報.がん種.OncoTree..名称.)), multiple = TRUE)
      })
      output$select_histology_group_1 = renderUI({ 
        pickerInput("histology_group_1", "Histology type 1 to be analyzed (if none, not selected)",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$histology, multiple = TRUE)
      })
      incProgress(1 / 8)
      output$select_histology_group_1_name = renderUI({ 
        pickerInput("histology_group_1_name", "Name for histology type 1",
                    options = list(`actions-box` = TRUE),
                    input$histology_group_1, multiple = FALSE)
      })
      output$select_histology_group_2 = renderUI({ 
        pickerInput("histology_group_2", "Histology type 2 to be analyzed",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$histology[-which(input$histology %in%
                                             input$histology_group_1)], multiple = TRUE)
      })
      output$select_histology_group_2_name = renderUI({ 
        pickerInput("histology_group_2_name", "Name for histology type 2",
                    input$histology_group_2, multiple = FALSE)
      })
      incProgress(1 / 8)
      output$select_histology_group_3 = renderUI({ 
        pickerInput("histology_group_3", "Histology type 3 to be analyzed",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$histology[-which(input$histology %in%
                                             c(input$histology_group_1,
                                               input$histology_group_2))], multiple = TRUE)
      })
      incProgress(1 / 8)
      output$select_histology_group_3_name = renderUI({ 
        pickerInput("histology_group_3_name", "Name for histology type 3",
                    options = list(`actions-box` = TRUE),
                    input$histology_group_3, multiple = FALSE)
      })
      output$select_histology_group_4 = renderUI({ 
        pickerInput("histology_group_4", "Histology type 4 to be analyzed",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$histology[-which(input$histology %in%
                                             c(input$histology_group_1,
                                               input$histology_group_2,
                                               input$histology_group_3))], multiple = TRUE)
      })
      output$select_histology_group_4_name = renderUI({ 
        pickerInput("histology_group_4_name", "Name for histology type 4",
                    options = list(`actions-box` = TRUE),
                    input$histology_group_4, multiple = FALSE)
      })
      output$select_sex = renderUI({ 
        pickerInput("sex", "Filter by sex",
                    choices = sort(unique(Data_case_raw()$症例.基本情報.性別.名称.)), 
                    selected = sort(unique(Data_case_raw()$症例.基本情報.性別.名称.)), multiple = TRUE)
      })
      output$select_minimum_pts = renderUI({ 
        numericInput("minimum_pts", "Minimum patients for each histology",
                     value = 1,
                     min = 1,
                     max = length((Data_case_raw() %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID))$C.CAT調査結果.基本項目.ハッシュID),
                     step = 1)
      })
      incProgress(1 / 8)
      output$select_eps = renderUI({ 
        numericInput("eps", "Distance value for DBSCAN clustering",
                     value = 1.0,
                     min = 0,
                     max = 10,
                     step = 0.1)
      })
      output$select_panel = renderUI({ 
        pickerInput("panel", "Filter by panel",
                    choices = sort(unique(Data_case_raw()$症例.検体情報.パネル.名称.)), 
                    selected = sort(unique(Data_case_raw()$症例.検体情報.パネル.名称.)), multiple = TRUE)
      })
      output$select_year = renderUI({ 
        pickerInput("year", "Filter by test-year",
                    choices = sort(unique(as.POSIXlt(as.Date(Data_case_raw()$症例.管理情報.登録日))$year + 1900)),
                    selected = sort(unique(as.POSIXlt(as.Date(Data_case_raw()$症例.管理情報.登録日))$year + 1900)), multiple = TRUE)
      })
      output$select_age = renderUI({ 
        sliderInput(inputId = "age", label = "Age for analysis",
                    value = c(
                      min(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE),
                      max(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE)),
                    min = min(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE),
                    max = max(Data_case_raw()$症例.基本情報.年齢, na.rm = TRUE),
                    step = 1)
      })
      output$select_mid_age = renderUI({
        tmp = Data_case_raw() %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, 症例.基本情報.年齢)
        sliderInput(inputId = "mid_age", label = "Threshold age for oncoprint",
                    value = round(median(tmp$症例.基本情報.年齢, na.rm = TRUE) / 5) * 5,
                    min = min(tmp$症例.基本情報.年齢, na.rm = TRUE),
                    max = max(tmp$症例.基本情報.年齢, na.rm = TRUE),
                    step = 1)
      })
      incProgress(1 / 8)
      output$select_PS = renderUI({ 
        pickerInput("PS", "Filter by performance status",
                    choices = sort(unique(Data_case_raw()$症例.背景情報.ECOG.PS.名称.)), 
                    selected = sort(unique(Data_case_raw()$症例.背景情報.ECOG.PS.名称.)), multiple = TRUE)
      })
      output$select_smoking = renderUI({ 
        pickerInput("smoking", "Filter by smoking status",
                    choices = sort(unique(Data_case_raw()$症例.背景情報.喫煙歴有無.名称.)), 
                    selected = sort(unique(Data_case_raw()$症例.背景情報.喫煙歴有無.名称.)), multiple = TRUE)
      })
      output$select_gene = renderUI({
        candidate = sort(unique(c(Data_report_raw()$Hugo_Symbol,
                      paste0(input$special_gene, "_", input$special_gene_mutation_1_name),
                      paste0(input$special_gene, "_", input$special_gene_mutation_2_name),
                      paste0(input$special_gene, "_NOS"))))
        candidate = candidate[!candidate %in% c("", "_", "_NOS", paste0(input$special_gene, "_"))]
        candidate = candidate[!is.na(candidate)]
        pickerInput("gene", "Genes of interest (if any)",
                    candidate,
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    #selected = c("MSI", "PMS2"),
                     selected = unique(names(sort(table(Data_report_raw()$Hugo_Symbol),
                                                  decreasing = T)))[1],
                    multiple = TRUE)
      })
      output$select_lolliplot_gene = renderUI({ 
        candidate = sort(unique(c(Data_report_raw()$Hugo_Symbol,
                                  paste0(input$special_gene, "_", input$special_gene_mutation_1_name),
                                  paste0(input$special_gene, "_", input$special_gene_mutation_2_name),
                                  paste0(input$special_gene, "_NOS"))))
        candidate = candidate[!candidate %in% c("", "_", "_NOS", paste0(input$special_gene, "_"))]
        candidate = candidate[!is.na(candidate)]
        pickerInput("lolliplot_gene", "Gene for lolliplot (if any)",
                    options = list(`live-search`=TRUE),
                    choices = candidate,
                    selected = unique(names(sort(table(Data_report_raw()$Hugo_Symbol),
                                                 decreasing = T)))[1],
                    multiple = FALSE)
      })
      output$select_lolliplot_no = renderUI({
        tmp = Data_case_raw() %>% dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, 症例.基本情報.年齢)
        sliderInput(inputId = "lolliplot_no", label = "Threshold mutation count for lolliplot",
                    value = 5,
                    min = 1,
                    max = 100,
                    step = 1)
      })
      
      
      output$select_gene_group_1 = renderUI({ 
        pickerInput("gene_group_1", "Gene-set 1 of interest (if any)",
                    input$gene,
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    multiple = TRUE)
      })
      output$select_gene_group_2 = renderUI({ 
        pickerInput("gene_group_2", "Gene-set 2 of interest (if any)",
                    input$gene[-which(input$gene %in% input$gene_group_1)],
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    multiple = TRUE)
      })
      incProgress(1 / 8)
      output$select_special_gene = renderUI({
        pickerInput("special_gene", "Gene to analyze (if any)",
                    c("", sort(unique(Data_report_raw()$Hugo_Symbol))),
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    multiple = FALSE)
      })
      output$select_special_gene_mutation_1 = renderUI({
        if(!is.null(input$special_gene)){
          pickerInput("special_gene_mutation_1", "Variants 1", sort(unique(
            (Data_report_raw() %>% dplyr::filter(Hugo_Symbol == input$special_gene | (input$special_gene != "" & str_detect(Hugo_Symbol, paste0(input$special_gene, "_")))))$amino.acid.change)),
            options = list(`actions-box` = TRUE, `live-search`=TRUE),
            multiple = TRUE)
        } else {
          pickerInput("special_gene_mutation_1", "Variants 1", NULL,
                      options = list(`actions-box` = TRUE, `live-search`=TRUE),
                      multiple = TRUE)
        }
      })
      output$select_special_gene_mutation_1_name = renderUI({
        if(!is.null(input$special_gene_mutation_1)){
          textInput(inputId = "special_gene_mutation_1_name", label = "Name for variants 1",
                  value = input$special_gene_mutation_1[[1]])
        } else {
          textInput(inputId = "special_gene_mutation_1_name", label = "Name for variants 1",
                    value = NULL)
        }
      })
      output$select_special_gene_mutation_2 = renderUI({
        if(!is.null(input$special_gene)){
          tmp_mut = sort(unique(
            (Data_report_raw() %>% dplyr::filter(Hugo_Symbol == input$special_gene | (input$special_gene != "" & str_detect(Hugo_Symbol, paste0(input$special_gene, "_")))))$amino.acid.change))
          pickerInput("special_gene_mutation_2", "Variants 2", tmp_mut[!tmp_mut %in% input$special_gene_mutation_1],
                      options = list(`actions-box` = TRUE, `live-search`=TRUE),
                      multiple = TRUE)
        } else{
          pickerInput("special_gene_mutation_2", "Variants 2", NULL,
                      options = list(`actions-box` = TRUE, `live-search`=TRUE),
                      multiple = TRUE)
        }
      })
      output$select_special_gene_mutation_2_name = renderUI({
        if(!is.null(input$special_gene_mutation_2)){
          textInput(inputId = "special_gene_mutation_2_name", label = "Name for variants 2",
                  value = input$special_gene_mutation_2[[1]])
        } else{
          textInput(inputId = "special_gene_mutation_2_name", label = "Name for variants 2",
                    value = NULL)
        }
      })
      output$select_gene_no = renderUI({ 
        numericInput(inputId = "gene_no", label = "Gene number for oncoprint",
                     value = min(GENE_NO_THRESHOLD, length(sort(unique(Data_report_raw()$Hugo_Symbol)))),
                     min = max(1, length(input$gene)),
                     max = length(sort(unique(Data_report_raw()$Hugo_Symbol))),
                     step = 1)
      })
      incProgress(1 / 8)
      output$select_patho = renderUI({ 
        radioButtons(
          "patho", "Variants for analysis",
          choices = c("Only pathogenic muts", "All muts"),
          #selected = "All muts")
          selected = "Only pathogenic muts")
      })
      output$select_oncoprint_option = renderUI({ 
        radioButtons(
          "oncoprint_option", "Oncoprint setting",
          choices = c("Sort by mutation frequency", "Sort by pathology", "Sort by age", "Sort by metastasis status at CGP"),
          selected = "Sort by mutation frequency")
      })
      output$select_target_line = renderUI({ 
        pickerInput("target_line", "CTx lines to analyze",
                    choices = c("1","2","3","4", "5~"), 
                    selected =  c("1","2","3","4", "5~"),
                    options = list(`actions-box` = TRUE),
                    multiple = TRUE)
      })
      output$select_gene_group_analysis = renderUI({ 
        radioButtons(
          "gene_group_analysis", "Case selection based on the mutations",
          choices = c("Only cases with mutations in the gene set are analyzed",
                      "Only cases without mutations in the gene set are analyzed",
                      "All cases"),
          selected = "All cases")
      })
      output$select_special_gene_annotation = renderUI({ 
        radioButtons(
          "special_gene_annotation", "Pathological significance of the genes",
          choices = c("Only pathological mutations are included in the analysis",
                      "This gene is also analyzed for VUS and Benign"),
          selected = "Only pathological mutations are included in the analysis")
      })
      output$select_special_gene_independent = renderUI({ 
        radioButtons(
          "special_gene_independent", "Treat specified variants independently?",
          choices = c("Treat as independent genes in analyses",
                      "No"),
          selected = "Treat as independent genes in analyses")
      })
      output$select_RMST_CGP = renderUI({ 
        numericInput("RMST_CGP", "Timing for RMST measuring in survival analysis after CGP (years)",
                     value = 2,
                     min = 1,
                     max = 5,
                     step = 1)
      })
      output$select_RMST_CTx = renderUI({ 
        numericInput("RMST_CTx", "Timing for RMST measuring in survival analysis after CTx (years)",
                     value = 2,
                     min = 1,
                     max = 10,
                     step = 1)
      })
      output$select_RMST_drug = renderUI({ 
        numericInput("RMST_drug", "Timing for RMST measuring in drug analysis (months)",
                     value = 12,
                     min = 1,
                     max = 60,
                     step = 1)
      })
      
      incProgress(1 / 8)
    })
  })
  
  output$select_new_analysis = renderUI({ 
    radioButtons(
      "new_analysis", "Analyze with new dataset",
      choices = c("Yes, input new csv files",
                  "No, use the previous dataset"),
      selected = "No, use the previous dataset")
  })
  output$select_intermediate_file = renderUI({ 
    radioButtons(
      "intermediate_file", "Save intermediate files",
      choices = c("Yes (not recommended)",
                  "No"),
      selected = "No")
  })
  output$select_histology_detail = renderUI({ 
    radioButtons(
      "histology_detail", "Analyze without detailed histology",
      choices = c("Yes, use oncotree 1st level",
                  "No"),
      selected = "No")
  })
  output$select_MSI = renderUI({ 
    radioButtons(
      "MSI", "Analyze MSI-PCR test results",
      choices = c("Yes, additional analysis in survival after CGP and drug response",
                  "No"),
      selected = "No")
  })
  output$select_MMR = renderUI({ 
    radioButtons(
      "MMR", "Analyze MMR-IHC test results",
      choices = c("Yes, additional analysis in survival after CGP and drug response",
                  "No"),
      selected = "No")
  })
  output$select_HER2 = renderUI({ 
    radioButtons(
      "HER2", "Analyze HER2-IHC test results",
      choices = c("Yes, additional analysis in survival after CGP and drug response",
                  "No"),
      selected = "No")
  })
  output$select_fusion = renderUI({ 
    radioButtons(
      "fusion", "How to analyze fusion genes",
      choices = c("Split into two and treat as mutation in each gene",
                  "All fusion genes are named as 'fusion_genes'",
                  "As is"),
      selected = "Split into two and treat as mutation in each gene")
  })
  output$select_amplification = renderUI({ 
    radioButtons(
      "amplification", "How to analyze gene amplification/loss",
      choices = c("Treat as indipendent gene",
                  "Amplification/loss is a kind of mutation"),
      selected = "Amplification/loss is a kind of mutation")
  })
  output$select_T_N = renderUI({ 
    radioButtons(
      "T_N", "Analysis for germline/somatic origin",
      choices = c("Only somatic for T/N panel", "All muts"),
      selected = "All muts")
  })
  output$download_ID_histology = downloadHandler(
    filename = "ID_histology.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/ID_histology.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_test_clinical_data = downloadHandler(
    filename = "sample_clinical.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/sample_clinical.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_treatment_reach_data = downloadHandler(
    filename = "treatment_reach_summary.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/treatment_reach_summary.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_test_report_data = downloadHandler(
    filename = "sample_report.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/sample_report.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_ID_drug_pre_CGP = downloadHandler(
    filename = "ID_drug_pre_CGP.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/ID_drug_pre_CGP.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_ID_drug_post_CGP = downloadHandler(
    filename = "ID_drug_post_CGP.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/ID_drug_post_CGP.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_drug_rename = downloadHandler(
    filename = "table_drug_rename.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/table_drug_rename.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_drug_combination_rename = downloadHandler(
    filename = "table_drug_combination_rename.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/table_drug_rename.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_regimen_rename = downloadHandler(
    filename = "table_regimen_rename.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/table_regimen_rename.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_histology_rename = downloadHandler(
    filename = "table_histology_rename.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/table_histology_rename.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_mutation_rename = downloadHandler(
    filename = "table_mutation_rename.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/table_mutation_rename.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$download_reannotation = downloadHandler(
    filename = "table_reannotation.csv",
    content = function(file) {
      csv_file = read.csv(header = TRUE, file("source/table_reannotation.csv",
                                              encoding='UTF-8-BOM'))
      write_excel_csv(csv_file, file)
    }
  )
  output$select_drug_volcano = renderUI({ 
    radioButtons(
      "drug_volcano", "Only volcano plot analysis",
      choices = c("Yes",
                  "No, all analyses"),
      selected = "No, all analyses")
  })
  output$select_volcano_pathology = renderUI({ 
    radioButtons(
      "volcano_pathology", "Considering pathology in volcano plot analysis",
      choices = c("Yes",
                  "No"),
      selected = "Yes")
  })
  output$select_drug_survuval_CTx = renderUI({ 
    radioButtons(
      "drug_survuval_CTx", "Survival analysis after CTx start date",
      choices = c("Yes",
                  "No"),
      selected = "No")
  })
  output$select_machine_learning = renderUI({ 
    radioButtons(
      "machine_learning", "Perform machine learning in CGP benefit prediction, time consuming",
      choices = c("Yes",
                  "No"),
      selected = "No")
  })
  output$select_importance = renderUI({ 
    radioButtons(
      "importance", "Importance of factors in machine learning",
      choices = c("SHAP value (heavy analysis)",
                  "Importance (very fast)"),
      selected = "Importance (very fast)")
  })
  output$select_clustering = renderUI({ 
    radioButtons(
      "clustering", "Clustering cohort",
      choices = c("Fixed to the 1st analysis",
                  "Vabiable according to selected pathology"),
      selected = "Fixed to the 1st analysis")
  })
  output$select_drug_other = renderUI({ 
    radioButtons(
      "drug_other", "Rename unselected drugs as others",
      choices = c("Yes",
                  "No"),
      selected = "No")
  })
  output$select_TMB_threshold = renderUI({ 
    numericInput("TMB_threshold", "Threshold for TMB",
                 value = 10,
                 min = 1,
                 max = 30,
                 step = 1)
  })
  
  
  Data_cluster_ID = reactive({
    if(is.null(input$patho)){
      return(NULL)
    } else {
      withProgress(message = "Clustering based on gene mutation pattern.", {
        if(input$clustering == "Fixed to the 1st analysis" &
           file.exists("source/Data_mutation_cord.qs")){
          Data_mutation_cord = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE),file="source/Data_mutation_cord.qs")
        } else {
          Data_case_target = Data_case() %>%
            dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = T)
          if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
            Data_case_target = Data_case_target %>%
              dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                              Data_report()$Tumor_Sample_Barcode)
          }
          Data_MAF_target = Data_report() %>%
            dplyr::filter(
              !str_detect(Hugo_Symbol, ",") &
                Hugo_Symbol != "" &
                Evidence_level %in% c("","A","B","C","D","E","F") &
                Variant_Classification != "expression"
            ) %>% 
            dplyr::arrange(desc(Evidence_level)) %>%
            dplyr::distinct(Tumor_Sample_Barcode,
                            Hugo_Symbol,
                            Start_Position,
                            .keep_all = TRUE)
          Data_MAF_target = Data_MAF_target %>%
            dplyr::filter(Tumor_Sample_Barcode %in%
                            Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
          if(input$patho == "Only pathogenic muts"){
            Data_MAF_target = Data_MAF_target %>%
              dplyr::filter(Evidence_level == "F")
          } else {
            Data_MAF_target = Data_MAF_target %>%
              dplyr::filter(!Hugo_Symbol %in% c("TMB", "MSI") | Evidence_level == "F")
          }
          Data_MAF_target = Data_MAF_target  %>%
            dplyr::distinct(Tumor_Sample_Barcode,
                            Hugo_Symbol,
                            .keep_all = TRUE)
          Gene_list = unique(names(sort(table(Data_MAF_target$Hugo_Symbol),
                                        decreasing = T)))
          Data_mutation = data.frame(matrix(0,
                                            nrow=length(Data_case_target$C.CAT調査結果.基本項目.ハッシュID),
                                            ncol=length(Gene_list)))
          colnames(Data_mutation) = Gene_list
          rownames(Data_mutation) = Data_case_target$C.CAT調査結果.基本項目.ハッシュID
          for(i in 1:length(Data_MAF_target$Hugo_Symbol)){
            Data_mutation[Data_MAF_target$Tumor_Sample_Barcode[i],
                          Data_MAF_target$Hugo_Symbol[i]] = 1
          }
          if(length(Data_mutation[is.na(Data_mutation)])> 0){
            Data_mutation[is.na(Data_mutation)] <- 0
          }
          n_neighbors_ = min(15, dim(Data_mutation)[1])
          set.seed(1)
          Data_mutation_umap = umap(Data_mutation, random_state=1, n_neighbors = n_neighbors_)
          Data_mutation_cord = Data_mutation_umap$layout %>% as.data.frame()
          EPS = input$eps
          MINPTS = 3
          set.seed(1)
          dbscan_res_changed <- dbscan(Data_mutation_cord, eps = EPS, minPts = MINPTS)
          if(min(dbscan_res_changed$cluster) == 0){
            dbscan_res_changed$cluster = (dbscan_res_changed$cluster + 1)
          }
          Data_mutation_cord$C.CAT調査結果.基本項目.ハッシュID = rownames(Data_mutation)
          Data_mutation_cord$driver_mutations = rowSums(Data_mutation)
          Data_mutation_cord$cluster = dbscan_res_changed$cluster
          if(input$clustering == "Fixed to the 1st analysis"){
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), Data_mutation_cord, file="source/Data_mutation_cord.qs")
          }
        }
        return(Data_mutation_cord)
      })
    }
  })
  
  observeEvent(input$summary_base, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      incProgress(1 / 8)
      Data_summary = Data_case_target %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
        dplyr::select(
          C.CAT調査結果.基本項目.ハッシュID,
          症例.基本情報.がん種.OncoTree.,
          症例.基本情報.がん種.OncoTree..名称.,
          症例.基本情報.がん種.OncoTree.LEVEL1.,
          症例.基本情報.性別.名称.,
          症例.基本情報.年齢,
          症例.背景情報.診断日,
          症例.管理情報.登録日,
          症例.背景情報.家族歴有無.名称.,
          症例.背景情報.喫煙歴有無.名称.,
          症例.背景情報.喫煙年数,
          症例.背景情報.アルコール多飲有無.名称.,
          症例.背景情報.ECOG.PS.名称.,
          症例.背景情報.重複がん有無.異なる臓器..名称.,
          症例.背景情報.多発がん有無.同一臓器..名称.,
          症例.検体情報.パネル.名称.,
          症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
          症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
          症例.がん種情報.肺.EGFR.名称.,
          症例.がん種情報.肺.ALK融合.名称.,
          症例.がん種情報.肺.ROS1.名称.,
          症例.がん種情報.肺.BRAF.V600..名称.,
          症例.がん種情報.肺.MET遺伝子エクソン14スキッピング変異.名称.,
          症例.がん種情報.肺.KRAS.G12C遺伝子変異.名称.,
          症例.がん種情報.肺.RET融合遺伝子.名称.,
          症例.がん種情報.肺.PD.L1.IHC..名称.,
          症例.がん種情報.乳.HER2.IHC..名称.,
          症例.がん種情報.乳.HER2.FISH..名称.,
          症例.がん種情報.乳.ER.名称.,
          症例.がん種情報.乳.PgR.名称.,
          症例.がん種情報.固形がん.マイクロサテライト不安定性.名称.,
          症例.がん種情報.固形がん.ミスマッチ修復機能欠損.名称.,
          症例.がん種情報.食道.胃.腸.HER2.名称.,
          症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称.,
          症例.がん種情報.登録時転移有無.名称.,
          Lymph_met,
          Brain_met,
          Lung_met,
          Bone_met,
          Liver_met,
          Other_met,
          Not_brain_bone_liver_met,
          YoungOld)
      incProgress(1 / 8)
      if(!"BOWEL" %in% Data_summary$症例.基本情報.がん種.OncoTree.LEVEL1. &
         !"STOMACH" %in% Data_summary$症例.基本情報.がん種.OncoTree.LEVEL1.){
        Data_summary = Data_summary %>% dplyr::select(
          -症例.がん種情報.食道.胃.腸.HER2.名称.,
          -症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称.
        )
      }
      if(!"LUNG" %in% Data_summary$症例.基本情報.がん種.OncoTree.LEVEL1.){
        Data_summary = Data_summary %>% dplyr::select(
          -症例.がん種情報.肺.EGFR.名称.,
          -症例.がん種情報.肺.ALK融合.名称.,
          -症例.がん種情報.肺.ROS1.名称.,
          -症例.がん種情報.肺.BRAF.V600..名称.,
          -症例.がん種情報.肺.MET遺伝子エクソン14スキッピング変異.名称.,
          -症例.がん種情報.肺.KRAS.G12C遺伝子変異.名称.,
          -症例.がん種情報.肺.RET融合遺伝子.名称.,
          -症例.がん種情報.肺.PD.L1.IHC..名称.
        )
      }
      if(!"BREAST" %in% Data_summary$症例.基本情報.がん種.OncoTree.LEVEL1.){
        Data_summary = Data_summary %>% dplyr::select(
          -症例.がん種情報.乳.HER2.IHC..名称.,
          -症例.がん種情報.乳.HER2.FISH..名称.,
          -症例.がん種情報.乳.ER.名称.,
          -症例.がん種情報.乳.PgR.名称.
        )
      }
      Data_summary = Data_summary %>% dplyr::select(
        -症例.基本情報.がん種.OncoTree.LEVEL1.,
      )
      Data_MAF = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","A","B","C","D","E","F") &
            Variant_Classification != "expression"
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
      if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
        Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
      }
      incProgress(1 / 8)

      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      if(input$patho == "Only pathogenic muts"){
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(Evidence_level == "F")
      } else {
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(!Hugo_Symbol %in% c("TMB", "MSI") | Evidence_level == "F")
      }
      incProgress(1 / 8)
      Data_MAF_target = Data_MAF_target %>%
        dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change, .keep_all = T) %>%
        dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
      incProgress(1 / 8)
      if(!is.null(input$gene_group_1)){
        if(!is.null(input$gene_group_2)){
          ID_special_gene_mutation_1 = (Data_MAF_target %>%
                                          dplyr::filter(Hugo_Symbol %in% input$gene_group_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_1, "_"),collapse ="|")) &
                                                          amino.acid.change %in% input$special_gene_mutation_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_1, "_"),collapse ="|")) &
                                                          amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
          ID_special_gene_mutation_2 = (Data_MAF_target %>%
                                          dplyr::filter(Hugo_Symbol %in% input$gene_group_2 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_2, "_"),collapse ="|")) &
                                                          amino.acid.change %in% input$special_gene_mutation_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_2, "_"),collapse ="|")) &
                                                          amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
          Data_summary = Data_summary %>% dplyr::mutate(
            separation_value = case_when(
              C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene_mutation_1 &
                C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene_mutation_2 ~ paste0(paste(input$gene_group_1, sep="/"), " mut, ", paste0(paste(input$gene_group_2, sep="/"), " mut")),
              C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene_mutation_1 ~ paste0(paste(input$gene_group_1, sep="/"), " mut, ", paste0(paste(input$gene_group_2, sep="/"), " WT")),
              C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene_mutation_2 ~ paste0(paste(input$gene_group_1, sep="/"), " WT, ", paste0(paste(input$gene_group_2, sep="/"), " mut")),
              TRUE ~ paste0("No mut in ",  paste(input$gene_group_1, sep="/"), "/", paste(input$gene_group_2, sep="/"))
            )
          )
        } else {
          ID_special_gene_mutation_1 = (Data_MAF_target %>%
                               dplyr::filter(Hugo_Symbol %in% input$gene_group_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_1, "_"),collapse ="|")) &
                                               amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
          Data_summary = Data_summary %>% dplyr::mutate(
            separation_value = case_when(
              C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene_mutation_1 ~ paste0(paste(input$gene_group_1, sep="/"), " mut"),
              TRUE ~ paste0("No mut in ", paste(input$gene_group_1, sep="/"))
            )
          )
        }
      } else if(!input$special_gene == ""){
        ID_special_gene = (Data_MAF_target %>%
                             dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_"))))$Tumor_Sample_Barcode
        ID_special_gene_mutation_1 = (Data_MAF_target %>%
                                        dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                        amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
        ID_special_gene_mutation_2 = (Data_MAF_target %>%
                                        dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                        amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
        Data_summary = Data_summary %>% dplyr::mutate(
          separation_value = case_when(
            C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene_mutation_1 ~ paste0(input$special_gene_mutation_1_name, " in ", input$special_gene),
            C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene_mutation_2 ~ paste0(input$special_gene_mutation_2_name, " in ", input$special_gene),
            C.CAT調査結果.基本項目.ハッシュID %in% ID_special_gene ~ paste0("Other mut in ", input$special_gene),
            TRUE ~ paste0("No mut in ", input$special_gene)
          )
        )
      } else {
        Data_summary = Data_summary %>% dplyr::mutate(
          separation_value = "No mutation pattern")
      }
      incProgress(1 / 8)
      Data_summary_ID = Data_summary$C.CAT調査結果.基本項目.ハッシュID
      Data_summary = Data_summary %>% dplyr::select(
        -C.CAT調査結果.基本項目.ハッシュID)

      Data_summary$Lymph_met = as.character(Data_summary$Lymph_met)
      Data_summary$Brain_met = as.character(Data_summary$Brain_met)
      Data_summary$Lung_met = as.character(Data_summary$Lung_met)
      Data_summary$Bone_met = as.character(Data_summary$Bone_met)
      Data_summary$Liver_met = as.character(Data_summary$Liver_met)
      Data_summary$Other_met = as.character(Data_summary$Other_met)
      Data_summary$Not_brain_bone_liver_met = as.character(Data_summary$Not_brain_bone_liver_met)
      
      Data_summary$症例.基本情報.年齢.診断時 = 
        Data_summary$症例.基本情報.年齢 - 
        ceiling(as.integer(as.Date(Data_summary$症例.管理情報.登録日) - as.Date(Data_summary$症例.背景情報.診断日))/365.25)
      Data_summary$症例.基本情報.年齢.診断時[Data_summary$症例.基本情報.年齢.診断時 < 0] = NA_integer_
      Data_summary = Data_summary %>% dplyr::select(
        -症例.背景情報.診断日,
        -症例.管理情報.登録日
      )
      
      Data_summary = Data_summary %>%
        dplyr::mutate(
          YoungOld = case_when(
            YoungOld == "Younger" ~ paste0("Younger than ", input$mid_age),
            TRUE ~ paste0(input$mid_age, " and older")
          ))
      colnames_tmp = colnames(Data_summary)
      colnames_tmp[colnames_tmp == "症例.基本情報.がん種.OncoTree."] = "Diagnosis"
      colnames_tmp[colnames_tmp == "症例.基本情報.がん種.OncoTree..名称."] = "Diagnosis (OncoTree)"
      colnames_tmp[colnames_tmp == "症例.基本情報.性別.名称."] = "Sex"
      colnames_tmp[colnames_tmp == "症例.基本情報.年齢"] = "Age at CGP"
      colnames_tmp[colnames_tmp == "症例.基本情報.年齢.診断時"] = "Age at diagnosis"
      colnames_tmp[colnames_tmp == "症例.背景情報.家族歴有無.名称."] = "Family cancer history"
      colnames_tmp[colnames_tmp == "症例.背景情報.喫煙歴有無.名称."] = "Smoking history"
      colnames_tmp[colnames_tmp == "症例.背景情報.喫煙年数"] = "Years of smoking"
      colnames_tmp[colnames_tmp == "症例.背景情報.アルコール多飲有無.名称."] = "Alcoholic history"
      colnames_tmp[colnames_tmp == "症例.背景情報.ECOG.PS.名称."] = "Performance status"
      colnames_tmp[colnames_tmp == "症例.背景情報.重複がん有無.異なる臓器..名称."] = "Double cancer in a different organ"
      colnames_tmp[colnames_tmp == "症例.背景情報.多発がん有無.同一臓器..名称."] = "Multiple tumor nodules in the organ"
      colnames_tmp[colnames_tmp == "症例.検体情報.パネル.名称."] = "Cancer gene panel"
      colnames_tmp[colnames_tmp == "症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称."] = "Treatment recommended by expert panel"
      colnames_tmp[colnames_tmp == "症例.EP後レジメン情報.提示された治療薬を投与した.名称."] = "Indicated treatment administered"
      colnames_tmp[colnames_tmp == "症例.がん種情報.登録時転移有無.名称."] = "Any distant metastasis"
      colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
      colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
      colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
      colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
      colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
      colnames_tmp[colnames_tmp == "Other_met"] = "Other organ metastasis"
      colnames_tmp[colnames_tmp == "Not_brain_bone_liver_met"] = "Metastasis other than brain/bone/liver"
      colnames_tmp[colnames_tmp == "症例.がん種情報.肺.EGFR.名称."] = "EGFR IHC in lung"
      colnames_tmp[colnames_tmp == "症例.がん種情報.肺.ALK融合.名称."] = "ALK fusion in lung"
      colnames_tmp[colnames_tmp == "症例.がん種情報.肺.ROS1.名称."] = "ROS1 fusion in lung"
      colnames_tmp[colnames_tmp == "症例.がん種情報.肺.BRAF.V600..名称."] = "BRAF V600E in lung"
      colnames_tmp[colnames_tmp == "症例.がん種情報.肺.MET遺伝子エクソン14スキッピング変異.名称."] = "MET ex14 skipping in lung"
      colnames_tmp[colnames_tmp == "症例.がん種情報.肺.KRAS.G12C遺伝子変異.名称."] = "KRAS G12C in lung"
      colnames_tmp[colnames_tmp == "症例.がん種情報.肺.RET融合遺伝子.名称."] = "RET fusion in lung"
      colnames_tmp[colnames_tmp == "症例.がん種情報.肺.PD.L1.IHC..名称."] = "PD-L1 IHC in lung"
      colnames_tmp[colnames_tmp == "症例.がん種情報.乳.HER2.IHC..名称."] = "HER2 IHC in breast"
      colnames_tmp[colnames_tmp == "症例.がん種情報.乳.HER2.FISH..名称."] = "HER2 FISH in breast"
      colnames_tmp[colnames_tmp == "症例.がん種情報.乳.ER.名称."] = "ER IHC in breast"
      colnames_tmp[colnames_tmp == "症例.がん種情報.乳.PgR.名称."] = "PgR IHC in breast"
      colnames_tmp[colnames_tmp == "症例.がん種情報.固形がん.マイクロサテライト不安定性.名称."] = "MSI PCR"
      colnames_tmp[colnames_tmp == "症例.がん種情報.固形がん.ミスマッチ修復機能欠損.名称."] = "MMR IHC"
      colnames_tmp[colnames_tmp == "症例.がん種情報.食道.胃.腸.HER2.名称."] = "HER2 IHC in stomach/bowel"
      colnames_tmp[colnames_tmp == "症例.がん種情報.食道.胃.腸.HER2遺伝子増幅..ISH法..名称."] = "HER2 FISH in stomach/bowel"
      colnames_tmp[colnames_tmp == "YoungOld"] = paste0(input$mid_age, " and older")

      colnames(Data_summary) = colnames_tmp

      Data_summary$`Any distant metastasis`[is.na(Data_summary$`Any distant metastasis`)] = "Unknown"
      Data_summary$`Any distant metastasis`[Data_summary$`Any distant metastasis` == "あり"] = "Yes"
      Data_summary$`Any distant metastasis`[Data_summary$`Any distant metastasis` == "なし"] = "No"
      Data_summary$`Any distant metastasis`[Data_summary$`Any distant metastasis` == "不明"] = "Unknown"
      if("EGFR IHC in lung" %in% colnames_tmp){
        Data_summary$`EGFR IHC in lung`[is.na(Data_summary$`EGFR IHC in lung`)] = "Unknown"
        Data_summary$`EGFR IHC in lung`[Data_summary$`EGFR IHC in lung` == "陽性"] = "Positive"
        Data_summary$`EGFR IHC in lung`[Data_summary$`EGFR IHC in lung` == "陰性"] = "Negative"
        Data_summary$`EGFR IHC in lung`[Data_summary$`EGFR IHC in lung` == "不明or未検査"] = "Unknown"
        Data_summary$`EGFR IHC in lung`[Data_summary$`EGFR IHC in lung` == "判定不能"] = "Undetermined"
      }
      if("ALK fusion in lung" %in% colnames_tmp){
        Data_summary$`ALK fusion in lung`[is.na(Data_summary$`ALK fusion in lung`)] = "Unknown"
        Data_summary$`ALK fusion in lung`[Data_summary$`ALK fusion in lung` == "陽性"] = "Positive"
        Data_summary$`ALK fusion in lung`[Data_summary$`ALK fusion in lung` == "陰性"] = "Negative"
        Data_summary$`ALK fusion in lung`[Data_summary$`ALK fusion in lung` == "不明or未検査"] = "Unknown"
        Data_summary$`ALK fusion in lung`[Data_summary$`ALK fusion in lung` == "判定不能"] = "Undetermined"
      }
      if("ROS1 fusion in lung" %in% colnames_tmp){
        Data_summary$`ROS1 fusion in lung`[is.na(Data_summary$`ROS1 fusion in lung`)] = "Unknown"
        Data_summary$`ROS1 fusion in lung`[Data_summary$`ROS1 fusion in lung` == "陽性"] = "Positive"
        Data_summary$`ROS1 fusion in lung`[Data_summary$`ROS1 fusion in lung` == "陰性"] = "Negative"
        Data_summary$`ROS1 fusion in lung`[Data_summary$`ROS1 fusion in lung` == "不明or未検査"] = "Unknown"
        Data_summary$`ROS1 fusion in lung`[Data_summary$`ROS1 fusion in lung` == "判定不能"] = "Undetermined"
      }
      if("BRAF V600E in lung" %in% colnames_tmp){
        Data_summary$`BRAF V600E in lung`[is.na(Data_summary$`BRAF V600E in lung`)] = "Unknown"
        Data_summary$`BRAF V600E in lung`[Data_summary$`BRAF V600E in lung` == "陽性"] = "Positive"
        Data_summary$`BRAF V600E in lung`[Data_summary$`BRAF V600E in lung` == "陰性"] = "Negative"
        Data_summary$`BRAF V600E in lung`[Data_summary$`BRAF V600E in lung` == "不明or未検査"] = "Unknown"
        Data_summary$`BRAF V600E in lung`[Data_summary$`BRAF V600E in lung` == "判定不能"] = "Undetermined"
      }
      if("MET ex14 skipping in lung" %in% colnames_tmp){
        Data_summary$`MET ex14 skipping in lung`[is.na(Data_summary$`MET ex14 skipping in lung`)] = "Unknown"
        Data_summary$`MET ex14 skipping in lung`[Data_summary$`MET ex14 skipping in lung` == "陽性"] = "Positive"
        Data_summary$`MET ex14 skipping in lung`[Data_summary$`MET ex14 skipping in lung` == "陰性"] = "Negative"
        Data_summary$`MET ex14 skipping in lung`[Data_summary$`MET ex14 skipping in lung` == "不明or未検査"] = "Unknown"
        Data_summary$`MET ex14 skipping in lung`[Data_summary$`MET ex14 skipping in lung` == "判定不能"] = "Undetermined"
      }
      if("KRAS G12C in lung" %in% colnames_tmp){
        Data_summary$`KRAS G12C in lung`[is.na(Data_summary$`KRAS G12C in lung`)] = "Unknown"
        Data_summary$`KRAS G12C in lung`[Data_summary$`KRAS G12C in lung` == "陽性"] = "Positive"
        Data_summary$`KRAS G12C in lung`[Data_summary$`KRAS G12C in lung` == "陰性"] = "Negative"
        Data_summary$`KRAS G12C in lung`[Data_summary$`KRAS G12C in lung` == "不明or未検査"] = "Unknown"
        Data_summary$`KRAS G12C in lung`[Data_summary$`KRAS G12C in lung` == "判定不能"] = "Undetermined"
      }
      if("RET fusion in lung" %in% colnames_tmp){
        Data_summary$`RET fusion in lung`[is.na(Data_summary$`RET fusion in lung`)] = "Unknown"
        Data_summary$`RET fusion in lung`[Data_summary$`RET fusion in lung` == "陽性"] = "Positive"
        Data_summary$`RET fusion in lung`[Data_summary$`RET fusion in lung` == "陰性"] = "Negative"
        Data_summary$`RET fusion in lung`[Data_summary$`RET fusion in lung` == "不明or未検査"] = "Unknown"
        Data_summary$`RET fusion in lung`[Data_summary$`RET fusion in lung` == "判定不能"] = "Undetermined"
      }
      if("PD-L1 IHC in lung" %in% colnames_tmp){
        Data_summary$`PD-L1 IHC in lung`[is.na(Data_summary$`PD-L1 IHC in lung`)] = "Unknown"
        Data_summary$`PD-L1 IHC in lung`[Data_summary$`PD-L1 IHC in lung` == "陽性"] = "Positive"
        Data_summary$`PD-L1 IHC in lung`[Data_summary$`PD-L1 IHC in lung` == "陰性"] = "Negative"
        Data_summary$`PD-L1 IHC in lung`[Data_summary$`PD-L1 IHC in lung` == "不明or未検査"] = "Unknown"
        Data_summary$`PD-L1 IHC in lung`[Data_summary$`PD-L1 IHC in lung` == "判定不能"] = "Undetermined"
      }
      if("HER2 IHC in breast" %in% colnames_tmp){
        Data_summary$`HER2 IHC in breast`[is.na(Data_summary$`HER2 IHC in breast`)] = "Unknown"
        Data_summary$`HER2 IHC in breast`[Data_summary$`HER2 IHC in breast` == "陽性（3+）"] = "Positive"
        Data_summary$`HER2 IHC in breast`[Data_summary$`HER2 IHC in breast` %in% c("陰性",  "陰性（1+）")] = "Negative"
        Data_summary$`HER2 IHC in breast`[Data_summary$`HER2 IHC in breast` == "境界域（2+）"] = "Intermediate"
        Data_summary$`HER2 IHC in breast`[Data_summary$`HER2 IHC in breast` == "不明or未検査"] = "Unknown"
        Data_summary$`HER2 IHC in breast`[Data_summary$`HER2 IHC in breast` == "判定不能"] = "Undetermined"
      }
      if("HER2 FISH in breast" %in% colnames_tmp){
        Data_summary$`HER2 FISH in breast`[is.na(Data_summary$`HER2 FISH in breast`)] = "Unknown"
        Data_summary$`HER2 FISH in breast`[Data_summary$`HER2 FISH in breast` == "陽性"] = "Positive"
        Data_summary$`HER2 FISH in breast`[Data_summary$`HER2 FISH in breast` == "陰性"] = "Negative"
        Data_summary$`HER2 FISH in breast`[Data_summary$`HER2 FISH in breast` == "不明or未検査"] = "Unknown"
        Data_summary$`HER2 FISH in breast`[Data_summary$`HER2 FISH in breast` == "equivocal"] = "Equivocal"
        Data_summary$`HER2 FISH in breast`[Data_summary$`HER2 FISH in breast` == "判定不能"] = "Undetermined"
      }
      if("ER IHC in breast" %in% colnames_tmp){
        Data_summary$`ER IHC in breast`[is.na(Data_summary$`ER IHC in breast`)] = "Unknown"
        Data_summary$`ER IHC in breast`[Data_summary$`ER IHC in breast` == "陽性"] = "Positive"
        Data_summary$`ER IHC in breast`[Data_summary$`ER IHC in breast` == "陰性"] = "Negative"
        Data_summary$`ER IHC in breast`[Data_summary$`ER IHC in breast` == "不明or未検査"] = "Unknown"
        Data_summary$`ER IHC in breast`[Data_summary$`ER IHC in breast` == "判定不能"] = "Undetermined"
      }
      if("PgR IHC in breast" %in% colnames_tmp){
        Data_summary$`PgR IHC in breast`[is.na(Data_summary$`PgR IHC in breast`)] = "Unknown"
        Data_summary$`PgR IHC in breast`[Data_summary$`PgR IHC in breast` == "陽性"] = "Positive"
        Data_summary$`PgR IHC in breast`[Data_summary$`PgR IHC in breast` == "陰性"] = "Negative"
        Data_summary$`PgR IHC in breast`[Data_summary$`PgR IHC in breast` == "不明or未検査"] = "Unknown"
        Data_summary$`PgR IHC in breast`[Data_summary$`PgR IHC in breast` == "判定不能"] = "Undetermined"
      }
      if("MSI PCR" %in% colnames_tmp){
        Data_summary$`MSI PCR`[is.na(Data_summary$`MSI PCR`)] = "Unknown"
        Data_summary$`MSI PCR`[Data_summary$`MSI PCR` == "陽性"] = "Positive"
        Data_summary$`MSI PCR`[Data_summary$`MSI PCR` == "陰性"] = "Negative"
        Data_summary$`MSI PCR`[Data_summary$`MSI PCR` == "不明or未検査"] = "Unknown"
        Data_summary$`MSI PCR`[Data_summary$`MSI PCR` == "判定不能"] = "Undetermined"
      }
      if("MMR IHC" %in% colnames_tmp){
        Data_summary$`MMR IHC`[is.na(Data_summary$`MMR IHC`)] = "Unknown"
        Data_summary$`MMR IHC`[Data_summary$`MMR IHC` == "dMMR(欠損)"] = "dMMR"
        Data_summary$`MMR IHC`[Data_summary$`MMR IHC` == "pMMR(正常)"] = "pMMR"
        Data_summary$`MMR IHC`[Data_summary$`MMR IHC` == "不明or未検査"] = "Unknown"
        Data_summary$`MMR IHC`[Data_summary$`MMR IHC` == "判定不能"] = "Undetermined"
      }
      if("HER2 IHC in stomach/bowel" %in% colnames_tmp){
        Data_summary$`HER2 IHC in stomach/bowel`[is.na(Data_summary$`HER2 IHC in stomach/bowel`)] = "Unknown"
        Data_summary$`HER2 IHC in stomach/bowel`[Data_summary$`HER2 IHC in stomach/bowel` == "陽性（3+）"] = "Positive"
        Data_summary$`HER2 IHC in stomach/bowel`[Data_summary$`HER2 IHC in stomach/bowel` == "境界域（2+）"] = "Intermediate"
        Data_summary$`HER2 IHC in stomach/bowel`[Data_summary$`HER2 IHC in stomach/bowel` %in% c("陰性",  "陰性（1+）")] = "Negative"
        Data_summary$`HER2 IHC in stomach/bowel`[Data_summary$`HER2 IHC in stomach/bowel` == "不明or未検査"] = "Unknown"
        Data_summary$`HER2 IHC in stomach/bowel`[Data_summary$`HER2 IHC in stomach/bowel` == "判定不能"] = "Undetermined"
      }
      if("HER2 FISH in stomach/bowel" %in% colnames_tmp){
        Data_summary$`HER2 FISH in stomach/bowel`[is.na(Data_summary$`HER2 FISH in stomach/bowel`)] = "Unknown"
        Data_summary$`HER2 FISH in stomach/bowel`[Data_summary$`HER2 FISH in stomach/bowel` == "陽性"] = "Positive"
        Data_summary$`HER2 FISH in stomach/bowel`[Data_summary$`HER2 FISH in stomach/bowel` == "陰性"] = "Negative"
        Data_summary$`HER2 FISH in stomach/bowel`[Data_summary$`HER2 FISH in stomach/bowel` == "equivocal"] = "Equivocal"
        Data_summary$`HER2 FISH in stomach/bowel`[Data_summary$`HER2 FISH in stomach/bowel` == "不明or未検査"] = "Unknown"
        Data_summary$`HER2 FISH in stomach/bowel`[Data_summary$`HER2 FISH in stomach/bowel` == "判定不能"] = "Undetermined"
      }
      Data_summary = Data_summary %>%
        dplyr::mutate(
          Sex = case_when(
            Sex == "女" ~ "Female",
            Sex == "未入力・不明" ~ "Unknown",
            Sex == "男" ~ "Male"
          ),
          `Family cancer history` = case_when(
            `Family cancer history` == "あり" ~ "Yes",
            `Family cancer history` == "不明" ~ "Unknown",
            `Family cancer history` == "なし" ~ "No"
          ),
          `Smoking history` = case_when(
            `Smoking history` == "あり" ~ "Yes",
            `Smoking history` == "不明" ~ "Unknown",
            `Smoking history` == "なし" ~ "No"
          ),
          `Alcoholic history` = case_when(
            `Alcoholic history` == "あり" ~ "Yes",
            `Alcoholic history` == "不明" ~ "Unknown",
            `Alcoholic history` == "なし" ~ "No"
          ),
          `Performance status` = case_when(
            `Performance status` == "不明" ~ "Unknown",
            TRUE ~ `Performance status`
          ),
          `Double cancer in a different organ` = case_when(
            `Double cancer in a different organ` == "あり" ~ "Yes",
            `Double cancer in a different organ` == "不明" ~ "Unknown",
            `Double cancer in a different organ` == "なし" ~ "No"
          ),
          `Multiple tumor nodules in the organ` = case_when(
            `Multiple tumor nodules in the organ` == "あり" ~ "Yes",
            `Multiple tumor nodules in the organ` == "不明" ~ "Unknown",
            `Multiple tumor nodules in the organ` == "なし" ~ "No"
          ),
          `Treatment recommended by expert panel` = case_when(
            `Treatment recommended by expert panel` == "いいえ" ~ "No",
            `Treatment recommended by expert panel` == "はい" ~ "Yes"
          ),
          `Indicated treatment administered` = case_when(
            `Indicated treatment administered` == "投与した" ~ "Yes",
            `Indicated treatment administered` == "投与しなかった" ~ "No"
          )
        )
      output$table_summary_1 <- render_gt({
        Data_summary %>% dplyr::select(
          -`Diagnosis`) %>%
          tbl_summary(
            by = "separation_value",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
                          type = list(all_continuous() ~ "continuous2",
                                      all_dichotomous() ~ "categorical"),
                          digits = all_continuous() ~ 2,
                          missing = "ifany") %>%
          modify_caption("**Patient Characteristics** (N = {N})") %>%
          add_n() %>%
          bold_labels() %>% as_gt()
      })
      incProgress(1 / 8)
      output$table_summary_2 <- render_gt({
        Data_summary %>% dplyr::select(
          -"Diagnosis (OncoTree)",
          -separation_value) %>%
          tbl_summary(
            by = "Diagnosis",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Patient Characteristics, by diagnosis** (N = {N})") %>%
          add_n() %>%
          bold_labels() %>% as_gt()
      })
      Data_summary_table = Data_summary
      Data_summary_table$ID = Data_summary_ID
      output$table_summary_3 = DT::renderDataTable(Data_summary_table,
                                                 server = FALSE,
                                                 filter = 'top', 
                                                 extensions = c('Buttons'), 
                                                 options = list(pageLength = 100, 
                                                                scrollX = TRUE,
                                                                scrollY = "1000px",
                                                                scrollCollapse = TRUE,
                                                                dom="Blfrtip",
                                                                buttons = c('csv', 'excel', 'copy')))
      
      incProgress(1 / 8)
    })
  })

  output$input_shape_llogis = renderUI({ 
    numericInput("shape_llogis", "Shape of log-logistic distribution followed by survival after chemotherapy induction",
                 value = 2,
                 min = 0.1,
                 max = 10,
                 step = 0.1)
  })
  output$input_scale_llogis = renderUI({ 
    numericInput("scale_llogis", "Scale of log-logistic distribution followed by survival after chemotherapy induction",
                 value = 3,
                 min = 0.1,
                 max = 10,
                 step = 0.1)
  })
  output$input_shape_CGP = renderUI({ 
    numericInput("shape_CGP", "Shape of log-logistic distribution followed by survival after CGP testing",
                 value = 3,
                 min = 0.1,
                 max = 10,
                 step = 0.1)
  })
  output$input_scale_CGP = renderUI({ 
    numericInput("scale_CGP", "Scale of log-logistic distribution followed by survival after CGP testing",
                 value = 1,
                 min = 0.1,
                 max = 10,
                 step = 0.1)
  })
  output$input_shape_weibull = renderUI({ 
    numericInput("shape_weibull", "Shape of Weibull distribution followed by the timing of CGP",
                 value = 2,
                 min = 0.1,
                 max = 10,
                 step = 0.1)
  })
  output$input_scale_weibull = renderUI({ 
    numericInput("scale_weibull", "Scale of Weibull distribution followed by the timing of CGP",
                 value = 3,
                 min = 0.1,
                 max = 10,
                 step = 0.1)
  })
  output$input_censor_rate = renderUI({ 
    numericInput("censor_rate", "Censoring rate after CGP",
                 value = 0.5,
                 min = 0,
                 max = 1,
                 step = 0.1)
  })
  observeEvent(input$survival_simuration, {
    withProgress(message = sample(nietzsche)[1], {
      library(cmdstanr)
      shape_llogis = input$shape_llogis
      median_os = 365.25*input$scale_llogis
      shape_CGP = input$shape_CGP
      scale_CGP = 365.25*input$scale_CGP
      shape_followup = input$shape_weibull
      scale_followup = 365.25*input$scale_weibull
      patient_no = 1000
      censor_rate = input$censor_rate
      immortal_length = 90

      Data_survival = as.data.frame(paste0("Patient_", 1:patient_no))
      colnames(Data_survival) = "ID"
      
      Data_survival$true_OS = qllogis(((patient_no-1):0)/patient_no, shape = shape_llogis, scale = median_os, log = FALSE)
      #Data_survival$true_OS = qweibull(((patient_no-1):0)/patient_no, shape = shape_weibull, scale = median_os/(log(2)^(1/shape_weibull)), log = FALSE)
      
      Data_survival$true_OS = ifelse(Data_survival$true_OS>=365.25*10, 365.25*10, Data_survival$true_OS)
      Data_survival$true_censor = 1
      
      #Data_survival$followup_date = rllogis(patient_no, shape = shape_followup, scale = scale_followup)
      Data_survival$followup_date = rweibull(patient_no, shape = shape_followup, scale = scale_followup/(log(2)^(1/shape_followup)))
      Data_survival$time_palliative_enroll = Data_survival$true_OS - rllogis(patient_no, shape = shape_CGP, scale = scale_CGP)
      Data_survival$time_enroll_final = Data_survival$true_OS - Data_survival$time_palliative_enroll
      
      #mean_survival = sum(Data_survival$time_enroll_final) / patient_no
      Data_survival$censor_date = runif(patient_no) * Data_survival$time_enroll_final
      Data_survival$censor = as.integer( runif(patient_no, min = 0, max = 1) + censor_rate)
      Data_survival$censor = 1 - ifelse(Data_survival$censor > 1, 1, Data_survival$censor)
      Data_survival$time_enroll_final = ifelse(Data_survival$censor == 0, Data_survival$censor_date, Data_survival$time_enroll_final)
      
      Data_survival$CGP_undergo = ifelse(Data_survival$time_enroll_final > 90 &
                                           Data_survival$time_enroll_final < Data_survival$followup_date &
                                           Data_survival$time_palliative_enroll > 0 &
                                           Data_survival$time_palliative_enroll < Data_survival$true_OS, 1, 0)
      
      Data_survival$observed_OS = Data_survival$time_palliative_enroll + Data_survival$time_enroll_final
      
      
      Data_survival_curve = Data_survival %>% dplyr::filter(CGP_undergo == 1 & observed_OS > immortal_length)
      if(sum(Data_survival_curve$censor) > 0){
        fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_curve)
        time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
        survival_rate_10 = summary(fit,times = time_10)[[6]]
        survival_rate_90 = summary(fit,times = max(Data_survival_curve$time_enroll_final)*0.9)[[6]]
      } else {
        time_10 = 90
        survival_rate_10 = 0
        survival_rate_90 = 1
      }
      time_90 = max(time_10 + 1, max(Data_survival_curve$time_enroll_final)*0.9)
      Data_survival_tmp = Data_survival_curve %>% dplyr::filter(
        time_enroll_final >= time_10 &
          time_enroll_final <= time_90)
      Data_survival_tmp$time_enroll_final =
        Data_survival_tmp$time_enroll_final - (time_10 - 1)
      Data_survival_tmp2 = Data_survival_curve %>% dplyr::filter(
        time_enroll_final > time_90) %>%
        dplyr::arrange(time_enroll_final)
      Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
      idx <- 1:nrow(Data_survival_tmp2)
      Data_drug_survival_1 = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
      
      if(sum(Data_survival_curve$censor) > 0){
        fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_curve)
        time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
        survival_rate_10 = summary(fit,times = time_10)[[6]]
        survival_rate_90 = summary(fit,times = max(Data_survival_curve$time_palliative_enroll)*0.9)[[6]]
      } else {
        time_10 = 90
        survival_rate_10 = 0
        survival_rate_90 = 1
      }
      time_90 = max(time_10 + 1, max(Data_survival_curve$time_palliative_enroll)*0.9)
      Data_survival_tmp5 = Data_survival_curve %>% dplyr::filter(
        time_palliative_enroll >= time_10 &
          time_palliative_enroll <= time_90)
      Data_survival_tmp5$time_palliative_enroll =
        Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
      Data_survival_tmp6 = Data_survival_curve %>% dplyr::filter(
        time_palliative_enroll > time_90) %>%
        dplyr::arrange(time_palliative_enroll)
      Data_survival_tmp6$time_palliative_enroll = time_90 - (time_10 - 1)
      idx <- 1:nrow(Data_survival_tmp6)
      Data_drug_survival_2 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
      
      Median_entry = median(Data_survival_tmp5$time_palliative_enroll)

      Data_drug_survival_1 = Data_drug_survival_1 %>%
        dplyr::mutate(delay_1 = case_when(
          Data_drug_survival_1$time_palliative_enroll <= Median_entry ~ "SPT to entry early",
          TRUE ~ "SPT to entry later"))

      stan_weibull_survival_model_data <-
        list(
          Median = as.integer(Median_entry),
          Nobs_early = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later"),
          Nobs_late = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later"),
          Ncen_early = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later"),
          Ncen_late = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later"),
          M_bg = 1,
          Nexp = length(Data_drug_survival_2$time_palliative_enroll),
          ybef_exp = Data_drug_survival_2$time_palliative_enroll,
          yobs_early = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later"],
          yobs_late = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later"],
          ycen_early = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later"],
          ycen_late = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later"]
        )
      
      stan_model_path <- "source/Stan_code_loglog_weibull.stan"
      stan_model_compiled <- cmdstan_model(stan_model_path)
      stan_weibull_survival_model_fit <- stan_model_compiled$sample(
        data = stan_weibull_survival_model_data,
        seed = 1234,
        chains = 4,
        parallel_chains = min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE)),
        iter_sampling = 2000,  # (iter-warmup)
        iter_warmup = 1000,
        thin = 1,
        refresh = 500
      )
      incProgress(1 / 3)
      
      stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
      stan_weibull_survival_model_draws_total_exp <-
        stan_weibull_survival_model_draws %>%
        dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_total")) %>%
        tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_exp_total")) %>%
        dplyr::select(-key) 
      stan_weibull_survival_model_draws_bef_exp <-
        stan_weibull_survival_model_draws %>%
        dplyr::select(.chain, .iteration, .draw, starts_with("y_exptmp")) %>%
        tidyr::gather(key = key, value = yhat_bef_exp, starts_with("y_exptmp")) %>%
        dplyr::select(-key) 
      stan_weibull_survival_model_draws_aft_exp <-
        stan_weibull_survival_model_draws %>%
        dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_uncens")) %>%
        tidyr::gather(key = key, value = yhat_aft_exp, starts_with("yhat_exp_uncens")) %>%
        dplyr::select(-key) 
      stan_weibull_survival_model_draws_ear <-
        stan_weibull_survival_model_draws %>%
        dplyr::select(.chain, .iteration, .draw, starts_with("yhat_early")) %>%
        tidyr::gather(key = key, value = yhat_ear, starts_with("yhat_early")) %>%
        dplyr::select(-key) 
      stan_weibull_survival_model_draws_lat <-
        stan_weibull_survival_model_draws %>%
        dplyr::select(.chain, .iteration, .draw, starts_with("yhat_late")) %>%
        tidyr::gather(key = key, value = yhat_lat, starts_with("yhat_late")) %>%
        dplyr::select(-key) 
      
      median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
      median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
      
      median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
      yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
      yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival_curve$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_survival_curve$censor))]
      
      Data_survival_curve$left_adj = yhat_weibull_sort_average_total_exp
      Data_survival_curve$left_adj[Data_survival_curve$left_adj > 10000] = 10000
      
      independence_check = with(Data_survival_curve, cKendall(
        trun = time_palliative_enroll,
        obs = observed_OS,
        delta = censor,
        trans = "linear"))
      
      simulation_fit = survfit(Surv(event = rep(1,length(Data_survival_curve$left_adj)),
                                      time = left_adj) ~ 1, 
                                 data = Data_survival_curve)
      survival_fit_true = survfit(Surv(true_OS, true_censor) ~ 1, data = Data_survival %>% dplyr::filter(true_OS > immortal_length))
      #survival_fit_observed = survfit(Surv(observed_OS, censor) ~ 1, data = Data_survival %>% dplyr::filter(true_OS > immortal_length & observed_OS > 0))
      survival_fit_CGP = survfit(Surv(observed_OS, censor) ~ 1, data = Data_survival %>% dplyr::filter(CGP_undergo == 1 & observed_OS > immortal_length))
      survival_fit_pre_CGP = survfit(Surv(time_palliative_enroll, true_censor) ~ 1, data = Data_survival %>% dplyr::filter(CGP_undergo == 1 & observed_OS > immortal_length))
      survival_fit_post_CGP = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival %>% dplyr::filter(CGP_undergo == 1 & observed_OS > immortal_length))
      survival_fit_CGP_risk_set = survfit(Surv(time = time_palliative_enroll,time2 = observed_OS, censor) ~ 1, data = Data_survival %>% dplyr::filter(CGP_undergo == 1 & observed_OS > immortal_length))
      g = ggsurvplot(
          fit = list(survival_fit_true = survival_fit_true,
                     simulation_fit = simulation_fit,
                     survival_fit_CGP = survival_fit_CGP,
                     survival_fit_pre_CGP = survival_fit_pre_CGP,
                     survival_fit_post_CGP = survival_fit_post_CGP,
                     survival_fit_CGP_risk_set = survival_fit_CGP_risk_set),
          combine = TRUE,
          # data = Data_survival,
          xlab = "simulation data, time from treatment start date (months)",
          ylab = "Survival Probability",
          censor = TRUE,
          conf.int = FALSE,
          surv.scale = "percent",
          font.title = 8,
          font.subtitle = 8,
          font.main = 8,
          font.submain = 8,
          font.caption = 8,
          font.legend = 8,
          cumevents = FALSE,
          cumcensor = FALSE,
          xscale = "d_m",
          pval = FALSE,
          surv.median.line = "hv",
          palette = "Dark2",
          risk.table = TRUE,
          risk.table.y.text = FALSE,
          tables.theme = clean_theme(),
          legend = c(0.8,0.90),
          xlim = c(0,2000),
          break.x.by = 365.25 * 2.5,
          legend.labs = c("Ideal data: all patients without censoring",
                          "Estimated survival with bias adjustment method",
                          "Conventional Kaplan-Meier estimater",
                          "Suvival before CGP",
                          "Suvival after CGP",
                          "Risk-set adjusted survival")
      ) +   
        labs(title = paste0("Toy survival data with left-truncation bias, cKendall tau=",
                            format(independence_check$PE,digits = 3),
                           ", Median OS: Ideal data, ",
                           round(summary(survival_fit_true)$table[[7]] / 365.25 * 12, digits = 1)," (",
                           round(summary(survival_fit_true)$table[[8]] / 365.25 * 12, digits = 1),"-",
                           round(summary(survival_fit_true)$table[[9]] / 365.25 * 12, digits = 1),"), months (95%CI)"),
             subtitle = paste0("bias-corrected, ",
                               round(summary(simulation_fit)$table[[7]] / 365.25 * 12, digits = 1)," (",
                               round(summary(simulation_fit)$table[[8]] / 365.25 * 12, digits = 1),"-",
                               round(summary(simulation_fit)$table[[9]] / 365.25 * 12, digits = 1),"), ",
                               "risk-set adjusted, ",
                               round(summary(survival_fit_CGP_risk_set)$table[[7]] / 365.25 * 12, digits = 1)," (",
                               round(summary(survival_fit_CGP_risk_set)$table[[8]] / 365.25 * 12, digits = 1),"-",
                               round(summary(survival_fit_CGP_risk_set)$table[[9]] / 365.25 * 12, digits = 1),"), ",
                               "conventional KM estimator, ",
                               round(summary(survival_fit_CGP)$table[[7]] / 365.25 * 12, digits = 1), " (",
                               round(summary(survival_fit_CGP)$table[[8]] / 365.25 * 12, digits = 1),"-",
                               round(summary(survival_fit_CGP)$table[[9]] / 365.25 * 12, digits = 1),")"))
      g$table <- g$table + theme(plot.title = element_blank(),
                                           plot.subtitle = element_blank())
      incProgress(1 / 3)
      
      
      output$figure_bias_correction_simulation = renderPlot({
        g
      })
      
      output$text_bias_correction_simulation = renderText({
      paste0("Median survival after chemotherapy induction is ", input$scale_llogis, " years",
             "\nSurvival after induction of chemotherapy follows a log-logistic distribution with shape=", input$shape_llogis,", scale=", input$scale_llogis, " years",
             "\nFinal follow-up date follows a log-logistic distribution with shape=", input$shape_CGP, ", scale=", input$scale_CGP, " years",
             "\nMaxcimum follow-up period after CGP testing follows a Weibull distribution with shape=", input$shape_weibull, ", scale=", input$scale_weibull, " years for remaining survival",
             "\nPatients surviving less than 3 months are not offered chemotherapy",
             "\nPatients with survival less than 3 months do not undergo CGP testing\n",
             input$censor_rate*100, "% of cases are right-censored at random timing after CGP testing"
          )
      })
    })
  })
  
  observeEvent(input$inter_panel, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      incProgress(1 / 8)
      Data_summary = Data_case_target %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,.keep_all=TRUE) %>%
        dplyr::select(
          C.CAT調査結果.基本項目.ハッシュID,
          症例.基本情報.がん種.OncoTree.,
          症例.検体情報.パネル.名称.
      )
      colnames(Data_summary) = c("Tumor_Sample_Barcode","Cancername","Panel")
      incProgress(1 / 8)
      Data_MAF = Data_report() %>%
        dplyr::select(Tumor_Sample_Barcode,
                      Hugo_Symbol,
                      Chromosome,
                      Start_Position,
                      End_Position,
                      Reference_Allele,
                      Tumor_Seq_Allele2,
                      Evidence_level,
                      VAF,
                      Variant_Classification,
                      TMB)
      Data_MAF_unique = Data_MAF %>%
        dplyr::select(Tumor_Sample_Barcode,TMB) %>%
        dplyr::distinct()
      Data_MAF = Data_MAF %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","F") &
            Variant_Classification %in% c("small_scale_variant,", "nonsense_frameshift")
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
      incProgress(1 / 8)
      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      Data_MAF_target$VAF = as.numeric(str_replace(Data_MAF_target$VAF, "\\ \\(.*\\)", ""))
      incProgress(1 / 8)
      Data_MAF_target = Data_MAF_target %>%
        dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol, Start_Position, .keep_all = T) %>%
        dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, Start_Position) %>%
        dplyr::select(Tumor_Sample_Barcode, Evidence_level, VAF)
      Data_MAF_target = Data_MAF_target %>% left_join(Data_summary, by="Tumor_Sample_Barcode")
      Data_summary = Data_summary %>% left_join(Data_MAF_unique, by="Tumor_Sample_Barcode")
      incProgress(1 / 8)
      Diseases = sort(unique(Data_summary$Cancername))
      Summary_Table = data.frame(Diseases)
      Summary_Table$Pts_All = 0
      Summary_Table$Pts_F1 = 0
      Summary_Table$Pts_F1L = 0
      Summary_Table$Pts_NOP = 0
      Summary_Table$Pts_GMT = 0
      Summary_Table$TMB_All = 0
      Summary_Table$TMB_F1 = 0
      Summary_Table$TMB_F1L = 0
      Summary_Table$TMB_NOP = 0
      Summary_Table$TMB_GMT = 0
      Summary_Table$AAC_All = 0
      Summary_Table$AAC_F1 = 0
      Summary_Table$AAC_F1L = 0
      Summary_Table$AAC_NOP = 0
      Summary_Table$AAC_GMT = 0
      Summary_Table$PG_All = 0
      Summary_Table$PG_F1 = 0
      Summary_Table$PG_F1L = 0
      Summary_Table$PG_NOP = 0
      Summary_Table$PG_GMT = 0
      for(i in 1:length(Diseases)){
        Summary_Table$Pts_All[i] =
          length((Data_summary %>%
                 dplyr::filter(Cancername == Diseases[i]))$TMB)
        Summary_Table$Pts_F1[i] =
          length((Data_summary %>%
                 dplyr::filter(Cancername == Diseases[i] & Panel == "FoundationOne CDx"))$TMB)
        Summary_Table$Pts_F1L[i] =
          length((Data_summary %>%
                 dplyr::filter(Cancername == Diseases[i] & Panel == "F1Liquid CDx"))$TMB)
        Summary_Table$Pts_NOP[i] =
          length((Data_summary %>%
                 dplyr::filter(Cancername == Diseases[i] & Panel == "NCC OncoPanel"))$TMB)
        Summary_Table$Pts_GMT[i] =
          length((Data_summary %>%
                 dplyr::filter(Cancername == Diseases[i] & Panel == "GenMineTOP"))$TMB)
        Summary_Table$TMB_All[i] =
          mean((Data_summary %>%
                    dplyr::filter(Cancername == Diseases[i]))$TMB,na.rm = T)
        Summary_Table$TMB_F1[i] =
          mean((Data_summary %>%
                 dplyr::filter(Cancername == Diseases[i] & Panel == "FoundationOne CDx"))$TMB,na.rm = T)
        Summary_Table$TMB_F1L[i] =
          mean((Data_summary %>%
                 dplyr::filter(Cancername == Diseases[i] & Panel == "F1Liquid CDx"))$TMB,na.rm = T)
        Summary_Table$TMB_NOP[i] =
          mean((Data_summary %>%
                 dplyr::filter(Cancername == Diseases[i] & Panel == "NCC OncoPanel"))$TMB,na.rm = T)
        Summary_Table$TMB_GMT[i] =
          mean((Data_summary %>%
                 dplyr::filter(Cancername == Diseases[i] & Panel == "GenMineTOP"))$TMB,na.rm = T)
        Summary_Table$AAC_All[i] =
          fun_zero(length((Data_MAF_target %>% dplyr::filter(Cancername == Diseases[i]))$VAF),
                   Summary_Table$Pts_All[i])
        Summary_Table$AAC_F1[i] =
          fun_zero(length((Data_MAF_target %>% dplyr::filter(Cancername == Diseases[i] & Panel == "FoundationOne CDx"))$VAF),
                   Summary_Table$Pts_F1[i])
        Summary_Table$AAC_F1L[i] =
          fun_zero(length((Data_MAF_target %>% dplyr::filter(Cancername == Diseases[i] & Panel == "F1Liquid CDx"))$VAF),
                   Summary_Table$Pts_F1L[i])
        Summary_Table$AAC_NOP[i] =
          fun_zero(length((Data_MAF_target %>% dplyr::filter(Cancername == Diseases[i] & Panel == "NCC OncoPanel"))$VAF),
                   Summary_Table$Pts_NOP[i])
        Summary_Table$AAC_GMT[i] =
          fun_zero(length((Data_MAF_target %>% dplyr::filter(Cancername == Diseases[i] & Panel == "GenMineTOP"))$VAF),
                   Summary_Table$Pts_GMT[i])
        Summary_Table$PG_All[i] =
          fun_zero(length((Data_MAF_target %>% dplyr::filter(Cancername == Diseases[i] & Evidence_level == "F"))$VAF),
                   Summary_Table$Pts_All[i])
        Summary_Table$PG_F1[i] =
          fun_zero(length((Data_MAF_target %>% dplyr::filter(Cancername == Diseases[i] & Panel == "FoundationOne CDx" & Evidence_level == "F"))$VAF),
                   Summary_Table$Pts_F1[i])
        Summary_Table$PG_F1L[i] =
          fun_zero(length((Data_MAF_target %>% dplyr::filter(Cancername == Diseases[i] & Panel == "F1Liquid CDx" & Evidence_level == "F"))$VAF),
                   Summary_Table$Pts_F1L[i])
        Summary_Table$PG_NOP[i] =
          fun_zero(length((Data_MAF_target %>% dplyr::filter(Cancername == Diseases[i] & Panel == "NCC OncoPanel" & Evidence_level == "F"))$VAF),
                   Summary_Table$Pts_NOP[i])
        Summary_Table$PG_GMT[i] =
          fun_zero(length((Data_MAF_target %>% dplyr::filter(Cancername == Diseases[i] & Panel == "GenMineTOP" & Evidence_level == "F"))$VAF),
                   Summary_Table$Pts_GMT[i])
      }
      for(i in 2:length(Summary_Table[1,])){
        Summary_Table[,i] = as.integer(Summary_Table[,i] * 100) / 100
      }
      output$figure_inter_panel = renderPlot({
        g1 = ggplot(Data_MAF_target, aes(x=VAF, fill=Panel)) +
          geom_density(alpha=.2) +
          ggtitle("Distribution of variant allele frequency for each panel") +
          theme_classic()
        g2 = gg_empty()
        grid.arrange(g1, g2, ncol = 1)
      })
      output$table_inter_panel = DT::renderDataTable(Summary_Table,
                                                    server = FALSE,
                                                    filter = 'top', 
                                                    extensions = c('Buttons'), 
                                                    options = list(pageLength = 100, 
                                                                   scrollX = TRUE,
                                                                   scrollY = "1000px",
                                                                   scrollCollapse = TRUE,
                                                                   dom="Blfrtip",
                                                                   buttons = c('csv', 'copy')))
      incProgress(1 / 13)
    })
  })

  observeEvent(input$oncoprint, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      Data_MAF = Data_report() %>%
       dplyr::filter(
         !str_detect(Hugo_Symbol, ",") &
         Hugo_Symbol != "" &
         Evidence_level %in% c("","A","B","C","D","E","F") &
         Variant_Classification != "expression"
       ) %>% 
       dplyr::arrange(desc(Evidence_level)) %>%
       dplyr::distinct(Tumor_Sample_Barcode,
                       Hugo_Symbol,
                       Start_Position,
                       .keep_all = TRUE)
     if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
       Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
     }

     incProgress(1 / 8)
     
     Data_MAF_target = Data_MAF %>%
       dplyr::filter(Tumor_Sample_Barcode %in%
                       Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
     if(input$patho == "Only pathogenic muts"){
       Data_MAF_target = Data_MAF_target %>%
         dplyr::filter(Evidence_level == "F")
     } else {
       Data_MAF_target = Data_MAF_target %>%
         dplyr::filter(!Hugo_Symbol %in% c("TMB", "MSI") | Evidence_level == "F")
     }

     incProgress(1 / 8)
     Data_MAF_target_tmp = Data_MAF_target %>%
       dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change, .keep_all = T) %>%
       dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
     Data_MAF_target = Data_MAF_target_tmp #%>%
     # dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol, .keep_all = T) %>%
     # dplyr::arrange(Hugo_Symbol, Tumor_Sample_Barcode)

     # ID__ = unique((Data_MAF_target_tmp %>% dplyr::filter(Hugo_Symbol == "MSI"))$Tumor_Sample_Barcode)
     # write_csv(Data_MAF_target_tmp %>% dplyr::filter(Hugo_Symbol == "PMS2"), file = "~/Desktop/PMS2_MSI.csv")
     # write_csv(Data_MAF_target_tmp %>% dplyr::filter(Tumor_Sample_Barcode %in% ID__ & Hugo_Symbol == "PMS2"), file = "~/Desktop/PMS2_MSI_high.csv")
     
     output$figure_oncoprint = renderPlot({
       gene = input$gene
       gene_no = input$gene_no
       oncogenic_genes = names(sort(table(Data_MAF_target$Hugo_Symbol),decreasing = T))
       gene_group_1 = input$gene_group_1
       gene_group_2 = input$gene_group_2
       oncoprint_option = input$oncoprint_option
       mid_age = input$mid_age
       withProgress(message = sample(nietzsche)[1], {
         color_mut = c("amplification" = "red",
                 "loss" = "blue",
                 "nonsense_frameshift" = "green4",
                 "TMB_MSI_high" = "orange",
                 "exon_skipping" = "red4",
                 "small_scale_variant" = "green",
                 "duplication" = "pink",
                 "rearrangement" = "purple",
                 "fusion" = "gray",
                 "truncation" = "black",
                 "inversion" = "coral2",
                 "deletion" = "yellow")
         height_ratios <- c(
           amplification = 1.0,
           loss = 0.95,
           nonsense_frameshift = 0.9,
           TMB_MSI_high = 0.85,
           exon_skipping = 0.8,
           small_scale_variant = 0.75,
           duplication = 0.7,
           rearrangement = 0.65,
           fusion = 0.6,
           truncation = 0.55,
           inversion = 0.45,
           deletion = 0.5
         )
         alter_fun = list(
           background = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0.5, "mm"), h - unit(0.5, "mm"), gp = gpar(fill = "#F4F4F4", col = NA))
           },
           amplification = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["amplification"],
                       gp = gpar(fill = color_mut["amplification"], col = NA))
           },
           loss = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["loss"],
                       gp = gpar(fill = color_mut["loss"], col = NA))
           },
           nonsense_frameshift = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["nonsense_frameshift"],
                       gp = gpar(fill = color_mut["nonsense_frameshift"], col = NA))
           },
           TMB_MSI_high = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["TMB_MSI_high"],
                       gp = gpar(fill = color_mut["TMB_MSI_high"], col = NA))
           },
           exon_skipping = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["exon_skipping"],
                       gp = gpar(fill = color_mut["exon_skipping"], col = NA))
           },
           small_scale_variant = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["small_scale_variant"],
                       gp = gpar(fill = color_mut["small_scale_variant"], col = NA))
           },
           duplication = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["duplication"],
                       gp = gpar(fill = color_mut["duplication"], col = NA))
           },
           rearrangement = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["rearrangement"],
                       gp = gpar(fill = color_mut["rearrangement"], col = NA))
           },
           fusion = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["fusion"],
                       gp = gpar(fill = color_mut["fusion"], col = NA))
           },
           truncation = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["truncation"],
                       gp = gpar(fill = color_mut["truncation"], col = NA))
           },
           inversion = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["inversion"],
                       gp = gpar(fill = color_mut["inversion"], col = NA))
           },
           deletion = function(x, y, w, h) {
             grid.rect(x, y, w - unit(0, "pt"), h * height_ratios["deletion"],
                       gp = gpar(fill = color_mut["deletion"], col = NA))
           }
         )
         Data_oncoprint = Data_case_target %>%
           dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                         症例.基本情報.がん種.OncoTree.,
                         YoungOld,
                         Lymph_met,
                         Brain_met,
                         Lung_met,
                         Bone_met,
                         Liver_met,
                         Other_met) %>%
           dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all=TRUE) %>%
           dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
         gene = gene[gene %in% Data_MAF_target$Hugo_Symbol]
         if(length(gene) > 0){
           oncogenic_genes = unique(c(gene, oncogenic_genes))
         }
         gene_no = min(length(oncogenic_genes),
                       gene_no)
         mat_oncoprint = matrix("",
                                nrow = length(Data_oncoprint$症例.基本情報.がん種.OncoTree.),
                                ncol = gene_no)
         rownames(mat_oncoprint) = Data_oncoprint$C.CAT調査結果.基本項目.ハッシュID
         colnames(mat_oncoprint) = oncogenic_genes[1:gene_no]
         Data_TMB = (Data_MAF_target %>%
                       dplyr::distinct(Tumor_Sample_Barcode, TMB))
         colnames(Data_TMB) = c("C.CAT調査結果.基本項目.ハッシュID", "TMB")
         Data_oncoprint = left_join(Data_oncoprint,
                                    Data_TMB,
                                    by = "C.CAT調査結果.基本項目.ハッシュID")
         Data_TMB = Data_oncoprint$TMB
         Data_TMB[is.na(Data_TMB)] <- 0
         incProgress(1 / 4)
         Data_major_gene = Data_MAF_target %>%
           dplyr::filter(Hugo_Symbol %in% colnames(mat_oncoprint)) %>%
           dplyr::arrange(Tumor_Sample_Barcode)
         
         Data_major_gene_agg <- Data_major_gene %>%
           group_by(Tumor_Sample_Barcode, Hugo_Symbol) %>%
           summarise(Variant_Classification = paste(unique(Variant_Classification), collapse = ";"), .groups = "drop")
         mat_oncoprint_update <- acast(Data_major_gene_agg, Tumor_Sample_Barcode ~ Hugo_Symbol, value.var = "Variant_Classification", fill = "")
         mat_oncoprint[rownames(mat_oncoprint_update), colnames(mat_oncoprint_update)] <- mat_oncoprint_update
         incProgress(1 / 4)
         rownames(mat_oncoprint) = Data_oncoprint$症例.基本情報.がん種.OncoTree.
         mat_oncoprint = t(mat_oncoprint)
         column_title = "OncoPrint for frequent gene mutations"
         tmp = mat_oncoprint
         tmp[tmp != ""] = 1
         tmp[tmp == ""] = 0
         tmp = matrix(as.numeric(tmp), ncol = ncol(tmp))
         tmp2 = order(apply(tmp, FUN=sum, MARGIN=1))
         if(!is.null(gene_group_1)){
           if(!is.null(gene_group_2)){
             tmp2 = c(tmp2[-which(tmp2 %in%
                                    c(which(oncogenic_genes %in% c(gene_group_1, gene_group_2))))],
                      tmp2[which(tmp2 %in% which(oncogenic_genes %in% gene_group_2))],
                      tmp2[which(tmp2 %in% which(oncogenic_genes %in%  gene_group_1))])
           } else{
             tmp2 = c(tmp2[-which(tmp2 %in% which(oncogenic_genes %in% gene_group_1))],
                      tmp2[which(tmp2 %in% which(oncogenic_genes %in% gene_group_1))])
           }
         }
         tmp = data.frame(tmp)
         colnames(tmp) = 1:ncol(tmp)
         sample_order = 1:ncol(tmp)
         colPal3 <- colorRampPalette(brewer.pal(11, "Spectral"))
         anno_diag = colnames(mat_oncoprint)
         anno_age = Data_oncoprint$YoungOld
         Data_oncoprint = Data_oncoprint %>%
           dplyr::mutate(anno_meta = case_when(
             Lymph_met == "Yes" ~ "(+)",
             Brain_met == "Yes" ~ "(+)", 
             Lung_met == "Yes" ~ "(+)",
             Bone_met == "Yes" ~ "(+)",
             Liver_met == "Yes" ~ "(+)",
             Other_met == "Yes" ~ "(+)",
             TRUE ~ "(-)"))
         anno_meta = Data_oncoprint$anno_meta
         diag_list = names(sort(table(anno_diag),decreasing = TRUE))
         diag_list = factor(diag_list, levels = diag_list)
         diag_col = colPal3(length(diag_list))
         meta_list = unique(anno_meta)
         age_list = unique(anno_age)
         names(diag_col) = diag_list
         diag_list2 = sort(unique(anno_diag))
         diag_col2 = diag_col
         names(diag_col2) = diag_list2
         tmp3 = factor(mat_oncoprint)
         tmp3 = relevel(tmp3, ref = "")
         tmp3 = matrix(as.numeric(as.logical(as.numeric(tmp3) - 1)),
                       ncol = ncol(mat_oncoprint))
         tmp3 = data.frame(tmp3)
         colnames(tmp3) = 1:ncol(tmp3)
         ord <- do.call(order, c(as.data.frame(t(tmp3[rev(tmp2), ])), list(decreasing = TRUE)))
         # 並び替え適用
         tmp3 <- tmp3[, ord]
         # 各列の移動先を取得
         sample_order <- rank(ord, ties.method = "first")
         sample_order_simple = sample_order
         if(oncoprint_option == "Sort by pathology"){
           sample_order_final = NULL
           for(l in diag_list){
             sample_order_final = c(sample_order_final,
                                    sample_order[sample_order %in% (1:length(sample_order))[(anno_diag == l)]])
           }
         } else if(oncoprint_option == "Sort by age"){
           sample_order_final_age = NULL
           for(l in age_list){
             sample_order_final_age = c(sample_order_final_age,
                                        sample_order[sample_order %in% (1:length(sample_order))[(anno_age == l)]])
           }
         } else if(oncoprint_option == "Sort by metastasis status at CGP"){
           sample_order_final_meta = NULL
           for(l in meta_list){
             sample_order_final_meta = c(sample_order_final_meta,
                                         sample_order[sample_order %in% (1:length(sample_order))[(anno_meta == l)]])
           }
         }
         incProgress(1 / 4)
         # 縦に並べる凡例: Mutation Types, Age, Metastasis
         lgd_mut <- Legend(
           at = names(color_mut), 
           labels = names(color_mut),
           title = "Mutation Types",
           legend_gp = gpar(fill = unlist(color_mut)),
           direction = "vertical"
         )
         lgd_age <- Legend(
           at = c("Older", "Younger"),
           labels = c("Older", "Younger"),
           title = paste0("Age <= ", mid_age, ", or older"),
           legend_gp = gpar(fill = c("green", "blue")),
           direction = "vertical"
         )
         lgd_meta <- Legend(
           at = c("(+)", "(-)"),
           labels = c("Positive", "Negative"),
           title = "Metastasis",
           legend_gp = gpar(fill = c("purple1", "orange")),
           direction = "vertical"
         )
         if(length(diag_col2) > 100){
           wrap_every <- max(14, ceil(length(diag_col2)/7))
           lgd_diag_chunks <- split(names(diag_col2), ceiling(seq_along(names(diag_col2)) / wrap_every))
           lgd_diag_rows <- lapply(seq_along(lgd_diag_chunks), function(i) {
             chunk <- lgd_diag_chunks[[i]]
             if (i == 1) {
               # 1つ目にタイトルを追加
               Legend(
                 at = chunk,                          # 凡例項目
                 labels = chunk,                      # ラベル
                 legend_gp = gpar(fill = diag_col2[chunk]),  # 色設定
                 title = "Diagnosis"                  # タイトル
               )
             } else {
               # 2つ目以降はタイトルなし
               Legend(
                 at = chunk,                          # 凡例項目
                 labels = chunk,                      # ラベル
                 legend_gp = gpar(fill = diag_col2[chunk]),   # 色設定
                 title = "\u200B"
               )
             }
           })
           # 横方向で分割された凡例をパッキング
           final_lgd_diag_with_title <- do.call(packLegend, c(lgd_diag_rows, list(direction = "horizontal", gap = unit(3, "mm"))))
           # 縦方向に並べる凡例を統合
           vertical_legends <- packLegend(
             lgd_mut,
             lgd_age,
             lgd_meta,
             direction = "vertical",  # 縦に並べる
             gap = unit(3, "mm")
           )
           final_legends = list(vertical_legends, final_lgd_diag_with_title)
         } else {
           wrap_every <- max(10, ceil(length(diag_col2)/6))
           lgd_diag_chunks <- split(names(diag_col2), ceiling(seq_along(names(diag_col2)) / wrap_every))
           lgd_diag_rows <- lapply(seq_along(lgd_diag_chunks), function(i) {
             chunk <- lgd_diag_chunks[[i]]
             if (i == 1) {
               # 1つ目にタイトルを追加
               Legend(
                 at = chunk,                          # 凡例項目
                 labels = chunk,                      # ラベル
                 legend_gp = gpar(fill = diag_col2[chunk]),  # 色設定
                 title = "Diagnosis"                  # タイトル
               )
             } else {
               # 2つ目以降はタイトルなし
               Legend(
                 at = chunk,                          # 凡例項目
                 labels = chunk,                      # ラベル
                 legend_gp = gpar(fill = diag_col2[chunk]),   # 色設定
                 title = "\u200B"
               )
             }
           })
           # 横方向で分割された凡例をパッキング
           final_lgd_diag_with_title <- do.call(packLegend, c(lgd_diag_rows, list(direction = "horizontal", gap = unit(3, "mm"))))
           # 縦方向に並べる凡例を統合
           vertical_legends <- packLegend(
             lgd_age,
             lgd_meta,
             direction = "vertical",  # 縦に並べる
             gap = unit(3, "mm")
           )
           final_legends = list(lgd_mut, vertical_legends, final_lgd_diag_with_title)
         }
         ha = HeatmapAnnotation(
           Diagnosis = anno_diag,
           Age = anno_age,
           Metastasis = anno_meta,
           col = list(Diagnosis = diag_col2,
                      Age = c("Older" = "green", "Younger" = "blue"),
                      Metastasis = c("(+)" = "purple1", "(-)" = "orange")),
           annotation_height = unit(15, "mm"),
           show_legend = FALSE
         )
         hb = HeatmapAnnotation(TMB = anno_points(Data_TMB, size = unit(0.6, "mm"),
                                                  ylim = c(0, 30),
                                                  axis_param = list(
                                                    side = "right",
                                                    at = c(0, 10, 20, 30), 
                                                    labels = c("0", "10", "20", ">30")
                                                  )),
                                annotation_height = unit(15, "mm"),
                                show_legend = FALSE)
         if(oncoprint_option == "Sort by mutation frequency"){
           ht = oncoPrint(mat_oncoprint, get_type = function(x) strsplit(x, ";")[[1]],
                          column_order = sample_order_simple, row_order = rev(tmp2),
                          alter_fun = alter_fun, col = color_mut,  pct_gp = gpar(fontsize = 10),
                          column_title = column_title,
                          remove_empty_columns = FALSE, remove_empty_rows = FALSE,
                          bottom_annotation = ha,
                          top_annotation = hb,
                          use_raster = FALSE,
                          #use_raster = TRUE,
                          #height = unit(200, "mm"),
                          alter_fun_is_vectorized = TRUE)
         } else if(oncoprint_option == "Sort by age"){
           ht = oncoPrint(mat_oncoprint, get_type = function(x) strsplit(x, ";")[[1]],
                          column_order = sample_order_final_age, row_order = rev(tmp2),
                          alter_fun = alter_fun, col = col,  pct_gp = gpar(fontsize = 10),
                          column_title = column_title,
                          remove_empty_columns = FALSE, remove_empty_rows = FALSE,
                          bottom_annotation = ha,
                          top_annotation = hb,
                          use_raster = FALSE,
                          #height = unit(200, "mm"),
                          alter_fun_is_vectorized = TRUE)
         } else if(oncoprint_option == "Sort by metastasis status at CGP"){
           ht = oncoPrint(mat_oncoprint, get_type = function(x) strsplit(x, ";")[[1]],
                          column_order = sample_order_final_meta, row_order = rev(tmp2),
                          alter_fun = alter_fun, col = col,  pct_gp = gpar(fontsize = 10),
                          column_title = column_title,
                          remove_empty_columns = FALSE, remove_empty_rows = FALSE,
                          bottom_annotation = ha,
                          top_annotation = hb,
                          use_raster = FALSE,
                          #height = unit(200, "mm"),
                          alter_fun_is_vectorized = TRUE)
         } else{
           ht = oncoPrint(mat_oncoprint, get_type = function(x) strsplit(x, ";")[[1]],
                          column_order = sample_order_final, row_order = rev(tmp2),
                          alter_fun = alter_fun, col = col,  pct_gp = gpar(fontsize = 10),
                          column_title = column_title,
                          remove_empty_columns = FALSE, remove_empty_rows = FALSE,
                          bottom_annotation = ha,
                          top_annotation = hb,
                          use_raster = FALSE,
                          #height = unit(200, "mm"),
                          alter_fun_is_vectorized = TRUE)
         }
         draw(
           ht,
           show_heatmap_legend = FALSE,
           annotation_legend_list = final_legends,  # 凡例を縦・横に配置
           annotation_legend_side = "bottom",  # 凡例を下部に配置
           heatmap_legend_side = "bottom"      # ヒートマップの凡例も下部に配置
         )
         incProgress(1 / 4)
       })
     })
     incProgress(1 / 8)
     output$figure_lolliplot2 = renderPlot({
       withProgress(message = "Internet access is necessary", {
         lolliplot_gene = input$lolliplot_gene
         if(is.null(lolliplot_gene)){
           lolliplot_gene = "TP53"
         } else if(lolliplot_gene == ""){
           lolliplot_gene = "TP53"
         }
         data_lolliplot = Data_MAF_target %>%
           dplyr::select(Chromosome,
                         Start_Position,
                         End_Position,
                         Reference_Allele,
                         Tumor_Seq_Allele2,
                         Tumor_Sample_Barcode,
                         Hugo_Symbol,
                         Variant_Classification,
                         amino.acid.change) %>%
           dplyr::filter(!Variant_Classification %in% c("rearrangement",
                                             "truncation",
                                             "amplification",
                                             "fusion",
                                             "TMB_MSI_high",
                                             "duplication",
                                             "exon_skipping",
                                             "deletion",
                                             "inversion",
                                             "loss")) %>%
           dplyr::filter(!str_detect(amino.acid.change, "c.")) %>%
           dplyr::filter(!str_detect(amino.acid.change, "_")) %>%
           dplyr::filter(Hugo_Symbol == lolliplot_gene) 
         if(length(data_lolliplot$Chromosome) == 0){
           gg_empty()
         } else {
           var<-unique(data_lolliplot[!duplicated(data_lolliplot),])#remove duplicates
           var[order(var$Hugo_Symbol,gsub("([A-Z]+)([0-9]+)","\\2",var$amino.acid.change)),]#order
           var.freq<-plyr::count(var,c("Hugo_Symbol","amino.acid.change","Variant_Classification"))
           var.aanum<-unlist(lapply(regmatches(var.freq$amino.acid.change,gregexpr(pattern="*(\\d+)",var.freq$amino.acid.change)),function(x) x[[1]]))
           var.plot.data<-cbind(var.freq,as.numeric(as.character(var.aanum)))#hard way to remove factor
           colnames(var.plot.data)[5]<-"var.aanum"
           var.plot.data = var.plot.data[var.plot.data$var.aanum >=1 & is.finite(var.plot.data$var.aanum) & !is.na(var.plot.data$var.aanum),]
           
           var.plot<-isolate(var.plot.data)
           var.plot<-var.plot[var.plot$Hugo_Symbol==lolliplot_gene,]
           var.plot<-var.plot[order(var.plot$var.aanum),]
           var.plot$amino.acid.change<-as.character(var.plot$amino.acid.change)
           
           incProgress(1 / 4)
  
           p<-ggplot(var.plot,aes(var.aanum,freq,color=Variant_Classification,label=amino.acid.change)) +
             geom_segment(aes(x=var.aanum,y=0,xend=var.aanum,yend=freq),
                          color=ifelse(var.plot$freq>=input$lolliplot_no,"orange","grey50"),
                          linewidth=ifelse(var.plot$freq>=input$lolliplot_no,1.3,0.7)) +
             geom_point(size=5) +
             geom_text_repel(nudge_y=0.2,
                             color=ifelse(var.plot$freq>=input$lolliplot_no,"orange","NA"),
                             size=ifelse(var.plot$freq>=input$lolliplot_no,5,5),
                             fontface="bold", max.overlaps=Inf) +
             theme_classic()
           proteins_acc<-read.table(file("source/UniProt.txt",
                                         encoding='UTF-8-BOM'),header=T)
          if(!lolliplot_gene %in% c("NKX2_1", "NKX3_1","H1_2", "H3_4","H3_5","H3_3A","H3_3B")){
            lolliplot_gene_raw = str_split(lolliplot_gene, "_",simplify = T)[[1]]
          }  
           if(length(proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene_raw,2]) == 1){
             proteins_acc<-proteins_acc[proteins_acc$Hugo_Symbol==lolliplot_gene_raw,2]
             proteins_acc_url<-gsub(" ","%2C",proteins_acc)
             baseurl<-"https://www.ebi.ac.uk/proteins/api/features?offset=0&size=100&accession="
             url<-paste0(baseurl,proteins_acc_url)
             incProgress(1 / 6)
             flag_lolliplot = 0
             if(file.exists(paste0("source/lolliplot/", lolliplot_gene_raw, ".rda"))){
               load(paste0("source/lolliplot/", lolliplot_gene_raw, ".rda"))
               flag_lolliplot = 1
             } else {
               prots_feat<-try(GET(url,accept_json()), silent = FALSE)
               if (class(prots_feat) != "try-error"){
                 flag_lolliplot = 1
                 save(prots_feat, file=paste0("source/lolliplot/", lolliplot_gene_raw, ".rda"))
               }
             }
             if (flag_lolliplot == 1){
               prots_feat_red<-(httr::content(prots_feat))
               incProgress(1 / 6)
               features_total_plot<-NULL
               for(i in 1:length(prots_feat_red)){ 
                 features_temp<-drawProteins::extract_feat_acc(prots_feat_red[[i]])#the extract_feat_acc() function takes features into a data.frame
                 features_temp$order<-i # this order is needed for plotting later
                 features_total_plot<-rbind(features_total_plot,features_temp)
               }
               plot_start<-0#starts 0
               plot_end<-max(features_total_plot$end)
               if("CHAIN"%in%(unique(features_total_plot$type))){
                 p <- p +
                   geom_rect(data=features_total_plot[features_total_plot$type=="CHAIN",],
                             mapping=aes(xmin=begin,xmax=end,ymin=-0.1*max(var.plot$freq),ymax=-0.04*max(var.plot$freq)),
                             colour="grey",
                             fill="grey",
                             inherit.aes=F)
               }
               incProgress(1 / 6)
               if("DNA_BIND"%in%(unique(features_total_plot$type))) {
                 features_total_plot[features_total_plot$type=="DNA_BIND","description"] <- features_total_plot[features_total_plot$type=="DNA_BIND","type"]
                 p <- p +
                   geom_rect(data=features_total_plot[features_total_plot$type=="DNA_BIND",],
                             mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                             inherit.aes=F)
               }
               if("DOMAIN"%in%(unique(features_total_plot$type))){
                 p <- p +
                   geom_rect(data=features_total_plot[features_total_plot$type=="DOMAIN",],
                             mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                             inherit.aes=F)
               }
               if("MOTIF"%in%(unique(features_total_plot$type))){
                 p <- p +
                   geom_rect(data=features_total_plot[features_total_plot$type=="MOTIF",],
                             mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                             inherit.aes=F)
               }
               if("REPEAT"%in%(unique(features_total_plot$type))){
                 p <- p +
                   geom_rect(data=features_total_plot[features_total_plot$type=="REPEAT",],
                             mapping=aes(xmin=begin,xmax=end,ymin=-0.12*max(var.plot$freq),ymax=-0.02*max(var.plot$freq),fill=description),
                             inherit.aes=F)
               }
               p <- p +
                 scale_x_continuous(breaks=round(seq(plot_start,plot_end,by=50)),name="Amino acid number")
             }
           }
           p <- p +
             theme(plot.title=element_text(face="bold",size=(15),hjust=0),axis.text.x=element_text(angle=90,hjust=1)) +
             labs(y="Mutation frequency in our dataset",
                  title=paste("Mutation frequency at ",lolliplot_gene," in the dataset",sep="")) +
             scale_y_continuous(breaks=seq(0,max(var.plot$freq),max(floor(max(var.plot$freq)/100)*10,5)),name="Frequency")
           print(p)
           incProgress(1 / 6)
         }
       })
     })
     incProgress(1 / 8)

     Data_case_target = Data_case_target %>%
       dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE) %>%
       dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                     症例.基本情報.年齢,
                     YoungOld,
                     Lymph_met,
                     Brain_met,
                     Lung_met,
                     Bone_met,
                     Liver_met,
                     Other_met,
                     EP_option,
                     EP_treat,
                     症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                     症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                     症例.基本情報.性別.名称.,
                     症例.背景情報.病理診断名,
                     症例.背景情報.臨床診断名,
                     症例.検体情報.病理診断名,
                     症例.基本情報.がん種.OncoTree.,
                     症例.基本情報.がん種.OncoTree..名称.,
                     症例.基本情報.がん種.OncoTree.LEVEL1.,
                     症例.検体情報.パネル.名称.,
                     症例.背景情報.ECOG.PS.名称.,
                     症例.背景情報.喫煙歴有無.名称.,
                     症例.背景情報.アルコール多飲有無.名称.,
                     症例.背景情報.重複がん有無.異なる臓器..名称.,
                     症例.背景情報.多発がん有無.同一臓器..名称.,
                     症例.がん種情報.登録時転移部位.名称.,
                     症例.検体情報.腫瘍細胞含有割合,
                     症例.検体情報.検体採取部位.名称.,
                     症例.転帰情報.転帰.名称.,
                     症例.背景情報.診断日,
                     症例.管理情報.登録日,
                     症例.転帰情報.最終生存確認日,
                     症例.転帰情報.死亡日
                     ) %>%
       dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE) %>%
       dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
     incProgress(1 / 8)
     Data_MAF_target_tmp =  Data_MAF_target_tmp %>%
       dplyr::mutate(
         Evidence = case_when(
           Evidence_level == "F" ~ "Oncogenic",
           TRUE ~ ""
         ))
     gene_list = unique(c(input$gene,
                   names(sort(table(Data_MAF_target$Hugo_Symbol),
                            decreasing = T))))
     gene_list = gene_list[gene_list %in% unique(Data_MAF_target$Hugo_Symbol)]
     gene_list = gene_list[1:min(input$gene_no, length(gene_list))]
     # ② Data_MAF_target_tmp から gene_list に含まれる遺伝子について、
     #    サンプルごとに amino.acid.change をカンマ区切りでまとめる
     gene_data <- Data_MAF_target_tmp %>%
       filter(Hugo_Symbol %in% gene_list) %>%
       group_by(Tumor_Sample_Barcode, Hugo_Symbol) %>%
       summarise(Variant_mut = paste(amino.acid.change, collapse = ","), .groups = "drop") %>%
       pivot_wider(names_from = Hugo_Symbol,
                   values_from = Variant_mut,
                   values_fill = list(Variant_mut = ""))
     
     # ③ gene_table の初期化：各サンプル（Data_case_target のハッシュID）を行とし、
     #     gene_list の各遺伝子を列とする。gene_table$target_gene_sequence は別途追加。
     gene_table <- Data_case_target %>%
       select(Tumor_Sample_Barcode = `C.CAT調査結果.基本項目.ハッシュID`)
     
     # gene_data のサンプルIDと gene_table の ID をキーに left_join
     gene_table <- gene_table %>%
       left_join(gene_data, by = c("Tumor_Sample_Barcode" = "Tumor_Sample_Barcode"))
     
     # gene_table に不足している列（gene_list に存在するが、gene_data にないサンプル）を空文字に補完
     gene_table <- gene_table %>%
       mutate(across(all_of(gene_list), ~ replace_na(.x, ""))) %>%
       mutate(target_gene_sequence = "")
     
     # gene_table = data.frame(matrix(rep(rep("", length(gene_list)),
     #                                    length(Data_case_target$C.CAT調査結果.基本項目.ハッシュID)),
     #                                nrow=length(Data_case_target$C.CAT調査結果.基本項目.ハッシュID)))
     # rownames(gene_table) = Data_case_target$C.CAT調査結果.基本項目.ハッシュID
     # colnames(gene_table) = gene_list
     # gene_table$target_gene_sequence = ""
     # incProgress(1 / 8)
     # for(i in 1:length(gene_list)){
     #   tmp_mut = Data_MAF_target_tmp %>%
     #     dplyr::filter(Hugo_Symbol == gene_list[i])
     #   ID_mut = tmp_mut$Tumor_Sample_Barcode
     #   Variant_mut = tmp_mut$amino.acid.change
     #   for(j in 1:length(ID_mut)){
     #     if(gene_table[ID_mut[j],i] == ""){
     #       gene_table[ID_mut[j],i] = Variant_mut[j]
     #     } else{
     #       gene_table[ID_mut[j],i] = paste(gene_table[ID_mut[j],i], Variant_mut[j], sep=",")
     #     }
     #   }
     # }
     
     # if(!is.null(input$gene_group_1)){
     #   tmp_mut = Data_MAF_target_tmp %>%
     #   dplyr::filter(Hugo_Symbol == input$gene_group_1[[1]])
     #   gene_name = input$gene_group_1[[1]]
     # } else {
     #   tmp_mut = Data_MAF_target_tmp %>%
     #     dplyr::filter(Hugo_Symbol == names(sort(table(Data_MAF_target_tmp$Hugo_Symbol),decreasing = T))[[1]])
     #   gene_name = names(sort(table(Data_MAF_target_tmp$Hugo_Symbol),decreasing = T))[[1]]
     # }
     # ID_mut = tmp_mut$Tumor_Sample_Barcode
     # Variant_mut = paste0("chr",
     #   tmp_mut$Chromosome,":",
     #   tmp_mut$Start_Position,"_",
     #   tmp_mut$Reference_Allele,">",
     #   tmp_mut$Tumor_Seq_Allele2,"_",
     #   tmp_mut$Evidence)
     # for(j in 1:length(ID_mut)){
     #   if(gene_table[ID_mut[j],"target_gene_sequence"] == ""){
     #     gene_table[ID_mut[j],"target_gene_sequence"] = Variant_mut[j]
     #   } else{
     #     gene_table[ID_mut[j],"target_gene_sequence"] = paste(gene_table[ID_mut[j],"target_gene_sequence"], Variant_mut[j], sep=",")
     #   }
     # }
     # 1. 最頻出の Hugo_Symbol を取得
     if(!is.null(input$gene_group_1)){
       most_common_gene <- input$gene_group_1[[1]]
       gene_name = input$gene_group_1[[1]]
     } else {
       most_common_gene <- names(sort(table(Data_MAF_target_tmp$Hugo_Symbol), decreasing = TRUE))[1]
       gene_name = names(sort(table(Data_MAF_target_tmp$Hugo_Symbol),decreasing = T))[[1]]
     }
     # 2. 該当遺伝子のデータを抽出し Variant_mut を作成
     tmp_mut <- Data_MAF_target_tmp %>%
       filter(Hugo_Symbol == most_common_gene) %>%
       mutate(Variant_mut = paste0("chr", Chromosome, ":", Start_Position, "_",
                                   Reference_Allele, ">", Tumor_Seq_Allele2, "_", Evidence))
     # 3. サンプルごとに Variant_mut を結合（カンマ区切り）
     mutations_summary <- tmp_mut %>%
       group_by(Tumor_Sample_Barcode) %>%
       summarise(variants = paste(Variant_mut, collapse = ","), .groups = "drop")
     # 4. gene_table に、すでに target_gene_sequence が空なら variants を、そうでなければ既存と結合した文字列に更新
     gene_table <- gene_table %>%
       left_join(mutations_summary, by = "Tumor_Sample_Barcode") %>%
       mutate(target_gene_sequence = if_else(
         target_gene_sequence == "" | is.na(target_gene_sequence),
         variants,
         paste(target_gene_sequence, variants, sep = ",")
       )) %>%
       select(-variants)  # 不要な列は削除
     incProgress(1 / 8)
     
     Data_case_target = left_join(Data_case_target,
                                  gene_table,
                                  by = c("C.CAT調査結果.基本項目.ハッシュID" = "Tumor_Sample_Barcode"))
     Data_case_target_table = Data_case_target %>% dplyr::select(-症例.基本情報.がん種.OncoTree.)
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                          "症例\\.検体情報\\.病理診断名",
     #                                          "症例\\.検体情報\\.提出検体の病理診断名")
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                          "C.CAT調査結果\\.基本項目\\.",
     #                                          "")
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                          "症例\\.EP後レジメン情報\\.",
     #                                          "")
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                          "症例\\.基本情報\\.",
     #                                          "")
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                          "症例\\.検体情報\\.",
     #                                          "")
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                          "症例\\.背景情報\\.",
     #                                          "")
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                          "症例\\.がん種情報\\.",
     #                                          "")
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                          "症例\\.転帰情報\\.",
     #                                          "")
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                          "症例\\.管理情報\\.",
     #                                          "")
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                                "\\.名称\\.",
     #                                                "")
     # colnames(Data_case_target_table) = str_replace(colnames(Data_case_target_table),
     #                                                "target_gene_sequence",
     #                                                paste0(gene_name, "_sequence"))
     replacements <- c(
       "症例\\.検体情報\\.病理診断名" = "症例.検体情報.提出検体の病理診断名",
       "C.CAT調査結果\\.基本項目\\." = "",
       "症例\\.EP後レジメン情報\\." = "",
       "症例\\.基本情報\\." = "",
       "症例\\.検体情報\\." = "",
       "症例\\.背景情報\\." = "",
       "症例\\.がん種情報\\." = "",
       "症例\\.転帰情報\\." = "",
       "症例\\.管理情報\\." = "",
       "\\.名称\\." = "",
       "target_gene_sequence" = paste0(gene_name, "_sequence")
     )
     
     # 一括置換
     new_names <- str_replace_all(colnames(Data_case_target_table), replacements)
     colnames(Data_case_target_table) <- new_names
     output$table_patient = DT::renderDataTable(Data_case_target_table,
                                                server = FALSE,
                                                filter = 'top', 
                                                extensions = c('Buttons'), 
                                                options = list(pageLength = 100, 
                                                               scrollX = TRUE,
                                                               scrollY = "1000px",
                                                               scrollCollapse = TRUE,
                                                               dom="Blfrtip",
                                                               buttons = c('csv', 'excel', 'copy')))
     incProgress(1 / 1)
    })
  })
  
  # Mutually exclusive or co-occurrence
  observeEvent(input$mutually_exclusive, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      Data_case_target = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.基本情報.年齢,
                      症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.基本情報.性別.名称.,
                      症例.基本情報.がん種.OncoTree.,
                      症例.基本情報.がん種.OncoTree..名称.,
                      症例.基本情報.がん種.OncoTree.LEVEL1.,
                      症例.検体情報.パネル.名称.,
                      症例.背景情報.ECOG.PS.名称.,
                      症例.背景情報.喫煙歴有無.名称.,
                      症例.背景情報.アルコール多飲有無.名称.,
                      症例.背景情報.重複がん有無.異なる臓器..名称.,
                      症例.背景情報.多発がん有無.同一臓器..名称.,
                      症例.がん種情報.登録時転移部位.名称.,
                      症例.検体情報.腫瘍細胞含有割合,
                      症例.検体情報.検体採取部位.名称.,
                      症例.転帰情報.転帰.名称.,
                      症例.背景情報.診断日,
                      症例.管理情報.登録日,
                      症例.転帰情報.最終生存確認日,
                      症例.転帰情報.死亡日
        )
      Data_MAF = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","A","B","C","D","E","F") &
            Variant_Classification != "expression"
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
      if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
        Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
      }

      
      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      if(input$patho == "Only pathogenic muts"){
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(Evidence_level == "F")
      } else {
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(!Hugo_Symbol %in% c("TMB", "MSI") | Evidence_level == "F")
        
      }
      Data_MAF_target = Data_MAF_target  %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        .keep_all = TRUE)
      Gene_names = unique(c(input$gene,
                            names(sort(table(Data_MAF_target$Hugo_Symbol),
                                       decreasing = T))))
      Gene_names = Gene_names[Gene_names %in% unique(Data_MAF_target$Hugo_Symbol)]
      Gene_names = Gene_names[1:min(length(unique(Data_MAF_target$Hugo_Symbol)),input$gene_no)]
      Patient_names = unique(Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      Mutation_matrix = matrix(rep(0, length(Gene_names) * length(Patient_names)),
                               nrow = length(Gene_names),
                               ncol = length(Patient_names))
      rownames(Mutation_matrix) = Gene_names
      colnames(Mutation_matrix) = Patient_names
      
      Mutation_list_for_matrix = Data_MAF_target %>%
        dplyr::filter(Hugo_Symbol %in% Gene_names)
      for(i in 1:length(Mutation_list_for_matrix$Hugo_Symbol)){
        Mutation_matrix[Mutation_list_for_matrix$Hugo_Symbol[i],
                        Mutation_list_for_matrix$Tumor_Sample_Barcode[i]] = 1
      }
      PMA <- getPM(Mutation_matrix)
      mutually_exclusive <- getMutex(A=Mutation_matrix,
                                     method = "ShiftedBinomial",
                                     #method = "Exact",
                                     #method = "Binomial",
                                     #method = "RefinedNormal",
                                     PM=PMA,
                                     lower.tail=TRUE
      )
      colnames(mutually_exclusive) = Gene_names
      rownames(mutually_exclusive) = Gene_names
      mutually_exclusive[lower.tri(mutually_exclusive,diag=TRUE)] <- NA
      Data_mut_ex = expand.grid(Gene_names, Gene_names)
      colnames(Data_mut_ex) = c("Gene_1", "Gene_2")
      Data_mut_ex$P_value = mutually_exclusive@x
      Data_mut_ex = Data_mut_ex %>%
        dplyr::mutate(P_value = case_when(
          P_value > 1-10^-10 ~ 1-10^-10,
          P_value < 10^-10 ~ 10^-10,
          TRUE ~ P_value
        ))
      Data_mut_ex = Data_mut_ex %>%
        dplyr::mutate(P_co_or_ex = case_when(
          P_value > 0.5 ~ log(1-P_value,base=10),
          TRUE ~ -log(P_value,base=10)
        ))
      Data_mut_ex = Data_mut_ex %>%
        dplyr::mutate(P_figure = case_when(
          P_co_or_ex > 3 ~ 3,
          P_co_or_ex < -3 ~ -3,
          TRUE ~ P_co_or_ex
        ))
      Data_mut_ex = Data_mut_ex %>%
        dplyr::mutate(P_signif = case_when(
          P_value <= 0.001 ~ "*",
          P_value >= 0.999 ~ "#",
          TRUE ~ ""
        ))
      Data_mut_ex$Gene_1 <- factor(Data_mut_ex$Gene_1, levels = Gene_names)
      Data_mut_ex$Gene_2 <- factor(Data_mut_ex$Gene_2, levels = Gene_names)
      plots = list()
      plots[[1]] = Data_mut_ex %>% 
        ggplot(aes(x = Gene_1, y = Gene_2, fill = P_figure)) + 
        geom_tile(color = "white") + 
        scale_fill_gradient2(low = "red", high = "blue", mid = "white", 
                             midpoint = 0, limit = c(-3, 3),  breaks = -3:3,
                             labels = c("<-3", "-2", "-1", "0", "1", "2", ">3"),
                             name = "-log10(P-value)", space = "Lab",
                             na.value = "white") +
        geom_text(aes(label = P_signif),
                  color = "black", size = 4)+
        scale_y_discrete(limits = rev(levels(Data_mut_ex$Gene_2))) +
        theme_classic() +
        ggtitle("P value: Mutually exclusive (blue) or co-occurrence (red)\n*: p<0.001, estimated with Rediscover package") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
      Data_mut_ex_table = Data_mut_ex %>%
        dplyr::select(Gene_1, Gene_2, P_co_or_ex)
      Data_mut_ex_table = Data_mut_ex_table %>%
        dplyr::mutate(P_value = 10^-abs(P_co_or_ex))
      Data_mut_ex_table$G1mutG2mut = 0
      Data_mut_ex_table$G1mutG2wt = 0
      Data_mut_ex_table$G1wtG2mut = 0
      Data_mut_ex_table$G1wtG2wt = 0
      Data_mut_ex_table$OddsRatio = 0
      Data_mut_ex_table$OR_999CI_LL = 0
      Data_mut_ex_table$OR_999CI_UL = 0
      Gene_num = length(Gene_names)
      Single_mut = rowSums(Mutation_matrix)
      Total_patients = length(Mutation_matrix[1,])
      for(i in 1:Gene_num){
        for(j in 1:Gene_num){
          tmp = sum(Mutation_matrix[i,] & Mutation_matrix[j,])
          Data_mut_ex_table$G1mutG2mut[(i-1)*Gene_num+j] = tmp
          Data_mut_ex_table$G1mutG2wt[(i-1)*Gene_num+j] = Single_mut[j] - tmp
          Data_mut_ex_table$G1wtG2mut[(i-1)*Gene_num+j] = Single_mut[i] - tmp
          Data_mut_ex_table$G1wtG2wt[(i-1)*Gene_num+j] = Total_patients - Single_mut[i] - Single_mut[j] + tmp
          tmp = fisher.test(matrix(c(tmp, Single_mut[i] - tmp, Single_mut[j] - tmp, Total_patients - Single_mut[i] - Single_mut[j] + tmp), nrow = 2), conf.level = 0.999)
          Data_mut_ex_table$OddsRatio[(i-1)*Gene_num+j] = tmp$estimate[[1]]
          Data_mut_ex_table$OR_999CI_LL[(i-1)*Gene_num+j] = tmp$conf.int[[1]]
          Data_mut_ex_table$OR_999CI_UL[(i-1)*Gene_num+j] = tmp$conf.int[[2]]
        }
      }
      Data_mut_ex_table = Data_mut_ex_table %>%
        dplyr::filter(!is.na(P_value))
      Data_mut_ex_table = Data_mut_ex_table %>%
        dplyr::select(-P_co_or_ex)
      output$table_mutually_exclusive = DT::renderDataTable(Data_mut_ex_table,
                                                    server = FALSE,
                                                    filter = 'top', 
                                                    extensions = c('Buttons'), 
                                                    options = list(pageLength = 100, 
                                                                   scrollX = TRUE,
                                                                   scrollY = "1000px",
                                                                   scrollCollapse = TRUE,
                                                                   dom="Blfrtip",
                                                                   buttons = c('csv', 'copy')))
      Data_mut_ex$OddsRatio = NA
      no_tmp = ((length(Data_mut_ex[,1]))^0.5)
      for(i in 1:length(Data_mut_ex[,1])){
        if(i %% no_tmp != 0 & i %% no_tmp <= floor(i/no_tmp)){
          Data_mut_ex[i,"OddsRatio"] = (Data_mut_ex_table %>% dplyr::filter(Gene_1 == Data_mut_ex[i,"Gene_1"] & Gene_2 == Data_mut_ex[i,"Gene_2"]))$OddsRatio
        }
      }
      Data_mut_ex$OddsRatio = -log10(Data_mut_ex$OddsRatio)
      Data_mut_ex = Data_mut_ex %>%
        dplyr::mutate(OddsRatio_figure = case_when(
          OddsRatio > 3 ~ 3,
          OddsRatio < -3 ~ -3,
          TRUE ~ OddsRatio
        ))
      plots[[2]] = Data_mut_ex %>% 
        ggplot(aes(x = Gene_1, y = Gene_2, fill = OddsRatio_figure)) + 
        geom_tile(color = "white") + 
        scale_fill_gradient2(low = "red", high = "blue", mid = "white", 
                             midpoint = 0, limit = c(-3, 3), breaks = -3:3,
                             labels = c("<-3", "-2", "-1", "0", "1", "2", ">3"),
                             name = "-log10(OddsRatio)", space = "Lab",
                             na.value = "white")+
        geom_text(aes(label = P_signif),
                  color = "black", size = 4)+
        scale_y_discrete(limits = rev(levels(Data_mut_ex$Gene_2)))+
        theme_classic() +
        ggtitle("Odds ratio: Mutually exclusive (blue) or co-occurrence (red)\n*: p<0.001 for mutually exclusive, #: p<0.001 for co-occurrence") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
      output$figure_mutually_exclusive = renderPlot({
        grid.arrange(plots[[1]], plots[[2]], ncol = 1)
      })
      incProgress(1 / 1)
    })
  })
  
  observeEvent(input$mut_subtype, {
    withProgress(message = sample(nietzsche)[1], {
      output$figure_mut_subtype = renderPlot({
        Data_case_target = Data_case()
        if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
          Data_case_target = Data_case_target %>%
            dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                            Data_report()$Tumor_Sample_Barcode)
        }
        Data_case_target = Data_case_target %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                        症例.基本情報.年齢,
                        Lymph_met,
                        Brain_met,
                        Lung_met,
                        Bone_met,
                        Liver_met,
                        Other_met,
                        EP_option,
                        EP_treat,
                        症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                        症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                        症例.基本情報.性別.名称.,
                        症例.基本情報.がん種.OncoTree.,
                        症例.基本情報.がん種.OncoTree..名称.,
                        症例.基本情報.がん種.OncoTree.LEVEL1.,
                        症例.検体情報.パネル.名称.,
                        症例.背景情報.ECOG.PS.名称.,
                        症例.背景情報.喫煙歴有無.名称.,
                        症例.背景情報.アルコール多飲有無.名称.,
                        症例.背景情報.重複がん有無.異なる臓器..名称.,
                        症例.背景情報.多発がん有無.同一臓器..名称.,
                        症例.がん種情報.登録時転移部位.名称.,
                        症例.検体情報.腫瘍細胞含有割合,
                        症例.検体情報.検体採取部位.名称.,
                        症例.転帰情報.転帰.名称.,
                        症例.背景情報.診断日,
                        症例.管理情報.登録日,
                        症例.転帰情報.最終生存確認日,
                        症例.転帰情報.死亡日
          )
        Data_MAF = Data_report() %>%
          dplyr::filter(
            !str_detect(Hugo_Symbol, ",") &
              Hugo_Symbol != "" &
              Evidence_level %in% c("","A","B","C","D","E","F") &
              Variant_Classification != "expression"
          ) %>% 
          dplyr::arrange(desc(Evidence_level)) %>%
          dplyr::distinct(Tumor_Sample_Barcode,
                          Hugo_Symbol,
                          Start_Position,
                          .keep_all = TRUE)
         if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
          Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
        }

        Data_MAF_target = Data_MAF %>%
          dplyr::filter(Tumor_Sample_Barcode %in%
                          Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
        if(input$patho == "Only pathogenic muts"){
          Data_MAF_target = Data_MAF_target %>%
            dplyr::filter(Evidence_level == "F")
        } else {
          Data_MAF_target = Data_MAF_target %>%
            dplyr::filter(!Hugo_Symbol %in% c("TMB", "MSI") | Evidence_level == "F")
          
        }
        Data_MAF_target = Data_MAF_target  %>%
          dplyr::distinct(Tumor_Sample_Barcode,
                          Hugo_Symbol,
                          .keep_all = TRUE)
        gene_to_analyze = unique(c(input$gene,
                              names(sort(table(Data_MAF_target$Hugo_Symbol),
                                         decreasing = T))))
        gene_to_analyze = gene_to_analyze[gene_to_analyze %in% unique(Data_MAF_target$Hugo_Symbol)]
        gene_to_analyze = gene_to_analyze[1:min(length(unique(Data_MAF_target$Hugo_Symbol)),input$gene_no)]
        Diseases = sort(unique(Data_case_target$症例.基本情報.がん種.OncoTree.))
        Summary_Gene_alteration = data.frame(matrix(0,
                                                    nrow=length(Diseases),
                                                    ncol=length(gene_to_analyze)))
        colnames(Summary_Gene_alteration) = gene_to_analyze
        rownames(Summary_Gene_alteration) = Diseases
        
        for(i in Diseases){
          Data_tmp = Data_case_target %>%
            dplyr::filter(症例.基本情報.がん種.OncoTree. == i)
          Data_tmp = unique(Data_tmp$C.CAT調査結果.基本項目.ハッシュID)
          patient_no = length(Data_tmp)
          for(j in gene_to_analyze){
            Summary_Gene_alteration[i,j] =
              length(unique((Data_MAF_target %>% dplyr::filter(
                Hugo_Symbol == j &
                  Tumor_Sample_Barcode %in% Data_tmp))$Tumor_Sample_Barcode)) / patient_no * 100
          }
        }
        
        Data_tmp = Summary_Gene_alteration
        Data_tmp$Disease = rownames(Data_tmp)
        Data_tmp = Data_tmp %>% gather(value = "freq", key = Gene, -Disease)
        Data_tmp <- transform(Data_tmp, Disease= factor(Disease, levels = sort(unique(Disease), decreasing = TRUE)))
        ggplot(Data_tmp, aes(as.factor(Gene), as.factor(Disease))) +
          geom_tile(aes(fill = freq), color = "black",
                    #            lwd = 1,
                    linetype = 1) + 
          geom_text(aes(label = round(freq, 0))) +
          scale_fill_gradient(low = "white", high = "red", limit=c(0,100), name="Frequency (%)") +
          theme_classic() +
          labs(title = "Frequently mutated genes and diagnoses") +
          theme(axis.text.x = element_text(angle = 45, hjust = 1),
                axis.title.x = element_blank(),
                axis.title.y = element_blank(),
                axis.ticks.x = element_blank(),
                axis.ticks.y = element_blank(),
                axis.line.x = element_blank(),
                axis.line.y = element_blank())
      })
    })
  })

  
  observeEvent(input$custering_analysis, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      Data_case_target = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.基本情報.年齢,
                      Lymph_met,
                      Brain_met,
                      Lung_met,
                      Bone_met,
                      Liver_met,
                      Other_met,
                      EP_option,
                      EP_treat,
                      YoungOld,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.基本情報.性別.名称.,
                      症例.基本情報.がん種.OncoTree.,
                      症例.基本情報.がん種.OncoTree..名称.,
                      症例.基本情報.がん種.OncoTree.LEVEL1.,
                      症例.検体情報.パネル.名称.,
                      症例.背景情報.ECOG.PS.名称.,
                      症例.背景情報.喫煙歴有無.名称.,
                      症例.背景情報.アルコール多飲有無.名称.,
                      症例.背景情報.重複がん有無.異なる臓器..名称.,
                      症例.背景情報.多発がん有無.同一臓器..名称.,
                      症例.がん種情報.登録時転移部位.名称.,
                      症例.検体情報.腫瘍細胞含有割合,
                      症例.検体情報.検体採取部位.名称.,
                      症例.転帰情報.転帰.名称.,
                      症例.背景情報.診断日,
                      症例.管理情報.登録日,
                      症例.転帰情報.最終生存確認日,
                      症例.転帰情報.死亡日
        )
      incProgress(1 / 13)      
      Data_case_target$Cancers = Data_case_target$症例.基本情報.がん種.OncoTree.
      Data_survival = Data_case_target %>%
        dplyr::mutate(final_observe = case_when(
          症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
          症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
          症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
          TRUE ~ 症例.管理情報.登録日))
      Data_survival = Data_survival %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::mutate(censor = case_when(
          症例.転帰情報.死亡日 != "" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
      
      Data_survival = Data_survival %>%
        dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
        dplyr::mutate(final_observe = ifelse(is.na(final_observe), 症例.管理情報.登録日, final_observe))
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, censor) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival = Data_survival %>%
        dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(Data_survival$症例.管理情報.登録日)) %>%
        dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                        as.Date(症例.EP前レジメン情報.投与開始日))
      
      Data_survival = Data_survival %>%
        dplyr::mutate(time_enroll_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.管理情報.登録日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.EP前レジメン情報.投与開始日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_enroll =
                        Data_survival$time_palliative_final -
                        Data_survival$time_enroll_final)
  
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, final_observe) %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
  
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_enroll_final) %>%
        dplyr::arrange(desc(time_enroll_final)) %>%
        dplyr::filter(time_enroll_final > 0 & !is.na(time_enroll_final) & is.finite(time_enroll_final)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      incProgress(1 / 13)
      Data_survival_tmp = Data_survival %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                        c("その他", "緩和")) %>%
        dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_palliative_enroll, time_palliative_final) %>%
        dplyr::filter(time_palliative_enroll > 0 & time_palliative_enroll < 10000 & !is.na(time_palliative_enroll) & is.finite(time_palliative_enroll)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
  
      Data_case_target = Data_case_target %>%
        dplyr::distinct(.keep_all = TRUE, C.CAT調査結果.基本項目.ハッシュID)
      Data_MAF = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","A","B","C","D","E","F") &
            Variant_Classification != "expression"
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
      if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
        Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
      }

      Data_report_TMB = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
        dplyr::select(Tumor_Sample_Barcode, TMB) %>%
        dplyr::distinct(Tumor_Sample_Barcode, .keep_all=T)
      colnames(Data_report_TMB) = c("C.CAT調査結果.基本項目.ハッシュID", "TMB")
      Data_case_target = left_join(Data_case_target, Data_report_TMB,
                                   by = "C.CAT調査結果.基本項目.ハッシュID")
      
      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      if(input$patho == "Only pathogenic muts"){
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(Evidence_level == "F")
      } else {
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(!Hugo_Symbol %in% c("TMB", "MSI") | Evidence_level == "F")
        
      }
      Data_MAF_target = Data_MAF_target  %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        .keep_all = TRUE)
      incProgress(1 / 13)
      Cancername = sort(unique(Data_case_target$Cancers))
      Total_pts = as.vector(table((Data_case_target %>%
                                     dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = T))$Cancers))
      Gene_list = unique(names(sort(table(Data_MAF_target$Hugo_Symbol),
                                      decreasing = T)))

      Data_mutation_cord = Data_cluster_ID() %>%
        dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
        dplyr::select(-C.CAT調査結果.基本項目.ハッシュID, -cluster, -driver_mutations)
      Data_case_target = left_join(Data_case_target,
                                   Data_cluster_ID() %>% dplyr::select(C.CAT調査結果.基本項目.ハッシュID, cluster, driver_mutations),
                                   by = "C.CAT調査結果.基本項目.ハッシュID")      
      
      Data_report_tmp = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Variant_Classification != "expression"
        )

      Data_report_tmp = Data_report_tmp %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      
      Data_disease = (Data_case_target %>%
                    dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID,
                                    症例.基本情報.がん種.OncoTree.))
      colnames(Data_disease) = c("Tumor_Sample_Barcode", "Cancers")
      Data_MAF_target = left_join(Data_MAF_target,
                                  Data_disease,
                                  by = "Tumor_Sample_Barcode")
      Data_MAF_target$Cancers = as.factor(Data_MAF_target$Cancers)
      Data_report_tmp = left_join(Data_report_tmp,
                                  Data_disease,
                                 by = "Tumor_Sample_Barcode")
      Data_report_tmp$Cancers = as.factor(Data_report_tmp$Cancers)
      Data_report_tmp = Data_report_tmp %>% dplyr::mutate(
        Evidence_level = case_when(
          Evidence_level == "F" & Hugo_Symbol == "TMB" ~ "A",
          Evidence_level == "" ~ "G",
          TRUE ~ Evidence_level)
        ) %>% dplyr::mutate(Drug = case_when(
          Evidence_level == "A" &
            Hugo_Symbol == "TMB" &
            Drug == "" ~ "pembrolizumab",
            TRUE ~ Drug)
        ) %>% dplyr::mutate(Resistance = case_when(
          Evidence_level %in% c("R2*","R1*","R3*","R2","R3","R","R1") ~ 1,
          TRUE ~ 0)
        ) %>%
        dplyr::distinct(Tumor_Sample_Barcode, Evidence_level, Drug, .keep_all = T)
      incProgress(1 / 13)
      
      
      
      Data_evidence_table_tmp = Data_report_tmp %>%
        dplyr::filter(Evidence_level %in% c("A","B","C")) %>%
        dplyr::select(Hugo_Symbol,
                      # Chromosome, Start_Position, Reference_Allele, Tumor_Seq_Allele2,
                      amino.acid.change, Evidence_level, Drug, Cancers)
      Data_evidence_table_tmp$tmp = paste(Data_evidence_table_tmp$Hugo_Symbol,"_",
                                          Data_evidence_table_tmp$amino.acid.change,"_",
                                          Data_evidence_table_tmp$Evidence_level,"_",
                                          Data_evidence_table_tmp$Drug,"_",
                                          Data_evidence_table_tmp$Cancers)
      Data_evidence_table = Data_evidence_table_tmp %>%
        dplyr::distinct()
      Data_evidence_table_tmp2 = data.frame(table(Data_evidence_table_tmp$tmp))
      colnames(Data_evidence_table_tmp2) = c("tmp", "Frequency")
      Data_evidence_table = Data_evidence_table %>%
        dplyr::left_join(Data_evidence_table_tmp2, by="tmp") %>%
        dplyr::arrange(desc(Frequency))
      
      output$table_evidence = DT::renderDataTable(Data_evidence_table,
                                                    server = FALSE,
                                                    filter = 'top', 
                                                    extensions = c('Buttons'), 
                                                    options = list(pageLength = 100, 
                                                                   scrollX = TRUE,
                                                                   scrollY = "1000px",
                                                                   scrollCollapse = TRUE,
                                                                   dom="Blfrtip",
                                                                   buttons = c('csv', 'copy')))
      
      Data_report_tmp = Data_report_tmp  %>%
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::filter(Resistance != 1) %>%
        dplyr::filter(!Evidence_level %in% c("F", "G")) %>%
        dplyr::distinct(Tumor_Sample_Barcode, .keep_all = T)
      Data_report_tmp$Level = as.factor(Data_report_tmp$Evidence_level)
      data_figure = Data_report_tmp %>% dplyr::select(Cancers,Level) %>% dplyr::filter(!is.na(Cancers)&!is.na(Level))
      Num_A = rep(0, length(Cancername))
      names(Num_A) = Cancername
      Num_tmp = table((data_figure %>% dplyr::filter(Level == "A"))$Cancers)
      if(length(Num_tmp) > 0){
        for(i in 1:length(Num_tmp)){
          Num_A[names(Num_tmp)[i]] = Num_tmp[i]
        }
        Num_A = fun_zero(Num_A, Total_pts)
      }
      Num_B = rep(0, length(Cancername))
      names(Num_B) =  Cancername
      Num_tmp = table((data_figure %>% dplyr::filter(Level == "B"))$Cancers)
      if(length(Num_tmp) > 0){
        for(i in 1:length(Num_tmp)){
          Num_B[names(Num_tmp)[i]] = Num_tmp[i]
        }
        Num_B = fun_zero(Num_B, Total_pts)
      }
      Num_C = rep(0, length(Cancername))
      names(Num_C) =  Cancername
      Num_tmp = table((data_figure %>% dplyr::filter(Level == "C"))$Cancers)
      if(length(Num_tmp) > 0){
        for(i in 1:length(Num_tmp)){
          Num_C[names(Num_tmp)[i]] = Num_tmp[i]
        }
        Num_C = fun_zero(Num_C, Total_pts)
      }
      Num_D = rep(0, length(Cancername))
      names(Num_D) =  Cancername
      Num_tmp = table((data_figure %>% dplyr::filter(Level == "D"))$Cancers)
      if(length(Num_tmp) > 0){
        for(i in 1:length(Num_tmp)){
          Num_D[names(Num_tmp)[i]] = Num_tmp[i]
        }
        Num_D = fun_zero(Num_D, Total_pts)
      }
      Num_E = rep(0, length(Cancername))
      names(Num_E) =  Cancername
      Num_tmp = table((data_figure %>% dplyr::filter(Level == "E"))$Cancers)
      if(length(Num_tmp) > 0){
        for(i in 1:length(Num_tmp)){
          Num_E[names(Num_tmp)[i]] = Num_tmp[i]
        }
        Num_E = fun_zero(Num_E, Total_pts)
      }
      Num_Option = rep(0, length(Cancername))
      names(Num_Option) =  Cancername
      Num_tmp = table((Data_case_target %>% dplyr::filter(症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. == "はい"))$Cancers)
      if(length(Num_tmp) > 0){
        for(i in 1:length(Num_tmp)){
          Num_Option[names(Num_tmp)[i]] = Num_tmp[i]
        }
      }
      Num_Option = fun_zero(Num_Option, Total_pts)
      Num_Receive_Tx = rep(0, length(Cancername))
      names(Num_Receive_Tx) =  Cancername
      Num_tmp =table((Data_case_target %>% dplyr::filter(症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した"))$Cancers)
      if(length(Num_tmp) > 0){
        for(i in 1:length(Num_tmp)){
          Num_Receive_Tx[names(Num_tmp)[i]] = Num_tmp[i]
        }
      }
      Num_Receive_Tx = fun_zero(Num_Receive_Tx, Total_pts)
      Num_Receive_Tx_per_Option = fun_zero(Num_Receive_Tx, Num_Option)
      Total_pts_short = rep(0, length(Cancername))
      Total_pts_long = rep(0, length(Cancername))
      Num_pre_CGP = rep(0, length(Cancername))
      Num_post_CGP = rep(0, length(Cancername))
  
      Total_pts_Bone_no = rep(0, length(Cancername))
      Total_pts_Bone = rep(0, length(Cancername))
      Total_pts_Brain_no = rep(0, length(Cancername))
      Total_pts_Brain = rep(0, length(Cancername))
      Total_pts_Lung_no = rep(0, length(Cancername))
      Total_pts_Lung = rep(0, length(Cancername))
      Total_pts_Liver_no = rep(0, length(Cancername))
      Total_pts_Liver = rep(0, length(Cancername))
  

      Disease_cluster = data.frame(matrix(rep(0, length(Cancername) * max(Data_case_target$cluster)),
                                          ncol = max(Data_case_target$cluster)))
      colnames(Disease_cluster) = seq(1,max(Data_case_target$cluster))
      rownames(Disease_cluster) = Cancername
      
      Data_case_target$diagnosis = Data_case_target$症例.基本情報.がん種.OncoTree.
      Data_survival = Data_case_target
  
      for(i in 1:length(Cancername)){
        Data_survival_tmp = Data_survival %>%
          dplyr::filter(Cancers == Cancername[i])
        if(length(Data_survival_tmp$Cancers) > 0){
          if(sum(Data_survival_tmp$time_palliative_enroll, na.rm = TRUE) > 0){
            survival_simple = survfit(Surv(time_palliative_enroll,
                                           rep(1, length(time_palliative_enroll)))~1,
                                      data=Data_survival_tmp)
            Num_pre_CGP[i] = median(survival_simple)[[1]]
          } else{
            Num_pre_CGP[i] = 0
          }
          if(sum(Data_survival_tmp$time_enroll_final, na.rm = TRUE) > 0){
            survival_simple = survfit(Surv(time_enroll_final,
                                         censor)~1,
                                    data=Data_survival_tmp)
            Num_post_CGP[i] = median(survival_simple)[[1]]
          } else{
            Num_post_CGP[i] = 0
          }
  
          Data_survival_tmp2 = Data_survival_tmp %>%
            dplyr::filter(Bone_met == "Yes")
          Total_pts_Bone[i] = dim(Data_survival_tmp2)[1]
          Data_survival_tmp2 = Data_survival_tmp %>%
            dplyr::filter(Bone_met == "No")
          Total_pts_Bone_no[i] = dim(Data_survival_tmp2)[1]
          Data_survival_tmp2 = Data_survival_tmp %>%
            dplyr::filter(Brain_met == "Yes")
          Total_pts_Brain[i] = dim(Data_survival_tmp2)[1]
          Data_survival_tmp2 = Data_survival_tmp %>%
            dplyr::filter(Brain_met == "No")
          Total_pts_Brain_no[i] = dim(Data_survival_tmp2)[1]
          Data_survival_tmp2 = Data_survival_tmp %>%
            dplyr::filter(Lung_met == "Yes")
          Total_pts_Lung[i] = dim(Data_survival_tmp2)[1]
          Data_survival_tmp2 = Data_survival_tmp %>%
            dplyr::filter(Lung_met == "No")
          Total_pts_Lung_no[i] = dim(Data_survival_tmp2)[1]
          Data_survival_tmp2 = Data_survival_tmp %>%
            dplyr::filter(Liver_met == "Yes")
          Total_pts_Liver[i] = dim(Data_survival_tmp2)[1]
          Data_survival_tmp2 = Data_survival_tmp %>%
            dplyr::filter(Liver_met == "No")
          Total_pts_Liver_no[i] = dim(Data_survival_tmp2)[1]
        }
      }
      incProgress(1 / 13)
      Data_survival$timing = "Pre"
      for(i in 1:length(Cancername)){
        Data_survival[Data_survival$Cancers == unique(Data_survival$Cancers)[i],]$timing = ifelse(
          Data_survival[Data_survival$Cancers == unique(Data_survival$Cancers)[i],]$time_palliative_enroll < Num_pre_CGP[i],
          "Pre", "Post"
        )
      }
      Diseases = Cancername
      Summary_Table = data.frame(Diseases)
      Summary_Table$sample_is_primary_tumor = 0
      Summary_Table$sample_is_metastatic_tumor = 0
      Summary_Table$sample_is_unknown_tumor = 0
      Summary_Table$tumor_purity_0_25 = 0
      Summary_Table$tumor_purity_25_50 = 0
      Summary_Table$tumor_purity_50_75 = 0
      Summary_Table$tumor_purity_75_100 = 0
      Summary_Table$tumor_purity_NA = 0
      Summary_Table$age_median = 0
      Summary_Table$age_lowest = 0
      Summary_Table$age_highest = 0
      Summary_Table$age_range_25 = 0
      Summary_Table$age_range_75 = 0
      Summary_Table$male = 0
      Summary_Table$female = 0
      Summary_Table$sex_unknown = 0
      Summary_Table$oncogenic_driver_plus = 0
      Summary_Table$oncogenic_driver_minus = 0
      Summary_Table$TMB_median = 0
      Summary_Table$TMB_lowest = 0
      Summary_Table$TMB_highest = 0
      Summary_Table$TMB_range_25 = 0
      Summary_Table$TMB_range_75 = 0
      Summary_Table$entropy = 0
  
      Summary_Table$brain = Total_pts_Brain
      Summary_Table$brain_no = Total_pts_Brain_no
      Summary_Table$bone = Total_pts_Bone
      Summary_Table$bone_no = Total_pts_Bone_no
      Summary_Table$lung = Total_pts_Lung
      Summary_Table$lung_no = Total_pts_Lung_no
      Summary_Table$liver = Total_pts_Liver
      Summary_Table$liver_no = Total_pts_Liver_no
  
      Summary_Table$option = Num_Option * 100
      Summary_Table$treat = Num_Receive_Tx * 100
      Summary_Table$time_before_CGP = Num_pre_CGP
      Summary_Table$time_after_CGP = Num_post_CGP
  
  
      for(i in 1:length(Diseases)){
        Summary_Table$sample_is_primary_tumor[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    症例.検体情報.検体採取部位.名称. == "原発巣"))$症例.検体情報.検体採取部位.名称.)
        Summary_Table$sample_is_metastatic_tumor[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    症例.検体情報.検体採取部位.名称. == "転移巣"))$症例.検体情報.検体採取部位.名称.)
        Summary_Table$sample_is_unknown_tumor[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    症例.検体情報.検体採取部位.名称. %in% c("不明","")))$症例.検体情報.検体採取部位.名称.)
        Summary_Table$tumor_purity_0_25[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    !is.na(症例.検体情報.腫瘍細胞含有割合) &
                                    症例.検体情報.腫瘍細胞含有割合 < 25))$症例.検体情報.腫瘍細胞含有割合)
        Summary_Table$tumor_purity_25_50[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    !is.na(症例.検体情報.腫瘍細胞含有割合) &
                                    症例.検体情報.腫瘍細胞含有割合 >= 25 &
                                    症例.検体情報.腫瘍細胞含有割合 < 50))$症例.検体情報.腫瘍細胞含有割合)
        Summary_Table$tumor_purity_50_75[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    !is.na(症例.検体情報.腫瘍細胞含有割合) &
                                    症例.検体情報.腫瘍細胞含有割合 >= 50 &
                                    症例.検体情報.腫瘍細胞含有割合 < 75))$症例.検体情報.腫瘍細胞含有割合)
        Summary_Table$tumor_purity_75_100[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    !is.na(症例.検体情報.腫瘍細胞含有割合) &
                                    症例.検体情報.腫瘍細胞含有割合 >= 75 &
                                    症例.検体情報.腫瘍細胞含有割合 <= 100))$症例.検体情報.腫瘍細胞含有割合)
        Summary_Table$tumor_purity_NA[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    is.na(症例.検体情報.腫瘍細胞含有割合)))$症例.検体情報.腫瘍細胞含有割合)
        Summary_Table$age_median[i] =
          quantile((Data_case_target %>%
                      dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                      !is.na(症例.基本情報.年齢)))$症例.基本情報.年齢)[[3]]
        Summary_Table$age_lowest[i] =
          quantile((Data_case_target %>%
                      dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                      !is.na(症例.基本情報.年齢)))$症例.基本情報.年齢)[[1]]
        Summary_Table$age_highest[i] =
          quantile((Data_case_target %>%
                      dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                      !is.na(症例.基本情報.年齢)))$症例.基本情報.年齢)[[5]]
        Summary_Table$age_range_25[i] =
          quantile((Data_case_target %>%
                      dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                      !is.na(症例.基本情報.年齢)))$症例.基本情報.年齢)[[2]]
        Summary_Table$age_range_75[i] =
          quantile((Data_case_target %>%
                      dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                      !is.na(症例.基本情報.年齢)))$症例.基本情報.年齢)[[4]]
        Summary_Table$male[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    !is.na(症例.基本情報.性別.名称.) &
                                    症例.基本情報.性別.名称. == "男"))$症例.基本情報.性別.名称.)
        Summary_Table$female[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    !is.na(症例.基本情報.性別.名称.) &
                                    症例.基本情報.性別.名称. == "女"))$症例.基本情報.性別.名称.)
        Summary_Table$sex_unknown[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    !is.na(症例.基本情報.性別.名称.) &
                                    症例.基本情報.性別.名称. == "未入力・不明"))$症例.基本情報.性別.名称.)
        Summary_Table$oncogenic_driver_plus[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    driver_mutations > 0))$driver_mutations)
        Summary_Table$oncogenic_driver_minus[i] =
          length((Data_case_target %>%
                    dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                    driver_mutations == 0))$driver_mutations)
        Summary_Table$TMB_median[i] =
          quantile((Data_case_target %>%
                      dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                      !is.na(TMB)))$TMB)[[3]]
        Summary_Table$TMB_lowest[i] =
          quantile((Data_case_target %>%
                      dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                      !is.na(TMB)))$TMB)[[1]]
        Summary_Table$TMB_highest[i] =
          quantile((Data_case_target %>%
                      dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                      !is.na(TMB)))$TMB)[[5]]
        Summary_Table$TMB_range_25[i] =
          quantile((Data_case_target %>%
                      dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                      !is.na(TMB)))$TMB)[[2]]
        Summary_Table$TMB_range_75[i] =
          quantile((Data_case_target %>%
                      dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i] &
                                      !is.na(TMB)))$TMB)[[4]]
        tmp_cluster = as.data.frame(table((Data_case_target %>%
                                             dplyr::filter(症例.基本情報.がん種.OncoTree. == Diseases[i]))$cluster))
        Summary_Table$entropy[i] =
          shannon.entropy(as.numeric(as.character(tmp_cluster[,2])), max(Data_case_target$cluster))
        for(j in 1:length(tmp_cluster[,1])){
          Disease_cluster[i,as.numeric(as.character(tmp_cluster[j,1]))] =
            as.numeric(as.character(tmp_cluster[j,2]))
        }
      }
      incProgress(1 / 13)
      
      testSummary = data.frame(as.vector(t(matrix(rep(1:max(Data_case_target$cluster),
                                                      length(Diseases)),
                                                      ncol =length(Diseases)))))
      colnames(testSummary) = "Cluster"
      testSummary$Histology = rep(Diseases, max(Data_case_target$cluster))
      testSummary$All_patients = rep(rep(0, length(Diseases)), max(Data_case_target$cluster))
      testSummary$Positive_patients = rep(rep(0, length(Diseases)), max(Data_case_target$cluster))
      testSummary$OddsRatio = rep(rep(1, length(Diseases)), max(Data_case_target$cluster))
      testSummary$Pvalue = rep(rep(1, length(Diseases)), max(Data_case_target$cluster))
      
      selected_genes = rep("", max(Data_survival$cluster))
      for(i in 1:max(Data_survival$cluster)){
        pos_num = data.frame(Diseases)
        pos_num$n = 0
        colnames(pos_num) = c("diagnosis", "n")
        Data_cluster_plus = Data_survival %>%
          dplyr::filter(cluster == i) %>%
          dplyr::count(diagnosis) %>%
          dplyr::arrange(-n)
        Data_cluster_plus = dplyr::full_join(Data_cluster_plus, pos_num, by = join_by(diagnosis))
        Data_cluster_plus = Data_cluster_plus[,1:2]
        colnames(Data_cluster_plus) = c("diagnosis", "n")
        Data_cluster_plus$n = replace_na(Data_cluster_plus$n, 0)
        Data_cluster_plus = Data_cluster_plus %>% dplyr::arrange(diagnosis)
        Data_cluster_plus = Data_cluster_plus$n
        pos_num = data.frame(Diseases)
        pos_num$n = 0
        colnames(pos_num) = c("diagnosis", "n")
        Data_cluster_minus = Data_survival %>%
          dplyr::filter(cluster != i) %>%
          dplyr::count(diagnosis) %>%
          dplyr::arrange(-n)
        Data_cluster_minus = dplyr::full_join(Data_cluster_minus, pos_num, by = join_by(diagnosis))
        Data_cluster_minus = Data_cluster_minus[,1:2]
        colnames(Data_cluster_minus) = c("diagnosis", "n")
        Data_cluster_minus$n = replace_na(Data_cluster_minus$n, 0)
        Data_cluster_minus = Data_cluster_minus %>% dplyr::arrange(diagnosis)
        Data_cluster_minus = Data_cluster_minus$n
        No_plus_neg = length((Data_survival %>% dplyr::filter(cluster == i))[,1][[1]]) -
          Data_cluster_plus
        No_minus_neg = length((Data_survival %>% dplyr::filter(cluster != i))[,1][[1]]) -
          Data_cluster_minus
        testSummary[((1+(i-1)*length(Diseases)):(i*length(Diseases))),3] = length((Data_survival %>% dplyr::filter(cluster == i))[,1][[1]])
        testSummary[((1+(i-1)*length(Diseases)):(i*length(Diseases))),4] = Data_cluster_plus
        for(j in 1:length(Data_cluster_plus)){
          mutation_matrix = matrix(c(Data_cluster_plus[j],
                                     Data_cluster_minus[j],
                                     No_plus_neg[j],
                                     No_minus_neg[j]),
                                   nrow = 2,
                                   dimnames = list(Cancer = c("CancerType", "OtherCancerType"),
                                                   Cluster = c(paste("cluster", i), "Other clusters")))
          testSummary[(j+(i-1)*length(Diseases)),5] =
            round(fisher.test(mutation_matrix)$estimate[[1]], digits = 3)
          testSummary[(j+(i-1)*length(Diseases)),6] =
            round(fisher.test(mutation_matrix)$p.value, digits = 3)
        }
        top_genes = testSummary[((1+(i-1)*length(Diseases)):(i*length(Diseases))),] %>%
          dplyr::arrange(desc(OddsRatio)) %>%
          dplyr::filter(OddsRatio > 1 & Pvalue < 0.05) 
        if(length(top_genes$OddsRatio) >= 1){
          top_genes = top_genes[1:min(3, length(top_genes$OddsRatio)),]$Histology
        } else{
          top_genes = ""
        }
        selected_genes[i] = paste0(i, ":", paste(top_genes,collapse = ";"))
      }
      testSummary[,"OddsRatio"] = ifelse(is.infinite(testSummary[,"OddsRatio"]), 99999, testSummary[,"OddsRatio"])
      testSummary_disease = testSummary
      
      incProgress(1 / 13)
  
      Data_cluster = Data_case_target %>% dplyr::select(C.CAT調査結果.基本項目.ハッシュID, cluster)
      colnames(Data_cluster) = c("Tumor_Sample_Barcode", "cluster")
      Data_MAF_target = left_join(Data_MAF_target, Data_cluster, by = "Tumor_Sample_Barcode")
      Mutations = sort(unique(Data_MAF_target$Hugo_Symbol))
      testSummary = data.frame(as.vector(t(matrix(rep(1:max(Data_MAF_target$cluster),
                                                      length(Mutations)),
                                                  ncol =length(Mutations)))))
      colnames(testSummary) = "Cluster"
      testSummary$mutation = rep(Mutations, max(Data_MAF_target$cluster))
      testSummary$All_patients = rep(rep(0, length(Mutations)), max(Data_MAF_target$cluster))
      testSummary$Positive_patients = rep(rep(0, length(Mutations)), max(Data_MAF_target$cluster))
      testSummary$OddsRatio = rep(rep(1, length(Mutations)), max(Data_MAF_target$cluster))
      testSummary$Pvalue = rep(rep(1, length(Mutations)), max(Data_MAF_target$cluster))
      
      for(i in 1:max(Data_MAF_target$cluster)){
        pos_num = data.frame(Mutations)
        pos_num$n = 0
        colnames(pos_num) = c("Hugo_Symbol", "n")
        Data_cluster_plus = Data_MAF_target %>%
          dplyr::filter(cluster == i) %>%
          dplyr::count(Hugo_Symbol) %>%
          dplyr::arrange(-n)
        Data_cluster_plus = dplyr::full_join(Data_cluster_plus, pos_num, by = join_by(Hugo_Symbol))
        Data_cluster_plus = Data_cluster_plus[,1:2]
        colnames(Data_cluster_plus) = c("Hugo_Symbol", "n")
        Data_cluster_plus$n = replace_na(Data_cluster_plus$n, 0)
        Data_cluster_plus = Data_cluster_plus %>% dplyr::arrange(Hugo_Symbol)
        Data_cluster_plus = Data_cluster_plus$n
        pos_num = data.frame(Mutations)
        pos_num$n = 0
        colnames(pos_num) = c("Hugo_Symbol", "n")
        Data_cluster_minus = Data_MAF_target %>%
          dplyr::filter(cluster != i) %>%
          dplyr::count(Hugo_Symbol) %>%
          dplyr::arrange(-n)
        Data_cluster_minus = dplyr::full_join(Data_cluster_minus, pos_num, by = join_by(Hugo_Symbol))
        Data_cluster_minus = Data_cluster_minus[,1:2]
        colnames(Data_cluster_minus) = c("Hugo_Symbol", "n")
        Data_cluster_minus$n = replace_na(Data_cluster_minus$n, 0)
        Data_cluster_minus = Data_cluster_minus %>% dplyr::arrange(Hugo_Symbol)
        Data_cluster_minus = Data_cluster_minus$n
        No_plus_neg = length((Data_MAF_target %>%
                                dplyr::filter(cluster == i) %>%
                                dplyr::distinct(Tumor_Sample_Barcode))[,1][[1]]) -
          Data_cluster_plus
        No_minus_neg = length((Data_MAF_target %>%
                                 dplyr::filter(cluster != i) %>%
                                 dplyr::distinct(Tumor_Sample_Barcode))[,1][[1]]) -
          Data_cluster_minus
        testSummary[((1+(i-1)*length(Mutations)):(i*length(Mutations))),3] =
          length((Data_MAF_target %>%
                   dplyr::filter(cluster == i) %>%
                   dplyr::distinct(Tumor_Sample_Barcode))[,1][[1]])
        testSummary[((1+(i-1)*length(Mutations)):(i*length(Mutations))),4] = Data_cluster_plus
        for(j in 1:length(Data_cluster_plus)){
          mutation_matrix = matrix(c(Data_cluster_plus[j],
                                     Data_cluster_minus[j],
                                     No_plus_neg[j],
                                     No_minus_neg[j]),
                                   nrow = 2,
                                   dimnames = list(Cancer = c("CancerType", "OtherCancerType"),
                                                   Cluster = c(paste("cluster", i), "Other clusters")))
          testSummary[(j+(i-1)*length(Mutations)),5] =
            round(fisher.test(mutation_matrix)$estimate[[1]], digits = 3)
          testSummary[(j+(i-1)*length(Mutations)),6] =
            round(fisher.test(mutation_matrix)$p.value, digits = 3)
        }
        top_genes = testSummary[((1+(i-1)*length(Mutations)):(i*length(Mutations))),] %>%
          dplyr::arrange(desc(OddsRatio)) %>%
          dplyr::filter(OddsRatio > 1 & Pvalue < 0.05) 
        if(length(top_genes$OddsRatio) >= 1){
          top_genes = top_genes[1:min(3, length(top_genes$OddsRatio)),]$mutation
        } else{
          top_genes = ""
        }
        selected_genes[i] = paste0(selected_genes[i], "/", paste(top_genes,collapse = ";"))
      }
      incProgress(1 / 13)
      testSummary[,"OddsRatio"] = ifelse(is.infinite(testSummary[,"OddsRatio"]), 99999, testSummary[,"OddsRatio"])
      testSummary_mutation = testSummary
      
      output$table_basic_data = DT::renderDataTable(Summary_Table,
                                                 server = FALSE,
                                                 filter = 'top', 
                                                 extensions = c('Buttons'), 
                                                 options = list(pageLength = 100, 
                                                                scrollX = TRUE,
                                                                scrollY = "1000px",
                                                                scrollCollapse = TRUE,
                                                                dom="Blfrtip",
                                                                buttons = c('csv', 'copy')))
      output$table_disease = DT::renderDataTable(testSummary_disease,
                                                 server = FALSE,
                                                 filter = 'top', 
                                                 extensions = c('Buttons'), 
                                         options = list(pageLength = 100, 
                                                        scrollX = TRUE,
                                                        scrollY = "1000px",
                                                        scrollCollapse = TRUE,
                                                        dom="Blfrtip",
                                                        buttons = c('csv', 'copy')))
      output$table_mutation = DT::renderDataTable(testSummary_mutation,
                                                 server = FALSE,
                                                 filter = 'top', 
                                                 extensions = c('Buttons'), 
                                                 options = list(pageLength = 100, 
                                                                scrollX = TRUE,
                                                                scrollY = "1000px",
                                                                scrollCollapse = TRUE,
                                                                dom="Blfrtip",
                                                                buttons = c('csv', 'copy')))
      
      Disease_cluster$Disease = rownames(Disease_cluster)
      Disease_cluster <- transform(Disease_cluster, Disease= factor(Disease, levels = sort(unique(Disease_cluster$Disease), decreasing = TRUE)))
      colnames(Disease_cluster) = c(seq(1,max(Data_survival$cluster)), "Disease")
      Data_entropy = gather(Disease_cluster, key = cluster, value = samples, -Disease)
      Data_entropy$cluster = as.numeric(Data_entropy$cluster)
      Summary_Table = transform(Summary_Table, Diseases= factor(Diseases, levels = sort(unique(Disease_cluster$Disease), decreasing = FALSE)))
      Data_entropy = transform(Data_entropy, Disease= factor(Disease, levels = sort(unique(Disease_cluster$Disease), decreasing = FALSE)))
      
      x <- data.frame(
        Cancers   = Diseases,
        Patients = Total_pts,
        Age = Summary_Table$age_median,
        Level_A = (Num_A * 100),
        Level_AB = (Num_A + Num_B) * 100,
        Level_ABC = (Num_A + Num_B + Num_C) * 100,
        Treatment_option = Num_Option * 100,
        Receive_Recom_Tx = Num_Receive_Tx * 100,
        Receive_Tx_rate_with_opt = Num_Receive_Tx_per_Option * 100,
        Survival_pre_CGP = Num_pre_CGP,
        Survival_post_CGP = Num_post_CGP
      )
      x$Cancers = factor(x$Cancers)
      x_max_patients = max(x$Patients)
      
      y <- data.frame(
        Cancers   = rep(Diseases,5),
        Patients   = rep(Total_pts,5),
        Patients_with_treatment_option   = rep(round(Total_pts*Num_Option),5),
        Patients_received_recommended_treatment   = rep(round(Total_pts*Num_Receive_Tx),5),
        Level = c(rep("A", length(Diseases)),
                  rep("B", length(Diseases)),
                  rep("C", length(Diseases)),
                  rep("D", length(Diseases)),
                  rep("E", length(Diseases))),
        Patients_with_evidence_level = c(round(Total_pts*Num_A),
                                         round(Total_pts*Num_B),
                                         round(Total_pts*Num_C),
                                         round(Total_pts*Num_D),
                                         round(Total_pts*Num_E)),
        weight = c(Num_A, Num_B, Num_C, Num_D, Num_E)
      )
      y$Cancers = factor(y$Cancers)
      y$Level = factor(y$Level)
      
      output$figure_entropy = renderPlot({
        g1 = ggplot(Summary_Table, aes(x=Diseases, y=entropy, fill="black")) +
          geom_point(stat = "identity") +
          coord_flip() +
          theme_classic() +
          theme(legend.position = "none")
        
        # g2 = ggplot(Data_entropy, aes(x=Disease, y=samples, fill=cluster)) +
        #   geom_bar(stat = "identity") +
        #   scale_fill_gradientn(colours = c("green", "black", "magenta", "darkred", "orange", "blue", "yellow")) +
        #   coord_flip() +
        #   theme_classic()
        g3 = ggplot(Data_entropy, aes(x=Disease, y=samples, fill=cluster)) +
          geom_bar(stat = "identity", position = "fill") +
          scale_fill_gradientn(colours = c("green", "black", "magenta", "darkred", "orange", "blue", "yellow")) +
          coord_flip() +
          theme_classic()
        grid.arrange(g1, g3, ncol = 2)
      })
      output$figure_base = renderPlot({
        # Data_tmp_1 =data.frame(Summary_Table$Diseases)
        # Data_tmp_1$primary = Summary_Table$sample_is_primary_tumor
        # Data_tmp_1$Primary = "Yes"
        # Data_tmp_2 =data.frame(Summary_Table$Diseases)
        # Data_tmp_2$primary = Summary_Table$sample_is_metastatic_tumor
        # Data_tmp_2$Primary = "No"
        # Data_tmp_3 =data.frame(Summary_Table$Diseases)
        # Data_tmp_3$primary = Summary_Table$sample_is_unknown_tumor
        # Data_tmp_3$Primary = "Unknown"
        # Data_tmp = rbind(Data_tmp_1, Data_tmp_2, Data_tmp_3)
        # cancer_freq_order = c("Unknown", "No", "Yes")
        # Data_tmp <- transform(Data_tmp, Primary= factor(Primary, levels = cancer_freq_order))
        # #Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Diseases), decreasing = TRUE)))
        # 
        # chart_1 = ggplot(data = Data_tmp, aes(x = Diseases, y = primary, fill=Primary)) +
        #   geom_col(position="fill") +
        #   coord_flip() +
        #   theme_classic() + 
        #   theme(legend.position = "none",
        #         axis.ticks.y =  element_blank(),
        #         axis.text.y = element_blank(),
        #         axis.title.y = element_blank(),
        #         axis.line.y = element_blank())
        # 
        # Data_tmp_1 =data.frame(Summary_Table$Diseases)
        # Data_tmp_1$purity = Summary_Table$tumor_purity_0_25
        # Data_tmp_1$Purity = "0-25%"
        # Data_tmp_2 =data.frame(Summary_Table$Diseases)
        # Data_tmp_2$purity = Summary_Table$tumor_purity_25_50
        # Data_tmp_2$Purity = "25-49%"
        # Data_tmp_3 =data.frame(Summary_Table$Diseases)
        # Data_tmp_3$purity = Summary_Table$tumor_purity_50_75
        # Data_tmp_3$Purity = "50-74%"
        # Data_tmp_4 =data.frame(Summary_Table$Diseases)
        # Data_tmp_4$purity = Summary_Table$tumor_purity_75_100
        # Data_tmp_4$Purity = "75-100%"
        # Data_tmp_5 =data.frame(Summary_Table$Diseases)
        # Data_tmp_5$purity = Summary_Table$tumor_purity_NA
        # Data_tmp_5$Purity = "NA"
        # Data_tmp = rbind(Data_tmp_1, Data_tmp_2, Data_tmp_3, Data_tmp_4, Data_tmp_5)
        # cancer_freq_order = c("NA", "75-100%", "50-74%", "25-49%", "0-25%")
        # Data_tmp <- transform(Data_tmp, Purity= factor(Purity, levels = cancer_freq_order))
        # Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Diseases), decreasing = TRUE)))
        # 
        # chart_2 = ggplot(data = Data_tmp, aes(x = Diseases, y = purity, fill=Purity)) +
        #   geom_col(position="fill") +
        #   coord_flip() +
        #   theme_classic() + 
        #   theme(legend.position = "none",
        #         axis.ticks.y =  element_blank(),
        #         axis.text.y = element_blank(),
        #         axis.title.y = element_blank(),
        #         axis.line.y = element_blank())
        # 
        Data_tmp = Data_case_target
        Data_tmp$Age = Data_tmp$症例.基本情報.年齢
        Data_tmp <- transform(Data_tmp, diagnosis= factor(diagnosis, levels = sort(unique(Data_tmp$diagnosis), decreasing = TRUE)))
        
        chart_3 = ggplot(data = Data_tmp, aes(x = diagnosis, y = Age)) +
          geom_boxplot(outlier.colour = NA) +
          coord_flip() +
          theme_classic() + 
          theme(legend.position = "none",
                axis.ticks.y =  element_blank(),
                #axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank())
  
        Data_tmp_1 =data.frame(Summary_Table$Diseases)
        Data_tmp_1$sex = Summary_Table$male
        Data_tmp_1$Sex = "Male"
        Data_tmp_2 =data.frame(Summary_Table$Diseases)
        Data_tmp_2$sex = Summary_Table$female
        Data_tmp_2$Sex = "Female"
        Data_tmp_3 =data.frame(Summary_Table$Diseases)
        Data_tmp_3$sex = Summary_Table$sex_unknown
        Data_tmp_3$Sex = "Unknown"
        Data_tmp = rbind(Data_tmp_1, Data_tmp_2, Data_tmp_3)
        Total_pts_num = sum(Data_tmp$sex)
        Data_tmp$sex = fun_zero(Data_tmp$sex, Total_pts_num)
        cancer_freq_order = c("Unknown", "Female", "Male")
        Data_tmp <- transform(Data_tmp, Sex= factor(Sex, levels = cancer_freq_order))
        Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Diseases), decreasing = TRUE)))
        
        chart_4 = ggplot(data = Data_tmp, aes(x = Diseases, y = sex, fill=Sex)) +
          geom_col(position="fill") +
          coord_flip() +
          theme_classic() + 
          theme(#legend.position = "none",
                axis.ticks.y =  element_blank(),
                axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(lim = c(0,1), labels = percent) +
          scale_fill_manual(values = rev(brewer.pal(5, "Paired")[1:length(unique(Data_tmp$Sex))])) +
          ylab("Sex") +
          guides(fill = guide_legend(reverse = TRUE))
  
        Data_tmp_1 =data.frame(Summary_Table$Diseases)
        Data_tmp_1$driver = Summary_Table$oncogenic_driver_plus
        Data_tmp_1$Driver = "Yes"
        Data_tmp_2 =data.frame(Summary_Table$Diseases)
        Data_tmp_2$driver = Summary_Table$oncogenic_driver_minus
        Data_tmp_2$Driver = "No"
        Data_tmp = rbind(Data_tmp_1, Data_tmp_2)
        Total_pts_num = sum(Data_tmp$driver)
        Data_tmp$driver = fun_zero(Data_tmp$driver, Total_pts_num)
        cancer_freq_order = c("Unknown", "No", "Yes")
        Data_tmp <- transform(Data_tmp, Driver= factor(Driver, levels = cancer_freq_order))
        Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Diseases), decreasing = TRUE)))
        
        chart_5 = ggplot(data = Data_tmp, aes(x = Diseases, y = driver, fill=Driver)) +
          geom_col(position="fill") +
          coord_flip() +
          theme_classic() + 
          theme(#legend.position = "none",
                axis.ticks.y =  element_blank(),
                axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(lim = c(0,1), labels = percent) +
          scale_fill_manual(values = rev(brewer.pal(5, "Paired")[1:length(unique(Data_tmp$Driver))])) +
          ylab("Oncogenic muts detected") +
          guides(fill = guide_legend(reverse = TRUE))
  
        Data_tmp = Data_case_target
        Data_tmp <- transform(Data_tmp, diagnosis= factor(diagnosis, levels = sort(unique(Diseases), decreasing = TRUE)))
        chart_6 = ggplot(data = Data_tmp, aes(x = diagnosis, y = TMB)) +
          geom_boxplot(outlier.colour = NA) +
          coord_flip() +
          theme_classic() + 
          theme(legend.position = "none",
                axis.ticks.y =  element_blank(),
                axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(limits = c(0, NA))
        
        Data_tmp_1 =data.frame(Summary_Table$Diseases)
        Data_tmp_1$brain_meta = Summary_Table$brain
        Data_tmp_1$Brain_meta = "(+)"
        Data_tmp_2 =data.frame(Summary_Table$Diseases)
        Data_tmp_2$brain_meta = Summary_Table$brain_no
        Data_tmp_2$Brain_meta = "(-)"
        Data_tmp = rbind(Data_tmp_1, Data_tmp_2)
        Total_pts_num = sum(Data_tmp$brain_meta)
        Data_tmp$brain_meta = fun_zero(Data_tmp$brain_meta, Total_pts_num)
        cancer_freq_order = c("(-)", "(+)")
        Data_tmp <- transform(Data_tmp, Brain_meta= factor(Brain_meta, levels = cancer_freq_order))
        Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_case_target$diagnosis), decreasing = TRUE)))
        
        chart_7 = ggplot(data = Data_tmp, aes(x = Diseases, y = brain_meta, fill=Brain_meta)) +
          geom_col(position="fill") +
          coord_flip() +
          theme_classic() + 
          theme(legend.position = "none",
                axis.ticks.y =  element_blank(),
                #axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(lim = c(0,1), labels = percent) +
          scale_fill_manual(values = rev(brewer.pal(5, "Paired")[1:length(unique(Data_tmp$Brain_meta))])) +
          ylab("Brain metastasis") +
          guides(fill = guide_legend(reverse = TRUE))
  
        Data_tmp_1 =data.frame(Summary_Table$Diseases)
        Data_tmp_1$bone_meta = Summary_Table$bone
        Data_tmp_1$Bone_meta = "(+)"
        Data_tmp_2 =data.frame(Summary_Table$Diseases)
        Data_tmp_2$bone_meta = Summary_Table$bone_no
        Data_tmp_2$Bone_meta = "(-)"
        Data_tmp = rbind(Data_tmp_1, Data_tmp_2)
        Total_pts_num = sum(Data_tmp$bone_meta)
        Data_tmp$bone_meta = fun_zero(Data_tmp$bone_meta, Total_pts_num)
        cancer_freq_order = c("(-)", "(+)")
        Data_tmp <- transform(Data_tmp, Bone_meta= factor(Bone_meta, levels = cancer_freq_order))
        Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_case_target$diagnosis), decreasing = TRUE)))
        
        chart_8 = ggplot(data = Data_tmp, aes(x = Diseases, y = bone_meta, fill=Bone_meta)) +
          geom_col(position="fill") +
          coord_flip() +
          theme_classic() + 
          theme(legend.position = "none",
                axis.ticks.y =  element_blank(),
                axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(lim = c(0,1), labels = percent) +
          scale_fill_manual(values = rev(brewer.pal(5, "Paired")[1:length(unique(Data_tmp$Bone_meta))])) +
          ylab("Bone metastasis") +
          guides(fill = guide_legend(reverse = TRUE))
  
        Data_tmp_1 =data.frame(Summary_Table$Diseases)
        Data_tmp_1$lung_meta = Summary_Table$lung
        Data_tmp_1$Lung_meta = "(+)"
        Data_tmp_2 =data.frame(Summary_Table$Diseases)
        Data_tmp_2$lung_meta = Summary_Table$lung_no
        Data_tmp_2$Lung_meta = "(-)"
        Data_tmp = rbind(Data_tmp_1, Data_tmp_2)
        Total_pts_num = sum(Data_tmp$lung_meta)
        Data_tmp$lung_meta = fun_zero(Data_tmp$lung_meta, Total_pts_num)
        cancer_freq_order = c("(-)", "(+)")
        Data_tmp <- transform(Data_tmp, Lung_meta= factor(Lung_meta, levels = cancer_freq_order))
        Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_case_target$diagnosis), decreasing = TRUE)))
        
        chart_9 = ggplot(data = Data_tmp, aes(x = Diseases, y = lung_meta, fill=Lung_meta)) +
          geom_col(position="fill") +
          coord_flip() +
          theme_classic() + 
          theme(legend.position = "none",
                axis.ticks.y =  element_blank(),
                axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(lim = c(0,1), labels = percent) +
          scale_fill_manual(values = rev(brewer.pal(5, "Paired")[1:length(unique(Data_tmp$Lung_meta))])) +
          ylab("Lung metastasis") +
          guides(fill = guide_legend(reverse = TRUE))
  
        Data_tmp_1 =data.frame(Summary_Table$Diseases)
        Data_tmp_1$liver_meta = Summary_Table$liver
        Data_tmp_1$Liver_meta = "(+)"
        Data_tmp_2 =data.frame(Summary_Table$Diseases)
        Data_tmp_2$liver_meta = Summary_Table$liver_no
        Data_tmp_2$Liver_meta = "(-)"
        Data_tmp = rbind(Data_tmp_1, Data_tmp_2)
        Total_pts_num = sum(Data_tmp$liver_meta)
        Data_tmp$liver_meta = fun_zero(Data_tmp$liver_meta, Total_pts_num)
        cancer_freq_order = c("(-)", "(+)")
        Data_tmp <- transform(Data_tmp, Liver_meta= factor(Liver_meta, levels = cancer_freq_order))
        Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_case_target$diagnosis), decreasing = TRUE)))
        
        chart_10 = ggplot(data = Data_tmp, aes(x = Diseases, y = liver_meta, fill=Liver_meta)) +
          geom_col(position="fill") +
          coord_flip() +
          theme_classic() + 
          theme(# legend.position = "none",
                axis.ticks.y =  element_blank(),
                axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(lim = c(0,1), labels = percent) +
          ylab("Liver metastasis") + 
          scale_fill_manual(values = rev(brewer.pal(5, "Paired")[1:length(unique(Data_tmp$Liver_meta))]), name = "Metastasis") +
          guides(fill = guide_legend(reverse = TRUE))
          
        Data_tmp =data.frame(Summary_Table$Diseases)
        Data_tmp$option = Summary_Table$option / 100
        Data_tmp$Option = "(+)"
        Data_tmp <- transform(Data_tmp, Option= factor(Option))
        Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_case_target$diagnosis), decreasing = TRUE)))
        
        chart_11 = ggplot(data = Data_tmp, aes(x = Diseases, y = option, fill=Option)) +
          geom_point() +
          coord_flip() +
          theme_classic() + 
          theme(legend.position = "none",
                axis.ticks.y =  element_blank(),
                #axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(lim = c(0,NA), labels = percent, name = "Pts with CTx recommendation")
  
        Data_tmp =data.frame(Summary_Table$Diseases)
        Data_tmp$treat = Summary_Table$treat / 100
        Data_tmp$Treat = "(+)"
        Data_tmp <- transform(Data_tmp, Treat= factor(Treat))
        Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_case_target$diagnosis), decreasing = TRUE)))
        
        chart_12 = ggplot(data = Data_tmp, aes(x = Diseases, y = treat, fill=Treat)) +
          geom_point() +
          coord_flip() +
          theme_classic() + 
          theme(legend.position = "none",
                axis.ticks.y =  element_blank(),
                axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(lim = c(0,NA), labels = percent, name = "Pts received recommended CTx")
          
  
        Data_tmp =data.frame(Summary_Table$Diseases)
        Data_tmp$time_before_CGP = Summary_Table$time_before_CGP
        Data_tmp$Time_before_CGP = "(+)"
        Data_tmp <- transform(Data_tmp, Time_before_CGP= factor(Time_before_CGP))
        Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_case_target$diagnosis), decreasing = TRUE)))
        
        chart_13 = ggplot(data = Data_tmp, aes(x = Diseases, y = time_before_CGP, fill=Time_before_CGP)) +
          geom_col() +
          coord_flip() +
          theme_classic() + 
          theme(legend.position = "none",
                axis.ticks.y =  element_blank(),
                axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(limits = c(0, NA)) +
          ylab("Median time from CTx to CGP (days)") + 
          scale_fill_manual(values = brewer.pal(3, "Paired")[1])
  
        Data_tmp =data.frame(Summary_Table$Diseases)
        Data_tmp$time_after_CGP = Summary_Table$time_after_CGP
        Data_tmp$Time_after_CGP = "(+)"
        Data_tmp <- transform(Data_tmp, Time_after_CGP= factor(Time_after_CGP))
        Data_tmp <- transform(Data_tmp, Diseases= factor(Diseases, levels = sort(unique(Data_case_target$diagnosis), decreasing = TRUE)))
        
        chart_14 = ggplot(data = Data_tmp, aes(x = Diseases, y = time_after_CGP, fill=Time_after_CGP)) +
          geom_col() +
          coord_flip() +
          theme_classic() + 
          theme(legend.position = "none",
                axis.ticks.y =  element_blank(),
                axis.text.y = element_blank(),
                axis.title.y = element_blank(),
                axis.line.y = element_blank()) +
          scale_y_continuous(limits = c(0, NA)) +
          ylab("Median time from CGP to death (days)") + 
          scale_fill_manual(values = brewer.pal(3, "Paired")[1])
  
        grid.arrange(chart_3, chart_4, chart_5, chart_6, chart_7, chart_8, chart_9, chart_10, chart_11, chart_12, chart_13, chart_14, ncol = 4)
      })
      incProgress(1 / 13)
      
      output$figure_drug_evidence = renderPlot({
        ggplot(y, aes(x= reorder(Cancers, desc(Cancers)), y = weight, fill = reorder(Level, desc(Level)))) +
          geom_bar(stat = "identity") +
          scale_y_continuous(labels = percent) +
          scale_fill_nejm() +
          theme_bw() +
          labs(y = "Frequency of druggable mutation",x = "Cancer type", fill = "Evidence level") +
          coord_flip() +
          guides(fill = guide_legend(reverse = TRUE))
      })
      output$table_drug_evidence = DT::renderDataTable(y,
                                                      server = FALSE,
                                                      filter = 'top', 
                                                      extensions = c('Buttons'), 
                                                      options = list(pageLength = 100, 
                                                                     scrollX = TRUE,
                                                                     scrollY = "1000px",
                                                                     scrollCollapse = TRUE,
                                                                     dom="Blfrtip",
                                                                     buttons = c('csv', 'excel', 'copy')))
      
      output$figure_CGP_pre_post = renderPlot({
        cor_A = x %>% 
          summarise(cor = round(cor.test(Survival_pre_CGP, Survival_post_CGP)$estimate,3), p = round(cor.test(Survival_pre_CGP, Survival_post_CGP)$p.value,3)) %>% 
          as.data.frame() #%>% mutate(p = ifelse(p < 0.05, "p < 0.05", paste("p =", p )))
        max_x = max(x$Survival_pre_CGP,na.rm = T)*1.05
        max_y = max(x$Survival_post_CGP,na.rm = T)*1.05
        x <- x %>% 
          mutate(point_size = Total_pts / x_max_patients * 10)
        
        g <- ggplot(x, aes(x = Survival_pre_CGP, y = Survival_post_CGP, color = Cancers)) +
          geom_point(aes(size = point_size), alpha = 0.5) +
          scale_size_identity() +  # size はすでに計算済みの値をそのまま使用
          scale_fill_nejm() +
          theme_bw() +
          coord_cartesian(xlim = c(0, max_x), ylim = c(0, max_y)) +
          labs(x = "Survival before CGP (days)",
               y = "Survival after CGP (days)",
               color = "Cancer type") +
          geom_smooth(method = "lm", se = FALSE, na.rm = TRUE, colour = "red") +
          geom_text(data = cor_A, 
                    aes(x = max_x * 0.5, y = 30, label = paste("r =", cor, ", p =", p)),
                    colour = "red")
      })
      incProgress(1 / 13)
      
      output$figure_patient_treatment = renderPlot({
        h = list()
        h[[1]] = gg_drug_plot(x, "Patients", "Level_A",
                              "Patients",
                              "Frequency of Evidence level A (%)",
                              Total_pts, x_max_patients)
        h[[2]] = gg_drug_plot(x, "Patients", "Level_AB",
                              "Patients",
                              "Frequency of Evidence level A/B (%)",
                              Total_pts, x_max_patients)
        h[[3]] = gg_drug_plot(x, "Patients", "Level_ABC",
                              "Patients",
                              "Frequency of Evidence level A/B/C (%)",
                              Total_pts, x_max_patients)
        h[[4]] = gg_drug_plot(x, "Patients", "Treatment_option",
                              "Patients",
                              "Patients with treatment option (%)",
                              Total_pts, x_max_patients)
        h[[5]] = gg_drug_plot(x, "Patients", "Receive_Recom_Tx",
                              "Patients",
                              "Patients received recommended treatment (%)",
                              Total_pts, x_max_patients)
        h[[6]] = gg_drug_plot(x, "Patients", "Receive_Tx_rate_with_opt",
                              "Patients",
                              "Rate of receiving recommended treatment (%)",
                              Total_pts, x_max_patients)
        grid.arrange(h[[1]], h[[2]], h[[3]], h[[4]], h[[5]], h[[6]], ncol = 2)
      })
      incProgress(1 / 13)
      
      output$figure_Pre_CGP_treatment = renderPlot({
        h = list()
        h[[1]] = gg_drug_plot(x, "Survival_pre_CGP", "Level_A",
                              "Survival before CGP (days)",
                              "Frequency of Evidence level A (%)",
                              Total_pts, x_max_patients)
        h[[2]] = gg_drug_plot(x, "Survival_pre_CGP", "Level_AB",
                              "Survival before CGP (days)",
                              "Frequency of Evidence level A/B (%)",
                              Total_pts, x_max_patients)
        h[[3]] = gg_drug_plot(x, "Survival_pre_CGP", "Level_ABC",
                              "Survival before CGP (days)",
                              "Frequency of Evidence level A/B/C (%)",
                              Total_pts, x_max_patients)
        h[[4]] = gg_drug_plot(x, "Survival_pre_CGP", "Treatment_option",
                              "Survival before CGP (days)",
                              "Patients with treatment option (%)",
                              Total_pts, x_max_patients)
        h[[5]] = gg_drug_plot(x, "Survival_pre_CGP", "Receive_Recom_Tx",
                              "Survival before CGP (days)",
                              "Patients received recommended treatment (%)",
                              Total_pts, x_max_patients)
        h[[6]] = gg_drug_plot(x, "Survival_pre_CGP", "Receive_Tx_rate_with_opt",
                              "Survival before CGP (days)",
                              "Rate of receiving recommended treatment (%)",
                              Total_pts, x_max_patients)
        grid.arrange(h[[1]], h[[2]], h[[3]], h[[4]], h[[5]], h[[6]], ncol = 2)
      })
      
      incProgress(1 / 13)
      
      output$figure_Post_CGP_treatment = renderPlot({
        h = list()
        h[[1]] = gg_drug_plot(x, "Survival_post_CGP", "Level_A",
                              "Survival after CGP (days)",
                              "Frequency of Evidence level A (%)",
                              Total_pts, x_max_patients)
        h[[2]] = gg_drug_plot(x, "Survival_post_CGP", "Level_AB",
                              "Survival after CGP (days)",
                              "Frequency of Evidence level A/B (%)",
                              Total_pts, x_max_patients)
        h[[3]] = gg_drug_plot(x, "Survival_post_CGP", "Level_ABC",
                              "Survival after CGP (days)",
                              "Frequency of Evidence level A/B/C (%)",
                              Total_pts, x_max_patients)
        h[[4]] = gg_drug_plot(x, "Survival_post_CGP", "Treatment_option",
                              "Survival after CGP (days)",
                              "Patients with treatment option (%)",
                              Total_pts, x_max_patients)
        h[[5]] = gg_drug_plot(x, "Survival_post_CGP", "Receive_Recom_Tx",
                              "Survival after CGP (days)",
                              "Patients received recommended treatment (%)",
                              Total_pts, x_max_patients)
        h[[6]] = gg_drug_plot(x, "Survival_post_CGP", "Receive_Tx_rate_with_opt",
                              "Survival after CGP (days)",
                              "Rate of receiving recommended treatment (%)",
                              Total_pts, x_max_patients)
        grid.arrange(h[[1]], h[[2]], h[[3]], h[[4]], h[[5]], h[[6]], ncol = 2)
      })
      
      output$figure_Age_and_treatment = renderPlot({
        h = list()
        h[[1]] = gg_drug_plot(x, "Age", "Level_A",
                              "Median age (years old)",
                              "Frequency of Evidence level A (%)",
                              Total_pts, x_max_patients)
        h[[2]] = gg_drug_plot(x, "Age", "Level_AB",
                              "Median age (years old)",
                              "Frequency of Evidence level A/B (%)",
                              Total_pts, x_max_patients)
        h[[3]] = gg_drug_plot(x, "Age", "Level_ABC",
                              "Median age (years old)",
                              "Frequency of Evidence level A/B/C (%)",
                              Total_pts, x_max_patients)
        h[[4]] = gg_drug_plot(x, "Age", "Treatment_option",
                              "Median age (years old)",
                              "Patients with treatment option (%)",
                              Total_pts, x_max_patients)
        h[[5]] = gg_drug_plot(x, "Age", "Receive_Recom_Tx",
                              "Median age (years old)",
                              "Patients received recommended treatment (%)",
                              Total_pts, x_max_patients)
        h[[6]] = gg_drug_plot(x, "Age", "Receive_Tx_rate_with_opt",
                              "Median age (years old)",
                              "Rate of receiving recommended treatment (%)",
                              Total_pts, x_max_patients)
        grid.arrange(h[[1]], h[[2]], h[[3]], h[[4]], h[[5]], h[[6]], ncol = 2)
      })

      output$figure_cluster_subtype = renderPlot({
        ggplot(Data_mutation_cord, aes(V1,V2,color = Data_case_target$Cancers)) +
          geom_point() + 
          labs(x = "UMAP 1", y = "UMAP 2",
               title = "Unsupervised clustering of oncogenic alterations") +
          theme_classic() +
          guides(color = guide_legend(title = "Histology", nrow=30))
      })
      output$figure_cluster_age = renderPlot({
        ggplot(Data_mutation_cord, aes(V1,V2,color = Data_case_target$YoungOld)) +
          geom_point() + 
          labs(x = "UMAP 1", y = "UMAP 2",
               title = "Unsupervised clustering of oncogenic alterations") +
          theme_classic() + labs(color = "年齢") +
          guides(color = guide_legend(title = "Age"))
      })
      output$figure_cluster = renderPlot({
        if(min(Data_case_target$cluster) == 0){
          Data_case_target$cluster = as.factor(Data_case_target$cluster + 1)
          Data_mutation_cord$cluster = Data_case_target$cluster
          ggplot(Data_mutation_cord, aes(V1,V2,color = cluster)) +
            geom_point() + 
            labs(x = "UMAP 1", y = "UMAP 2",
                 title = paste(
                   "DBSCAN clustering, cluster-1 is NOT-ASSIGNED, EPS:",
                   input$eps, ", minPts:", 3, sep="")) +
            theme_classic() +
            guides(color = guide_legend(title = "TOP3 histology/mutated gene", nrow=30)) +
            scale_color_discrete(labels = selected_genes) 
        } else{
          Data_case_target$cluster = as.factor(Data_case_target$cluster + 1)
          Data_mutation_cord$cluster = Data_case_target$cluster
          ggplot(Data_mutation_cord, aes(V1,V2,color = cluster)) +
            geom_point() + 
            labs(x = "UMAP 1", y = "UMAP 2",
                 title = paste(
                   "DBSCAN clustering, EPS:",
                   input$eps, ", minPts:", 3, sep="")) +
            theme_classic() +
            guides(color = guide_legend(title = "TOP3 histology/mutated gene"), nrow=30) +
            scale_color_discrete(labels = selected_genes)
        }
      })
      # output$figure_cluster_2 = renderPlot({
      #   if(min(Data_case_target$cluster) == 0){
      #     Data_case_target$cluster = as.factor(Data_case_target$cluster + 1)
      #     Data_mutation_cord$cluster = Data_case_target$cluster
      #     ggplot(Data_mutation_cord, aes(V1,V2,color = cluster)) +
      #       geom_point() + 
      #       labs(x = "UMAP 1", y = "UMAP 2",
      #            title = paste(
      #              "DBSCAN clustering, cluster-1 is NOT-ASSIGNED, EPS:",
      #              input$eps, ", minPts:", 3, sep="")) +
      #       theme_classic() +
      #       guides(color = guide_legend(title = "TOP3 histology/mutated gene", nrow=30)) +
      #       scale_color_discrete(labels = selected_genes) + theme(legend.position = "none")
      #   } else{
      #     Data_case_target$cluster = as.factor(Data_case_target$cluster + 1)
      #     Data_mutation_cord$cluster = Data_case_target$cluster
      #     ggplot(Data_mutation_cord, aes(V1,V2,color = cluster)) +
      #       geom_point() + 
      #       labs(x = "UMAP 1", y = "UMAP 2",
      #            title = paste(
      #              "DBSCAN clustering, EPS:",
      #              input$eps, ", minPts:", 3, sep="")) +
      #       theme_classic() +
      #       guides(color = guide_legend(title = "TOP3 histology/mutated gene"), nrow=30) +
      #       scale_color_discrete(labels = selected_genes) + theme(legend.position = "none")
      #   }
      # })
      incProgress(1 / 13)
    })
  })
  
  
  
  observeEvent(input$survival_CGP_analysis, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      Data_case_target = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.基本情報.年齢,
                      Lymph_met,
                      Brain_met,
                      Lung_met,
                      Bone_met,
                      Liver_met,
                      Other_met,
                      EP_option,
                      EP_treat,
                      YoungOld,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      症例.EP前レジメン情報.最良総合効果.名称.,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.化学療法レジメン名称,
                      症例.EP前レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.EP後レジメン情報.治療ライン,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.化学療法レジメン名称,
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      症例.EP後レジメン情報.レジメン継続区分.名称.,
                      症例.EP後レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.最良総合効果.名称.,
                      症例.基本情報.性別.名称.,
                      症例.基本情報.がん種.OncoTree.,
                      症例.基本情報.がん種.OncoTree..名称.,
                      症例.基本情報.がん種.OncoTree.LEVEL1.,
                      症例.検体情報.パネル.名称.,
                      症例.背景情報.ECOG.PS.名称.,
                      症例.背景情報.喫煙歴有無.名称.,
                      症例.背景情報.アルコール多飲有無.名称.,
                      症例.背景情報.重複がん有無.異なる臓器..名称.,
                      症例.背景情報.多発がん有無.同一臓器..名称.,
                      症例.がん種情報.登録時転移部位.名称.,
                      症例.検体情報.腫瘍細胞含有割合,
                      症例.検体情報.検体採取部位.名称.,
                      症例.転帰情報.転帰.名称.,
                      症例.背景情報.診断日,
                      症例.管理情報.登録日,
                      症例.転帰情報.最終生存確認日,
                      症例.転帰情報.死亡日,
                      HER2_IHC,
                      MSI_PCR,
                      MMR_IHC
        )
      if(input$HER2 == "No"){
        Data_case_target = Data_case_target %>%
          dplyr::select(-HER2_IHC)
      } else {
        Data_case_target = Data_case_target %>%
          dplyr::filter(
            HER2_IHC != "Unknown"
          )
      }
      if(input$MSI == "No"){
        Data_case_target = Data_case_target %>%
          dplyr::select(-MSI_PCR)
      } else {
        Data_case_target = Data_case_target %>%
          dplyr::filter(
            MSI_PCR != "Unknown"
          )
      }
      if(input$MMR == "No"){
        Data_case_target = Data_case_target %>%
          dplyr::select(-MMR_IHC)
      } else {
        Data_case_target = Data_case_target %>%
          dplyr::filter(
            MMR_IHC != "Unknown"
          )
      }
      Data_case_target$Cancers = Data_case_target$症例.基本情報.がん種.OncoTree.
      Data_survival = Data_case_target %>%
        dplyr::mutate(final_observe = case_when(
          症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
          症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
          症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
          TRUE ~ 症例.管理情報.登録日))
      Data_survival = Data_survival %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::mutate(censor = case_when(
          症例.転帰情報.死亡日 != "" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
      
      Data_survival = Data_survival %>%
        dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
        dplyr::mutate(final_observe = ifelse(is.na(final_observe), 症例.管理情報.登録日, final_observe))
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, censor) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival = Data_survival %>%
        dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(Data_survival$症例.管理情報.登録日)) %>%
        dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                        as.Date(症例.EP前レジメン情報.投与開始日))
      incProgress(1 / 13)
      
      Data_survival = Data_survival %>%
        dplyr::mutate(time_enroll_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.管理情報.登録日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.EP前レジメン情報.投与開始日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_enroll =
                        Data_survival$time_palliative_final -
                        Data_survival$time_enroll_final)
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, final_observe) %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_enroll_final) %>%
        dplyr::arrange(desc(time_enroll_final)) %>%
        dplyr::filter(time_enroll_final > 0 & !is.na(time_enroll_final) & is.finite(time_enroll_final)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival_tmp = Data_survival %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                        c("その他", "緩和")) %>%
        dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_palliative_enroll, time_palliative_final) %>%
        dplyr::filter(time_palliative_enroll > 0 & time_palliative_enroll < 10000 & !is.na(time_palliative_enroll) & is.finite(time_palliative_enroll)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      tmp1 = Data_case_target %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                        c("その他", "緩和")) %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.化学療法レジメン名称,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      症例.管理情報.登録日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      final_observe,
                      censor,
                      症例.EP前レジメン情報.終了理由.名称.,
                      症例.EP前レジメン情報.最良総合効果.名称.) %>%
        dplyr::distinct()
      tmp1 = tmp1 %>%
        dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                        as.Date(tmp1$症例.EP前レジメン情報.投与開始日)) %>%
        dplyr::mutate(症例.EP前レジメン情報.投与終了日 =
                        as.Date(tmp1$症例.EP前レジメン情報.投与終了日)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(tmp1$症例.管理情報.登録日))
      tmp1 = tmp1 %>%
        dplyr::mutate(Drug_length =
                        as.integer(difftime(tmp1$症例.EP前レジメン情報.投与終了日,
                                            tmp1$症例.EP前レジメン情報.投与開始日,
                                            units = "days")))
      colnames(tmp1) = c("ID",
                         "Drug",
                         "レジメン",
                         "治療ライン",
                         "投与開始日",
                         "投与終了日",
                         "Ongoing",
                         "登録日",
                         "実施目的",
                         "final_observe",
                         "censor",
                         "終了理由",
                         "最良総合効果",
                         "Drug_length")
      tmp1$TxCGP = "Pre"
      tmp2 = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.化学療法レジメン名称,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      症例.EP後レジメン情報.レジメン継続区分.名称.,
                      症例.管理情報.登録日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      final_observe,
                      censor,
                      症例.EP後レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.最良総合効果.名称.) %>%
        dplyr::distinct()
      tmp2$症例.EP前レジメン情報.実施目的.名称. = "緩和"
      tmp2 = tmp2 %>%
        dplyr::mutate(症例.EP後レジメン情報.投与開始日 =
                        as.Date(tmp2$症例.EP後レジメン情報.投与開始日)) %>%
        dplyr::mutate(症例.EP後レジメン情報.投与終了日 =
                        as.Date(tmp2$症例.EP後レジメン情報.投与終了日)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(tmp2$症例.管理情報.登録日))
      
      tmp2 = tmp2 %>%
        dplyr::mutate(Drug_length =
                        as.integer(difftime(tmp2$症例.EP後レジメン情報.投与終了日,
                                            tmp2$症例.EP後レジメン情報.投与開始日,
                                            units = "days")))
      colnames(tmp2) = c("ID",
                         "Drug",
                         "レジメン",
                         "治療ライン",
                         "投与開始日",
                         "投与終了日",
                         "Ongoing",
                         "登録日",
                         "実施目的",
                         "final_observe",
                         "censor",
                         "終了理由",
                         "最良総合効果",
                         "Drug_length")
      tmp2$TxCGP = "Post"
      Data_drug = rbind(tmp1, tmp2)
      
      if(!is.null(input$ID_drug)){
        Data_drug = data.frame(NULL)
        for(i in 1:length(input$ID_drug[,1])){
          tmp = read.csv(header = TRUE, file(input$ID_drug[[i, 'datapath']],
                                             encoding='UTF-8-BOM'))
          if("症例.EP前レジメン情報.投与開始日" %in% colnames(tmp)){
            colnames(tmp) = c("ID",
                              "登録日",
                              "治療ライン",
                              "実施目的",
                              "レジメン",
                              "Drug",
                              "投与開始日",
                              "投与終了日",
                              "Ongoing",
                              "終了理由",
                              "最良総合効果",
                              "最終生存確認日",
                              "死亡日")
            tmp$投与開始日 = as.Date(tmp$投与開始日)
            tmp$投与終了日 = as.Date(tmp$投与終了日)
            tmp = tmp %>%
              dplyr::mutate(
                Drug_length = as.integer(difftime(tmp$投与終了日,
                                                  tmp$投与開始日,
                                                  units = "days")),
                censor = case_when(
                  死亡日 != "" ~ 1,
                  TRUE ~ 0
                ),
                final_observe = case_when(
                  最終生存確認日 == "           " ~ 死亡日,
                  最終生存確認日 != "" ~ 最終生存確認日,
                  死亡日 != "" ~ 死亡日,
                  TRUE ~ 登録日)
              )
            tmp$登録日 = as.Date(tmp$登録日)
            tmp$最終生存確認日 = as.Date(tmp$最終生存確認日)
            tmp$死亡日 = as.Date(tmp$死亡日)
            tmp$final_observe = as.Date(tmp$final_observe)
            tmp = tmp %>%
              dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
              dplyr::mutate(final_observe = ifelse(is.na(final_observe), 登録日, final_observe))
            tmp = tmp %>% dplyr::select(-最終生存確認日, -死亡日)
            tmp$TxCGP = "Pre"
          } else{
            colnames(tmp) = c("ID",
                              "登録日",
                              "治療ライン",
                              "レジメン",
                              "Drug",
                              "投与開始日",
                              "投与終了日",
                              "Ongoing",
                              "終了理由",
                              "最良総合効果",
                              "最終生存確認日",
                              "死亡日")
            tmp$投与開始日 = as.Date(tmp$投与開始日)
            tmp$投与終了日 = as.Date(tmp$投与終了日)
            tmp = tmp %>%
              dplyr::mutate(
                Drug_length = as.integer(difftime(tmp$投与終了日,
                                                  tmp$投与開始日,
                                                  units = "days")),
                censor = case_when(
                  死亡日 != "" ~ 1,
                  TRUE ~ 0
                ),
                final_observe = case_when(
                  最終生存確認日 == "           " ~ 死亡日,
                  最終生存確認日 != "" ~ 最終生存確認日,
                  死亡日 != "" ~ 死亡日,
                  TRUE ~ 登録日)
              )
            tmp$登録日 = as.Date(tmp$登録日)
            tmp$最終生存確認日 = as.Date(tmp$最終生存確認日)
            tmp$死亡日 = as.Date(tmp$死亡日)
            tmp$final_observe = as.Date(tmp$final_observe)
            tmp = tmp %>%
              dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
              dplyr::mutate(final_observe = ifelse(is.na(final_observe), 登録日, final_observe))
            tmp = tmp %>% dplyr::select(-最終生存確認日, -死亡日)
            tmp$実施目的 = "緩和"
            tmp$TxCGP = "Post"
          }
          Data_drug = rbind(Data_drug, tmp)
        }
      }
      
      Data_drug = Data_drug %>%
        dplyr::mutate(final_observe =
                        as.Date(Data_drug$final_observe))
      Data_drug = Data_drug %>%
        dplyr::arrange(投与開始日) %>%
        dplyr::arrange(治療ライン) %>%
        dplyr::arrange(ID)
      
      Data_drug = Data_drug %>% dplyr::mutate(RECIST = case_when(
        最良総合効果 == "CR" ~ "5",
        最良総合効果 == "PR" ~ "4",
        最良総合効果 == "SD" ~ "3",
        最良総合効果 == "PD" ~ "2",
        最良総合効果 == "NE" ~ "1",
        最良総合効果 == "Unknown" ~ "0",
        TRUE ~ "0"
      ))

      Data_drug = Data_drug %>%
        dplyr::filter(治療ライン %in% c("１次治療","２次治療","３次治療","４次治療"))
      
      Data_drug_RECIST = (Data_drug %>%
                            dplyr::arrange(desc(RECIST),ID,desc(レジメン),治療ライン) %>%
                            dplyr::distinct(ID,治療ライン,.keep_all = T)) %>%
        dplyr::select(ID,治療ライン,RECIST)

      Data_drug = Data_drug %>% dplyr::select(-投与開始日, -投与終了日, -RECIST, -Ongoing, -Drug)
      Data_drug = left_join(Data_drug, Data_drug_RECIST, by = c('ID','治療ライン'))
      Data_drug = Data_drug %>%
        dplyr::arrange(ID) %>%
        dplyr::distinct(ID,治療ライン, .keep_all = T)
      
      Data_drug <- data.frame(Data_drug %>% 
                                dplyr::arrange(ID) %>%
                                dplyr::group_by(ID) %>%
                                dplyr::mutate(治療ライン = row_number()))
      Data_drug$治療ライン = as.character(Data_drug$治療ライン)

      ID_pre_CGP_lines = Data_drug %>%
        dplyr::filter(TxCGP == "Pre") %>%
        dplyr::select(ID, 治療ライン) %>%
        dplyr::arrange(desc(治療ライン)) %>%
      dplyr::distinct()
      colnames(ID_pre_CGP_lines) = c("ID", "CTx_lines_before_CGP")
      
      ID_pre_CGP_RECIST = Data_drug %>%
        dplyr::filter(TxCGP == "Pre") %>%
        dplyr::select(ID, RECIST) %>%
        dplyr::arrange(desc(RECIST)) %>%
      dplyr::distinct()
      colnames(ID_pre_CGP_RECIST) = c("ID", "RECIST")
      
      ID_pre_CGP_RECIST = ID_pre_CGP_RECIST %>% dplyr::mutate(RECIST = case_when(
        RECIST == "5" ~ "CR",
        RECIST == "4" ~ "PR",
        RECIST == "3" ~ "SD",
        RECIST == "2" ~ "PD",
        RECIST == "1" ~ "NE",
        RECIST == "0" ~ "NE",
        TRUE ~ "NE"
      ))
      Data_case_target$CTx_lines_before_CGP = unlist(lapply(list(Data_case_target$C.CAT調査結果.基本項目.ハッシュID), function(x) {
        as.vector(ID_pre_CGP_lines$CTx_lines_before_CGP[match(x, ID_pre_CGP_lines$ID)])}))
      Data_case_target$RECIST = unlist(lapply(list(Data_case_target$C.CAT調査結果.基本項目.ハッシュID), function(x) {
        as.vector(ID_pre_CGP_RECIST$RECIST[match(x, ID_pre_CGP_RECIST$ID)])}))
      if(sum(is.na(Data_case_target$CTx_lines_before_CGP)>0)){
        Data_case_target[is.na(Data_case_target$CTx_lines_before_CGP),]$CTx_lines_before_CGP = "0"
      }
      if(sum(is.na(Data_case_target$RECIST)>0)){
        Data_case_target[is.na(Data_case_target$RECIST),]$RECIST = "CTx_naive"
      }
      
      Data_case_target = Data_case_target %>%
        dplyr::distinct(.keep_all = TRUE, C.CAT調査結果.基本項目.ハッシュID)
      Data_MAF = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","A","B","C","D","E","F") &
            Variant_Classification != "expression"
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
        if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
        Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
      }

      Data_report_TMB = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
        dplyr::select(Tumor_Sample_Barcode, TMB) %>%
        dplyr::distinct(Tumor_Sample_Barcode, .keep_all=T)
      colnames(Data_report_TMB) = c("C.CAT調査結果.基本項目.ハッシュID", "TMB")
      Data_case_target = left_join(Data_case_target, Data_report_TMB,
                                   by = "C.CAT調査結果.基本項目.ハッシュID")
      
      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      if(input$patho == "Only pathogenic muts"){
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(Evidence_level == "F")
      }
      Data_MAF_target = Data_MAF_target %>%
        dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol, .keep_all = T)
      Gene_list = unique(names(sort(table(Data_MAF_target$Hugo_Symbol),
                                    decreasing = T)))
      if(length(Data_case_target$censor[is.na(Data_case_target$censor)])> 0){
        Data_case_target$censor[is.na(Data_case_target$censor)] = 0
      }
      incProgress(1 / 13)
      
      Data_case_target = Data_case_target %>%
      dplyr::mutate(EP_option = case_when(
        症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. ==
          "はい" ~ 1,
        TRUE ~ 0)) %>%
      dplyr::mutate(EP_treat = case_when(
        症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
          "投与した" ~ 1,
        TRUE ~ 0))
      
      Data_tmp = Data_case_target %>% dplyr::select(
        C.CAT調査結果.基本項目.ハッシュID,
        症例.EP後レジメン情報.治療ライン,
        症例.EP後レジメン情報.化学療法レジメン名称,
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
        症例.EP後レジメン情報.提示された治療薬を投与した.名称.
        ) %>%
        dplyr::distinct()
      Data_tmp = tidyr::replace_na(Data_tmp,
                                   replace = list(症例.EP後レジメン情報.治療ライン = 0,
                                                  症例.EP後レジメン情報.化学療法レジメン名称 = "",
                                                  症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = ""))
      Data_tmp = Data_tmp %>%
        dplyr::filter(症例.EP後レジメン情報.治療ライン != 0 &(
          症例.EP後レジメン情報.化学療法レジメン名称 != "" |
            症例.EP後レジメン情報.薬剤名.YJ一般名.EN. != "" |
            症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した"))
      ID_tmp = unique(Data_tmp$C.CAT調査結果.基本項目.ハッシュID)
      Data_survival = Data_case_target %>% dplyr::filter(
          !is.na(time_enroll_final) &
          is.finite(time_enroll_final) & 
          time_enroll_final > 0 &
          !is.na(censor) &
          is.finite(censor)
      )  
      Data_survival$afterCGPtreat = Data_survival$EP_treat
      Data_survival$afterCGPtreat[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp] <- 1
      Data_survival = Data_survival %>%
        dplyr::mutate(treat_group = case_when(
          afterCGPtreat == 1 & EP_treat == 1 ~ "Option(+)RecomTreat(+)",
          afterCGPtreat == 1 & EP_option == 1 ~ "Option(+)UnrecomTreat(+)",
          afterCGPtreat == 1 & EP_option == 0 ~ "Option(-)UnrecomTreat(+)",
          afterCGPtreat == 0 & EP_option == 1 ~ "Option(+)Treat(-)",
          afterCGPtreat == 0 & EP_option == 0 ~ "Option(-)Treat(-)"
        ))
      Data_survival = Data_survival %>%
        dplyr::mutate(treat_group_2 = case_when(
          afterCGPtreat == 1 & EP_treat == 1 ~ "RecomTreat(+)",
          afterCGPtreat == 1 & EP_option == 1 ~ "UnrecomTreat(+)",
          afterCGPtreat == 1 & EP_option == 0 ~ "UnrecomTreat(+)",
          afterCGPtreat == 0 & EP_option == 1 ~ "Treat(-)",
          afterCGPtreat == 0 & EP_option == 0 ~ "Treat(-)"
        ))
      Data_survival = Data_survival %>%
        dplyr::mutate(treat_group_3 = case_when(
          afterCGPtreat == 1 & EP_treat == 1 ~ "Treat(+)",
          afterCGPtreat == 1 & EP_option == 1 ~ "Treat(+)",
          afterCGPtreat == 1 & EP_option == 0 ~ "Treat(+)",
          afterCGPtreat == 0 & EP_option == 1 ~ "Treat(-)",
          afterCGPtreat == 0 & EP_option == 0 ~ "Treat(-)"
        ))
      Data_survival = Data_survival %>%
        dplyr::filter(time_enroll_final > 0 & !is.na(time_enroll_final) & is.finite(time_enroll_final))
      Cancername = sort(unique(Data_survival$Cancers))
      Total_pts = as.vector(table((Data_survival %>%
                                     dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = T))$Cancers))
      Data_MAF_target = Data_MAF_target %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_survival$C.CAT調査結果.基本項目.ハッシュID)
      incProgress(1 / 13)
      
      if(file.exists("source/treatment_reach_summary.csv")){
        survival_table = read_csv("source/treatment_reach_summary.csv",show_col_types=FALSE)
      } else{
        survival_table = data.frame(Organ = as.character(),
                                    Histology = as.character(),
                                    Patient_no = as.numeric(),
                                    Deceased_patient_no = as.numeric(),
                                    Day_180_survival = as.numeric(),
                                    Year_1_survival = as.numeric(),
                                    Year_2_survival = as.numeric(),
                                    Year_3_survival = as.numeric(),
                                    Year_4_survival = as.numeric(),
                                    Time_50percent = as.numeric(),
                                    Time_25percent = as.numeric(),
                                    Year_1_reach_rate = as.numeric(),
                                    Year_2_reach_rate = as.numeric(),
                                    Year_3_reach_rate = as.numeric(),
                                    Year_4_reach_rate = as.numeric(),
                                    All_treatment_reach_no = as.numeric(),
                                    Deceased_treatment_reach_no = as.numeric(),
                                    Treatment_rate_crossing_date = as.numeric(),
                                    Promising_survival_ratio = as.numeric()
        )
      }
      survival_table_tmp = data.frame(paste0(unique(Data_survival$症例.基本情報.がん種.OncoTree.LEVEL1.), collapse = ";"))
      colnames(survival_table_tmp) = "Organ"
      survival_table_tmp$Histology = paste0(unique(Data_survival$Cancers), collapse = ";")
      survival_table_tmp$Patient_no = length(Data_survival$Cancers)
      survival_table_data = survfit(Surv(time_enroll_final, censor)~1,
                                    data=Data_survival,conf.type = "log-log")
      if(max(survival_table_data$time) >= 180){
        survival_table_tmp$Day_180_survival = round(summary(survival_table_data, times = 180)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_1_survival = NA
      }
      if(max(survival_table_data$time) >= 365.25*1){
        survival_table_tmp$Year_1_survival = round(summary(survival_table_data, times = 365.25*1)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_1_survival = NA
      }
      if(max(survival_table_data$time) >= 365.25*2){
        survival_table_tmp$Year_2_survival = round(summary(survival_table_data, times = 365.25*2)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_2_survival = NA
      }
      if(max(survival_table_data$time) >= 365.25*3){
        survival_table_tmp$Year_3_survival = round(summary(survival_table_data, times = 365.25*3)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_3_survival = NA
      }
      if(max(survival_table_data$time) >= 365.25*4){
        survival_table_tmp$Year_4_survival = round(summary(survival_table_data, times = 365.25*4)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_4_survival = NA
      }
      survival_table_tmp$Time_50percent = quantile(survival_table_data, 0.5)$quantile[[1]]
      survival_table_tmp$Time_25percent = quantile(survival_table_data, 0.75)$quantile[[1]]
      survp = surv_curv_entry(survival_table_data, Data_survival, paste0(unique(Data_survival$症例.基本情報.がん種.OncoTree.LEVEL1.)[[1]], " ",
                                                                         survival_table_data[[1]], " patients"), "All enrolled patients", NULL, NULL)
      pdf(paste0("source/", unique(Data_survival$症例.基本情報.がん種.OncoTree.LEVEL1.)[[1]], "_survival_after_CGP.pdf"))
      print(survp, newpage = FALSE)
      dev.off()
      
      output$figure_survival_CGP_1 = renderPlot({
        g = list()
        # Survival analysis for all patients
        survival_simple_all = survfit(Surv(time_enroll_final, censor)~1,
                                  data=Data_survival,conf.type = "log-log")
        g[[1]] = surv_curv_entry(survival_simple_all, Data_survival, paste(survival_simple_all[[1]], "patients"), "All enrolled patients", NULL, NULL)
        kk=2
        # Survival analysis for EP recommendation
        survival_simple = survfit(Surv(time_enroll_final, censor)~EP_option,
                                  data=Data_survival)
        if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~EP_option, data=Data_survival, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~EP_option, data=Data_survival, rho=1)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival, paste("EP option existed for", 
                                                                     (survival_simple)[[1]][[2]], "patients"), paste("Expert panel recommended option existed:",
                                                                                                                     c("No", "Yes"), sep=""), diff_0, diff_1)
          kk = kk + 1
        }
        
        # Survival analysis for EP recommended treatment
        survival_simple = survfit(Surv(time_enroll_final, censor)~EP_treat,
                                  data=Data_survival)
        if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                            data=Data_survival, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                            data=Data_survival, rho=1)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival, paste("Recommended treatment done for", 
                                                                     (survival_simple)[[1]][[2]], "patients"),
                               paste("Recommended treatment done: ",
                                     c("No", "Yes"), sep=""), diff_0, diff_1)
          kk = kk + 1
        }
        
        # Survival analysis for EP recommended treatment only patients with treatment option
        Data_survival_option = Data_survival %>%
          dplyr::filter(EP_option == 1)
        survival_simple = survfit(Surv(time_enroll_final, censor)~EP_treat,
                                  data=Data_survival_option)
        if(length(Data_survival_option[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                            data=Data_survival_option, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~EP_treat,
                            data=Data_survival_option, rho=1)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival_option,  paste("Recommended treatment done for", 
                                                                             (survival_simple)[[1]][[2]], "patients with treatment option"), c("EP option exists but recommended treatment not done",
                                                                                                                                               "EP option exists and recommended treatment done"), diff_0, diff_1)
          kk = kk + 1
        }
        
        # Survival analysis for EP recommended treatment
        survival_simple = survfit(Surv(time_enroll_final, censor)~treat_group,
                                  data=Data_survival)
        if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~treat_group,
                            data=Data_survival, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~treat_group,
                            data=Data_survival, rho=1)
          if(length(unique(Data_survival$treat_group))==5){
            g[[kk]] = surv_curv_entry(survival_simple, Data_survival,  paste("EP option and treatment"),
                                     c("Targetable mutation (+) and recommended treatment done",
                                       "Targetable mutation (+) and other treatment done",
                                       "Targetable mutation (-) and treatment done",
                                       "Targetable mutation (+) and no treatment done",
                                       "Targetable mutation (-) and no treatment done"),
                                     diff_0, diff_1)
          } else {
            g[[kk]] = surv_curv_entry(survival_simple, Data_survival,  paste("EP option and treatment"),
                                      NULL,
                                      diff_0, diff_1)
          }
          kk = kk + 1
        }

        # Survival analysis for EP recommended treatment
        survival_simple = survfit(Surv(time_enroll_final, censor)~treat_group_2,
                                  data=Data_survival)
        if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~treat_group_2,
                            data=Data_survival, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~treat_group_2,
                            data=Data_survival, rho=1)
          if(length(unique(Data_survival$treat_group))==3){
            g[[kk]] = surv_curv_entry(survival_simple, Data_survival, paste("EP option and treatment"),
                                   c("Recommended treatment done",
                                     "No treatment done",
                                     "Not-recommended treatment done"),
                                   diff_0, diff_1)
          } else {
            g[[kk]] = surv_curv_entry(survival_simple, Data_survival, paste("EP option and treatment"),
                                      NULL,
                                      diff_0, diff_1)
            
          }
          kk = kk + 1
        }
        if(kk != 7){
          for(kkk in kk:6){
            g[[kkk]] = ggsurvplot_empty()
          }
        }
        arrange_ggsurvplots(g,print=TRUE,ncol=2,nrow=3)
      })
      incProgress(1 / 13)
       
      Data_survival_PS_matched = Data_survival %>%
        dplyr::filter(症例.背景情報.ECOG.PS.名称. != "不明" &
                        症例.基本情報.性別.名称. != "未入力・不明" &
                        !is.na(YoungOld) &
                        treat_group_2 != "UnrecomTreat(+)") %>%
        dplyr::mutate(
          PS_modified = case_when(
            症例.背景情報.ECOG.PS.名称. == "0" ~ "0",
            症例.背景情報.ECOG.PS.名称. == "1" ~ "1",
            TRUE ~ "2_4"
          ),
          treatment = case_when(
            treat_group_2 == "RecomTreat(+)" ~ 1,
            TRUE ~ 0
          ),
          Lines = case_when(
            CTx_lines_before_CGP %in% c("2", "3", "4")  ~ "2 or later",
            TRUE ~ CTx_lines_before_CGP
          ),
          Panel = case_when(
            症例.検体情報.パネル.名称. %in% c("NCC OncoPanel", "FoundationOne CDx", "GenMineTOP") ~ "Solid",
            TRUE ~ "Liquid"
          )
        )
      
      # propensity score matching considering
      # sex, age, performance status, histology, lines, metastasis
      Factor_PS = NULL
      if(length(unique(Data_survival_PS_matched$Cancers)) > 1){
        Factor_PS = c(Factor_PS, "Cancers")
      }
      if(length(unique(Data_survival_PS_matched$症例.基本情報.性別.名称.)) > 1){
        Factor_PS = c(Factor_PS, "症例.基本情報.性別.名称.")
      }
      if(length(unique(Data_survival_PS_matched$YoungOld)) > 1){
        Factor_PS = c(Factor_PS, "YoungOld")
      }
      if(length(unique(Data_survival_PS_matched$Panel)) > 1){
        Factor_PS = c(Factor_PS, "Panel")
      }
      if(length(unique(Data_survival_PS_matched$PS_modified)) > 1){
        Factor_PS = c(Factor_PS, "PS_modified")
      }
      if(length(unique(Data_survival_PS_matched$Lines)) > 1){
        Factor_PS = c(Factor_PS, "Lines")
      }
      if(length(unique(Data_survival_PS_matched$Lymph_met)) > 1){
        Factor_PS = c(Factor_PS, "Lymph_met")
      }
      if(length(unique(Data_survival_PS_matched$Brain_met)) > 1){
        Factor_PS = c(Factor_PS, "Brain_met")
      }
      if(length(unique(Data_survival_PS_matched$Lung_met)) > 1){
        Factor_PS = c(Factor_PS, "Lung_met")
      }
      if(length(unique(Data_survival_PS_matched$Bone_met)) > 1){
        Factor_PS = c(Factor_PS, "Bone_met")
      }
      if(length(unique(Data_survival_PS_matched$Liver_met)) > 1){
        Factor_PS = c(Factor_PS, "Liver_met")
      }
      formula_PS = formula(
        paste0("treatment ~", paste(Factor_PS, collapse = "+"))
      )
      matching = NA
      if(length(Factor_PS)>0 & !all(Data_survival_PS_matched$treatment == Data_survival_PS_matched$treatment[1])){
        propensityScore = glm(formula_PS,
                              family  = binomial(link = "logit"),
                              data    = Data_survival_PS_matched)
        
        Data_survival_PS_matched$PScore = propensityScore$fitted.values
        matching = Matching::Match(Y = Data_survival_PS_matched$time_enroll_final,
                         Tr = Data_survival_PS_matched$treatment,
                         X = Data_survival_PS_matched$PScore,
                         caliper=0.25,
                         ties=F,
                         replace=F)
      }
      
      output$figure_survival_CGP_CTx = renderPlot({
        g = list()
        kk = 1
        # Survival analysis for treatment effect before CGP
        Data_survival_treated = Data_survival %>%
          dplyr::filter(RECIST != "NE")
        survival_simple = survfit(Surv(time_enroll_final, censor)~RECIST,
                                  data=Data_survival_treated)
        if(length(Data_survival_treated[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~RECIST,
                            data=Data_survival_treated, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~RECIST,
                            data=Data_survival_treated, rho=1)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival_treated, "Best CTx effect before CGP",
                                   NULL, diff_0, diff_1)
          kk = kk + 1
        }
        Data_survival_treated = Data_survival
        survival_simple = survfit(Surv(time_enroll_final, censor)~CTx_lines_before_CGP,
                                  data=Data_survival_treated)
        if(length(Data_survival_treated[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~CTx_lines_before_CGP,
                            data=Data_survival_treated, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~CTx_lines_before_CGP,
                            data=Data_survival_treated, rho=1)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival_treated, "CTx lines before CGP",
                                   NULL, diff_0, diff_1)
          kk = kk + 1
        }
        
        if(!is.na(matching)[[1]]){
          matched_data = Data_survival_PS_matched[c(matching$index.treated, matching$index.control),]
          matched_fit = survfit(Surv(event = censor,
                                     time = time_enroll_final) ~ treatment, 
                                data = matched_data)
          if(sum(matched_fit$n.event) > 0){
            if(summary(matched_fit)$table[7] > 2 & summary(matched_fit)$table[8] > 2){
              tau0 = (max((matched_data %>% dplyr::filter(treatment == 0))$time_enroll_final)/365.25)
              tau1 = (max((matched_data %>% dplyr::filter(treatment == 1))$time_enroll_final)/365.25)
              tau = floor(min(tau0, tau1, input$RMST_CGP) * 10) / 10
              verify <- survRM2::rmst2(
                time = matched_data$time_enroll_final,
                status = matched_data$censor,
                arm = matched_data$treatment,
                tau = tau * 365.25,
                alpha = 0.05
              )
              
              surv_diff_est = round(verify$unadjusted.result[1], digits = 1)
              surv_diff_CI_LL = round(verify$unadjusted.result[4], digits = 1)
              surv_diff_CI_UL = round(verify$unadjusted.result[7], digits = 1)
              
              diff_0 = survdiff(Surv(time_enroll_final, censor)~treatment, data=matched_data, rho=0)
              diff_1 = survdiff(Surv(time_enroll_final, censor)~treatment, data=matched_data, rho=1)
              g[[kk]] = surv_curv_entry(matched_fit, matched_data,
                                  paste0("PS matched: age/sex/PS/metastasis/diagnosis/CTx-line/panel, ", tau, "-year RMST diff.: ",
                                         surv_diff_est, " (", surv_diff_CI_LL, "-", surv_diff_CI_UL, ")"),
                                  c("No-treatment, PS matched",
                                    "Recommended treatment, PS matched"),
                                  diff_0, diff_1)
              kk = kk + 1
            }
          }
        }
        if(input$HER2 != "No"){
          Data_survival_treated = Data_survival %>%
            dplyr::filter(!is.na(HER2_IHC))
          survival_simple = survfit(Surv(time_enroll_final, censor)~HER2_IHC,
                                    data=Data_survival_treated)
          if(length(Data_survival_treated[,1][[1]]) != survival_simple[[1]][[1]]){
            diff_0 = survdiff(Surv(time_enroll_final, censor)~HER2_IHC,
                              data=Data_survival_treated, rho=0)
            diff_1 = survdiff(Surv(time_enroll_final, censor)~HER2_IHC,
                              data=Data_survival_treated, rho=1)
            g[[kk]] = surv_curv_entry(survival_simple, Data_survival_treated, "HER2 IHC",
                                      NULL, diff_0, diff_1)
            kk = kk + 1
          }
        }
        if(input$MSI != "No"){
          Data_survival_treated = Data_survival %>%
            dplyr::filter(!is.na(MSI_PCR))
          survival_simple = survfit(Surv(time_enroll_final, censor)~MSI_PCR,
                                    data=Data_survival_treated)
          if(length(Data_survival_treated[,1][[1]]) != survival_simple[[1]][[1]]){
            diff_0 = survdiff(Surv(time_enroll_final, censor)~MSI_PCR,
                              data=Data_survival_treated, rho=0)
            diff_1 = survdiff(Surv(time_enroll_final, censor)~MSI_PCR,
                              data=Data_survival_treated, rho=1)
            g[[kk]] = surv_curv_entry(survival_simple, Data_survival_treated, "MSI PCR",
                                      NULL, diff_0, diff_1)
            kk = kk + 1
          }
        }
        if(input$MMR != "No"){
          Data_survival_treated = Data_survival %>%
            dplyr::filter(!is.na(MMR_IHC))
          survival_simple = survfit(Surv(time_enroll_final, censor)~MMR_IHC,
                                    data=Data_survival_treated)
          if(length(Data_survival_treated[,1][[1]]) != survival_simple[[1]][[1]]){
            diff_0 = survdiff(Surv(time_enroll_final, censor)~MMR_IHC,
                              data=Data_survival_treated, rho=0)
            diff_1 = survdiff(Surv(time_enroll_final, censor)~MMR_IHC,
                              data=Data_survival_treated, rho=1)
            g[[kk]] = surv_curv_entry(survival_simple, Data_survival_treated, "MMR IHC",
                                      NULL, diff_0, diff_1)
            kk = kk + 1
          }
        }
        if(kk != 7){
          for(kkk in kk:6){
            g[[kkk]] = ggsurvplot_empty()
          }
        }
        arrange_ggsurvplots(g,print=TRUE,ncol=2,nrow=3)
      })
          
      output$figure_survival_CGP_2 = renderPlot({
        g = list()
        kk = 1
        # Survival analysis for EP recommended treatment
        Data_survival_treated = Data_survival %>%
          dplyr::filter(treat_group_2 != "Treat(-)")
        survival_simple = survfit(Surv(time_enroll_final, censor)~treat_group_2,
                                  data=Data_survival_treated)
        if(length(Data_survival_treated[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~treat_group_2,
                            data=Data_survival_treated, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~treat_group_2,
                            data=Data_survival_treated, rho=1)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival_treated, paste("EP option and treatment"),
                                   c("Recommended treatment done",
                                     "Not-recommended treatment done"),
                                   diff_0, diff_1)
          kk = kk + 1
        }
        
        # Survival analysis for EP recommended treatment
        survival_simple = survfit(Surv(time_enroll_final, censor)~treat_group_3,
                                  data=Data_survival)
        if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~treat_group_3,
                            data=Data_survival, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~treat_group_3,
                            data=Data_survival, rho=1)
          #summary(survival_simple)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival, paste("EP option and treatment"),
                                   c("No treatment done",
                                     "Treatment done"),
                                   diff_0, diff_1)
          kk = kk + 1
        }
        
        # Survival analysis for ECOG performance status
        Data_survival_Performance = Data_survival %>%
          dplyr::mutate(症例.背景情報.ECOG.PS.名称. = case_when(
            症例.背景情報.ECOG.PS.名称. == "不明" ~ "Unknown",
            TRUE ~ 症例.背景情報.ECOG.PS.名称.
          ))
        survival_simple = survfit(Surv(time_enroll_final, censor)~症例.背景情報.ECOG.PS.名称.,
                                  data=Data_survival_Performance)
        if(length(Data_survival_Performance[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~症例.背景情報.ECOG.PS.名称.,
                            data=Data_survival_Performance, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~症例.背景情報.ECOG.PS.名称.,
                            data=Data_survival_Performance, rho=1)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival_Performance, "ECOG PS",
                                   paste0("Performance status:",
                                         sort(unique(Data_survival_Performance$症例.背景情報.ECOG.PS.名称.))), diff_0, diff_1)
          kk = kk + 1
        }
        
        # Survival analysis for ECOG performance status, PS 0 vs 1 vs 2/3/4
        Data_survival_PS = Data_survival %>%
          dplyr::filter(症例.背景情報.ECOG.PS.名称. != "不明") %>%
          dplyr::mutate(PS_modified = case_when(
            症例.背景情報.ECOG.PS.名称. == "0" ~ "0",
            症例.背景情報.ECOG.PS.名称. == "1" ~ "1",
            TRUE ~ "2_4"
          ))
        survival_simple = survfit(Surv(time_enroll_final, censor)~PS_modified,
                                  data=Data_survival_PS)
        if(length(Data_survival_PS[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~PS_modified,
                            data=Data_survival_PS, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~PS_modified,
                            data=Data_survival_PS, rho=1)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival_PS,
                                   paste0("ECOG PS (PS unknown patients excluded)"), paste("Performance status:",
                                          sort(unique(Data_survival_PS$PS_modified))), diff_0, diff_1)
          kk = kk + 1
        }
        
        
        # Survival analysis for diagnosis
        Data_survival_tmp = Data_survival %>% dplyr::filter(
          Cancers %in% sort(names(table(Data_survival$Cancers)))[1:min(7,length(unique(Data_survival$Cancers)))])
        Data_survival_tmp2 = Data_survival
        Data_survival_tmp2$Cancers = " ALL"
        Data_survival_tmp2 = rbind(Data_survival_tmp, Data_survival_tmp2)
        if(length(unique(Data_survival_tmp$Cancers))>1){
          survival_simple = survfit(Surv(time_enroll_final, censor)~Cancers,
                                    data=Data_survival_tmp)
          if(length(Data_survival_tmp[,1][[1]]) != survival_simple[[1]][[1]]){
            diff_0 = survdiff(Surv(time_enroll_final, censor)~Cancers,
                              data=Data_survival_tmp, rho=0)
            diff_1 = survdiff(Surv(time_enroll_final, censor)~Cancers,
                              data=Data_survival_tmp, rho=1)
            g[[kk]] = surv_curv_entry(survival_simple, Data_survival_tmp, paste("diagnosis"), sort(unique(Data_survival_tmp$Cancers)), diff_0, diff_1)
            kk = kk + 1
          }
          survival_simple = survfit(Surv(time_enroll_final, censor)~Cancers,
                                    data=Data_survival_tmp2)
          if(length(Data_survival_tmp2[,1][[1]]) != survival_simple[[1]][[1]]){
            diff_0 = survdiff(Surv(time_enroll_final, censor)~Cancers,
                              data=Data_survival_tmp2, rho=0)
            diff_1 = survdiff(Surv(time_enroll_final, censor)~Cancers,
                              data=Data_survival_tmp2, rho=1)
            g[[kk]] = surv_curv_entry(survival_simple, Data_survival_tmp2, paste("diagnosis"), sort(unique(Data_survival_tmp2$Cancers)), diff_0, diff_1)
            kk = kk + 1
          }
          
        } else{
          survival_simple = survfit(Surv(time_enroll_final, censor)~1,
                                    data=Data_survival_tmp)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival_PS, paste("Only 1 diagnosis"), unique(Data_survival_tmp$Cancers)[[1]], diff_0, diff_1)
          kk = kk + 1
        }
        
        # Survival analysis for mutation, any
        mut_gene = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
        if(length(mut_gene)==0){
          mut_gene = "TP53"
        }
        ID_mutation = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene))$Tumor_Sample_Barcode
        Data_survival = Data_survival %>% dplyr::mutate(
          mutation = case_when(
            C.CAT調査結果.基本項目.ハッシュID %in% ID_mutation ~ 0,
            TRUE ~ 1
          )
        )
        survival_simple = survfit(Surv(time_enroll_final, censor)~mutation,
                                  data=Data_survival)
        if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~mutation,
                            data=Data_survival, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~mutation,
                            data=Data_survival, rho=1)
          g[[kk]] = surv_curv_entry(survival_simple, Data_survival, paste(paste(mut_gene, collapse = "/")),
                                c(paste0(paste(mut_gene, collapse = "/"), " mut in any"), "Not mut"), diff_0, diff_1)
          kk = kk + 1
        }
        # Survival analysis for mutation, all
        mut_gene = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
        if(length(mut_gene)==0){
          mut_gene = "TP53"
        }
        ID_mutation = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene))$Tumor_Sample_Barcode
        for(jj in 1:length(mut_gene)){
          ID_mutation = ID_mutation[ID_mutation %in% (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene[jj]))$Tumor_Sample_Barcode]
        }
        if(length(ID_mutation) > 0){
          Data_survival = Data_survival %>% dplyr::mutate(
            mutation = case_when(
              C.CAT調査結果.基本項目.ハッシュID %in% ID_mutation ~ 1,
              C.CAT調査結果.基本項目.ハッシュID %in% (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene))$Tumor_Sample_Barcode ~ 2,
              TRUE ~ 4
            )
          )
          survival_simple = survfit(Surv(time_enroll_final, censor)~mutation,
                                    data=Data_survival)
          if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
            if(sum(unique(Data_survival$mutation)) == 7){
              diff_0 = survdiff(Surv(time_enroll_final, censor)~mutation,
                                data=Data_survival, rho=0)
              diff_1 = survdiff(Surv(time_enroll_final, censor)~mutation,
                                data=Data_survival, rho=1)
              g[[kk]] = surv_curv_entry(survival_simple, Data_survival, paste(paste(mut_gene, collapse = "/")),
                                        c(paste0(paste(mut_gene, collapse = "/"), " mut in all"),
                                          paste0(paste(mut_gene, collapse = "/"), " mut in some"),
                                          "Not mut"), diff_0, diff_1)
              kk = kk + 1
            } else if(sum(unique(Data_survival$mutation)) == 6){
              diff_0 = survdiff(Surv(time_enroll_final, censor)~mutation,
                                data=Data_survival, rho=0)
              diff_1 = survdiff(Surv(time_enroll_final, censor)~mutation,
                                data=Data_survival, rho=1)
              g[[kk]] = surv_curv_entry(survival_simple, Data_survival, paste(paste(mut_gene, collapse = "/")),
                                        c(paste0(paste(mut_gene, collapse = "/"), " mut in some"), "Not mut"), diff_0, diff_1)
              kk = kk + 1
            } else if(sum(unique(Data_survival$mutation)) == 5){
              diff_0 = survdiff(Surv(time_enroll_final, censor)~mutation,
                                data=Data_survival, rho=0)
              diff_1 = survdiff(Surv(time_enroll_final, censor)~mutation,
                                data=Data_survival, rho=1)
              g[[kk]] = surv_curv_entry(survival_simple, Data_survival, paste(paste(mut_gene, collapse = "/")),
                                        c(paste0(paste(mut_gene, collapse = "/"), " mut in all"), "Not mut"), diff_0, diff_1)
              kk = kk + 1
            } else {
              diff_0 = survdiff(Surv(time_enroll_final, censor)~mutation,
                                data=Data_survival, rho=0)
              diff_1 = survdiff(Surv(time_enroll_final, censor)~mutation,
                                data=Data_survival, rho=1)
              g[[kk]] = surv_curv_entry(survival_simple, Data_survival, paste(paste(mut_gene, collapse = "/")),
                                        c(paste0(paste(mut_gene, collapse = "/"), " mut in all"), paste0(paste(mut_gene, collapse = "/"), " mut in some")), diff_0, diff_1)
              kk = kk + 1
            }
          }
        }
        if(kk != 9){
          for(kkk in kk:8){
            g[[kkk]] = ggsurvplot_empty()
          }
        }
        arrange_ggsurvplots(g,print=TRUE,ncol=2,nrow=4)
      })
      incProgress(1 / 13)
      # analysis for common oncogenic mutations
      h = list()
      hs = list()
      oncogenic_genes = Data_MAF_target %>%
        dplyr::select(Tumor_Sample_Barcode, Hugo_Symbol) %>%
        dplyr::distinct() %>%
        dplyr::count(Hugo_Symbol) %>%
        dplyr::arrange(-n)
      colnames(oncogenic_genes) = c("gene_mutation", "all_patients")
      oncogenic_genes = rbind(oncogenic_genes %>% dplyr::filter(gene_mutation %in% input$gene),
                              oncogenic_genes %>% dplyr::filter(!gene_mutation %in% input$gene))
      oncogenic_genes = oncogenic_genes[1:min(length(oncogenic_genes$all_patients), input$gene_no),]
      
      gene_table = data.frame(oncogenic_genes$gene_mutation)
      colnames(gene_table) = c("Gene")
      gene_table$positive_patients = oncogenic_genes$all_patients
      gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)) / 10
      gene_table$positive_median = 0
      gene_table$positive_LL = 0
      gene_table$positive_UL = 0
      gene_table$negative_median = 0
      gene_table$negative_LL = 0
      gene_table$negative_UL = 0
      gene_table$diff_median = 0
      gene_table$diff_LL = 0
      gene_table$diff_UL = 0
      
      k = 1
      for(i in 1:length(oncogenic_genes$all_patients)){
        ID_mutation = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% oncogenic_genes$gene_mutation[i]))$Tumor_Sample_Barcode
        Data_survival = Data_survival %>% dplyr::mutate(
          gene_mut = case_when(
            C.CAT調査結果.基本項目.ハッシュID %in% ID_mutation ~ 0,
            TRUE ~ 1
          )
        )
        traditional_fit = survfit(Surv(event = censor,
                                       time = time_enroll_final) ~ gene_mut, 
                                  data = Data_survival)
        if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
          tau0 = ((max((Data_survival %>% dplyr::filter(gene_mut == 0))$time_enroll_final) - 1)/365.25)
          tau1 = ((max((Data_survival %>% dplyr::filter(gene_mut == 1))$time_enroll_final) - 1)/365.25)
          tau = floor(min(tau0, tau1, input$RMST_CGP) * 10) / 10
          verify <- survRM2::rmst2(
            time = Data_survival$time_enroll_final,
            status = Data_survival$censor,
            arm = Data_survival$gene_mut,
            tau = tau * 365.25,
            alpha = 0.05
          )

          diff_0 = survdiff(Surv(time_enroll_final, censor)~gene_mut,
                            data=Data_survival, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~gene_mut,
                            data=Data_survival, rho=1)
          mut_gene = paste(oncogenic_genes$gene_mutation[i],
                           ", top ", i, " gene", sep="")
          tmp = data.frame(summary(traditional_fit)$table)
          gene_table$positive_median[i] = round(tmp$median[1] / 365.25 * 12, digits = 1)
          gene_table$positive_LL[i] = round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1)
          gene_table$positive_UL[i] = round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1)
          gene_table$negative_median[i] = round(tmp$median[2] / 365.25 * 12, digits = 1)
          gene_table$negative_LL[i] = round(tmp$X0.95LCL[2] / 365.25 * 12, digits = 1)
          gene_table$negative_UL[i] = round(tmp$X0.95UCL[2] / 365.25 * 12, digits = 1)
          gene_table$diff_median[i] = -round(verify$unadjusted.result[1], digits = 1)
          gene_table$diff_LL[i] = -round(verify$unadjusted.result[7], digits = 1)
          gene_table$diff_UL[i] = -round(verify$unadjusted.result[4], digits = 1)
          hs[[k]] = surv_curv_entry(traditional_fit, Data_survival,
                                    paste0(tau, "-year RMST diff.: ",
                                           gene_table$diff_median[i],
                                           " (", gene_table$diff_LL[i], "-", gene_table$diff_UL[i], ") months"),
                                   c(paste0(oncogenic_genes$gene_mutation[i], " mut"), "Not mut"), diff_0, diff_1)
          k = k + 1
        } else{
          tmp = summary(traditional_fit)$table
          legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
          gene_table$negative_median[i] = round(tmp[[7]] / 365.25 * 12, digits = 1)
          gene_table$negative_LL[i] = round(tmp[[8]] / 365.25 * 12, digits = 1)
          gene_table$negative_UL[i] = round(tmp[[9]] / 365.25 * 12, digits = 1)
        }
      }
      incProgress(1 / 13)
      Gene_arrange = gene_table$Gene
      Gene_data = gene_table
      gene_table = gene_table %>% dplyr::mutate(
        Gene = factor(gene_table$Gene, levels = Gene_arrange))
      p <- 
        gene_table |>
        ggplot(aes(y = fct_rev(Gene))) + 
        theme_classic()
      p <- p +
        geom_point(aes(x=diff_median), shape=15, size=3) +
        geom_linerange(aes(xmin=diff_LL, xmax=diff_UL))
      p <- p +
        geom_vline(xintercept = 0, linetype="dashed") +
        labs(x="Median OS and difference in restricted mean survival time (months) in 2 years", y="Genes with oncogenic alteration")
      p <- p +
        coord_cartesian(ylim=c(1,length(Gene_arrange) + 1),
                        xlim=c( - max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) - 0.5,
                                max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) + 0.5))
      p <- p +
        annotate("text",
                 x = (-max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) - 0.5)/2,
                 y = length(Gene_arrange) + 1,
                 label = "mut=short survival") +
        annotate("text",
                 x = (max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) + 0.5)/2,
                 y = length(Gene_arrange) + 1,
                 label = "mut=long survival")
      p_mid <- p + 
        theme(legend.position = "none",
              axis.line.y = element_blank(),
              axis.ticks.y= element_blank(),
              axis.text.y= element_blank(),
              axis.title.y= element_blank())
      gene_table <- gene_table %>%
        dplyr::mutate(
          estimate_lab_1 = paste0(positive_median, " (", positive_LL, "-", positive_UL, ")")) %>%
        dplyr::mutate(
          estimate_lab_2 = paste0(negative_median, " (", negative_LL, "-", negative_UL, ")")) %>%
        dplyr::mutate(
          estimate_lab_3 = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
        dplyr::mutate(
          patients = paste0(positive_patients, " (", positive_freq, "%)"))
      gene_table = 
        dplyr::bind_rows(
          data.frame(
            Gene = "Gene",
            estimate_lab_1 = "Survival, mut (+)",
            estimate_lab_2 = "Survival, mut (-)",
            estimate_lab_3 = "RMST difference",
            patients = "Patients"
          ), gene_table)
      Gene_arrange = gene_table$Gene
      gene_table = gene_table %>% dplyr::mutate(
        Gene = factor(gene_table$Gene, levels = Gene_arrange))
      
      p_left <- gene_table  %>% ggplot(aes(y = fct_rev(Gene)))
      p_left <- p_left + geom_text(aes(x = 0, label = Gene), hjust = 0,
                                   fontface = ifelse(gene_table$Gene == "Gene", "bold", "bold.italic"))
      p_left <- p_left + geom_text(aes(x = 1.8, label = estimate_lab_1), hjust = 0,
                                   fontface = ifelse(gene_table$estimate_lab_1 == "Survival, mut (+)", "bold", "plain"))
      p_left <- p_left + geom_text(aes(x = 3.4, label = estimate_lab_2), hjust = 0,
                                   fontface = ifelse(gene_table$estimate_lab_2 == "Survival, mut (-)", "bold", "plain"))
      p_left <- p_left + geom_text(aes(x = 5.0, label = estimate_lab_3), hjust = 0,
                                   fontface = ifelse(gene_table$estimate_lab_3 == "RMST difference", "bold", "plain"))
      p_left <- p_left + theme_void() + coord_cartesian(xlim = c(0, 7.5))
      
      # right side of plot - pvalues
      p_right <- gene_table  |> ggplot() +
        geom_text(aes(x = 0, y = fct_rev(Gene), label = patients), hjust = 0,
                  fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")) +
        theme_void()
      
      for(kk in k:32){
        hs[[kk]] = ggsurvplot_empty()
      }
      
      # final plot arrangement
      layout <- c(patchwork::area(t = 0, l = 0, b = 30, r = 7.5),
                  patchwork::area(t = 1, l = 7.5, b = 30, r = 12.5),
                  patchwork::area(t = 0, l = 13.5, b = 30, r = 14))
      h[[1]] = p_left + p_mid + p_right + plot_layout(design = layout)

      incProgress(1 / 13)
      
      output$figure_survival_CGP_3 = renderPlot({
        h[[1]]
      })
      
      output$figure_survival_CGP_3_data = DT::renderDataTable(gene_table,
                                                                  server = FALSE,
                                                                  filter = 'top', 
                                                                  extensions = c('Buttons'), 
                                                                  options = list(pageLength = 100, 
                                                                                 scrollX = TRUE,
                                                                                 scrollY = "1000px",
                                                                                 scrollCollapse = TRUE,
                                                                                 dom="Blfrtip",
                                                                                 buttons = c('csv', 'excel', 'copy')))
      output$figure_survival_CGP_4 = renderPlot({
        arrange_ggsurvplots(hs,print=TRUE,ncol=4,nrow=8,surv.plot.height = 0.8,risk.table.height = 0.2)
      })
      incProgress(1 / 13)
      
      Data_cluster_ID_list = Data_cluster_ID() %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, cluster)
      Data_case_target = left_join(Data_case_target,
                                   Data_cluster_ID_list,
                                   by = "C.CAT調査結果.基本項目.ハッシュID")
      
      Data_forest = Data_case_target %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                        YoungOld,
                        Lymph_met,
                        Brain_met,
                        Lung_met,
                        Bone_met,
                        Liver_met,
                        #Other_met,
                        EP_option,
                        EP_treat,
                        症例.基本情報.性別.名称.,
                        症例.基本情報.がん種.OncoTree.,
                        症例.基本情報.がん種.OncoTree.LEVEL1.,
                        症例.背景情報.ECOG.PS.名称.,
                        #症例.背景情報.喫煙歴有無.名称.,
                        #症例.背景情報.アルコール多飲有無.名称.,
                        time_enroll_final,
                        censor,
                        CTx_lines_before_CGP,
                        RECIST,
                        症例.検体情報.パネル.名称.,
                        症例.基本情報.年齢,
                        cluster)
      colnames(Data_forest) = 
          c('ID', 'Age',
            'Lymph_met',
            'Brain_met',
            'Lung_met',
            'Bone_met',
            'Liver_met',
            #'Other_met',
            'EP_option',
            'EP_treat',
            'Sex', 'Histology', 'Histology_dummy', 'PS',
            #'Smoking_history', 'Alcoholic_history',
            'time_enroll_final', 'censor', "Lines", "Best_effect", "Panel","Age_raw","Cluster")
      #Data_forest$Histology_dummy = paste0(Data_forest$Histology_dummy,"_NOS")
      col_added = 0
      if(input$HER2 != "No"){
        Data_forest$HER2_IHC = Data_case_target$HER2_IHC
        Data_forest = Data_forest %>%
          dplyr::filter(
            HER2_IHC != "Unknown"
          )
        col_added = col_added + 1
      }
      if(input$MSI != "No"){
        Data_forest$MSI_PCR = Data_case_target$MSI_PCR
        Data_forest = Data_forest %>%
          dplyr::filter(
            MSI_PCR != "Unknown"
          )
        col_added = col_added + 1
      }
      if(input$MMR != "No"){
        Data_forest$MMR_IHC = Data_case_target$MMR_IHC
        Data_forest = Data_forest %>%
          dplyr::filter(
            MMR_IHC != "Unknown"
          )
        col_added = col_added + 1
      }
      
      Data_forest = Data_forest %>%
        dplyr::filter(PS != "不明" & Sex != "未入力・不明") %>%
        dplyr::mutate(
          PS = case_when(
            PS == 0 ~ "0",
            PS == 1 ~ "1",
            TRUE ~ "2_4"
          ),
          Lines = case_when(
            Lines == "0" ~ "0",
            Lines == "1" ~ "1",
            TRUE ~ "2~"
          ),
          Panel = case_when(
            Panel %in% c("NCC OncoPanel", "FoundationOne CDx", "GenMineTOP") ~ "Solid",
            TRUE ~ "Liquid"
          )
        )
      Data_forest$Panel = factor(Data_forest$Panel)
      Data_forest$Cluster = factor(Data_forest$Cluster)
      Data_forest$Best_effect[Data_forest$Best_effect == "Unknown"] = "NE"
      Data_forest$Best_effect[Data_forest$Best_effect == "NE"] = "NE"
      Data_forest$Best_effect[is.na(Data_forest$Best_effect)] = "NE"
      Data_forest$Best_effect = factor(Data_forest$Best_effect, levels = c("SD","CR","PR","PD","CTx_naive","NE"))
      if(length(unique(Data_forest$Panel))>1){
        Data_forest$Panel =  relevel(Data_forest$Panel, ref="Solid") 
      }
      frequent_organs = sort(table(Data_forest$Histology),decreasing = T)
      frequent_organs_names = names(frequent_organs[frequent_organs>=50])
      Data_forest = Data_forest %>% dplyr::mutate(
        Histology = case_when(
          Histology %in% frequent_organs_names ~ Histology,
          TRUE ~ Histology_dummy
        )
      )
      Data_ML =  Data_forest %>%
        dplyr::select(-Age, -time_enroll_final, -censor, -EP_option, -ID, -Histology_dummy)
      Data_forest =  Data_forest %>%
        dplyr::select(-Age_raw)
      gene_names = unique(c(input$gene,
                          as.character(Gene_data$Gene)))
      gene_names = gene_names[gene_names %in% unique(Data_MAF_target$Hugo_Symbol)]
      
      gene_names = gene_names[1:min(12, length(gene_names))]
      for(i in 1:length(gene_names)){
        ID_mutation = unique((Data_MAF_target %>% dplyr::filter(Hugo_Symbol == gene_names[i]))$Tumor_Sample_Barcode)
        Data_forest_tmp = Data_forest %>% 
          dplyr::select(ID) %>%
          dplyr::distinct() %>%
          dplyr::mutate(
            XXX = case_when(
              ID %in% ID_mutation ~ "mut(+)",
              TRUE ~ "mut(-)"
          )
        )
        colnames(Data_forest_tmp) = c("ID", gene_names[i])
        Data_forest = left_join(Data_forest, Data_forest_tmp, by=c("ID"))
      }
      Factor_names = colnames(Data_forest)
      Factor_names = Factor_names[!Factor_names %in% c('ID', 'time_enroll_final', 'censor', 'EP_treat')]
      
      Data_forest_tmp = Data_forest
      Disease_tmp = unique(sort(Data_forest_tmp$Histology))
      for(i in length(gene_names):1){
        if(sum(Data_forest_tmp[[(19+col_added+i)]] == "mut(+)" & Data_forest_tmp$censor == 1) < 3){
          Factor_names = Factor_names[-(15+col_added+i)]
        } else if(sum(Data_forest_tmp[[(19+col_added+i)]] != "mut(+)" & Data_forest_tmp$censor == 1) < 3){
          Factor_names = Factor_names[-(15+col_added+i)]
        }
      }
      for(i in 1:length(Disease_tmp)){
        Data_forest_tmp2 = Data_forest_tmp %>% dplyr::filter(Histology == Disease_tmp[i])
        if(sum(Data_forest_tmp2$censor == 1) < 3){
          Data_forest_tmp = Data_forest_tmp %>%
            dplyr::mutate(Histology = case_when(
              Histology == Disease_tmp[i] ~ Histology_dummy,
              TRUE ~ Histology
            )
          )
        }
      }
      Data_forest_tmp$Histology = factor(Data_forest_tmp$Histology)
      if(length(unique(Data_forest_tmp$Histology))<2){
        Factor_names = Factor_names[!Factor_names %in% c('Histology')]
      } else {
        Data_forest_tmp$Histology = relevel(Data_forest_tmp$Histology,
                                            ref=names(sort(table(Data_forest_tmp$Histology),decreasing = T))[[1]])
      }
      if(length(unique(Data_forest_tmp$Cluster))<2){
        Factor_names = Factor_names[!Factor_names %in% c('Cluster')]
      } else {
        Data_forest_tmp$Cluster = relevel(Data_forest_tmp$Cluster,
                                            ref=names(sort(table(Data_forest_tmp$Cluster),decreasing = T))[[1]])
      }
      if(sum(Data_forest_tmp$EP_option == 0 & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('EP_option')]
      } else if(sum(Data_forest_tmp$EP_option == 1 & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('EP_option')]
      }
      if(sum(Data_forest_tmp$Age == "Younger" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Age')]
      } else if(sum(Data_forest_tmp$Age == "Older" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Age')]
      }
      if(sum(Data_forest_tmp$Sex == "男" & Data_forest_tmp$censor == 1,na.rm = T) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Sex')]
        } else if(sum(Data_forest_tmp$Sex != "男" & Data_forest_tmp$censor == 1,na.rm = T) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Sex')]
      }
      if(sum(Data_forest_tmp$Panel == "Solid" & Data_forest_tmp$censor == 1,na.rm = T) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Panel')]
      } else if(sum(Data_forest_tmp$Panel != "Solid" & Data_forest_tmp$censor == 1,na.rm = T) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Panel')]
      }
      if(sum(Data_forest_tmp$PS == "0" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('PS')]
      } else if(sum(Data_forest_tmp$PS != "0" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('PS')]
      } else if(sum(Data_forest_tmp$PS == "2_4" & Data_forest_tmp$censor == 1) < 3){
        Data_forest_tmp$Lines[Data_forest_tmp$PS == "1"] = "1_4"
        Data_forest_tmp$Lines[Data_forest_tmp$PS == "2_4"] = "1_4"
      }
      if(sum(Data_forest_tmp$Lines == "1" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Lines')]
      } else if(sum(Data_forest_tmp$Lines != "1" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Lines')]
      } else if(sum(Data_forest_tmp$Lines == "0" & Data_forest_tmp$censor == 1) < 3){
        Data_forest_tmp$Lines[Data_forest_tmp$Lines == "0"] = "0~1"
        Data_forest_tmp$Lines[Data_forest_tmp$Lines == "1"] = "0~1"
      } else if(sum(Data_forest_tmp$Lines == "2~" & Data_forest_tmp$censor == 1) < 3){
        Data_forest_tmp$Lines[Data_forest_tmp$Lines == "2~"] = "1~"
        Data_forest_tmp$Lines[Data_forest_tmp$Lines == "1"] = "1~"
      }
      if(sum(Data_forest_tmp$Best_effect == "SD" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Best_effect')]
      } else if(sum(Data_forest_tmp$Best_effect != "SD" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Best_effect')]
      }
      if(sum(Data_forest_tmp$Lymph_met == "Yes" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Lymph_met')]
      } else if(sum(Data_forest_tmp$Lymph_met != "Yes" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Lymph_met')]
      }
      if(sum(Data_forest_tmp$Lung_met == "Yes" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Lung_met')]
      } else if(sum(Data_forest_tmp$Lung_met != "Yes" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Lung_met')]
      }
      if(sum(Data_forest_tmp$Brain_met == "Yes" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Brain_met')]
      } else if(sum(Data_forest_tmp$Brain_met != "Yes" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Brain_met')]
      }
      if(sum(Data_forest_tmp$Bone_met == "Yes" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Bone_met')]
      } else if(sum(Data_forest_tmp$Bone_met != "Yes" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Bone_met')]
      }
      if(sum(Data_forest_tmp$Liver_met == "Yes" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Liver_met')]
      } else if(sum(Data_forest_tmp$Liver_met != "Yes" & Data_forest_tmp$censor == 1) < 3){
        Factor_names = Factor_names[!Factor_names %in% c('Liver_met')]
      }
      if(input$HER2 != "No"){
        if(sum(Data_forest_tmp$HER2_IHC == "Positive" & Data_forest_tmp$censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('HER2_IHC')]
        } else if(sum(Data_forest_tmp$HER2_IHC != "Positive" & Data_forest_tmp$censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('HER2_IHC')]
        }
      }
      if(input$MSI != "No"){
        if(sum(Data_forest_tmp$MSI_PCR == "Positive" & Data_forest_tmp$censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('MSI_PCR')]
        } else if(sum(Data_forest_tmp$MSI_PCR != "Positive" & Data_forest_tmp$censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('MSI_PCR')]
        }
      }
      if(input$MMR != "No"){
        if(sum(Data_forest_tmp$MMR_IHC == "Positive" & Data_forest_tmp$censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('MMR_IHC')]
        } else if(sum(Data_forest_tmp$MMR_IHC != "Positive" & Data_forest_tmp$censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('MMR_IHC')]
        }
      }
      
      if(length(Factor_names) > 1){
        for(i in length(Factor_names):1){
          if(all(as.character(Data_forest_tmp[[Factor_names[i]]]) ==
                 as.character(Data_forest_tmp[[Factor_names[i]]][1]))){
            Factor_names = Factor_names[Factor_names != Factor_names[i]]
          }
        }
      }
      Factor_names = Factor_names[!Factor_names %in% c('Histology_dummy')]
      Factor_names_univariant = Factor_names
      if(length(Factor_names) > 1){
        Factor_names_tmp = Factor_names
        for(i in 1:(length(Factor_names_tmp) - 1)){
          for(j in (i+1):length(Factor_names_tmp)){
            if(Factor_names_tmp[i] %in% colnames(Data_forest_tmp) &
               Factor_names_tmp[j] %in% colnames(Data_forest_tmp)){
              if(abs(cor(as.numeric(as.factor(as.factor(Data_forest_tmp[[Factor_names_tmp[i]]]))),
                         as.numeric(as.factor(as.factor(Data_forest_tmp[[Factor_names_tmp[j]]]))))) > 0.95){
                Factor_names = Factor_names[Factor_names != Factor_names_tmp[j]]
              }
            }
          }
        }
      }
          
      incProgress(1 / 13)
      output$figure_survival_CGP_5_1 = render_gt({
        Factor_names = Factor_names[Factor_names != "Cluster"]
        Factor_names_univariant = Factor_names_univariant[Factor_names_univariant != "Cluster"]
        if(length(Factor_names) > 0){
          if("Sex" %in% colnames(Data_forest_tmp)){
            Data_forest_tmp = Data_forest_tmp %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          Data_forest_tmp_table = Data_forest_tmp %>%
            dplyr::mutate(EP_option=case_when(
              EP_option == 1 ~ "Yes",
              TRUE ~ "No"
            ))
          colnames_tmp = colnames(Data_forest_tmp_table)
          colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
          colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
          colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
          colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
          colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
          colnames_tmp[colnames_tmp == "EP_option"] = "Treatment recommended"
          colnames_tmp[colnames_tmp == "PS"] = "Performance status"
          colnames_tmp[colnames_tmp == "Lines"] = "CTx lines before CGP"
          colnames_tmp[colnames_tmp == "Best_effect"] = "Best CTx effect before CGP"
          Factor_names[Factor_names == "Lymph_met"] = "Lymphatic metastasis"
          Factor_names[Factor_names == "Liver_met"] = "Liver metastasis"
          Factor_names[Factor_names == "Lung_met"] = "Lung metastasis"
          Factor_names[Factor_names == "Bone_met"] = "Bone metastasis"
          Factor_names[Factor_names == "Brain_met"] = "Brain metastasis"
          Factor_names[Factor_names == "EP_option"] = "Treatment recommended"
          Factor_names[Factor_names == "PS"] = "Performance status"
          Factor_names[Factor_names == "Lines"] = "CTx lines before CGP"
          Factor_names[Factor_names == "Best_effect"] = "Best CTx effect before CGP"
          Factor_names_univariant[Factor_names_univariant == "Lymph_met"] = "Lymphatic metastasis"
          Factor_names_univariant[Factor_names_univariant == "Liver_met"] = "Liver metastasis"
          Factor_names_univariant[Factor_names_univariant == "Lung_met"] = "Lung metastasis"
          Factor_names_univariant[Factor_names_univariant == "Bone_met"] = "Bone metastasis"
          Factor_names_univariant[Factor_names_univariant == "Brain_met"] = "Brain metastasis"
          Factor_names_univariant[Factor_names_univariant == "EP_option"] = "Treatment recommended"
          Factor_names_univariant[Factor_names_univariant == "PS"] = "Performance status"
          Factor_names_univariant[Factor_names_univariant == "Lines"] = "CTx lines before CGP"
          Factor_names_univariant[Factor_names_univariant == "Best_effect"] = "Best CTx effect before CGP"
          
          colnames(Data_forest_tmp_table) = colnames_tmp
          Factor_names = Factor_names[Factor_names != "CTx lines before CGP"]
          Formula = as.formula(paste0("Surv(time_enroll_final, censor) ~ ",
                                        paste( paste0("`",Factor_names,"`"), collapse = " + ")))
          linelistsurv_cox <-  coxph(formula = Formula,
                                     data = Data_forest_tmp_table,
                                     control = coxph.control(iter.max = 50))

          univ_tab <- Data_forest_tmp_table %>% 
            tbl_uvregression(                         ## 単変量解析の表を生成
              method = coxph,                           ## 実行したい回帰（一般化線形モデル）を定義
              y = Surv(time = time_enroll_final, event = censor),
              include = Factor_names_univariant,                            ## アウトカム変数を定義
              exponentiate = TRUE                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
            ) |>
            add_global_p() |> # add global p-value
            add_n(location = "level") |>
            add_nevent(location = "level") |> # add number of events of the outcome
            add_q() |> # adjusts global p-values for multiple testing
            bold_p() |> # bold p-values under a given threshold (default 0.05)
            modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
            modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
            modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
            bold_labels()

          final_mv_reg <- linelistsurv_cox %>%
            stats::step(direction = "backward", trace = FALSE)
          if(!is.null(final_mv_reg$xlevels)){
            mv_tab = final_mv_reg |>
              tbl_regression(                         ## 多変量解析の表を生成
                exponentiate = TRUE                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
              ) |>
              add_global_p() |> # add global p-value
              add_q() |> # adjusts global p-values for multiple testing
              bold_p() |> # bold p-values under a given threshold (default 0.05)
              modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              bold_labels()
            tbl_merge(
              tbls = list(univ_tab, mv_tab),
              tab_spanner = c("**Univariate**", "**Multivariable**")) |>
              modify_caption("Hazard ratio for death after CGP (Only factors with >2 events observed)") |> as_gt()
          } else {
            univ_tab |>
              modify_caption("Hazard ratio for death after CGP, no significant factor in multivariable analysis (Only factors with >2 events observed)") |> as_gt()
          }
        }
      })

      output$figure_survival_CGP_5_2 = render_gt({
        Factor_names = Factor_names[!Factor_names %in% gene_names]
        Factor_names_univariant = Factor_names_univariant[!Factor_names_univariant %in% gene_names]
        if(length(Factor_names) > 0){
          if("Sex" %in% colnames(Data_forest_tmp)){
            Data_forest_tmp = Data_forest_tmp %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          Data_forest_tmp_table = Data_forest_tmp %>%
            dplyr::mutate(EP_option=case_when(
              EP_option == 1 ~ "Yes",
              TRUE ~ "No"
            ))
          colnames_tmp = colnames(Data_forest_tmp_table)
          colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
          colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
          colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
          colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
          colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
          colnames_tmp[colnames_tmp == "EP_option"] = "Treatment recommended"
          colnames_tmp[colnames_tmp == "PS"] = "Performance status"
          colnames_tmp[colnames_tmp == "Lines"] = "CTx lines before CGP"
          colnames_tmp[colnames_tmp == "Best_effect"] = "Best CTx effect before CGP"
          Factor_names[Factor_names == "Lymph_met"] = "Lymphatic metastasis"
          Factor_names[Factor_names == "Liver_met"] = "Liver metastasis"
          Factor_names[Factor_names == "Lung_met"] = "Lung metastasis"
          Factor_names[Factor_names == "Bone_met"] = "Bone metastasis"
          Factor_names[Factor_names == "Brain_met"] = "Brain metastasis"
          Factor_names[Factor_names == "EP_option"] = "Treatment recommended"
          Factor_names[Factor_names == "PS"] = "Performance status"
          Factor_names[Factor_names == "Lines"] = "CTx lines before CGP"
          Factor_names[Factor_names == "Best_effect"] = "Best CTx effect before CGP"
          Factor_names_univariant[Factor_names_univariant == "Lymph_met"] = "Lymphatic metastasis"
          Factor_names_univariant[Factor_names_univariant == "Liver_met"] = "Liver metastasis"
          Factor_names_univariant[Factor_names_univariant == "Lung_met"] = "Lung metastasis"
          Factor_names_univariant[Factor_names_univariant == "Bone_met"] = "Bone metastasis"
          Factor_names_univariant[Factor_names_univariant == "Brain_met"] = "Brain metastasis"
          Factor_names_univariant[Factor_names_univariant == "EP_option"] = "Treatment recommended"
          Factor_names_univariant[Factor_names_univariant == "PS"] = "Performance status"
          Factor_names_univariant[Factor_names_univariant == "Lines"] = "CTx lines before CGP"
          Factor_names_univariant[Factor_names_univariant == "Best_effect"] = "Best CTx effect before CGP"
          
          colnames(Data_forest_tmp_table) = colnames_tmp
          Factor_names = Factor_names[Factor_names != "CTx lines before CGP"]
          Formula = as.formula(paste0("Surv(time_enroll_final, censor) ~ ",
                                      paste( paste0("`",Factor_names,"`"), collapse = " + ")))
          linelistsurv_cox <-  coxph(formula = Formula,
                                     data = Data_forest_tmp_table,
                                     control = coxph.control(iter.max = 50))
          
          univ_tab <- Data_forest_tmp_table %>% 
            tbl_uvregression(                         ## 単変量解析の表を生成
              method = coxph,                           ## 実行したい回帰（一般化線形モデル）を定義
              y = Surv(time = time_enroll_final, event = censor),
              include = Factor_names_univariant,                            ## アウトカム変数を定義
              exponentiate = TRUE                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
            ) |>
            add_global_p() |> # add global p-value
            add_n(location = "level") |>
            add_nevent(location = "level") |> # add number of events of the outcome
            add_q() |> # adjusts global p-values for multiple testing
            bold_p() |> # bold p-values under a given threshold (default 0.05)
            modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
            modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
            modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
            bold_labels()
          
          final_mv_reg <- linelistsurv_cox %>%
            stats::step(direction = "backward", trace = FALSE)
          if(!is.null(final_mv_reg$xlevels)){
            mv_tab = final_mv_reg |>
              tbl_regression(                         ## 多変量解析の表を生成
                exponentiate = TRUE                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
              ) |>
              add_global_p() |> # add global p-value
              add_q() |> # adjusts global p-values for multiple testing
              bold_p() |> # bold p-values under a given threshold (default 0.05)
              modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              bold_labels()
            tbl_merge(
              tbls = list(univ_tab, mv_tab),
              tab_spanner = c("**Univariate**", "**Multivariable**")) |>
              modify_caption("Hazard ratio for death after CGP (Only factors with >2 events observed)") |> as_gt()
          } else {
            univ_tab |>
              modify_caption("Hazard ratio for death after CGP, no significant factor in multivariable analysis (Only factors with >2 events observed)") |> as_gt()
          }
        }
      })

      Data_reach_rate_all = Data_survival %>%
        dplyr::filter(censor == 1)  %>%
        dplyr::arrange(time_enroll_final)%>%
        dplyr::mutate(Treatment_reach = case_when(
          treat_group_2 == "RecomTreat(+)" ~ 1,
          TRUE ~ 0
        ))
      Data_reach_rate_all$reach_rate = 0
      reached_pts_all = 0
      for(pts in 1:length(Data_reach_rate_all$Treatment_reach)){
        if(Data_reach_rate_all$Treatment_reach[pts] == 1){
          reached_pts_all = reached_pts_all + 1
        }
        Data_reach_rate_all$reach_rate[pts] = reached_pts_all / pts
      }
      Overall_average = sum(Data_survival$treat_group_2 == "RecomTreat(+)") / length(Data_survival$treat_group_2)
      Data_reach_rate_all = Data_reach_rate_all %>% dplyr::arrange(-time_enroll_final)
      if(min(Data_reach_rate_all$time_enroll_final) <= 365.25*1){
        survival_table_tmp$Year_1_reach_rate = round((Data_reach_rate_all %>% dplyr::filter(time_enroll_final <= 365.25*1))[1,"reach_rate"], digits = 3)[[1]]
      } else{
        survival_table_tmp$Year_1_reach_rate = NA
      }
      if(min(Data_reach_rate_all$time_enroll_final) <= 365.25*2){
        survival_table_tmp$Year_2_reach_rate = round((Data_reach_rate_all %>% dplyr::filter(time_enroll_final <= 365.25*2))[1,"reach_rate"], digits = 3)[[1]]
      } else{
        survival_table_tmp$Year_2_reach_rate = NA
      }
      if(min(Data_reach_rate_all$time_enroll_final) <= 365.25*3){
        survival_table_tmp$Year_3_reach_rate = round((Data_reach_rate_all %>% dplyr::filter(time_enroll_final <= 365.25*3))[1,"reach_rate"], digits = 3)[[1]]
      } else{
        survival_table_tmp$Year_3_reach_rate = NA
      }
      if(min(Data_reach_rate_all$time_enroll_final) <= 365.25*4){
        survival_table_tmp$Year_4_reach_rate = round((Data_reach_rate_all %>% dplyr::filter(time_enroll_final <= 365.25*4))[1,"reach_rate"], digits = 3)[[1]]
      } else{
        survival_table_tmp$Year_4_reach_rate = NA
      }
      reachp_all = Data_reach_rate_all %>%
        ggplot() +
        geom_line(aes(x = time_enroll_final, 
                      y = reach_rate)) +
        labs(x = paste0("Time from enrollment (days), ",
                        unique(Data_survival$症例.基本情報.がん種.OncoTree.LEVEL1.)[[1]],
                        " deceased ",
                        length(Data_reach_rate_all$Treatment_reach),
                        " patients"),
             y = "Cummulative average treatment reach rate") +
        ggtitle("Survival period after CGP") +
        theme_bw()
      pdf(paste0("source/", unique(Data_survival$症例.基本情報.がん種.OncoTree.LEVEL1.)[[1]], "_treatment_reach_rate_after_CGP_cumulative.pdf"))
      print(reachp_all, newpage = FALSE)
      dev.off()

      Data_reach_rate = Data_survival %>% dplyr::filter(censor == 1)
      Data_reach_rate = Data_reach_rate %>%
        dplyr::mutate(Treatment_reach = case_when(
          treat_group_2 == "RecomTreat(+)" ~ 1,
          TRUE ~ 0
        ))
      Data_reach_rate = Data_reach_rate %>% dplyr::arrange(time_enroll_final)
      Data_reach_rate$reach_rate = 0
      reached_pts = 0
      for(pts in 1:length(Data_reach_rate$Treatment_reach)){
        if(Data_reach_rate$Treatment_reach[pts] == 1){
          reached_pts = reached_pts + 1
        }
        Data_reach_rate$reach_rate[pts] = reached_pts / pts
      }

      reachp = Data_reach_rate %>% 
        ggplot(aes(x = time_enroll_final,
                   y = Treatment_reach)) +
        geom_smooth(method = "loess", span = 0.75, se = F) +
        geom_hline(yintercept=Overall_average, linetype=2, alpha=0.7, linewidth=0.5, color='red') +
        geom_point() + 
        labs(x = paste0("Time from enrollment (days), ",
                        unique(Data_survival$症例.基本情報.がん種.OncoTree.LEVEL1.)[[1]],
                        " deceased ",
        length(Data_reach_rate$Treatment_reach), " patients"), y = "Average treatment reach rate") +
        theme_bw()
      reachp_info <- ggplot_build(reachp)
      threshold_x = as.integer(reachp_info$data[[1]]$x[reachp_info$data[[1]]$y > Overall_average][1])
      reachp = reachp +
        geom_vline(xintercept=threshold_x, linetype=2, alpha=0.7, linewidth=0.5, color='green') +
        geom_vline(xintercept=survival_table_tmp$Time_50percent, linetype=2, alpha=0.7, linewidth=0.5, color='purple') +
        ggtitle("Survival period after CGP and average treatment reach rate",
                subtitle = paste0("Red line: overall average rate including censored patients, crossing at ", threshold_x ," days\n",
                                  "Green line: the day treatment reach rate reached the average (", round(Overall_average, digits = 3) * 100, "%)\n",
                                  "Purple line: median overall survival (", survival_table_tmp$Time_50percent," days)"))
        
      pdf(paste0("source/", unique(Data_survival$症例.基本情報.がん種.OncoTree.LEVEL1.)[[1]], "_treatment_reach_rate_after_CGP_loess.pdf"))
      print(reachp, newpage = FALSE)
      dev.off()

      Data_survival_180 = Data_reach_rate %>% dplyr::filter(time_enroll_final <= 180)
      Data_survival_181 = Data_reach_rate %>% dplyr::filter(time_enroll_final > 180)
      Data_survival_365 = Data_reach_rate %>% dplyr::filter(time_enroll_final <= 365)
      Data_survival_366 = Data_reach_rate %>% dplyr::filter(time_enroll_final > 365)
      
      survival_table_tmp$Deceased_patient_no = length(Data_reach_rate$Cancers)
      survival_table_tmp$All_treatment_reach_no = sum(Data_survival$treat_group_2 == "RecomTreat(+)")
      survival_table_tmp$Deceased_treatment_reach_no = sum(Data_reach_rate$Treatment_reach)
      survival_table_tmp$Treatment_reach_rate_all = round(survival_table_tmp$All_treatment_reach_no / survival_table_tmp$Patient_no, digits = 3)
      survival_table_tmp$Treatment_reach_rate_deceased = round(survival_table_tmp$Deceased_treatment_reach_no / survival_table_tmp$Deceased_patient_no, digits = 3)
      survival_table_tmp$Treatment_rate_crossing_date = threshold_x
      survival_table_tmp$Promising_survival_ratio = round(threshold_x / survival_table_tmp$Time_50percent, digits = 3)
      survival_table_tmp$Patient_no_deceased_1_180 = length(Data_survival_180$Cancers)
      survival_table_tmp$Patient_no_deceased_181_ = length(Data_survival_181$Cancers)
      survival_table_tmp$Treatment_reach_no_deceased_1_180 = sum(Data_survival_180$Treatment_reach)
      survival_table_tmp$Treatment_reach_no_deceased_181_ = sum(Data_survival_181$Treatment_reach)
      survival_table_tmp$Treatment_reach_rate_deceased_1_180 = round(survival_table_tmp$Treatment_reach_no_deceased_1_180 / survival_table_tmp$Patient_no_deceased_1_180, digits = 3)
      survival_table_tmp$Treatment_reach_rate_deceased_181_ = round(survival_table_tmp$Treatment_reach_no_deceased_181_ / survival_table_tmp$Patient_no_deceased_181_, digits = 3)
      
      mx=matrix(c(
        survival_table_tmp$Treatment_reach_no_deceased_181_,
        survival_table_tmp$Patient_no_deceased_181_,
        survival_table_tmp$Treatment_reach_no_deceased_1_180,
        survival_table_tmp$Patient_no_deceased_1_180), 
        nrow=2, byrow=T)
      survival_table_tmp$Fisher_odds_ratio_estimate_180_vs_181 = round(fisher.test(mx)$estimate[[1]], digits = 3)
      survival_table_tmp$Fisher_odds_ratio_95CI_lower_limit_180_vs_181 = round(fisher.test(mx)$conf.int[[1]], digits = 3)
      survival_table_tmp$Fisher_odds_ratio_95CI_upper_limit_180_vs_181 = round(fisher.test(mx)$conf.int[[2]], digits = 3)
      survival_table_tmp$Fisher_p_value_180_vs_181 = round(fisher.test(mx)$p.value, digits = 3)

      survival_table_tmp$Patient_no_deceased_1_365 = length(Data_survival_365$Cancers)
      survival_table_tmp$Patient_no_deceased_366_ = length(Data_survival_366$Cancers)
      survival_table_tmp$Treatment_reach_no_deceased_1_365 = sum(Data_survival_365$Treatment_reach)
      survival_table_tmp$Treatment_reach_no_deceased_366_ = sum(Data_survival_366$Treatment_reach)
      survival_table_tmp$Treatment_reach_rate_deceased_1_365 = round(survival_table_tmp$Treatment_reach_no_deceased_1_365 / survival_table_tmp$Patient_no_deceased_1_365, digits = 3)
      survival_table_tmp$Treatment_reach_rate_deceased_366_ = round(survival_table_tmp$Treatment_reach_no_deceased_366_ / survival_table_tmp$Patient_no_deceased_366_, digits = 3)
      
      mx=matrix(c(
        survival_table_tmp$Treatment_reach_no_deceased_366_,
        survival_table_tmp$Patient_no_deceased_366_,
        survival_table_tmp$Treatment_reach_no_deceased_1_365,
        survival_table_tmp$Patient_no_deceased_1_365), 
        nrow=2, byrow=T)
      survival_table_tmp$Fisher_odds_ratio_estimate_365_vs_366 = round(fisher.test(mx)$estimate[[1]], digits = 3)
      survival_table_tmp$Fisher_odds_ratio_95CI_lower_limit_365_vs_366 = round(fisher.test(mx)$conf.int[[1]], digits = 3)
      survival_table_tmp$Fisher_odds_ratio_95CI_upper_limit_365_vs_366 = round(fisher.test(mx)$conf.int[[2]], digits = 3)
      survival_table_tmp$Fisher_p_value_365_vs_366 = round(fisher.test(mx)$p.value, digits = 3)
      survival_table = rbind(survival_table, survival_table_tmp)
      write_csv(survival_table, file="source/treatment_reach_summary.csv")
      
      
    
      output$figure_survival_treatment_reach = renderPlot({
        grid.arrange(reachp,reachp_all, nrow = 2)
      })
      output$table_survival_treatment_reach = DT::renderDataTable(survival_table,
                                                   server = FALSE,
                                                   filter = 'top', 
                                                   extensions = c('Buttons'), 
                                                   options = list(pageLength = 100, 
                                                                  scrollX = TRUE,
                                                                  scrollY = "1000px",
                                                                  scrollCollapse = TRUE,
                                                                  dom="Blfrtip",
                                                                  buttons = c('csv', 'excel', 'copy')))
      incProgress(1 / 13)
    })
  })
  
  observeEvent(input$survival_CTx_analysis2, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      Data_case_target = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.基本情報.年齢,
                      Lymph_met,
                      Brain_met,
                      Lung_met,
                      Bone_met,
                      Liver_met,
                      Other_met,
                      EP_option,
                      EP_treat,
                      YoungOld,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      症例.EP前レジメン情報.最良総合効果.名称.,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.化学療法レジメン名称,
                      症例.EP前レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.EP後レジメン情報.治療ライン,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.化学療法レジメン名称,
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      症例.EP後レジメン情報.レジメン継続区分.名称.,
                      症例.EP後レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.最良総合効果.名称.,
                      症例.基本情報.性別.名称.,
                      症例.基本情報.がん種.OncoTree.,
                      症例.基本情報.がん種.OncoTree..名称.,
                      症例.基本情報.がん種.OncoTree.LEVEL1.,
                      症例.検体情報.パネル.名称.,
                      症例.背景情報.ECOG.PS.名称.,
                      症例.背景情報.喫煙歴有無.名称.,
                      症例.背景情報.アルコール多飲有無.名称.,
                      症例.背景情報.重複がん有無.異なる臓器..名称.,
                      症例.背景情報.多発がん有無.同一臓器..名称.,
                      症例.がん種情報.登録時転移部位.名称.,
                      症例.検体情報.腫瘍細胞含有割合,
                      症例.検体情報.検体採取部位.名称.,
                      症例.転帰情報.転帰.名称.,
                      症例.背景情報.診断日,
                      症例.管理情報.登録日,
                      症例.転帰情報.最終生存確認日,
                      症例.転帰情報.死亡日,
                      HER2_IHC,
                      MSI_PCR,
                      MMR_IHC
        )
      if(input$HER2 == "No"){
        Data_case_target = Data_case_target %>%
          dplyr::select(-HER2_IHC)
      } else {
        Data_case_target = Data_case_target %>%
          dplyr::filter(
            HER2_IHC != "Unknown"
          )
      }
      if(input$MSI == "No"){
        Data_case_target = Data_case_target %>%
          dplyr::select(-MSI_PCR)
      } else {
        Data_case_target = Data_case_target %>%
          dplyr::filter(
            MSI_PCR != "Unknown"
          )
      }
      if(input$MMR == "No"){
        Data_case_target = Data_case_target %>%
          dplyr::select(-MMR_IHC)
      } else {
        Data_case_target = Data_case_target %>%
          dplyr::filter(
            MMR_IHC != "Unknown"
          )
      }
      Data_case_target$Cancers = Data_case_target$症例.基本情報.がん種.OncoTree.
      Data_survival = Data_case_target %>%
        dplyr::mutate(final_observe = case_when(
          症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
          症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
          症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
          TRUE ~ 症例.管理情報.登録日))
      Data_survival = Data_survival %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::mutate(censor = case_when(
          症例.転帰情報.死亡日 != "" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
      
      Data_survival = Data_survival %>%
        dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
        dplyr::mutate(final_observe = ifelse(is.na(final_observe), 症例.管理情報.登録日, final_observe))
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, censor) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival = Data_survival %>%
        dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(Data_survival$症例.管理情報.登録日)) %>%
        dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                        as.Date(症例.EP前レジメン情報.投与開始日))
      incProgress(1 / 13)
      
      Data_survival = Data_survival %>%
        dplyr::mutate(time_enroll_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.管理情報.登録日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.EP前レジメン情報.投与開始日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_enroll =
                        Data_survival$time_palliative_final -
                        Data_survival$time_enroll_final)
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, final_observe) %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_enroll_final) %>%
        dplyr::arrange(desc(time_enroll_final)) %>%
        dplyr::filter(time_enroll_final > 0 & !is.na(time_enroll_final) & is.finite(time_enroll_final)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival_tmp = Data_survival %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                        c("その他", "緩和")) %>%
        dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_palliative_enroll, time_palliative_final) %>%
        dplyr::filter(time_palliative_enroll > 0 & time_palliative_enroll < 10000 & !is.na(time_palliative_enroll) & is.finite(time_palliative_enroll)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      tmp1 = Data_case_target %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                        c("その他", "緩和")) %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.化学療法レジメン名称,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      症例.管理情報.登録日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      final_observe,
                      censor,
                      症例.EP前レジメン情報.終了理由.名称.,
                      症例.EP前レジメン情報.最良総合効果.名称.) %>%
        dplyr::distinct()
      tmp1 = tmp1 %>%
        dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                        as.Date(tmp1$症例.EP前レジメン情報.投与開始日)) %>%
        dplyr::mutate(症例.EP前レジメン情報.投与終了日 =
                        as.Date(tmp1$症例.EP前レジメン情報.投与終了日)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(tmp1$症例.管理情報.登録日))
      tmp1 = tmp1 %>%
        dplyr::mutate(Drug_length =
                        as.integer(difftime(tmp1$症例.EP前レジメン情報.投与終了日,
                                            tmp1$症例.EP前レジメン情報.投与開始日,
                                            units = "days")))
      colnames(tmp1) = c("ID",
                         "Drug",
                         "レジメン",
                         "治療ライン",
                         "投与開始日",
                         "投与終了日",
                         "Ongoing",
                         "登録日",
                         "実施目的",
                         "final_observe",
                         "censor",
                         "終了理由",
                         "最良総合効果",
                         "Drug_length")
      tmp1$TxCGP = "Pre"
      tmp2 = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.化学療法レジメン名称,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      症例.EP後レジメン情報.レジメン継続区分.名称.,
                      症例.管理情報.登録日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      final_observe,
                      censor,
                      症例.EP後レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.最良総合効果.名称.) %>%
        dplyr::distinct()
      tmp2$症例.EP前レジメン情報.実施目的.名称. = "緩和"
      tmp2 = tmp2 %>%
        dplyr::mutate(症例.EP後レジメン情報.投与開始日 =
                        as.Date(tmp2$症例.EP後レジメン情報.投与開始日)) %>%
        dplyr::mutate(症例.EP後レジメン情報.投与終了日 =
                        as.Date(tmp2$症例.EP後レジメン情報.投与終了日)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(tmp2$症例.管理情報.登録日))
      
      tmp2 = tmp2 %>%
        dplyr::mutate(Drug_length =
                        as.integer(difftime(tmp2$症例.EP後レジメン情報.投与終了日,
                                            tmp2$症例.EP後レジメン情報.投与開始日,
                                            units = "days")))
      colnames(tmp2) = c("ID",
                         "Drug",
                         "レジメン",
                         "治療ライン",
                         "投与開始日",
                         "投与終了日",
                         "Ongoing",
                         "登録日",
                         "実施目的",
                         "final_observe",
                         "censor",
                         "終了理由",
                         "最良総合効果",
                         "Drug_length")
      tmp2$TxCGP = "Post"
      Data_drug = rbind(tmp1, tmp2)
      
      if(!is.null(input$ID_drug)){
        Data_drug = data.frame(NULL)
        for(i in 1:length(input$ID_drug[,1])){
          tmp = read.csv(header = TRUE, file(input$ID_drug[[i, 'datapath']],
                                             encoding='UTF-8-BOM'))
          if("症例.EP前レジメン情報.投与開始日" %in% colnames(tmp)){
            colnames(tmp) = c("ID",
                              "登録日",
                              "治療ライン",
                              "実施目的",
                              "レジメン",
                              "Drug",
                              "投与開始日",
                              "投与終了日",
                              "Ongoing",
                              "終了理由",
                              "最良総合効果",
                              "最終生存確認日",
                              "死亡日")
            tmp$投与開始日 = as.Date(tmp$投与開始日)
            tmp$投与終了日 = as.Date(tmp$投与終了日)
            tmp = tmp %>%
              dplyr::mutate(
                Drug_length = as.integer(difftime(tmp$投与終了日,
                                                  tmp$投与開始日,
                                                  units = "days")),
                censor = case_when(
                  死亡日 != "" ~ 1,
                  TRUE ~ 0
                ),
                final_observe = case_when(
                  最終生存確認日 == "           " ~ 死亡日,
                  最終生存確認日 != "" ~ 最終生存確認日,
                  死亡日 != "" ~ 死亡日,
                  TRUE ~ 登録日)
              )
            tmp$登録日 = as.Date(tmp$登録日)
            tmp$最終生存確認日 = as.Date(tmp$最終生存確認日)
            tmp$死亡日 = as.Date(tmp$死亡日)
            tmp$final_observe = as.Date(tmp$final_observe)
            tmp = tmp %>%
              dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
              dplyr::mutate(final_observe = ifelse(is.na(final_observe), 登録日, final_observe))
            tmp = tmp %>% dplyr::select(-最終生存確認日, -死亡日)
            tmp$TxCGP = "Pre"
          } else{
            colnames(tmp) = c("ID",
                              "登録日",
                              "治療ライン",
                              "レジメン",
                              "Drug",
                              "投与開始日",
                              "投与終了日",
                              "Ongoing",
                              "終了理由",
                              "最良総合効果",
                              "最終生存確認日",
                              "死亡日")
            tmp$投与開始日 = as.Date(tmp$投与開始日)
            tmp$投与終了日 = as.Date(tmp$投与終了日)
            tmp = tmp %>%
              dplyr::mutate(
                Drug_length = as.integer(difftime(tmp$投与終了日,
                                                  tmp$投与開始日,
                                                  units = "days")),
                censor = case_when(
                  死亡日 != "" ~ 1,
                  TRUE ~ 0
                ),
                final_observe = case_when(
                  最終生存確認日 == "           " ~ 死亡日,
                  最終生存確認日 != "" ~ 最終生存確認日,
                  死亡日 != "" ~ 死亡日,
                  TRUE ~ 登録日)
              )
            tmp$登録日 = as.Date(tmp$登録日)
            tmp$最終生存確認日 = as.Date(tmp$最終生存確認日)
            tmp$死亡日 = as.Date(tmp$死亡日)
            tmp$final_observe = as.Date(tmp$final_observe)
            tmp = tmp %>%
              dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
              dplyr::mutate(final_observe = ifelse(is.na(final_observe), 登録日, final_observe))
            tmp = tmp %>% dplyr::select(-最終生存確認日, -死亡日)
            tmp$実施目的 = "緩和"
            tmp$TxCGP = "Post"
          }
          Data_drug = rbind(Data_drug, tmp)
        }
      }
      
      Data_drug = Data_drug %>%
        dplyr::mutate(final_observe =
                        as.Date(Data_drug$final_observe))
      Data_drug = Data_drug %>%
        dplyr::arrange(投与開始日) %>%
        dplyr::arrange(治療ライン) %>%
        dplyr::arrange(ID)
      
      Data_drug = Data_drug %>% dplyr::mutate(RECIST = case_when(
        最良総合効果 == "CR" ~ "5",
        最良総合効果 == "PR" ~ "4",
        最良総合効果 == "SD" ~ "3",
        最良総合効果 == "PD" ~ "2",
        最良総合効果 == "NE" ~ "1",
        最良総合効果 == "Unknown" ~ "0",
        TRUE ~ "0"
      ))
      
      Data_drug = Data_drug %>%
        dplyr::filter(治療ライン %in% c("１次治療","２次治療","３次治療","４次治療"))
      
      Data_drug_RECIST = (Data_drug %>%
                            dplyr::arrange(desc(RECIST),ID,desc(レジメン),治療ライン) %>%
                            dplyr::distinct(ID,治療ライン,.keep_all = T)) %>%
        dplyr::select(ID,治療ライン,RECIST)
      
      Data_drug = Data_drug %>% dplyr::select(-投与開始日, -投与終了日, -RECIST, -Ongoing, -Drug)
      Data_drug = left_join(Data_drug, Data_drug_RECIST, by = c('ID','治療ライン'))
      Data_drug = Data_drug %>%
        dplyr::arrange(ID) %>%
        dplyr::distinct(ID,治療ライン, .keep_all = T)
      
      Data_drug <- data.frame(Data_drug %>% 
                                dplyr::arrange(ID) %>%
                                dplyr::group_by(ID) %>%
                                dplyr::mutate(治療ライン = row_number()))
      Data_drug$治療ライン = as.character(Data_drug$治療ライン)
      
      ID_RECIST = Data_drug %>%
        dplyr::select(ID, RECIST) %>%
        dplyr::arrange(desc(RECIST)) %>%
        dplyr::distinct()
      colnames(ID_RECIST) = c("ID", "RECIST")
      
      ID_RECIST = ID_RECIST %>% dplyr::mutate(RECIST = case_when(
        RECIST == "5" ~ "CR",
        RECIST == "4" ~ "PR",
        RECIST == "3" ~ "SD",
        RECIST == "2" ~ "PD",
        RECIST == "1" ~ "NE",
        RECIST == "0" ~ "NE",
        TRUE ~ "NE"
      ))
      Data_case_target$RECIST = unlist(lapply(list(Data_case_target$C.CAT調査結果.基本項目.ハッシュID), function(x) {
        as.vector(ID_RECIST$RECIST[match(x, ID_RECIST$ID)])}))
      if(sum(is.na(Data_case_target$RECIST)>0)){
        Data_case_target[is.na(Data_case_target$RECIST),]$RECIST = "CTx_naive"
      }
      
      Data_case_target = Data_case_target %>%
        dplyr::distinct(.keep_all = TRUE, C.CAT調査結果.基本項目.ハッシュID)
      Data_MAF = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","A","B","C","D","E","F") &
            Variant_Classification != "expression"
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
      if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
        Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
      }
      
      Data_report_TMB = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
        dplyr::select(Tumor_Sample_Barcode, TMB) %>%
        dplyr::distinct(Tumor_Sample_Barcode, .keep_all=T)
      colnames(Data_report_TMB) = c("C.CAT調査結果.基本項目.ハッシュID", "TMB")
      Data_case_target = left_join(Data_case_target, Data_report_TMB,
                                   by = "C.CAT調査結果.基本項目.ハッシュID")
      
      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      if(input$patho == "Only pathogenic muts"){
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(Evidence_level == "F")
      }
      Data_MAF_target = Data_MAF_target %>%
        dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol, .keep_all = T)
      Gene_list = unique(names(sort(table(Data_MAF_target$Hugo_Symbol),
                                    decreasing = T)))
      if(length(Data_case_target$censor[is.na(Data_case_target$censor)])> 0){
        Data_case_target$censor[is.na(Data_case_target$censor)] = 0
      }
      incProgress(1 / 13)
      
      Data_case_target = Data_case_target %>%
        dplyr::mutate(EP_option = case_when(
          症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. ==
            "はい" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::mutate(EP_treat = case_when(
          症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
            "投与した" ~ 1,
          TRUE ~ 0))
      
      Data_tmp = Data_case_target %>% dplyr::select(
        C.CAT調査結果.基本項目.ハッシュID,
        症例.EP後レジメン情報.治療ライン,
        症例.EP後レジメン情報.化学療法レジメン名称,
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
        症例.EP後レジメン情報.提示された治療薬を投与した.名称.
      ) %>%
        dplyr::distinct()
      Data_tmp = tidyr::replace_na(Data_tmp,
                                   replace = list(症例.EP後レジメン情報.治療ライン = 0,
                                                  症例.EP後レジメン情報.化学療法レジメン名称 = "",
                                                  症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = ""))
      Data_tmp = Data_tmp %>%
        dplyr::filter(症例.EP後レジメン情報.治療ライン != 0 &(
          症例.EP後レジメン情報.化学療法レジメン名称 != "" |
            症例.EP後レジメン情報.薬剤名.YJ一般名.EN. != "" |
            症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した"))
      ID_tmp = unique(Data_tmp$C.CAT調査結果.基本項目.ハッシュID)
      Data_case_target = Data_case_target %>% dplyr::filter(
        !is.na(time_enroll_final) &
          is.finite(time_enroll_final) & 
          time_enroll_final > 0 &
          !is.na(time_palliative_enroll) &
          is.finite(time_palliative_enroll) & 
          time_palliative_enroll > 0 &
          !is.na(censor) &
          is.finite(censor)
      )  
      Data_survival = Data_case_target
      Data_survival$afterCGPtreat = Data_survival$EP_treat
      Data_survival$afterCGPtreat[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp] <- 1
      Data_survival = Data_survival %>%
        dplyr::mutate(treat_group = case_when(
          afterCGPtreat == 1 & EP_treat == 1 ~ "Option(+)RecomTreat(+)",
          afterCGPtreat == 1 & EP_option == 1 ~ "Option(+)UnrecomTreat(+)",
          afterCGPtreat == 1 & EP_option == 0 ~ "Option(-)UnrecomTreat(+)",
          afterCGPtreat == 0 & EP_option == 1 ~ "Option(+)Treat(-)",
          afterCGPtreat == 0 & EP_option == 0 ~ "Option(-)Treat(-)"
        ))
      Data_survival = Data_survival %>%
        dplyr::mutate(treat_group_2 = case_when(
          afterCGPtreat == 1 & EP_treat == 1 ~ "RecomTreat(+)",
          afterCGPtreat == 1 & EP_option == 1 ~ "UnrecomTreat(+)",
          afterCGPtreat == 1 & EP_option == 0 ~ "UnrecomTreat(+)",
          afterCGPtreat == 0 & EP_option == 1 ~ "Treat(-)",
          afterCGPtreat == 0 & EP_option == 0 ~ "Treat(-)"
        ))
      Data_survival = Data_survival %>%
        dplyr::mutate(treat_group_3 = case_when(
          afterCGPtreat == 1 & EP_treat == 1 ~ "Treat(+)",
          afterCGPtreat == 1 & EP_option == 1 ~ "Treat(+)",
          afterCGPtreat == 1 & EP_option == 0 ~ "Treat(+)",
          afterCGPtreat == 0 & EP_option == 1 ~ "Treat(-)",
          afterCGPtreat == 0 & EP_option == 0 ~ "Treat(-)"
        ))
      Data_survival = Data_survival %>%
        dplyr::filter(time_enroll_final > 0 & !is.na(time_enroll_final) & is.finite(time_enroll_final))
      Cancername = sort(unique(Data_survival$Cancers))
      Total_pts = as.vector(table((Data_survival %>%
                                     dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = T))$Cancers))
      Data_MAF_target = Data_MAF_target %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_survival$C.CAT調査結果.基本項目.ハッシュID)
      incProgress(1 / 13)
      
      if(file.exists("source/treatment_reach_summary_CTx.csv")){
        survival_table = read_csv("source/treatment_reach_summary_CTx.csv",show_col_types=FALSE)
      } else{
        survival_table = data.frame(Organ = as.character(),
                                    Histology = as.character(),
                                    Patient_no = as.numeric(),
                                    Deceased_patient_no = as.numeric(),
                                    Day_180_survival = as.numeric(),
                                    Year_1_survival = as.numeric(),
                                    Year_2_survival = as.numeric(),
                                    Year_3_survival = as.numeric(),
                                    Year_4_survival = as.numeric(),
                                    Year_5_survival = as.numeric(),
                                    Time_50percent = as.numeric(),
                                    Time_25percent = as.numeric(),
                                    Year_1_reach_rate = as.numeric(),
                                    Year_2_reach_rate = as.numeric(),
                                    Year_3_reach_rate = as.numeric(),
                                    Year_4_reach_rate = as.numeric(),
                                    Year_5_reach_rate = as.numeric(),
                                    All_treatment_reach_no = as.numeric(),
                                    Deceased_treatment_reach_no = as.numeric(),
                                    Treatment_rate_crossing_date = as.numeric(),
                                    Promising_survival_ratio = as.numeric()
        )
      }
      survival_table_tmp = data.frame(paste0(unique(Data_survival$症例.基本情報.がん種.OncoTree.LEVEL1.), collapse = ";"))
      colnames(survival_table_tmp) = "Organ"
      survival_table_tmp$Histology = paste0(unique(Data_survival$Cancers), collapse = ";")
      survival_table_tmp$Patient_no = length(Data_survival$Cancers)
      survival_table_data = survfit(Surv(time = time_palliative_enroll,
                                         time2 = time_palliative_final,
                                         censor)~1,
                                    data=Data_survival,conf.type = "log-log")
      if(max(survival_table_data$time) >= 180){
        survival_table_tmp$Day_180_survival = round(summary(survival_table_data, times = 180)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_1_survival = NA
      }
      if(max(survival_table_data$time) >= 365.25*1){
        survival_table_tmp$Year_1_survival = round(summary(survival_table_data, times = 365.25*1)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_1_survival = NA
      }
      if(max(survival_table_data$time) >= 365.25*2){
        survival_table_tmp$Year_2_survival = round(summary(survival_table_data, times = 365.25*2)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_2_survival = NA
      }
      if(max(survival_table_data$time) >= 365.25*3){
        survival_table_tmp$Year_3_survival = round(summary(survival_table_data, times = 365.25*3)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_3_survival = NA
      }
      if(max(survival_table_data$time) >= 365.25*4){
        survival_table_tmp$Year_4_survival = round(summary(survival_table_data, times = 365.25*4)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_4_survival = NA
      }
      if(max(survival_table_data$time) >= 365.25*5){
        survival_table_tmp$Year_5_survival = round(summary(survival_table_data, times = 365.25*5)$surv, digits = 3)
      } else{
        survival_table_tmp$Year_5_survival = NA
      }
      survival_table_tmp$Time_50percent = quantile(survival_table_data, 0.5)$quantile[[1]]
      survival_table_tmp$Time_25percent = quantile(survival_table_data, 0.75)$quantile[[1]]


      g = list()
      # Survival analysis for all patients
      survival_simple_all = survfit(Surv(time = time_palliative_enroll,
                                         time2 = time_palliative_final,
                                         censor)~1,
                                    data=Data_survival,conf.type = "log-log")
      independence_check = with(Data_survival %>% sample_n(min(length(Data_survival$C.CAT調査結果.基本項目.ハッシュID), 40000)), cKendall(
        trun = time_palliative_enroll,
        obs = time_palliative_final,
        delta = censor,
        method = "MB",
        trans = "linear"))

      g[[1]] = surv_curv_CTx(survival_simple_all, Data_survival,
                             paste0(survival_simple_all[[1]], " patients, cKendall tau= ",
                                    format(independence_check$PE,digits=2),
                                    ", SE= ", format(independence_check$SE,digits=2),
                                    ", p= ", format(independence_check$p.value,digits=2)),
                             "All enrolled patients, risk-set adjusted", NULL)
      kk=2
      # Survival analysis for EP recommendation
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                         time2 = time_palliative_final,
                                         censor)~EP_option,
                                    data=Data_survival,conf.type = "log-log")
      if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
        diff_0 = coxph(Surv(time = time_palliative_enroll,
                               time2 = time_palliative_final,
                               censor)~EP_option, data=Data_survival)
        g[[kk]] = surv_curv_CTx(survival_simple, Data_survival, paste(paste(survival_simple[[1]],collapse = "/"),"patients,","EP option existed for", 
                                                                        (survival_simple)[[1]][[2]], "patients"), paste("Expert panel recommended option existed:",
                                                                                                                        c("No", "Yes"), sep=""), diff_0)
        kk = kk + 1
      }
      
      # Survival analysis for EP recommended treatment
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final,
                                     censor)~EP_treat,
                                data=Data_survival,conf.type = "log-log")
      if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
        diff_0 = coxph(Surv(time = time_palliative_enroll,
                            time2 = time_palliative_final,
                            censor)~EP_treat, data=Data_survival)
        g[[kk]] = surv_curv_CTx(survival_simple, Data_survival, paste(paste(survival_simple[[1]],collapse = "/"),"patients,","Recommended treatment done for", 
                                                                        (survival_simple)[[1]][[2]], "patients"),
                                  paste("Recommended treatment done: ",
                                        c("No", "Yes"), sep=""), diff_0)
        kk = kk + 1
      }
      
      # Survival analysis for EP recommended treatment only patients with treatment option
      Data_survival_option = Data_survival %>%
        dplyr::filter(EP_option == 1)
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~EP_treat,
                                data=Data_survival_option)
      if(length(Data_survival_option[,1][[1]]) != survival_simple[[1]][[1]]){
        diff_0 = coxph(Surv(time = time_palliative_enroll,
                               time2 = time_palliative_final, censor)~EP_treat,
                          data=Data_survival_option)
        g[[kk]] = surv_curv_CTx(survival_simple, Data_survival_option,  paste(paste(survival_simple[[1]],collapse = "/"),"patients,","Recommended treatment done for", 
                                                                                (survival_simple)[[1]][[2]], "patients with treatment option"), c("EP option exists but recommended treatment not done",
                                                                                                                                                  "EP option exists and recommended treatment done"), diff_0)
        kk = kk + 1
      }
      
      # Survival analysis for EP recommended treatment
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~treat_group,
                                data=Data_survival)
      if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
        diff_0 = coxph(Surv(time = time_palliative_enroll,
                               time2 = time_palliative_final, censor)~treat_group,
                          data=Data_survival)
        if(length(unique(Data_survival$treat_group))==5){
          g[[kk]] = surv_curv_CTx(survival_simple, Data_survival,  paste(paste(survival_simple[[1]],collapse = "/"),"patients,","EP option and treatment"),
                                    c("Targetable mutation (+) and recommended treatment done",
                                      "Targetable mutation (+) and other treatment done",
                                      "Targetable mutation (-) and treatment done",
                                      "Targetable mutation (+) and no treatment done",
                                      "Targetable mutation (-) and no treatment done"),
                                    diff_0)
        } else {
          g[[kk]] = surv_curv_CTx(survival_simple, Data_survival,  paste(paste(survival_simple[[1]],collapse = "/"),"patients,","EP option and treatment"),
                                    NULL,
                                    diff_0)
        }
        kk = kk + 1
      }
      
      # Survival analysis for EP recommended treatment
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~treat_group_2,
                                data=Data_survival)
      if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
        diff_0 = coxph(Surv(time = time_palliative_enroll,
                               time2 = time_palliative_final, censor)~treat_group_2,
                          data=Data_survival)
        if(length(unique(Data_survival$treat_group))==3){
          g[[kk]] = surv_curv_CTx(survival_simple, Data_survival, paste(paste(survival_simple[[1]],collapse = "/"),"patients,","EP option and treatment"),
                                    c("Recommended treatment done",
                                      "No treatment done",
                                      "Not-recommended treatment done"),
                                    diff_0)
        } else {
          g[[kk]] = surv_curv_CTx(survival_simple, Data_survival, paste(paste(survival_simple[[1]],collapse = "/"),"patients,","EP option and treatment"),
                                    NULL,
                                    diff_0)
          
        }
        kk = kk + 1
      }

      Data_survival_treated = Data_survival %>%
        dplyr::filter(RECIST != "NE")
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~RECIST,
                                data=Data_survival_treated)
      if(length(Data_survival_treated[,1][[1]]) != survival_simple[[1]][[1]]){
        diff_0 = coxph(Surv(time = time_palliative_enroll,
                               time2 = time_palliative_final, censor)~RECIST,
                          data=Data_survival_treated)
        g[[kk]] = surv_curv_CTx(survival_simple, Data_survival_treated, paste(paste(survival_simple[[1]],collapse = "/"),"patients,","Best CTx effect of all CTx"),
                                  NULL, diff_0)
        kk = kk + 1
      }
      # Survival analysis for EP recommended treatment
      Data_survival_treated = Data_survival %>%
        dplyr::filter(treat_group_2 != "Treat(-)")
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~treat_group_2,
                                data=Data_survival_treated)
      if(length(Data_survival_treated[,1][[1]]) != survival_simple[[1]][[1]]){
        diff_0 = coxph(Surv(time = time_palliative_enroll,
                               time2 = time_palliative_final, censor)~treat_group_2,
                          data=Data_survival_treated)
        g[[kk]] = surv_curv_CTx(survival_simple, Data_survival_treated, paste(paste(survival_simple[[1]],collapse = "/"),"patients,","EP option and treatment"),
                                  c("Recommended treatment done",
                                    "Not-recommended treatment done"),
                                  diff_0)
        kk = kk + 1
      }
      
      # Survival analysis for EP recommended treatment
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~treat_group_3,
                                data=Data_survival)
      if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
        diff_0 = coxph(Surv(time = time_palliative_enroll,
                               time2 = time_palliative_final, censor)~treat_group_3,
                          data=Data_survival)
        #summary(survival_simple)
        g[[kk]] = surv_curv_CTx(survival_simple, Data_survival, paste(paste(survival_simple[[1]],collapse = "/"),"patients,","EP option and treatment"),
                                  c("No treatment done",
                                    "Treatment done"),
                                  diff_0)
        kk = kk + 1
      }
      # Survival analysis for diagnosis
      Data_survival_tmp = Data_survival %>% dplyr::filter(
        Cancers %in% sort(names(table(Data_survival$Cancers)))[1:min(7,length(unique(Data_survival$Cancers)))])
      Data_survival_tmp2 = Data_survival
      Data_survival_tmp2$Cancers = " ALL"
      Data_survival_tmp2 = rbind(Data_survival_tmp, Data_survival_tmp2)
      if(length(unique(Data_survival_tmp$Cancers))>1){
        survival_simple = survfit(Surv(time = time_palliative_enroll,
                                       time2 = time_palliative_final, censor)~Cancers,
                                  data=Data_survival_tmp)
        if(length(Data_survival_tmp[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = coxph(Surv(time = time_palliative_enroll,
                                 time2 = time_palliative_final, censor)~Cancers,
                            data=Data_survival_tmp)
          g[[kk]] = surv_curv_CTx(survival_simple, Data_survival_tmp, paste(paste(survival_simple[[1]],collapse = "/"),"patients,","diagnosis"), sort(unique(Data_survival_tmp$Cancers)), diff_0)
          kk = kk + 1
        }
        survival_simple = survfit(Surv(time_enroll_final, censor)~Cancers,
                                  data=Data_survival_tmp2)
        if(length(Data_survival_tmp2[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = coxph(Surv(time = time_palliative_enroll,
                                 time2 = time_palliative_final, censor)~Cancers,
                            data=Data_survival_tmp2)
          g[[kk]] = surv_curv_CTx(survival_simple, Data_survival_tmp2, paste(paste(survival_simple[[1]],collapse = "/"),"patients,","diagnosis"), sort(unique(Data_survival_tmp2$Cancers)), diff_0)
          kk = kk + 1
        }
        
      } else{
        survival_simple = survfit(Surv(time = time_palliative_enroll,
                                       time2 = time_palliative_final, censor)~1,
                                  data=Data_survival_tmp)
        g[[kk]] = surv_curv_CTx(survival_simple, Data_survival_PS, paste(paste(survival_simple[[1]],collapse = "/"),"patients,","Only 1 diagnosis"), unique(Data_survival_tmp$Cancers)[[1]], diff_0)
        kk = kk + 1
      }
      
      # Survival analysis for mutation, any
      mut_gene = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
      if(length(mut_gene)==0){
        mut_gene = "TP53"
      }
      ID_mutation = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene))$Tumor_Sample_Barcode
      Data_survival = Data_survival %>% dplyr::mutate(
        mutation = case_when(
          C.CAT調査結果.基本項目.ハッシュID %in% ID_mutation ~ 0,
          TRUE ~ 1
        )
      )
      survival_simple = survfit(Surv(time = time_palliative_enroll,
                                     time2 = time_palliative_final, censor)~mutation,
                                data=Data_survival)
      if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
        diff_0 = coxph(Surv(time = time_palliative_enroll,
                               time2 = time_palliative_final, censor)~mutation,
                          data=Data_survival)
        g[[kk]] = surv_curv_CTx(survival_simple, Data_survival, paste(paste(survival_simple[[1]],collapse = "/"),"patients,",paste(mut_gene, collapse = "/")),
                                  c(paste0(paste(mut_gene, collapse = "/"), " mut in any"), "Not mut"), diff_0)
        kk = kk + 1
      }
      # Survival analysis for mutation, all
      mut_gene = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
      if(length(mut_gene)==0){
        mut_gene = "TP53"
      }
      ID_mutation = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene))$Tumor_Sample_Barcode
      for(jj in 1:length(mut_gene)){
        ID_mutation = ID_mutation[ID_mutation %in% (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene[jj]))$Tumor_Sample_Barcode]
      }
      if(length(ID_mutation) > 0){
        Data_survival = Data_survival %>% dplyr::mutate(
          mutation = case_when(
            C.CAT調査結果.基本項目.ハッシュID %in% ID_mutation ~ 1,
            C.CAT調査結果.基本項目.ハッシュID %in% (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene))$Tumor_Sample_Barcode ~ 2,
            TRUE ~ 4
          )
        )
        survival_simple = survfit(Surv(time = time_palliative_enroll,
                                       time2 = time_palliative_final, censor)~mutation,
                                  data=Data_survival)
        if(length(Data_survival[,1][[1]]) != survival_simple[[1]][[1]]){
          if(sum(unique(Data_survival$mutation)) == 7){
            diff_0 = coxph(Surv(time = time_palliative_enroll,
                                   time2 = time_palliative_final, censor)~mutation,
                              data=Data_survival)
            g[[kk]] = surv_curv_CTx(survival_simple, Data_survival, paste(paste(survival_simple[[1]],collapse = "/"),"patients,",paste(mut_gene, collapse = "/")),
                                      c(paste0(paste(mut_gene, collapse = "/"), " mut in all"),
                                        paste0(paste(mut_gene, collapse = "/"), " mut in some"),
                                        "Not mut"), diff_0)
            kk = kk + 1
          } else if(sum(unique(Data_survival$mutation)) == 6){
            diff_0 = coxph(Surv(time = time_palliative_enroll,
                                   time2 = time_palliative_final, censor)~mutation,
                              data=Data_survival)
            g[[kk]] = surv_curv_CTx(survival_simple, Data_survival, paste(paste(survival_simple[[1]],collapse = "/"),"patients,",paste(mut_gene, collapse = "/")),
                                      c(paste0(paste(mut_gene, collapse = "/"), " mut in some"), "Not mut"), diff_0)
            kk = kk + 1
          } else if(sum(unique(Data_survival$mutation)) == 5){
            diff_0 = coxph(Surv(time = time_palliative_enroll,
                                   time2 = time_palliative_final, censor)~mutation,
                              data=Data_survival)
            g[[kk]] = surv_curv_CTx(survival_simple, Data_survival, paste(paste(survival_simple[[1]],collapse = "/"),"patients,",paste(mut_gene, collapse = "/")),
                                      c(paste0(paste(mut_gene, collapse = "/"), " mut in all"), "Not mut"), diff_0)
            kk = kk + 1
          } else {
            diff_0 = coxph(Surv(time = time_palliative_enroll,
                                   time2 = time_palliative_final, censor)~mutation,
                              data=Data_survival)
            g[[kk]] = surv_curv_CTx(survival_simple, Data_survival, paste(paste(survival_simple[[1]],collapse = "/"),"patients,",paste(mut_gene, collapse = "/")),
                                      c(paste0(paste(mut_gene, collapse = "/"), " mut in all"), paste0(paste(mut_gene, collapse = "/"), " mut in some")), diff_0)
            kk = kk + 1
          }
        }
      }
      
      if(kk != 17){
        for(kkk in kk:16){
          g[[kkk]] = ggsurvplot_empty()
        }
      }
      output$figure_survival_CTx_1_2 = renderPlot({
        arrange_ggsurvplots(g,print=TRUE,ncol=4,nrow=4)
      })
      incProgress(1 / 13)
      # analysis for common oncogenic mutations
      h = list()
      hs = list()
      oncogenic_genes = Data_MAF_target %>%
        dplyr::select(Tumor_Sample_Barcode, Hugo_Symbol) %>%
        dplyr::distinct() %>%
        dplyr::count(Hugo_Symbol) %>%
        dplyr::arrange(-n)
      colnames(oncogenic_genes) = c("gene_mutation", "all_patients")
      oncogenic_genes = rbind(oncogenic_genes %>% dplyr::filter(gene_mutation %in% input$gene),
                              oncogenic_genes %>% dplyr::filter(!gene_mutation %in% input$gene))
      oncogenic_genes = oncogenic_genes[1:min(length(oncogenic_genes$all_patients), input$gene_no),]
      
      gene_table = data.frame(oncogenic_genes$gene_mutation)
      colnames(gene_table) = c("Gene")
      gene_table$positive_patients = oncogenic_genes$all_patients
      gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)) / 10
      gene_table$positive_median = 0
      gene_table$positive_LL = 0
      gene_table$positive_UL = 0
      gene_table$negative_median = 0
      gene_table$negative_LL = 0
      gene_table$negative_UL = 0
      gene_table$diff_median = 0
      gene_table$diff_LL = 0
      gene_table$diff_UL = 0
      
      k = 1
      for(i in 1:length(oncogenic_genes$all_patients)){
        ID_mutation = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% oncogenic_genes$gene_mutation[i]))$Tumor_Sample_Barcode
        Data_survival = Data_survival %>% dplyr::mutate(
          gene_mut = case_when(
            C.CAT調査結果.基本項目.ハッシュID %in% ID_mutation ~ 0,
            TRUE ~ 1
          )
        )
        traditional_fit = survfit(Surv(time = time_palliative_enroll,
                                       time2 = time_palliative_final, censor) ~ gene_mut, 
                                  data = Data_survival)
        if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
          diff_0 = coxph(Surv(time = time_palliative_enroll,
                              time2 = time_palliative_final, censor)~gene_mut,
                            data=Data_survival)
          mut_gene = paste(oncogenic_genes$gene_mutation[i],
                           ", top ", i, " gene", sep="")
          tmp = data.frame(summary(traditional_fit)$table)
          gene_table$positive_median[i] = round(tmp$median[1] / 365.25 * 12, digits = 1)
          gene_table$positive_LL[i] = round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1)
          gene_table$positive_UL[i] = round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1)
          gene_table$negative_median[i] = round(tmp$median[2] / 365.25 * 12, digits = 1)
          gene_table$negative_LL[i] = round(tmp$X0.95LCL[2] / 365.25 * 12, digits = 1)
          gene_table$negative_UL[i] = round(tmp$X0.95UCL[2] / 365.25 * 12, digits = 1)
          data_tmp = tidy(diff_0, exponentiate=TRUE, conf.int=TRUE)
          gene_table$diff_median[i] = round(1/data_tmp[[2]], digits = 3)
          gene_table$diff_LL[i] = round(1/data_tmp[[7]], digits = 3)
          gene_table$diff_UL[i] = round(1/data_tmp[[6]], digits = 3)
               
          hs[[k]] = surv_curv_CTx(traditional_fit, Data_survival,
                                  paste0(oncogenic_genes$gene_mutation[i], " mutation and overall survival"),
                                    c(paste0(oncogenic_genes$gene_mutation[i], " mut"), "Not mut"), diff_0)
          k = k + 1
        } else{
          tmp = summary(traditional_fit)$table
          legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (", round(tmp[[8]] / 365.25 * 12, digits = 1), "-", round(tmp[[9]] / 365.25 * 12, digits = 1),")")
          gene_table$negative_median[i] = round(tmp[[7]] / 365.25 * 12, digits = 1)
          gene_table$negative_LL[i] = round(tmp[[8]] / 365.25 * 12, digits = 1)
          gene_table$negative_UL[i] = round(tmp[[9]] / 365.25 * 12, digits = 1)
        }
      }
      incProgress(1 / 13)
      Gene_arrange = gene_table$Gene
      Gene_data = gene_table
      gene_table = gene_table %>% dplyr::mutate(
        Gene = factor(gene_table$Gene, levels = Gene_arrange))
      p <- 
        gene_table |>
        ggplot(aes(y = fct_rev(Gene))) + 
        theme_classic() +  scale_x_log10()
      p <- p +
        geom_point(aes(x=diff_median), shape=15, size=3) +
        geom_linerange(aes(xmin=diff_LL, xmax=diff_UL))
      p <- p +
        geom_vline(xintercept = 1, linetype="dashed") +
        labs(x="Hazard ratio", y="Genes with oncogenic alteration")
      p <- p +
        coord_cartesian(ylim=c(1,length(Gene_arrange) + 1),
                        xlim=c( min(gene_table$diff_LL) ^ 1.1,
                                max(gene_table$diff_UL) ^ 1.1))
      p <- p +
        annotate("text",
                 x = sqrt(max(gene_table$diff_UL) ^ 1.1),
                 y = length(Gene_arrange) + 1,
                 label = "mut=short survival") +
        annotate("text",
                 x = sqrt(min(gene_table$diff_LL) ^ 1.1),
                 y = length(Gene_arrange) + 1,
                 label = "mut=long survival")
      p_mid <- p + 
        theme(legend.position = "none",
              axis.line.y = element_blank(),
              axis.ticks.y= element_blank(),
              axis.text.y= element_blank(),
              axis.title.y= element_blank())
      gene_table <- gene_table %>%
        dplyr::mutate(
          estimate_lab_1 = paste0(positive_median, " (", positive_LL, "-", positive_UL, ")")) %>%
        dplyr::mutate(
          estimate_lab_2 = paste0(negative_median, " (", negative_LL, "-", negative_UL, ")")) %>%
        dplyr::mutate(
          estimate_lab_3 = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
        dplyr::mutate(
          patients = paste0(positive_patients, " (", positive_freq, "%)"))
      gene_table = 
        dplyr::bind_rows(
          data.frame(
            Gene = "Gene",
            estimate_lab_1 = "Survival, mut (+)",
            estimate_lab_2 = "Survival, mut (-)",
            estimate_lab_3 = "Hazard ratio",
            patients = "Patients"
          ), gene_table)
      Gene_arrange = gene_table$Gene
      gene_table = gene_table %>% dplyr::mutate(
        Gene = factor(gene_table$Gene, levels = Gene_arrange))
      
      p_left <- gene_table  %>% ggplot(aes(y = fct_rev(Gene)))
      p_left <- p_left + geom_text(aes(x = 0, label = Gene), hjust = 0,
                                   fontface = ifelse(gene_table$Gene == "Gene", "bold", "bold.italic"))
      p_left <- p_left + geom_text(aes(x = 1.8, label = estimate_lab_1), hjust = 0,
                                   fontface = ifelse(gene_table$estimate_lab_1 == "Survival, mut (+)", "bold", "plain"))
      p_left <- p_left + geom_text(aes(x = 3.3, label = estimate_lab_2), hjust = 0,
                                   fontface = ifelse(gene_table$estimate_lab_2 == "Survival, mut (-)", "bold", "plain"))
      p_left <- p_left + geom_text(aes(x = 4.8, label = estimate_lab_3), hjust = 0,
                                   fontface = ifelse(gene_table$estimate_lab_3 == "Hazard ratio", "bold", "plain"))
      p_left <- p_left + theme_void() + coord_cartesian(xlim = c(0, 7.5))
      
      # right side of plot - pvalues
      p_right <- gene_table  |> ggplot() +
        geom_text(aes(x = 0, y = fct_rev(Gene), label = patients), hjust = 0,
                  fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")) +
        theme_void()
      
      for(kk in k:32){
        hs[[kk]] = ggsurvplot_empty()
      }
      
      # final plot arrangement
      layout <- c(patchwork::area(t = 0, l = 0, b = 30, r = 7.5),
                  patchwork::area(t = 1, l = 7.5, b = 30, r = 12.5),
                  patchwork::area(t = 0, l = 13.5, b = 30, r = 14))
      h[[1]] = p_left + p_mid + p_right + plot_layout(design = layout)
      
      incProgress(1 / 13)
      
      output$figure_survival_CTx_2_2 = renderPlot({
        h[[1]]
      })
      gene_table_gene = gene_table
      output$figure_survival_CTx_2_data_2 = DT::renderDataTable(gene_table_gene,
                                                              server = FALSE,
                                                              filter = 'top', 
                                                              extensions = c('Buttons'), 
                                                              options = list(pageLength = 100, 
                                                                             scrollX = TRUE,
                                                                             scrollY = "1000px",
                                                                             scrollCollapse = TRUE,
                                                                             dom="Blfrtip",
                                                                             buttons = c('csv', 'excel', 'copy')))
      output$figure_survival_CTx_3_2 = renderPlot({
        arrange_ggsurvplots(hs,print=TRUE,ncol=4,nrow=8,surv.plot.height = 0.8,risk.table.height = 0.2)
      })
      incProgress(1 / 13)
      
      Data_cluster_ID_list = Data_cluster_ID() %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, cluster)
      Data_survival = left_join(Data_survival,
                                   Data_cluster_ID_list,
                                   by = "C.CAT調査結果.基本項目.ハッシュID")
      Data_survival = Data_survival %>%
        dplyr::mutate(Cancer_name_event = case_when(
          censor == 1 ~ Cancers,
          TRUE ~ "Others"
        ))
      Cancer_table = sort(table(Data_survival$Cancer_name_event),decreasing = T)
      Cancer_names = names(Cancer_table[Cancer_table>=3])
      Cancer_names = Cancer_names[Cancer_names != "Others"]
      Data_survival = Data_survival %>%
        dplyr::mutate(Cancer_name = case_when(
          Cancers %in% Cancer_names ~ Cancers,
          TRUE ~ "Others"
        ))
      Cancer_names = c(Cancer_names, "Others")
      Data_survival$Cancer_name = as.factor(Data_survival$Cancer_name)
      Data_survival$Cancer_name = relevel(Data_survival$Cancer_name,
              ref=Cancer_names[[1]])
      Data_survival$cluster = as.character(Data_survival$cluster)
      Data_survival = Data_survival %>%
        dplyr::mutate(cluster_name_event = case_when(
          censor == 1 ~ cluster,
          TRUE ~ "Others"
        ))
      cluster_table = sort(table(Data_survival$cluster_name_event),decreasing = T)
      cluster_names = names(cluster_table[cluster_table>=3])
      cluster_names = cluster_names[cluster_names != "Others"]
      Data_survival = Data_survival %>%
        dplyr::mutate(cluster = case_when(
          cluster %in% cluster_names ~ cluster,
          TRUE ~ "Others"
        ))
      cluster_names = c(cluster_names, "Others")
      Data_survival$cluster = as.factor(Data_survival$cluster)
      Data_survival$cluster = relevel(Data_survival$cluster,
                                          ref=cluster_names[[1]])
      
      Data_survival_cancer = Data_survival %>% dplyr::filter(Cancer_name != "Others")
      Data_survival_cluster = Data_survival %>% dplyr::filter(cluster != "Others")
      Data_survival_cancer_cluster = Data_survival_cancer %>% dplyr::filter(cluster != "Others")
      output$figure_survival_CTx_4_2 = renderPlot({
        ggforest(coxph(Surv(time = time_palliative_enroll,
                            time2 = time_palliative_final, censor)~Cancer_name,
                       data=Data_survival_cancer), data=Data_survival_cancer)
      })
      output$figure_survival_CTx_5_2 = renderPlot({
        ggforest(coxph(Surv(time = time_palliative_enroll,
                            time2 = time_palliative_final, censor)~cluster,
                       data=Data_survival_cluster), data=Data_survival_cluster)
      })
      output$figure_survival_CTx_5_1_2 = renderPlot({
        ggforest(coxph(Surv(time = time_palliative_enroll,
                            time2 = time_palliative_final, censor)~cluster+Cancer_name,
                       data=Data_survival_cancer_cluster), data=Data_survival_cancer_cluster)
      })
      
      incProgress(1 / 13)
    })
  })
  
  
  observeEvent(input$CGP_benefit_analysis, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      Data_case_target = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.基本情報.年齢,
                      Lymph_met,
                      Brain_met,
                      Lung_met,
                      Bone_met,
                      Liver_met,
                      Other_met,
                      EP_option,
                      EP_treat,
                      YoungOld,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      症例.EP前レジメン情報.最良総合効果.名称.,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.化学療法レジメン名称,
                      症例.EP前レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.EP後レジメン情報.治療ライン,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.化学療法レジメン名称,
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      症例.EP後レジメン情報.レジメン継続区分.名称.,
                      症例.EP後レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.最良総合効果.名称.,
                      症例.基本情報.性別.名称.,
                      症例.基本情報.がん種.OncoTree.,
                      症例.基本情報.がん種.OncoTree..名称.,
                      症例.基本情報.がん種.OncoTree.LEVEL1.,
                      症例.検体情報.パネル.名称.,
                      症例.背景情報.ECOG.PS.名称.,
                      症例.背景情報.喫煙歴有無.名称.,
                      症例.背景情報.アルコール多飲有無.名称.,
                      症例.背景情報.重複がん有無.異なる臓器..名称.,
                      症例.背景情報.多発がん有無.同一臓器..名称.,
                      症例.がん種情報.登録時転移部位.名称.,
                      症例.検体情報.腫瘍細胞含有割合,
                      症例.検体情報.検体採取部位.名称.,
                      症例.転帰情報.転帰.名称.,
                      症例.背景情報.診断日,
                      症例.管理情報.登録日,
                      症例.転帰情報.最終生存確認日,
                      症例.転帰情報.死亡日,
                      HER2_IHC,
                      MSI_PCR,
                      MMR_IHC
        )
      if(input$HER2 == "No"){
        Data_case_target = Data_case_target %>%
          dplyr::select(-HER2_IHC)
      } else {
        Data_case_target = Data_case_target %>%
          dplyr::filter(
            HER2_IHC != "Unknown"
          )
      }
      if(input$MSI == "No"){
        Data_case_target = Data_case_target %>%
          dplyr::select(-MSI_PCR)
      } else {
        Data_case_target = Data_case_target %>%
          dplyr::filter(
            MSI_PCR != "Unknown"
          )
      }
      if(input$MMR == "No"){
        Data_case_target = Data_case_target %>%
          dplyr::select(-MMR_IHC)
      } else {
        Data_case_target = Data_case_target %>%
          dplyr::filter(
            MMR_IHC != "Unknown"
          )
      }
      Data_case_target = Data_case_target %>%
      dplyr::mutate(
        症例.背景情報.喫煙歴有無.名称. = case_when(
          症例.背景情報.喫煙歴有無.名称. == "あり" ~ "Yes",
          症例.背景情報.喫煙歴有無.名称. == "なし" ~ "No",
          TRUE ~ "Unknown"
        ),
        症例.背景情報.アルコール多飲有無.名称. = case_when(
          症例.背景情報.アルコール多飲有無.名称. == "あり" ~ "Yes",
          症例.背景情報.アルコール多飲有無.名称. == "なし" ~ "No",
          TRUE ~ "Unknown"
        )
      )
      Data_case_target$Cancers = Data_case_target$症例.基本情報.がん種.OncoTree.
      Data_survival = Data_case_target %>%
        dplyr::mutate(final_observe = case_when(
          症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
          症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
          症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
          TRUE ~ 症例.管理情報.登録日))
      Data_survival = Data_survival %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::mutate(censor = case_when(
          症例.転帰情報.死亡日 != "" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
      
      Data_survival = Data_survival %>%
        dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
        dplyr::mutate(final_observe = ifelse(is.na(final_observe), 症例.管理情報.登録日, final_observe))
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, censor) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival = Data_survival %>%
        dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(Data_survival$症例.管理情報.登録日)) %>%
        dplyr::mutate(症例.背景情報.診断日 =
                        as.Date(Data_survival$症例.背景情報.診断日)) %>%
        dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                        as.Date(症例.EP前レジメン情報.投与開始日))
      incProgress(1 / 13)
      
      Data_survival = Data_survival %>%
        dplyr::mutate(time_diagnosis_enroll =
                        as.integer(difftime(Data_survival$症例.管理情報.登録日,
                                            Data_survival$症例.背景情報.診断日,
                                            units = "days")))      
      Data_survival = Data_survival %>%
        dplyr::mutate(time_enroll_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.管理情報.登録日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.EP前レジメン情報.投与開始日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_enroll =
                        Data_survival$time_palliative_final -
                        Data_survival$time_enroll_final)
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, final_observe) %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_diagnosis_enroll) %>%
        dplyr::arrange(desc(time_diagnosis_enroll)) %>%
        dplyr::filter(time_diagnosis_enroll > 0 & !is.na(time_diagnosis_enroll) & is.finite(time_diagnosis_enroll)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")

      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_enroll_final) %>%
        dplyr::arrange(desc(time_enroll_final)) %>%
        dplyr::filter(time_enroll_final > 0 & !is.na(time_enroll_final) & is.finite(time_enroll_final)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival_tmp = Data_survival %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                        c("その他", "緩和")) %>%
        dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_palliative_enroll, time_palliative_final) %>%
        dplyr::filter(time_palliative_enroll > 0 & time_palliative_enroll < 10000 & !is.na(time_palliative_enroll) & is.finite(time_palliative_enroll)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      tmp1 = Data_case_target %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                        c("その他", "緩和")) %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.化学療法レジメン名称,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      症例.管理情報.登録日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      final_observe,
                      censor,
                      症例.EP前レジメン情報.終了理由.名称.,
                      症例.EP前レジメン情報.最良総合効果.名称.) %>%
        dplyr::distinct()
      tmp1 = tmp1 %>%
        dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                        as.Date(tmp1$症例.EP前レジメン情報.投与開始日)) %>%
        dplyr::mutate(症例.EP前レジメン情報.投与終了日 =
                        as.Date(tmp1$症例.EP前レジメン情報.投与終了日)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(tmp1$症例.管理情報.登録日))
      tmp1 = tmp1 %>%
        dplyr::mutate(Drug_length =
                        as.integer(difftime(tmp1$症例.EP前レジメン情報.投与終了日,
                                            tmp1$症例.EP前レジメン情報.投与開始日,
                                            units = "days")))
      colnames(tmp1) = c("ID",
                         "Drug",
                         "レジメン",
                         "治療ライン",
                         "投与開始日",
                         "投与終了日",
                         "Ongoing",
                         "登録日",
                         "実施目的",
                         "final_observe",
                         "censor",
                         "終了理由",
                         "最良総合効果",
                         "Drug_length")
      tmp1$TxCGP = "Pre"
      tmp2 = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.化学療法レジメン名称,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      症例.EP後レジメン情報.レジメン継続区分.名称.,
                      症例.管理情報.登録日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      final_observe,
                      censor,
                      症例.EP後レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.最良総合効果.名称.) %>%
        dplyr::distinct()
      tmp2$症例.EP前レジメン情報.実施目的.名称. = "緩和"
      tmp2 = tmp2 %>%
        dplyr::mutate(症例.EP後レジメン情報.投与開始日 =
                        as.Date(tmp2$症例.EP後レジメン情報.投与開始日)) %>%
        dplyr::mutate(症例.EP後レジメン情報.投与終了日 =
                        as.Date(tmp2$症例.EP後レジメン情報.投与終了日)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(tmp2$症例.管理情報.登録日))
      
      tmp2 = tmp2 %>%
        dplyr::mutate(Drug_length =
                        as.integer(difftime(tmp2$症例.EP後レジメン情報.投与終了日,
                                            tmp2$症例.EP後レジメン情報.投与開始日,
                                            units = "days")))
      colnames(tmp2) = c("ID",
                         "Drug",
                         "レジメン",
                         "治療ライン",
                         "投与開始日",
                         "投与終了日",
                         "Ongoing",
                         "登録日",
                         "実施目的",
                         "final_observe",
                         "censor",
                         "終了理由",
                         "最良総合効果",
                         "Drug_length")
      tmp2$TxCGP = "Post"
      Data_drug = rbind(tmp1, tmp2)
      
      if(!is.null(input$ID_drug)){
        Data_drug = data.frame(NULL)
        for(i in 1:length(input$ID_drug[,1])){
          tmp = read.csv(header = TRUE, file(input$ID_drug[[i, 'datapath']],
                                             encoding='UTF-8-BOM'))
          if("症例.EP前レジメン情報.投与開始日" %in% colnames(tmp)){
            colnames(tmp) = c("ID",
                              "登録日",
                              "治療ライン",
                              "実施目的",
                              "レジメン",
                              "Drug",
                              "投与開始日",
                              "投与終了日",
                              "Ongoing",
                              "終了理由",
                              "最良総合効果",
                              "最終生存確認日",
                              "死亡日")
            tmp$投与開始日 = as.Date(tmp$投与開始日)
            tmp$投与終了日 = as.Date(tmp$投与終了日)
            tmp = tmp %>%
              dplyr::mutate(
                Drug_length = as.integer(difftime(tmp$投与終了日,
                                                  tmp$投与開始日,
                                                  units = "days")),
                censor = case_when(
                  死亡日 != "" ~ 1,
                  TRUE ~ 0
                ),
                final_observe = case_when(
                  最終生存確認日 == "           " ~ 死亡日,
                  最終生存確認日 != "" ~ 最終生存確認日,
                  死亡日 != "" ~ 死亡日,
                  TRUE ~ 登録日)
              )
            tmp$登録日 = as.Date(tmp$登録日)
            tmp$最終生存確認日 = as.Date(tmp$最終生存確認日)
            tmp$死亡日 = as.Date(tmp$死亡日)
            tmp$final_observe = as.Date(tmp$final_observe)
            tmp = tmp %>%
              dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
              dplyr::mutate(final_observe = ifelse(is.na(final_observe), 登録日, final_observe))
            tmp = tmp %>% dplyr::select(-最終生存確認日, -死亡日)
            tmp$TxCGP = "Pre"
          } else{
            colnames(tmp) = c("ID",
                              "登録日",
                              "治療ライン",
                              "レジメン",
                              "Drug",
                              "投与開始日",
                              "投与終了日",
                              "Ongoing",
                              "終了理由",
                              "最良総合効果",
                              "最終生存確認日",
                              "死亡日")
            tmp$投与開始日 = as.Date(tmp$投与開始日)
            tmp$投与終了日 = as.Date(tmp$投与終了日)
            tmp = tmp %>%
              dplyr::mutate(
                Drug_length = as.integer(difftime(tmp$投与終了日,
                                                  tmp$投与開始日,
                                                  units = "days")),
                censor = case_when(
                  死亡日 != "" ~ 1,
                  TRUE ~ 0
                ),
                final_observe = case_when(
                  最終生存確認日 == "           " ~ 死亡日,
                  最終生存確認日 != "" ~ 最終生存確認日,
                  死亡日 != "" ~ 死亡日,
                  TRUE ~ 登録日)
              )
            tmp$登録日 = as.Date(tmp$登録日)
            tmp$最終生存確認日 = as.Date(tmp$最終生存確認日)
            tmp$死亡日 = as.Date(tmp$死亡日)
            tmp$final_observe = as.Date(tmp$final_observe)
            tmp = tmp %>%
              dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
              dplyr::mutate(final_observe = ifelse(is.na(final_observe), 登録日, final_observe))
            tmp = tmp %>% dplyr::select(-最終生存確認日, -死亡日)
            tmp$実施目的 = "緩和"
            tmp$TxCGP = "Post"
          }
          Data_drug = rbind(Data_drug, tmp)
        }
      }
      
      Data_drug = Data_drug %>%
        dplyr::mutate(final_observe =
                        as.Date(Data_drug$final_observe))
      Data_drug = Data_drug %>%
        dplyr::arrange(投与開始日) %>%
        dplyr::arrange(治療ライン) %>%
        dplyr::arrange(ID)
      
      Data_drug = Data_drug %>% dplyr::mutate(RECIST = case_when(
        最良総合効果 == "CR" ~ "5",
        最良総合効果 == "PR" ~ "4",
        最良総合効果 == "SD" ~ "3",
        最良総合効果 == "PD" ~ "2",
        最良総合効果 == "NE" ~ "1",
        最良総合効果 == "Unknown" ~ "0",
        TRUE ~ "0"
      ))
      Data_drug = Data_drug %>%
        dplyr::filter(治療ライン %in% c("１次治療","２次治療","３次治療","４次治療"))
      
      Data_drug_RECIST = (Data_drug %>%
                            dplyr::arrange(desc(RECIST),ID,desc(レジメン),治療ライン) %>%
                            dplyr::distinct(ID,治療ライン,.keep_all = T)) %>%
        dplyr::select(ID,治療ライン,RECIST)
      
      Data_drug = Data_drug %>% dplyr::select(-投与開始日, -投与終了日, -RECIST, -Ongoing, -Drug)
      Data_drug = left_join(Data_drug, Data_drug_RECIST, by = c('ID','治療ライン'))
      Data_drug = Data_drug %>%
        dplyr::arrange(ID) %>%
        dplyr::distinct(ID,治療ライン, .keep_all = T)
      
      Data_drug <- data.frame(Data_drug %>% 
                                dplyr::arrange(ID) %>%
                                dplyr::group_by(ID) %>%
                                dplyr::mutate(治療ライン = row_number()))
      Data_drug$治療ライン = as.character(Data_drug$治療ライン)
      
      ID_pre_CGP_lines = Data_drug %>%
        dplyr::filter(TxCGP == "Pre") %>%
        dplyr::select(ID, 治療ライン) %>%
        dplyr::arrange(desc(治療ライン)) %>%
        dplyr::distinct()
      colnames(ID_pre_CGP_lines) = c("ID", "CTx_lines_before_CGP")
      
      ID_pre_CGP_RECIST = Data_drug %>%
        dplyr::filter(TxCGP == "Pre") %>%
        dplyr::select(ID, RECIST) %>%
        dplyr::arrange(desc(RECIST)) %>%
        dplyr::distinct()
      colnames(ID_pre_CGP_RECIST) = c("ID", "RECIST")
      
      ID_pre_CGP_RECIST = ID_pre_CGP_RECIST %>% dplyr::mutate(RECIST = case_when(
        RECIST == "5" ~ "CR",
        RECIST == "4" ~ "PR",
        RECIST == "3" ~ "SD",
        RECIST == "2" ~ "PD",
        RECIST == "1" ~ "NE",
        RECIST == "0" ~ "Unknown",
        TRUE ~ "Unknown"
      ))
      Data_case_target$CTx_lines_before_CGP = unlist(lapply(list(Data_case_target$C.CAT調査結果.基本項目.ハッシュID), function(x) {
        as.vector(ID_pre_CGP_lines$CTx_lines_before_CGP[match(x, ID_pre_CGP_lines$ID)])}))
      Data_case_target$RECIST = unlist(lapply(list(Data_case_target$C.CAT調査結果.基本項目.ハッシュID), function(x) {
        as.vector(ID_pre_CGP_RECIST$RECIST[match(x, ID_pre_CGP_RECIST$ID)])}))
      if(sum(is.na(Data_case_target$CTx_lines_before_CGP)>0)){
        Data_case_target[is.na(Data_case_target$CTx_lines_before_CGP),]$CTx_lines_before_CGP = "0"
      }
      if(sum(is.na(Data_case_target$RECIST)>0)){
        Data_case_target[is.na(Data_case_target$RECIST),]$RECIST = "CTx_naive"
      }
      
      Data_case_target = Data_case_target %>%
        dplyr::distinct(.keep_all = TRUE, C.CAT調査結果.基本項目.ハッシュID)
      Data_MAF = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","A","B","C","D","E","F") &
            Variant_Classification != "expression"
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
      if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
        Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
      }
      ID_evidence_AB = unique((Data_report() %>%
                                 dplyr::filter(
                                   Evidence_level %in% c("A","B")) %>% 
                                 dplyr::arrange(Tumor_Sample_Barcode))$Tumor_Sample_Barcode)
      Data_report_TMB = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
        dplyr::select(Tumor_Sample_Barcode, TMB) %>%
        dplyr::distinct(Tumor_Sample_Barcode, .keep_all=T)
      colnames(Data_report_TMB) = c("C.CAT調査結果.基本項目.ハッシュID", "TMB")
      Data_case_target = left_join(Data_case_target, Data_report_TMB,
                                   by = "C.CAT調査結果.基本項目.ハッシュID")
      
      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      if(input$patho == "Only pathogenic muts"){
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(Evidence_level == "F")
      }
      Data_MAF_target = Data_MAF_target %>%
        dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol, .keep_all = T)
      Gene_list = unique(names(sort(table(Data_MAF_target$Hugo_Symbol),
                                    decreasing = T)))
      if(length(Data_case_target$censor[is.na(Data_case_target$censor)])> 0){
        Data_case_target$censor[is.na(Data_case_target$censor)] = 0
      }
      incProgress(1 / 13)
      
      Data_case_target = Data_case_target %>%
        dplyr::mutate(EP_option = case_when(
          症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称. ==
            "はい" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::mutate(EP_treat = case_when(
          症例.EP後レジメン情報.提示された治療薬を投与した.名称. ==
            "投与した" ~ 1,
          TRUE ~ 0))
      
      Data_tmp = Data_case_target %>% dplyr::select(
        C.CAT調査結果.基本項目.ハッシュID,
        症例.EP後レジメン情報.治療ライン,
        症例.EP後レジメン情報.化学療法レジメン名称,
        症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
        症例.EP後レジメン情報.提示された治療薬を投与した.名称.
      ) %>%
        dplyr::distinct()
      Data_tmp = tidyr::replace_na(Data_tmp,
                                   replace = list(症例.EP後レジメン情報.治療ライン = 0,
                                                  症例.EP後レジメン情報.化学療法レジメン名称 = "",
                                                  症例.EP後レジメン情報.薬剤名.YJ一般名.EN. = ""))
      Data_tmp = Data_tmp %>%
        dplyr::filter(症例.EP後レジメン情報.治療ライン != 0 &(
          症例.EP後レジメン情報.化学療法レジメン名称 != "" |
            症例.EP後レジメン情報.薬剤名.YJ一般名.EN. != "" |
            症例.EP後レジメン情報.提示された治療薬を投与した.名称. == "投与した"))
      ID_tmp = unique(Data_tmp$C.CAT調査結果.基本項目.ハッシュID)
      Data_survival = Data_case_target %>% dplyr::filter(
        !is.na(time_enroll_final) &
          is.finite(time_enroll_final) & 
          time_enroll_final > 0 &
          !is.na(censor) &
          is.finite(censor)
      )  
      Data_survival$afterCGPtreat = Data_survival$EP_treat
      Data_survival$afterCGPtreat[Data_survival$C.CAT調査結果.基本項目.ハッシュID %in% ID_tmp] <- 1
      
      Data_survival = Data_survival %>%
        dplyr::filter(time_enroll_final > 0 & !is.na(time_enroll_final) & is.finite(time_enroll_final))
      Cancername = sort(unique(Data_survival$Cancers))
      Data_MAF_target = Data_MAF_target %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      incProgress(1 / 13)
      Data_case_target$症例.管理情報.登録日 =
        as.Date(Data_case_target$症例.管理情報.登録日)
      
      Data_forest = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      YoungOld,
                      Lymph_met,
                      Brain_met,
                      Lung_met,
                      Bone_met,
                      Liver_met,
                      #Other_met,
                      EP_option,
                      EP_treat,
                      症例.基本情報.性別.名称.,
                      症例.基本情報.がん種.OncoTree.,
                      症例.基本情報.がん種.OncoTree.LEVEL1.,
                      症例.背景情報.ECOG.PS.名称.,
                      症例.背景情報.喫煙歴有無.名称.,
                      症例.背景情報.アルコール多飲有無.名称.,
                      症例.管理情報.登録日,
                      time_enroll_final,
                      time_diagnosis_enroll,
                      censor,
                      CTx_lines_before_CGP,
                      RECIST,
                      症例.検体情報.パネル.名称.,
                      症例.基本情報.年齢)
      colnames(Data_forest) = 
        c('ID', 'Age',
          'Lymph_met',
          'Brain_met',
          'Lung_met',
          'Bone_met',
          'Liver_met',
          #'Other_met',
          'EP_option',
          'EP_treat',
          'Sex', 'Histology', 'Histology_dummy', 'PS',
          'Smoking_history', 'Alcoholic_history',
          'Enroll_date',
          'time_enroll_final',
          'time_diagnosis_enroll',
          'censor', "Lines", "Best_effect", "Panel","Age_raw")
      #Data_forest$Histology_dummy = paste0(Data_forest$Histology_dummy,"_NOS")
      Data_forest$time_diagnosis_enroll = ifelse(Data_forest$time_diagnosis_enroll > 365.25*1, ">1-year", "<=1-year")
      Data_forest$Enroll_date <- format(as.Date(Data_forest$Enroll_date), "%Y")
      col_added = 4
      if(input$HER2 != "No"){
        Data_forest$HER2_IHC = Data_case_target$HER2_IHC
        Data_forest = Data_forest %>%
          dplyr::filter(
            HER2_IHC != "Unknown"
          )
        col_added = col_added + 1
      }
      if(input$MSI != "No"){
        Data_forest$MSI_PCR = Data_case_target$MSI_PCR
        Data_forest = Data_forest %>%
          dplyr::filter(
            MSI_PCR != "Unknown"
          )
        col_added = col_added + 1
      }
      if(input$MMR != "No"){
        Data_forest$MMR_IHC = Data_case_target$MMR_IHC
        Data_forest = Data_forest %>%
          dplyr::filter(
            MMR_IHC != "Unknown"
          )
        col_added = col_added + 1
      }
      
      Data_forest = Data_forest %>%
        dplyr::filter(PS != "不明" & Sex != "未入力・不明" & Smoking_history != "Unknown" & Alcoholic_history != "Unknown" & !is.na(time_diagnosis_enroll) & !is.na(Enroll_date)) %>%
        dplyr::mutate(
          PS = case_when(
            PS == 0 ~ "0",
            PS == 1 ~ "1",
            TRUE ~ "2_4"
          ),
          Panel = case_when(
            Panel %in% c("NCC OncoPanel", "FoundationOne CDx", "GenMineTOP") ~ Panel,
            TRUE ~ "Liquid"
          ),
          Lines = case_when(
            Lines == "0" ~ "0",
            Lines == "1" ~ "1",
            TRUE ~ "2~"
          )
        )
      # Data_forest$Panel = factor(Data_forest$Panel)
      Data_forest$Best_effect[Data_forest$Best_effect == "Unknown"] = "NE"
      Data_forest$Best_effect = factor(Data_forest$Best_effect, levels = c("SD","CR","PR","PD","NE","CTx_naive"))
      # if(length(unique(Data_forest$Panel))>1){
      #   Data_forest$Panel =  relevel(Data_forest$Panel, ref=names(sort(table(Data_forest$Panel)))[[1]]) 
      # }
      frequent_organs = sort(table(Data_forest$Histology),decreasing = T)
      frequent_organs_names = names(frequent_organs[frequent_organs>=50])
      Data_forest = Data_forest %>% dplyr::mutate(
        Histology = case_when(
          Histology %in% frequent_organs_names ~ Histology,
          TRUE ~ Histology_dummy
        )
      )
      oncogenic_genes = Data_MAF_target %>%
        dplyr::select(Tumor_Sample_Barcode, Hugo_Symbol) %>%
        dplyr::distinct() %>%
        dplyr::count(Hugo_Symbol) %>%
        dplyr::arrange(-n)
      colnames(oncogenic_genes) = c("gene_mutation", "all_patients")
      oncogenic_genes = rbind(oncogenic_genes %>% dplyr::filter(gene_mutation %in% input$gene),
                              oncogenic_genes %>% dplyr::filter(!gene_mutation %in% input$gene))
      oncogenic_genes = oncogenic_genes[1:min(length(oncogenic_genes$all_patients), input$gene_no),]
      Gene_data = data.frame(oncogenic_genes$gene_mutation)
      colnames(Gene_data) = c("Gene")
      

      Data_forest =  Data_forest %>%
        dplyr::select(-Age_raw)
      gene_names = unique(c(input$gene,
                            as.character(Gene_data$Gene)))
      gene_names = gene_names[gene_names %in% unique(Data_MAF_target$Hugo_Symbol)]
      gene_names = gene_names[1:min(input$gene_no, length(gene_names))]
      for(i in 1:length(gene_names)){
        ID_mutation = unique((Data_MAF_target %>% dplyr::filter(Hugo_Symbol == gene_names[i]))$Tumor_Sample_Barcode)
        Data_forest_tmp = Data_forest %>% 
          dplyr::select(ID) %>%
          dplyr::distinct() %>%
          dplyr::mutate(
            XXX = case_when(
              ID %in% ID_mutation ~ "mut(+)",
              TRUE ~ "mut(-)"
            )
          )
        colnames(Data_forest_tmp) = c("ID", gene_names[i])
        Data_forest = left_join(Data_forest, Data_forest_tmp, by=c("ID"))
      }
      # Factor_names =
      # Factor_names = Factor_names[!Factor_names %in% c('ID', 'time_enroll_final', 'censor', 'EP_treat', 'EP_option')]
      
      Data_forest_tmp__ = Data_forest %>%
        dplyr::select(-time_enroll_final, -censor, -EP_option, -ID)
      Disease_tmp = unique(sort(Data_forest_tmp__$Histology))
      for(i in 1:length(Disease_tmp)){
        Data_forest_tmp_2_ = Data_forest_tmp__ %>% dplyr::filter(Histology == Disease_tmp[i])
        if(sum(Data_forest_tmp_2_$EP_treat == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::mutate(Histology = case_when(
              Histology == Disease_tmp[i] ~ Histology_dummy,
              TRUE ~ Histology
            )
            )
        }
      }
      for(i in length(gene_names):1){
        kk = 14+col_added+i
        if(sum(Data_forest_tmp__[[kk]] == "mut(+)" & Data_forest_tmp__$EP_treat == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__[,-..kk]
        } else if(sum(Data_forest_tmp__[[kk]] != "mut(+)" & Data_forest_tmp__$EP_treat == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__[,-..kk]
        }
      }
      if(sum(Data_forest_tmp__$time_diagnosis_enroll == ">1-year" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-time_diagnosis_enroll)
      } else if(sum(Data_forest_tmp__$time_diagnosis_enroll != ">1-year" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-time_diagnosis_enroll)
      }
      if(sum(Data_forest_tmp__$PS == "0" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-PS)
      } else if(sum(Data_forest_tmp__$PS != "0" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-PS)
      } else if(sum(Data_forest_tmp__$PS == "2_4" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__$PS[Data_forest_tmp__$PS == "1"] = "1_4"
        Data_forest_tmp__$PS[Data_forest_tmp__$PS == "2_4"] = "1_4"
      }
      if((sum(Data_forest_tmp__$Panel == "NCC OncoPanel" & Data_forest_tmp__$EP_treat == 1) < 3) |
         (sum(Data_forest_tmp__$Panel == "FoundationOne CDx" & Data_forest_tmp__$EP_treat == 1) < 3) |
         (sum(Data_forest_tmp__$Panel == "GenMineTOP" & Data_forest_tmp__$EP_treat == 1) < 3)){
          Data_forest_tmp__$Panel[Data_forest_tmp__$Panel == "NCC OncoPanel"] = "Solid"
          Data_forest_tmp__$Panel[Data_forest_tmp__$Panel == "FoundationOne CDx"] = "Solid"
          Data_forest_tmp__$Panel[Data_forest_tmp__$Panel == "GenMineTOP"] = "Solid"
          if(sum(Data_forest_tmp__$Panel == "Solid" & Data_forest_tmp__$EP_treat == 1) < 3){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Panel)
          } else if(sum(Data_forest_tmp__$Panel != "Solid" & Data_forest_tmp__$EP_treat == 1) < 3){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Panel)
          }
      }
      if(sum(Data_forest_tmp__$Lines == "1" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Lines)
      } else if(sum(Data_forest_tmp__$Lines != "1" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Lines)
      } else if(sum(Data_forest_tmp__$Lines == "0" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "0"] = "0~1"
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "1"] = "0~1"
      } else if(sum(Data_forest_tmp__$Lines == "2~" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "2~"] = "1~"
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "1"] = "1~"
      }
      if(sum(Data_forest_tmp__$Enroll_date == "2019" & Data_forest_tmp__$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2019"] = "2019/20"
        Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2020"] = "2019/20"
        if(sum(Data_forest_tmp__$Enroll_date == "2019/20" & Data_forest_tmp__$EP_treat == 1,na.rm = T) < 3){
          Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2019/20"] = "2019-21"
          Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2021"] = "2019-21"
          if(sum(Data_forest_tmp__$Enroll_date == "2019-21" & Data_forest_tmp__$EP_treat == 1,na.rm = T) < 3){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Enroll_date)
          } else if(sum(Data_forest_tmp__$Enroll_date != "2019-21" & Data_forest_tmp__$EP_treat == 1,na.rm = T) < 3){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Enroll_date)
          }
        } else if(sum(Data_forest_tmp__$Enroll_date != "2019/20" & Data_forest_tmp__$EP_treat == 1,na.rm = T) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-Enroll_date)
        }
      } else if(sum(Data_forest_tmp__$Enroll_date != "2019" & Data_forest_tmp__$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Enroll_date)
      }
      if(sum(Data_forest_tmp__$Best_effect == "SD" & Data_forest_tmp__$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Best_effect)
      } else if(sum(Data_forest_tmp__$Best_effect != "SD" & Data_forest_tmp__$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Best_effect)
      }
      if(sum(Data_forest_tmp__$Age == "Younger" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Age)
      } else if(sum(Data_forest_tmp__$Age == "Older" & Data_forest_tmp__$EP_treat == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Age)
      }
      if(sum(Data_forest_tmp__$Sex == "男" & Data_forest_tmp__$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Sex)
      } else if(sum(Data_forest_tmp__$Sex != "男" & Data_forest_tmp__$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Sex)
      }
      metastasis_columns <- c("Smoking_history", "Alcoholic_history", "Lymph_met", "Lung_met", "Brain_met", "Bone_met", "Liver_met")
      
      # 削除対象の列名を取得（EP_treat == 1 の行において、"Yes" の数または "Yes" でない数が3未満なら削除）
      cols_to_drop <- metastasis_columns[
        sapply(metastasis_columns, function(col) {
          yes_count <- Data_forest_tmp__[EP_treat == 1, sum(get(col) == "Yes", na.rm = TRUE)]
          no_count  <- Data_forest_tmp__[EP_treat == 1, sum(get(col) != "Yes", na.rm = TRUE)]
          (yes_count < 3 || no_count < 3)
        })
      ]
      
      # 削除対象の列がある場合、一括削除
      if (length(cols_to_drop) > 0) {
        Data_forest_tmp__[, (cols_to_drop) := NULL]
      }
      if(input$HER2 != "No"){
        if(sum(Data_forest_tmp__$HER2_IHC == "Positive" & Data_forest_tmp__$EP_treat == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-HER2_IHC)
        } else if(sum(Data_forest_tmp__$HER2_IHC != "Positive" & Data_forest_tmp__$EP_treat == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-HER2_IHC)
        }
      }
      if(input$MSI != "No"){
        if(sum(Data_forest_tmp__$MSI_PCR == "Positive" & Data_forest_tmp__$EP_treat == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MSI_PCR)
        } else if(sum(Data_forest_tmp__$MSI_PCR != "Positive" & Data_forest_tmp__$EP_treat == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MSI_PCR)
        }
      }
      if(input$MMR != "No"){
        if(sum(Data_forest_tmp__$MMR_IHC == "dMMR" & Data_forest_tmp__$EP_treat == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MMR_IHC)
        } else if(sum(Data_forest_tmp__$MMR_IHC != "dMMR" & Data_forest_tmp__$EP_treat == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MMR_IHC)
        }
      }
      Data_forest_tmp__$Histology = factor(Data_forest_tmp__$Histology)
      if(length(unique(Data_forest_tmp__$Histology))<2){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Histology)
      } else {
        Data_forest_tmp__$Histology = relevel(Data_forest_tmp__$Histology,
                                            ref=names(sort(table(Data_forest_tmp__$Histology),decreasing = T))[[1]])
      }
      Data_forest_tmp__ = Data_forest_tmp__ %>%
        dplyr::select(-Histology_dummy)
      
      if(length(colnames(Data_forest_tmp__)) > 1){
        Factor_names_tmp__ = colnames(Data_forest_tmp__)
        Factor_names_tmp__ = Factor_names_tmp__[Factor_names_tmp__ != "EP_treat"]
        for(i in length(Factor_names_tmp__):1){
          if(all(as.character(Data_forest_tmp__[[Factor_names_tmp__[i]]]) ==
                              as.character(Data_forest_tmp__[[Factor_names_tmp__[i]]][1]))){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Factor_names_tmp__[i])
          }
        }
      }
      Data_forest_tmp__univariant = Data_forest_tmp__
      if(length(colnames(Data_forest_tmp__)) > 2){
        Factor_names_tmp__ = colnames(Data_forest_tmp__)
        Factor_names_tmp__ = Factor_names_tmp__[Factor_names_tmp__ != "EP_treat"]
        for(i in 1:(length(Factor_names_tmp__) - 1)){
          for(j in (i+1):length(Factor_names_tmp__)){
            if(Factor_names_tmp__[i] %in% colnames(Data_forest_tmp__) &
               Factor_names_tmp__[j] %in% colnames(Data_forest_tmp__)){
              if(abs(cor(as.numeric(as.factor(Data_forest_tmp__[[Factor_names_tmp__[i]]])),
                         as.numeric(as.factor(Data_forest_tmp__[[Factor_names_tmp__[j]]])))) > 0.95){
                Data_forest_tmp__ = Data_forest_tmp__ %>%
                  dplyr::select(-Factor_names_tmp__[j])
              }
            }
          }
        }
      }
      
      Data_forest_tmp_7 = data.frame(Data_forest_tmp__)
      Data_forest_tmp_7__univariant = data.frame(Data_forest_tmp__univariant)
      Data_forest_tmp_7 = Data_forest_tmp_7[,colnames(Data_forest_tmp_7) %in% 
                                              c("Age","Lymph_met","Brain_met","Lung_met","Bone_met","Liver_met","Sex","Histology","PS","Lines","Best_effect","EP_treat","Panel",
                                                "HER2_IHC", "MSI_PCR", "MMR_IHC","time_diagnosis_enroll", "Smoking_history", "Alcoholic_history", "Enroll_date"), drop=FALSE]
      Data_forest_tmp_7__univariant = Data_forest_tmp_7__univariant[,colnames(Data_forest_tmp_7__univariant) %in% 
                                                                    c("Age","Lymph_met","Brain_met","Lung_met","Bone_met","Liver_met","Sex","Histology","PS","Lines","Best_effect","EP_treat","Panel",
                                                                      "HER2_IHC", "MSI_PCR", "MMR_IHC","time_diagnosis_enroll", "Smoking_history", "Alcoholic_history", "Enroll_date"), drop=FALSE]
      output$figure_survival_CGP_7 = render_gt({
        if(length(colnames(Data_forest_tmp_7)) > 1){
          Data_forest_tmp_7 = data.frame( lapply(Data_forest_tmp_7, as.factor) )
          Data_forest_tmp_7__univariant = data.frame( lapply(Data_forest_tmp_7__univariant, as.factor) )
          if("Histology" %in% colnames(Data_forest_tmp_7)){
            Data_forest_tmp_7$Histology <- relevel(Data_forest_tmp_7$Histology, ref=names(sort(table(Data_forest_tmp_7$Histology),decreasing = T))[[1]]) 
          }
          if("Histology" %in% colnames(Data_forest_tmp_7__univariant)){
            Data_forest_tmp_7__univariant$Histology <- relevel(Data_forest_tmp_7__univariant$Histology, ref=names(sort(table(Data_forest_tmp_7__univariant$Histology),decreasing = T))[[1]]) 
          }
          if("Sex" %in% colnames(Data_forest_tmp_7)){
            Data_forest_tmp_7 = Data_forest_tmp_7 %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          if("Sex" %in% colnames(Data_forest_tmp_7__univariant)){
            Data_forest_tmp_7__univariant = Data_forest_tmp_7__univariant %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          colnames_tmp = colnames(Data_forest_tmp_7__univariant)
          colnames_tmp[colnames_tmp == "Age"] = paste0("Age ~", input$mid_age-1,"/",input$mid_age, "~")
          colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
          colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
          colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
          colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
          colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
          colnames_tmp[colnames_tmp == "PS"] = "Performance status"
          colnames_tmp[colnames_tmp == "Lines"] = "CTx lines before CGP"
          colnames_tmp[colnames_tmp == "Best_effect"] = "Best CTx effect before CGP"
          colnames(Data_forest_tmp_7__univariant) = colnames_tmp
          if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m3_uni.qs")){
            univ_tab = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m3_uni.qs")
          } else {
            # 並列処理で実行する関数
            run_univariate_model <- function(var, data) {
              fml <- as.formula(paste("EP_treat ~", paste("`", var, "`", sep = "")))
              model <- glm(fml, data = data, family = binomial)
              tbl_regression(model, exponentiate = TRUE, conf.int = FALSE) %>%
                add_global_p() %>%
                add_n(location = "level") %>%
                add_nevent(location = "level") |>
                add_q() |>
                bold_p() |>
                # modify_fmt_fun(conf.low ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                # modify_fmt_fun(conf.high ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels() %>%
                tbl_butcher()
            }
            # 並列処理する変数リスト
            univ_vars <- setdiff(names(Data_forest_tmp_7__univariant), "EP_treat")
            # 利用可能なコア数を取得
            num_cores <- min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE))
            # 並列処理の実行
            univ_models_parallel <- mclapply(
              univ_vars,
              run_univariate_model,
              data = Data_forest_tmp_7__univariant,
              mc.cores = num_cores
            )
            names(univ_models_parallel) <- univ_vars # 結果に変数名を付与 (mclapply の場合)
            univ_tab <- tbl_stack(univ_models_parallel)
            if(input$intermediate_file != "No"){ 
              QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), univ_tab, file = "source/m3_uni.qs")
            }
          }  
          final_mv_regxlevels <- (prepare_data(as.numeric(as.character(Data_forest_tmp_7$EP_treat)),
                                               as.matrix(data.frame(lapply(Data_forest_tmp_7[, setdiff(names(Data_forest_tmp_7), "EP_treat")],
                                                                           function(x) as.numeric(as.factor(x))))), type = "logistic") %>% stepwise(aic))$model
          Data_forest_tmp_7 = Data_forest_tmp_7 %>% select(c(final_mv_regxlevels, "EP_treat"))
          colnames_tmp = colnames(Data_forest_tmp_7)
          colnames_tmp[colnames_tmp == "Age"] = paste0("Age ~", input$mid_age-1,"/",input$mid_age, "~")
          colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
          colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
          colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
          colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
          colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
          colnames_tmp[colnames_tmp == "PS"] = "Performance status"
          colnames_tmp[colnames_tmp == "Lines"] = "CTx lines before CGP"
          colnames_tmp[colnames_tmp == "Best_effect"] = "Best CTx effect before CGP"
          colnames(Data_forest_tmp_7) = colnames_tmp
          Data_forest_tmp_7 = Data_forest_tmp_7[,colnames(Data_forest_tmp_7) != "CTx lines before CGP"]
          if(length(final_mv_regxlevels) != 0){
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m3.qs")){
              mv_tab = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m3.qs")
            } else {
              mv_tab = glm(EP_treat ~.,
                           data = Data_forest_tmp_7,
                           family = binomial(link = "logit")) |>
                tbl_regression(                         ## 多変量解析の表を生成
                  exponentiate = TRUE,                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
                  conf.int = FALSE
                ) |>
                add_global_p() |> # add global p-value
                add_q() |> # adjusts global p-values for multiple testing
                bold_p() |> # bold p-values under a given threshold (default 0.05)
                # modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                # modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels() %>%
                tbl_butcher()
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), mv_tab, file = "source/m3.qs")
              }
            }  
            tbl_merge(
              tbls = list(univ_tab, mv_tab),
              tab_spanner = c("**Univariate**", "**Multivariable**")) |>
              modify_caption("Factors that led patients to receive the recommended treatment (Only factors with >2 events observed)") |> as_gt()
          } else {
            univ_tab |>
              modify_caption("Factors that led patients to receive the recommended treatment, no significant factor in multivariable analysis (Only factors with >2 events observed)") |> as_gt()
          }
        }
      })
      
      # Factor_names = colnames(Data_forest)
      # Factor_names = Factor_names[!Factor_names %in% c('ID', 'time_enroll_final', 'censor', 'EP_treat', 'EP_option')]
      
      Data_forest_tmp__ = Data_forest %>%
        dplyr::select(-time_enroll_final, -censor, -EP_treat, -ID)
      Disease_tmp = unique(sort(Data_forest_tmp__$Histology))
      for(i in 1:length(Disease_tmp)){
        Data_forest_tmp_2_ = Data_forest_tmp__ %>% dplyr::filter(Histology == Disease_tmp[i])
        if(sum(Data_forest_tmp_2_$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::mutate(Histology = case_when(
              Histology == Disease_tmp[i] ~ Histology_dummy,
              TRUE ~ Histology
            )
            )
        }
      }
      for(i in length(gene_names):1){
        kk = 14+col_added+i
        if(sum(Data_forest_tmp__[[kk]] == "mut(+)" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__[,-..kk]
        } else if(sum(Data_forest_tmp__[[kk]] != "mut(+)" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__[,-..kk]
        }
      }

      if(sum(Data_forest_tmp__$time_diagnosis_enroll == ">1-year" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-time_diagnosis_enroll)
      } else if(sum(Data_forest_tmp__$time_diagnosis_enroll != ">1-year" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-time_diagnosis_enroll)
      }
      
      
      
      if(sum(Data_forest_tmp__$PS == "0" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-PS)
      } else if(sum(Data_forest_tmp__$PS != "0" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-PS)
      } else if(sum(Data_forest_tmp__$PS == "2_4" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__$PS[Data_forest_tmp__$PS == "1"] = "1_4"
        Data_forest_tmp__$PS[Data_forest_tmp__$PS == "2_4"] = "1_4"
      }
      if((sum(Data_forest_tmp__$Panel == "NCC OncoPanel" & Data_forest_tmp__$EP_option == 1) < 3) |
         (sum(Data_forest_tmp__$Panel == "FoundationOne CDx" & Data_forest_tmp__$EP_option == 1) < 3) |
         (sum(Data_forest_tmp__$Panel == "GenMineTOP" & Data_forest_tmp__$EP_option == 1) < 3)){
        Data_forest_tmp__$Panel[Data_forest_tmp__$Panel == "NCC OncoPanel"] = "Solid"
        Data_forest_tmp__$Panel[Data_forest_tmp__$Panel == "FoundationOne CDx"] = "Solid"
        Data_forest_tmp__$Panel[Data_forest_tmp__$Panel == "GenMineTOP"] = "Solid"
        if(sum(Data_forest_tmp__$Panel == "Solid" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-Panel)
        } else if(sum(Data_forest_tmp__$Panel != "Solid" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-Panel)
        }
      }
      if(sum(Data_forest_tmp__$Enroll_date == "2019" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2019"] = "2019/20"
        Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2020"] = "2019/20"
        if(sum(Data_forest_tmp__$Enroll_date == "2019/20" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
          Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2019/20"] = "2019-21"
          Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2021"] = "2019-21"
          if(sum(Data_forest_tmp__$Enroll_date == "2019-21" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Enroll_date)
          } else if(sum(Data_forest_tmp__$Enroll_date != "2019-21" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Enroll_date)
          }
        } else if(sum(Data_forest_tmp__$Enroll_date != "2019/20" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-Enroll_date)
        }
      } else if(sum(Data_forest_tmp__$Enroll_date != "2019" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Enroll_date)
      }
      if(sum(Data_forest_tmp__$Lines == "1" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Lines)
      } else if(sum(Data_forest_tmp__$Lines != "1" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Lines)
      } else if(sum(Data_forest_tmp__$Lines == "0" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "0"] = "0~1"
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "1"] = "0~1"
      } else if(sum(Data_forest_tmp__$Lines == "2~" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "2~"] = "1~"
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "1"] = "1~"
      }
      if(sum(Data_forest_tmp__$Best_effect == "SD" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Best_effect)
      } else if(sum(Data_forest_tmp__$Best_effect != "SD" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Best_effect)
      }
      if(sum(Data_forest_tmp__$Age == "Younger" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Age)
      } else if(sum(Data_forest_tmp__$Age == "Older" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Age)
      }
      if(sum(Data_forest_tmp__$Sex == "男" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Sex)
      } else if(sum(Data_forest_tmp__$Sex != "男" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Sex)
      }
      metastasis_columns <- c("Smoking_history", "Alcoholic_history", "Lymph_met", "Lung_met", "Brain_met", "Bone_met", "Liver_met")
      
      # 削除対象の列名を取得（EP_option == 1 の行において、"Yes" の数または "Yes" でない数が3未満なら削除）
      cols_to_drop <- metastasis_columns[
        sapply(metastasis_columns, function(col) {
          yes_count <- Data_forest_tmp__[EP_option == 1, sum(get(col) == "Yes", na.rm = TRUE)]
          no_count  <- Data_forest_tmp__[EP_option == 1, sum(get(col) != "Yes", na.rm = TRUE)]
          (yes_count < 3 || no_count < 3)
        })
      ]
      
      # 削除対象の列がある場合、一括削除
      if (length(cols_to_drop) > 0) {
        Data_forest_tmp__[, (cols_to_drop) := NULL]
      }
      if(input$HER2 != "No"){
        if(sum(Data_forest_tmp__$HER2_IHC == "Positive" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-HER2_IHC)
        } else if(sum(Data_forest_tmp__$HER2_IHC != "Positive" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-HER2_IHC)
        }
      }
      if(input$MSI != "No"){
        if(sum(Data_forest_tmp__$MSI_PCR == "Positive" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MSI_PCR)
        } else if(sum(Data_forest_tmp__$MSI_PCR != "Positive" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MSI_PCR)
        }
      }
      if(input$MMR != "No"){
        if(sum(Data_forest_tmp__$MMR_IHC == "dMMR" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MMR_IHC)
        } else if(sum(Data_forest_tmp__$MMR_IHC != "dMMR" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MMR_IHC)
        }
      }
      Data_forest_tmp__$Histology = factor(Data_forest_tmp__$Histology)
      if(length(unique(Data_forest_tmp__$Histology))<2){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Histology)
      } else {
        Data_forest_tmp__$Histology = relevel(Data_forest_tmp__$Histology,
                                              ref=names(sort(table(Data_forest_tmp__$Histology),decreasing = T))[[1]])
      }
      Data_forest_tmp__ = Data_forest_tmp__ %>%
        dplyr::select(-Histology_dummy)
      
      if(length(colnames(Data_forest_tmp__)) > 1){
        Factor_names_tmp__ = colnames(Data_forest_tmp__)
        Factor_names_tmp__ = Factor_names_tmp__[Factor_names_tmp__ != "EP_option"]
        for(i in length(Factor_names_tmp__):1){
          if(all(as.character(Data_forest_tmp__[[Factor_names_tmp__[i]]]) ==
                 as.character(Data_forest_tmp__[[Factor_names_tmp__[i]]][1]))){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Factor_names_tmp__[i])
          }
        }
      }
      Data_forest_tmp__univariant = Data_forest_tmp__
      if(length(colnames(Data_forest_tmp__)) > 2){
        Factor_names_tmp__ = colnames(Data_forest_tmp__)
        Factor_names_tmp__ = Factor_names_tmp__[Factor_names_tmp__ != "EP_option"]
        for(i in 1:(length(Factor_names_tmp__) - 1)){
          for(j in (i+1):length(Factor_names_tmp__)){
            if(Factor_names_tmp__[i] %in% colnames(Data_forest_tmp__) &
               Factor_names_tmp__[j] %in% colnames(Data_forest_tmp__)){
              if(abs(cor(as.numeric(as.factor(Data_forest_tmp__[[Factor_names_tmp__[i]]])),
                         as.numeric(as.factor(Data_forest_tmp__[[Factor_names_tmp__[j]]])))) > 0.95){
                Data_forest_tmp__ = Data_forest_tmp__ %>%
                  dplyr::select(-Factor_names_tmp__[j])
              }
            }
          }
        }
      }
      Data_forest_tmp_7_rec = data.frame(Data_forest_tmp__)
      Data_forest_tmp_7__univariant_rec = data.frame(Data_forest_tmp__univariant)
      Data_forest_tmp_7_rec = Data_forest_tmp_7_rec[,colnames(Data_forest_tmp_7_rec) %in% 
                                              c("Age","Lymph_met","Brain_met","Lung_met","Bone_met","Liver_met","Sex","Histology","PS","Lines","Best_effect","EP_option","Panel",
                                                "HER2_IHC", "MSI_PCR", "MMR_IHC", "Smoking_history", "Alcoholic_history", "time_diagnosis_enroll", "Enroll_date"), drop=FALSE]
      Data_forest_tmp_7__univariant_rec = Data_forest_tmp_7__univariant_rec[,colnames(Data_forest_tmp_7__univariant_rec) %in% 
                                                                      c("Age","Lymph_met","Brain_met","Lung_met","Bone_met","Liver_met","Sex","Histology","PS","Lines","Best_effect","EP_option","Panel",
                                                                        "HER2_IHC", "MSI_PCR", "MMR_IHC", "Smoking_history", "Alcoholic_history", "time_diagnosis_enroll", "Enroll_date"), drop=FALSE]
      output$figure_survival_CGP_7_rec = render_gt({
        if(length(colnames(Data_forest_tmp_7_rec)) > 1){
          Data_forest_tmp_7_rec = data.frame( lapply(Data_forest_tmp_7_rec, as.factor) )
          Data_forest_tmp_7__univariant_rec = data.frame( lapply(Data_forest_tmp_7__univariant_rec, as.factor) )
          if("Histology" %in% colnames(Data_forest_tmp_7_rec)){
            Data_forest_tmp_7_rec$Histology <- relevel(Data_forest_tmp_7_rec$Histology, ref=names(sort(table(Data_forest_tmp_7_rec$Histology),decreasing = T))[[1]]) 
          }
          if("Histology" %in% colnames(Data_forest_tmp_7__univariant_rec)){
            Data_forest_tmp_7__univariant_rec$Histology <- relevel(Data_forest_tmp_7__univariant_rec$Histology, ref=names(sort(table(Data_forest_tmp_7__univariant_rec$Histology),decreasing = T))[[1]]) 
          }
          if("Sex" %in% colnames(Data_forest_tmp_7_rec)){
            Data_forest_tmp_7_rec = Data_forest_tmp_7_rec %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          if("Sex" %in% colnames(Data_forest_tmp_7__univariant_rec)){
            Data_forest_tmp_7__univariant_rec = Data_forest_tmp_7__univariant_rec %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          colnames_tmp = colnames(Data_forest_tmp_7__univariant_rec)
          colnames_tmp[colnames_tmp == "Age"] = paste0("Age ~", input$mid_age-1,"/",input$mid_age, "~")
          colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
          colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
          colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
          colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
          colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
          colnames_tmp[colnames_tmp == "PS"] = "Performance status"
          colnames_tmp[colnames_tmp == "Lines"] = "CTx lines before CGP"
          colnames_tmp[colnames_tmp == "Best_effect"] = "Best CTx effect before CGP"
          colnames(Data_forest_tmp_7__univariant_rec) = colnames_tmp
          if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m3_rec_uni.qs")){
            univ_tab = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m3_rec_uni.qs")
          } else {
            # 並列処理で実行する関数
            run_univariate_model <- function(var, data) {
              fml <- as.formula(paste("EP_option ~", paste("`", var, "`", sep = "")))
              model <- glm(fml, data = data, family = binomial)
              tbl_regression(model, exponentiate = TRUE, conf.int = FALSE) %>%
                add_global_p() %>%
                add_n(location = "level") %>%
                add_nevent(location = "level") |>
                add_q() |>
                bold_p() |>
                # modify_fmt_fun(conf.low ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                # modify_fmt_fun(conf.high ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels() %>%
                tbl_butcher()
            }
            # 並列処理する変数リスト
            univ_vars <- setdiff(names(Data_forest_tmp_7__univariant_rec), "EP_option")
            # 利用可能なコア数を取得
            num_cores <- min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE))
            # 並列処理の実行
            univ_models_parallel <- mclapply(
              univ_vars,
              run_univariate_model,
              data = Data_forest_tmp_7__univariant_rec,
              mc.cores = num_cores
            )
            names(univ_models_parallel) <- univ_vars # 結果に変数名を付与 (mclapply の場合)
            univ_tab <- tbl_stack(univ_models_parallel)
            if(input$intermediate_file != "No"){ 
              QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), univ_tab, file = "source/m3_rec_uni.qs")
            }
          }  
          final_mv_regxlevels <- (prepare_data(as.numeric(as.character(Data_forest_tmp_7_rec$EP_option)),
                                               as.matrix(data.frame(lapply(Data_forest_tmp_7_rec[, setdiff(names(Data_forest_tmp_7_rec), "EP_option")],
                                                                           function(x) as.numeric(as.factor(x))))), type = "logistic") %>% stepwise(aic))$model
          Data_forest_tmp_7_rec = Data_forest_tmp_7_rec %>% select(c(final_mv_regxlevels, "EP_option"))
          colnames_tmp = colnames(Data_forest_tmp_7_rec)
          colnames_tmp[colnames_tmp == "Age"] = paste0("Age ~", input$mid_age-1,"/",input$mid_age, "~")
          colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
          colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
          colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
          colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
          colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
          colnames_tmp[colnames_tmp == "PS"] = "Performance status"
          colnames_tmp[colnames_tmp == "Lines"] = "CTx lines before CGP"
          colnames_tmp[colnames_tmp == "Best_effect"] = "Best CTx effect before CGP"
          colnames(Data_forest_tmp_7_rec) = colnames_tmp
          Data_forest_tmp_7_rec = Data_forest_tmp_7_rec[,colnames(Data_forest_tmp_7_rec) != "CTx lines before CGP"]
          if(length(final_mv_regxlevels) != 0){
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m3_rec.qs")){
              mv_tab = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m3_rec.qs")
            } else {
              mv_tab = glm(EP_option ~.,
                           data = Data_forest_tmp_7_rec,
                           family = binomial(link = "logit")) |>
                tbl_regression(                         ## 多変量解析の表を生成
                  exponentiate = TRUE,                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
                  conf.int = FALSE
                ) |>
                add_global_p() |> # add global p-value
                add_q() |> # adjusts global p-values for multiple testing
                bold_p() |> # bold p-values under a given threshold (default 0.05)
                # modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                # modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels() %>%
                tbl_butcher()
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), mv_tab, file = "source/m3_rec.qs")
              }
            }  
            tbl_merge(
              tbls = list(univ_tab, mv_tab),
              tab_spanner = c("**Univariate**", "**Multivariable**")) |>
              modify_caption("Factors that recommended treatment (Only factors with >2 events observed)") |> as_gt()
          } else {
            univ_tab |>
              modify_caption("Factors that recommended treatment, no significant factor in multivariable analysis (Only factors with >2 events observed)") |> as_gt()
          }
        }
      })

      # Factor_names = colnames(Data_forest)
      # Factor_names = Factor_names[!Factor_names %in% c('ID', 'time_enroll_final', 'censor', 'EP_treat', 'EP_option')]
      
      Data_forest_tmp__ = Data_forest %>%
        dplyr::mutate(EP_option = case_when(
          ID %in% ID_evidence_AB ~ 1,
          TRUE ~ 0
        )) %>%
        dplyr::select(-time_enroll_final, -censor, -EP_treat, -ID)
      Disease_tmp = unique(sort(Data_forest_tmp__$Histology))
      for(i in 1:length(Disease_tmp)){
        Data_forest_tmp_2_ = Data_forest_tmp__ %>% dplyr::filter(Histology == Disease_tmp[i])
        if(sum(Data_forest_tmp_2_$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::mutate(Histology = case_when(
              Histology == Disease_tmp[i] ~ Histology_dummy,
              TRUE ~ Histology
            )
            )
        }
      }
      for(i in length(gene_names):1){
        kk = 14+col_added+i
        if(sum(Data_forest_tmp__[[kk]] == "mut(+)" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__[,-..kk]
        } else if(sum(Data_forest_tmp__[[kk]] != "mut(+)" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__[,-..kk]
        }
      }
      if(sum(Data_forest_tmp__$PS == "0" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-PS)
      } else if(sum(Data_forest_tmp__$PS != "0" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-PS)
      } else if(sum(Data_forest_tmp__$PS == "2_4" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__$PS[Data_forest_tmp__$PS == "1"] = "1_4"
        Data_forest_tmp__$PS[Data_forest_tmp__$PS == "2_4"] = "1_4"
      }
      if(sum(Data_forest_tmp__$time_diagnosis_enroll == ">1-year" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-time_diagnosis_enroll)
      } else if(sum(Data_forest_tmp__$time_diagnosis_enroll != ">1-year" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-time_diagnosis_enroll)
      }
      if((sum(Data_forest_tmp__$Panel == "NCC OncoPanel" & Data_forest_tmp__$EP_option == 1) < 3) |
         (sum(Data_forest_tmp__$Panel == "FoundationOne CDx" & Data_forest_tmp__$EP_option == 1) < 3) |
         (sum(Data_forest_tmp__$Panel == "GenMineTOP" & Data_forest_tmp__$EP_option == 1) < 3)){
        Data_forest_tmp__$Panel[Data_forest_tmp__$Panel == "NCC OncoPanel"] = "Solid"
        Data_forest_tmp__$Panel[Data_forest_tmp__$Panel == "FoundationOne CDx"] = "Solid"
        Data_forest_tmp__$Panel[Data_forest_tmp__$Panel == "GenMineTOP"] = "Solid"
        if(sum(Data_forest_tmp__$Panel == "Solid" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-Panel)
        } else if(sum(Data_forest_tmp__$Panel != "Solid" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-Panel)
        }
      }
      if(sum(Data_forest_tmp__$Enroll_date == "2019" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2019"] = "2019/20"
        Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2020"] = "2019/20"
        if(sum(Data_forest_tmp__$Enroll_date == "2019/20" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
          Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2019/20"] = "2019-21"
          Data_forest_tmp__$Enroll_date[Data_forest_tmp__$Enroll_date == "2021"] = "2019-21"
          if(sum(Data_forest_tmp__$Enroll_date == "2019-21" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Enroll_date)
          } else if(sum(Data_forest_tmp__$Enroll_date != "2019-21" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Enroll_date)
          }
        } else if(sum(Data_forest_tmp__$Enroll_date != "2019/20" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-Enroll_date)
        }
      } else if(sum(Data_forest_tmp__$Enroll_date != "2019" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Enroll_date)
      }
      if(sum(Data_forest_tmp__$Lines == "1" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Lines)
      } else if(sum(Data_forest_tmp__$Lines != "1" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Lines)
      } else if(sum(Data_forest_tmp__$Lines == "0" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "0"] = "0~1"
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "1"] = "0~1"
      } else if(sum(Data_forest_tmp__$Lines == "2~" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "2~"] = "1~"
        Data_forest_tmp__$Lines[Data_forest_tmp__$Lines == "1"] = "1~"
      }
      if(sum(Data_forest_tmp__$Best_effect == "SD" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Best_effect)
      } else if(sum(Data_forest_tmp__$Best_effect != "SD" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Best_effect)
      }
      if(sum(Data_forest_tmp__$Age == "Younger" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Age)
      } else if(sum(Data_forest_tmp__$Age == "Older" & Data_forest_tmp__$EP_option == 1) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Age)
      }
      if(sum(Data_forest_tmp__$Sex == "男" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Sex)
      } else if(sum(Data_forest_tmp__$Sex != "男" & Data_forest_tmp__$EP_option == 1,na.rm = T) < 3){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Sex)
      }
      # if(sum(Data_forest_tmp__$Smoking_history == "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Smoking_history)
      # } else if(sum(Data_forest_tmp__$Smoking_history != "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Smoking_history)
      # }
      # if(sum(Data_forest_tmp__$Alcoholic_history == "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Alcoholic_history)
      # } else if(sum(Data_forest_tmp__$Alcoholic_history != "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Alcoholic_history)
      # }
      # if(sum(Data_forest_tmp__$Lymph_met == "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Lymph_met)
      # } else if(sum(Data_forest_tmp__$Lymph_met != "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Lymph_met)
      # }
      # if(sum(Data_forest_tmp__$Lung_met == "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Lung_met)
      # } else if(sum(Data_forest_tmp__$Lung_met != "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Lung_met)
      # }
      # if(sum(Data_forest_tmp__$Brain_met == "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Brain_met)
      # } else if(sum(Data_forest_tmp__$Brain_met != "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Brain_met)
      # }
      # if(sum(Data_forest_tmp__$Bone_met == "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Bone_met)
      # } else if(sum(Data_forest_tmp__$Bone_met != "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Bone_met)
      # }
      # if(sum(Data_forest_tmp__$Liver_met == "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Liver_met)
      # } else if(sum(Data_forest_tmp__$Liver_met != "Yes" & Data_forest_tmp__$EP_option == 1) < 3){
      #   Data_forest_tmp__ = Data_forest_tmp__ %>%
      #     dplyr::select(-Liver_met)
      # }
      metastasis_columns <- c("Smoking_history", "Alcoholic_history", "Lymph_met", "Lung_met", "Brain_met", "Bone_met", "Liver_met")
      
      # 削除対象の列名を取得（EP_option == 1 の行において、"Yes" の数または "Yes" でない数が3未満なら削除）
      cols_to_drop <- metastasis_columns[
        sapply(metastasis_columns, function(col) {
          yes_count <- Data_forest_tmp__[EP_option == 1, sum(get(col) == "Yes", na.rm = TRUE)]
          no_count  <- Data_forest_tmp__[EP_option == 1, sum(get(col) != "Yes", na.rm = TRUE)]
          (yes_count < 3 || no_count < 3)
        })
      ]
      
      # 削除対象の列がある場合、一括削除
      if (length(cols_to_drop) > 0) {
        Data_forest_tmp__[, (cols_to_drop) := NULL]
      }
      if(input$HER2 != "No"){
        if(sum(Data_forest_tmp__$HER2_IHC == "Positive" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-HER2_IHC)
        } else if(sum(Data_forest_tmp__$HER2_IHC != "Positive" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-HER2_IHC)
        }
      }
      if(input$MSI != "No"){
        if(sum(Data_forest_tmp__$MSI_PCR == "Positive" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MSI_PCR)
        } else if(sum(Data_forest_tmp__$MSI_PCR != "Positive" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MSI_PCR)
        }
      }
      if(input$MMR != "No"){
        if(sum(Data_forest_tmp__$MMR_IHC == "dMMR" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MMR_IHC)
        } else if(sum(Data_forest_tmp__$MMR_IHC != "dMMR" & Data_forest_tmp__$EP_option == 1) < 3){
          Data_forest_tmp__ = Data_forest_tmp__ %>%
            dplyr::select(-MMR_IHC)
        }
      }
      Data_forest_tmp__$Histology = factor(Data_forest_tmp__$Histology)
      if(length(unique(Data_forest_tmp__$Histology))<2){
        Data_forest_tmp__ = Data_forest_tmp__ %>%
          dplyr::select(-Histology)
      } else {
        Data_forest_tmp__$Histology = relevel(Data_forest_tmp__$Histology,
                                              ref=names(sort(table(Data_forest_tmp__$Histology),decreasing = T))[[1]])
      }
      Data_forest_tmp__ = Data_forest_tmp__ %>%
        dplyr::select(-Histology_dummy)
      
      if(length(colnames(Data_forest_tmp__)) > 1){
        Factor_names_tmp__ = colnames(Data_forest_tmp__)
        Factor_names_tmp__ = Factor_names_tmp__[Factor_names_tmp__ != "EP_option"]
        for(i in length(Factor_names_tmp__):1){
          if(all(as.character(Data_forest_tmp__[[Factor_names_tmp__[i]]]) ==
                 as.character(Data_forest_tmp__[[Factor_names_tmp__[i]]][1]))){
            Data_forest_tmp__ = Data_forest_tmp__ %>%
              dplyr::select(-Factor_names_tmp__[i])
          }
        }
      }
      Data_forest_tmp__univariant = Data_forest_tmp__
      if(length(colnames(Data_forest_tmp__)) > 2){
        Factor_names_tmp__ = colnames(Data_forest_tmp__)
        Factor_names_tmp__ = Factor_names_tmp__[Factor_names_tmp__ != "EP_option"]
        for(i in 1:(length(Factor_names_tmp__) - 1)){
          for(j in (i+1):length(Factor_names_tmp__)){
            if(Factor_names_tmp__[i] %in% colnames(Data_forest_tmp__) &
               Factor_names_tmp__[j] %in% colnames(Data_forest_tmp__)){
              if(abs(cor(as.numeric(as.factor(Data_forest_tmp__[[Factor_names_tmp__[i]]])),
                         as.numeric(as.factor(Data_forest_tmp__[[Factor_names_tmp__[j]]])))) > 0.95){
                Data_forest_tmp__ = Data_forest_tmp__ %>%
                  dplyr::select(-Factor_names_tmp__[j])
              }
            }
          }
        }
      }
       
      Data_forest_tmp_7_GMT = data.frame(Data_forest_tmp__)
      Data_forest_tmp_7__univariant_GMT = data.frame(Data_forest_tmp__univariant)
      Data_forest_tmp_7_GMT = Data_forest_tmp_7_GMT[,colnames(Data_forest_tmp_7_GMT) %in% 
                                                      c("Age","Lymph_met","Brain_met","Lung_met","Bone_met","Liver_met","Sex","Histology","PS","Lines","Best_effect","EP_option","Panel",
                                                        "HER2_IHC", "MSI_PCR", "MMR_IHC","Smoking_history","Alcoholic_history","time_diagnosis_enroll", "Enroll_date"), drop=FALSE]
      Data_forest_tmp_7__univariant_GMT = Data_forest_tmp_7__univariant_GMT[,colnames(Data_forest_tmp_7__univariant_GMT) %in% 
                                                                              c("Age","Lymph_met","Brain_met","Lung_met","Bone_met","Liver_met","Sex","Histology","PS","Lines","Best_effect","EP_option","Panel",
                                                                                "HER2_IHC", "MSI_PCR", "MMR_IHC","Alcoholic_history","time_diagnosis_enroll", "Enroll_date"), drop=FALSE]
      output$figure_survival_CGP_7_GMT = render_gt({
        if(length(colnames(Data_forest_tmp_7_GMT)) > 1){
          Data_forest_tmp_7_GMT = data.frame( lapply(Data_forest_tmp_7_GMT, as.factor) )
          Data_forest_tmp_7__univariant_GMT = data.frame( lapply(Data_forest_tmp_7__univariant_GMT, as.factor) )
          if("Histology" %in% colnames(Data_forest_tmp_7_GMT)){
            Data_forest_tmp_7_GMT$Histology <- relevel(Data_forest_tmp_7_GMT$Histology, ref=names(sort(table(Data_forest_tmp_7_GMT$Histology),decreasing = T))[[1]]) 
          }
          if("Histology" %in% colnames(Data_forest_tmp_7__univariant_GMT)){
            Data_forest_tmp_7__univariant_GMT$Histology <- relevel(Data_forest_tmp_7__univariant_GMT$Histology, ref=names(sort(table(Data_forest_tmp_7__univariant_GMT$Histology),decreasing = T))[[1]]) 
          }
          if("Sex" %in% colnames(Data_forest_tmp_7_GMT)){
            Data_forest_tmp_7_GMT = Data_forest_tmp_7_GMT %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          if("Sex" %in% colnames(Data_forest_tmp_7__univariant_GMT)){
            Data_forest_tmp_7__univariant_GMT = Data_forest_tmp_7__univariant_GMT %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          colnames_tmp = colnames(Data_forest_tmp_7__univariant_GMT)
          colnames_tmp[colnames_tmp == "Age"] = paste0("Age ~", input$mid_age-1,"/",input$mid_age, "~")
          colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
          colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
          colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
          colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
          colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
          colnames_tmp[colnames_tmp == "PS"] = "Performance status"
          colnames_tmp[colnames_tmp == "Lines"] = "CTx lines before CGP"
          colnames_tmp[colnames_tmp == "Best_effect"] = "Best CTx effect before CGP"
          colnames(Data_forest_tmp_7__univariant_GMT) = colnames_tmp
          if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m3_GMT_uni.qs")){
            univ_tab = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m3_GMT_uni.qs")
          } else {
            # 並列処理で実行する関数
            run_univariate_model <- function(var, data) {
              fml <- as.formula(paste("EP_option ~", paste("`", var, "`", sep = "")))
              model <- glm(fml, data = data, family = binomial)
              tbl_regression(model, exponentiate = TRUE, conf.int = FALSE) %>%
                add_global_p() %>%
                add_n(location = "level") %>%
                add_nevent(location = "level") |>
                add_q() |>
                bold_p() |>
                # modify_fmt_fun(conf.low ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                # modify_fmt_fun(conf.high ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels() %>%
                tbl_butcher()
            }
            # 並列処理する変数リスト
            univ_vars <- setdiff(names(Data_forest_tmp_7__univariant_GMT), "EP_option")
            # 利用可能なコア数を取得
            num_cores <- min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE))
            # 並列処理の実行
            univ_models_parallel <- mclapply(
              univ_vars,
              run_univariate_model,
              data = Data_forest_tmp_7__univariant_GMT,
              mc.cores = num_cores
            )
            names(univ_models_parallel) <- univ_vars # 結果に変数名を付与 (mclapply の場合)
            univ_tab <- tbl_stack(univ_models_parallel)
            if(input$intermediate_file != "No"){ 
              QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), univ_tab, file = "source/m3_GMT_uni.qs")
            }
          }  
          final_mv_regxlevels <- (prepare_data(as.numeric(as.character(Data_forest_tmp_7_GMT$EP_option)),
                                               as.matrix(data.frame(lapply(Data_forest_tmp_7_GMT[, setdiff(names(Data_forest_tmp_7_GMT), "EP_option")],
                                                                           function(x) as.numeric(as.factor(x))))), type = "logistic") %>% stepwise(aic))$model
          Data_forest_tmp_7_GMT = Data_forest_tmp_7_GMT %>% select(c(final_mv_regxlevels, "EP_option"))
          colnames_tmp = colnames(Data_forest_tmp_7_GMT)
          colnames_tmp[colnames_tmp == "Age"] = paste0("Age ~", input$mid_age-1,"/",input$mid_age, "~")
          colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
          colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
          colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
          colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
          colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
          colnames_tmp[colnames_tmp == "PS"] = "Performance status"
          colnames_tmp[colnames_tmp == "Lines"] = "CTx lines before CGP"
          colnames_tmp[colnames_tmp == "Best_effect"] = "Best CTx effect before CGP"
          colnames(Data_forest_tmp_7_GMT) = colnames_tmp
          Data_forest_tmp_7_GMT = Data_forest_tmp_7_GMT[,colnames(Data_forest_tmp_7_GMT) != "CTx lines before CGP"]
          if(length(final_mv_regxlevels) != 0){
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m3_GMT.qs")){
              mv_tab = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m3_GMT.qs")
            } else {
              mv_tab = glm(EP_option ~.,
                           data = Data_forest_tmp_7_GMT,
                           family = binomial(link = "logit")) |>
                tbl_regression(                         ## 多変量解析の表を生成
                  exponentiate = TRUE,                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
                  conf.int = FALSE
                ) |>
                add_global_p() |> # add global p-value
                add_q() |> # adjusts global p-values for multiple testing
                bold_p() |> # bold p-values under a given threshold (default 0.05)
                # modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                # modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels() %>%
                tbl_butcher()
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), mv_tab, file = "source/m3_GMT.qs")
              }
            }  
            tbl_merge(
              tbls = list(univ_tab, mv_tab),
              tab_spanner = c("**Univariate**", "**Multivariable**")) |>
              modify_caption("Factors for genome-matched therapy (Evidence level A/B) (Only factors with >2 events observed)") |> as_gt()
          } else {
            univ_tab |>
              modify_caption("Factors for genome-matched therapy (Evidence level A/B), no significant factor in multivariable analysis (Only factors with >2 events observed)") |> as_gt()
          }
        }
      })
      
      Data_forest_tmp_ = Data_forest  %>%
        dplyr::filter(EP_option == 1) %>%
        dplyr::select(-time_enroll_final, -censor, -EP_option, -ID)
      Disease_tmp = unique(sort(Data_forest_tmp_$Histology))
      for(i in 1:length(Disease_tmp)){
        Data_forest_tmp_2 = Data_forest_tmp_ %>% dplyr::filter(Histology == Disease_tmp[i])
        if(sum(Data_forest_tmp_2$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_ %>%
            dplyr::mutate(Histology = case_when(
              Histology == Disease_tmp[i] ~ Histology_dummy,
              TRUE ~ Histology
            )
            )
        }
      }
      for(i in length(gene_names):1){
        kk = 14+col_added+i
        if(sum(Data_forest_tmp_[[kk]] == "mut(+)" & Data_forest_tmp_$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_[,-..kk]
        } else if(sum(Data_forest_tmp_[[kk]] != "mut(+)" & Data_forest_tmp_$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_[,-..kk]
        }
      }
      if(sum(Data_forest_tmp_$PS == "0" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-PS)
      } else if(sum(Data_forest_tmp_$PS != "0" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-PS)
      } else if(sum(Data_forest_tmp_$PS == "2_4" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_$PS[Data_forest_tmp_$PS == "1"] = "1_4"
        Data_forest_tmp_$PS[Data_forest_tmp_$PS == "2_4"] = "1_4"
      }
      if(sum(Data_forest_tmp_$Smoking_history == "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Smoking_history)
      } else if(sum(Data_forest_tmp_$Smoking_history != "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Smoking_history)
      }
      if(sum(Data_forest_tmp_$Alcoholic_history == "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Alcoholic_history)
      } else if(sum(Data_forest_tmp_$Alcoholic_history != "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Alcoholic_history)
      }
      if(sum(Data_forest_tmp_$time_diagnosis_enroll == ">1-year" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-time_diagnosis_enroll)
      } else if(sum(Data_forest_tmp_$time_diagnosis_enroll != ">1-year" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-time_diagnosis_enroll)
      }
      if((sum(Data_forest_tmp_$Panel == "NCC OncoPanel" & Data_forest_tmp_$EP_treat == 1) < 3) |
         (sum(Data_forest_tmp_$Panel == "FoundationOne CDx" & Data_forest_tmp_$EP_treat == 1) < 3) |
         (sum(Data_forest_tmp_$Panel == "GenMineTOP" & Data_forest_tmp_$EP_treat == 1) < 3)){
        Data_forest_tmp_$Panel[Data_forest_tmp_$Panel == "NCC OncoPanel"] = "Solid"
        Data_forest_tmp_$Panel[Data_forest_tmp_$Panel == "FoundationOne CDx"] = "Solid"
        Data_forest_tmp_$Panel[Data_forest_tmp_$Panel == "GenMineTOP"] = "Solid"
        if(sum(Data_forest_tmp_$Panel == "Solid" & Data_forest_tmp_$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_ %>%
            dplyr::select(-Panel)
        } else if(sum(Data_forest_tmp_$Panel != "Solid" & Data_forest_tmp_$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_ %>%
            dplyr::select(-Panel)
        }
      }
      if(sum(Data_forest_tmp_$Enroll_date == "2019" & Data_forest_tmp_$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp_$Enroll_date[Data_forest_tmp_$Enroll_date == "2019"] = "2019/20"
        Data_forest_tmp_$Enroll_date[Data_forest_tmp_$Enroll_date == "2020"] = "2019/20"
        if(sum(Data_forest_tmp_$Enroll_date == "2019/20" & Data_forest_tmp_$EP_treat == 1,na.rm = T) < 3){
          Data_forest_tmp_$Enroll_date[Data_forest_tmp_$Enroll_date == "2019/20"] = "2019-21"
          Data_forest_tmp_$Enroll_date[Data_forest_tmp_$Enroll_date == "2021"] = "2019-21"
          if(sum(Data_forest_tmp_$Enroll_date == "2019-21" & Data_forest_tmp_$EP_treat == 1,na.rm = T) < 3){
            Data_forest_tmp_ = Data_forest_tmp_ %>%
              dplyr::select(-Enroll_date)
          } else if(sum(Data_forest_tmp_$Enroll_date != "2019-21" & Data_forest_tmp_$EP_treat == 1,na.rm = T) < 3){
            Data_forest_tmp_ = Data_forest_tmp_ %>%
              dplyr::select(-Enroll_date)
          }
        } else if(sum(Data_forest_tmp_$Enroll_date != "2019/20" & Data_forest_tmp_$EP_treat == 1,na.rm = T) < 3){
          Data_forest_tmp_ = Data_forest_tmp_ %>%
            dplyr::select(-Enroll_date)
        }
      } else if(sum(Data_forest_tmp_$Enroll_date != "2019" & Data_forest_tmp_$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Enroll_date)
      }
      if(sum(Data_forest_tmp_$Age == "Younger" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Age)
      } else if(sum(Data_forest_tmp_$Age == "Older" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Age)
      }
      if(sum(Data_forest_tmp_$Best_effect == "SD" & Data_forest_tmp_$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Best_effect)
      } else if(sum(Data_forest_tmp_$Best_effect != "SD" & Data_forest_tmp_$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(Best_effect)
      }
      if(sum(Data_forest_tmp_$Sex == "男" & Data_forest_tmp_$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Sex)
      } else if(sum(Data_forest_tmp_$Sex != "男" & Data_forest_tmp_$EP_treat == 1,na.rm = T) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Sex)
      }
      if(sum(Data_forest_tmp_$Lines == "1" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Lines)
      } else if(sum(Data_forest_tmp_$Lines != "1" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Lines)
      } else if(sum(Data_forest_tmp_$Lines == "0" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_$Lines[Data_forest_tmp_$Lines == "0"] = "0~1"
        Data_forest_tmp_$Lines[Data_forest_tmp_$Lines == "1"] = "0~1"
      } else if(sum(Data_forest_tmp_$Lines == "2~" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_$Lines[Data_forest_tmp_$Lines == "2~"] = "1~"
        Data_forest_tmp_$Lines[Data_forest_tmp_$Lines == "1"] = "1~"
      } 
      if(sum(Data_forest_tmp_$Lymph_met == "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Lymph_met)
      } else if(sum(Data_forest_tmp_$Lymph_met != "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Lymph_met)
      }
      if(sum(Data_forest_tmp_$Lung_met == "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Lung_met)
      } else if(sum(Data_forest_tmp_$Lung_met != "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Lung_met)
      }
      if(sum(Data_forest_tmp_$Brain_met == "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Brain_met)
      } else if(sum(Data_forest_tmp_$Brain_met != "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Brain_met)
      }
      if(sum(Data_forest_tmp_$Bone_met == "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Bone_met)
      } else if(sum(Data_forest_tmp_$Bone_met != "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Bone_met)
      }
      if(sum(Data_forest_tmp_$Liver_met == "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Liver_met)
      } else if(sum(Data_forest_tmp_$Liver_met != "Yes" & Data_forest_tmp_$EP_treat == 1) < 3){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Liver_met)
      }
      if(input$HER2 != "No"){
        if(sum(Data_forest_tmp_$HER2_IHC == "Positive" & Data_forest_tmp_$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_ %>%
            dplyr::select(-HER2_IHC)
        } else if(sum(Data_forest_tmp_$HER2_IHC != "Positive" & Data_forest_tmp_$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_ %>%
            dplyr::select(-HER2_IHC)
        }
      }
      if(input$MSI != "No"){
        if(sum(Data_forest_tmp_$MSI_PCR == "Positive" & Data_forest_tmp_$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_ %>%
            dplyr::select(-MSI_PCR)
        } else if(sum(Data_forest_tmp_$MSI_PCR != "Positive" & Data_forest_tmp_$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_ %>%
            dplyr::select(-MSI_PCR)
        }
      }
      if(input$MMR != "No"){
        if(sum(Data_forest_tmp_$MMR_IHC == "dMMR" & Data_forest_tmp_$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_ %>%
            dplyr::select(-MMR_IHC)
        } else if(sum(Data_forest_tmp_$MMR_IHC != "dMMR" & Data_forest_tmp_$EP_treat == 1) < 3){
          Data_forest_tmp_ = Data_forest_tmp_ %>%
            dplyr::select(-MMR_IHC)
        }
      }
      Data_forest_tmp_$Histology = factor(Data_forest_tmp_$Histology)
      if(length(unique(Data_forest_tmp_$Histology))<2){
        Data_forest_tmp_ = Data_forest_tmp_ %>%
          dplyr::select(-Histology)
      } else {
        Data_forest_tmp_$Histology = relevel(Data_forest_tmp_$Histology,
                                              ref=names(sort(table(Data_forest_tmp_$Histology),decreasing = T))[[1]])
      }
      Data_forest_tmp_ = Data_forest_tmp_ %>%
        dplyr::select(-Histology_dummy)
      if(length(colnames(Data_forest_tmp_)) > 1){
        Factor_names_tmp_ = colnames(Data_forest_tmp_)
        Factor_names_tmp_ = Factor_names_tmp_[Factor_names_tmp_ != "EP_treat"]
        for(i in length(Factor_names_tmp_):1){
          if(all(as.character(Data_forest_tmp_[[Factor_names_tmp_[i]]]) ==
                 as.character(Data_forest_tmp_[[Factor_names_tmp_[i]]][1]))){
            Data_forest_tmp_ = Data_forest_tmp_ %>%
              dplyr::select(-Factor_names_tmp_[i])
          }
        }
      }
      
      Data_forest_tmp_univariant = Data_forest_tmp_
      
      if(length(colnames(Data_forest_tmp_)) > 2){
        Factor_names_tmp_ = colnames(Data_forest_tmp_)
        Factor_names_tmp_ = Factor_names_tmp_[Factor_names_tmp_ != "EP_treat"]
        for(i in 1:(length(Factor_names_tmp_) - 1)){
          for(j in (i+1):length(Factor_names_tmp_)){
            if(Factor_names_tmp_[i] %in% colnames(Data_forest_tmp_) &
               Factor_names_tmp_[j] %in% colnames(Data_forest_tmp_)){
              if(abs(cor(as.numeric(as.factor(Data_forest_tmp_[[Factor_names_tmp_[i]]])),
                         as.numeric(as.factor(Data_forest_tmp_[[Factor_names_tmp_[j]]])))) > 0.95){
                Data_forest_tmp_ = Data_forest_tmp_ %>%
                  dplyr::select(-Factor_names_tmp_[j])
              }
            }
          }
        }
      }
      incProgress(1 / 13)
      
      Data_forest_tmp_6 = Data_forest_tmp_
      Data_forest_tmp_6_univariant = Data_forest_tmp_univariant
      output$figure_survival_CGP_6 = render_gt({
        if(length(colnames(Data_forest_tmp_6)) > 1){
          Data_forest_tmp_6 = data.frame( lapply(Data_forest_tmp_6, as.factor) )
          Data_forest_tmp_6_univariant = data.frame( lapply(Data_forest_tmp_6_univariant, as.factor) )
          if("Histology" %in% colnames(Data_forest_tmp_6)){
            Data_forest_tmp_6$Histology <- relevel(Data_forest_tmp_6$Histology, ref=names(sort(table(Data_forest_tmp_6$Histology),decreasing = T))[[1]]) 
          }
          if("Histology" %in% colnames(Data_forest_tmp_6_univariant)){
            Data_forest_tmp_6_univariant$Histology <- relevel(Data_forest_tmp_6_univariant$Histology, ref=names(sort(table(Data_forest_tmp_6_univariant$Histology),decreasing = T))[[1]]) 
          }
          if("Sex" %in% colnames(Data_forest_tmp_6)){
            Data_forest_tmp_6 = Data_forest_tmp_6 %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          if("Sex" %in% colnames(Data_forest_tmp_6_univariant)){
            Data_forest_tmp_6_univariant = Data_forest_tmp_6_univariant %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          colnames_tmp = colnames(Data_forest_tmp_6_univariant)
          colnames_tmp[colnames_tmp == "Age"] = paste0("Age ~", input$mid_age-1,"/",input$mid_age, "~")
          colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
          colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
          colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
          colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
          colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
          colnames_tmp[colnames_tmp == "PS"] = "Performance status"
          colnames_tmp[colnames_tmp == "Lines"] = "CTx lines before CGP"
          colnames_tmp[colnames_tmp == "Best_effect"] = "Best CTx effect before CGP"
          colnames(Data_forest_tmp_6_univariant) = colnames_tmp
          if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m3_post_uni.qs")){
            univ_tab = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m3_post_uni.qs")
          } else {
            # 並列処理で実行する関数
            run_univariate_model <- function(var, data) {
              fml <- as.formula(paste("EP_treat ~", paste("`", var, "`", sep = "")))
              model <- glm(fml, data = data, family = binomial)
              tbl_regression(model, exponentiate = TRUE, conf.int = FALSE) %>%
                add_global_p() %>%
                add_n(location = "level") %>%
                add_nevent(location = "level") |>
                add_q() |>
                bold_p() |>
                # modify_fmt_fun(conf.low ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                # modify_fmt_fun(conf.high ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(abs(x) > 100 | abs(x) < 0.01, format(x, digits = 3, scientific = TRUE), ifelse(abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels() %>%
                tbl_butcher()
            }
            # 並列処理する変数リスト
            univ_vars <- setdiff(names(Data_forest_tmp_6_univariant), "EP_treat")
            # 利用可能なコア数を取得
            num_cores <- min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE))
            # 並列処理の実行
            univ_models_parallel <- mclapply(
              univ_vars,
              run_univariate_model,
              data = Data_forest_tmp_6_univariant,
              mc.cores = num_cores
            )
            names(univ_models_parallel) <- univ_vars # 結果に変数名を付与 (mclapply の場合)
            univ_tab <- tbl_stack(univ_models_parallel)
            if(input$intermediate_file != "No"){ 
              QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), univ_tab, file = "source/m3_post_uni.qs")
            }
          }  
          final_mv_regxlevels <- (prepare_data(as.numeric(as.character(Data_forest_tmp_6$EP_treat)),
                                               as.matrix(data.frame(lapply(Data_forest_tmp_6[, setdiff(names(Data_forest_tmp_6), "EP_treat")],
                                                                           function(x) as.numeric(as.factor(x))))), type = "logistic") %>% stepwise(aic))$model
          Data_forest_tmp_6 = Data_forest_tmp_6 %>% select(c(final_mv_regxlevels, "EP_treat"))
          colnames_tmp = colnames(Data_forest_tmp_6)
          colnames_tmp[colnames_tmp == "Age"] = paste0("Age ~", input$mid_age-1,"/",input$mid_age, "~")
          colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
          colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
          colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
          colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
          colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
          colnames_tmp[colnames_tmp == "PS"] = "Performance status"
          colnames_tmp[colnames_tmp == "Lines"] = "CTx lines before CGP"
          colnames_tmp[colnames_tmp == "Best_effect"] = "Best CTx effect before CGP"
          colnames(Data_forest_tmp_6) = colnames_tmp
          Data_forest_tmp_6 = Data_forest_tmp_6[,colnames(Data_forest_tmp_6) != "CTx lines before CGP"]
          if(length(final_mv_regxlevels) != 0){
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m3_post.qs")){
              mv_tab = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m3_post.qs")
            } else {
              mv_tab = glm(EP_treat ~.,
                           data = Data_forest_tmp_6,
                           family = binomial(link = "logit")) |>
                tbl_regression(                         ## 多変量解析の表を生成
                  exponentiate = TRUE,                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
                  conf.int = FALSE
                ) |>
                add_global_p() |> # add global p-value
                add_q() |> # adjusts global p-values for multiple testing
                bold_p() |> # bold p-values under a given threshold (default 0.05)
                # modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                # modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels() %>%
                tbl_butcher()
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), mv_tab, file = "source/m3_post.qs")
              }
            }  
            tbl_merge(
              tbls = list(univ_tab, mv_tab),
              tab_spanner = c("**Univariate**", "**Multivariable**")) |>
              modify_caption("Factors that led patients with recommended treatment to receive the recommended treatment (Only factors with >2 events observed)") |> as_gt()
          } else {
            univ_tab |>
              modify_caption("Factors that led patients with recommended treatment to receive the recommended treatment, no significant factor in multivariable analysis (Only factors with >2 events observed)") |> as_gt()
          }
          
        }
      })
      
      tmp_post <- reactiveValues()
      output$figure_nomogram_treat_post_CGP = renderPlot({
        if(length(colnames(Data_forest_tmp_6)) > 1){
          Data_forest_tmp_6 = data.frame( lapply(Data_forest_tmp_6, as.factor) )
          if("Histology" %in% colnames(Data_forest_tmp_6)){
            Data_forest_tmp_6$Histology <- relevel(Data_forest_tmp_6$Histology, ref=names(sort(table(Data_forest_tmp_6$Histology),decreasing = T))[[1]]) 
          }
          if("Sex" %in% colnames(Data_forest_tmp_6)){
            Data_forest_tmp_6 = Data_forest_tmp_6 %>%
              dplyr::mutate(Sex=case_when(
                Sex == "女" ~ "Female",
                TRUE ~ "Male"
              ))
          }
          Data_forest_tmp_6 = Data_forest_tmp_6[,colnames(Data_forest_tmp_6) != "Lines"]
          final_mv_regxlevels <- (prepare_data(as.numeric(as.character(Data_forest_tmp_6$EP_treat)),
                                                       as.matrix(data.frame(lapply(Data_forest_tmp_6[, setdiff(names(Data_forest_tmp_6), "EP_treat")],
                                                                                   function(x) as.numeric(as.factor(x))))), type = "logistic") %>% stepwise(aic))$model
          if(length(final_mv_regxlevels) != 0){
            Data_forest_tmp_treat = Data_forest_tmp_6 %>% dplyr::select(c("EP_treat", final_mv_regxlevels))
          # m1 <- glm(EP_treat ~., data = Data_forest_tmp_6, family = binomial(link = "logit"))
          # final_mv_reg <- m1 %>%
          #   stats::step(direction = "backward", trace = FALSE)
          # if(!is.null(final_mv_reg$xlevels)){
          #   Data_forest_tmp_treat = Data_forest_tmp_6 %>% dplyr::select(c("EP_treat", names(final_mv_reg$xlevels)))
            dd <- datadist(Data_forest_tmp_treat) 
            options(datadist=dd) 
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m2_post.qs")){
              m2 = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m2_post.qs")
            } else {
              m2 <- lrm(formula(paste0("EP_treat ~ ",paste(colnames(Data_forest_tmp_treat)[colnames(Data_forest_tmp_treat)!="EP_treat"], collapse="+"))),
                        data = Data_forest_tmp_treat,x=TRUE,y=TRUE,penalty = Penal)
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), m2, file = "source/m2_post.qs")
              }
            }
            BootNo = min(500, max(25, as.integer(50000 / nrow(Data_forest_tmp_treat))*100))
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/val_post.qs")){
              val = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/val_post.qs")
            } else {
              # error in rms package
              val <- try(rms::validate(m2, B=BootNo), silent = FALSE)
              if (class(val) == "try-error") {
                val = rms::validate(m2, B=BootNo)
              }
              # do not delete
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), val, file = "source/val_post.qs")
              }
            }
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/cal_post.qs")){
              cal = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/cal_post.qs")
            } else {
              cal = rms::calibrate(m2, B=BootNo)
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), cal, file = "source/cal_post.qs")
              }
            }
            log_nomogram <- nomogram(m2, fun = plogis, lp=F, funlabel="Matched therapy")
            options(datadist=NULL) 
            par(mfcol = c(2, 1))
            plot(cal, lwd=2, lty=1, 
                 cex.lab=1.2, cex.axis=1, cex.main=1.2, cex.sub=0.6, 
                 xlim=c(0, 1), ylim= c(0, 1), 
                 xlab="Nomogram-Predicted Probability of recommended treatment receive", 
                 ylab="Actual recommended treatment receive (proportion)", 
                 legend=FALSE)
            
            #lines(cal[, c(1:3)], type ="o", lwd=1, pch=16, col=c("#00468BFF"))
            text(x=.4, y=.0, adj = c(0,0), paste0("Matched therapy receiving ratio, only patients with CTx recommendation, post CGP information","\n",
                                                  "Model lokelihood ratio test P-value: ", format(m2$stats["P"][[1]], nsmall=3),'\n',
                                                  BootNo, "-time-bootstrapped-Brier score: ", round(val[9,5], digits=3),'\n',
                                                  BootNo, "-time-bootstrapped-Nagelkerke R2 = ", round(val[2,5], digits=3),'\n',
                                                  BootNo, "-time-bootstrapped-concordance index = ", round(((val[1,5] + 1) / 2), digits=3),"\n",
                                                  "Intercept = ",round(val[3,5], digits=3),"\n",
                                                  "Slope = ", round(val[4,5], digits=3)))
            
            abline(0, 1, lty=3, lwd=2,  col=c("#224444")) 
            legend(x=.8, y=.2, legend=c("Apparent", "Bias-corrected", "Ideal"), 
                   lty=c(3, 1, 2), bty="n")
            plot(log_nomogram)
          } else {
            gg_empty()
          }
          tmp_post$log_nomogram = log_nomogram
        }
      })
      
      output$histology_nomogram_post_CGP = renderText({
        if(!is.null(tmp_post$log_nomogram$Histology$points)){
          paste0("Histology and points\n",            
          paste(paste(tmp_post$log_nomogram$Histology$Histology, round(tmp_post$log_nomogram$Histology$points,digits = 0), sep="\t"), collapse = "\n"))
        } else {
          ""
        }
      })
      
      Data_forest_tmp_8 = Data_forest_tmp_7
      if("Lines" %in% colnames(Data_forest_tmp_8)){
        Data_forest_tmp_8 = Data_forest_tmp_8 %>% select(-Lines)
      }
      if("Sex" %in% colnames(Data_forest_tmp_8)){
        Data_forest_tmp_8 = Data_forest_tmp_8 %>%
          dplyr::mutate(Sex=case_when(
            Sex == "女" ~ "Female",
            TRUE ~ "Male"
          ))
      }
      Data_ML_full = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      Lymph_met,
                      Brain_met,
                      Lung_met,
                      Bone_met,
                      Liver_met,
                      #Other_met,
                      EP_option,
                      EP_treat,
                      症例.基本情報.性別.名称.,
                      症例.基本情報.がん種.OncoTree.,
                      症例.基本情報.がん種.OncoTree.LEVEL1.,
                      症例.背景情報.ECOG.PS.名称.,
                      症例.背景情報.喫煙歴有無.名称.,
                      症例.背景情報.アルコール多飲有無.名称.,
                      症例.管理情報.登録日,
                      time_enroll_final,
                      time_diagnosis_enroll,
                      censor,
                      CTx_lines_before_CGP,
                      RECIST,
                      症例.検体情報.パネル.名称.,
                      YoungOld)
      colnames(Data_ML_full) = 
        c('ID',
          'Lymph_met',
          'Brain_met',
          'Lung_met',
          'Bone_met',
          'Liver_met',
          #'Other_met',
          'EP_option',
          'EP_treat',
          'Sex', 'Histology', 'Histology_dummy', 'PS',
          'Smoking_history', 'Alcoholic_history',
          'Enroll_date',
          'time_enroll_final',
          'time_diagnosis_enroll',
          'censor', "Lines", "Best_effect", "Panel","Age")
      if(input$HER2 != "No"){
        Data_ML_full$HER2_IHC = Data_case_target$HER2_IHC
        Data_ML_full = Data_ML_full %>%
          dplyr::filter(
            HER2_IHC != "Unknown"
          )
      }
      if(input$MSI != "No"){
        Data_ML_full$MSI_PCR = Data_case_target$MSI_PCR
        Data_ML_full = Data_ML_full %>%
          dplyr::filter(
            MSI_PCR != "Unknown"
          )
      }
      if(input$MMR != "No"){
        Data_ML_full$MMR_IHC = Data_case_target$MMR_IHC
        Data_ML_full = Data_ML_full %>%
          dplyr::filter(
            MMR_IHC != "Unknown"
          )
      }
      Data_ML_full$Enroll_date = as.numeric(Data_ML_full$Enroll_date)
      Data_ML_full$time_diagnosis_enroll = ifelse(Data_ML_full$time_diagnosis_enroll > 365.25*1, ">1-year", "<=1-year")
      Data_ML_full$Enroll_date <- format(as.Date(Data_ML_full$Enroll_date), "%Y")
      Data_ML_full = Data_ML_full %>%
        dplyr::filter(PS != "不明" & Sex != "未入力・不明" & Smoking_history != "Unknown" & Alcoholic_history != "Unknown" & !is.na(time_diagnosis_enroll) & !is.na(Enroll_date)) %>%
        dplyr::mutate(
          PS = case_when(
            PS == 0 ~ "0",
            PS == 1 ~ "1",
            TRUE ~ "2_4"
          ),
          Panel = case_when(
            Panel %in% c("NCC OncoPanel", "FoundationOne CDx", "GenMineTOP") ~ Panel,
            TRUE ~ "Liquid"
          ),
          Lines = case_when(
            Lines == "0" ~ "0",
            Lines == "1" ~ "1",
            TRUE ~ "2~"
          )
        )
      if("Sex" %in% colnames(Data_ML_full)){
        Data_ML_full = Data_ML_full %>%
          dplyr::mutate(Sex=case_when(
            Sex == "女" ~ "Female",
            TRUE ~ "Male"
          ))
      }
      frequent_organs = sort(table(Data_ML_full$Histology),decreasing = T)
      frequent_organs_names = names(frequent_organs[frequent_organs>=50])
      Data_ML_full = Data_ML_full %>% dplyr::mutate(
        Histology = case_when(
          Histology %in% frequent_organs_names ~ Histology,
          TRUE ~ Histology_dummy
        )
      )
      Data_ML_full$Panel = as.factor(Data_ML_full$Panel)
      Data_ML_full$Best_effect[Data_ML_full$Best_effect == "Unknown"] = "NE"
      Data_ML_full$Best_effect = factor(Data_ML_full$Best_effect, levels = c("SD","CR","PR","PD","NE","CTx_naive"))
      if(length(unique(Data_ML_full$Panel))>1){
        Data_ML_full$Panel =  relevel(Data_ML_full$Panel, ref=names(sort(table(Data_ML_full$Panel),decreasing = T))[[1]])
      }
      Data_ML_full$Histology = as.factor(Data_ML_full$Histology)
      Data_ML_full$Histology =  relevel(Data_ML_full$Histology, ref=names(sort(table(Data_ML_full$Histology),decreasing = T))[[1]])
      Data_ML_full = Data_ML_full %>%
        dplyr::select(-time_enroll_final, -censor, -Histology_dummy)

      Data_ML = Data_ML_full %>%
        dplyr::select(-EP_option, -ID)
      if("Histology" %in% colnames(Data_forest_tmp_8)){
        Organ_data = data.frame(Data_ML$Histology, Data_forest_tmp_8$Histology) %>% dplyr::distinct()
      } else {
        Organ_data = data.frame(Data_ML$Histology, Data_ML$Histology) %>% dplyr::distinct()
      }
      colnames(Organ_data) = c("Organ_type_raw", "Organ_type")
      save(file = "source/Organ_data.rda", Organ_data)
      if(length(colnames(Data_forest_tmp_8)) > 1){
        Data_forest_tmp_8 = data.frame( lapply(Data_forest_tmp_8, as.factor) )
        if("Histology" %in% colnames(Data_forest_tmp_8)){
          Data_forest_tmp_8$Histology <- relevel(Data_forest_tmp_8$Histology, ref=names(sort(table(Data_forest_tmp_8$Histology),decreasing = T))[[1]]) 
        }
        final_mv_regxlevels <- (prepare_data(as.numeric(as.character(Data_forest_tmp_8$EP_treat)),
                             as.matrix(data.frame(lapply(Data_forest_tmp_8[, setdiff(names(Data_forest_tmp_8), "EP_treat")],
                                                         function(x) as.numeric(as.factor(x))))), type = "logistic") %>% stepwise(aic))$model
        # m1 <- glm(EP_treat ~., data = Data_forest_tmp_8, family = binomial(link = "logit"))
        # final_mv_reg <- m1 %>%
        #   stats::step(direction = "backward", trace = FALSE)
        # if(!is.null(final_mv_reg$xlevels)){
        # Data_forest_tmp_treat = Data_forest_tmp_8 %>% dplyr::select(c("EP_treat", names(final_mv_reg$xlevels)))
        if(length(final_mv_regxlevels) != 0){
          Data_forest_tmp_treat = Data_forest_tmp_8 %>% dplyr::select(c("EP_treat", final_mv_regxlevels))
          if(length(colnames(Data_forest_tmp_treat)) > 1){
            dd <- datadist(Data_forest_tmp_treat) 
            options(datadist=dd)
            formula_treat = formula(paste0("EP_treat ~ `",paste(colnames(Data_forest_tmp_treat)[colnames(Data_forest_tmp_treat)!="EP_treat"], collapse="` + `"),"`"))
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m2.qs")){
              m2 = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m2.qs")
            } else {
              m2 <- lrm(formula_treat,
                        data = Data_forest_tmp_treat,x=TRUE,y=TRUE,penalty = Penal, tol=Toler)
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), m2, file = "source/m2.qs")
              }
            }
            tmp_post$m2 = m2
            BootNo = min(500, max(25, as.integer(50000 / nrow(Data_forest_tmp_treat))*100))
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/val.qs")){
              val = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/val.qs")
            } else {
              # error in rms package
              val <- try(rms::validate(m2, B=BootNo), silent = FALSE)
              if (class(val) == "try-error") {
                val = rms::validate(m2, B=BootNo)
              }
              # do not delete
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), val, file = "source/val.qs")
              }
            }
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/cal.qs")){
              cal = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/cal.qs")
            } else {
              cal = rms::calibrate(m2, B=BootNo)
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), cal, file = "source/cal.qs")
              }
            }
            options(datadist=NULL)
            log_nomogram <- nomogram(m2, fun = plogis, lp=F, funlabel="Matched therapy")
            m2_pvalue = m2$stats["P"][[1]]
            m2_val_1 = val[9,5]
            m2_val_2 = val[2,5]
            m2_val_3 = val[1,5]
            m2_val_4 = val[3,5]
            m2_val_5 = val[4,5]
            rm(m2)
            rm(val)
          }
        }
      }
      output$figure_nomogram_treat_pre_CGP = renderPlot({
        if(length(colnames(Data_forest_tmp_8)) > 1){
          if(length(final_mv_regxlevels) != 0){
          # if(!is.null(final_mv_reg$xlevels)){
            if(length(colnames(Data_forest_tmp_treat)) > 1){
              par(mfcol = c(2, 1))
              plot(cal, lwd=2, lty=1, 
                   cex.lab=1.2, cex.axis=1, cex.main=1.2, cex.sub=0.6, 
                   xlim=c(0, 1), ylim= c(0, 1), 
                   xlab="Nomogram-Predicted Probability of recommended treatment receive", 
                   ylab="Actual recommended treatment receive (proportion)", 
                   legend=FALSE)
              abline(0, 1, lty=3, lwd=2,  col=c("#224444")) 
              text(x=.4, y=.0, adj = c(0,0), paste0("Matched therapy receiving ratio, pre CGP information (other than mutation)","\n",
                                                    "Model lokelihood ratio test P-value: ", format(m2_pvalue, nsmall=3),'\n',
                                                    BootNo, "-time-bootstrapped-Brier score: ", round(m2_val_1, digits=3),'\n',
                                                    BootNo, "-time-bootstrapped-Nagelkerke R2 = ", round(m2_val_2, digits=3),'\n',
                                                    BootNo, "-time-bootstrapped-concordance index = ", round(((m2_val_3 + 1) / 2), digits=3),"\n",
                                                    "Intercept = ",round(m2_val_4, digits=3),"\n",
                                                    "Slope = ", round(m2_val_5, digits=3)))
              legend(x=.8, y=.2, legend=c("Apparent", "Bias-corrected", "Ideal"), 
                     lty=c(3, 1, 2), bty="n")
              plot(log_nomogram)
            } else {
              gg_empty()
            }
          } else {
            gg_empty()
          }
        } else {
          gg_empty()
        }
      })
      output$histology_nomogram_pre_CGP = renderText({
        if(length(colnames(Data_forest_tmp_8)) > 1){
          # if(!is.null(final_mv_reg$xlevels)){
          if(length(final_mv_regxlevels) != 0){
            if(length(colnames(Data_forest_tmp_treat)) > 1){
              if(!is.null(log_nomogram$Histology$points)){
                paste0("Histology and points\n",            
                      paste(paste(log_nomogram$Histology$Histology, round(log_nomogram$Histology$points,digits = 0), sep="\t"), collapse = "\n"))
              } else {
                ""
              }
            }
          }
        }
      })
      Data_ML = Data_ML_full %>%
        dplyr::select(-EP_treat, -ID)
      Data_forest_tmp_8_rec = Data_forest_tmp_7_rec
      if("Lines" %in% colnames(Data_forest_tmp_8_rec)){
        Data_forest_tmp_8_rec = Data_forest_tmp_8_rec %>% select(-Lines)
      }
      if("Sex" %in% colnames(Data_forest_tmp_8_rec)){
        Data_forest_tmp_8_rec = Data_forest_tmp_8_rec %>%
          dplyr::mutate(Sex=case_when(
            Sex == "女" ~ "Female",
            TRUE ~ "Male"
          ))
      }
      if("Histology" %in% colnames(Data_forest_tmp_8_rec)){
        Organ_data_rec = data.frame(Data_ML$Histology, Data_forest_tmp_8_rec$Histology) %>% dplyr::distinct()
      } else {
        Organ_data_rec = data.frame(Data_ML$Histology, Data_ML$Histology) %>% dplyr::distinct()
      }
      colnames(Organ_data_rec) = c("Organ_type_raw", "Organ_type")
      save(file = "source/Organ_data_rec.rda", Organ_data_rec)
      if(length(colnames(Data_forest_tmp_8_rec)) > 1){
        Data_forest_tmp_8_rec = data.frame( lapply(Data_forest_tmp_8_rec, as.factor) )
        if("Histology" %in% colnames(Data_forest_tmp_8_rec)){
          Data_forest_tmp_8_rec$Histology <- relevel(Data_forest_tmp_8_rec$Histology, ref=names(sort(table(Data_forest_tmp_8_rec$Histology),decreasing = T))[[1]]) 
        }
        final_mv_regxlevels_rec <- (prepare_data(as.numeric(as.character(Data_forest_tmp_8_rec$EP_option)),
                                                     as.matrix(data.frame(lapply(Data_forest_tmp_8_rec[, setdiff(names(Data_forest_tmp_8_rec), "EP_option")],
                                                                                 function(x) as.numeric(as.factor(x))))), type = "logistic") %>% stepwise(aic))$model
        if(length(final_mv_regxlevels_rec) != 0){
          Data_forest_tmp_treat_rec = Data_forest_tmp_8_rec %>% dplyr::select(c("EP_option", final_mv_regxlevels_rec))
        # m1_rec <- glm(EP_option ~., data = Data_forest_tmp_8_rec, family = binomial(link = "logit"))
        # final_mv_reg_rec <- m1_rec %>%
        #   stats::step(direction = "backward", trace = FALSE)
        # if(!is.null(final_mv_reg_rec$xlevels)){
        #   Data_forest_tmp_treat_rec = Data_forest_tmp_8_rec %>% dplyr::select(c("EP_option", names(final_mv_reg_rec$xlevels)))
          if(length(colnames(Data_forest_tmp_treat_rec)) > 1){
            dd <- datadist(Data_forest_tmp_treat_rec) 
            options(datadist=dd)
            formula_treat_rec = formula(paste0("EP_option ~ `",paste(colnames(Data_forest_tmp_treat_rec)[colnames(Data_forest_tmp_treat_rec)!="EP_option"], collapse="` + `"),"`"))
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m2_rec.qs")){
              m2_rec = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m2_rec.qs")
            } else {
              m2_rec <- lrm(formula_treat_rec,
                        data = Data_forest_tmp_treat_rec,x=TRUE,y=TRUE,penalty = Penal, tol=Toler)
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), m2_rec, file = "source/m2_rec.qs")
              }
            }
            BootNo = min(500, max(25, as.integer(50000 / nrow(Data_forest_tmp_treat_rec))*100))
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/val_rec.qs")){
              val_rec = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/val_rec.qs")
            } else {
              # error in rms package
              val_rec <- try(rms::validate(m2_rec, B=BootNo), silent = FALSE)
              if (class(val_rec) == "try-error") {
                val_rec = rms::validate(m2_rec, B=BootNo)
              }
              # do not delete
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), val_rec, file = "source/val_rec.qs")
              }
            }
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/cal_rec.qs")){
              cal_rec = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/cal_rec.qs")
            } else {
              cal_rec = rms::calibrate(m2_rec, B=BootNo)
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), cal_rec, file = "source/cal_rec.qs")
              }
            }
            options(datadist=NULL)
            log_nomogram_rec <- nomogram(m2_rec, fun = plogis, lp=F, funlabel="Treatment recommendation")
            m2_rec_pvalue = m2_rec$stats["P"][[1]]
            m2_rec_val_1 = val_rec[9,5]
            m2_rec_val_2 = val_rec[2,5]
            m2_rec_val_3 = val_rec[1,5]
            m2_rec_val_4 = val_rec[3,5]
            m2_rec_val_5 = val_rec[4,5]
            rm(m2_rec)
            rm(val_rec)
          }
        }
      }
      output$figure_nomogram_treat_pre_CGP_rec = renderPlot({
        if(length(colnames(Data_forest_tmp_8_rec)) > 1){
          if(length(final_mv_regxlevels_rec) != 0){
          # if(!is.null(final_mv_reg_rec$xlevels)){
            if(length(colnames(Data_forest_tmp_treat_rec)) > 1){
              par(mfcol = c(2, 1))
              plot(cal_rec, lwd=2, lty=1, 
                   cex.lab=1.2, cex.axis=1, cex.main=1.2, cex.sub=0.6, 
                   xlim=c(0, 1), ylim= c(0, 1), 
                   xlab="Nomogram-Predicted Probability of treatment recommendation", 
                   ylab="Actual treatment recommendation (proportion)", 
                   legend=FALSE)
              
              abline(0, 1, lty=3, lwd=2,  col=c("#224444")) 
              text(x=.4, y=.0, adj = c(0,0), paste0("Matched therapy recommendation ratio, pre CGP information (other than mutation)","\n",
                                                    "Model lokelihood ratio test P-value: ", format(m2_rec_pvalue, nsmall=3),'\n',
                                                    BootNo, "-time-bootstrapped-Brier score: ", round(1, digits=3),'\n',
                                                    BootNo, "-time-bootstrapped-Nagelkerke R2 = ", round(m2_rec_val_2, digits=3),'\n',
                                                    BootNo, "-time-bootstrapped-concordance index = ", round(((m2_rec_val_3 + 1) / 2), digits=3),"\n",
                                                    "Intercept = ",round(m2_rec_val_4, digits=3),"\n",
                                                    "Slope = ", round(m2_rec_val_5, digits=3)))
              legend(x=.8, y=.2, legend=c("Apparent", "Bias-corrected", "Ideal"), 
                     lty=c(3, 1, 2), bty="n")
              plot(log_nomogram_rec)
            } else {
              gg_empty()
            }
          } else {
            gg_empty()
          }
        } else {
          gg_empty()
        }
      })
      output$histology_nomogram_pre_CGP_rec = renderText({
        if(length(colnames(Data_forest_tmp_8_rec)) > 1){
          if(length(final_mv_regxlevels_rec) != 0){
          # if(!is.null(final_mv_reg_rec$xlevels)){
            if(length(colnames(Data_forest_tmp_treat_rec)) > 1){
              if(!is.null(log_nomogram_rec$Histology$points)){
                paste0("Histology and points\n",            
                       paste(paste(log_nomogram_rec$Histology$Histology, round(log_nomogram_rec$Histology$points,digits = 0), sep="\t"), collapse = "\n"))
              } else {
                ""
              }
            }
          }
        }
      })
      
      Data_forest_tmp_8_GMT = Data_forest_tmp_7_GMT
      if("Lines" %in% colnames(Data_forest_tmp_8_GMT)){
        Data_forest_tmp_8_GMT = Data_forest_tmp_8_GMT %>% select(-Lines)
      }
      if("Sex" %in% colnames(Data_forest_tmp_8_GMT)){
        Data_forest_tmp_8_GMT = Data_forest_tmp_8_GMT %>%
          dplyr::mutate(Sex=case_when(
            Sex == "女" ~ "Female",
            TRUE ~ "Male"
          ))
      }
      if("Histology" %in% colnames(Data_forest_tmp_8_GMT)){
        Organ_data_GMT = data.frame(Data_ML$Histology, Data_forest_tmp_8_GMT$Histology) %>% dplyr::distinct()
      } else {
        Organ_data_GMT = data.frame(Data_ML$Histology, Data_ML$Histology) %>% dplyr::distinct()
      }
      colnames(Organ_data_GMT) = c("Organ_type_raw", "Organ_type")
      save(file = "source/Organ_data_GMT.rda", Organ_data_GMT)
      if(length(colnames(Data_forest_tmp_8_GMT)) > 1){
        Data_forest_tmp_8_GMT = data.frame( lapply(Data_forest_tmp_8_GMT, as.factor) )
        if("Histology" %in% colnames(Data_forest_tmp_8_GMT)){
          Data_forest_tmp_8_GMT$Histology <- relevel(Data_forest_tmp_8_GMT$Histology, ref=names(sort(table(Data_forest_tmp_8_GMT$Histology),decreasing = T))[[1]]) 
        }
        final_mv_regxlevels_GMT <- (prepare_data(as.numeric(as.character(Data_forest_tmp_8_GMT$EP_option)),
                                                     as.matrix(data.frame(lapply(Data_forest_tmp_8_GMT[, setdiff(names(Data_forest_tmp_8_GMT), "EP_option")],
                                                                                 function(x) as.numeric(as.factor(x))))), type = "logistic") %>% stepwise(aic))$model
        if(length(final_mv_regxlevels_GMT) != 0){
          Data_forest_tmp_treat_GMT = Data_forest_tmp_8_GMT %>% dplyr::select(c("EP_option", final_mv_regxlevels_GMT))
        # m1_GMT <- glm(EP_option ~., data = Data_forest_tmp_8_GMT, family = binomial(link = "logit"))
        # final_mv_reg_GMT <- m1_GMT %>%
        #   stats::step(direction = "backward", trace = FALSE)
        # if(!is.null(final_mv_reg_GMT$xlevels)){
        #   Data_forest_tmp_treat_GMT = Data_forest_tmp_8_GMT %>% dplyr::select(c("EP_option", names(final_mv_reg_GMT$xlevels)))
          if(length(colnames(Data_forest_tmp_treat_GMT)) > 1){
            dd <- datadist(Data_forest_tmp_treat_GMT) 
            options(datadist=dd)
            formula_treat_GMT = formula(paste0("EP_option ~ `",paste(colnames(Data_forest_tmp_treat_GMT)[colnames(Data_forest_tmp_treat_GMT)!="EP_option"], collapse="` + `"),"`"))
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/m2_GMT.qs")){
              m2_GMT = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m2_GMT.qs")
            } else {
              m2_GMT <- lrm(formula_treat_GMT,
                            data = Data_forest_tmp_treat_GMT,x=TRUE,y=TRUE,penalty = Penal, tol=Toler)
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), m2_GMT, file = "source/m2_GMT.qs")
              }
            }
            BootNo = min(500, max(25, as.integer(50000 / nrow(Data_forest_tmp_treat_GMT))*100))
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/val_GMT.qs")){
              val_GMT = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/val_GMT.qs")
            } else {
              # error in rms package
              val_GMT <- try(rms::validate(m2_GMT, B=BootNo), silent = FALSE)
              if (class(val_GMT) == "try-error") {
                val_GMT = rms::validate(m2_GMT, B=BootNo)
              }
              # do not delete
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), val_GMT, file = "source/val_GMT.qs")
              }
            }
            if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/cal_GMT.qs")){
              cal_GMT = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/cal_GMT.qs")
            } else {
              cal_GMT = rms::calibrate(m2_GMT, B=BootNo)
              if(input$intermediate_file != "No"){ 
                QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), cal_GMT, file = "source/cal_GMT.qs")
              }
            }
            options(datadist=NULL)
            log_nomogram_GMT <- nomogram(m2_GMT, fun = plogis, lp=F, funlabel="Evidence level A/B")
            m2_GMT_pvalue = m2_GMT$stats["P"][[1]]
            m2_GMT_val_1 = val_GMT[9,5]
            m2_GMT_val_2 = val_GMT[2,5]
            m2_GMT_val_3 = val_GMT[1,5]
            m2_GMT_val_4 = val_GMT[3,5]
            m2_GMT_val_5 = val_GMT[4,5]
            rm(m2_GMT)
            rm(val_GMT)
          }
        }
      }
      output$figure_nomogram_treat_pre_CGP_GMT = renderPlot({
        if(length(colnames(Data_forest_tmp_8_GMT)) > 1){
          if(length(colnames(Data_forest_tmp_treat_GMT)) > 1){
            par(mfcol = c(2, 1))
            plot(cal_GMT, lwd=2, lty=1, 
                 cex.lab=1.2, cex.axis=1, cex.main=1.2, cex.sub=0.6, 
                 xlim=c(0, 1), ylim= c(0, 1), 
                 xlab="Nomogram-Predicted Probability of Evidence level A/B", 
                 ylab="Actual evidence level A/B (proportion)", 
                 legend=FALSE)
            
            abline(0, 1, lty=3, lwd=2,  col=c("#224444")) 
            text(x=.4, y=.0, adj = c(0,0), paste0("Matched therapy recommendation ratio, pre CGP information (other than mutation)","\n",
                                                  "Model lokelihood ratio test P-value: ", format(m2_GMT_pvalue, nsmall=3),'\n',
                                                  BootNo, "-time-bootstrapped-Brier score: ", round(m2_GMT_val_1, digits=3),'\n',
                                                  BootNo, "-time-bootstrapped-Nagelkerke R2 = ", round(m2_GMT_val_2, digits=3),'\n',
                                                  BootNo, "-time-bootstrapped-concordance index = ", round(((m2_GMT_val_3 + 1) / 2), digits=3),"\n",
                                                  "Intercept = ",round(m2_GMT_val_4, digits=3),"\n",
                                                  "Slope = ", round(m2_GMT_val_5, digits=3)))
            legend(x=.8, y=.2, legend=c("Apparent", "Bias-corrected", "Ideal"), 
                   lty=c(3, 1, 2), bty="n")
            plot(log_nomogram_GMT)
          } else {
            gg_empty()
          }
        } else {
          gg_empty()
        }
      })
      output$histology_nomogram_pre_CGP_GMT = renderText({
        if(length(colnames(Data_forest_tmp_8_GMT)) > 1){
          if(length(colnames(Data_forest_tmp_treat_GMT)) > 1){
            if(!is.null(log_nomogram_GMT$Histology$points)){
              paste0("Histology and points\n",            
                     paste(paste(log_nomogram_GMT$Histology$Histology, round(log_nomogram_GMT$Histology$points,digits = 0), sep="\t"), collapse = "\n"))
            } else {
              ""
            }
          }
        }
      })
      if(length(colnames(Data_forest_tmp_8)) > 1 & input$machine_learning == "Yes"){
        Data_forest_tmp_8$patientid = 1:length(Data_forest$EP_treat)
        dca_thresholds = seq(0, 0.25, 0.01)
        set.seed(1212)            
        # create a 5-fold cross validation set, 1 repeat which is the base case, change to suit your use case
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/glm_param.qs")){
          df_prediction = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/glm_param.qs")
        } else {
          cross_validation_samples <- rsample::vfold_cv(Data_forest_tmp_8, strata = "EP_treat", v = 5, repeats = 1)
          formula_treat = formula(paste0("EP_treat ~ `",paste(colnames(Data_forest_tmp_8)[!colnames(Data_forest_tmp_8) %in% c("EP_treat", "patientid")], collapse="` + `"),"`"))
          df_crossval_predictions <-
            cross_validation_samples %>%
            dplyr::mutate(
              glm_analysis = c(list(lrm(formula = formula_treat, data = rsample::analysis(splits[[1]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[2]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[3]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[4]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[5]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler))#,
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[6]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[7]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[8]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[9]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[10]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler))
                               )
            )
          print("ok")

          df_prediction_id = c(rsample::assessment(df_crossval_predictions$splits[[1]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[2]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[3]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[4]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[5]])$patientid#,
                               # rsample::assessment(df_crossval_predictions$splits[[6]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[7]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[8]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[9]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[10]])$patientid
          )
          df_prediction_score = c(predict(df_crossval_predictions$glm_analysis[[1]], rsample::assessment(df_crossval_predictions$splits[[1]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[2]], rsample::assessment(df_crossval_predictions$splits[[2]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[3]], rsample::assessment(df_crossval_predictions$splits[[3]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[4]], rsample::assessment(df_crossval_predictions$splits[[4]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[5]], rsample::assessment(df_crossval_predictions$splits[[5]]),type = "fitted")#,
                                  # predict(df_crossval_predictions$glm_analysis[[6]], rsample::assessment(df_crossval_predictions$splits[[6]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[7]], rsample::assessment(df_crossval_predictions$splits[[7]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[8]], rsample::assessment(df_crossval_predictions$splits[[8]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[9]], rsample::assessment(df_crossval_predictions$splits[[9]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[10]], rsample::assessment(df_crossval_predictions$splits[[10]]),type = "fitted")
          )
          rm(df_crossval_predictions)
          
          df_prediction = data.frame(df_prediction_id, df_prediction_score)
          colnames(df_prediction) = c("patientid", ".fitted")
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), df_prediction, file="source/glm_param.qs")
          }
        }
        
        df_cv_pred <-
          Data_forest_tmp_8 %>%
          dplyr::left_join(
            df_prediction,
            by = 'patientid'
          )
        if("PS" %in% colnames(Data_forest_tmp_8)){
          PS_data = data.frame(sort(unique(Data_forest_tmp_8$PS)))
          colnames(PS_data) = c("PS_type")
          save(file = "source/PS_data.rda", PS_data)
        } else {
          print("no PS data")
        }
        if("Lines" %in% colnames(Data_forest_tmp_8)){
          Lines_data = data.frame(sort(unique(Data_forest_tmp_8$Lines)))
          colnames(Lines_data) = c("Lines_type")
          save(file = "source/Lines_data.rda", Lines_data)
        } else {
          print("no Line data")
        }
        
        set.seed(1212)
        Data_ML = Data_ML_full %>%
          dplyr::select(-EP_option, -ID)
        Data_ML$EP_treat = factor(Data_ML$EP_treat)
        common_cols = colnames(Data_forest_tmp_8)
        common_cols = common_cols[common_cols != "patientid"]
        Data_ML = Data_ML[, ..common_cols]
        Data_ML[, (names(Data_ML)) := lapply(.SD, function(x) if (is.character(x)) as.factor(x) else x)]
        factor_cols <- names(Data_ML)[sapply(Data_ML, is.factor)]
        factor_cols = factor_cols[factor_cols != "EP_treat"]
        factor_cols = factor_cols[factor_cols %in% c("PS", "Panel", "Histology", "Best_effect", "Lines")]
        if (length(factor_cols) > 0) {
          Data_ML <- one_hot(Data_ML, cols = factor_cols, sparsifyNAs = FALSE, dropCols = TRUE)
        }
        cls_met <- metric_set(roc_auc)
        Data_cv <- vfold_cv(v=5, Data_ML, strata = EP_treat)
        Data_train = vfold_cv(v=5,
                              rsample::analysis(Data_cv$splits[[1]]) %>%
                                dplyr::slice_sample(n = min(nrow(rsample::analysis(Data_cv$splits[[1]])), 20000)),
                              strata = EP_treat)
        Data_parameter_train = rsample::analysis(Data_train$splits[[1]])
        Data_parameter_test = rsample::assessment(Data_train$splits[[1]])
        Data_parameter_cv = Data_parameter_train |> 
          rsample::vfold_cv(v = 5, strata = EP_treat)
        Data_recipe_train <- Data_parameter_train |>
          recipe(EP_treat ~ .) |>
          step_dummy(all_nominal_predictors()) |>
          step_nzv(all_predictors())# |>
        Data_recipe <- Data_ML |>
          recipe(EP_treat ~ .) |>
          step_dummy(all_nominal_predictors()) |>
          step_nzv(all_predictors()) |>
          step_dummy(all_nominal_predictors())
        # cv_fit_pred <- function(recipe, spec, df_train, df_test, df_cv = NULL, grid = 8) {
        #   if(is.null(df_cv)) {
        #     params_grid = NULL
        #     wf <- 
        #       workflow() |> 
        #       workflows::add_recipe(recipe = recipe) |> 
        #       workflows::add_model(spec = spec)
        #   } else {
        #     cv_wf <- 
        #       workflow() |> 
        #       workflows::add_recipe(recipe = recipe) |> 
        #       workflows::add_model(spec = spec)
        #     params_grid <- 
        #       cv_wf |> 
        #       tune::tune_grid(
        #         resamples = df_cv,
        #         grid = grid,
        #         control = tune::control_grid(save_pred = TRUE)
        #       )
        #     wf <- 
        #       cv_wf |> 
        #       tune::finalize_workflow(
        #         parameters = params_grid |> tune::select_best(metric = "roc_auc")
        #       )
        #   }
        #   wf_fit <- 
        #     wf |> 
        #     parsnip::fit(data = df_train)
        #   pred <- 
        #     wf_fit |> 
        #     tune::augment(new_data = df_test)
        #   return(
        #     list(
        #       params_grid = params_grid, 
        #       wf_fit = wf_fit, 
        #       pred = pred
        #     )
        #   )
        # }
        rf_param_space <- parameters(
          mtry(range = c(1, 5)),
          min_n(range = c(2, 50))
          )
        lgb_param_space <- parameters(
          mtry(range = c(1, 5)),
          tree_depth(range = c(2, 5)),
          min_n(range = c(2, 50)),
          num_leaves(range = c(10, 100))
        )
        cv_fit_pred <- function(recipe, spec, df_train, df_test, df_cv = NULL, iter = 500, init_grid_n = 8, param_space = NULL) {
          if (is.null(df_cv)) {
            wf <- workflow() |> 
              add_recipe(recipe) |> 
              add_model(spec)
            params_grid <- NULL
          } else {
            cv_wf <- workflow() |> 
              add_recipe(recipe) |> 
              add_model(spec)
            
            # 初期グリッド（ランダム or グリッド）
            initial_grid <- tune::tune_grid(
              cv_wf,
              resamples = df_cv,
              grid = init_grid_n,
              metrics = metric_set(roc_auc),
              control = control_grid(save_pred = TRUE)
            )
            
            # ベイズ最適化
            params_grid <- tune::tune_bayes(
              cv_wf,
              resamples = df_cv,
              param_info = param_space,
              initial = initial_grid,
              iter = iter,
              metrics = metric_set(roc_auc),
              control = tune::control_bayes(
                no_improve = 10,          # 改善がないと停止
                verbose = FALSE,
                save_pred = TRUE
              )
            )
            
            best_params <- params_grid |> tune::select_best(metric = "roc_auc")
            
            wf <- cv_wf |> tune::finalize_workflow(best_params)
          }
          
          wf_fit <- wf |> fit(data = df_train)
          pred <- wf_fit |> augment(new_data = df_test)
          
          return(list(
            params_grid = params_grid,
            wf_fit = wf_fit,
            pred = pred
          ))
        }
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/lgb_param.qs")){
          lgb_parameter = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/lgb_param.qs")
        } else {
          lgb_spec <-
            parsnip::boost_tree(mode = "classification") %>%
            parsnip::set_args(
              tree_depth = tune(), min_n = tune(), mtry = tune(), trees = 2000, learn_rate = 0.05, lambda_l1 = 0.9) |> 
            set_engine("lightgbm", num_leaves = tune(),
                       nthread = max(1, parallel::detectCores() - 1, na.rm = TRUE))
          lgb_result <- 
            Data_recipe_train |> 
            cv_fit_pred(lgb_spec, Data_parameter_train, Data_parameter_test, Data_parameter_cv, param_space = lgb_param_space)
          lgb_parameter = (lgb_result$params_grid %>% show_best(metric = "roc_auc"))[1,]
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), lgb_parameter, file="source/lgb_param.qs")
          }
        }  
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/rf_param.qs")){
          rf_parameter = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/rf_param.qs")
        } else {
          rf_spec <-
            parsnip::rand_forest() |> 
            parsnip::set_args(mtry = tune(), min_n = tune(), trees = 1000) |> 
            parsnip::set_engine(
              'ranger',
              num.threads = max(1, parallel::detectCores() - 1, na.rm = TRUE)
            ) |> 
            parsnip::set_mode('classification')            
          rf_result <- 
            Data_recipe_train |> 
            cv_fit_pred(rf_spec, Data_parameter_train, Data_parameter_test, Data_parameter_cv, param_space = rf_param_space)
          rf_parameter = (rf_result$params_grid %>% show_best(metric = "roc_auc"))[1,]
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), rf_parameter, file="source/rf_param.qs")
          }
        }  
        set.seed(1212)
        lgb_model <-
          parsnip::boost_tree(mode = "classification") %>%
          set_engine("lightgbm", num_leaves = lgb_parameter$num_leaves,
                     importance = TRUE,  # 変数重要度を有効化
                     lambda_l1 = .9,
                     min_n = lgb_parameter$min_n,
                     tree_depth = lgb_parameter$tree_depth,
                     learn_rate = 0.05,
                     trees = 2000,
                     mtry = lgb_parameter$mtry,
                     nthread = max(1, parallel::detectCores() - 1, na.rm = TRUE))
        lgb_wflow <- 
          workflow() %>% 
          add_model(lgb_model) %>% 
          add_recipe(Data_recipe)
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/lgb_m2.qs")){
          lgb_m2 = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/lgb_m2.qs")
        } else {
          lgb_m2 = lgb_wflow %>% fit(Data_ML) 
          # lgb_m2 = fit_resamples(lgb_wflow, resamples = Data_cv,
          #               control = control_resamples(save_pred = TRUE))
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), lgb_m2, file = "source/lgb_m2.qs")
          }
        }
        tmp_post$lgb_m2 = lgb_m2
        rf_model <-
          parsnip::rand_forest() |>
          parsnip::set_args(trees = 2000, min_n = rf_parameter$min_n, mtry = rf_parameter$mtry) |>
          parsnip::set_engine(
            'ranger', keep.inbag = TRUE, importance = "impurity",
            num.threads = max(1, parallel::detectCores() - 1, na.rm = TRUE)
          ) |>
          parsnip::set_mode('classification')
        rf_wflow <- 
          workflow() %>% 
          add_model(rf_model) %>% 
          add_recipe(Data_recipe)
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/rf_m2.qs")){
          rf_m2 = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/rf_m2.qs")
        } else {
          rf_m2 = rf_wflow %>% fit(Data_ML) 
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), rf_m2, file = "source/rf_m2.qs")
          }
        }
        tmp_post$rf_m2 = rf_m2
        X <- Data_ML %>% select(-EP_treat)
        X <- if (nrow(X) > 10000) {
          X[sample(nrow(X), 10000), , drop = FALSE]
        } else {
          X
        }
        pred_wrapper <- function(object, newdata) {
          predict(object, new_data = newdata, type = "prob") %>% dplyr::pull(2)
        }
        set.seed(123)
        if(input$importance == "SHAP value (heavy analysis)"){
          if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/shap_values_lgb.qs")){
            shap_values = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/shap_values_lgb.qs")
          } else {
            shap_values <- fastshap::explain(parallel = ifelse(nrow(Data_ML) < 10000, TRUE, FALSE),
                                             # ncores = max(1, parallel::detectCores() - 1, na.rm = TRUE),
                                             object = lgb_m2,
                                             X = X[seq_len(min(nrow(X), 5000)),],
                                             pred_wrapper = pred_wrapper,
                                             nsim = 3  # モンテカルロシミュレーション回数（計算精度と時間のバランスで調整）
            )
            if(input$intermediate_file != "No"){ 
              QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), shap_values, file = "source/shap_values_lgb.qs")
            }
          }
          rm(lgb_m2)
          # 各特徴量ごとの平均絶対値を計算
          mean_abs_shap <- colMeans(abs(shap_values))
          # データフレームに変換
          importance_df <- data.frame(
            feature = names(mean_abs_shap),
            mean_abs_shap = mean_abs_shap
          )
          importance_df = importance_df %>%
            dplyr::arrange(mean_abs_shap)
          shap_long <- as.data.frame(shap_values) %>%
            mutate(id = row_number()) %>%   # サンプルIDを追加
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "shap_value"
            )
          # ② Xの元データもlong形式に変換（行番号を付与）
          X_long <- X %>%
            mutate(id = row_number()) %>%
            mutate(across(where(is.factor),
                          ~ case_when(
                            . %in% c("Yes", "Male") ~ 1,
                            . %in% c("No", "Female") ~ 0,
                            TRUE ~ as.numeric(as.character(.))
                          ))) %>%
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "orig_value"
            ) %>%
            group_by(feature) %>%
            group_by(feature) %>%
            mutate(norm_value = case_when(
              max(orig_value) == min(orig_value) ~ orig_value,  # すべて同じならそのまま
              TRUE ~ (orig_value - min(orig_value)) / (max(orig_value) - min(orig_value))
            )) %>%
            ungroup()
          # ③ shap_longとX_longをidとfeatureで結合
          shap_long <- shap_long %>%
            left_join(X_long, by = c("id", "feature"))
          shap_long$feature = factor(shap_long$feature, levels = importance_df$feature)
          importance_df = importance_df[max(1, nrow(importance_df)-29):nrow(importance_df),]
          shap_long = shap_long %>%
            filter(feature %in% importance_df$feature)
          # 棒グラフでプロット（横向き）
          gSHAP1_lgb <- ggplot(importance_df, aes(x = reorder(feature, mean_abs_shap), y = mean_abs_shap)) +
            geom_bar(stat = "identity") +
            coord_flip() +
            labs(title = "LightGBM, Mean |SHAP value| of Features", x = "Features", y = "Mean |SHAP value|") +
            scale_fill_manual(values = c("Positive" = "steelblue", "Negative" = "tomato")) +  # 色の指定
            theme_minimal() +
            theme(legend.title = element_blank(),  # 凡例タイトルを削除
                  legend.position = "bottom")  # 凡例を下に配置
          gSHAP2_lgb = ggplot(shap_long, aes(x = shap_value, y = feature, color = norm_value)) +
            geom_quasirandom(alpha = 0.2, width = 0.2) +
            scale_color_gradient(low = "blue", high = "red", 
                                 limits = c(0, 1), 
                                 breaks = c(0, 1),
                                 labels = c("Low/No/Female", "High/Yes/Male")) +
            theme_minimal() +
            labs(title = "SHAP Value Distribution", color = "Feature Value")
          if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/shap_values_rf.qs")){
            shap_values = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/shap_values_rf.qs")
          } else {
            shap_values <- fastshap::explain(parallel = ifelse(nrow(Data_ML) < 10000, TRUE, FALSE),
                                             # ncores = max(1, parallel::detectCores() - 1, na.rm = TRUE),
                                             object = rf_m2,
                                             X = X[seq_len(min(nrow(X), 5000)),],
                                             pred_wrapper = pred_wrapper,
                                             nsim = 3  # モンテカルロシミュレーション回数（計算精度と時間のバランスで調整）
            )
            if(input$intermediate_file != "No"){ 
              QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), shap_values, file = "source/shap_values_rf.qs")
            }
          }
          rm(rf_m2)
          # 各特徴量のSHAP値の符号（正/負）を判別し、色を付ける
          shap_values_with_sign <- shap_values
          shap_values_with_sign[] <- ifelse(shap_values_with_sign >= 0, "Positive", "Negative")
          # 各特徴量ごとの平均絶対値を計算
          mean_abs_shap <- colMeans(abs(shap_values))
          # データフレームに変換
          importance_df <- data.frame(
            feature = names(mean_abs_shap),
            mean_abs_shap = mean_abs_shap,
            sign = shap_values_with_sign[1,]  # ここでSHAP値の符号を代表として使う
          )
          importance_df = importance_df %>%
            dplyr::arrange(mean_abs_shap)
          shap_long <- as.data.frame(shap_values) %>%
            mutate(id = row_number()) %>%   # サンプルIDを追加
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "shap_value"
            )
          # ② Xの元データもlong形式に変換（行番号を付与）
          X_long <- X %>%
            mutate(id = row_number()) %>%
            mutate(across(where(is.factor),
                          ~ case_when(
                            . %in% c("Yes", "Male") ~ 1,
                            . %in% c("No", "Female") ~ 0,
                            TRUE ~ as.numeric(as.character(.))
                          ))) %>%
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "orig_value"
            ) %>%
            group_by(feature) %>%
            group_by(feature) %>%
            mutate(norm_value = case_when(
              max(orig_value) == min(orig_value) ~ orig_value,  # すべて同じならそのまま
              TRUE ~ (orig_value - min(orig_value)) / (max(orig_value) - min(orig_value))
            )) %>%
            ungroup()
          # ③ shap_longとX_longをidとfeatureで結合
          shap_long <- shap_long %>%
            left_join(X_long, by = c("id", "feature"))
          shap_long$feature = factor(shap_long$feature, levels = importance_df$feature)
          importance_df = importance_df[max(1, nrow(importance_df)-29):nrow(importance_df),]
          shap_long = shap_long %>%
            filter(feature %in% importance_df$feature)
          # 棒グラフでプロット（横向き）
          gSHAP1_rf <- ggplot(importance_df, aes(x = reorder(feature, mean_abs_shap), y = mean_abs_shap)) +
            geom_bar(stat = "identity") +
            coord_flip() +
            labs(title = "RandomForest, Mean |SHAP value| of Features", x = "Features", y = "Mean |SHAP value|") +
            scale_fill_manual(values = c("Positive" = "steelblue", "Negative" = "tomato")) +  # 色の指定
            theme_minimal() +
            theme(legend.title = element_blank(),  # 凡例タイトルを削除
                  legend.position = "bottom")  # 凡例を下に配置
          gSHAP2_rf = ggplot(shap_long, aes(x = shap_value, y = feature, color = norm_value)) +
            geom_quasirandom(alpha = 0.2, width = 0.2) +
            scale_color_gradient(low = "blue", high = "red", 
                                 limits = c(0, 1), 
                                 breaks = c(0, 1),
                                 labels = c("Low/No/Female", "High/Yes/Male")) +
            theme_minimal() +
            labs(title = "SHAP Value Distribution", color = "Feature Value")
        } else {
          importances <- map(Data_cv$splits, function(split) {
            lgb_fit <- fit(lgb_wflow, data = analysis(split))
            lgb_booster <- extract_fit_engine(lgb_fit)
            imp <- lgb.importance(lgb_booster)
            imp %>% select(Feature, Gain)  # 必要な列だけ
          })
          # 2. importance を平均化（split ごとの dataframe を結合 → group_by → summarise）
          lgb_vi <- bind_rows(importances) %>%
            group_by(Feature) %>%
            summarise(importance = mean(Gain, na.rm = TRUE)) %>%
            arrange(desc(importance))
          # 3. 可視化
          gSHAP1_lgb <- ggplot(lgb_vi[seq_len(min(30,nrow(lgb_vi))),], aes(x = reorder(Feature, importance), y = importance)) +
            geom_bar(stat = "identity", fill = "steelblue") +
            coord_flip() +
            labs(
              title = "Top 30 Variable Importance from LightGBM (CV Averaged)",
              x = "Feature",
              y = "Importance"
            ) +
            theme_minimal()
          gSHAP2_lgb = gg_empty()
          rf_importances <- map(Data_cv$splits, function(split) {
            rf_fit <- fit(rf_wflow, data = analysis(split))
            rf_model_fitted <- extract_fit_engine(rf_fit)
            # 変数重要度を tibble 形式で取得
            tibble(
              Feature = names(rf_model_fitted$variable.importance),
              Importance = rf_model_fitted$variable.importance
            )
          })
          # 重要度を平均化
          rf_vi <- bind_rows(rf_importances) %>%
            group_by(Feature) %>%
            summarise(importance = mean(Importance, na.rm = TRUE)) %>%
            arrange(desc(importance))
          # 3. 可視化
          gSHAP1_rf <- ggplot(rf_vi[seq_len(min(30,nrow(rf_vi))),], aes(x = reorder(Feature, importance), y = importance)) +
            geom_bar(stat = "identity", fill = "steelblue") +
            coord_flip() +
            labs(
              title = "Top 30 Variable Importance from Random Forest (CV Averaged)",
              x = "Feature",
              y = "Importance"
            ) +
            theme_minimal()
          gSHAP2_rf = gg_empty()
        }
        keep_pred <- control_resamples(save_pred = TRUE, save_workflow = TRUE)
        
        set.seed(1212)
        rf_res <- rf_wflow %>% fit_resamples(resamples = Data_cv, metrics = cls_met, control = keep_pred)
        rf_res_all = rf_res %>%
          pull(.predictions) %>%
          bind_rows() %>%
          dplyr::arrange(.row)
        # df_cv_pred$rf  = rf_res_all$.pred_1
        ROC_rf <- roc(EP_treat ~ .pred_1, data = rf_res_all, ci = TRUE)
        gROC_rf = ggroc(ROC_rf, 
                        size = 1, #サイズ
                        legacy.axes = TRUE) + 
          geom_abline(color = "dark grey", size = 0.5) +
          theme_classic() +
          labs(
            title = 
              paste0("Random forest, AUC: ", round(ROC_rf$auc[1],3), " (95%CI:",round(ROC_rf$ci[1],3), "-", round(ROC_rf$ci[3],3), ")"),
            subtitle = paste0("mtry/min_n=",rf_parameter$mtry,"/",rf_parameter$min_n)
          )
        
        set.seed(1212)
        lgb_res <- lgb_wflow %>% fit_resamples(resamples = Data_cv, metrics = cls_met, control = keep_pred)
        lgb_res_all = lgb_res %>%
          pull(.predictions) %>%
          bind_rows() %>%
          dplyr::arrange(.row)
        # df_cv_pred$lgb  = lgb_res_all$.pred_1
        ROC_lgb <- roc(EP_treat ~ .pred_1, data = lgb_res_all, ci = TRUE)
        gROC_lgb = ggroc(ROC_lgb, 
                         size = 1, #サイズ
                         legacy.axes = TRUE) + 
          geom_abline(color = "dark grey", size = 0.5) +
          theme_classic() +
          labs(
            title = 
              paste0("LightGBM, AUC: ", round(ROC_lgb$auc[1],3), " (95%CI:",round(ROC_lgb$ci[1],3), "-", round(ROC_lgb$ci[3],3), ")"),
            subtitle = paste0("mtry/min_n/tree_depth/num_leaves=",lgb_parameter$mtry,"/",lgb_parameter$min_n,"/",lgb_parameter$tree_depth,"/",lgb_parameter$num_leaves)
          )
        
        g1 = collect_predictions(lgb_res) %>%
          ggplot(aes(.pred_1)) +
          geom_histogram(col = "white", bins = 40) +
          facet_wrap(~ EP_treat, ncol = 1) +
          geom_rug(col = "blue", alpha = 1 / 2) + 
          labs(x = "LightGBM: Probability estimate of treatment reach")
        g2 = lgb_res %>% cal_plot_windowed(truth = EP_treat, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("LightGBM: Calibration curve")
        set.seed(1212)
        iso_val <- cal_validate_logistic(lgb_res, metrics = cls_met,
                                         save_pred = TRUE, times = 25)
        # collect_metrics(iso_val)
        g3= collect_predictions(iso_val) %>%
          filter(.type == "calibrated") %>%
          cal_plot_windowed(truth = EP_treat, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("LightGBM: Logistic calibration via GAM")
        
        df_cv_pred$lgb  = (iso_val %>%
                             pull(.predictions_cal) %>%
                             bind_rows() %>%
                             dplyr::arrange(.row))$.pred_1
        # beta_val <- lgb_res %>% cal_validate_beta(metrics = cls_met, save_pred = TRUE)
        # # collect_metrics(beta_val)
        # g3= collect_predictions(beta_val) %>%
        #   filter(.type == "calibrated") %>%
        #   cal_plot_windowed(truth = EP_treat, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
        #   ggtitle("LightGBM: Beta calibration")
        # df_cv_pred$lgb  = (beta_val %>%
        #                      pull(.predictions_cal) %>%
        #                      bind_rows() %>%
        #                      dplyr::arrange(.row))$.pred_1
        g4 = collect_predictions(rf_res) %>%
          ggplot(aes(.pred_1)) +
          geom_histogram(col = "white", bins = 40) +
          facet_wrap(~ EP_treat, ncol = 1) +
          geom_rug(col = "blue", alpha = 1 / 2) + 
          labs(x = "Random forest: Probability estimate of treatment reach")
        g5 = rf_res %>% cal_plot_windowed(truth = EP_treat, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("Random forest: Calibration curve")
        
        iso_val <- cal_validate_logistic(rf_res, metrics = cls_met,
                                         save_pred = TRUE, times = 25)
        collect_metrics(iso_val)
        g6= collect_predictions(iso_val) %>%
          filter(.type == "calibrated") %>%
          cal_plot_windowed(truth = EP_treat, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("Random forest: Logistic calibration via GAM")
        df_cv_pred$rf  = (iso_val %>%
                            pull(.predictions_cal) %>%
                            bind_rows() %>%
                            dplyr::arrange(.row))$.pred_1
        # beta_val <- rf_res %>% cal_validate_beta(metrics = cls_met, save_pred = TRUE)
        # collect_metrics(beta_val)
        # g6= collect_predictions(beta_val) %>%
        #   filter(.type == "calibrated") %>%
        #   cal_plot_windowed(truth = EP_treat, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
        #   ggtitle("Random forest: Beta calibration")
        # df_cv_pred$rf  = (beta_val %>%
        #                      pull(.predictions_cal) %>%
        #                      bind_rows() %>%
        #                      dplyr::arrange(.row))$.pred_1
        
        output$figure_DCA_pre_CGP_1 = renderPlot({
          dcurves::dca( # calculate net benefit scores on mean cross validation predictions
            data = df_cv_pred,
            formula = EP_treat ~ .fitted + rf + lgb,
            thresholds = dca_thresholds,
            label = list(
              .fitted = "5-fold cross-validated logistic regression model",
              rf = "5-fold cross-validated random forest model",
              lgb = "5-fold cross-validated lightGBM model"
            ))$dca |> 
            # plot cross validated net benefit values
            ggplot(aes(x = threshold, y = net_benefit, color = label)) +
            stat_smooth(method = "loess", se = FALSE, formula = "y ~ x", span = 0.2, fullrange = TRUE) +
            scale_x_continuous(labels = scales::percent_format(accuracy = 1)) +
            scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
            labs(x = "Threshold probability", y = "Net Benefit", color = "") +
            coord_cartesian(xlim = c(0, 0.25), ylim = c(-0.01, NA)) + 
            theme_bw()
        })
        
        output$figure_DCA_pre_CGP_ROC = renderPlot({
          gNomogram = df_cv_pred %>%
            ggplot(aes(.fitted)) +
            theme_classic() +
            geom_histogram(col = "white", bins = 40) +
            facet_wrap(~ EP_treat, ncol = 1) +
            geom_rug(col = "blue", alpha = 1 / 2) + 
            labs(x = "Logistic regression: Probability estimate of treatment reach")
          ROC <- roc(EP_treat ~ .fitted, data = df_cv_pred, ci = TRUE)
          gROC2 = ggroc(ROC, 
                        size = 1, #サイズ
                        legacy.axes = TRUE) + 
            geom_abline(color = "dark grey", size = 0.5) +
            theme_classic() +
            labs(
              title = 
                paste0("Logistic regression, AUC: ", round(ROC$auc[1],3), " (95%CI:",round(ROC$ci[1],3), "-", round(ROC$ci[3],3), ")")
            )
          ROC3 <- roc(EP_treat ~ .fitted + lgb + rf, data = df_cv_pred, ci = TRUE)
          gROC3 = ggroc(ROC3, 
                        size = 1, #サイズ
                        legacy.axes = TRUE) + 
            geom_abline(color = "dark grey", size = 0.5) +
            theme_classic() +
            scale_color_hue(name = "Prediction methods",
                            labels = c(.fitted = "Logistic regression",
                                       lgb = "Light GBM",
                                       rf  = "Random forest")) +
            labs(
              title = 
                paste0("ROC curves for treatment reach")
            )
          grid.arrange(gNomogram,gROC2,gROC3, nrow = 3)
        })
        
        output$figure_DCA_pre_CGP_Machine_learning = renderPlot({
          grid.arrange(g1,g2,g3,gROC_lgb,gSHAP1_lgb,gSHAP2_lgb,g4,g5,g6,gROC_rf,gSHAP1_rf,gSHAP2_rf, nrow = 4, ncol=3)
        })          
        output$table_DCA_pre_CGP_2 = render_gt({
          (dcurves::dca( # calculate net benefit scores on mean cross validation predictions
            data = df_cv_pred,
            formula = EP_treat ~ .fitted + rf + lgb,
            thresholds = dca_thresholds,
            label = list(
              .fitted = "5-fold cross-validated logistic regression model",
              rf = "5-fold cross-validated random forest model",
              lgb = "5-fold cross-validated lightGBM model"
            ))) %>%
            net_intervention_avoided() %>%
            as_tibble() %>%
            gt::gt() %>%
            gt::fmt_percent(columns = threshold, decimals = 0) %>%
            gt::fmt(columns = net_benefit, fns = function(x) style_sigfig(x, digits = 3)) %>%
            gt::cols_label(
              label = "Strategy",
              threshold = "Decision Threshold",
              net_benefit = "Net Benefit"
            ) %>%
            gt::cols_align("left", columns = label)
        })
        
        output$table_prediction = DT::renderDataTable(df_cv_pred,
                                                      server = FALSE,
                                                      filter = 'top', 
                                                      extensions = c('Buttons'), 
                                                      options = list(pageLength = 100, 
                                                                     scrollX = TRUE,
                                                                     scrollY = "1000px",
                                                                     scrollCollapse = TRUE,
                                                                     dom="Blfrtip",
                                                                     buttons = c('csv', 'copy')))
      }
      if(length(colnames(Data_forest_tmp_8_rec)) > 1 & input$machine_learning == "Yes"){
        Data_forest_tmp_8_rec$patientid = 1:length(Data_forest_tmp_8_rec$EP_option)
        dca_thresholds = seq(0, 1, 0.01)
        set.seed(1212)            
        # create a 5-fold cross validation set, 1 repeat which is the base case, change to suit your use case
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/glm_param_rec.qs")){
          df_prediction_rec = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/glm_param_rec.qs")
        } else {
          cross_validation_samples <- rsample::vfold_cv(Data_forest_tmp_8_rec, strata = "EP_option", v = 5, repeats = 1)
          formula_treat = formula(paste0("EP_option ~ `",paste(colnames(Data_forest_tmp_8_rec)[!colnames(Data_forest_tmp_8_rec) %in% c("EP_option", "patientid")], collapse="` + `"),"`"))
          df_crossval_predictions <-
            cross_validation_samples %>%
            dplyr::mutate(
              glm_analysis = c(list(lrm(formula = formula_treat, data = rsample::analysis(splits[[1]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[2]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[3]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[4]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[5]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler))#,
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[6]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[7]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[8]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[9]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[10]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler))
              )
            )
          
          df_prediction_id = c(rsample::assessment(df_crossval_predictions$splits[[1]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[2]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[3]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[4]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[5]])$patientid#,
                               # rsample::assessment(df_crossval_predictions$splits[[6]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[7]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[8]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[9]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[10]])$patientid
          )
          df_prediction_score = c(predict(df_crossval_predictions$glm_analysis[[1]], rsample::assessment(df_crossval_predictions$splits[[1]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[2]], rsample::assessment(df_crossval_predictions$splits[[2]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[3]], rsample::assessment(df_crossval_predictions$splits[[3]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[4]], rsample::assessment(df_crossval_predictions$splits[[4]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[5]], rsample::assessment(df_crossval_predictions$splits[[5]]),type = "fitted")#,
                                  # predict(df_crossval_predictions$glm_analysis[[6]], rsample::assessment(df_crossval_predictions$splits[[6]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[7]], rsample::assessment(df_crossval_predictions$splits[[7]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[8]], rsample::assessment(df_crossval_predictions$splits[[8]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[9]], rsample::assessment(df_crossval_predictions$splits[[9]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[10]], rsample::assessment(df_crossval_predictions$splits[[10]]),type = "fitted")
          )
          rm(df_crossval_predictions)
          
          df_prediction_rec = data.frame(df_prediction_id, df_prediction_score)
          colnames(df_prediction_rec) = c("patientid", ".fitted")
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), df_prediction_rec, file="source/glm_param_rec.qs")
          }
        }
        
        df_cv_pred_rec <-
          Data_forest_tmp_8_rec %>%
          dplyr::left_join(
            df_prediction_rec,
            by = 'patientid'
          )
        if("PS" %in% colnames(Data_forest_tmp_8_rec)){
          PS_data_rec = data.frame(sort(unique(Data_forest_tmp_8_rec$PS)))
          colnames(PS_data_rec) = c("PS_type")
          save(file = "source/PS_data_rec.rda", PS_data_rec)
        }
        if("Lines" %in% colnames(Data_forest_tmp_8_rec)){
          Lines_data_rec = data.frame(sort(unique(Data_forest_tmp_8_rec$Lines)))
          colnames(Lines_data_rec) = c("Lines_type")
          save(file = "source/Lines_data_rec.rda", Lines_data_rec)
        }
        
        set.seed(1212)
        Data_ML = Data_ML_full %>%
          dplyr::select(-EP_treat, -ID)
        common_cols = colnames(Data_forest_tmp_8_rec)
        common_cols = common_cols[common_cols != "patientid"]
        Data_ML = Data_ML[, ..common_cols]
        Data_ML[, (names(Data_ML)) := lapply(.SD, function(x) if (is.character(x)) as.factor(x) else x)]
        factor_cols <- names(Data_ML)[sapply(Data_ML, is.factor)]
        factor_cols = factor_cols[factor_cols != "EP_option"]
        factor_cols = factor_cols[factor_cols %in% c("PS", "Panel", "Histology", "Best_effect", "Lines")]
        if (length(factor_cols) > 0) {
          Data_ML <- one_hot(Data_ML, cols = factor_cols, sparsifyNAs = FALSE, dropCols = TRUE)
        }
        Data_ML$EP_option = factor(Data_ML$EP_option)
        cls_met <- metric_set(roc_auc)
        Data_cv <- vfold_cv(v=5, Data_ML, strata = EP_option)
        Data_train = vfold_cv(v=5,
                              rsample::analysis(Data_cv$splits[[1]]) %>%
                                dplyr::slice_sample(n = min(nrow(rsample::analysis(Data_cv$splits[[1]])), 20000)),
                              strata = EP_option)
        Data_parameter_train = rsample::analysis(Data_train$splits[[1]])
        Data_parameter_test = rsample::assessment(Data_train$splits[[1]])
        Data_parameter_cv = Data_parameter_train |> 
          rsample::vfold_cv(v = 5, strata = EP_option)
        Data_recipe_train <- Data_parameter_train |>
          recipe(EP_option ~ .) |>
          step_nzv(all_predictors()) |>
          step_dummy(all_nominal_predictors())
        Data_recipe <- Data_ML |>
          recipe(EP_option ~ .) |>
          step_nzv(all_predictors()) |>
          step_dummy(all_nominal_predictors())
        # cv_fit_pred <- function(recipe, spec, df_train, df_test, df_cv = NULL, grid = 8) {
        #   if(is.null(df_cv)) {
        #     params_grid = NULL
        #     wf <- 
        #       workflow() |> 
        #       workflows::add_recipe(recipe = recipe) |> 
        #       workflows::add_model(spec = spec)
        #   } else {
        #     cv_wf <- 
        #       workflow() |> 
        #       workflows::add_recipe(recipe = recipe) |> 
        #       workflows::add_model(spec = spec)
        #     params_grid <- 
        #       cv_wf |> 
        #       tune::tune_grid(
        #         resamples = df_cv,
        #         grid = grid,
        #         control = tune::control_grid(save_pred = TRUE)
        #       )
        #     wf <- 
        #       cv_wf |> 
        #       tune::finalize_workflow(
        #         parameters = params_grid |> tune::select_best(metric = "roc_auc")
        #       )
        #   }
        #   wf_fit <- 
        #     wf |> 
        #     parsnip::fit(data = df_train)
        #   pred <- 
        #     wf_fit |> 
        #     tune::augment(new_data = df_test)
        #   return(
        #     list(
        #       params_grid = params_grid, 
        #       wf_fit = wf_fit, 
        #       pred = pred
        #     )
        #   )
        # }
        rf_param_space <- parameters(
          mtry(range = c(1, 5)),
          min_n(range = c(2, 50))
        )
        lgb_param_space <- parameters(
          mtry(range = c(1, 5)),
          tree_depth(range = c(2, 5)),
          min_n(range = c(2, 50)),
          num_leaves(range = c(10, 100))
        )
        cv_fit_pred <- function(recipe, spec, df_train, df_test, df_cv = NULL, iter = 500, init_grid_n = 8, param_space = NULL) {
          if (is.null(df_cv)) {
            wf <- workflow() |> 
              add_recipe(recipe) |> 
              add_model(spec)
            params_grid <- NULL
          } else {
            cv_wf <- workflow() |> 
              add_recipe(recipe) |> 
              add_model(spec)
            
            # 初期グリッド（ランダム or グリッド）
            initial_grid <- tune::tune_grid(
              cv_wf,
              resamples = df_cv,
              grid = init_grid_n,
              metrics = metric_set(roc_auc),
              control = control_grid(save_pred = TRUE)
            )
            
            # ベイズ最適化
            params_grid <- tune::tune_bayes(
              cv_wf,
              resamples = df_cv,
              param_info = param_space,
              initial = initial_grid,
              iter = iter,
              metrics = metric_set(roc_auc),
              control = tune::control_bayes(
                no_improve = 10,          # 改善がないと停止
                verbose = FALSE,
                save_pred = TRUE
              )
            )
            
            best_params <- params_grid |> tune::select_best(metric = "roc_auc")
            
            wf <- cv_wf |> tune::finalize_workflow(best_params)
          }
          
          wf_fit <- wf |> fit(data = df_train)
          pred <- wf_fit |> augment(new_data = df_test)
          
          return(list(
            params_grid = params_grid,
            wf_fit = wf_fit,
            pred = pred
          ))
        }
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/lgb_param_rec.qs")){
          lgb_parameter_rec = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/lgb_param_rec.qs")
        } else {
          lgb_spec <-
            parsnip::boost_tree(mode = "classification") %>%
            parsnip::set_args(
              tree_depth = tune(), min_n = tune(), mtry = tune(), trees = 2000, learn_rate = 0.05, lambda_l1 = 0.9) |> 
            set_engine("lightgbm", num_leaves = tune(),
                       nthread = max(1, parallel::detectCores() - 1, na.rm = TRUE))
          lgb_result <- 
            Data_recipe_train |> 
            cv_fit_pred(lgb_spec, Data_parameter_train, Data_parameter_test, Data_parameter_cv, param_space = lgb_param_space)
          lgb_parameter_rec = (lgb_result$params_grid %>% show_best(metric = "roc_auc"))[1,]
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), lgb_parameter_rec, file="source/lgb_param_rec.qs")
          }
        }  
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/rf_param_rec.qs")){
          rf_parameter_rec = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/rf_param_rec.qs")
        } else {
          rf_spec <-
            parsnip::rand_forest() |> 
            parsnip::set_args(mtry = tune(), min_n = tune(), trees = 1000) |> 
            parsnip::set_engine(
              'ranger',
              num.threads = max(1, parallel::detectCores() - 1, na.rm = TRUE)
            ) |> 
            parsnip::set_mode('classification')            
          rf_result <- 
            Data_recipe_train |> 
            cv_fit_pred(rf_spec, Data_parameter_train, Data_parameter_test, Data_parameter_cv, param_space = rf_param_space)
          rf_parameter_rec = (rf_result$params_grid %>% show_best(metric = "roc_auc"))[1,]
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), rf_parameter_rec, file="source/rf_param_rec.qs")
          }
        }  
        set.seed(1212)
        lgb_model <-
          parsnip::boost_tree(mode = "classification") %>%
          set_engine("lightgbm", num_leaves = lgb_parameter_rec$num_leaves,
                     importance = TRUE,  # 変数重要度を有効化              
                     lambda_l1 = .9,
                     min_n = lgb_parameter_rec$min_n,
                     tree_depth = lgb_parameter_rec$tree_depth,
                     learn_rate = 0.05,
                     trees = 2000,
                     mtry = lgb_parameter_rec$mtry,
                     nthread = max(1, parallel::detectCores() - 1, na.rm = TRUE))
        lgb_wflow <- 
          workflow() %>% 
          add_model(lgb_model) %>% 
          add_recipe(Data_recipe)
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/lgb_m2_rec.qs")){
          lgb_m2_rec = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/lgb_m2_rec.qs")
        } else {
          lgb_m2_rec = lgb_wflow %>% fit(Data_ML) 
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), lgb_m2_rec, file = "source/lgb_m2_rec.qs")
          }
        }
        rf_model <-
          parsnip::rand_forest() |>
          parsnip::set_args(trees = 2000, min_n = rf_parameter_rec$min_n, mtry = rf_parameter_rec$mtry) |>
          parsnip::set_engine(
            'ranger', keep.inbag = TRUE, importance = "impurity",
            num.threads = max(1, parallel::detectCores() - 1, na.rm = TRUE)
          ) |>
          parsnip::set_mode('classification')
        rf_wflow <- 
          workflow() %>% 
          add_model(rf_model) %>% 
          add_recipe(Data_recipe)
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/rf_m2_rec.qs")){
          rf_m2_rec = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/rf_m2_rec.qs")
        } else {
          rf_m2_rec = rf_wflow %>% fit(Data_ML) 
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), rf_m2_rec, file = "source/rf_m2_rec.qs")
          }
        }
        X <- Data_ML %>% select(-EP_option)
        X <- if (nrow(X) > 10000) {
          X[sample(nrow(X), 10000), , drop = FALSE]
        } else {
          X
        }
        pred_wrapper <- function(object, newdata) {
          predict(object, new_data = newdata, type = "prob") %>% dplyr::pull(2)
        }
        set.seed(123)
        if(input$importance == "SHAP value (heavy analysis)"){
          if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/shap_values_lgb_rec.qs")){
            shap_values = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/shap_values_lgb_rec.qs")
          } else {
            shap_values <- fastshap::explain(parallel = ifelse(nrow(Data_ML) < 10000, TRUE, FALSE),
                                             # ncores = max(1, parallel::detectCores() - 1, na.rm = TRUE),
                                             object = lgb_m2_rec,
                                             X = X[seq_len(min(nrow(X), 5000)),],
                                             pred_wrapper = pred_wrapper,
                                             nsim = 3  # モンテカルロシミュレーション回数（計算精度と時間のバランスで調整）
            )
            if(input$intermediate_file != "No"){ 
              QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), shap_values, file = "source/shap_values_lgb_rec.qs")
            }
          }
          rm(lgb_m2_rec)
          # 各特徴量のSHAP値の符号（正/負）を判別し、色を付ける
          shap_values_with_sign <- shap_values
          shap_values_with_sign[] <- ifelse(shap_values_with_sign >= 0, "Positive", "Negative")
          # 各特徴量ごとの平均絶対値を計算
          mean_abs_shap <- colMeans(abs(shap_values))
          # データフレームに変換
          importance_df <- data.frame(
            feature = names(mean_abs_shap),
            mean_abs_shap = mean_abs_shap,
            sign = shap_values_with_sign[1,]  # ここでSHAP値の符号を代表として使う
          )
          importance_df = importance_df %>%
            dplyr::arrange(mean_abs_shap)
          shap_long <- as.data.frame(shap_values) %>%
            mutate(id = row_number()) %>%   # サンプルIDを追加
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "shap_value"
            )
          # ② Xの元データもlong形式に変換（行番号を付与）
          X_long <- X %>%
            mutate(id = row_number()) %>%
            mutate(across(where(is.factor),
                          ~ case_when(
                            . %in% c("Yes", "Male") ~ 1,
                            . %in% c("No", "Female") ~ 0,
                            TRUE ~ as.numeric(as.character(.))
                          ))) %>%
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "orig_value"
            ) %>%
            group_by(feature) %>%
            group_by(feature) %>%
            mutate(norm_value = case_when(
              max(orig_value) == min(orig_value) ~ orig_value,  # すべて同じならそのまま
              TRUE ~ (orig_value - min(orig_value)) / (max(orig_value) - min(orig_value))
            )) %>%
            ungroup()
          # ③ shap_longとX_longをidとfeatureで結合
          shap_long <- shap_long %>%
            left_join(X_long, by = c("id", "feature"))
          shap_long$feature = factor(shap_long$feature, levels = importance_df$feature)
          importance_df = importance_df[max(1, nrow(importance_df)-29):nrow(importance_df),]
          shap_long = shap_long %>%
            filter(feature %in% importance_df$feature)
          # 棒グラフでプロット（横向き）
          gSHAP1_lgb_rec <- ggplot(importance_df, aes(x = reorder(feature, mean_abs_shap), y = mean_abs_shap)) +
            geom_bar(stat = "identity") +
            coord_flip() +
            labs(title = "LightGBM, Mean |SHAP value| of Features", x = "Features", y = "Mean |SHAP value|") +
            scale_fill_manual(values = c("Positive" = "steelblue", "Negative" = "tomato")) +  # 色の指定
            theme_minimal() +
            theme(legend.title = element_blank(),  # 凡例タイトルを削除
                  legend.position = "bottom")  # 凡例を下に配置
          gSHAP2_lgb_rec = ggplot(shap_long, aes(x = shap_value, y = feature, color = norm_value)) +
            geom_quasirandom(alpha = 0.2, width = 0.2) +
            scale_color_gradient(low = "blue", high = "red", 
                                 limits = c(0, 1), 
                                 breaks = c(0, 1),
                                 labels = c("Low/No/Female", "High/Yes/Male")) +
            theme_minimal() +
            labs(title = "SHAP Value Distribution", color = "Feature Value")
          if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/shap_values_rf_rec.qs")){
            shap_values = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/shap_values_rf_rec.qs")
          } else {
            shap_values <- fastshap::explain(parallel = ifelse(nrow(Data_ML) < 10000, TRUE, FALSE),
                                             # ncores = max(1, parallel::detectCores() - 1, na.rm = TRUE),
                                             object = rf_m2_rec,
                                             X = X[seq_len(min(nrow(X), 5000)),],
                                             pred_wrapper = pred_wrapper,
                                             nsim = 3  # モンテカルロシミュレーション回数（計算精度と時間のバランスで調整）
            )
            if(input$intermediate_file != "No"){ 
              QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), shap_values, file = "source/shap_values_rf_rec.qs")
            }
          }
          rm(rf_m2_rec)
          # 各特徴量のSHAP値の符号（正/負）を判別し、色を付ける
          shap_values_with_sign <- shap_values
          shap_values_with_sign[] <- ifelse(shap_values_with_sign >= 0, "Positive", "Negative")
          # 各特徴量ごとの平均絶対値を計算
          mean_abs_shap <- colMeans(abs(shap_values))
          # データフレームに変換
          importance_df <- data.frame(
            feature = names(mean_abs_shap),
            mean_abs_shap = mean_abs_shap,
            sign = shap_values_with_sign[1,]  # ここでSHAP値の符号を代表として使う
          )
          importance_df = importance_df %>%
            dplyr::arrange(mean_abs_shap)
          shap_long <- as.data.frame(shap_values) %>%
            mutate(id = row_number()) %>%   # サンプルIDを追加
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "shap_value"
            )
          # ② Xの元データもlong形式に変換（行番号を付与）
          X_long <- X %>%
            mutate(id = row_number()) %>%
            mutate(across(where(is.factor),
                          ~ case_when(
                            . %in% c("Yes", "Male") ~ 1,
                            . %in% c("No", "Female") ~ 0,
                            TRUE ~ as.numeric(as.character(.))
                          ))) %>%
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "orig_value"
            ) %>%
            group_by(feature) %>%
            group_by(feature) %>%
            mutate(norm_value = case_when(
              max(orig_value) == min(orig_value) ~ orig_value,  # すべて同じならそのまま
              TRUE ~ (orig_value - min(orig_value)) / (max(orig_value) - min(orig_value))
            )) %>%
            ungroup()
          # ③ shap_longとX_longをidとfeatureで結合
          shap_long <- shap_long %>%
            left_join(X_long, by = c("id", "feature"))
          shap_long$feature = factor(shap_long$feature, levels = importance_df$feature)
          importance_df = importance_df[max(1, nrow(importance_df)-29):nrow(importance_df),]
          shap_long = shap_long %>%
            filter(feature %in% importance_df$feature)
          # 棒グラフでプロット（横向き）
          gSHAP1_rf_rec <- ggplot(importance_df, aes(x = reorder(feature, mean_abs_shap), y = mean_abs_shap)) +
            geom_bar(stat = "identity") +
            coord_flip() +
            labs(title = "RandomForest, Mean |SHAP value| of Features", x = "Features", y = "Mean |SHAP value|") +
            scale_fill_manual(values = c("Positive" = "steelblue", "Negative" = "tomato")) +  # 色の指定
            theme_minimal() +
            theme(legend.title = element_blank(),  # 凡例タイトルを削除
                  legend.position = "bottom")  # 凡例を下に配置
          gSHAP2_rf_rec = ggplot(shap_long, aes(x = shap_value, y = feature, color = norm_value)) +
            geom_quasirandom(alpha = 0.2, width = 0.2) +
            scale_color_gradient(low = "blue", high = "red", 
                                 limits = c(0, 1), 
                                 breaks = c(0, 1),
                                 labels = c("Low/No/Female", "High/Yes/Male")) +
            theme_minimal() +
            labs(title = "SHAP Value Distribution", color = "Feature Value")
        } else {
          importances <- map(Data_cv$splits, function(split) {
            lgb_fit <- fit(lgb_wflow, data = analysis(split))
            lgb_booster <- extract_fit_engine(lgb_fit)
            imp <- lgb.importance(lgb_booster)
            imp %>% select(Feature, Gain)  # 必要な列だけ
          })
          # 2. importance を平均化（split ごとの dataframe を結合 → group_by → summarise）
          lgb_vi <- bind_rows(importances) %>%
            group_by(Feature) %>%
            summarise(importance = mean(Gain, na.rm = TRUE)) %>%
            arrange(desc(importance))
          # 3. 可視化
          gSHAP1_lgb_rec <- ggplot(lgb_vi[seq_len(min(30,nrow(lgb_vi))),], aes(x = reorder(Feature, importance), y = importance)) +
            geom_bar(stat = "identity", fill = "steelblue") +
            coord_flip() +
            labs(
              title = "Top 30 Variable Importance from LightGBM (CV Averaged)",
              x = "Feature",
              y = "Importance"
            ) +
            theme_minimal()
          gSHAP2_lgb_rec = gg_empty()
          rf_importances <- map(Data_cv$splits, function(split) {
            rf_fit <- fit(rf_wflow, data = analysis(split))
            rf_model_fitted <- extract_fit_engine(rf_fit)
            # 変数重要度を tibble 形式で取得
            tibble(
              Feature = names(rf_model_fitted$variable.importance),
              Importance = rf_model_fitted$variable.importance
            )
          })
          # 重要度を平均化
          rf_vi <- bind_rows(rf_importances) %>%
            group_by(Feature) %>%
            summarise(importance = mean(Importance, na.rm = TRUE)) %>%
            arrange(desc(importance))
          # 3. 可視化
          gSHAP1_rf_rec <- ggplot(rf_vi[seq_len(min(30,nrow(rf_vi))),], aes(x = reorder(Feature, importance), y = importance)) +
            geom_bar(stat = "identity", fill = "steelblue") +
            coord_flip() +
            labs(
              title = "Top 30 Variable Importance from Random Forest (CV Averaged)",
              x = "Feature",
              y = "Importance"
            ) +
            theme_minimal()
          gSHAP2_rf_rec = gg_empty()
        }
        keep_pred <- control_resamples(save_pred = TRUE, save_workflow = TRUE)
        
        set.seed(1212)
        rf_res <- rf_wflow %>% fit_resamples(resamples = Data_cv, metrics = cls_met, control = keep_pred)
        rf_res_all = rf_res %>%
          pull(.predictions) %>%
          bind_rows() %>%
          dplyr::arrange(.row)
        # df_cv_pred_rec$rf  = rf_res_all$.pred_1
        ROC_rf_rec <- roc(EP_option ~ .pred_1, data = rf_res_all, ci = TRUE)
        gROC_rf_rec = ggroc(ROC_rf_rec, 
                        size = 1, #サイズ
                        legacy.axes = TRUE) + 
          geom_abline(color = "dark grey", size = 0.5) +
          theme_classic() +
          labs(
            title = 
              paste0("Random forest, AUC: ", round(ROC_rf_rec$auc[1],3), " (95%CI:",round(ROC_rf_rec$ci[1],3), "-", round(ROC_rf_rec$ci[3],3), ")"),
            subtitle = paste0("mtry/min_n=",rf_parameter_rec$mtry,"/",rf_parameter_rec$min_n)
          )
        
        set.seed(1212)
        lgb_res <- lgb_wflow %>% fit_resamples(resamples = Data_cv, metrics = cls_met, control = keep_pred)
        lgb_res_all = lgb_res %>%
          pull(.predictions) %>%
          bind_rows() %>%
          dplyr::arrange(.row)
        # df_cv_pred_rec$lgb  = lgb_res_all$.pred_1
        ROC_lgb_rec <- roc(EP_option ~ .pred_1, data = lgb_res_all, ci = TRUE)
        gROC_lgb_rec = ggroc(ROC_lgb_rec, 
                         size = 1, #サイズ
                         legacy.axes = TRUE) + 
          geom_abline(color = "dark grey", size = 0.5) +
          theme_classic() +
          labs(
            title = 
              paste0("LightGBM, AUC: ", round(ROC_lgb_rec$auc[1],3), " (95%CI:",round(ROC_lgb_rec$ci[1],3), "-", round(ROC_lgb_rec$ci[3],3), ")"),
            subtitle =   paste0("mtry/min_n/tree_depth/num_leaves=",lgb_parameter_rec$mtry,"/",lgb_parameter_rec$min_n,"/",lgb_parameter_rec$tree_depth,"/",lgb_parameter_rec$num_leaves)
          )
        
        g1_rec = collect_predictions(lgb_res) %>%
          ggplot(aes(.pred_1)) +
          geom_histogram(col = "white", bins = 40) +
          facet_wrap(~ EP_option, ncol = 1) +
          geom_rug(col = "blue", alpha = 1 / 2) + 
          labs(x = "LightGBM: Probability estimate of treatment recommendation")
        g2_rec = lgb_res %>% cal_plot_windowed(truth = EP_option, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("LightGBM: Calibration curve")
        set.seed(1212)
        iso_val <- cal_validate_logistic(lgb_res, metrics = cls_met,
                                         save_pred = TRUE, times = 25)
        # collect_metrics(iso_val)
        g3_rec = collect_predictions(iso_val) %>%
          filter(.type == "calibrated") %>%
          cal_plot_windowed(truth = EP_option, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("LightGBM: Logistic calibration via GAM")
        
        df_cv_pred_rec$lgb  = (iso_val %>%
                             pull(.predictions_cal) %>%
                             bind_rows() %>%
                             dplyr::arrange(.row))$.pred_1
        # beta_val <- lgb_res %>% cal_validate_beta(metrics = cls_met, save_pred = TRUE)
        # # collect_metrics(beta_val)
        # g3= collect_predictions(beta_val) %>%
        #   filter(.type == "calibrated") %>%
        #   cal_plot_windowed(truth = EP_option, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
        #   ggtitle("LightGBM: Beta calibration")
        # df_cv_pred_rec$lgb  = (beta_val %>%
        #                      pull(.predictions_cal) %>%
        #                      bind_rows() %>%
        #                      dplyr::arrange(.row))$.pred_1
        g4_rec = collect_predictions(rf_res) %>%
          ggplot(aes(.pred_1)) +
          geom_histogram(col = "white", bins = 40) +
          facet_wrap(~ EP_option, ncol = 1) +
          geom_rug(col = "blue", alpha = 1 / 2) + 
          labs(x = "Random forest: Probability estimate of treatment recommendation")
        g5_rec = rf_res %>% cal_plot_windowed(truth = EP_option, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("Random forest: Calibration curve")
        
        iso_val <- cal_validate_logistic(rf_res, metrics = cls_met,
                                         save_pred = TRUE, times = 25)
        collect_metrics(iso_val)
        g6_rec = collect_predictions(iso_val) %>%
          filter(.type == "calibrated") %>%
          cal_plot_windowed(truth = EP_option, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("Random forest: Logistic calibration via GAM")
        df_cv_pred_rec$rf  = (iso_val %>%
                            pull(.predictions_cal) %>%
                            bind_rows() %>%
                            dplyr::arrange(.row))$.pred_1
        # beta_val <- rf_res %>% cal_validate_beta(metrics = cls_met, save_pred = TRUE)
        # collect_metrics(beta_val)
        # g6= collect_predictions(beta_val) %>%
        #   filter(.type == "calibrated") %>%
        #   cal_plot_windowed(truth = EP_option, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
        #   ggtitle("Random forest: Beta calibration")
        # df_cv_pred_rec$rf  = (beta_val %>%
        #                      pull(.predictions_cal) %>%
        #                      bind_rows() %>%
        #                      dplyr::arrange(.row))$.pred_1
        
        output$figure_DCA_pre_CGP_1_rec = renderPlot({
          dcurves::dca( # calculate net benefit scores on mean cross validation predictions
            data = df_cv_pred_rec,
            formula = EP_option ~ .fitted + rf + lgb,
            thresholds = dca_thresholds,
            label = list(
              .fitted = "5-fold cross-validated logistic regression model",
              rf = "5-fold cross-validated random forest model",
              lgb = "5-fold cross-validated lightGBM model"
            ))$dca |> 
            # plot cross validated net benefit values
            ggplot(aes(x = threshold, y = net_benefit, color = label)) +
            stat_smooth(method = "loess", se = FALSE, formula = "y ~ x", span = 0.2, fullrange = TRUE) +
            scale_x_continuous(labels = scales::percent_format(accuracy = 1)) +
            scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
            labs(x = "Threshold probability", y = "Net Benefit", color = "") +
            coord_cartesian(xlim = c(0, 1), ylim = c(-0.01, NA)) + 
            theme_bw()
        })
        
        output$figure_DCA_pre_CGP_ROC_rec = renderPlot({
          gNomogram = df_cv_pred_rec %>%
            ggplot(aes(.fitted)) +
            theme_classic() +
            geom_histogram(col = "white", bins = 40) +
            facet_wrap(~ EP_option, ncol = 1) +
            geom_rug(col = "blue", alpha = 1 / 2) + 
            labs(x = "Logistic regression: Probability estimate of treatment recommendation")
          ROC <- roc(EP_option ~ .fitted, data = df_cv_pred_rec, ci = TRUE)
          gROC2 = ggroc(ROC, 
                        size = 1, #サイズ
                        legacy.axes = TRUE) + 
            geom_abline(color = "dark grey", size = 0.5) +
            theme_classic() +
            labs(
              title = 
                paste0("Logistic regression, AUC: ", round(ROC$auc[1],3), " (95%CI:",round(ROC$ci[1],3), "-", round(ROC$ci[3],3), ")")
            )
          ROC3 <- roc(EP_option ~ .fitted + lgb + rf, data = df_cv_pred_rec, ci = TRUE)
          gROC3 = ggroc(ROC3, 
                        size = 1, #サイズ
                        legacy.axes = TRUE) + 
            geom_abline(color = "dark grey", size = 0.5) +
            theme_classic() +
            scale_color_hue(name = "Prediction methods",
                            labels = c(.fitted = "Logistic regression",
                                       lgb = "Light GBM",
                                       rf  = "Random forest")) +
            labs(
              title = 
                paste0("ROC curves for treatment recommendation")
            )
          grid.arrange(gNomogram,gROC2,gROC3, nrow = 3)
        })
        
        output$figure_DCA_pre_CGP_Machine_learning_rec = renderPlot({
          grid.arrange(g1_rec,g2_rec,g3_rec,gROC_lgb_rec,gSHAP1_lgb_rec,gSHAP2_lgb_rec,g4_rec,g5_rec,g6_rec,gROC_rf_rec,gSHAP1_rf_rec,gSHAP2_rf_rec, nrow = 4, ncol=3)
        })          
        output$table_DCA_pre_CGP_2_rec = render_gt({
          (dcurves::dca( # calculate net benefit scores on mean cross validation predictions
            data = df_cv_pred_rec,
            formula = EP_option ~ .fitted + rf + lgb,
            thresholds = dca_thresholds,
            label = list(
              .fitted = "5-fold cross-validated logistic regression model",
              rf = "5-fold cross-validated random forest model",
              lgb = "5-fold cross-validated lightGBM model"
            ))) %>%
            net_intervention_avoided() %>%
            as_tibble() %>%
            gt::gt() %>%
            gt::fmt_percent(columns = threshold, decimals = 0) %>%
            gt::fmt(columns = net_benefit, fns = function(x) style_sigfig(x, digits = 3)) %>%
            gt::cols_label(
              label = "Strategy",
              threshold = "Decision Threshold",
              net_benefit = "Net Benefit"
            ) %>%
            gt::cols_align("left", columns = label)
        })
        
        output$table_prediction_rec = DT::renderDataTable(df_cv_pred_rec,
                                                      server = FALSE,
                                                      filter = 'top', 
                                                      extensions = c('Buttons'), 
                                                      options = list(pageLength = 100, 
                                                                     scrollX = TRUE,
                                                                     scrollY = "1000px",
                                                                     scrollCollapse = TRUE,
                                                                     dom="Blfrtip",
                                                                     buttons = c('csv', 'copy')))
      }
      if(length(colnames(Data_forest_tmp_8_GMT)) > 1 & input$machine_learning == "Yes"){
        Data_forest_tmp_8_GMT$patientid = 1:length(Data_forest_tmp_8_GMT$EP_option)
        dca_thresholds = seq(0, 1, 0.01)
        set.seed(1212)            
        # create a 5-fold cross validation set, 1 repeat which is the base case, change to suit your use case
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/glm_param_GMT.qs")){
          df_prediction_GMT = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/glm_param_GMT.qs")
        } else {
          cross_validation_samples <- rsample::vfold_cv(Data_forest_tmp_8_GMT, strata = "EP_option", v = 5, repeats = 1)
          formula_treat = formula(paste0("EP_option ~ `",paste(colnames(Data_forest_tmp_8_GMT)[!colnames(Data_forest_tmp_8_GMT) %in% c("EP_option", "patientid")], collapse="` + `"),"`"))
          df_crossval_predictions <-
            cross_validation_samples %>%
            dplyr::mutate(
              glm_analysis = c(list(lrm(formula = formula_treat, data = rsample::analysis(splits[[1]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[2]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[3]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[4]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               list(lrm(formula = formula_treat, data = rsample::analysis(splits[[5]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler))#,
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[6]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[7]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[8]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[9]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler)),
                               # list(lrm(formula = formula_treat, data = rsample::analysis(splits[[10]]),x=TRUE,y=TRUE,penalty = Penal, tol=Toler))
              )
            )
          
          df_prediction_id = c(rsample::assessment(df_crossval_predictions$splits[[1]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[2]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[3]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[4]])$patientid,
                               rsample::assessment(df_crossval_predictions$splits[[5]])$patientid#,
                               # rsample::assessment(df_crossval_predictions$splits[[6]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[7]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[8]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[9]])$patientid,
                               # rsample::assessment(df_crossval_predictions$splits[[10]])$patientid
          )
          df_prediction_score = c(predict(df_crossval_predictions$glm_analysis[[1]], rsample::assessment(df_crossval_predictions$splits[[1]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[2]], rsample::assessment(df_crossval_predictions$splits[[2]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[3]], rsample::assessment(df_crossval_predictions$splits[[3]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[4]], rsample::assessment(df_crossval_predictions$splits[[4]]),type = "fitted"),
                                  predict(df_crossval_predictions$glm_analysis[[5]], rsample::assessment(df_crossval_predictions$splits[[5]]),type = "fitted")#,
                                  # predict(df_crossval_predictions$glm_analysis[[6]], rsample::assessment(df_crossval_predictions$splits[[6]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[7]], rsample::assessment(df_crossval_predictions$splits[[7]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[8]], rsample::assessment(df_crossval_predictions$splits[[8]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[9]], rsample::assessment(df_crossval_predictions$splits[[9]]),type = "fitted"),
                                  # predict(df_crossval_predictions$glm_analysis[[10]], rsample::assessment(df_crossval_predictions$splits[[10]]),type = "fitted")
          )
          rm(df_crossval_predictions)
          
          df_prediction_GMT = data.frame(df_prediction_id, df_prediction_score)
          colnames(df_prediction_GMT) = c("patientid", ".fitted")
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), df_prediction_GMT, file="source/glm_param_GMT.qs")
          }
        }
        
        df_cv_pred_GMT <-
          Data_forest_tmp_8_GMT %>%
          dplyr::left_join(
            df_prediction_GMT,
            by = 'patientid'
          )
        if("PS" %in% colnames(Data_forest_tmp_8_GMT)){
          PS_data_GMT = data.frame(sort(unique(Data_forest_tmp_8_GMT$PS)))
          colnames(PS_data_GMT) = c("PS_type")
          save(file = "source/PS_data_GMT.rda", PS_data_GMT)
        }
        if("Lines" %in% colnames(Data_forest_tmp_8_GMT)){
          Lines_data_GMT = data.frame(sort(unique(Data_forest_tmp_8_GMT$Lines)))
          colnames(Lines_data_GMT) = c("Lines_type")
          save(file = "source/Lines_data_GMT.rda", Lines_data_GMT)
        }
        set.seed(1212)
        Data_ML = Data_ML_full %>%
          dplyr::mutate(EP_option = case_when(
            ID %in% ID_evidence_AB ~ 1,
            TRUE ~ 0
          ))
        Data_ML = Data_ML %>%
          dplyr::select(-EP_treat, -ID)
        common_cols = colnames(Data_forest_tmp_8_GMT)
        common_cols = common_cols[common_cols != "patientid"]
        Data_ML = Data_ML[, ..common_cols]
        Data_ML[, (names(Data_ML)) := lapply(.SD, function(x) if (is.character(x)) as.factor(x) else x)]
        factor_cols <- names(Data_ML)[sapply(Data_ML, is.factor)]
        factor_cols = factor_cols[factor_cols != "EP_option"]
        factor_cols = factor_cols[factor_cols %in% c("PS", "Panel", "Histology", "Best_effect", "Lines")]
        if (length(factor_cols) > 0) {
          Data_ML <- one_hot(Data_ML, cols = factor_cols, sparsifyNAs = FALSE, dropCols = TRUE)
        }
        Data_ML$EP_option = factor(Data_ML$EP_option)
        cls_met <- metric_set(roc_auc)
        Data_cv <- vfold_cv(v=5, Data_ML, strata = EP_option)
        Data_train = vfold_cv(v=5,
                              rsample::analysis(Data_cv$splits[[1]]) %>%
                                dplyr::slice_sample(n = min(nrow(rsample::analysis(Data_cv$splits[[1]])), 20000)),
                              strata = EP_option)
        Data_parameter_train = rsample::analysis(Data_train$splits[[1]])
        Data_parameter_test = rsample::assessment(Data_train$splits[[1]])
        Data_parameter_cv = Data_parameter_train |> 
          rsample::vfold_cv(v = 5, strata = EP_option)
        Data_recipe_train <- Data_parameter_train |>
          recipe(EP_option ~ .) |>
          step_nzv(all_predictors()) |>
          step_dummy(all_nominal_predictors())
        Data_recipe <- Data_ML |>
          recipe(EP_option ~ .) |>
          step_nzv(all_predictors()) |>
          step_dummy(all_nominal_predictors())
        # cv_fit_pred <- function(recipe, spec, df_train, df_test, df_cv = NULL, grid = 8) {
        #   if(is.null(df_cv)) {
        #     params_grid = NULL
        #     wf <- 
        #       workflow() |> 
        #       workflows::add_recipe(recipe = recipe) |> 
        #       workflows::add_model(spec = spec)
        #   } else {
        #     cv_wf <- 
        #       workflow() |> 
        #       workflows::add_recipe(recipe = recipe) |> 
        #       workflows::add_model(spec = spec)
        #     params_grid <- 
        #       cv_wf |> 
        #       tune::tune_grid(
        #         resamples = df_cv,
        #         grid = grid,
        #         control = tune::control_grid(save_pred = TRUE)
        #       )
        #     wf <- 
        #       cv_wf |> 
        #       tune::finalize_workflow(
        #         parameters = params_grid |> tune::select_best(metric = "roc_auc")
        #       )
        #   }
        #   wf_fit <- 
        #     wf |> 
        #     parsnip::fit(data = df_train)
        #   pred <- 
        #     wf_fit |> 
        #     tune::augment(new_data = df_test)
        #   return(
        #     list(
        #       params_grid = params_grid, 
        #       wf_fit = wf_fit, 
        #       pred = pred
        #     )
        #   )
        # }
        rf_param_space <- parameters(
          mtry(range = c(1, 5)),
          min_n(range = c(2, 50))
        )
        lgb_param_space <- parameters(
          mtry(range = c(1, 5)),
          tree_depth(range = c(2, 5)),
          min_n(range = c(2, 50)),
          num_leaves(range = c(10, 100))
        )
        cv_fit_pred <- function(recipe, spec, df_train, df_test, df_cv = NULL, iter = 500, init_grid_n = 8, param_space = NULL) {
          if (is.null(df_cv)) {
            wf <- workflow() |> 
              add_recipe(recipe) |> 
              add_model(spec)
            params_grid <- NULL
          } else {
            cv_wf <- workflow() |> 
              add_recipe(recipe) |> 
              add_model(spec)
            
            # 初期グリッド（ランダム or グリッド）
            initial_grid <- tune::tune_grid(
              cv_wf,
              resamples = df_cv,
              grid = init_grid_n,
              metrics = metric_set(roc_auc),
              control = control_grid(save_pred = TRUE)
            )
            
            # ベイズ最適化
            params_grid <- tune::tune_bayes(
              cv_wf,
              resamples = df_cv,
              param_info = param_space,
              initial = initial_grid,
              iter = iter,
              metrics = metric_set(roc_auc),
              control = tune::control_bayes(
                no_improve = 10,          # 改善がないと停止
                verbose = FALSE,
                save_pred = TRUE
              )
            )
            
            best_params <- params_grid |> tune::select_best(metric = "roc_auc")
            
            wf <- cv_wf |> tune::finalize_workflow(best_params)
          }
          
          wf_fit <- wf |> fit(data = df_train)
          pred <- wf_fit |> augment(new_data = df_test)
          
          return(list(
            params_grid = params_grid,
            wf_fit = wf_fit,
            pred = pred
          ))
        }
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/lgb_param_GMT.qs")){
          lgb_parameter_GMT = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/lgb_param_GMT.qs")
        } else {
          lgb_spec <-
            parsnip::boost_tree(mode = "classification") %>%
            parsnip::set_args(
              tree_depth = tune(), min_n = tune(), mtry = tune(), trees = 2000, learn_rate = 0.05, lambda_l1 = 0.9) |> 
            set_engine("lightgbm", num_leaves = tune(),
                       nthread = max(1, parallel::detectCores() - 1, na.rm = TRUE))
          lgb_result <- 
            Data_recipe_train |> 
            cv_fit_pred(lgb_spec, Data_parameter_train, Data_parameter_test, Data_parameter_cv, param_space = lgb_param_space)
          lgb_parameter_GMT = (lgb_result$params_grid %>% show_best(metric = "roc_auc"))[1,]
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), lgb_parameter_GMT, file="source/lgb_param_GMT.qs")
          }
        }  
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/rf_param_GMT.qs")){
          rf_parameter_GMT = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/rf_param_GMT.qs")
        } else {
          rf_spec <-
            parsnip::rand_forest() |> 
            parsnip::set_args(mtry = tune(), min_n = tune(), trees = 1000) |> 
            parsnip::set_engine(
              'ranger',
              num.threads = max(1, parallel::detectCores() - 1, na.rm = TRUE)
            ) |> 
            parsnip::set_mode('classification')            
          rf_result <- 
            Data_recipe_train |> 
            cv_fit_pred(rf_spec, Data_parameter_train, Data_parameter_test, Data_parameter_cv, param_space = rf_param_space)
          rf_parameter_GMT = (rf_result$params_grid %>% show_best(metric = "roc_auc"))[1,]
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), rf_parameter_GMT, file="source/rf_param_GMT.qs")
          }
        }  
        set.seed(1212)
        lgb_model <-
          parsnip::boost_tree(mode = "classification") %>%
          set_engine("lightgbm", num_leaves = lgb_parameter_GMT$num_leaves,
                     importance = TRUE,  # 変数重要度を有効化              
                     lambda_l1 = .9,
                     min_n = lgb_parameter_GMT$min_n,
                     tree_depth = lgb_parameter_GMT$tree_depth,
                     learn_rate = 0.05,
                     trees = 2000,
                     mtry = lgb_parameter_GMT$mtry,
                     nthread = max(1, parallel::detectCores() - 1, na.rm = TRUE))
        lgb_wflow <- 
          workflow() %>% 
          add_model(lgb_model) %>% 
          add_recipe(Data_recipe)
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/lgb_m2_GMT.qs")){
          lgb_m2_GMT = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/lgb_m2_GMT.qs")
        } else {
          lgb_m2_GMT = lgb_wflow %>% fit(Data_ML) 
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), lgb_m2_GMT, file = "source/lgb_m2_GMT.qs")
          }
        }
        rf_model <-
          parsnip::rand_forest() |>
          parsnip::set_args(trees = 2000, min_n = rf_parameter_GMT$min_n, mtry = rf_parameter_GMT$mtry) |>
          parsnip::set_engine(
            'ranger', keep.inbag = TRUE, importance = "impurity",
            num.threads = max(1, parallel::detectCores() - 1, na.rm = TRUE)
          ) |>
          parsnip::set_mode('classification')
        rf_wflow <- 
          workflow() %>% 
          add_model(rf_model) %>% 
          add_recipe(Data_recipe)
        if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/rf_m2_GMT.qs")){
          rf_m2_GMT = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/rf_m2_GMT.qs")
        } else {
          rf_m2_GMT = rf_wflow %>% fit(Data_ML) 
          if(input$intermediate_file != "No"){ 
            QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), rf_m2_GMT, file = "source/rf_m2_GMT.qs")
          }
        }
        X <- Data_ML %>% select(-EP_option)
        X <- if (nrow(X) > 10000) {
          X[sample(nrow(X), 10000), , drop = FALSE]
        } else {
          X
        }
        pred_wrapper <- function(object, newdata) {
          predict(object, new_data = newdata, type = "prob") %>% dplyr::pull(2)
        }
        set.seed(123)
        if(input$importance == "SHAP value (heavy analysis)"){
          if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/shap_values_lgb_GMT.qs")){
            shap_values = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/shap_values_lgb_GMT.qs")
          } else {
            shap_values <- fastshap::explain(parallel = ifelse(nrow(Data_ML) < 10000, TRUE, FALSE),
                                             # ncores = max(1, parallel::detectCores() - 1, na.rm = TRUE),
                                             object = lgb_m2_GMT,
                                             X = X[seq_len(min(nrow(X), 5000)),],
                                             pred_wrapper = pred_wrapper,
                                             nsim = 3  # モンテカルロシミュレーション回数（計算精度と時間のバランスで調整）
            )
            if(input$intermediate_file != "No"){ 
              QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), shap_values, file = "source/shap_values_lgb_GMT.qs")
            }
          }
          rm(lgb_m2_GMT)
          # 各特徴量のSHAP値の符号（正/負）を判別し、色を付ける
          shap_values_with_sign <- shap_values
          shap_values_with_sign[] <- ifelse(shap_values_with_sign >= 0, "Positive", "Negative")
          # 各特徴量ごとの平均絶対値を計算
          mean_abs_shap <- colMeans(abs(shap_values))
          # データフレームに変換
          importance_df <- data.frame(
            feature = names(mean_abs_shap),
            mean_abs_shap = mean_abs_shap,
            sign = shap_values_with_sign[1,]  # ここでSHAP値の符号を代表として使う
          )
          importance_df = importance_df %>%
            dplyr::arrange(mean_abs_shap)
          shap_long <- as.data.frame(shap_values) %>%
            mutate(id = row_number()) %>%   # サンプルIDを追加
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "shap_value"
            )
          # ② Xの元データもlong形式に変換（行番号を付与）
          X_long <- X %>%
            mutate(id = row_number()) %>%
            mutate(across(where(is.factor),
                          ~ case_when(
                            . %in% c("Yes", "Male") ~ 1,
                            . %in% c("No", "Female") ~ 0,
                            TRUE ~ as.numeric(as.character(.))
                          ))) %>%
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "orig_value"
            ) %>%
            group_by(feature) %>%
            group_by(feature) %>%
            mutate(norm_value = case_when(
              max(orig_value) == min(orig_value) ~ orig_value,  # すべて同じならそのまま
              TRUE ~ (orig_value - min(orig_value)) / (max(orig_value) - min(orig_value))
            )) %>%
            ungroup()
          # ③ shap_longとX_longをidとfeatureで結合
          shap_long <- shap_long %>%
            left_join(X_long, by = c("id", "feature"))
          shap_long$feature = factor(shap_long$feature, levels = importance_df$feature)
          importance_df = importance_df[max(1, nrow(importance_df)-29):nrow(importance_df),]
          shap_long = shap_long %>%
            filter(feature %in% importance_df$feature)
          # 棒グラフでプロット（横向き）
          gSHAP1_lgb_GMT <- ggplot(importance_df, aes(x = reorder(feature, mean_abs_shap), y = mean_abs_shap)) +
            geom_bar(stat = "identity") +
            coord_flip() +
            labs(title = "LightGBM, Mean |SHAP value| of Features", x = "Features", y = "Mean |SHAP value|") +
            scale_fill_manual(values = c("Positive" = "steelblue", "Negative" = "tomato")) +  # 色の指定
            theme_minimal() +
            theme(legend.title = element_blank(),  # 凡例タイトルを削除
                  legend.position = "bottom")  # 凡例を下に配置
          gSHAP2_lgb_GMT = ggplot(shap_long, aes(x = shap_value, y = feature, color = norm_value)) +
            geom_quasirandom(alpha = 0.2, width = 0.2) +
            scale_color_gradient(low = "blue", high = "red", 
                                 limits = c(0, 1), 
                                 breaks = c(0, 1),
                                 labels = c("Low/No/Female", "High/Yes/Male")) +
            theme_minimal() +
            labs(title = "SHAP Value Distribution", color = "Feature Value")
          if(input$new_analysis =="No, use the previous dataset" & input$intermediate_file != "No" & file.exists("source/shap_values_rf_GMT.qs")){
            shap_values = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/shap_values_rf_GMT.qs")
          } else {
            shap_values <- fastshap::explain(parallel = ifelse(nrow(Data_ML) < 10000, TRUE, FALSE),
                                             # ncores = max(1, parallel::detectCores() - 1, na.rm = TRUE),
                                             object = rf_m2_GMT,
                                             X = X[seq_len(min(nrow(X), 5000)),],
                                             pred_wrapper = pred_wrapper,
                                             nsim = 3  # モンテカルロシミュレーション回数（計算精度と時間のバランスで調整）
            )
            if(input$intermediate_file != "No"){ 
              QS_SAVE(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), shap_values, file = "source/shap_values_rf_GMT.qs")
            }
          }
          rm(rf_m2_GMT)
          # 各特徴量のSHAP値の符号（正/負）を判別し、色を付ける
          shap_values_with_sign <- shap_values
          shap_values_with_sign[] <- ifelse(shap_values_with_sign >= 0, "Positive", "Negative")
          # 各特徴量ごとの平均絶対値を計算
          mean_abs_shap <- colMeans(abs(shap_values))
          # データフレームに変換
          importance_df <- data.frame(
            feature = names(mean_abs_shap),
            mean_abs_shap = mean_abs_shap,
            sign = shap_values_with_sign[1,]  # ここでSHAP値の符号を代表として使う
          )
          importance_df = importance_df %>%
            dplyr::arrange(mean_abs_shap)
          shap_long <- as.data.frame(shap_values) %>%
            mutate(id = row_number()) %>%   # サンプルIDを追加
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "shap_value"
            )
          # ② Xの元データもlong形式に変換（行番号を付与）
          X_long <- X %>%
            mutate(id = row_number()) %>%
            mutate(across(where(is.factor),
                          ~ case_when(
                            . %in% c("Yes", "Male") ~ 1,
                            . %in% c("No", "Female") ~ 0,
                            TRUE ~ as.numeric(as.character(.))
                          ))) %>%
            pivot_longer(
              cols = -id,
              names_to = "feature",
              values_to = "orig_value"
            ) %>%
            group_by(feature) %>%
            group_by(feature) %>%
            mutate(norm_value = case_when(
              max(orig_value) == min(orig_value) ~ orig_value,  # すべて同じならそのまま
              TRUE ~ (orig_value - min(orig_value)) / (max(orig_value) - min(orig_value))
            )) %>%
            ungroup()
          # ③ shap_longとX_longをidとfeatureで結合
          shap_long <- shap_long %>%
            left_join(X_long, by = c("id", "feature"))
          shap_long$feature = factor(shap_long$feature, levels = importance_df$feature)
          importance_df = importance_df[max(1, nrow(importance_df)-29):nrow(importance_df),]
          shap_long = shap_long %>%
            filter(feature %in% importance_df$feature)
          # 棒グラフでプロット（横向き）
          gSHAP1_rf_GMT <- ggplot(importance_df, aes(x = reorder(feature, mean_abs_shap), y = mean_abs_shap)) +
            geom_bar(stat = "identity") +
            coord_flip() +
            labs(title = "RandomForest, Mean |SHAP value| of Features", x = "Features", y = "Mean |SHAP value|") +
            scale_fill_manual(values = c("Positive" = "steelblue", "Negative" = "tomato")) +  # 色の指定
            theme_minimal() +
            theme(legend.title = element_blank(),  # 凡例タイトルを削除
                  legend.position = "bottom")  # 凡例を下に配置
          gSHAP2_rf_GMT = ggplot(shap_long, aes(x = shap_value, y = feature, color = norm_value)) +
            geom_quasirandom(alpha = 0.2, width = 0.2) +
            scale_color_gradient(low = "blue", high = "red", 
                                 limits = c(0, 1), 
                                 breaks = c(0, 1),
                                 labels = c("Low/No/Female", "High/Yes/Male")) +
            theme_minimal() +
            labs(title = "SHAP Value Distribution", color = "Feature Value")
        } else {
          importances <- map(Data_cv$splits, function(split) {
            lgb_fit <- fit(lgb_wflow, data = analysis(split))
            lgb_booster <- extract_fit_engine(lgb_fit)
            imp <- lgb.importance(lgb_booster)
            imp %>% select(Feature, Gain)  # 必要な列だけ
          })
          # 2. importance を平均化（split ごとの dataframe を結合 → group_by → summarise）
          lgb_vi <- bind_rows(importances) %>%
            group_by(Feature) %>%
            summarise(importance = mean(Gain, na.rm = TRUE)) %>%
            arrange(desc(importance))
          # 3. 可視化
          gSHAP1_lgb_GMT <- ggplot(lgb_vi[seq_len(min(30,nrow(lgb_vi))),], aes(x = reorder(Feature, importance), y = importance)) +
            geom_bar(stat = "identity", fill = "steelblue") +
            coord_flip() +
            labs(
              title = "Top 30 Variable Importance from LightGBM (CV Averaged)",
              x = "Feature",
              y = "Importance"
            ) +
            theme_minimal()
          gSHAP2_lgb_GMT = gg_empty()
          rf_importances <- map(Data_cv$splits, function(split) {
            rf_fit <- fit(rf_wflow, data = analysis(split))
            rf_model_fitted <- extract_fit_engine(rf_fit)
            # 変数重要度を tibble 形式で取得
            tibble(
              Feature = names(rf_model_fitted$variable.importance),
              Importance = rf_model_fitted$variable.importance
            )
          })
          # 重要度を平均化
          rf_vi <- bind_rows(rf_importances) %>%
            group_by(Feature) %>%
            summarise(importance = mean(Importance, na.rm = TRUE)) %>%
            arrange(desc(importance))
          # 3. 可視化
          gSHAP1_rf_GMT <- ggplot(rf_vi[seq_len(min(30,nrow(rf_vi))),], aes(x = reorder(Feature, importance), y = importance)) +
            geom_bar(stat = "identity", fill = "steelblue") +
            coord_flip() +
            labs(
              title = "Top 30 Variable Importance from Random Forest (CV Averaged)",
              x = "Feature",
              y = "Importance"
            ) +
            theme_minimal()
          gSHAP2_rf_GMT = gg_empty()
        }
        keep_pred <- control_resamples(save_pred = TRUE, save_workflow = TRUE)
        
        set.seed(1212)
        rf_res <- rf_wflow %>% fit_resamples(resamples = Data_cv, metrics = cls_met, control = keep_pred)
        rf_res_all = rf_res %>%
          pull(.predictions) %>%
          bind_rows() %>%
          dplyr::arrange(.row)
        # df_cv_pred_GMT$rf  = rf_res_all$.pred_1
        ROC_rf_GMT <- roc(EP_option ~ .pred_1, data = rf_res_all, ci = TRUE)
        gROC_rf_GMT = ggroc(ROC_rf_GMT, 
                            size = 1, #サイズ
                            legacy.axes = TRUE) + 
          geom_abline(color = "dark grey", size = 0.5) +
          theme_classic() +
          labs(
            title = 
              paste0("Random forest, AUC: ", round(ROC_rf_GMT$auc[1],3), " (95%CI:",round(ROC_rf_GMT$ci[1],3), "-", round(ROC_rf_GMT$ci[3],3), ")"),
            subtitle = paste0("mtry/min_n=",rf_parameter_GMT$mtry,"/",rf_parameter_GMT$min_n)
          )
        
        set.seed(1212)
        lgb_res <- lgb_wflow %>% fit_resamples(resamples = Data_cv, metrics = cls_met, control = keep_pred)
        lgb_res_all = lgb_res %>%
          pull(.predictions) %>%
          bind_rows() %>%
          dplyr::arrange(.row)
        # df_cv_pred_rec$lgb  = lgb_res_all$.pred_1
        ROC_lgb_GMT <- roc(EP_option ~ .pred_1, data = lgb_res_all, ci = TRUE)
        gROC_lgb_GMT = ggroc(ROC_lgb_GMT, 
                             size = 1, #サイズ
                             legacy.axes = TRUE) + 
          geom_abline(color = "dark grey", size = 0.5) +
          theme_classic() +
          labs(
            title = 
              paste0("LightGBM, AUC: ", round(ROC_lgb_GMT$auc[1],3), " (95%CI:",round(ROC_lgb_GMT$ci[1],3), "-", round(ROC_lgb_GMT$ci[3],3), ")"),
            subtitle = paste0("mtry/min_n/tree_depth/num_leaves=",lgb_parameter_GMT$mtry,"/",lgb_parameter_GMT$min_n,"/",lgb_parameter_GMT$tree_depth,"/",lgb_parameter_GMT$num_leaves)
          )
        
        g1_GMT = collect_predictions(lgb_res) %>%
          ggplot(aes(.pred_1)) +
          geom_histogram(col = "white", bins = 40) +
          facet_wrap(~ EP_option, ncol = 1) +
          geom_rug(col = "blue", alpha = 1 / 2) + 
          labs(x = "LightGBM: Probability estimate of evidence level A/B")
        g2_GMT = lgb_res %>% cal_plot_windowed(truth = EP_option, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("LightGBM: Calibration curve")
        set.seed(1212)
        iso_val <- cal_validate_logistic(lgb_res, metrics = cls_met,
                                         save_pred = TRUE, times = 25)
        # collect_metrics(iso_val)
        g3_GMT = collect_predictions(iso_val) %>%
          filter(.type == "calibrated") %>%
          cal_plot_windowed(truth = EP_option, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("LightGBM: Logistic calibration via GAM")
        
        df_cv_pred_GMT$lgb  = (iso_val %>%
                                 pull(.predictions_cal) %>%
                                 bind_rows() %>%
                                 dplyr::arrange(.row))$.pred_1
        # beta_val <- lgb_res %>% cal_validate_beta(metrics = cls_met, save_pred = TRUE)
        # # collect_metrics(beta_val)
        # g3= collect_predictions(beta_val) %>%
        #   filter(.type == "calibrated") %>%
        #   cal_plot_windowed(truth = EP_treat, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
        #   ggtitle("LightGBM: Beta calibration")
        # df_cv_pred_GMT$lgb  = (beta_val %>%
        #                      pull(.predictions_cal) %>%
        #                      bind_rows() %>%
        #                      dplyr::arrange(.row))$.pred_1
        g4_GMT = collect_predictions(rf_res) %>%
          ggplot(aes(.pred_1)) +
          geom_histogram(col = "white", bins = 40) +
          facet_wrap(~ EP_option, ncol = 1) +
          geom_rug(col = "blue", alpha = 1 / 2) + 
          labs(x = "Random forest: Probability estimate of evidence level A/B")
        g5_GMT = rf_res %>% cal_plot_windowed(truth = EP_option, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("Random forest: Calibration curve")
        
        iso_val <- cal_validate_logistic(rf_res, metrics = cls_met,
                                         save_pred = TRUE, times = 25)
        collect_metrics(iso_val)
        g6_GMT = collect_predictions(iso_val) %>%
          filter(.type == "calibrated") %>%
          cal_plot_windowed(truth = EP_option, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
          ggtitle("Random forest: Logistic calibration via GAM")
        df_cv_pred_GMT$rf  = (iso_val %>%
                                pull(.predictions_cal) %>%
                                bind_rows() %>%
                                dplyr::arrange(.row))$.pred_1
        # beta_val <- rf_res %>% cal_validate_beta(metrics = cls_met, save_pred = TRUE)
        # collect_metrics(beta_val)
        # g6= collect_predictions(beta_val) %>%
        #   filter(.type == "calibrated") %>%
        #   cal_plot_windowed(truth = EP_treat, event_level = 'second', estimate = .pred_1, step_size = 0.025) +
        #   ggtitle("Random forest: Beta calibration")
        # df_cv_pred_GMT$rf  = (beta_val %>%
        #                      pull(.predictions_cal) %>%
        #                      bind_rows() %>%
        #                      dplyr::arrange(.row))$.pred_1
        
        output$figure_DCA_pre_CGP_1_GMT = renderPlot({
          dcurves::dca( # calculate net benefit scores on mean cross validation predictions
            data = df_cv_pred_GMT,
            formula = EP_option ~ .fitted + rf + lgb,
            thresholds = dca_thresholds,
            label = list(
              .fitted = "5-fold cross-validated logistic regression model",
              rf = "5-fold cross-validated random forest model",
              lgb = "5-fold cross-validated lightGBM model"
            ))$dca |> 
            # plot cross validated net benefit values
            ggplot(aes(x = threshold, y = net_benefit, color = label)) +
            stat_smooth(method = "loess", se = FALSE, formula = "y ~ x", span = 0.2, fullrange = TRUE) +
            scale_x_continuous(labels = scales::percent_format(accuracy = 1)) +
            scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
            labs(x = "Threshold probability", y = "Net Benefit", color = "") +
            coord_cartesian(xlim = c(0, 0.6), ylim = c(-0.01, NA)) + 
            theme_bw()
        })
        
        output$figure_DCA_pre_CGP_ROC_GMT = renderPlot({
          gNomogram = df_cv_pred_GMT %>%
            ggplot(aes(.fitted)) +
            theme_classic() +
            geom_histogram(col = "white", bins = 40) +
            facet_wrap(~ EP_option, ncol = 1) +
            geom_rug(col = "blue", alpha = 1 / 2) + 
            labs(x = "Logistic regression: Probability estimate of evidence level A/B")
          ROC <- roc(EP_option ~ .fitted, data = df_cv_pred_GMT, ci = TRUE)
          gROC2 = ggroc(ROC, 
                        size = 1, #サイズ
                        legacy.axes = TRUE) + 
            geom_abline(color = "dark grey", size = 0.5) +
            theme_classic() +
            labs(
              title = 
                paste0("Logistic regression, AUC: ", round(ROC$auc[1],3), " (95%CI:",round(ROC$ci[1],3), "-", round(ROC$ci[3],3), ")")
            )
          ROC3 <- roc(EP_option ~ .fitted + lgb + rf, data = df_cv_pred_GMT, ci = TRUE)
          gROC3 = ggroc(ROC3, 
                        size = 1, #サイズ
                        legacy.axes = TRUE) + 
            geom_abline(color = "dark grey", size = 0.5) +
            theme_classic() +
            scale_color_hue(name = "Prediction methods",
                            labels = c(.fitted = "Logistic regression",
                                       lgb = "Light GBM",
                                       rf  = "Random forest")) +
            labs(
              title = 
                paste0("ROC curves for evidence level A/B")
            )
          grid.arrange(gNomogram,gROC2,gROC3, nrow = 3)
        })
        
        output$figure_DCA_pre_CGP_Machine_learning_GMT = renderPlot({
          grid.arrange(g1_GMT,g2_GMT,g3_GMT,gROC_lgb_GMT,gSHAP1_lgb_GMT,gSHAP2_lgb_GMT,g4_GMT,g5_GMT,g6_GMT,gROC_rf_GMT,gSHAP1_rf_GMT,gSHAP2_rf_GMT, nrow = 4, ncol=3)
        })          
        output$table_DCA_pre_CGP_2_GMT = render_gt({
          (dcurves::dca( # calculate net benefit scores on mean cross validation predictions
            data = df_cv_pred_GMT,
            formula = EP_option ~ .fitted + rf + lgb,
            thresholds = dca_thresholds,
            label = list(
              .fitted = "5-fold cross-validated logistic regression model",
              rf = "5-fold cross-validated random forest model",
              lgb = "5-fold cross-validated lightGBM model"
            ))) %>%
            net_intervention_avoided() %>%
            as_tibble() %>%
            gt::gt() %>%
            gt::fmt_percent(columns = threshold, decimals = 0) %>%
            gt::fmt(columns = net_benefit, fns = function(x) style_sigfig(x, digits = 3)) %>%
            gt::cols_label(
              label = "Strategy",
              threshold = "Decision Threshold",
              net_benefit = "Net Benefit"
            ) %>%
            gt::cols_align("left", columns = label)
        })
        
        output$table_prediction_GMT = DT::renderDataTable(df_cv_pred_GMT,
                                                          server = FALSE,
                                                          filter = 'top', 
                                                          extensions = c('Buttons'), 
                                                          options = list(pageLength = 100, 
                                                                         scrollX = TRUE,
                                                                         scrollY = "1000px",
                                                                         scrollCollapse = TRUE,
                                                                         dom="Blfrtip",
                                                                         buttons = c('csv', 'copy')))
      }
      incProgress(1 / 13)
    })
  })


  observeEvent(input$predict_nomogram_setting, {
    withProgress(message = sample(nietzsche)[1], {
      incProgress(1 / 13)
      if(input$intermediate_file != "No" & file.exists("source/m2.qs")){
        tmp_post$m2 = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/m2.qs")
      }
      if(input$intermediate_file != "No" & file.exists("source/lgb_m2.qs")){
        tmp_post$lgb_m2 = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/lgb_m2.qs")
      }
      if(input$intermediate_file != "No" & file.exists("source/rf_m2.qs")){
        tmp_post$rf_m2 = QS_READ(nthreads = max(1, parallel::detectCores() - 1, na.rm = TRUE), file="source/rf_m2.qs")
      }
      load("source/Organ_data.rda")
      output$input_age = renderUI({
        numericInput("predict_age", "Age",
                     value = 60,
                     min = 1,
                     max = 120,
                     step = 1)
      })
      output$input_sex = renderUI({ 
        radioButtons(
          "predict_Sex", "Sex",
          choices = c("Male", "Female"),
          selected = "Male")
      })
      output$input_Lymph_met = renderUI({ 
        radioButtons(
          "predict_Lymph_met", "Lymph node metastasis",
          choices = c("Yes", "No"),
          selected = "No")
      })
      output$input_Brain_met = renderUI({ 
        radioButtons(
          "predict_Brain_met", "Brain metastasis",
          choices = c("Yes", "No"),
          selected = "No")
      })
      output$input_Lung_met = renderUI({ 
        radioButtons(
          "predict_Lung_met", "Lung metastasis",
          choices = c("Yes", "No"),
          selected = "No")
      })
      output$input_Bone_met = renderUI({ 
        radioButtons(
          "predict_Bone_met", "Bone metastasis",
          choices = c("Yes", "No"),
          selected = "No")
      })
      output$input_Liver_met = renderUI({ 
        radioButtons(
          "predict_Liver_met", "Liver metastasis",
          choices = c("Yes", "No"),
          selected = "No")
      })
      output$input_PS = renderUI({ 
        radioButtons(
          "predict_PS", "Performance status",
          choices = c("0", "1", "2_4"),
          selected = "0")
      })
      output$input_Lines = renderUI({ 
        radioButtons(
          "predict_Lines", "CTx line",
          choices = c("1", "2", "3~"),
          selected = "1")
      })
      output$input_Best_effect = renderUI({ 
        radioButtons(
          "predict_Best_effect", "Best CTx effect in previous lines",
          choices = c("CR", "PR", "SD", "PD", "Unknown", "CTx_naive"),
          selected = "SD")
      })
      output$input_organ = renderUI({
        pickerInput("predict_organ", "Histology",
                    choices = Organ_data$Organ_type_raw,
                    selected = Organ_data$Organ_type_raw[1], multiple = FALSE)
      })
      output$input_Panel = renderUI({ 
        radioButtons(
          "predict_Panel", "Panel",
          choices = c("Solid", "Liquid"),
          selected = "Solid")
      })
    })
  })
  observeEvent(input$start_prediction, {
    withProgress(message = sample(nietzsche)[1], {
      incProgress(1 / 13)
      load("source/Organ_data.rda")
      predict_age = ifelse(input$predict_age <= input$mid_age, "Younger", "Older")
      predict_PS = input$predict_PS
      predict_Lines = input$predict_Lines
      predict_Histology = Organ_data[Organ_data$Organ_type_raw == input$predict_organ,]$Organ_type
      
      if(file.exists("source/PS_data.rda")){
        load("source/PS_data.rda")
        PS_choice = PS_data$PS_type
      } else {
        PS_choice = c("0", "1", "2_4")
      }
      if(file.exists("source/Lines_data.rda")){
        load("source/Lines_data.rda")
        Lines_choice = Lines_data$Lines_type
      } else {
        Lines_choice = c("1", "2", "3~")
      }
      if(length(PS_choice) == 2){
        predict_PS = ifelse(predict_PS == "0", "0", "1_4")
      }
      if(length(Lines_choice) == 2){
        if(all(Lines_choice == c("1~2", "3~"))){
          predict_Lines = ifelse(predict_PS == "3~", "3~", "1~2")
        } else {
          predict_Lines = ifelse(predict_Lines == "1", "1", "2~")
        }
      }
      input_data = data.frame(
          Age = predict_age,
          Lymph_met = input$predict_Lymph_met,
          Brain_met = input$predict_Brain_met,
          Lung_met = input$predict_Lung_met,
          Bone_met = input$predict_Bone_met,
          Liver_met = input$predict_Liver_met,
          Sex = input$predict_Sex,
          Histology = predict_Histology,
          PS = predict_PS,
          Lines = predict_Lines,
          Best_effect = input$predict_Best_effect,
          Panel = input$predict_Panel
      )
      input_data_ML = data.frame(
        Age_raw = input$predict_age,
        Lymph_met = input$predict_Lymph_met,
        Brain_met = input$predict_Brain_met,
        Lung_met = input$predict_Lung_met,
        Bone_met = input$predict_Bone_met,
        Liver_met = input$predict_Liver_met,
        Sex = input$predict_Sex,
        Histology = input$predict_organ,
        PS = input$predict_PS,
        Lines = input$predict_Lines,
        Best_effect = input$predict_Best_effect,
        Panel = input$predict_Panel
      )
      if(!is.null(tmp_post$m2)){
        predicted_score = predict(tmp_post$m2, input_data, se.fit=TRUE)
      }
      if(!is.null(tmp_post$lgb_m2)){
        predicted_score_lgb = predict(tmp_post$lgb_m2, input_data_ML, type = "prob")
      }
      if(!is.null(tmp_post$rf_m2)){
        predicted_score_rf = predict(tmp_post$rf_m2, input_data_ML, type = "prob")
        predicted_score_rf_CI = predict(tmp_post$rf_m2, input_data_ML, type = "conf_int")
      }

      output$prediction_nomogram = renderText({
        if(!is.null(tmp_post$m2) &
           !is.null(tmp_post$lgb_m2) &
           !is.null(tmp_post$rf_m2)){
          paste0("Predicted targeted therapy reach rate\n",
               "Logistic regression: ", round(exp(predicted_score[[1]][[1]])/(1+exp(predicted_score[[1]][[1]])), digits=2)*100,
               "% (95%CI: ",
               round(exp(predicted_score[[1]][[1]] - 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])/(1+exp(predicted_score[[1]][[1]] - 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])), digits=2)*100, "-",
               round(exp(predicted_score[[1]][[1]] + 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])/(1+exp(predicted_score[[1]][[1]] + 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])), digits=2)*100,")\n",
               "LightGBM: ", round(predicted_score_lgb[2], digits=2)*100, "%\n",
               "Random forest: ", round(predicted_score_rf[2], digits=2)*100, "% (95%CI: ",
               round(predicted_score_rf_CI[[3]], digits=2)*100, "-",
               round(predicted_score_rf_CI[[4]], digits=2)*100,")\n")
        } else if(!is.null(tmp_post$m2) &
                  !is.null(tmp_post$lgb_m2)){
          paste0("Predicted targeted therapy reach rate\n",
                 "Logistic regression: ", round(exp(predicted_score[[1]][[1]])/(1+exp(predicted_score[[1]][[1]])), digits=2)*100,
                 "% (95%CI: ",
                 round(exp(predicted_score[[1]][[1]] - 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])/(1+exp(predicted_score[[1]][[1]] - 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])), digits=2)*100, "-",
                 round(exp(predicted_score[[1]][[1]] + 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])/(1+exp(predicted_score[[1]][[1]] + 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])), digits=2)*100,")\n",
                 "LightGBM: ", round(predicted_score_lgb[2], digits=2)*100, "%\n")
        } else if(!is.null(tmp_post$m2) &
                  !is.null(tmp_post$rf_m2)){
          paste0("Predicted targeted therapy reach rate\n",
                 "Logistic regression: ", round(exp(predicted_score[[1]][[1]])/(1+exp(predicted_score[[1]][[1]])), digits=2)*100,
                 "% (95%CI: ",
                 round(exp(predicted_score[[1]][[1]] - 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])/(1+exp(predicted_score[[1]][[1]] - 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])), digits=2)*100, "-",
                 round(exp(predicted_score[[1]][[1]] + 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])/(1+exp(predicted_score[[1]][[1]] + 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])), digits=2)*100,")\n",
                 "Random forest: ", round(predicted_score_rf[2], digits=2)*100, "% (95%CI: ",
                 round(predicted_score_rf_CI[[3]], digits=2)*100, "-",
                 round(predicted_score_rf_CI[[4]], digits=2)*100,")\n")
        } else {
          paste0("Predicted targeted therapy reach rate\n",
                 "Logistic regression: ", round(exp(predicted_score[[1]][[1]])/(1+exp(predicted_score[[1]][[1]])), digits=2)*100,
                 "% (95%CI: ",
                 round(exp(predicted_score[[1]][[1]] - 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])/(1+exp(predicted_score[[1]][[1]] - 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])), digits=2)*100, "-",
                 round(exp(predicted_score[[1]][[1]] + 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])/(1+exp(predicted_score[[1]][[1]] + 1.96 * predict(tmp_post$m2, input_data, se.fit=TRUE)[[2]][[1]])), digits=2)*100,")\n")
        }
      })
      incProgress(1 / 13)
    })
  })
  
  observeEvent(input$survival_CTx_analysis, {
    withProgress(message = sample(nietzsche)[1], {
      library(cmdstanr)
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      Data_case_target = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.基本情報.年齢,
                      Lymph_met,
                      Brain_met,
                      Lung_met,
                      Bone_met,
                      Liver_met,
                      Other_met,
                      EP_option,
                      EP_treat,
                      YoungOld,
                      症例.EP前レジメン情報.化学療法レジメン名称,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.終了理由.名称.,
                      症例.EP前レジメン情報.最良総合効果.名称.,
                      症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                      症例.EP後レジメン情報.治療ライン,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.化学療法レジメン名称,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      症例.EP後レジメン情報.レジメン継続区分.名称.,
                      症例.EP後レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.最良総合効果.名称.,
                      症例.基本情報.性別.名称.,
                      症例.基本情報.がん種.OncoTree.,
                      症例.基本情報.がん種.OncoTree..名称.,
                      症例.基本情報.がん種.OncoTree.LEVEL1.,
                      症例.検体情報.パネル.名称.,
                      症例.背景情報.ECOG.PS.名称.,
                      症例.背景情報.喫煙歴有無.名称.,
                      症例.背景情報.アルコール多飲有無.名称.,
                      症例.背景情報.重複がん有無.異なる臓器..名称.,
                      症例.背景情報.多発がん有無.同一臓器..名称.,
                      症例.がん種情報.登録時転移部位.名称.,
                      症例.検体情報.腫瘍細胞含有割合,
                      症例.検体情報.検体採取部位.名称.,
                      症例.転帰情報.転帰.名称.,
                      症例.背景情報.診断日,
                      症例.管理情報.登録日,
                      症例.転帰情報.最終生存確認日,
                      症例.転帰情報.死亡日
        )
      incProgress(1 / 13)

      Data_case_target$Cancers = Data_case_target$症例.基本情報.がん種.OncoTree.
      Data_survival = Data_case_target %>%
        dplyr::mutate(final_observe = case_when(
          症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
          症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
          症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
          TRUE ~ 症例.管理情報.登録日))
      Data_survival = Data_survival %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::mutate(censor = case_when(
          症例.転帰情報.死亡日 != "" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
      
      Data_survival = Data_survival %>%
        dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
        dplyr::mutate(final_observe = ifelse(is.na(final_observe), 症例.管理情報.登録日, final_observe))
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, censor) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival = Data_survival %>%
        dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(Data_survival$症例.管理情報.登録日)) %>%
        dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                        as.Date(症例.EP前レジメン情報.投与開始日))
      
      Data_survival = Data_survival %>%
        dplyr::mutate(time_enroll_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.管理情報.登録日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.EP前レジメン情報.投与開始日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_enroll =
                        Data_survival$time_palliative_final -
                        Data_survival$time_enroll_final)
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, final_observe) %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_enroll_final) %>%
        dplyr::arrange(desc(time_enroll_final)) %>%
        dplyr::filter(time_enroll_final > 0 & !is.na(time_enroll_final) & is.finite(time_enroll_final)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      incProgress(1 / 13)
      
      Data_survival_tmp = Data_survival %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                        c("その他", "緩和")) %>%
        dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_palliative_enroll, time_palliative_final) %>%
        dplyr::filter(time_palliative_enroll > 0 & time_palliative_enroll < 10000 & !is.na(time_palliative_enroll) & is.finite(time_palliative_enroll)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_case_target = Data_case_target %>%
        dplyr::distinct(.keep_all = TRUE, C.CAT調査結果.基本項目.ハッシュID)
      Data_MAF = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","A","B","C","D","E","F") &
            Variant_Classification != "expression"
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
       if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
        Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
      }

      Data_report_TMB = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
        dplyr::select(Tumor_Sample_Barcode, TMB) %>%
        dplyr::distinct(Tumor_Sample_Barcode, .keep_all=T)
      colnames(Data_report_TMB) = c("C.CAT調査結果.基本項目.ハッシュID", "TMB")
      Data_case_target = left_join(Data_case_target, Data_report_TMB,
                                   by = "C.CAT調査結果.基本項目.ハッシュID")
      
      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      if(input$patho == "Only pathogenic muts"){
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(Evidence_level == "F")
      }
      Data_MAF_target = Data_MAF_target %>%
        dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol, .keep_all = T)
      incProgress(1 / 13)
      Gene_list = unique(names(sort(table(Data_MAF_target$Hugo_Symbol),
                                    decreasing = T)))

      if(length(Data_case_target$censor[is.na(Data_case_target$censor)])> 0){
        Data_case_target$censor[is.na(Data_case_target$censor)] = 0
      }
      Data_survival = Data_case_target %>% dplyr::filter(
        !is.na(time_palliative_enroll) &
        !is.na(time_enroll_final) &
        is.finite(time_palliative_enroll) &
        is.finite(time_enroll_final) & 
        time_palliative_enroll > 0 &
        time_enroll_final > 0 &
        !is.na(censor) &
        is.finite(censor)
      )
      Data_survival_curve = Data_survival
      if(sum(Data_survival_curve$censor) > 0){
        fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_curve)
        time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
        survival_rate_10 = summary(fit,times = time_10)[[6]]
        survival_rate_90 = summary(fit,times = max(Data_survival_curve$time_enroll_final)*0.9)[[6]]
      } else {
        time_10 = 90
        survival_rate_10 = 0
        survival_rate_90 = 1
      }
      time_90 = max(time_10 + 1, max(Data_survival_curve$time_enroll_final)*0.9)
      Data_survival_tmp = Data_survival_curve %>% dplyr::filter(
        time_enroll_final >= time_10 &
          time_enroll_final <= time_90)
      Data_survival_tmp$time_enroll_final =
        Data_survival_tmp$time_enroll_final - (time_10 - 1)
      Data_survival_tmp2 = Data_survival_curve %>% dplyr::filter(
        time_enroll_final > time_90) %>%
        dplyr::arrange(time_enroll_final)
      Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
      idx <- 1:nrow(Data_survival_tmp2)
      Data_drug_survival_1 = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
      #Data_survival$time_palliative_final = Data_survival$time_palliative_enroll + Data_survival$time_enroll_final

      if(sum(Data_survival_curve$censor) > 0){
        fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_curve)
        time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
        survival_rate_10 = summary(fit,times = time_10)[[6]]
        survival_rate_90 = summary(fit,times = max(Data_survival_curve$time_palliative_enroll)*0.9)[[6]]
      } else {
        time_10 = 90
        survival_rate_10 = 0
        survival_rate_90 = 1
      }
      time_90 = max(time_10 + 1, max(Data_survival_curve$time_palliative_enroll)*0.9)
      Data_survival_tmp5 = Data_survival_curve %>% dplyr::filter(
        time_palliative_enroll >= time_10 &
          time_palliative_enroll <= time_90)
      Data_survival_tmp5$time_palliative_enroll =
        Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
      Data_survival_tmp6 = Data_survival_curve %>% dplyr::filter(
        time_palliative_enroll > time_90) %>%
        dplyr::arrange(time_palliative_enroll)
      Data_survival_tmp6$time_palliative_enroll = time_90 - (time_10 - 1)
      idx <- 1:nrow(Data_survival_tmp6)
      Data_drug_survival_2 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
      
      Median_entry = median(Data_survival_tmp5$time_palliative_enroll)
      Data_MAF_target = Data_MAF_target %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_survival_curve$C.CAT調査結果.基本項目.ハッシュID)
      
      Data_drug_survival_1 = Data_drug_survival_1 %>%
        dplyr::mutate(delay_1 = case_when(
          Data_drug_survival_1$time_palliative_enroll <= Median_entry ~ "SPT to entry early",
          TRUE ~ "SPT to entry later"))
      Cancername = sort(unique(Data_survival_curve$Cancers))
      Total_pts = as.vector(table((Data_survival_curve %>%
                                     dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = T))$Cancers))

      incProgress(1 / 13)
      stan_weibull_survival_model_data <-
        list(
          Median = as.integer(Median_entry),
          Nobs_early = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later"),
          Nobs_late = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later"),
          Ncen_early = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later"),
          Ncen_late = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later"),
          M_bg = 1,
          Nexp = length(Data_drug_survival_2$time_palliative_enroll),
          ybef_exp = Data_drug_survival_2$time_palliative_enroll,
          yobs_early = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later"],
          yobs_late = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later"],
          ycen_early = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later"],
          ycen_late = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later"]
        )
      
      stan_model_path <- "source/Stan_code_loglog_weibull.stan"
      stan_model_compiled <- cmdstan_model(stan_model_path)
      stan_weibull_survival_model_fit <- stan_model_compiled$sample(
        data = stan_weibull_survival_model_data,
        seed = 1234,
        chains = 4,
        parallel_chains = min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE)),
        iter_sampling = 2000,  # (iter-warmup)
        iter_warmup = 1000,
        thin = 1,
        refresh = 500
      )
      incProgress(1 / 13)
      
      stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
      stan_weibull_survival_model_draws_total_exp <-
        stan_weibull_survival_model_draws %>%
        dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_total")) %>%
        tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_exp_total")) %>%
        dplyr::select(-key) 
      stan_weibull_survival_model_draws_bef_exp <-
        stan_weibull_survival_model_draws %>%
        dplyr::select(.chain, .iteration, .draw, starts_with("y_exptmp")) %>%
        tidyr::gather(key = key, value = yhat_bef_exp, starts_with("y_exptmp")) %>%
        dplyr::select(-key) 
      stan_weibull_survival_model_draws_aft_exp <-
        stan_weibull_survival_model_draws %>%
        dplyr::select(.chain, .iteration, .draw, starts_with("yhat_exp_uncens")) %>%
        tidyr::gather(key = key, value = yhat_aft_exp, starts_with("yhat_exp_uncens")) %>%
        dplyr::select(-key) 
      stan_weibull_survival_model_draws_ear <-
        stan_weibull_survival_model_draws %>%
        dplyr::select(.chain, .iteration, .draw, starts_with("yhat_early")) %>%
        tidyr::gather(key = key, value = yhat_ear, starts_with("yhat_early")) %>%
        dplyr::select(-key) 
      stan_weibull_survival_model_draws_lat <-
        stan_weibull_survival_model_draws %>%
        dplyr::select(.chain, .iteration, .draw, starts_with("yhat_late")) %>%
        tidyr::gather(key = key, value = yhat_lat, starts_with("yhat_late")) %>%
        dplyr::select(-key) 
      
      median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
      median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
  
      median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
      yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
      yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival_curve$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_survival_curve$censor))]
  
      Data_survival_curve$left_adj = yhat_weibull_sort_average_total_exp
      Data_survival_curve$left_adj[Data_survival_curve$left_adj > 10000] = 10000
  
      
      independence_check = with(Data_survival_curve, cKendall(
        trun = time_palliative_enroll,
        obs = time_palliative_final,
        delta = censor,
        trans = "linear"))
      
      
      traditional_fit = survfit(Surv(event = censor,
                                     time = time_palliative_final) ~ 1, 
                                data = Data_survival_curve)
      delayed_entry_fit = survfit(Surv(event = censor,
                                       time = time_palliative_enroll,
                                       time2 = time_palliative_final) ~ 1,
                                  data = Data_survival_curve)
      simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival_curve$left_adj)),
                                      time = left_adj) ~ 1, 
                                 data = Data_survival_curve)
      hsb_base = Data_survival_curve
      hsb_base$diagnosis = " ALL"
      g = list()
      g[[1]] = ggsurvplot(
        fit = list(traditional_fit = traditional_fit,
                   delayed_entry_fit = delayed_entry_fit,
                   simulation_fit_2 = simulation_fit_2
        ),
        combine = TRUE,
        data = Data_survival_curve,
        xlab = "Time from chemotherapy induction to final observation (months)",
        ylab = "Survival Probability",
        censor = TRUE,
        conf.int = FALSE,
        font.title = 8,
        font.subtitle = 8,
        font.main = 8,
        font.submain = 8,
        font.caption = 8,
        font.legend = 8,
        pval = FALSE,
        surv.median.line = "hv",
        risk.table = TRUE,
        risk.table.y.text = FALSE,
        tables.theme = clean_theme(), 
        legend = c(0.8,0.90),
        cumevents = FALSE,
        cumcensor = FALSE,
        break.x.by = 365.25 * 2.5,
        xscale = "d_m",
        xlim = c(0, max(Data_survival_curve$time_palliative_final) * 1.05),
        #break.x.by = ceiling(max(Data_survival_curve$time_palliative_final) / 500) * 100,
        legend.labs = c("All patients, Unadjusted",
                        "All patients, Number at risk adjusted",
                        "All patients, Adj. right cens. & left trunc.")
      ) + 
        labs(title = paste(traditional_fit[[1]], " patients; Median OS ",
                           round(summary(traditional_fit)$table[[7]] / 365.25 * 12, digits = 1)," (",
                           round(summary(traditional_fit)$table[[8]] / 365.25 * 12, digits = 1),"-",
                           round(summary(traditional_fit)$table[[9]] / 365.25 * 12, digits = 1),") (unadj.)/",
                           round(summary(delayed_entry_fit)$table[[7]] / 365.25 * 12, digits = 1), " (",
                           round(summary(delayed_entry_fit)$table[[8]] / 365.25 * 12, digits = 1),"-",
                           round(summary(delayed_entry_fit)$table[[9]] / 365.25 * 12, digits = 1),") (risk adj.)/",
                           "\n",
                           round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                           round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                           round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1),
                           ") (adj. right & left) months",
                           sep=""),
             subtitle = paste(2, "-year rate ",
                              format(summary(traditional_fit, time = 2 * 365.25)$surv * 100,digits=3),
                              "% (unadj.)/",
                              format(summary(delayed_entry_fit, time = 2 * 365.25)$surv * 100,digits=3),
                              "% (risk adj.)/",
                              format(summary(simulation_fit_2, time = 2 * 365.25)$surv * 100,digits=3),
                              "% (adj. right & left), ",
                              "cKendall tau= ", format(independence_check$PE,digits=2),
                              ", SE= ", format(independence_check$SE,digits=2),
                              ", p= ", format(independence_check$p.value,digits=2), sep=""))
      g[[1]]$table <- g[[1]]$table + theme(plot.title = element_blank(),
                                           plot.subtitle = element_blank())
      g[[2]] = ggsurvplot(
        fit = list(traditional_fit = traditional_fit,
                   # delayed_entry_fit = delayed_entry_fit,
                   simulation_fit_2 = simulation_fit_2
        ),
        combine = TRUE,
        data = Data_survival_curve,
        xlab = "Time from chemotherapy induction to final observation (months)",
        ylab = "Survival Probability",
        censor = TRUE,
        conf.int = FALSE,
        font.title = 8,
        font.subtitle = 8,
        font.main = 8,
        font.submain = 8,
        font.caption = 8,
        font.legend = 8,
        pval = FALSE,
        surv.median.line = "hv",
        risk.table = TRUE,
        risk.table.y.text = FALSE,
        tables.theme = clean_theme(), 
        legend = c(0.8,0.90),
        cumevents = FALSE,
        cumcensor = FALSE,
        break.x.by = 365.25 * 2.5,
        xscale = "d_m",
        xlim = c(0, max(Data_survival_curve$time_palliative_final) * 1.05),
        #break.x.by = ceiling(max(Data_survival_curve$time_palliative_final) / 500) * 100,
        legend.labs = c("All patients, Unadjusted",
                        # "All patients, Number at risk adjusted",
                        "All patients, Adj. right cens. & left trunc.")
      ) + 
        labs(title = paste(traditional_fit[[1]], " patients; Median OS ",
                           round(summary(traditional_fit)$table[[7]] / 365.25 * 12, digits = 1)," (",
                           round(summary(traditional_fit)$table[[8]] / 365.25 * 12, digits = 1),"-",
                           round(summary(traditional_fit)$table[[9]] / 365.25 * 12, digits = 1),") (unadj.)/",
                           "\n",
                           round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                           round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                           round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1),
                           ") (adj. right & left) months",
                           sep=""),
             subtitle = paste(2, "-year rate ",
                              format(summary(traditional_fit, time = 2 * 365.25)$surv * 100,digits=3),
                              "% (unadj.)/",
                              format(summary(simulation_fit_2, time = 2 * 365.25)$surv * 100,digits=3),
                              "% (adj. right & left), ",
                              "cKendall tau= ", format(independence_check$PE,digits=2),
                              ", SE= ", format(independence_check$SE,digits=2),
                              ", p= ", format(independence_check$p.value,digits=2), sep=""))
      g[[2]]$table <- g[[2]]$table + theme(plot.title = element_blank(),
                                 plot.subtitle = element_blank())
      text_median_base = paste0(round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                                round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                                round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1),
                                ")")
      incProgress(1 / 13)
      # Survival analysis for mutation
      mut_gene = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
      if(length(mut_gene)==0){
        mut_gene = "TP53"
      }
      ID_mutation = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene))$Tumor_Sample_Barcode
      Data_survival = Data_survival %>% dplyr::mutate(
        mutation = case_when(
          C.CAT調査結果.基本項目.ハッシュID %in% ID_mutation ~ paste(mut_gene, collapse = ";"),
          TRUE ~ "no-mut"
        )
      )
      traditional_fit = survfit(Surv(event = censor,
                                     time = time_palliative_final) ~ mutation, 
                                data = Data_survival)
      if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
        name_1 = paste("mutation", sort(unique(Data_survival$mutation))[[1]])
        name_2 = paste("mutation", sort(unique(Data_survival$mutation))[[2]])
        name_1_short = sort(unique(Data_survival$mutation))[[1]]
        name_2_short = sort(unique(Data_survival$mutation))[[2]]
        
        Data_survival$gene_mut = Data_survival$mutation == sort(unique(Data_survival$mutation))[[2]]
        Data_survival_curve = Data_survival
        Data_survival_tmp_pos = Data_survival_curve %>%
          dplyr::filter(gene_mut == TRUE)
        if(sum(Data_survival_tmp_pos$censor) > 0){
          fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos)
          time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
          survival_rate_10 = summary(fit,times = time_10)[[6]]
          survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_enroll_final)*0.9)[[6]]
        } else {
          time_10 = 90
          survival_rate_10 = 0
          survival_rate_90 = 1
        }
        time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_enroll_final)*0.9)
        Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
          time_enroll_final >= time_10 &
            time_enroll_final <= time_90)
        Data_survival_tmp$time_enroll_final =
          Data_survival_tmp$time_enroll_final - (time_10 - 1)
        Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
          time_enroll_final > time_90) %>%
          dplyr::arrange(time_enroll_final)
        
        Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
        idx <- 1:nrow(Data_survival_tmp2)
        Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
        
        Data_survival_tmp_neg = Data_survival_curve %>%
          dplyr::filter(gene_mut == FALSE)
        if(sum(Data_survival_tmp_neg$censor) > 0){
          fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_neg)
          time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
          survival_rate_10 = summary(fit,times = time_10)[[6]]
          survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_enroll_final)*0.9)[[6]]
        } else {
          time_10 = 90
          survival_rate_10 = 0
          survival_rate_90 = 1
        }
        time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_enroll_final)*0.9)
        Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
          time_enroll_final >= time_10 &
            time_enroll_final <= time_90)
        Data_survival_tmp3$time_enroll_final =
          Data_survival_tmp3$time_enroll_final - (time_10 - 1)
        Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
          time_enroll_final > time_90) %>%
          dplyr::arrange(time_enroll_final)
        
        Data_survival_tmp4$time_enroll_final = time_90 - (time_10 - 1)
        idx <- 1:nrow(Data_survival_tmp4)
        Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx < max(idx)*(1-survival_rate_10)],])
        Data_drug_survival_1 = rbind(Data_survival_tmp, Data_survival_tmp3)  
        
        Data_survival_tmp_pos = Data_survival_curve %>%
          dplyr::filter(gene_mut == TRUE)
        if(sum(Data_survival_tmp_pos$censor) > 0){
          fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_pos)
          time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
          survival_rate_10 = summary(fit,times = time_10)[[6]]
          survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)[[6]]
        } else {
          time_10 = 90
          survival_rate_10 = 0
          survival_rate_90 = 1
        }
        time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)
        Data_survival_tmp5 = Data_survival_tmp_pos %>% dplyr::filter(
          time_palliative_enroll >= time_10 &
            time_palliative_enroll <= time_90)
        Data_survival_tmp5$time_palliative_enroll =
          Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
        Data_survival_tmp6 = Data_survival_tmp_pos %>% dplyr::filter(
          time_palliative_enroll > time_90) %>%
          dplyr::arrange(time_palliative_enroll)
        Data_survival_tmp6$time_palliative_enroll = time_90 - (time_10 - 1)
        idx <- 1:nrow(Data_survival_tmp6)
        Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
        
        Data_survival_tmp_neg = Data_survival_curve %>%
          dplyr::filter(gene_mut == FALSE)
        
        if(sum(Data_survival_tmp_neg$censor) > 0){
          fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg)
          time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
          survival_rate_10 = summary(fit,times = time_10)[[6]]
          survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)[[6]]
        } else {
          time_10 = 90
          survival_rate_10 = 0
          survival_rate_90 = 1
        }
        time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)
        Data_survival_tmp7 = Data_survival_tmp_neg %>% dplyr::filter(
          time_palliative_enroll >= time_10 &
            time_palliative_enroll <= time_90)
        Data_survival_tmp7$time_palliative_enroll =
          Data_survival_tmp7$time_palliative_enroll - (time_10 - 1)
        Data_survival_tmp8 = Data_survival_tmp_neg %>% dplyr::filter(
          time_palliative_enroll > time_90) %>%
          dplyr::arrange(time_palliative_enroll)
        Data_survival_tmp8$time_palliative_enroll = time_90 - (time_10 - 1)
        idx <- 1:nrow(Data_survival_tmp8)
        Data_survival_tmp7 = rbind(Data_survival_tmp7, Data_survival_tmp8[idx[idx < max(idx)*(1-survival_rate_10)],])
        
        Median_entry_pos = median(Data_survival_tmp5$time_palliative_enroll)
        Median_entry_neg = median(Data_survival_tmp7$time_palliative_enroll)
        Data_drug_survival_1_2 = rbind(Data_survival_tmp5, Data_survival_tmp7)  
        
        Data_drug_survival_1 = Data_drug_survival_1 %>%
          dplyr::mutate(delay_1 = case_when(
            time_palliative_enroll <= Median_entry_pos &
              gene_mut == TRUE ~ "SPT to entry early",
            time_palliative_enroll > Median_entry_pos &
              gene_mut == TRUE ~ "SPT to entry later",
            time_palliative_enroll <= Median_entry_neg &
              gene_mut == FALSE ~ "SPT to entry early",
            time_palliative_enroll > Median_entry_neg &
              gene_mut == FALSE ~ "SPT to entry later"))
        
        if(sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
           sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
           sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
           sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
          
          stan_weibull_survival_model_data <-
            list(
              Median_pos = as.integer(Median_entry_pos),
              Median_neg = as.integer(Median_entry_neg),
              Nobs_early_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
              Nobs_late_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
              Ncen_early_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
              Ncen_late_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
              Nobs_early_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
              Nobs_late_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
              Ncen_early_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
              Ncen_late_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
              Nexp_pos = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE]),
              Nexp_neg = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE]),
              ybef_exp_pos = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE],
              ybef_exp_neg = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE],
              yobs_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
              yobs_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
              ycen_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
              ycen_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
              yobs_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
              yobs_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
              ycen_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
              ycen_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE]
            )
          
          stan_model_path <- "source/Stan_code_loglog_weibull_factor.stan"
          stan_model_compiled <- cmdstan_model(stan_model_path)
          stan_weibull_survival_model_fit <- stan_model_compiled$sample(
            data = stan_weibull_survival_model_data,
            seed = 1234,
            chains = 4,
            parallel_chains = min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE)),
            iter_sampling = 2000,  # (iter-warmup)
            iter_warmup = 1000,
            thin = 1,
            refresh = 500
          )

         stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
          stan_weibull_survival_model_draws_total <-
            stan_weibull_survival_model_draws %>%
            dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
            tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
            dplyr::select(-key) 
          stan_weibull_survival_model_draws_total_exp <-
            stan_weibull_survival_model_draws %>%
            dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
            tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
            dplyr::select(-key) 
          
          median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
          median_os_list_weibull_total = matrix(stan_weibull_survival_model_draws_total$yhat_total, ncol=4000)
          median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
          median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
          median_os_list_weibull_total = colMedians(median_os_list_weibull_total,na.rm = T)
          median_os_list_weibull_bias = colMedians(median_os_list_weibull_bias,na.rm = T)
          
          median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
          median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
          median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
          yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
          yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
          yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival_curve$censor) * as.integer(length(yhat_weibull_sort_total)/length(Data_survival_curve$censor))]
          yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival_curve$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_survival_curve$censor))]
          
          Data_survival_curve$factor_neg = yhat_weibull_sort_average_total
          Data_survival_curve$factor_pos = yhat_weibull_sort_average_total_exp
          Data_survival_curve$factor_neg[Data_survival_curve$factor_neg > 10000] = 10000
          Data_survival_curve$factor_pos[Data_survival_curve$factor_pos > 10000] = 10000
  
          traditional_fit = survfit(Surv(event = censor,
                                         time = time_palliative_final) ~ gene_mut,
                                    data = Data_survival_curve)
          simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival_curve$factor_neg)),
                                            time = factor_neg) ~ 1, 
                                       data = Data_survival_curve)
          simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival_curve$factor_pos)),
                                            time = factor_pos) ~ 1, 
                                       data = Data_survival_curve)
          
          g[[3]] = ggsurvplot(
            fit = list(traditional_fit = traditional_fit,
                       simulation_fit_neg = simulation_fit_neg,
                       simulation_fit_pos = simulation_fit_pos),
            combine = TRUE,
            data = Data_survival_curve,
            xlab = "Time from chemotherapy induction to final observation (months)",
            ylab = "Survival Probability",
            censor = TRUE,
            font.title = 8,
            font.subtitle = 8,
            font.main = 8,
            font.submain = 8,
            font.caption = 8,
            font.legend = 8,
            surv.median.line = "hv",
            conf.int = FALSE,
            pval = FALSE,
            risk.table = TRUE,
            risk.table.y.text = FALSE,
            tables.theme = clean_theme(), 
            legend = c(0.8,0.8),
            xlim = c(0, max(Data_survival_curve$time_palliative_final) * 1.05),
            cumevents = FALSE,
            cumcensor = FALSE,
            break.x.by = 365.25 * 2.5,
            xscale = "d_m",
            #break.x.by = ceiling(max(Data_survival_curve$time_palliative_final) / 500) * 100,
            legend.labs = c(paste0(name_1, ", Unadjusted"),
                            paste0(name_2, ", Unadjusted"),
                            paste0(name_1, ", Adjusted for Delayed Entry"),
                            paste0(name_2, ", Adjusted for Delayed Entry"))
          ) +   
            labs(title = paste(sum(traditional_fit[[1]]), " patients ",
                               "Median OS ",
                               round(summary(traditional_fit)$table[[13]] / 365.25 * 12, digits = 1)," (",
                               round(summary(traditional_fit)$table[[15]] / 365.25 * 12, digits = 1),"-",
                               round(summary(traditional_fit)$table[[17]] / 365.25 * 12, digits = 1),") (",
                               name_1_short,", unadj.)/",
                               round(summary(traditional_fit)$table[[14]] / 365.25 * 12, digits = 1), " (",
                               round(summary(traditional_fit)$table[[16]] / 365.25 * 12, digits = 1),"-",
                               round(summary(traditional_fit)$table[[18]] / 365.25 * 12, digits = 1),") (",
                               name_2_short,", unadj.)/",
                               round(median_os_summary_weibull_total[[2]] / 365.25 * 12, digits = 1), 
                               " (", round(median_os_summary_weibull_total[[1]] / 365.25 * 12, digits = 1), "-",
                               round(median_os_summary_weibull_total[[3]] / 365.25 * 12, digits = 1), ") (",
                               name_1_short,", adj.)/",
                               round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                               round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                               round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1),
                               ") (", name_2_short,", adj.) months",
                               sep=""),
                 subtitle = paste("Survival difference, ",name_1_short, "-", name_2_short, ": ",
                                  round(median_os_summary_weibull_bias[[2]] / 365.25 * 12, digits = 1), " (",
                                  round(median_os_summary_weibull_bias[[1]] / 365.25 * 12, digits = 1), "-", 
                                  round(median_os_summary_weibull_bias[[3]] / 365.25 * 12, digits = 1), 
                                  ") months, median (95% CI)",
                                  sep=""))
          g[[3]]$table <- g[[3]]$table + theme(plot.title = element_blank(),
                                               plot.subtitle = element_blank())
        }
      }
      incProgress(1 / 13)
      
      # comparing between diagnoses
      # if(length(unique(Data_survival$diagnosis)) == 2){
      #   traditional_fit = survfit(Surv(event = censor,
      #                                  time = time_palliative_final) ~ diagnosis, 
      #                             data = Data_survival)
      #   if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
      #     name_1 = sort(unique(Data_survival$diagnosis))[[1]]
      #     name_2 = sort(unique(Data_survival$diagnosis))[[2]]
      #     name_1_short = str_sub(name_1,1,15)
      #     name_2_short = str_sub(name_2,1,15)
      # 
      #     Data_survival$gene_mut = Data_survival$diagnosis == sort(unique(Data_survival$diagnosis))[[2]]
      #     Data_survival_curve = Data_survival
      #     Data_survival_tmp_pos = Data_survival_curve %>%
      #       dplyr::filter(gene_mut == TRUE)
      #     fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos)
      #     time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
      #     survival_rate_10 = summary(fit,times = time_10)[[6]]
      #     survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_enroll_final)*0.9)[[6]]
      #     time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_enroll_final)*0.9)
      #     Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
      #       time_enroll_final >= time_10 &
      #         time_enroll_final <= time_90)
      #     Data_survival_tmp$time_enroll_final =
      #       Data_survival_tmp$time_enroll_final - (time_10 - 1)
      #     Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
      #       time_enroll_final > time_90) %>%
      #       dplyr::arrange(time_enroll_final)
      #     
      #     Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
      #     idx <- 1:nrow(Data_survival_tmp2)
      #     Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
      #     
      #     Data_survival_tmp_neg = Data_survival_curve %>%
      #       dplyr::filter(gene_mut == FALSE)
      #     fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_neg)
      #     time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
      #     survival_rate_10 = summary(fit,times = time_10)[[6]]
      #     survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_enroll_final)*0.9)[[6]]
      #     time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_enroll_final)*0.9)
      #     Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
      #       time_enroll_final >= time_10 &
      #         time_enroll_final <= time_90)
      #     Data_survival_tmp3$time_enroll_final =
      #       Data_survival_tmp3$time_enroll_final - (time_10 - 1)
      #     Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
      #       time_enroll_final > time_90) %>%
      #       dplyr::arrange(time_enroll_final)
      #     
      #     Data_survival_tmp4$time_enroll_final = time_90 - (time_10 - 1)
      #     idx <- 1:nrow(Data_survival_tmp4)
      #     Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx < max(idx)*(1-survival_rate_10)],])
      #     Data_drug_survival_1 = rbind(Data_survival_tmp, Data_survival_tmp3)  
      #     
      #     Data_survival_tmp_pos = Data_survival_curve %>%
      #       dplyr::filter(gene_mut == TRUE)
      #     fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_pos)
      #     time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
      #     survival_rate_10 = summary(fit,times = time_10)[[6]]
      #     survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)[[6]]
      #     time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)
      #     Data_survival_tmp5 = Data_survival_tmp_pos %>% dplyr::filter(
      #       time_palliative_enroll >= time_10 &
      #         time_palliative_enroll <= time_90)
      #     Data_survival_tmp5$time_palliative_enroll =
      #       Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
      #     Data_survival_tmp6 = Data_survival_tmp_pos %>% dplyr::filter(
      #       time_palliative_enroll > time_90) %>%
      #       dplyr::arrange(time_palliative_enroll)
      #     Data_survival_tmp6$time_palliative_enroll = time_90 - (time_10 - 1)
      #     idx <- 1:nrow(Data_survival_tmp6)
      #     Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
      #     
      #     Data_survival_tmp_neg = Data_survival_curve %>%
      #       dplyr::filter(gene_mut == FALSE)
      #     
      #     fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg)
      #     time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
      #     survival_rate_10 = summary(fit,times = time_10)[[6]]
      #     survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)[[6]]
      #     time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)
      #     Data_survival_tmp7 = Data_survival_tmp_neg %>% dplyr::filter(
      #       time_palliative_enroll >= time_10 &
      #         time_palliative_enroll <= time_90)
      #     Data_survival_tmp7$time_palliative_enroll =
      #       Data_survival_tmp7$time_palliative_enroll - (time_10 - 1)
      #     Data_survival_tmp8 = Data_survival_tmp_neg %>% dplyr::filter(
      #       time_palliative_enroll > time_90) %>%
      #       dplyr::arrange(time_palliative_enroll)
      #     Data_survival_tmp8$time_palliative_enroll = time_90 - (time_10 - 1)
      #     idx <- 1:nrow(Data_survival_tmp8)
      #     Data_survival_tmp7 = rbind(Data_survival_tmp7, Data_survival_tmp8[idx[idx < max(idx)*(1-survival_rate_10)],])
      #     
      #     Median_entry_pos = median(Data_survival_tmp5$time_palliative_enroll)
      #     Median_entry_neg = median(Data_survival_tmp7$time_palliative_enroll)
      #     Data_drug_survival_1_2 = rbind(Data_survival_tmp5, Data_survival_tmp7)  
      #     
      #     Data_drug_survival_1 = Data_drug_survival_1 %>%
      #       dplyr::mutate(delay_1 = case_when(
      #         time_palliative_enroll <= Median_entry_pos &
      #           gene_mut == TRUE ~ "SPT to entry early",
      #         time_palliative_enroll > Median_entry_pos &
      #           gene_mut == TRUE ~ "SPT to entry later",
      #         time_palliative_enroll <= Median_entry_neg &
      #           gene_mut == FALSE ~ "SPT to entry early",
      #         time_palliative_enroll > Median_entry_neg &
      #           gene_mut == FALSE ~ "SPT to entry later"))
      #     
      #     if(sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
      #        sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
      #        sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
      #        sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
      #       
      #       stan_weibull_survival_model_data <-
      #         list(
      #           Median_pos = as.integer(Median_entry_pos),
      #           Median_neg = as.integer(Median_entry_neg),
      #           Nobs = sum(Data_drug_survival_1$censor == 1),
      #           Ncen = sum(Data_drug_survival_1$censor == 0),
      #           Ntot = length(Data_drug_survival_1$censor),
      #           M_bg = 2,
      #           Nexp = length(Data_drug_survival_1_2$time_palliative_enroll),
      #           ybef_exp = Data_drug_survival_1_2$time_palliative_enroll,
      #           xfactor = as.numeric(Data_drug_survival_1_2$gene_mut),
      #           yobs = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1],
      #           ycen = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0],
      #           Xobs_bg = matrix(c(as.numeric(Data_drug_survival_1$delay_1 == "SPT to entry later")[Data_drug_survival_1$censor == 1],
      #                              as.numeric(Data_drug_survival_1$gene_mut)[Data_drug_survival_1$censor == 1]),ncol = 2),
      #           Xcen_bg = matrix(c(as.numeric(Data_drug_survival_1$delay_1 == "SPT to entry later")[Data_drug_survival_1$censor == 0],
      #                              as.numeric(Data_drug_survival_1$gene_mut)[Data_drug_survival_1$censor == 0]),ncol = 2)
      #         )
      #       
      #       stan_weibull_survival_model_fit <-
      #         rstan::stan(file = "source/Stan_code_loglog_weibull_factor.stan",
      #                     data = stan_weibull_survival_model_data,
      #                     control=list(adapt_delta=0.99, max_treedepth = 10),
      #                     seed=1234, chains=4, iter=3000, warmup=1000, thin=1, refresh=500, verbose = FALSE,
      #                     pars = c("alpha","mu", "shape_exp", "scale_exp", "factor",
      #                              "yhat_total", "yhat_factor_total"))
      #       
      #       stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
      #       stan_weibull_survival_model_draws_total <-
      #         stan_weibull_survival_model_draws %>%
      #         dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      #         tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      #         dplyr::select(-key) 
      #       stan_weibull_survival_model_draws_total_exp <-
      #         stan_weibull_survival_model_draws %>%
      #         dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      #         tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      #         dplyr::select(-key) 
      #       
      #       median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
      #       median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
      #       
      #       median_os_list_weibull_total = matrix(stan_weibull_survival_model_draws_total$yhat_total, ncol=4000)
      #       median_os_list_weibull_total = colMedians(median_os_list_weibull_total,na.rm = T)
      #       
      #       median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
      #       
      #       median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
      #       median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
      #       median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
      #       yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
      #       yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
      #       yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival_curve$censor) * as.integer(length(yhat_weibull_sort_total)/length(Data_survival_curve$censor))]
      #       yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival_curve$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_survival_curve$censor))]
      #       
      #       Data_survival_curve$factor_neg = yhat_weibull_sort_average_total
      #       Data_survival_curve$factor_pos = yhat_weibull_sort_average_total_exp
      #       Data_survival_curve$factor_neg[Data_survival_curve$factor_neg > 10000] = 10000
      #       Data_survival_curve$factor_pos[Data_survival_curve$factor_pos > 10000] = 10000
      #       
      #       traditional_fit = survfit(Surv(event = censor,
      #                                      time = time_palliative_final) ~ gene_mut,
      #                                 data = Data_survival_curve)
      #       simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival_curve$factor_neg)),
      #                                         time = factor_neg) ~ 1, 
      #                                    data = Data_survival_curve)
      #       simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival_curve$factor_pos)),
      #                                         time = factor_pos) ~ 1, 
      #                                    data = Data_survival_curve)
      #       
      #       g[[k_g]] = ggsurvplot(
      #         fit = list(traditional_fit = traditional_fit,
      #                    simulation_fit_neg = simulation_fit_neg,
      #                    simulation_fit_pos = simulation_fit_pos),
      #         combine = TRUE,
      #         data = Data_survival_curve,
      #         xlab = "Time from chemotherapy induction to final observation (months)",
      #         ylab = "Survival Probability",
      #         censor = TRUE,
      #         font.title = 8,
      #         font.subtitle = 8,
      #         font.main = 8,
      #         font.submain = 8,
      #         font.caption = 8,
      #         font.legend = 8,
      #         surv.median.line = "hv",
      #         conf.int = FALSE,
      #         pval = FALSE,
      #         risk.table = TRUE,
      #         risk.table.y.text = FALSE,
      #         tables.theme = clean_theme(), 
      #         legend = c(0.8,0.8),
      #         xlim = c(0, max(Data_survival_curve$time_palliative_final) * 1.05),
      #         break.x.by = ceiling(max(Data_survival_curve$time_palliative_final) / 500) * 100,
      #         legend.labs = c(paste0(name_1, ", Unadjusted"),
      #                         paste0(name_2, ", Unadjusted"),
      #                         paste0(name_1, ", Adjusted for Delayed Entry"),
      #                         paste0(name_2, ", Adjusted for Delayed Entry"))
      #       ) +   
      #         labs(title = paste(sum(traditional_fit[[1]]), " patients ",
      #                            "Median OS ",
      #                            as.integer(summary(traditional_fit)$table[[13]])," (",
      #                            as.integer(summary(traditional_fit)$table[[15]]),"-",
      #                            as.integer(summary(traditional_fit)$table[[17]]),") (",
      #                            name_1_short,", unadj.)/",
      #                            as.integer(summary(traditional_fit)$table[[14]]), " (",
      #                            as.integer(summary(traditional_fit)$table[[16]]),"-",
      #                            as.integer(summary(traditional_fit)$table[[18]]),") (",
      #                            name_2_short,", unadj.)/",
      #                            as.integer(median_os_summary_weibull_total[[2]]), 
      #                            " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
      #                            as.integer(median_os_summary_weibull_total[[3]]), ") (",
      #                            name_1_short,", adj.)/",
      #                            as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
      #                            as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
      #                            as.integer(median_os_summary_weibull_total_exp[[3]]),
      #                            ") (", name_2_short,", adj.) days",
      #                            sep=""),
      #              subtitle = paste("Survival difference, ",name_1_short, "-", name_2_short, ": ",
      #                               as.integer(median_os_summary_weibull_bias[[2]]), " (",
      #                               as.integer(median_os_summary_weibull_bias[[1]]), "-", 
      #                               as.integer(median_os_summary_weibull_bias[[3]]), 
      #                               ") days, median (95% CI)",
      #                               sep=""))
      #       g[[k_g]]$table <- g[[k_g]]$table + theme(plot.title = element_blank(),
      #                                            plot.subtitle = element_blank())
      #       k_g = k_g + 1
      #     }
      #   }
      # }
      Data_survival$diagnosis = Data_survival$Cancers
      if(length(unique(Data_survival$diagnosis)) > 1){
        hb = list()
        hsb = list()
        hsba = list()
        kb = 1
        text_median = NULL
        oncogenic_genes = data.frame(names(table(Data_survival$diagnosis)),
                                     as.integer(table(Data_survival$diagnosis)))
        colnames(oncogenic_genes) = c("gene_mutation", "all_patients")
        oncogenic_genes = oncogenic_genes[1:min(length(oncogenic_genes$all_patients), 12),]
        
        gene_table = data.frame(oncogenic_genes$gene_mutation)
        colnames(gene_table) = c("Gene")
        gene_table$positive_patients = oncogenic_genes$all_patients
        gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)) / 10
        gene_table$positive_median = 0
        gene_table$positive_LL = 0
        gene_table$positive_UL = 0
        gene_table$negative_median = 0
        gene_table$negative_LL = 0
        gene_table$negative_UL = 0
        gene_table$diff_median = 0
        gene_table$diff_LL = 0
        gene_table$diff_UL = 0
        gene_table$positive_mortal_event = 0
        gene_table$negative_mortal_event = 0
        withProgress(message = sample(nietzsche)[1], {
          for(i in 1:length(oncogenic_genes$all_patients)){
            ID_mutation = (Data_survival %>% dplyr::filter(diagnosis %in% oncogenic_genes$gene_mutation[i]))$C.CAT調査結果.基本項目.ハッシュID
            Data_survival = Data_survival %>% dplyr::mutate(
              gene_mut = case_when(
                C.CAT調査結果.基本項目.ハッシュID %in% ID_mutation ~ TRUE,
                TRUE ~ FALSE
              )
            )
            
            mut_gene = oncogenic_genes$gene_mutation[i]
            traditional_fit = survfit(Surv(event = censor,
                                           time = time_palliative_final) ~ gene_mut, 
                                      data = Data_survival)
            if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
              gene_table$negative_mortal_event[i] = summary(traditional_fit)$table[7]
              gene_table$positive_mortal_event[i] = summary(traditional_fit)$table[8]
              
              Data_survival_curve = Data_survival
              Data_survival_tmp_pos = Data_survival_curve %>%
                dplyr::filter(gene_mut == TRUE)
              if(sum(Data_survival_tmp_pos$censor) > 0){
                fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_enroll_final)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_enroll_final)*0.9)
              Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
                time_enroll_final >= time_10 &
                  time_enroll_final <= time_90)
              Data_survival_tmp$time_enroll_final =
                Data_survival_tmp$time_enroll_final - (time_10 - 1)
              Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
                time_enroll_final > time_90) %>%
                dplyr::arrange(time_enroll_final)
              
              Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp2)
              Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
              
              Data_survival_tmp_neg = Data_survival_curve %>%
                dplyr::filter(gene_mut == FALSE)
              if(sum(Data_survival_tmp_neg$censor) > 0){
                fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_neg)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_enroll_final)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_enroll_final)*0.9)
              Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
                time_enroll_final >= time_10 &
                  time_enroll_final <= time_90)
              Data_survival_tmp3$time_enroll_final =
                Data_survival_tmp3$time_enroll_final - (time_10 - 1)
              Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
                time_enroll_final > time_90) %>%
                dplyr::arrange(time_enroll_final)
              
              Data_survival_tmp4$time_enroll_final = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp4)
              Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx < max(idx)*(1-survival_rate_10)],])
              Data_drug_survival_1 = rbind(Data_survival_tmp, Data_survival_tmp3)  
              
              Data_survival_tmp_pos = Data_survival_curve %>%
                dplyr::filter(gene_mut == TRUE)
              if(sum(Data_survival_tmp_pos$censor) > 0){
                fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_pos)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)
              Data_survival_tmp5 = Data_survival_tmp_pos %>% dplyr::filter(
                time_palliative_enroll >= time_10 &
                  time_palliative_enroll <= time_90)
              Data_survival_tmp5$time_palliative_enroll =
                Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
              Data_survival_tmp6 = Data_survival_tmp_pos %>% dplyr::filter(
                time_palliative_enroll > time_90) %>%
                dplyr::arrange(time_palliative_enroll)
              Data_survival_tmp6$time_palliative_enroll = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp6)
              Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
              
              Data_survival_tmp_neg = Data_survival_curve %>%
                dplyr::filter(gene_mut == FALSE)
              
              if(sum(Data_survival_tmp_neg$censor) > 0){
                fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)
              Data_survival_tmp7 = Data_survival_tmp_neg %>% dplyr::filter(
                time_palliative_enroll >= time_10 &
                  time_palliative_enroll <= time_90)
              Data_survival_tmp7$time_palliative_enroll =
                Data_survival_tmp7$time_palliative_enroll - (time_10 - 1)
              Data_survival_tmp8 = Data_survival_tmp_neg %>% dplyr::filter(
                time_palliative_enroll > time_90) %>%
                dplyr::arrange(time_palliative_enroll)
              Data_survival_tmp8$time_palliative_enroll = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp8)
              Data_survival_tmp7 = rbind(Data_survival_tmp7, Data_survival_tmp8[idx[idx < max(idx)*(1-survival_rate_10)],])
              
              Median_entry_pos = median(Data_survival_tmp5$time_palliative_enroll)
              Median_entry_neg = median(Data_survival_tmp7$time_palliative_enroll)
              Data_drug_survival_1_2 = rbind(Data_survival_tmp5, Data_survival_tmp7)  
              
              Data_drug_survival_1 = Data_drug_survival_1 %>%
                dplyr::mutate(delay_1 = case_when(
                  time_palliative_enroll <= Median_entry_pos &
                    gene_mut == TRUE ~ "SPT to entry early",
                  time_palliative_enroll > Median_entry_pos &
                    gene_mut == TRUE ~ "SPT to entry later",
                  time_palliative_enroll <= Median_entry_neg &
                    gene_mut == FALSE ~ "SPT to entry early",
                  time_palliative_enroll > Median_entry_neg &
                    gene_mut == FALSE ~ "SPT to entry later"))
              
              if(sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
                 sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
                 sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
                 sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
                
                stan_weibull_survival_model_data <-
                  list(
                    Median_pos = as.integer(Median_entry_pos),
                    Median_neg = as.integer(Median_entry_neg),
                    Nobs_early_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Nobs_late_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Ncen_early_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Ncen_late_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Nobs_early_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Nobs_late_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Ncen_early_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Ncen_late_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Nexp_pos = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE]),
                    Nexp_neg = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE]),
                    ybef_exp_pos = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE],
                    ybef_exp_neg = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE],
                    yobs_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    yobs_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    ycen_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    ycen_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    yobs_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                    yobs_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                    ycen_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                    ycen_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE]
                  )
                
                stan_model_path <- "source/Stan_code_loglog_weibull_factor.stan"
                stan_model_compiled <- cmdstan_model(stan_model_path)
                stan_weibull_survival_model_fit <- stan_model_compiled$sample(
                  data = stan_weibull_survival_model_data,
                  seed = 1234,
                  chains = 4,
                  parallel_chains = min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE)),
                  iter_sampling = 2000,  # (iter-warmup)
                  iter_warmup = 1000,
                  thin = 1,
                  refresh = 500
                )
                stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
                stan_weibull_survival_model_draws_total <-
                  stan_weibull_survival_model_draws %>%
                  dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
                  tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
                  dplyr::select(-key) 
                stan_weibull_survival_model_draws_total_exp <-
                  stan_weibull_survival_model_draws %>%
                  dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
                  tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
                  dplyr::select(-key) 
                
                median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
                median_os_list_weibull_total = matrix(stan_weibull_survival_model_draws_total$yhat_total, ncol=4000)
                median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
                median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
                median_os_list_weibull_total = colMedians(median_os_list_weibull_total,na.rm = T)
                median_os_list_weibull_bias = colMedians(median_os_list_weibull_bias,na.rm = T)
                
                
                median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
                median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
                median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
                yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
                yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
                yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival_curve$censor) * as.integer(length(yhat_weibull_sort_total)/length(Data_survival_curve$censor))]
                yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival_curve$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_survival_curve$censor))]
                
                Data_survival_curve$factor_neg = yhat_weibull_sort_average_total
                Data_survival_curve$factor_pos = yhat_weibull_sort_average_total_exp
                Data_survival_curve$factor_neg[Data_survival_curve$factor_neg > 10000] = 10000
                Data_survival_curve$factor_pos[Data_survival_curve$factor_pos > 10000] = 10000
                
                traditional_fit = survfit(Surv(event = censor,
                                               time = time_palliative_final) ~ gene_mut,
                                          data = Data_survival_curve)
                simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival_curve$factor_neg)),
                                                  time = factor_neg) ~ 1, 
                                             data = Data_survival_curve)
                simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival_curve$factor_pos)),
                                                  time = factor_pos) ~ 1, 
                                             data = Data_survival_curve)
                hsba[[kb]] = Data_survival_curve
                hsba[[kb]]$diagnosis = mut_gene
                hsb[[kb]] = ggsurvplot(
                  fit = list(traditional_fit = traditional_fit,
                             simulation_fit_neg = simulation_fit_neg,
                             simulation_fit_pos = simulation_fit_pos),
                  combine = TRUE,
                  data = Data_survival_curve,
                  xlab = "Time from chemotherapy induction to final observation (months)",
                  ylab = "Survival Probability",
                  censor = TRUE,
                  surv.scale = "percent",
                  font.title = 8,
                  font.subtitle = 8,
                  font.main = 8,
                  font.submain = 8,
                  font.caption = 8,
                  font.legend = 8,
                  surv.median.line = "v",
                  palette = "Dark2",
                  conf.int = FALSE,
                  pval = FALSE,
                  risk.table = TRUE,
                  risk.table.y.text = FALSE,
                  tables.theme = clean_theme(), 
                  legend = c(0.8,0.8),
                  xlim = c(0, max(Data_survival_curve$time_palliative_final) * 1.05),
                  cumevents = FALSE,
                  cumcensor = FALSE,
                  break.x.by = 365.25 * 2.5,
                  xscale = "d_m",
                  #break.x.by = ceiling(max(Data_survival_curve$time_palliative_final) / 500) * 100,
                  legend.labs = c(paste0("Others, Unadjusted"),
                                  paste0(mut_gene, ", Unadjusted"),
                                  paste0("Others, Adjusted for Delayed Entry"),
                                  paste0(mut_gene, ", Adjusted for Delayed Entry"))
                ) +   
                  labs(title = paste(sum(traditional_fit[[1]]), " patients ",
                                     "Median OS ",
                                     round(summary(traditional_fit)$table[[13]] / 365.25 * 12, digits = 1)," (",
                                     round(summary(traditional_fit)$table[[15]] / 365.25 * 12, digits = 1),"-",
                                     round(summary(traditional_fit)$table[[17]] / 365.25 * 12, digits = 1),") (others, unadj.)/",
                                     round(summary(traditional_fit)$table[[14]] / 365.25 * 12, digits = 1), " (",
                                     round(summary(traditional_fit)$table[[16]] / 365.25 * 12, digits = 1),"-",
                                     round(summary(traditional_fit)$table[[18]] / 365.25 * 12, digits = 1),") (", mut_gene, ", unadj.)/\n",
                                     round(median_os_summary_weibull_total[[2]] / 365.25 * 12, digits = 1), 
                                     " (", round(median_os_summary_weibull_total[[1]] / 365.25 * 12, digits = 1), "-",
                                     round(median_os_summary_weibull_total[[3]] / 365.25 * 12, digits = 1), ") (others, adj.)/",
                                     round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                                     round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                                     round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1),
                                     ") (", mut_gene, ", adj.) months",
                                     sep=""),
                       subtitle = paste("Survival difference with diagnosis: ",
                                        round(median_os_summary_weibull_bias[[2]] / 365.25 * 12, digits = 1), " (",
                                        round(median_os_summary_weibull_bias[[1]] / 365.25 * 12, digits = 1), "-", 
                                        round(median_os_summary_weibull_bias[[3]] / 365.25 * 12, digits = 1), 
                                        ") months, median (95% CI)",
                                        sep=""))
                hsb[[kb]]$table <- hsb[[kb]]$table + theme(plot.title = element_blank(),
                                                       plot.subtitle = element_blank())
                kb = kb + 1
                text_median = c(text_median, paste0(round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                                     round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                                     round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1), ")"))
                gene_table$positive_median[i] = round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1)
                gene_table$positive_LL[i] = round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1)
                gene_table$positive_UL[i] = round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1)
                gene_table$negative_median[i] = round(median_os_summary_weibull_total[[2]] / 365.25 * 12, digits = 1)
                gene_table$negative_LL[i] = round(median_os_summary_weibull_total[[1]] / 365.25 * 12, digits = 1)
                gene_table$negative_UL[i] = round(median_os_summary_weibull_total[[3]] / 365.25 * 12, digits = 1)
                gene_table$diff_median[i] = -round(median_os_summary_weibull_bias[[2]] / 365.25 * 12, digits = 1)
                gene_table$diff_LL[i] = -round(median_os_summary_weibull_bias[[3]] / 365.25 * 12, digits = 1)
                gene_table$diff_UL[i] = -round(median_os_summary_weibull_bias[[1]] / 365.25 * 12, digits = 1)
              }
            }
            incProgress(1 / length(oncogenic_genes$all_patients))
          }
          if(kb>1){
            surv_data = NULL
            for(kb_i in 1:(kb - 1)){
              surv_data = rbind(surv_data, hsba[[kb_i]])
            }
            surv_data = surv_data %>% dplyr::select(factor_pos, diagnosis)
            surv_data$censor = 1
            simulation_fit_diagnosis = survfit(Surv(event = censor,
                                                    time = factor_pos) ~ diagnosis, 
                                               data = surv_data)
            survdiff_diagnosis = surv_median(simulation_fit_diagnosis)
            text_median = paste0(paste(text_median, collapse = "/"), " months")
            hsb[[kb]] = ggsurvplot(
              fit = simulation_fit_diagnosis,
              combine = TRUE,
              data = surv_data,
              xlab = "Time from chemotherapy induction to final observation (months)",
              ylab = "Survival Probability",
              censor = TRUE,
              surv.scale = "percent",
              font.title = 8,
              font.subtitle = 8,
              font.main = 8,
              font.submain = 8,
              font.caption = 8,
              font.legend = 8,
              surv.median.line = "v",
              palette = "Paired",
              conf.int = FALSE,
              pval = FALSE,
              risk.table = TRUE,
              risk.table.y.text = FALSE,
              tables.theme = clean_theme(), 
              legend = c(0.8,0.8),
              xlim = c(0, max(Data_survival_curve$time_palliative_final) * 1.05),
              cumevents = FALSE,
              cumcensor = FALSE,
              break.x.by = 365.25 * 2.5,
              xscale = "d_m"
            ) +   
              labs(title = "Median OS by diagnosis, left-truncation bias adjusted",
                   subtitle = text_median
              )
            hsb[[kb]]$table <- hsb[[kb]]$table + theme(plot.title = element_blank(),
                                                       plot.subtitle = element_blank())
            kb = kb + 1
            surv_data = NULL
            for(kb_i in 1:(kb - 2)){
              surv_data = rbind(surv_data, hsba[[kb_i]])
            }
            surv_data = surv_data %>% dplyr::select(factor_pos, diagnosis)
            surv_data_tmp = hsb_base %>% dplyr::select(left_adj, diagnosis)
            colnames(surv_data_tmp) = c("factor_pos", "diagnosis")
            surv_data = rbind(surv_data, surv_data_tmp)
            surv_data$censor = 1
            simulation_fit_diagnosis = survfit(Surv(event = censor,
                                                    time = factor_pos) ~ diagnosis, 
                                               data = surv_data)
            survdiff_diagnosis = surv_median(simulation_fit_diagnosis)
            text_median = paste0(text_median_base, "/", paste(text_median, collapse = "/"), " months")
            hsb[[kb]] = ggsurvplot(
              fit = simulation_fit_diagnosis,
              combine = TRUE,
              data = surv_data,
              xlab = "Time from chemotherapy induction to final observation (months)",
              ylab = "Survival Probability",
              censor = TRUE,
              surv.scale = "percent",
              font.title = 8,
              font.subtitle = 8,
              font.main = 8,
              font.submain = 8,
              font.caption = 8,
              font.legend = 8,
              surv.median.line = "v",
              palette = "Dark2",
              conf.int = FALSE,
              pval = FALSE,
              risk.table = TRUE,
              risk.table.y.text = FALSE,
              tables.theme = clean_theme(), 
              legend = c(0.8,0.8),
              xlim = c(0, max(Data_survival_curve$time_palliative_final) * 1.05),
              cumevents = FALSE,
              cumcensor = FALSE,
              break.x.by = 365.25 * 2.5,
              xscale = "d_m"
            ) +   
              labs(title = "Median OS by diagnosis, left-truncation bias adjusted",
                   subtitle = text_median
              )
            hsb[[kb]]$table <- hsb[[kb]]$table + theme(plot.title = element_blank(),
                                                       plot.subtitle = element_blank())
            kb = kb + 1
            table_ToT <- 
              data.frame(Cohort = as.character(),
                         Duration_months = as.character(),
                         No_total = as.numeric(),
                         No_at_risk = as.numeric(),
                         No_event = as.numeric(),
                         No_censor = as.numeric(),
                         Survival_rate = as.numeric(),
                         Survival_rate_95LL = as.numeric(),
                         Survival_rate_95UL = as.numeric())
            row_no = 1
            for(cohort in 1:length(summary(simulation_fit_diagnosis, times=365.25 / 12, extend=TRUE)$time)){
              for(mths in (0:10)*30){
                dataset = summary(simulation_fit_diagnosis, times=365.25 / 12* mths, extend=TRUE)
                if(length(unique(surv_data$diagnosis)) > 1){
                  table_ToT[row_no,"Cohort"] = rownames(dataset$table)[cohort]
                } else{
                  table_ToT[row_no,"Cohort"] = unique(surv_data$diagnosis)
                }
                table_ToT[row_no,"Duration_months"] = mths
                table_ToT[row_no,"No_total"] = dataset$n[cohort]
                table_ToT[row_no,"No_at_risk"] = dataset$n.risk[cohort]
                table_ToT[row_no,"No_event"] = dataset$n.event[cohort]
                table_ToT[row_no,"No_censor"] = dataset$n.censor[cohort]
                table_ToT[row_no,"Survival_rate"] = dataset$surv[cohort]
                table_ToT[row_no,"Survival_rate_95LL"] = dataset$lower[cohort]
                table_ToT[row_no,"Survival_rate_95UL"] = dataset$upper[cohort]
                row_no = row_no + 1
              }
            }
            surv_data = NULL
            for(kb_i in 1:(kb - 3)){
              surv_data = rbind(surv_data, hsba[[kb_i]])
            }
            surv_data = surv_data  %>% dplyr::filter(gene_mut == TRUE) %>%
              dplyr::select(time_palliative_final, diagnosis, censor)
            surv_data_tmp = hsb_base %>% dplyr::select(time_palliative_final, diagnosis, censor)
            surv_data = rbind(surv_data, surv_data_tmp)
            simulation_fit_diagnosis = survfit(Surv(event = censor,
                                                    time = time_palliative_final) ~ diagnosis, 
                                               data = surv_data)
            for(cohort in 1:length(summary(simulation_fit_diagnosis, times=365.25 / 12, extend=TRUE)$time)){
              for(mths in (0:10)*30){
                dataset = summary(simulation_fit_diagnosis, times=365.25 / 12* mths, extend=TRUE)
                if(length(unique(surv_data$diagnosis)) > 1){
                  table_ToT[row_no,"Cohort"] = paste0(rownames(dataset$table)[cohort], "_raw_data")
                } else{
                  table_ToT[row_no,"Cohort"] = paste0(unique(surv_data$diagnosis), "_raw_data")
                }
                table_ToT[row_no,"Duration_months"] = mths
                table_ToT[row_no,"No_total"] = dataset$n[cohort]
                table_ToT[row_no,"No_at_risk"] = dataset$n.risk[cohort]
                table_ToT[row_no,"No_event"] = dataset$n.event[cohort]
                table_ToT[row_no,"No_censor"] = dataset$n.censor[cohort]
                table_ToT[row_no,"Survival_rate"] = dataset$surv[cohort]
                table_ToT[row_no,"Survival_rate_95LL"] = dataset$lower[cohort]
                table_ToT[row_no,"Survival_rate_95UL"] = dataset$upper[cohort]
                row_no = row_no + 1
              }
            }
          }
          if(kb < 17){
            for(kkb in kb:16){
              hsb[[kkb]] = ggsurvplot_empty()
            }
          }
        })
        incProgress(1 / 13)
        
        Gene_arrange = gene_table$Gene
        gene_table = gene_table %>% dplyr::mutate(
          Gene = factor(gene_table$Gene, levels = Gene_arrange))
        p <- 
          gene_table |>
          ggplot(aes(y = fct_rev(Gene))) + 
          theme_classic()
        p <- p +
          geom_point(aes(x=diff_median), shape=15, size=3) +
          geom_linerange(aes(xmin=diff_LL, xmax=diff_UL)) 
        p <- p +
          geom_vline(xintercept = 0, linetype="dashed") +
          labs(x="Survival difference by diagnosis", y="Diagnosis")
        p <- p +
          coord_cartesian(ylim=c(1,length(Gene_arrange) + 1),
                          xlim=c(-max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) - 0.5,
                                 max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) + 0.5))
        p <- p +
          annotate("text",
                   x = (-max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) - 0.5)/2,
                   y = length(Gene_arrange) + 1,
                   label = "diagnosis=short survival") +
          annotate("text",
                   x = (max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) + 0.5)/2,
                   y = length(Gene_arrange) + 1,
                   label = "diagnosis=long survival")
        p_mid <- p + 
          theme(axis.line.y = element_blank(),
                axis.ticks.y= element_blank(),
                axis.text.y= element_blank(),
                axis.title.y= element_blank())
        
        gene_table <- gene_table %>%
          dplyr::mutate(
            estimate_lab = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
          dplyr::mutate(
            patients = paste0(positive_patients, " (", positive_freq, "%)"))
        gene_table = 
          dplyr::bind_rows(
            data.frame(
              Gene = "Diagnosis",
              estimate_lab = "Survival difference",
              patients = "Patients"
            ), gene_table  )
        Gene_arrange = gene_table$Gene
        gene_table = gene_table %>% dplyr::mutate(
          Gene = factor(gene_table$Gene, levels = Gene_arrange))
        
        p_left <- gene_table  %>% ggplot(aes(y = fct_rev(Gene)))
        p_left <- p_left + geom_text(aes(x = 0, label = Gene), hjust = 0,
                                     fontface = ifelse(gene_table$Gene == "Diagnosis", "bold", "bold.italic"))
        p_left <- p_left + geom_text(aes(x = 2.4, label = estimate_lab), hjust = 0,
                                     fontface = ifelse(gene_table$estimate_lab == "Survival difference", "bold", "plain"))
        p_left <- p_left + theme_void() + coord_cartesian(xlim = c(0, 5))
        
        # right side of plot - pvalues
        p_right <- gene_table  |> ggplot() +
          geom_text(aes(x = 0, y = fct_rev(Gene), label = patients), hjust = 0,
                    fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")) +
          theme_void()
        
        # final plot arrangement
        layout <- c(patchwork::area(t = 0, l = 0, b = 30, r = 5),
                    patchwork::area(t = 1, l = 5, b = 30, r = 10),
                    patchwork::area(t = 0, l = 10, b = 30, r = 12))
        hb[[1]] = p_left + p_mid + p_right + plot_layout(design = layout)
        gene_table_hb = gene_table
        incProgress(1 / 13)
      }
      # if(length(unique(Data_survival$diagnosis)) == 3){
      #   traditional_fit = survfit(Surv(event = censor,
      #                                  time = time_palliative_final) ~ diagnosis, 
      #                             data = Data_survival)
      #   if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
      #     name_1 = sort(unique(Data_survival$diagnosis))[[1]]
      #     name_2 = sort(unique(Data_survival$diagnosis))[[2]]
      #     name_3 = sort(unique(Data_survival$diagnosis))[[3]]
      #     name_1_short = str_sub(name_1,1,15)
      #     name_2_short = str_sub(name_2,1,15)
      #     name_3_short = str_sub(name_3,1,15)
      #     
      #     Data_survival$gene_select =
      #       as.numeric(Data_survival$diagnosis == name_2) +
      #       2 * as.numeric(Data_survival$diagnosis == name_3)
      #     
      #     Data_survival_tmp_pos_2 = Data_survival %>%
      #       dplyr::filter(gene_select == 2)
      #     fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos_2)
      #     time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
      #     survival_rate_10 = summary(fit,times = time_10)[[6]]
      #     survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos_2$time_enroll_final)*0.9)[[6]]
      #     time_90 = max(time_10 + 1, max(Data_survival_tmp_pos_2$time_enroll_final)*0.9)
      #     Data_survival_tmp = Data_survival_tmp_pos_2 %>% dplyr::filter(
      #       time_enroll_final >= time_10 &
      #         time_enroll_final <= time_90)
      #     Data_survival_tmp$time_enroll_final =
      #       Data_survival_tmp$time_enroll_final - (time_10 - 1)
      #     Data_survival_tmp2 = Data_survival_tmp_pos_2 %>% dplyr::filter(
      #       time_enroll_final > time_90) %>%
      #       dplyr::arrange(time_enroll_final)
      #     Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
      #     idx <- 1:nrow(Data_survival_tmp2)
      #     Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
      #     
      #     Data_survival_tmp_pos_1 = Data_survival %>%
      #       dplyr::filter(gene_select == 1)
      #     fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos_1)
      #     time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
      #     survival_rate_10 = summary(fit,times = time_10)[[6]]
      #     survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos_1$time_enroll_final)*0.9)[[6]]
      #     time_90 = max(time_10 + 1, max(Data_survival_tmp_pos_1$time_enroll_final)*0.9)
      #     Data_survival_tmp3 = Data_survival_tmp_pos_1 %>% dplyr::filter(
      #       time_enroll_final >= time_10 &
      #         time_enroll_final <= time_90)
      #     Data_survival_tmp3$time_enroll_final =
      #       Data_survival_tmp3$time_enroll_final - (time_10 - 1)
      #     Data_survival_tmp4 = Data_survival_tmp_pos_1 %>% dplyr::filter(
      #       time_enroll_final > time_90) %>%
      #       dplyr::arrange(time_enroll_final)
      #     Data_survival_tmp4$time_enroll_final = time_90 - (time_10 - 1)
      #     idx <- 1:nrow(Data_survival_tmp4)
      #     Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx < max(idx)*(1-survival_rate_10)],])
      #     
      #     Data_survival_tmp_pos_0 = Data_survival %>%
      #       dplyr::filter(gene_select == 0)
      #     fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos_0)
      #     time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
      #     survival_rate_10 = summary(fit,times = time_10)[[6]]
      #     survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos_0$time_enroll_final)*0.9)[[6]]
      #     time_90 = max(time_10 + 1, max(Data_survival_tmp_pos_0$time_enroll_final)*0.9)
      #     Data_survival_tmp5 = Data_survival_tmp_pos_0 %>% dplyr::filter(
      #       time_enroll_final >= time_10 &
      #         time_enroll_final <= time_90)
      #     Data_survival_tmp5$time_enroll_final =
      #       Data_survival_tmp5$time_enroll_final - (time_10 - 1)
      #     Data_survival_tmp6 = Data_survival_tmp_pos_0 %>% dplyr::filter(
      #       time_enroll_final > time_90) %>%
      #       dplyr::arrange(time_enroll_final)
      #     Data_survival_tmp6$time_enroll_final = time_90 - (time_10 - 1)
      #     idx <- 1:nrow(Data_survival_tmp6)
      #     Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
      #     Data_drug_survival_1 = rbind(Data_survival_tmp, Data_survival_tmp3, Data_survival_tmp5)
      #     
      #     
      #     Data_survival_tmp_neg_2 = Data_survival %>%
      #       dplyr::filter(gene_select == 2)
      #     fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg_2)
      #     time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
      #     survival_rate_10 = summary(fit,times = time_10)[[6]]
      #     survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg_2$time_palliative_enroll)*0.9)[[6]]
      #     time_90 = max(time_10 + 1, max(Data_survival_tmp_neg_2$time_palliative_enroll)*0.9)
      #     Data_survival_tmp7 = Data_survival_tmp_neg_2 %>% dplyr::filter(
      #       time_palliative_enroll >= time_10 &
      #         time_palliative_enroll <= time_90)
      #     Data_survival_tmp7$time_palliative_enroll =
      #       Data_survival_tmp7$time_palliative_enroll - (time_10 - 1)
      #     Data_survival_tmp8 = Data_survival_tmp_neg_2 %>% dplyr::filter(
      #       time_palliative_enroll > time_90) %>%
      #       dplyr::arrange(time_palliative_enroll)
      #     Data_survival_tmp8$time_palliative_enroll = time_90 - (time_10 - 1)
      #     idx <- 1:nrow(Data_survival_tmp8)
      #     Data_survival_tmp7 = rbind(Data_survival_tmp7, Data_survival_tmp8[idx[idx < max(idx)*(1-survival_rate_10)],])
      #     
      #     Data_survival_tmp_neg_1 = Data_survival %>%
      #       dplyr::filter(gene_select == 2)
      #     fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg_1)
      #     time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
      #     survival_rate_10 = summary(fit,times = time_10)[[6]]
      #     survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg_1$time_palliative_enroll)*0.9)[[6]]
      #     time_90 = max(time_10 + 1, max(Data_survival_tmp_neg_1$time_palliative_enroll)*0.9)
      #     Data_survival_tmp9 = Data_survival_tmp_neg_1 %>% dplyr::filter(
      #       time_palliative_enroll >= time_10 &
      #         time_palliative_enroll <= time_90)
      #     Data_survival_tmp9$time_palliative_enroll =
      #       Data_survival_tmp9$time_palliative_enroll - (time_10 - 1)
      #     Data_survival_tmp10 = Data_survival_tmp_neg_1 %>% dplyr::filter(
      #       time_palliative_enroll > time_90) %>%
      #       dplyr::arrange(time_palliative_enroll)
      #     Data_survival_tmp10$time_palliative_enroll = time_90 - (time_10 - 1)
      #     idx <- 1:nrow(Data_survival_tmp10)
      #     Data_survival_tmp9 = rbind(Data_survival_tmp9, Data_survival_tmp10[idx[idx < max(idx)*(1-survival_rate_10)],])
      #     
      #     Data_survival_tmp_neg_0 = Data_survival %>%
      #       dplyr::filter(gene_select == 2)
      #     fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg_0)
      #     time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
      #     survival_rate_10 = summary(fit,times = time_10)[[6]]
      #     survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg_0$time_palliative_enroll)*0.9)[[6]]
      #     time_90 = max(time_10 + 1, max(Data_survival_tmp_neg_0$time_palliative_enroll)*0.9)
      #     Data_survival_tmp11 = Data_survival_tmp_neg_0 %>% dplyr::filter(
      #       time_palliative_enroll >= time_10 &
      #         time_palliative_enroll <= time_90)
      #     Data_survival_tmp11$time_palliative_enroll =
      #       Data_survival_tmp11$time_palliative_enroll - (time_10 - 1)
      #     Data_survival_tmp12 = Data_survival_tmp_neg_0 %>% dplyr::filter(
      #       time_palliative_enroll > time_90) %>%
      #       dplyr::arrange(time_palliative_enroll)
      #     Data_survival_tmp12$time_palliative_enroll = time_90 - (time_10 - 1)
      #     idx <- 1:nrow(Data_survival_tmp12)
      #     Data_survival_tmp11 = rbind(Data_survival_tmp11, Data_survival_tmp12[idx[idx < max(idx)*(1-survival_rate_10)],])
      #     
      #     Median_entry_2 = median(Data_survival_tmp7$time_palliative_enroll)
      #     Median_entry_1 = median(Data_survival_tmp9$time_palliative_enroll)
      #     Median_entry_0 = median(Data_survival_tmp11$time_palliative_enroll)
      #     Data_drug_survival_1_2 = rbind(Data_survival_tmp7, Data_survival_tmp9, Data_survival_tmp11)  
      #     
      #     Data_drug_survival_1 = Data_drug_survival_1 %>%
      #       dplyr::mutate(delay_1 = case_when(
      #         Data_drug_survival_1$time_palliative_enroll <= Median_entry_2 &
      #           gene_select == 2 ~ "SPT to entry early",
      #         Data_drug_survival_1$time_palliative_enroll > Median_entry_1 &
      #           gene_select == 1 ~ "SPT to entry later",
      #         Data_drug_survival_1$time_palliative_enroll <= Median_entry_0 &
      #           gene_select == 0 ~ "SPT to entry early",
      #         Data_drug_survival_1$time_palliative_enroll > Median_entry_2 &
      #           gene_select == 2 ~ "SPT to entry later",
      #         Data_drug_survival_1$time_palliative_enroll <= Median_entry_1 &
      #           gene_select == 1 ~ "SPT to entry early",
      #         Data_drug_survival_1$time_palliative_enroll > Median_entry_0 &
      #           gene_select == 0 ~ "SPT to entry later"))
      #     
      #     if(sum((Data_drug_survival_1 %>% dplyr::filter(gene_select == 2 & delay_1 == "SPT to entry later"))$censor) >= 1 &
      #        sum((Data_drug_survival_1 %>% dplyr::filter(gene_select == 1 & delay_1 == "SPT to entry later"))$censor) >= 1 &
      #        sum((Data_drug_survival_1 %>% dplyr::filter(gene_select == 0 & delay_1 == "SPT to entry later"))$censor) >= 1 &
      #        sum((Data_drug_survival_1 %>% dplyr::filter(gene_select == 2 & delay_1 != "SPT to entry later"))$censor) >= 1 &
      #        sum((Data_drug_survival_1 %>% dplyr::filter(gene_select == 1 & delay_1 != "SPT to entry later"))$censor) >= 1 &
      #        sum((Data_drug_survival_1 %>% dplyr::filter(gene_select == 0 & delay_1 != "SPT to entry later"))$censor) >= 1){
      #       
      #       stan_weibull_survival_model_data <-
      #         list(
      #           Median_2 = as.integer(Median_entry_2),
      #           Median_1 = as.integer(Median_entry_1),
      #           Median_0 = as.integer(Median_entry_0),
      #           ## Number of event individuals
      #           Nobs = sum(Data_drug_survival_1$censor == 1),
      #           ## Number of censored individuals
      #           Ncen = sum(Data_drug_survival_1$censor == 0),
      #           ## Number of total individuals
      #           Ntot = length(Data_drug_survival_1$censor),
      #           ## Number of covariates
      #           M_bg = 3,
      #           Nexp = length(Data_drug_survival_1_2$time_palliative_enroll),
      #           ## Times for CGP testing
      #           ybef_exp = Data_drug_survival_1_2$time_palliative_enroll,
      #           xfactor = as.numeric(Data_drug_survival_1_2$gene_select == 1),
      #           xfactor2 = as.numeric(Data_drug_survival_1_2$gene_select == 2),
      #           ## Times for event individuals
      #           yobs = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1],
      #           ## Times for censored individuals
      #           ycen = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0],
      #           ## Covariates for event individuals as a matrix
      #           Xobs_bg = matrix(c(as.numeric(Data_drug_survival_1$delay_1 == "SPT to entry later")[Data_drug_survival_1$censor == 1],
      #                              as.numeric(Data_drug_survival_1$gene_select == 1)[Data_drug_survival_1$censor == 1],
      #                              as.numeric(Data_drug_survival_1$gene_select == 2)[Data_drug_survival_1$censor == 1]),
      #                            ncol = 3),
      #           ## Covariates for censored individuals as a matrix
      #           Xcen_bg = matrix(c(as.numeric(Data_drug_survival_1$delay_1 == "SPT to entry later")[Data_drug_survival_1$censor == 0],
      #                              as.numeric(Data_drug_survival_1$gene_select == 1)[Data_drug_survival_1$censor == 0],
      #                              as.numeric(Data_drug_survival_1$gene_select == 2)[Data_drug_survival_1$censor == 0]),
      #                            ncol = 3)
      #         )
      #       
      #       stan_weibull_survival_model_fit <-
      #         rstan::stan(file = "source/Stan_code_loglog_weibull_3_factor.stan",
      #                     data = stan_weibull_survival_model_data,
      #                     control=list(adapt_delta=0.99, max_treedepth = 10),
      #                     seed=1234, chains=8, iter=2000, warmup=1000, thin=1, refresh=500, verbose = FALSE,
      #                     pars = c("alpha","mu", "shape_exp", "scale_exp", "factor", "factor2",
      #                              "yhat_total", "yhat_factor_total", "yhat_factor2_total"))
      #       
      #       stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
      #       stan_weibull_survival_model_draws_total <-
      #         stan_weibull_survival_model_draws %>%
      #         dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
      #         tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
      #         dplyr::select(-key) 
      #       stan_weibull_survival_model_draws_total_exp <-
      #         stan_weibull_survival_model_draws %>%
      #         dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
      #         tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
      #         dplyr::select(-key) 
      #       stan_weibull_survival_model_draws_total_exp2 <-
      #         stan_weibull_survival_model_draws %>%
      #         dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor2_total")) %>%
      #         tidyr::gather(key = key, value = yhat_total_exp2, starts_with("yhat_factor2_total")) %>%
      #         dplyr::select(-key) 
      #       
      #       median_os_list_weibull_total_exp2 = matrix(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2, ncol=4000)
      #       median_os_list_weibull_total_exp2 = colMedians(median_os_list_weibull_total_exp2,na.rm = T)
      #       median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
      #       median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
      #       median_os_list_weibull_total = matrix(stan_weibull_survival_model_draws_total$yhat_total, ncol=4000)
      #       median_os_list_weibull_total = colMedians(median_os_list_weibull_total,na.rm = T)
      #       
      #       median_os_list_weibull_bias_0_1 = median_os_list_weibull_total - median_os_list_weibull_total_exp
      #       median_os_list_weibull_bias_0_2 = median_os_list_weibull_total - median_os_list_weibull_total_exp2
      #       median_os_list_weibull_bias_1_2 = median_os_list_weibull_total_exp - median_os_list_weibull_total_exp2
      #       
      #       median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
      #       median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
      #       median_os_summary_weibull_total_exp2 = quantile(median_os_list_weibull_total_exp2, probs = c(0.025, 0.5, 0.975))
      #       median_os_summary_weibull_bias_0_1 = quantile(median_os_list_weibull_bias_0_1, probs = c(0.025, 0.5, 1 - 0.025))
      #       median_os_summary_weibull_bias_0_2 = quantile(median_os_list_weibull_bias_0_2, probs = c(0.025, 0.5, 1 - 0.025))
      #       median_os_summary_weibull_bias_1_2 = quantile(median_os_list_weibull_bias_1_2, probs = c(0.025, 0.5, 1 - 0.025))
      #       yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
      #       yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
      #       yhat_weibull_sort_total_exp2 = sort(stan_weibull_survival_model_draws_total_exp2$yhat_total_exp2)
      #       yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival$censor) * as.integer(length(yhat_weibull_sort_total)/length(Data_survival$censor))]
      #       yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_survival$censor))]
      #       yhat_weibull_sort_average_total_exp2 = yhat_weibull_sort_total_exp2[1:length(Data_survival$censor) * as.integer(length(yhat_weibull_sort_total_exp2)/length(Data_survival$censor))]
      #       
      #       Data_survival$factor_0 = yhat_weibull_sort_average_total
      #       Data_survival$factor_1 = yhat_weibull_sort_average_total_exp
      #       Data_survival$factor_2 = yhat_weibull_sort_average_total_exp2
      #       Data_survival$factor_0[Data_survival$factor_0 > 10000] = 10000
      #       Data_survival$factor_1[Data_survival$factor_1 > 10000] = 10000
      #       Data_survival$factor_2[Data_survival$factor_2 > 10000] = 10000
      #       
      #       traditional_fit = survfit(Surv(event = censor,
      #                                      time = time_palliative_final) ~ gene_select,
      #                                 data = Data_survival)
      #       simulation_fit_0 = survfit(Surv(event = rep(1,length(Data_survival$factor_0)),
      #                                       time = factor_0) ~ 1, 
      #                                  data = Data_survival)
      #       simulation_fit_1 = survfit(Surv(event = rep(1,length(Data_survival$factor_1)),
      #                                       time = factor_1) ~ 1, 
      #                                  data = Data_survival)
      #       simulation_fit_2 = survfit(Surv(event = rep(1,length(Data_survival$factor_2)),
      #                                       time = factor_2) ~ 1, 
      #                                  data = Data_survival)
      #       
      #       labs(title = paste(sum(traditional_fit[[1]]), " patients ",
      #                          "Median OS ",
      #                          as.integer(summary(traditional_fit)$table[[13]])," (",
      #                          as.integer(summary(traditional_fit)$table[[15]]),"-",
      #                          as.integer(summary(traditional_fit)$table[[17]]),") (",
      #                          name_1_short,", unadj.)/",
      #                          as.integer(summary(traditional_fit)$table[[14]]), " (",
      #                          as.integer(summary(traditional_fit)$table[[16]]),"-",
      #                          as.integer(summary(traditional_fit)$table[[18]]),") (",
      #                          name_2_short,", unadj.)/",
      #                          as.integer(median_os_summary_weibull_total[[2]]), 
      #                          " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
      #                          as.integer(median_os_summary_weibull_total[[3]]), ") (",
      #                          name_1_short,", adj.)/",
      #                          as.integer(median_os_summary_weibull_total_exp[[2]]), " (",
      #                          as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
      #                          as.integer(median_os_summary_weibull_total_exp[[3]]),
      #                          ") (", name_2_short,", adj.) days",
      #                          sep=""),
      #            subtitle = paste("Survival difference, ",name_1_short, "-", name_2_short, ": ",
      #                             as.integer(median_os_summary_weibull_bias[[2]]), " (",
      #                             as.integer(median_os_summary_weibull_bias[[1]]), "-", 
      #                             as.integer(median_os_summary_weibull_bias[[3]]), 
      #                             ") days, median (95% CI)",
      #                             sep=""))
      #       
      #       
      #       g[[k_g]] = ggsurvplot(
      #         fit = list(traditional_fit = traditional_fit,
      #                    simulation_fit_0 = simulation_fit_0,
      #                    simulation_fit_1 = simulation_fit_1,
      #                    simulation_fit_2 = simulation_fit_2),
      #         combine = TRUE,
      #         data = Data_survival,
      #         xlab = "Time from chemotherapy induction to final observation (months)",
      #         ylab = "Survival Probability",
      #         censor = TRUE,
      #         font.title = 8,
      #         font.subtitle = 8,
      #         font.main = 8,
      #         font.submain = 8,
      #         font.caption = 8,
      #         font.legend = 8,
      #         surv.median.line = "hv",
      #         conf.int = FALSE,
      #         pval = FALSE,
      #         risk.table = TRUE,
      #         risk.table.y.text = FALSE,
      #         tables.theme = clean_theme(), 
      #         legend = c(0.8,0.8),
      #         xlim = c(0, max(Data_survival$time_palliative_final) * 1.05),
      #         break.x.by = ceiling(max(Data_survival$time_palliative_final) / 500) * 100,
      #         legend.labs = c(paste(name_1, ", Unadjusted"),
      #                         paste(name_2, ", Unadjusted"),
      #                         paste(name_3, ", Unadjusted"),
      #                         paste(name_1, ", Adjusted for Delayed Entry"),
      #                         paste(name_2, ", Adjusted for Delayed Entry"),
      #                         paste(name_3, ", Adjusted for Delayed Entry"))
      #       ) +
      #         labs(title = paste(sum(traditional_fit[[1]]), " patients ",
      #                            "Median OS ",
      #                            as.integer(summary(traditional_fit)$table[[19]])," (",
      #                            as.integer(summary(traditional_fit)$table[[22]]),"-",
      #                            as.integer(summary(traditional_fit)$table[[25]]),") (",
      #                            name_1_short,", unadj.)/",
      #                            as.integer(summary(traditional_fit)$table[[20]]), " (",
      #                            as.integer(summary(traditional_fit)$table[[23]]),"-",
      #                            as.integer(summary(traditional_fit)$table[[26]]),") (",
      #                            name_2_short,", unadj.)/",
      #                            as.integer(summary(traditional_fit)$table[[21]]), " (",
      #                            as.integer(summary(traditional_fit)$table[[24]]),"-",
      #                            as.integer(summary(traditional_fit)$table[[27]]),") (",
      #                            name_3_short,", unadj.)/\n",
      #                            as.integer(median_os_summary_weibull_total[[2]]), 
      #                            " (", as.integer(median_os_summary_weibull_total[[1]]), "-",
      #                            as.integer(median_os_summary_weibull_total[[3]]), ") (",
      #                            name_1_short,", adj.)/",
      #                            as.integer(median_os_summary_weibull_total_exp[[2]]), 
      #                            " (", as.integer(median_os_summary_weibull_total_exp[[1]]), "-",
      #                            as.integer(median_os_summary_weibull_total_exp[[3]]), ") (",
      #                            name_2_short,", adj.)/",
      #                            as.integer(median_os_summary_weibull_total_exp2[[2]]), " (",
      #                            as.integer(median_os_summary_weibull_total_exp2[[1]]), "-",
      #                            as.integer(median_os_summary_weibull_total_exp2[[3]]),
      #                            ") (", name_3_short,", adj.) days",
      #                            sep=""),
      #              subtitle = paste("Survival difference (95% CI)\n",
      #                               name_1_short, "-", name_2_short, ": ",
      #                               as.integer(median_os_summary_weibull_bias_0_1[[2]]), " (",
      #                               as.integer(median_os_summary_weibull_bias_0_1[[1]]), "-", 
      #                               as.integer(median_os_summary_weibull_bias_0_1[[3]]), "), ",
      #                               name_1_short, "-", name_3_short, ": ",
      #                               as.integer(median_os_summary_weibull_bias_0_2[[2]]), " (",
      #                               as.integer(median_os_summary_weibull_bias_0_2[[1]]), "-", 
      #                               as.integer(median_os_summary_weibull_bias_0_2[[3]]), "), ",
      #                               name_2_short, "-", name_3_short, ": ",
      #                               as.integer(median_os_summary_weibull_bias_1_2[[2]]), " (",
      #                               as.integer(median_os_summary_weibull_bias_1_2[[1]]), "-", 
      #                               as.integer(median_os_summary_weibull_bias_1_2[[3]]),
      #                               ") days, median (95% CI)",
      #                               sep=""))
      #       g[[k_g]]$table <- g[[k_g]]$table + theme(plot.title = element_blank(),
      #                                                plot.subtitle = element_blank())
      #       k_g = k_g + 1
      #     }
      #   }
      # }
      # if(k_g != 4){
      #   for(k_g_ in k_g:3){
      #     g[[k_g_]] = ggsurvplot_empty()
      #   }
      # }
      

      # analysis for common oncogenic mutations
      h = list()
      hs=list()
      k = 1
      oncogenic_genes = Data_MAF_target %>%
        dplyr::select(Tumor_Sample_Barcode, Hugo_Symbol) %>%
        dplyr::distinct() %>%
        dplyr::count(Hugo_Symbol) %>%
        dplyr::arrange(-n)
      colnames(oncogenic_genes) = c("gene_mutation", "all_patients")
      oncogenic_genes = rbind(oncogenic_genes %>% dplyr::filter(gene_mutation %in% input$gene),
                              oncogenic_genes %>% dplyr::filter(!gene_mutation %in% input$gene))
      oncogenic_genes = oncogenic_genes[1:min(length(oncogenic_genes$all_patients), input$gene_no),]
      
      gene_table = data.frame(oncogenic_genes$gene_mutation)
      colnames(gene_table) = c("Gene")
      gene_table$positive_patients = oncogenic_genes$all_patients
      gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(Data_survival$C.CAT調査結果.基本項目.ハッシュID)) / 10
      gene_table$positive_median = 0
      gene_table$positive_LL = 0
      gene_table$positive_UL = 0
      gene_table$negative_median = 0
      gene_table$negative_LL = 0
      gene_table$negative_UL = 0
      gene_table$diff_median = 0
      gene_table$diff_LL = 0
      gene_table$diff_UL = 0
      gene_table$positive_mortal_event = 0
      gene_table$negative_mortal_event = 0
      withProgress(message = sample(nietzsche)[1], {
        for(i in 1:length(oncogenic_genes$all_patients)){
          ID_mutation = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% oncogenic_genes$gene_mutation[i]))$Tumor_Sample_Barcode
          Data_survival = Data_survival %>% dplyr::mutate(
            gene_mut = case_when(
              C.CAT調査結果.基本項目.ハッシュID %in% ID_mutation ~ TRUE,
              TRUE ~ FALSE
            )
          )
          
          mut_gene = paste0(oncogenic_genes$gene_mutation[i],
                           ", top ", i, " gene")
          traditional_fit = survfit(Surv(event = censor,
                                         time = time_palliative_final) ~ gene_mut, 
                                    data = Data_survival)
          if(length(Data_survival[,1]) != traditional_fit[[1]][[1]]){
            gene_table$negative_mortal_event[i] = summary(traditional_fit)$table[7]
            gene_table$positive_mortal_event[i] = summary(traditional_fit)$table[8]

            Data_survival_curve = Data_survival
            Data_survival_tmp_pos = Data_survival_curve %>%
              dplyr::filter(gene_mut == TRUE)
            if(sum(Data_survival_tmp_pos$censor) > 0){
              fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_enroll_final)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_enroll_final)*0.9)
            Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
              time_enroll_final >= time_10 &
                time_enroll_final <= time_90)
            Data_survival_tmp$time_enroll_final =
              Data_survival_tmp$time_enroll_final - (time_10 - 1)
            Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
              time_enroll_final > time_90) %>%
              dplyr::arrange(time_enroll_final)
            
            Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp2)
            Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
            
            Data_survival_tmp_neg = Data_survival_curve %>%
              dplyr::filter(gene_mut == FALSE)
            if(sum(Data_survival_tmp_neg$censor) > 0){
              fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_neg)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_enroll_final)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_enroll_final)*0.9)
            Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
              time_enroll_final >= time_10 &
                time_enroll_final <= time_90)
            Data_survival_tmp3$time_enroll_final =
              Data_survival_tmp3$time_enroll_final - (time_10 - 1)
            Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
              time_enroll_final > time_90) %>%
              dplyr::arrange(time_enroll_final)
            
            Data_survival_tmp4$time_enroll_final = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp4)
            Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx < max(idx)*(1-survival_rate_10)],])
            Data_drug_survival_1 = rbind(Data_survival_tmp, Data_survival_tmp3)  
            
            Data_survival_tmp_pos = Data_survival_curve %>%
              dplyr::filter(gene_mut == TRUE)
            if(sum(Data_survival_tmp_pos$censor) > 0){
              fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_pos)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)
            Data_survival_tmp5 = Data_survival_tmp_pos %>% dplyr::filter(
              time_palliative_enroll >= time_10 &
                time_palliative_enroll <= time_90)
            Data_survival_tmp5$time_palliative_enroll =
              Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
            Data_survival_tmp6 = Data_survival_tmp_pos %>% dplyr::filter(
              time_palliative_enroll > time_90) %>%
              dplyr::arrange(time_palliative_enroll)
            Data_survival_tmp6$time_palliative_enroll = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp6)
            Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
            
            Data_survival_tmp_neg = Data_survival_curve %>%
              dplyr::filter(gene_mut == FALSE)
            
            if(sum(Data_survival_tmp_neg$censor) > 0){
              fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)
            Data_survival_tmp7 = Data_survival_tmp_neg %>% dplyr::filter(
              time_palliative_enroll >= time_10 &
                time_palliative_enroll <= time_90)
            Data_survival_tmp7$time_palliative_enroll =
              Data_survival_tmp7$time_palliative_enroll - (time_10 - 1)
            Data_survival_tmp8 = Data_survival_tmp_neg %>% dplyr::filter(
              time_palliative_enroll > time_90) %>%
              dplyr::arrange(time_palliative_enroll)
            Data_survival_tmp8$time_palliative_enroll = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp8)
            Data_survival_tmp7 = rbind(Data_survival_tmp7, Data_survival_tmp8[idx[idx < max(idx)*(1-survival_rate_10)],])
            
            Median_entry_pos = median(Data_survival_tmp5$time_palliative_enroll)
            Median_entry_neg = median(Data_survival_tmp7$time_palliative_enroll)
            Data_drug_survival_1_2 = rbind(Data_survival_tmp5, Data_survival_tmp7)  
            
            Data_drug_survival_1 = Data_drug_survival_1 %>%
              dplyr::mutate(delay_1 = case_when(
                time_palliative_enroll <= Median_entry_pos &
                  gene_mut == TRUE ~ "SPT to entry early",
                time_palliative_enroll > Median_entry_pos &
                  gene_mut == TRUE ~ "SPT to entry later",
                time_palliative_enroll <= Median_entry_neg &
                  gene_mut == FALSE ~ "SPT to entry early",
                time_palliative_enroll > Median_entry_neg &
                  gene_mut == FALSE ~ "SPT to entry later"))
            
            if(sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
               sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
               sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
               sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
              
              stan_weibull_survival_model_data <-
                list(
                  Median_pos = as.integer(Median_entry_pos),
                  Median_neg = as.integer(Median_entry_neg),
                  Nobs_early_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                  Nobs_late_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                  Ncen_early_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                  Ncen_late_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                  Nobs_early_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                  Nobs_late_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                  Ncen_early_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                  Ncen_late_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                  Nexp_pos = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE]),
                  Nexp_neg = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE]),
                  ybef_exp_pos = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE],
                  ybef_exp_neg = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE],
                  yobs_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                  yobs_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                  ycen_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                  ycen_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                  yobs_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                  yobs_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                  ycen_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                  ycen_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE]
                )
              
              stan_model_path <- "source/Stan_code_loglog_weibull_factor.stan"
              stan_model_compiled <- cmdstan_model(stan_model_path)
              stan_weibull_survival_model_fit <- stan_model_compiled$sample(
                data = stan_weibull_survival_model_data,
                seed = 1234,
                chains = 4,
                parallel_chains = min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE)),
                iter_sampling = 2000,  # (iter-warmup)
                iter_warmup = 1000,
                thin = 1,
                refresh = 500
              )
              stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
              stan_weibull_survival_model_draws_total <-
                stan_weibull_survival_model_draws %>%
                dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
                tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
                dplyr::select(-key) 
              stan_weibull_survival_model_draws_total_exp <-
                stan_weibull_survival_model_draws %>%
                dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
                tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
                dplyr::select(-key) 
              
              median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
              median_os_list_weibull_total = matrix(stan_weibull_survival_model_draws_total$yhat_total, ncol=4000)
              median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
              median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
              median_os_list_weibull_total = colMedians(median_os_list_weibull_total,na.rm = T)
              median_os_list_weibull_bias = colMedians(median_os_list_weibull_bias,na.rm = T)

              median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
              median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
              median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
              yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
              yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
              yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_survival_curve$censor) * as.integer(length(yhat_weibull_sort_total)/length(Data_survival_curve$censor))]
              yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_survival_curve$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_survival_curve$censor))]
              
              Data_survival_curve$factor_neg = yhat_weibull_sort_average_total
              Data_survival_curve$factor_pos = yhat_weibull_sort_average_total_exp
              Data_survival_curve$factor_neg[Data_survival_curve$factor_neg > 10000] = 10000
              Data_survival_curve$factor_pos[Data_survival_curve$factor_pos > 10000] = 10000
              
              traditional_fit = survfit(Surv(event = censor,
                                             time = time_palliative_final) ~ gene_mut,
                                        data = Data_survival_curve)
              simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_survival_curve$factor_neg)),
                                                time = factor_neg) ~ 1, 
                                           data = Data_survival_curve)
              simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_survival_curve$factor_pos)),
                                                time = factor_pos) ~ 1, 
                                           data = Data_survival_curve)
              hs[[k]] = ggsurvplot(
                fit = list(traditional_fit = traditional_fit,
                           simulation_fit_neg = simulation_fit_neg,
                           simulation_fit_pos = simulation_fit_pos),
                combine = TRUE,
                data = Data_survival_curve,
                xlab = "Time from chemotherapy induction to final observation (months)",
                ylab = "Survival Probability",
                censor = TRUE,
                surv.scale = "percent",
                font.title = 8,
                font.subtitle = 8,
                font.main = 8,
                font.submain = 8,
                font.caption = 8,
                font.legend = 8,
                surv.median.line = "v",
                palette = "Dark2",
                conf.int = FALSE,
                pval = FALSE,
                risk.table = TRUE,
                risk.table.y.text = FALSE,
                tables.theme = clean_theme(), 
                legend = c(0.8,0.8),
                xlim = c(0, max(Data_survival_curve$time_palliative_final) * 1.05),
                cumevents = FALSE,
                cumcensor = FALSE,
                break.x.by = 365.25 * 2.5,
                xscale = "d_m",
                #break.x.by = ceiling(max(Data_survival_curve$time_palliative_final) / 500) * 100,
                legend.labs = c(paste(mut_gene, "(-), Unadjusted"),
                                paste(mut_gene, "(+), Unadjusted"),
                                paste(mut_gene, "(-), Adjusted for Delayed Entry"),
                                paste(mut_gene, "(+), Adjusted for Delayed Entry"))
              ) +   
                labs(title = paste(sum(traditional_fit[[1]]), " patients ",
                                   "Median OS ",
                                   round(summary(traditional_fit)$table[[13]] / 365.25 * 12, digits = 1)," (",
                                   round(summary(traditional_fit)$table[[15]] / 365.25 * 12, digits = 1),"-",
                                   round(summary(traditional_fit)$table[[17]] / 365.25 * 12, digits = 1),") (mut(-), unadj.)/",
                                   round(summary(traditional_fit)$table[[14]] / 365.25 * 12, digits = 1), " (",
                                   round(summary(traditional_fit)$table[[16]] / 365.25 * 12, digits = 1),"-",
                                   round(summary(traditional_fit)$table[[18]] / 365.25 * 12, digits = 1),") (mut(+), unadj.)/\n",
                                   round(median_os_summary_weibull_total[[2]] / 365.25 * 12, digits = 1), 
                                   " (", round(median_os_summary_weibull_total[[1]] / 365.25 * 12, digits = 1), "-",
                                   round(median_os_summary_weibull_total[[3]] / 365.25 * 12, digits = 1), ") (mut(-), adj.)/",
                                   round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                                   round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                                   round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1),
                                   ") (mut(+), adj.) months",
                                   sep=""),
                     subtitle = paste("Survival difference with gene mutation: ",
                                      round(median_os_summary_weibull_bias[[2]] / 365.25 * 12, digits = 1), " (",
                                      round(median_os_summary_weibull_bias[[1]] / 365.25 * 12, digits = 1), "-", 
                                      round(median_os_summary_weibull_bias[[3]] / 365.25 * 12, digits = 1), 
                                      ") months, median (95% CI)",
                                      sep=""))
              hs[[k]]$table <- hs[[k]]$table + theme(plot.title = element_blank(),
                                                   plot.subtitle = element_blank())
              k = k + 1
              gene_table$positive_median[i] = round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1)
              gene_table$positive_LL[i] = round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1)
              gene_table$positive_UL[i] = round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1)
              gene_table$negative_median[i] = round(median_os_summary_weibull_total[[2]] / 365.25 * 12, digits = 1)
              gene_table$negative_LL[i] = round(median_os_summary_weibull_total[[1]] / 365.25 * 12, digits = 1)
              gene_table$negative_UL[i] = round(median_os_summary_weibull_total[[3]] / 365.25 * 12, digits = 1)
              gene_table$diff_median[i] = -round(median_os_summary_weibull_bias[[2]] / 365.25 * 12, digits = 1)
              gene_table$diff_LL[i] = -round(median_os_summary_weibull_bias[[3]] / 365.25 * 12, digits = 1)
              gene_table$diff_UL[i] = -round(median_os_summary_weibull_bias[[1]] / 365.25 * 12, digits = 1)
            }
          }
          incProgress(1 / length(oncogenic_genes$all_patients))
        }
        for(kk in k:32){
          hs[[kk]] = ggsurvplot_empty()
        }
      })
      incProgress(1 / 13)
      
      Gene_arrange = gene_table$Gene
      gene_table = gene_table %>% dplyr::mutate(
        Gene = factor(gene_table$Gene, levels = Gene_arrange))
      p <- 
        gene_table |>
        ggplot(aes(y = fct_rev(Gene))) + 
        theme_classic()
      p <- p +
        geom_point(aes(x=diff_median), shape=15, size=3) +
        geom_linerange(aes(xmin=diff_LL, xmax=diff_UL)) 
      p <- p +
        geom_vline(xintercept = 0, linetype="dashed") +
        labs(x="Survival difference by mutation", y="Genes with oncogenic alteration")
      p <- p +
        coord_cartesian(ylim=c(1,length(Gene_arrange) + 1),
                        xlim=c(-max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) - 0.5,
                                max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) + 0.5))
      p <- p +
        annotate("text",
                 x = (-max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) - 0.5)/2,
                 y = length(Gene_arrange) + 1,
                 label = "mut=short survival") +
        annotate("text",
                 x = (max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) + 0.5)/2,
                 y = length(Gene_arrange) + 1,
                 label = "mut=long survival")
      p_mid <- p + 
        theme(axis.line.y = element_blank(),
              axis.ticks.y= element_blank(),
              axis.text.y= element_blank(),
              axis.title.y= element_blank())
      
      gene_table <- gene_table %>%
        dplyr::mutate(
          estimate_lab = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
        dplyr::mutate(
          patients = paste0(positive_patients, " (", positive_freq, "%)"))
      gene_table = 
        dplyr::bind_rows(
          data.frame(
            Gene = "Gene",
            estimate_lab = "Survival difference",
            patients = "Patients"
          ), gene_table  )
      Gene_arrange = gene_table$Gene
      gene_table = gene_table %>% dplyr::mutate(
        Gene = factor(gene_table$Gene, levels = Gene_arrange))
      
      p_left <- gene_table  %>% ggplot(aes(y = fct_rev(Gene)))
      p_left <- p_left + geom_text(aes(x = 0, label = Gene), hjust = 0,
                                   fontface = ifelse(gene_table$Gene == "Gene", "bold", "bold.italic"))
      p_left <- p_left + geom_text(aes(x = 1.4, label = estimate_lab), hjust = 0,
                                   fontface = ifelse(gene_table$estimate_lab == "Survival difference", "bold", "plain"))
      p_left <- p_left + theme_void() + coord_cartesian(xlim = c(0, 4))
      
      # right side of plot - pvalues
      p_right <- gene_table  |> ggplot() +
        geom_text(aes(x = 0, y = fct_rev(Gene), label = patients), hjust = 0,
                  fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")) +
        theme_void()
      
      # final plot arrangement
      layout <- c(patchwork::area(t = 0, l = 0, b = 30, r = 4),
                  patchwork::area(t = 1, l = 4, b = 30, r = 9),
                  patchwork::area(t = 0, l = 9, b = 30, r = 11))
      h[[1]] = p_left + p_mid + p_right + plot_layout(design = layout)
      incProgress(1 / 13)
      
      output$figure_survival_CTx_1 = renderPlot({
        arrange_ggsurvplots(g,print=TRUE,ncol=1,nrow=3)
      })
      incProgress(1 / 13)
      
      output$figure_survival_CTx_2 = renderPlot({
        h[[1]]
      })
      output$figure_survival_CTx_2_data = DT::renderDataTable(gene_table,
                                                              server = FALSE,
                                                              filter = 'top', 
                                                              extensions = c('Buttons'), 
                                                              options = list(pageLength = 100, 
                                                                             scrollX = TRUE,
                                                                             scrollY = "1000px",
                                                                             scrollCollapse = TRUE,
                                                                             dom="Blfrtip",
                                                                             buttons = c('csv', 'excel', 'copy')))
      incProgress(1 / 13)
      output$figure_survival_CTx_3 = renderPlot({
        arrange_ggsurvplots(hs,print=TRUE,ncol=4,nrow=8,surv.plot.height = 0.75,risk.table.height = 0.25)
      })
      incProgress(1 / 13)
      output$figure_survival_CTx_4 = renderPlot({
        hb[[1]]
      })
      output$figure_survival_CTx_4_data = DT::renderDataTable(gene_table_hb,
                                                      server = FALSE,
                                                      filter = 'top', 
                                                      extensions = c('Buttons'), 
                                                      options = list(pageLength = 100, 
                                                                     scrollX = TRUE,
                                                                     scrollY = "1000px",
                                                                     scrollCollapse = TRUE,
                                                                     dom="Blfrtip",
                                                                     buttons = c('csv', 'excel', 'copy')))
      
      incProgress(1 / 13)
      output$figure_survival_CTx_5 = renderPlot({
        arrange_ggsurvplots(hsb,print=TRUE,ncol=4,nrow=4,surv.plot.height = 0.75,risk.table.height = 0.25)
      })
      output$figure_survival_CTx_5_table = DT::renderDataTable(table_ToT,
                                                       server = FALSE,
                                                       filter = 'top', 
                                                       extensions = c('Buttons'), 
                                                       options = list(pageLength = 100, 
                                                                      scrollX = TRUE,
                                                                      scrollY = "1000px",
                                                                      scrollCollapse = TRUE,
                                                                      dom="Blfrtip",
                                                                      buttons = c('csv', 'excel', 'copy')))
      
      incProgress(1 / 13)
    })
  })

  drug_data <- reactiveValues()
  observeEvent(input$drug_table, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed", "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      Data_case_target = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.基本情報.年齢,
                      Lymph_met,
                      Brain_met,
                      Lung_met,
                      Bone_met,
                      Liver_met,
                      Other_met,
                      EP_option,
                      EP_treat,
                      症例.EP前レジメン情報.化学療法レジメン名称,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.終了理由.名称.,
                      症例.EP前レジメン情報.最良総合効果.名称.,
                      症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                      症例.EP後レジメン情報.治療ライン,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.化学療法レジメン名称,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      症例.EP後レジメン情報.レジメン継続区分.名称.,
                      症例.EP後レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.最良総合効果.名称.,
                      症例.基本情報.性別.名称.,
                      症例.基本情報.がん種.OncoTree.,
                      症例.基本情報.がん種.OncoTree..名称.,
                      症例.基本情報.がん種.OncoTree.LEVEL1.,
                      症例.検体情報.パネル.名称.,
                      症例.背景情報.ECOG.PS.名称.,
                      症例.背景情報.喫煙歴有無.名称.,
                      症例.背景情報.アルコール多飲有無.名称.,
                      症例.背景情報.重複がん有無.異なる臓器..名称.,
                      症例.背景情報.多発がん有無.同一臓器..名称.,
                      症例.がん種情報.登録時転移部位.名称.,
                      症例.検体情報.腫瘍細胞含有割合,
                      症例.検体情報.検体採取部位.名称.,
                      症例.転帰情報.転帰.名称.,
                      症例.背景情報.診断日,
                      症例.管理情報.登録日,
                      症例.転帰情報.最終生存確認日,
                      症例.転帰情報.死亡日
        )
      Data_MAF = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","A","B","C","D","E","F") &
            Variant_Classification != "expression"
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
       if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
        Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
      }

      
      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      if(input$patho == "Only pathogenic muts"){
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(Evidence_level == "F")
      } else {
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(!Hugo_Symbol %in% c("TMB", "MSI") | Evidence_level == "F")
        
      }
      Data_MAF_target_tmp = Data_MAF_target  %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        amino.acid.change,
                        .keep_all = TRUE)
      Data_MAF_target = Data_MAF_target  %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        .keep_all = TRUE)

      if(length(Data_case_target[is.na(Data_case_target$症例.EP前レジメン情報.化学療法レジメン名称),]$症例.EP前レジメン情報.化学療法レジメン名称) != 0){
        Data_case_target[is.na(Data_case_target$症例.EP前レジメン情報.化学療法レジメン名称),]$症例.EP前レジメン情報.化学療法レジメン名称 = ""
      }
      if(length(Data_case_target[is.na(Data_case_target$症例.EP後レジメン情報.化学療法レジメン名称),]$症例.EP後レジメン情報.化学療法レジメン名称) != 0){
        Data_case_target[is.na(Data_case_target$症例.EP後レジメン情報.化学療法レジメン名称),]$症例.EP後レジメン情報.化学療法レジメン名称 = ""
      }
      if(length(Data_case_target[is.na(Data_case_target$症例.EP前レジメン情報.最良総合効果.名称.),]$症例.EP前レジメン情報.最良総合効果.名称.) != 0){
        Data_case_target[is.na(Data_case_target$症例.EP前レジメン情報.最良総合効果.名称.),]$症例.EP前レジメン情報.最良総合効果.名称. = "Unknown"
      }
      if(length(Data_case_target[is.na(Data_case_target$症例.EP後レジメン情報.最良総合効果.名称.),]$症例.EP後レジメン情報.最良総合効果.名称.) != 0){
        Data_case_target[is.na(Data_case_target$症例.EP後レジメン情報.最良総合効果.名称.),]$症例.EP後レジメン情報.最良総合効果.名称. = "Unknown"
      }
      
      Data_survival = Data_case_target %>%
        dplyr::mutate(final_observe = case_when(
          症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
          症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
          症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
          TRUE ~ 症例.管理情報.登録日))
      Data_survival = Data_survival %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::mutate(censor = case_when(
          症例.転帰情報.死亡日 != "" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
      
      Data_survival = Data_survival %>%
        dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
        dplyr::mutate(final_observe = ifelse(is.na(final_observe), 症例.管理情報.登録日, final_observe))
      tmp1 = Data_survival %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                        c("その他", "緩和")) %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.化学療法レジメン名称,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      症例.管理情報.登録日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      final_observe,
                      censor,
                      症例.EP前レジメン情報.終了理由.名称.,
                      症例.EP前レジメン情報.最良総合効果.名称.) %>%
        dplyr::distinct()
      tmp1 = tmp1 %>%
        dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                        as.Date(tmp1$症例.EP前レジメン情報.投与開始日)) %>%
        dplyr::mutate(症例.EP前レジメン情報.投与終了日 =
                        as.Date(tmp1$症例.EP前レジメン情報.投与終了日)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(tmp1$症例.管理情報.登録日))
      tmp1 = tmp1 %>%
        dplyr::mutate(Drug_length =
                        as.integer(difftime(tmp1$症例.EP前レジメン情報.投与終了日,
                                            tmp1$症例.EP前レジメン情報.投与開始日,
                                            units = "days")))
      colnames(tmp1) = c("ID",
                         "Drug",
                         "レジメン",
                         "治療ライン",
                         "投与開始日",
                         "投与終了日",
                         "Ongoing",
                         "登録日",
                         "実施目的",
                         "final_observe",
                         "censor",
                         "終了理由",
                         "最良総合効果",
                         "Drug_length")
      tmp1$TxCGP = "Pre"
      tmp2 = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.化学療法レジメン名称,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      症例.EP後レジメン情報.レジメン継続区分.名称.,
                      症例.管理情報.登録日,
                      症例.EP前レジメン情報.実施目的.名称.,
                      final_observe,
                      censor,
                      症例.EP後レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.最良総合効果.名称.) %>%
        dplyr::distinct()
      tmp2$症例.EP前レジメン情報.実施目的.名称. = "緩和"
      tmp2 = tmp2 %>%
        dplyr::mutate(症例.EP後レジメン情報.投与開始日 =
                        as.Date(tmp2$症例.EP後レジメン情報.投与開始日)) %>%
        dplyr::mutate(症例.EP後レジメン情報.投与終了日 =
                        as.Date(tmp2$症例.EP後レジメン情報.投与終了日)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(tmp2$症例.管理情報.登録日))
      
      tmp2 = tmp2 %>%
        dplyr::mutate(Drug_length =
                        as.integer(difftime(tmp2$症例.EP後レジメン情報.投与終了日,
                                            tmp2$症例.EP後レジメン情報.投与開始日,
                                            units = "days")))
      colnames(tmp2) = c("ID",
                         "Drug",
                         "レジメン",
                         "治療ライン",
                         "投与開始日",
                         "投与終了日",
                         "Ongoing",
                         "登録日",
                         "実施目的",
                         "final_observe",
                         "censor",
                         "終了理由",
                         "最良総合効果",
                         "Drug_length")
      tmp2$TxCGP = "Post"
      Data_drug = rbind(tmp1, tmp2)
      
      if(!is.null(input$ID_drug)){
        Data_drug = data.frame(NULL)
        for(i in 1:length(input$ID_drug[,1])){
          tmp = read.csv(header = TRUE, file(input$ID_drug[[i, 'datapath']],
                                   encoding='UTF-8-BOM'))
          if("症例.EP前レジメン情報.投与開始日" %in% colnames(tmp)){
            colnames(tmp) = c("ID",
                               "登録日",
                               "治療ライン",
                               "実施目的",
                               "レジメン",
                               "Drug",
                               "投与開始日",
                               "投与終了日",
                               "Ongoing",
                               "終了理由",
                               "最良総合効果",
                               "最終生存確認日",
                               "死亡日")
            tmp$投与開始日 = as.Date(tmp$投与開始日)
            tmp$投与終了日 = as.Date(tmp$投与終了日)
            tmp = tmp %>%
              dplyr::mutate(
                Drug_length = as.integer(difftime(tmp$投与終了日,
                                                  tmp$投与開始日,
                                                  units = "days")),
                censor = case_when(
                  死亡日 != "" ~ 1,
                  TRUE ~ 0
                ),
                final_observe = case_when(
                  最終生存確認日 == "           " ~ 死亡日,
                  最終生存確認日 != "" ~ 最終生存確認日,
                  死亡日 != "" ~ 死亡日,
                  TRUE ~ 登録日)
              )
            tmp$登録日 = as.Date(tmp$登録日)
            tmp$最終生存確認日 = as.Date(tmp$最終生存確認日)
            tmp$死亡日 = as.Date(tmp$死亡日)
            tmp$final_observe = as.Date(tmp$final_observe)
            
            tmp = tmp %>%
              dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
              dplyr::mutate(final_observe = ifelse(is.na(final_observe), 登録日, final_observe))
            tmp = tmp %>% dplyr::select(-最終生存確認日, -死亡日)
            tmp$TxCGP = "Pre"
          } else{
            colnames(tmp) = c("ID",
                              "登録日",
                              "治療ライン",
                              "レジメン",
                              "Drug",
                              "投与開始日",
                              "投与終了日",
                              "Ongoing",
                              "終了理由",
                              "最良総合効果",
                              "最終生存確認日",
                              "死亡日")
            tmp$投与開始日 = as.Date(tmp$投与開始日)
            tmp$投与終了日 = as.Date(tmp$投与終了日)
            tmp = tmp %>%
              dplyr::mutate(
                Drug_length = as.integer(difftime(tmp$投与終了日,
                                                  tmp$投与開始日,
                                                  units = "days")),
                censor = case_when(
                  死亡日 != "" ~ 1,
                  TRUE ~ 0
                ),
                final_observe = case_when(
                  最終生存確認日 == "           " ~ 死亡日,
                  最終生存確認日 != "" ~ 最終生存確認日,
                  死亡日 != "" ~ 死亡日,
                  TRUE ~ 登録日)
              )
            tmp$登録日 = as.Date(tmp$登録日)
            tmp$最終生存確認日 = as.Date(tmp$最終生存確認日)
            tmp$死亡日 = as.Date(tmp$死亡日)
            tmp$final_observe = as.Date(tmp$final_observe)
            
            tmp = tmp %>%
              dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
              dplyr::mutate(final_observe = ifelse(is.na(final_observe), 登録日, final_observe))
            tmp = tmp %>% dplyr::select(-最終生存確認日, -死亡日)
            tmp$実施目的 = "緩和"
            tmp$TxCGP = "Post"
          }
          Data_drug = rbind(Data_drug, tmp)
          
        }
      }
      
      
      incProgress(1 / 3)
      Data_drug = Data_drug %>%
        dplyr::filter(ID %in% unique(Data_case_target$C.CAT調査結果.基本項目.ハッシュID))
      Data_drug = Data_drug %>%
        dplyr::mutate(final_observe =
                        as.Date(Data_drug$final_observe))
      Data_drug = Data_drug %>%
        dplyr::arrange(投与開始日) %>%
        dplyr::arrange(治療ライン) %>%
        dplyr::arrange(ID)
      
      Data_drug = Data_drug %>% dplyr::mutate(RECIST = case_when(
        最良総合効果 == "CR" ~ "5",
        最良総合効果 == "PR" ~ "4",
        最良総合効果 == "SD" ~ "3",
        最良総合効果 == "PD" ~ "2",
        最良総合効果 == "NE" ~ "1",
        最良総合効果 == "Unknown" ~ "0",
        TRUE ~ "0"
      ))
      Data_drug_5_w_start_date = Data_drug %>%
        dplyr::filter(治療ライン == "５次治療以降" & !is.na(投与開始日))
      Data_drug_5_w_regimen = Data_drug %>%
        dplyr::filter(治療ライン == "５次治療以降" & is.na(投与開始日) & !is.na(レジメン))
      Data_drug_5_wo_info = Data_drug %>%
        dplyr::filter(治療ライン == "５次治療以降" & is.na(投与開始日) & is.na(レジメン))
      Data_drug = Data_drug %>%
        dplyr::filter(治療ライン %in% c("１次治療","２次治療","３次治療","４次治療"))
      
      # Data_drug_Start = Data_drug_5_w_start_date %>%
      #   dplyr::arrange(投与開始日,ID) %>%
      #   dplyr::distinct(ID,投与開始日, .keep_all = T) %>%
      #   dplyr::select(ID,投与開始日)
      Data_drug_End = Data_drug_5_w_start_date %>%
        dplyr::arrange(desc(投与終了日),投与開始日,ID) %>%
        dplyr::distinct(ID,投与開始日, .keep_all = T) %>%
        dplyr::select(ID,投与開始日,投与終了日)
      Data_drug_RECIST = Data_drug_5_w_start_date %>%
        dplyr::arrange(desc(RECIST),投与開始日,ID) %>%
        dplyr::distinct(ID,投与開始日, .keep_all = T) %>%
        dplyr::select(ID,投与開始日, RECIST)
      Data_drug_Ongoing = Data_drug_5_w_start_date %>%
        dplyr::arrange(desc(Ongoing),投与開始日,ID) %>%
        dplyr::distinct(ID,投与開始日, .keep_all = T) %>%
        dplyr::select(ID,投与開始日,Ongoing)
      if(length(Data_drug_Ongoing$ID)>0){
        Data_drug_Ongoing$Drug = ""
        for(samples in 1:length(Data_drug_Ongoing$ID)){
          tmp = sort(unique(str_split(Data_drug_5_w_start_date[Data_drug_5_w_start_date$ID == Data_drug_Ongoing$ID[samples] &
                                                                 Data_drug_5_w_start_date$投与開始日 == Data_drug_Ongoing$投与開始日[samples],]$Drug, ",", simplify = T)[1,]))
          if("" %in% tmp){
            tmp = tmp[tmp != ""]
          }
          Data_drug_Ongoing$Drug[samples] = paste0(tmp, collapse = ",")
        }
      }
      if(length(Data_drug_Ongoing[Data_drug_Ongoing$Drug == "",]$Drug)>0){
        Data_drug_Ongoing$Drug[Data_drug_Ongoing$Drug == ""] = "Unknown"
      }
      
      # if(length(Data_drug_Start[is.na(Data_drug_Start$投与開始日),]$投与開始日)>0){
      #   Data_drug_Start[is.na(Data_drug_Start$投与開始日),]$投与開始日 <- -Inf
      # }
      if(length(Data_drug_End[is.na(Data_drug_End$投与終了日),]$投与終了日)>0){
        Data_drug_End[is.na(Data_drug_End$投与終了日),]$投与終了日 <- -Inf
      }
      
      Data_drug_5_w_start_date = Data_drug_5_w_start_date %>% dplyr::select(-投与終了日, -RECIST, -Ongoing, -Drug)
      # Data_drug_5_w_start_date = left_join(Data_drug_5_w_start_date, Data_drug_Start, by = c('ID','投与開始日'))
      Data_drug_5_w_start_date = left_join(Data_drug_5_w_start_date, Data_drug_End, by = c('ID','投与開始日'))
      Data_drug_5_w_start_date = left_join(Data_drug_5_w_start_date, Data_drug_RECIST, by = c('ID','投与開始日'))
      Data_drug_5_w_start_date = left_join(Data_drug_5_w_start_date, Data_drug_Ongoing, by = c('ID','投与開始日'))
      Data_drug_5_w_start_date = Data_drug_5_w_start_date %>%
        dplyr::arrange(ID) %>%
        dplyr::distinct(ID, 投与開始日, .keep_all = T)
      
      # Data_drug_Start = Data_drug_5_w_regimen %>%
      #   dplyr::arrange(レジメン,ID) %>%
      #   dplyr::distinct(ID,投与開始日, .keep_all = T) %>%
      #   dplyr::select(ID,投与開始日)
      # Data_drug_End = Data_drug_5_w_start_date %>%
      #   dplyr::arrange(desc(投与終了日),投与開始日,ID) %>%
      #   dplyr::distinct(ID,投与開始日, .keep_all = T) %>%
      #   dplyr::select(ID,治療ライン,投与終了日)
      Data_drug_RECIST = Data_drug_5_w_regimen %>%
        dplyr::arrange(desc(RECIST),レジメン,ID) %>%
        dplyr::distinct(ID,レジメン, .keep_all = T) %>%
        dplyr::select(ID,レジメン, RECIST)
      # Data_drug_Ongoing = Data_drug_5_w_start_date %>%
      #   dplyr::arrange(desc(Ongoing),投与開始日,ID) %>%
      #   dplyr::distinct(ID,投与開始日, .keep_all = T) %>%
      #   dplyr::select(ID,治療ライン,Ongoing)
      if(length(Data_drug_RECIST$ID)>0){
        Data_drug_RECIST$Drug = ""
        for(samples in 1:length(Data_drug_RECIST$ID)){
          tmp = sort(unique(str_split(Data_drug_5_w_regimen[Data_drug_5_w_regimen$ID == Data_drug_RECIST$ID[samples] &
                                                              Data_drug_5_w_regimen$レジメン == Data_drug_RECIST$レジメン[samples],]$Drug, ",", simplify = T)[1,]))
          if("" %in% tmp){
            tmp = tmp[tmp != ""]
          }
          Data_drug_RECIST$Drug[samples] = paste0(tmp, collapse = ",")
        }
      }
      if(length(Data_drug_RECIST[Data_drug_RECIST$Drug == "",]$Drug)>0){
        Data_drug_RECIST$Drug[Data_drug_RECIST$Drug == ""] = "Unknown"
      }

      Data_drug_5_w_regimen = Data_drug_5_w_regimen %>% dplyr::select(-RECIST, -Drug)
      Data_drug_5_w_regimen = left_join(Data_drug_5_w_regimen, Data_drug_RECIST, by = c('ID','レジメン'))
      Data_drug_5_w_regimen = Data_drug_5_w_regimen %>%
        dplyr::arrange(ID) %>%
        dplyr::distinct(ID, レジメン, .keep_all = T)
      
      if(length(Data_drug_5_wo_info[Data_drug_5_wo_info$Drug == "",]$Drug)>0){
        Data_drug_5_wo_info$Drug[Data_drug_5_wo_info$Drug == ""] = "Unknown"
      }

      Data_drug_Start = Data_drug %>%
        dplyr::arrange(投与開始日,ID,desc(レジメン),治療ライン) %>%
        dplyr::distinct(ID,治療ライン, .keep_all = T) %>%
        dplyr::select(ID,治療ライン, 投与開始日)
      Data_drug_End = Data_drug %>%
        dplyr::arrange(desc(投与終了日),ID,desc(レジメン),治療ライン) %>%
        dplyr::distinct(ID,治療ライン, .keep_all = T) %>%
        dplyr::select(ID,治療ライン,投与終了日)
      Data_drug_RECIST = Data_drug %>%
        dplyr::arrange(desc(RECIST),ID,desc(レジメン),治療ライン) %>%
        dplyr::distinct(ID,治療ライン,.keep_all = T) %>%
        dplyr::select(ID,治療ライン, RECIST)
      Data_drug_Ongoing = Data_drug %>%
        dplyr::arrange(desc(Ongoing),ID,desc(レジメン),治療ライン) %>%
        dplyr::distinct(ID,治療ライン, .keep_all = T) %>%
        dplyr::select(ID,治療ライン,Ongoing)
      if(length(Data_drug_Ongoing$ID)>0){
        Data_drug_Ongoing$Drug = ""
        for(samples in 1:length(Data_drug_Ongoing$ID)){
          tmp = sort(unique(str_split(Data_drug[Data_drug$ID == Data_drug_Ongoing$ID[samples] &
                                                  Data_drug$治療ライン == Data_drug_Ongoing$治療ライン[samples],]$Drug, ",", simplify = T)[1,]))
          if("" %in% tmp){
            tmp = tmp[tmp != ""]
          }
          Data_drug_Ongoing$Drug[samples] = paste0(tmp, collapse = ",")
        }
      }
      if(length(Data_drug_Ongoing[Data_drug_Ongoing$Drug == "",]$Drug)>0){
        Data_drug_Ongoing$Drug[Data_drug_Ongoing$Drug == ""] = "Unknown"
      }
      
      if(length(Data_drug_Start[is.na(Data_drug_Start$投与開始日),]$投与開始日)>0){
        Data_drug_Start[is.na(Data_drug_Start$投与開始日),]$投与開始日 <- -Inf
      }
      if(length(Data_drug_End[is.na(Data_drug_End$投与終了日),]$投与終了日)>0){
        Data_drug_End[is.na(Data_drug_End$投与終了日),]$投与終了日 <- -Inf
      }
      
      Data_drug = Data_drug %>% dplyr::select(-投与開始日, -投与終了日, -RECIST, -Ongoing, -Drug)
      Data_drug = left_join(Data_drug, Data_drug_Start, by = c('ID','治療ライン'))
      Data_drug = left_join(Data_drug, Data_drug_End, by = c('ID','治療ライン'))
      Data_drug = left_join(Data_drug, Data_drug_RECIST, by = c('ID','治療ライン'))
      Data_drug = left_join(Data_drug, Data_drug_Ongoing, by = c('ID','治療ライン'))
      Data_drug = Data_drug %>%
        dplyr::arrange(ID) %>%
        dplyr::distinct(ID, 治療ライン, .keep_all = T)
      
      if(length(Data_drug_5_w_start_date$ID) == 0){
        Data_drug_5_w_start_date = NULL
      }
      if(length(Data_drug_5_w_regimen$ID) == 0){
        Data_drug_5_w_regimen = NULL
      }
      if(length(Data_drug_5_wo_info$ID) == 0){
        Data_drug_5_wo_info = NULL
      }
      
      Data_drug = rbind(Data_drug, Data_drug_5_w_start_date, Data_drug_5_w_regimen, Data_drug_5_wo_info)
      
      Data_drug = Data_drug %>% dplyr::mutate(最良総合効果 = case_when(
        RECIST == "5" ~ "CR",
        RECIST == "4" ~ "PR",
        RECIST == "3" ~ "SD",
        RECIST == "2" ~ "PD",
        RECIST == "1" ~ "NE",
        RECIST == "0" ~ "Unknown",
        TRUE ~ RECIST
      ))
      
      Data_drug <- data.frame(Data_drug %>% 
                                dplyr::arrange(ID) %>%
                                dplyr::group_by(ID) %>%
                                dplyr::mutate(治療ライン_tmp = row_number()))
      Data_drug$治療ライン_tmp = as.character(Data_drug$治療ライン_tmp)
      Data_drug_tmp = Data_drug %>%
        dplyr::select(ID, 治療ライン) %>%
        dplyr::arrange(治療ライン) %>%
        dplyr::distinct(ID, .keep_all = T)
      colnames(Data_drug_tmp) = c("ID", "input_first_line")
      Data_drug = Data_drug %>% dplyr::left_join(Data_drug_tmp, by="ID")
      Data_drug = Data_drug %>% dplyr::mutate(治療ライン = case_when(
        input_first_line == "１次治療" & 治療ライン_tmp == "1" ~ "1",
        input_first_line == "１次治療" & 治療ライン_tmp == "2" ~ "2",
        input_first_line == "１次治療" & 治療ライン_tmp == "3" ~ "3",
        input_first_line == "１次治療" & 治療ライン_tmp == "4" ~ "4",
        input_first_line == "２次治療" & 治療ライン_tmp == "1" ~ "2",
        input_first_line == "２次治療" & 治療ライン_tmp == "2" ~ "3",
        input_first_line == "２次治療" & 治療ライン_tmp == "3" ~ "4",
        input_first_line == "３次治療" & 治療ライン_tmp == "1" ~ "3",
        input_first_line == "３次治療" & 治療ライン_tmp == "2" ~ "4",
        input_first_line == "４次治療" & 治療ライン_tmp == "1" ~ "4",
        TRUE ~ "5~"
      ))


      Data_drug = Data_drug %>% dplyr::mutate(TTF_censor = case_when(
        投与終了日 < 0 & Ongoing == "継続中" & TxCGP =="Pre" ~ 0,
        TRUE ~ 1
      ))
      Data_drug = Data_drug %>% dplyr::mutate(投与終了日 = case_when(
        投与終了日 < 0 & Ongoing == "継続中" & TxCGP =="Pre" ~ 登録日,
        TRUE ~ 投与終了日
      ))
      Data_drug$投与開始日 = as.Date(Data_drug$投与開始日)
      Data_drug$投与終了日 = as.Date(Data_drug$投与終了日)
      Drug_length_tmp = difftime(Data_drug$投与終了日, Data_drug$投与開始日, units = "days")
      Drug_length_tmp[Drug_length_tmp == -Inf] = NA
      Drug_length_tmp[Drug_length_tmp == Inf] = NA
      Data_drug$Drug_length = as.integer(Drug_length_tmp)
      #Data_drug = Data_drug %>% dplyr::filter(投与開始日 >=0 & 投与終了日 >= 0)
      Data_drug = Data_drug %>%
        dplyr::mutate(
          Drug_length = case_when(
            投与開始日 <0 | 投与終了日 < 0 ~ NA_integer_,
            投与開始日 >投与終了日 ~ NA_integer_,
            is.infinite(Drug_length) ~ NA_integer_,
            TRUE ~ Drug_length
        )
      )
      Data_drug = Data_drug %>%
        dplyr::mutate(
          TTF_censor = case_when(
            is.na(Drug_length) ~ NA_integer_,
            TRUE ~ TTF_censor
          )
        )
      #Data_drug = Data_drug %>% dplyr::filter(Drug_length>0 & !is.na(Drug_length) & is.finite(Drug_length))
      Data_drug = Data_drug %>%
        dplyr::arrange(ID, 治療ライン)
  
      if(!is.null(input$drug_combination_rename)){
        drug_combination_rename_list = data.frame(NULL)
        for(i in 1:length(input$drug_combination_rename[,1])){
          drug_combination_rename_list <- rbind(drug_combination_rename_list,
                                                read.csv(header = TRUE,
                                                         file(input$drug_combination_rename[[i, 'datapath']],
                                                              encoding='UTF-8-BOM')))
        }
        Data_drug$Drug_tmp = unlist(lapply(list(Data_drug$Drug), function(x) {
          as.vector(drug_combination_rename_list$Rename[match(x, drug_combination_rename_list$Drug)])}))
        Data_drug = Data_drug %>% dplyr::mutate(
          Drug = case_when(
            !is.na(Drug_tmp) ~ Drug_tmp,
            TRUE ~ Drug
          )
        )
        Data_drug = Data_drug %>% dplyr::select(-Drug_tmp)
      }

      drug_data$Data_drug = Data_drug
      
      Data_drug = Data_drug %>% dplyr::mutate(最良総合効果 = case_when(
        RECIST == "5" ~ "CR",
        RECIST == "4" ~ "PR",
        RECIST == "3" ~ "SD",
        RECIST == "2" ~ "PD",
        RECIST == "1" ~ "NE",
        RECIST == "0" ~ "Unknown",
        TRUE ~ "Unknown"
      ))
  
      Data_drug$治療ライン = as.character(Data_drug$治療ライン)
      
      Drug_summary =  Data_drug %>%
        dplyr::select(
          ID,
          Drug,
          治療ライン,
          最良総合効果,
          終了理由,
          Drug_length,
          TTF_censor,
      )
      colnames(Drug_summary)=
        c("ID", "Drugs", "CTx line", "RECIST", "Reason to finish treatment", "Time on treatment", "Treatment finished or censored")
      Drug_summary$`Reason to finish treatment`[is.na(Drug_summary$`Reason to finish treatment`)] = "Unknown"
      Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "その他理由で中止"] = "Other reason"
      Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "不明"] = "Unknown"
      Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "副作用等で中止"] = "Side effect"
      Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "本人希望により中止"] = "Patient will"
      Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "無効中止"] = "Not effective"
      Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "死亡中止"] = "Death"
      Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "計画通り終了"] = "As planned"
      Drug_summary$`Treatment finished or censored` = as.character(Drug_summary$`Treatment finished or censored`)
      Drug_summary$`Treatment finished or censored`[is.na(Drug_summary$`Treatment finished or censored`)] = "Unknown"
      Drug_summary$`Treatment finished or censored`[Drug_summary$`Treatment finished or censored` == "1"] = "Finished"
      Drug_summary$`Treatment finished or censored`[Drug_summary$`Treatment finished or censored` == "0"] = "Censored"
      Drug_summary_renamed = Drug_summary
      output$table_drug_all_1 <- render_gt({
        Drug_summary %>%
          dplyr::select(-ID) %>%
          tbl_summary(
            by = "CTx line",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                                "{mean} ({sd})",
                                                                "{median} ({p25}, {p75})", 
                                                                "{min}, {max}"),
                                           all_categorical() ~ "{n} / {N} ({p}%)"),
                          type = list(all_continuous() ~ "continuous2",
                                      all_dichotomous() ~ "categorical"),
                          digits = all_continuous() ~ 2,
                          missing = "ifany") %>%
          modify_caption("**Palliative CTx with treatment duration information, 1st-4th lines**") %>%
          add_n() %>% bold_labels() %>% as_gt()
      })
      output$table_drug_all_2 <- render_gt({
        Drug_summary %>%
          dplyr::select(-ID) %>%
          tbl_summary(
            by = "RECIST",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Palliative CTx with treatment duration information, 1st-4th lines**") %>%
          add_n() %>% bold_labels() %>% as_gt()
      })
      if(!is.null(input$gene_group_1)){
        if(!is.null(input$gene_group_2)){
          Data_MAF_target_tmp = Data_MAF_target_tmp %>%
            dplyr::filter(Tumor_Sample_Barcode %in%
                            Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
            dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
          
          ID_special_gene_mutation_1 = (Data_MAF_target_tmp %>%
                                          dplyr::filter(Hugo_Symbol %in% input$gene_group_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_1, "_"),collapse ="|")) &
                                                          amino.acid.change %in% input$special_gene_mutation_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_1, "_"),collapse ="|")) &
                                                          amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
          ID_special_gene_mutation_2 = (Data_MAF_target_tmp %>%
                                          dplyr::filter(Hugo_Symbol %in% input$gene_group_2 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_2, "_"),collapse ="|")) &
                                                          amino.acid.change %in% input$special_gene_mutation_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_2, "_"),collapse ="|")) &
                                                          amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
          Drug_summary_line_renamed = Drug_summary_renamed %>% dplyr::mutate(
            separation_value = case_when(
              ID %in% ID_special_gene_mutation_1 &
                ID %in% ID_special_gene_mutation_2 ~ paste0(paste(input$gene_group_1, sep="/"), " mut, ", paste0(paste(input$gene_group_2, sep="/"), " mut")),
              ID %in% ID_special_gene_mutation_1 ~ paste0(paste(input$gene_group_1, sep="/"), " mut, ", paste0(paste(input$gene_group_2, sep="/"), " WT")),
              ID %in% ID_special_gene_mutation_2 ~ paste0(paste(input$gene_group_1, sep="/"), " WT, ", paste0(paste(input$gene_group_2, sep="/"), " mut")),
              TRUE ~ paste0("No mut in ",  paste(input$gene_group_1, sep="/"), "/", paste(input$gene_group_2, sep="/"))
            )
          )
        } else {
          Data_MAF_target_tmp = Data_MAF_target_tmp %>%
            dplyr::filter(Tumor_Sample_Barcode %in%
                            Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
            dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
          
          ID_special_gene_mutation_1 = (Data_MAF_target_tmp %>%
                                          dplyr::filter(Hugo_Symbol %in% input$gene_group_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_1, "_"),collapse ="|")) &
                                                          amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
          Drug_summary_line_renamed = Drug_summary_renamed %>% dplyr::mutate(
            separation_value = case_when(
              ID %in% ID_special_gene_mutation_1 ~ paste0(paste(input$gene_group_1, sep="/"), " mut"),
              TRUE ~ paste0("No mut in ", paste(input$gene_group_1, sep="/"))
            )
          )
        }
      } else if(!input$special_gene == ""){
        Data_MAF_target_tmp = Data_MAF_target_tmp %>%
          dplyr::filter(Tumor_Sample_Barcode %in%
                          Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
          dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
        
        ID_special_gene = (Data_MAF_target_tmp %>%
                             dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_"))))$Tumor_Sample_Barcode
        ID_special_gene_mutation_1 = (Data_MAF_target_tmp %>%
                                        dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                        amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
        ID_special_gene_mutation_2 = (Data_MAF_target_tmp %>%
                                        dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                        amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
        Drug_summary_line_renamed = Drug_summary_renamed %>% dplyr::mutate(
          separation_value = case_when(
            ID %in% ID_special_gene_mutation_1 ~ paste0(input$special_gene_mutation_1_name, " in ", input$special_gene),
            ID %in% ID_special_gene_mutation_2 ~ paste0(input$special_gene_mutation_2_name, " in ", input$special_gene),
            ID %in% ID_special_gene ~ paste0("Other mut in ", input$special_gene),
            TRUE ~ paste0("No mut in ", input$special_gene)
          )
        )
      } else {
        Drug_summary_line_renamed = Drug_summary_renamed %>% dplyr::mutate(
          separation_value = "No mutation pattern")
      }

      Drug_summary_line_renamed = Drug_summary_line_renamed %>% dplyr::filter(`CTx line` %in% input$target_line)

      output$table_drug_line_1 <- render_gt({
        Drug_summary_line_renamed %>%
          dplyr::select(-ID) %>%
          tbl_summary(
            by = "separation_value",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Palliative CTx, selected lines, by mutation pattern**") %>%
          add_n() %>% bold_labels() %>% as_gt()
      })
      incProgress(1 / 3)
      
      Drug_summary_diagnosis_renamed = Drug_summary_line_renamed
      ID_diagnosis_list = Data_case_target %>% dplyr::select(C.CAT調査結果.基本項目.ハッシュID,症例.基本情報.がん種.OncoTree.) %>% dplyr::distinct()
      Drug_summary_diagnosis_renamed$diagnosis = unlist(lapply(list(Drug_summary_diagnosis_renamed$ID), function(x) {
          as.vector(ID_diagnosis_list$症例.基本情報.がん種.OncoTree.[match(x, ID_diagnosis_list$C.CAT調査結果.基本項目.ハッシュID)])}))
      output$table_drug_line_2 <- render_gt({
        Drug_summary_diagnosis_renamed %>%
          dplyr::select(-ID, -separation_value) %>%
          tbl_summary(
            by = "diagnosis",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Palliative CTx, selected lines, by diagnosis**") %>%
          add_n() %>% bold_labels() %>% as_gt()
      })
      mut_gene_ = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
      if(length(mut_gene_)==0){
        mut_gene_ = "TP53"
        ID_mutation_ = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene_))$Tumor_Sample_Barcode
        Drug_summary_mutation = Drug_summary_diagnosis_renamed %>% dplyr::mutate(
          mutation = case_when(
            ID %in% ID_mutation_ ~ "TP53 mut(+)",
            TRUE ~ "TP53 mut(-)"
          )
        )
      } else{
        if(is.null(input$gene_group_2)){
          ID_mutation_ = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene_))$Tumor_Sample_Barcode
          Drug_summary_mutation = Drug_summary_diagnosis_renamed %>% dplyr::mutate(
            mutation = case_when(
              ID %in% ID_mutation_ ~ paste0(paste(mut_gene_, collapse = "/"), " mut(+)"),
              TRUE ~ paste0(paste(mut_gene_, collapse = "/"), " mut(-)")
            )
          )
        } else{
          ID_mutation_1 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_1))$Tumor_Sample_Barcode
          ID_mutation_2 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_2))$Tumor_Sample_Barcode
          Drug_summary_mutation = Drug_summary_diagnosis_renamed %>% dplyr::mutate(
            mutation = case_when(
              ID %in% ID_mutation_1 & ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+) & ", paste(input$gene_group_2, collapse = "/"), " mut(+)"),
              ID %in% ID_mutation_1 & !ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+) & ", paste(input$gene_group_2, collapse = "/"), " mut(-)"),
              !ID %in% ID_mutation_1 & ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-) & ", paste(input$gene_group_2, collapse = "/"), " mut(+)"),
              !ID %in% ID_mutation_1 & !ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-) & ", paste(input$gene_group_2, collapse = "/"), " mut(-)"),
            )
          )
        }
      }
      
      output$table_drug_line_3 <- render_gt({
        Drug_summary_mutation %>%
          dplyr::select(-ID, -separation_value, -diagnosis) %>%
          tbl_summary(
            by = "mutation",
            statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                  "{mean} ({sd})",
                                                  "{median} ({p25}, {p75})", 
                                                  "{min}, {max}"),
                             all_categorical() ~ "{n} / {N} ({p}%)"),
            type = list(all_continuous() ~ "continuous2",
                        all_dichotomous() ~ "categorical"),
            digits = all_continuous() ~ 2,
            missing = "ifany") %>%
          modify_caption("**Palliative CTx, selected lines, by mutation**") %>%
          add_n() %>% bold_labels() %>% as_gt()
      })
  
      output$select_drug = renderUI({ 
        pickerInput("drug", "Choose drugs for treatment effect analysis",
                    choices = sort(unique(Drug_summary$Drug)), 
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    selected = names(sort(table(Drug_summary$Drug), decreasing = TRUE))[[1]], multiple = TRUE)
      })
      output$select_drug_group_1 = renderUI({ 
        pickerInput("drug_group_1", "Drug set 1 to be analyzed (if any)",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$drug, multiple = TRUE)
      })
      output$select_drug_group_1_name = renderUI({ 
        textInput(inputId = "drug_group_1_name", label = "Name for Drug set 1",
                  value = input$drug_group_1[[1]])
      })
      output$select_drug_group_2 = renderUI({ 
        pickerInput("drug_group_2", "Drug set 2 to be analyzed (if any)",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$drug[-which(input$drug %in% input$drug_group_1)], multiple = TRUE)
      })
      output$select_drug_group_2_name = renderUI({ 
        textInput(inputId = "drug_group_2_name", label = "Name for Drug set 2",
                  value = input$drug_group_2[[1]])
      })
      output$select_drug_group_3 = renderUI({ 
        pickerInput("drug_group_3", "Drug set 3 to be analyzed (if any)",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$drug[-which(input$drug %in% 
                                        c(input$drug_group_1, input$drug_group_2))], multiple = TRUE)
      })
      output$select_drug_group_3_name = renderUI({ 
        textInput(inputId = "drug_group_3_name", label = "Name for Drug set 3",
                  value = input$drug_group_3[[1]])
      })
      output$select_drug_group_4 = renderUI({ 
        pickerInput("drug_group_4", "Drug set 4 to be analyzed (if any)",
                    options = list(`actions-box` = TRUE, `live-search`=TRUE),
                    input$drug[-which(input$drug %in% 
                                        c(input$drug_group_1, input$drug_group_2, input$drug_group_3))], multiple = TRUE)
      })
      output$select_drug_group_4_name = renderUI({ 
        textInput(inputId = "drug_group_4_name", label = "Name for Drug set 4",
                  value = input$drug_group_4[[1]])
      })
      incProgress(1 / 3)
    })
  })

  observeEvent(input$drug_analysis, {
    withProgress(message = sample(nietzsche)[1], {
      Data_case_target = Data_case()
      if(input$gene_group_analysis %in% c("Only cases with mutations in the gene set are analyzed",
                                          "Only cases without mutations in the gene set are analyzed")){
        Data_case_target = Data_case_target %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                          Data_report()$Tumor_Sample_Barcode)
      }
      Data_case_target = Data_case_target %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                      症例.基本情報.年齢,
                      Lymph_met,
                      Brain_met,
                      Lung_met,
                      Bone_met,
                      Liver_met,
                      Other_met,
                      EP_option,
                      EP_treat,
                      YoungOld,
                      症例.EP前レジメン情報.化学療法レジメン名称,
                      症例.EP前レジメン情報.治療ライン.名称.,
                      症例.EP前レジメン情報.投与開始日,
                      症例.EP前レジメン情報.投与終了日,
                      症例.EP前レジメン情報.レジメン継続区分.名称.,
                      症例.EP前レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP前レジメン情報.実施目的.名称.,
                      症例.EP前レジメン情報.終了理由.名称.,
                      症例.EP前レジメン情報.最良総合効果.名称.,
                      症例.EP後レジメン情報.EPの結果新たな治療薬の選択肢が提示された.名称.,
                      症例.EP後レジメン情報.治療ライン,
                      症例.EP後レジメン情報.治療ライン.名称.,
                      症例.EP後レジメン情報.化学療法レジメン名称,
                      症例.EP後レジメン情報.提示された治療薬を投与した.名称.,
                      症例.EP後レジメン情報.薬剤名.YJ一般名.EN.,
                      症例.EP後レジメン情報.投与開始日,
                      症例.EP後レジメン情報.投与終了日,
                      症例.EP後レジメン情報.レジメン継続区分.名称.,
                      症例.EP後レジメン情報.終了理由.名称.,
                      症例.EP後レジメン情報.最良総合効果.名称.,
                      症例.基本情報.性別.名称.,
                      症例.基本情報.がん種.OncoTree.,
                      症例.基本情報.がん種.OncoTree..名称.,
                      症例.基本情報.がん種.OncoTree.LEVEL1.,
                      症例.検体情報.パネル.名称.,
                      症例.背景情報.ECOG.PS.名称.,
                      症例.背景情報.喫煙歴有無.名称.,
                      症例.背景情報.アルコール多飲有無.名称.,
                      症例.背景情報.重複がん有無.異なる臓器..名称.,
                      症例.背景情報.多発がん有無.同一臓器..名称.,
                      症例.がん種情報.登録時転移部位.名称.,
                      症例.検体情報.腫瘍細胞含有割合,
                      症例.検体情報.検体採取部位.名称.,
                      症例.転帰情報.転帰.名称.,
                      症例.背景情報.診断日,
                      症例.管理情報.登録日,
                      症例.転帰情報.最終生存確認日,
                      症例.転帰情報.死亡日,
                      HER2_IHC,
                      MSI_PCR,
                      MMR_IHC
        )
      Data_MAF = Data_report() %>%
        dplyr::filter(
          !str_detect(Hugo_Symbol, ",") &
            Hugo_Symbol != "" &
            Evidence_level %in% c("","A","B","C","D","E","F") &
            Variant_Classification != "expression"
        ) %>% 
        dplyr::arrange(desc(Evidence_level)) %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        Start_Position,
                        .keep_all = TRUE)
        if(length(Data_MAF[Data_MAF$TMB > 30,]$TMB) > 0){
        Data_MAF[Data_MAF$TMB > 30,]$TMB = 30
      }
      
      Data_MAF_target = Data_MAF %>%
        dplyr::filter(Tumor_Sample_Barcode %in%
                        Data_case_target$C.CAT調査結果.基本項目.ハッシュID)
      if(input$patho == "Only pathogenic muts"){
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(Evidence_level == "F")
      } else {
        Data_MAF_target = Data_MAF_target %>%
          dplyr::filter(!Hugo_Symbol %in% c("TMB", "MSI") | Evidence_level == "F")
      }
      Data_MAF_target_tmp = Data_MAF_target  %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        amino.acid.change,
                        .keep_all = TRUE)
      Data_MAF_target = Data_MAF_target  %>%
        dplyr::distinct(Tumor_Sample_Barcode,
                        Hugo_Symbol,
                        .keep_all = TRUE)
      if(input$drug_volcano != "Yes"){
        Data_cluster_ID_list = Data_cluster_ID() %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID, cluster, driver_mutations)
        Data_case_target = left_join(Data_case_target,
                                     Data_cluster_ID_list,
                                     by = "C.CAT調査結果.基本項目.ハッシュID")
      }
      if(length(Data_case_target[is.na(Data_case_target$症例.EP前レジメン情報.化学療法レジメン名称),]$症例.EP前レジメン情報.化学療法レジメン名称) != 0){
        Data_case_target[is.na(Data_case_target$症例.EP前レジメン情報.化学療法レジメン名称),]$症例.EP前レジメン情報.化学療法レジメン名称 = ""
      }
      if(length(Data_case_target[is.na(Data_case_target$症例.EP後レジメン情報.化学療法レジメン名称),]$症例.EP後レジメン情報.化学療法レジメン名称) != 0){
        Data_case_target[is.na(Data_case_target$症例.EP後レジメン情報.化学療法レジメン名称),]$症例.EP後レジメン情報.化学療法レジメン名称 = ""
      }
      if(length(Data_case_target[is.na(Data_case_target$症例.EP前レジメン情報.最良総合効果.名称.),]$症例.EP前レジメン情報.最良総合効果.名称.) != 0){
        Data_case_target[is.na(Data_case_target$症例.EP前レジメン情報.最良総合効果.名称.),]$症例.EP前レジメン情報.最良総合効果.名称. = "Unknown"}
      if(length(Data_case_target[is.na(Data_case_target$症例.EP後レジメン情報.最良総合効果.名称.),]$症例.EP後レジメン情報.最良総合効果.名称.) != 0){
        Data_case_target[is.na(Data_case_target$症例.EP後レジメン情報.最良総合効果.名称.),]$症例.EP後レジメン情報.最良総合効果.名称. = "Unknown"}
      
      Data_survival = Data_case_target %>%
        dplyr::mutate(final_observe = case_when(
          症例.転帰情報.最終生存確認日 == "           " ~ 症例.転帰情報.死亡日,
          症例.転帰情報.最終生存確認日 != "" ~ 症例.転帰情報.最終生存確認日,
          症例.転帰情報.死亡日 != "" ~ 症例.転帰情報.死亡日,
          TRUE ~ 症例.管理情報.登録日))
      Data_survival = Data_survival %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::mutate(censor = case_when(
          症例.転帰情報.死亡日 != "" ~ 1,
          TRUE ~ 0)) %>%
        dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
      
      Data_survival = Data_survival %>%
        dplyr::mutate(censor = ifelse(is.na(censor), 0, censor)) %>%
        dplyr::mutate(final_observe = ifelse(is.na(final_observe), 症例.管理情報.登録日, final_observe))
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, censor) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival = Data_survival %>%
        dplyr::mutate(final_observe = as.Date(Data_survival$final_observe)) %>%
        dplyr::mutate(症例.管理情報.登録日 =
                        as.Date(Data_survival$症例.管理情報.登録日)) %>%
        dplyr::mutate(症例.EP前レジメン情報.投与開始日 =
                        as.Date(症例.EP前レジメン情報.投与開始日))
      
      Data_survival = Data_survival %>%
        dplyr::mutate(time_enroll_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.管理情報.登録日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_final =
                        as.integer(difftime(Data_survival$final_observe,
                                            Data_survival$症例.EP前レジメン情報.投与開始日,
                                            units = "days")))
      Data_survival = Data_survival %>%
        dplyr::mutate(time_palliative_enroll =
                        Data_survival$time_palliative_final -
                        Data_survival$time_enroll_final)
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, final_observe) %>%
        dplyr::arrange(desc(final_observe)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival_tmp = Data_survival %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_enroll_final) %>%
        dplyr::arrange(desc(time_enroll_final)) %>%
        dplyr::filter(time_enroll_final > 0 & !is.na(time_enroll_final) & is.finite(time_enroll_final)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      incProgress(1 / 13)
      
      Data_survival_tmp = Data_survival %>%
        dplyr::filter(症例.EP前レジメン情報.実施目的.名称. %in%
                        c("その他", "緩和")) %>%
        dplyr::arrange(症例.EP前レジメン情報.投与開始日) %>%
        dplyr::select(C.CAT調査結果.基本項目.ハッシュID, time_palliative_enroll, time_palliative_final, 症例.EP前レジメン情報.投与開始日) %>%
        dplyr::filter(time_palliative_enroll > 0 & time_palliative_enroll < 10000 & !is.na(time_palliative_enroll) & is.finite(time_palliative_enroll)) %>%
        dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE)
      colnames(Data_survival_tmp) = c("C.CAT調査結果.基本項目.ハッシュID", "time_palliative_enroll", "time_palliative_final", "CTx_start_date")
      Data_case_target = left_join(Data_case_target, Data_survival_tmp, by="C.CAT調査結果.基本項目.ハッシュID")
      
      Data_survival = Data_case_target
      incProgress(1 / 13)
      
      Data_drug = drug_data$Data_drug
      significant_genes = NULL
      
      if(input$drug_volcano != "Yes"){
        if(input$drug_other == "Yes"){
          Data_drug = Data_drug %>%
            dplyr::mutate(
              Drug = case_when(
                Drug %in% input$drug ~ Drug,
                TRUE ~ "Other"
              ))
        }
      }
      

      Data_drug_RECIST = Data_drug
      Data_drug_original = Data_drug
      Data_drug = Data_drug %>% dplyr::filter(投与開始日 >=0 & 投与終了日 >= 0 & is.finite(投与開始日) & is.finite(投与終了日))
      Data_drug = Data_drug %>% dplyr::filter(Drug_length>0 & !is.na(Drug_length) & is.finite(Drug_length))
      Data_drug = Data_drug %>%
        dplyr::arrange(ID, 治療ライン)
      
      if(input$drug_volcano != "Yes"){
        Data_drug$TTF_pre_censor = -1
        Data_drug$TTF_pre = -1
        for(tx in 2:(length(Data_drug$ID))){
          if(Data_drug[tx,]$ID == Data_drug[tx-1,]$ID){
            Data_drug[tx,]$TTF_pre = Data_drug[tx-1,]$Drug_length
            Data_drug[tx,]$TTF_pre_censor = Data_drug[tx-1,]$TTF_censor
          }
        }
    
        Data_drug = Data_drug %>%
          dplyr::filter(治療ライン %in% input$target_line)
        
        Drug_summary =  Data_drug%>%
          dplyr::filter(Drug %in% input$drug) %>%
          dplyr::select(
            ID,
            Drug,
            治療ライン,
            最良総合効果,
            終了理由,
            Drug_length,
            TTF_censor,
          )
  
        colnames(Drug_summary)=
          c("ID", "Drugs", "CTx line", "RECIST", "Reason to finish treatment", "Time on treatment", "Treatment finished or censored")
        Drug_summary$`Reason to finish treatment`[is.na(Drug_summary$`Reason to finish treatment`)] = "Unknown"
        Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "その他理由で中止"] = "Other reason"
        Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "不明"] = "Unknown"
        Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "副作用等で中止"] = "Side effect"
        Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "本人希望により中止"] = "Patient will"
        Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "無効中止"] = "Not effective"
        Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "死亡中止"] = "Death"
        Drug_summary$`Reason to finish treatment`[Drug_summary$`Reason to finish treatment` == "計画通り終了"] = "As planned"
        Drug_summary$`Treatment finished or censored` = as.character(Drug_summary$`Treatment finished or censored`)
        Drug_summary$`Treatment finished or censored`[is.na(Drug_summary$`Treatment finished or censored`)] = "Unknown"
        Drug_summary$`Treatment finished or censored`[Drug_summary$`Treatment finished or censored` == "1"] = "Finished"
        Drug_summary$`Treatment finished or censored`[Drug_summary$`Treatment finished or censored` == "0"] = "Censored"
        
        if(!input$special_gene == ""){
          Data_MAF_target_tmp = Data_MAF_target_tmp %>%
            dplyr::filter(Tumor_Sample_Barcode %in%
                            Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
            dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
          
          ID_special_gene = (Data_MAF_target_tmp %>%
                               dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_"))))$Tumor_Sample_Barcode
          ID_special_gene_mutation_1 = (Data_MAF_target_tmp %>%
                                          dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                          amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
          ID_special_gene_mutation_2 = (Data_MAF_target_tmp %>%
                                          dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                          amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
          Drug_summary_line = Drug_summary %>% dplyr::mutate(
            separation_value = case_when(
              ID %in% ID_special_gene_mutation_1 ~ paste0(input$special_gene_mutation_1_name, " in ", input$special_gene),
              ID %in% ID_special_gene_mutation_2 ~ paste0(input$special_gene_mutation_2_name, " in ", input$special_gene),
              ID %in% ID_special_gene ~ paste0("Other mut in ", input$special_gene),
              TRUE ~ paste0("No mut in ", input$special_gene)
            )
          )
        } else{
          Drug_summary_line = Drug_summary %>% dplyr::mutate(
            separation_value = "No mutation pattern")
        }
        output$table_drug_ToT_1 <- render_gt({
          Drug_summary_line %>%
            dplyr::select(-ID) %>%
            tbl_summary(
              by = "separation_value",
              statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                    "{mean} ({sd})",
                                                    "{median} ({p25}, {p75})", 
                                                    "{min}, {max}"),
                               all_categorical() ~ "{n} / {N} ({p}%)"),
              type = list(all_continuous() ~ "continuous2",
                          all_dichotomous() ~ "categorical"),
              digits = all_continuous() ~ 2,
              missing = "ifany") %>%
            modify_caption("**Palliative CTx with treatment duration information, selected lines, by mutation pattern**") %>%
            add_n() %>% bold_labels() %>% as_gt()
        })
        
        Drug_summary_diagnosis = Drug_summary_line
        ID_diagnosis_list = Data_case_target %>% dplyr::select(C.CAT調査結果.基本項目.ハッシュID,症例.基本情報.がん種.OncoTree.) %>% dplyr::distinct()
        Drug_summary_diagnosis$diagnosis = unlist(lapply(list(Drug_summary_diagnosis$ID), function(x) {
          as.vector(ID_diagnosis_list$症例.基本情報.がん種.OncoTree.[match(x, ID_diagnosis_list$C.CAT調査結果.基本項目.ハッシュID)])}))
        output$table_drug_ToT_2 <- render_gt({
          Drug_summary_diagnosis %>%
            dplyr::select(-ID, -separation_value) %>%
            tbl_summary(
              by = "diagnosis",
              statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                    "{mean} ({sd})",
                                                    "{median} ({p25}, {p75})", 
                                                    "{min}, {max}"),
                               all_categorical() ~ "{n} / {N} ({p}%)"),
              type = list(all_continuous() ~ "continuous2",
                          all_dichotomous() ~ "categorical"),
              digits = all_continuous() ~ 2,
              missing = "ifany") %>%
            modify_caption("**Palliative CTx with treatment duration information, selected lines, by diagnosis**") %>%
            add_n() %>% bold_labels() %>% as_gt()
        })
        mut_gene_ = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
        if(length(mut_gene_)==0){
          mut_gene_ = "TP53"
          ID_mutation_ = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene_))$Tumor_Sample_Barcode
          Drug_summary_mutation = Drug_summary_diagnosis %>% dplyr::mutate(
            mutation = case_when(
              ID %in% ID_mutation_ ~ "TP53 mut(+)",
              TRUE ~ "TP53 mut(-)"
            )
          )
        } else{
          if(is.null(input$gene_group_2)){
            ID_mutation_ = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene_))$Tumor_Sample_Barcode
            Drug_summary_mutation = Drug_summary_diagnosis %>% dplyr::mutate(
              mutation = case_when(
                ID %in% ID_mutation_ ~ paste0(paste(mut_gene_, collapse = "/"), " mut(+)"),
                TRUE ~ paste0(paste(mut_gene_, collapse = "/"), " mut(-)")
              )
            )
          } else{
            ID_mutation_1 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_1))$Tumor_Sample_Barcode
            ID_mutation_2 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_2))$Tumor_Sample_Barcode
            Drug_summary_mutation = Drug_summary_diagnosis %>% dplyr::mutate(
              mutation = case_when(
                ID %in% ID_mutation_1 & ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+) & ", paste(input$gene_group_2, collapse = "/"), " mut(+)"),
                ID %in% ID_mutation_1 & !ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+) & ", paste(input$gene_group_2, collapse = "/"), " mut(-)"),
                !ID %in% ID_mutation_1 & ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-) & ", paste(input$gene_group_2, collapse = "/"), " mut(+)"),
                !ID %in% ID_mutation_1 & !ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-) & ", paste(input$gene_group_2, collapse = "/"), " mut(-)"),
              )
            )
          }
        }
        incProgress(1 / 13)
        
        output$table_drug_ToT_3 <- render_gt({
          Drug_summary_mutation %>%
            dplyr::select(-ID, -separation_value, -diagnosis) %>%
            tbl_summary(
              by = "mutation",
              statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                    "{mean} ({sd})",
                                                    "{median} ({p25}, {p75})", 
                                                    "{min}, {max}"),
                               all_categorical() ~ "{n} / {N} ({p}%)"),
              type = list(all_continuous() ~ "continuous2",
                          all_dichotomous() ~ "categorical"),
              digits = all_continuous() ~ 2,
              missing = "ifany") %>%
            modify_caption("**Palliative CTx with treatment duration information, selected lines, by mutation**") %>%
            add_n() %>% bold_labels() %>% as_gt()
        })
        
        Data_drug$serial = 1:length(Data_drug$ID)
    
        Data_drug_TTF = Data_drug %>% dplyr::distinct(ID, Drug, .keep_all = T)
        ga = list()
        gb = list()
        gc = list()
        kk = 1
        Data_TTF_compare = Data_drug %>%
          dplyr::filter(TTF_pre>0) %>%
          dplyr::distinct(ID, Drug, .keep_all = T)
        if(length(Data_TTF_compare$TTF_pre)>0){
          Data_TTF_compare$Drug_length = as.numeric(Data_TTF_compare$Drug_length)
          Data_TTF_compare$TTF_pre = as.numeric(Data_TTF_compare$TTF_pre)
          Data_TTF_compare$TTF_censor_shape = as.character(4 + 16 * Data_TTF_compare$TTF_censor)
          Data_TTF_compare$TTF_pre_censor_shape = as.character(4 + 16 * Data_TTF_compare$TTF_pre_censor)
          Data_TTF_compare = Data_TTF_compare %>% dplyr::mutate(
            TTF_censor_color = case_when(
              TTF_censor == 0 ~ "censored",
              TTF_censor == 1 ~ "finished",
              TRUE ~ "Unknown"
            )
          )
          Data_TTF_compare = Data_TTF_compare %>% dplyr::mutate(
            TTF_pre_censor_color = case_when(
              TTF_pre_censor == 0 ~ "censored",
              TTF_pre_censor == 1 ~ "finished",
              TRUE ~ "Unknown"
            )
          )
          Data_TTF_compare = Data_TTF_compare %>% dplyr::mutate(
            TTF_censor_double_color = case_when(
              TTF_censor == 0  & TTF_pre_censor == 0 ~ "target:censored, previous:censored",
              TTF_censor == 1  & TTF_pre_censor == 0 ~ "target:finished, previous:censored",
              TTF_censor == 0  & TTF_pre_censor == 1 ~ "target:censored, previous:finished",
              TTF_censor == 1  & TTF_pre_censor == 1 ~ "target:finished, previous:finished",
              TRUE ~ as.character(TTF_censor)
            )
          )
          Data_TTF_compare_tmp = Data_TTF_compare %>% dplyr::filter(Drug %in% input$drug) %>%
            dplyr::arrange(Drug_length)
          if(dim(Data_TTF_compare_tmp)[1]>0){
            Data_TTF_compare_tmp$ind <- factor(seq(1:length(Data_TTF_compare_tmp$ID)))
            Data_TTF_compare_tmp = Data_TTF_compare_tmp %>%
              dplyr::arrange(TTF_pre)
            Data_TTF_compare_tmp$ind2 <- factor(seq(1:length(Data_TTF_compare_tmp$ID)))
            g.mid<-ggplot(Data_TTF_compare_tmp,aes(x=1,y=ind))+
              geom_segment(aes(x=0.94,xend=0.96,yend=ind))+
              geom_segment(aes(x=1.04,xend=1.065,yend=ind))+
              ggtitle("")+
              ylab(NULL)+
              scale_x_continuous(expand=c(0,0),limits=c(0.94,1.065))+
              theme(axis.title=element_blank(),
                    legend.position = 'none',
                    panel.grid=element_blank(),
                    axis.text.y=element_blank(),
                    axis.ticks.y=element_blank(),
                    panel.background=element_blank(),
                    axis.text.x=element_text(color=NA),
                    axis.ticks.x=element_line(color=NA),
                    plot.margin = unit(c(1,-1,1,-1), "mm"))
            g1 <- ggplot(data = Data_TTF_compare_tmp, aes(x = ind, y = TTF_pre)) +
              geom_bar(stat = "identity") + ggtitle("Previous line, red:finished, blue:censored") +
              geom_point(aes(x = ind, y = TTF_pre,shape = TTF_pre_censor_shape,colour = TTF_pre_censor_color),size=2)+
              theme(axis.title.x = element_blank(),
                    legend.position = 'none',
                    axis.title.y = element_blank(),
                    axis.text.y = element_blank(),
                    axis.ticks.y = element_blank(),
                    plot.margin = unit(c(1,-1,1,0), "mm")) +
              scale_y_reverse() + coord_flip() +
              scale_colour_brewer(palette = "Dark2")
            g2 <- ggplot(data = Data_TTF_compare_tmp, aes(x = ind, y = Drug_length)) +xlab(NULL)+
              geom_bar(stat = "identity") + ggtitle(paste0("Time on treatment: ", paste(input$drug, collapse = ";"))) +
              geom_point(aes(x = ind, y = Drug_length,shape = TTF_censor_shape,colour = TTF_censor_color),size=2)+
              theme(axis.title.x = element_blank(), axis.title.y = element_blank(),
                    legend.position = 'none',
                    axis.text.y = element_blank(), axis.ticks.y = element_blank(),
                    plot.margin = unit(c(1,0,1,-1), "mm")) +
              coord_flip() +
              scale_colour_brewer(palette = "Dark2")
            gg1 <- ggplot_gtable(ggplot_build(g1))
            gg2 <- ggplot_gtable(ggplot_build(g2))
            gg.mid <- ggplot_gtable(ggplot_build(g.mid))
            ga[[1]] = grid.arrange(gg1,gg.mid,gg2,ncol=3,widths=c(9/19,1/19,9/19))
        
            Data_tmp_1 = Data_TTF_compare_tmp %>% dplyr::select(
              Drug_length, TTF_censor)
            Data_tmp_2 = Data_TTF_compare_tmp %>% dplyr::select(
              TTF_pre, TTF_pre_censor)
            colnames(Data_tmp_2) = colnames(Data_tmp_1)
            Data_tmp_1$line = "Time on treatment"
            Data_tmp_2$line = "Time on treatment, previous-line"
            Data_tmp = rbind(Data_tmp_1, Data_tmp_2)
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~line, data=Data_tmp,type = "kaplan-meier", conf.type = "log-log")
            diff_0 = survdiff(Surv(Drug_length, TTF_censor)~line,
                              data=Data_tmp, rho=0)
            diff_1 = survdiff(Surv(Drug_length, TTF_censor)~line,
                              data=Data_tmp, rho=1)
            diff_2 = coxph(Surv(Drug_length, TTF_censor)~line,
                              data=Data_tmp)
            gb[[kk]] = surv_curv_drug(survfit_t, Data_tmp, paste0("Time on treatment, ", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
            kk=kk+1
            Data_TTF_compare_tmp = Data_TTF_compare_tmp %>% dplyr::filter(TTF_pre_censor == 1 & TTF_censor == 1)
            
            if(length(Data_TTF_compare_tmp$TTF_pre)>2){
              RegModel.1 <- lm(TTF_pre~Drug_length, data=Data_TTF_compare_tmp)
              equation <- sprintf("Y = %.3g + %.3g * X", coef(RegModel.1)[1], coef(RegModel.1)[2])
              res <- cor.test(Data_TTF_compare_tmp$TTF_pre, Data_TTF_compare_tmp$Drug_length, alternative="two.sided", method="pearson")
              res_1 = ifelse(is.null(res$estimate), NA, round(res$estimate,3))
              res_2 = ifelse(is.null(res$conf.int[1]), NA, round(res$conf.int[1],3))
              res_3 = ifelse(is.null(res$conf.int[2]), NA, round(res$conf.int[2],3))
              equation = paste0("r=", res_1, " (95%CI: ", res_2, "-", res_3, "), ", equation)
              ga[[2]] = Data_TTF_compare_tmp %>%
                ggplot(aes(x = Drug_length, y = TTF_pre, colour = TTF_censor_double_color)) +
                geom_point()+
                geom_smooth(method=lm, se=FALSE, color="black", show.legend=FALSE)+
                theme_classic()+
                annotate("text", x = max(Data_TTF_compare_tmp$Drug_length)/2, y=-max(Data_TTF_compare_tmp$TTF_pre)/60, label=equation, parse=F) +
                labs(x = "Time on treatment of designated treatment, only patients without censoring", y = "Time on treatment of the previous line, only patients without censoring", title = paste0("ToT of target-line/previous-line:",paste(input$drug, collapse = ";"))) +
                scale_colour_brewer(palette = "Dark2")
            } else {
              ga[[2]] = Data_TTF_compare_tmp %>%
                ggplot(aes(x = Drug_length, y = TTF_pre, colour = TTF_censor_double_color)) +
                geom_point()+
                theme_classic()+
                labs(x = "Time on treatment, only pts without censoring", y = "Previous line, only pts without censoring", title = paste0("ToT of target-line/previous-line:",paste(input$drug, collapse = ";"))) +
                scale_colour_brewer(palette = "Dark2")
            }
          }
        } else {
          Data_TTF_compare_tmp = NULL
        }
        incProgress(1 / 13)
        
        if(length(Data_drug_TTF$TTF_censor) > 0){
          survfit_t <- survfit(Surv(Drug_length, TTF_censor)~1, data=Data_drug_TTF,type = "kaplan-meier", conf.type = "log-log")
          gb[[kk]] = surv_curv_drug(survfit_t, Data_drug_TTF, paste0("Time on treatment, all drugs"), NULL, NULL, NULL)
          kk = kk + 1
        }
        mut_gene = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
        if(is.null(input$gene)){
          mut_gene = "TP53"
        }
    
        Data_drug_TTF = Data_drug_TTF %>% dplyr::mutate(mut = case_when(
          ID %in% unique((Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene))$Tumor_Sample_Barcode) ~ "+",
          TRUE ~ "-"))
        if(length(unique(Data_drug_TTF$mut)) > 1){
          survfit_t <- survfit(Surv(Drug_length, TTF_censor)~mut, data=Data_drug_TTF,type = "kaplan-meier", conf.type = "log-log")
          if(length(Data_drug_TTF[,1]) != survfit_t[[1]][[1]]){
              
            diff_0 = survdiff(Surv(Drug_length, TTF_censor)~mut,
                              data=Data_drug_TTF, rho=0)
            diff_1 = survdiff(Surv(Drug_length, TTF_censor)~mut,
                              data=Data_drug_TTF, rho=1)
            diff_2 = coxph(Surv(Drug_length, TTF_censor)~mut,
                              data=Data_drug_TTF)
            gb[[kk]] = surv_curv_drug(survfit_t, Data_drug_TTF, paste("Time on treatment, all drugs,", paste0(collapse = ";", mut_gene) , "mut"), diff_0, diff_1 ,diff_2)
            kk = kk + 1
          }
        }
        
        Data_drug_TTF = Data_drug_TTF %>% dplyr::mutate(Regimen_type = case_when(
          Drug %in% input$drug ~ paste(input$drug, collapse = ";"),
          TRUE ~ "Other drugs"))
        if(length(unique(Data_drug_TTF$Regimen_type)) > 1){
          survfit_t <- survfit(Surv(Drug_length, TTF_censor)~Regimen_type, data=Data_drug_TTF,type = "kaplan-meier", conf.type = "log-log")
          if(length(Data_drug_TTF[,1]) != survfit_t[[1]][[1]]){
            diff_0 = survdiff(Surv(Drug_length, TTF_censor)~Regimen_type,
                              data=Data_drug_TTF, rho=0)
            diff_1 = survdiff(Surv(Drug_length, TTF_censor)~Regimen_type,
                              data=Data_drug_TTF, rho=1)
            diff_2 = coxph(Surv(Drug_length, TTF_censor)~Regimen_type,
                              data=Data_drug_TTF)
            gb[[kk]] = surv_curv_drug(survfit_t, Data_drug_TTF, paste0("Time on treatment, ", paste(input$drug, collapse = ";")), diff_0, diff_1 ,diff_2)
            kk = kk + 1
          }
        }
        
        Data_drug_TTF = Data_drug_TTF %>% dplyr::mutate(Regimen = case_when(
          Drug %in% input$drug_group_1 ~ input$drug_group_1_name,
          Drug %in% input$drug_group_2 ~ input$drug_group_2_name,
          Drug %in% input$drug_group_3 ~ input$drug_group_3_name,
          Drug %in% input$drug_group_4 ~ input$drug_group_4_name,
          TRUE ~ "Other drugs"))
        Data_drug_TTF_drug_select = Data_drug_TTF %>% dplyr::filter(Drug != "Other drugs")
        if(length(unique(Data_drug_TTF_drug_select$Regimen)) > 1){
          survfit_t <- survfit(Surv(Drug_length, TTF_censor)~Regimen, data=Data_drug_TTF_drug_select,type = "kaplan-meier", conf.type = "log-log")
          if(length(Data_drug_TTF_drug_select[,1]) != survfit_t[[1]][[1]]){
            diff_0 = survdiff(Surv(Drug_length, TTF_censor)~Regimen,
                              data=Data_drug_TTF_drug_select, rho=0)
            diff_1 = survdiff(Surv(Drug_length, TTF_censor)~Regimen,
                              data=Data_drug_TTF_drug_select, rho=1)
            diff_2 = coxph(Surv(Drug_length, TTF_censor)~Regimen,
                              data=Data_drug_TTF_drug_select)
            gb[[kk]] = surv_curv_drug(survfit_t, Data_drug_TTF_drug_select, "Time on treatment, selected drugs", diff_0, diff_1, diff_2)
            kk = kk + 1
          }
        }
        
        if(length(Data_drug_TTF$mut) > 0){
          survfit_t <- survfit(Surv(Drug_length, TTF_censor)~mut+Regimen_type, data=Data_drug_TTF,type = "kaplan-meier", conf.type = "log-log")
          if(length(unique(Data_drug_TTF$mut)) > 1 & length(unique(Data_drug_TTF$Regimen_type)) > 1){
            diff_0 = survdiff(Surv(Drug_length, TTF_censor)~mut+Regimen_type,
                              data=Data_drug_TTF, rho=0)
            diff_1 = survdiff(Surv(Drug_length, TTF_censor)~mut+Regimen_type,
                              data=Data_drug_TTF, rho=1)
            diff_2 = coxph(Surv(Drug_length, TTF_censor)~mut+Regimen_type,
                              data=Data_drug_TTF)
            gb[[kk]] = surv_curv_drug(survfit_t, Data_drug_TTF, paste("Time on treatment,", paste0(collapse = ";", mut_gene), "and", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
            kk = kk + 1
          }
        }
        
        # Survival analysis for mutation, all
        ID_mutation = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene))$Tumor_Sample_Barcode
        for(jj in 1:length(mut_gene)){
          ID_mutation = ID_mutation[ID_mutation %in% (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene[jj]))$Tumor_Sample_Barcode]
        }
        if(length(ID_mutation) > 0){
          Data_drug_TTF = Data_drug_TTF %>% dplyr::mutate(
            mut = case_when(
              ID %in% ID_mutation ~ "All genes",
              ID %in% (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene))$Tumor_Sample_Barcode ~ "Any genes",
              TRUE ~ "No genes"
            )
          )
          Data_drug_TTF_mut = Data_drug_TTF %>% dplyr::filter(Regimen_type == paste(input$drug, collapse = ";"))
          if(length(unique(Data_drug_TTF_mut$mut)) > 1){
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~mut, data=Data_drug_TTF_mut,type = "kaplan-meier", conf.type = "log-log")
            if(length(Data_drug_TTF_mut[,1]) != survfit_t[[1]][[1]]){
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~mut,
                                data=Data_drug_TTF_mut, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~mut,
                                data=Data_drug_TTF_mut, rho=1)
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~mut,
                                data=Data_drug_TTF_mut)
              gb[[kk]] = surv_curv_drug(survfit_t, Data_drug_TTF_mut, paste("Time on treatment,", paste0(collapse = ";", mut_gene), "and", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
              kk = kk + 1
            }
          }
        }
        # Survival analysis for lymph node metastasis
        ID_mutation_lymph_yes = (Data_case_target %>% dplyr::filter(Lymph_met %in% c("Yes")))$C.CAT調査結果.基本項目.ハッシュID
        ID_mutation_lymph_no = (Data_case_target %>% dplyr::filter(Lymph_met %in% c("No")))$C.CAT調査結果.基本項目.ハッシュID
        Data_drug_TTF = Data_drug_TTF %>% dplyr::mutate(
          met = case_when(
              ID %in% ID_mutation_lymph_yes ~ "lymph met (+)",
              ID %in% ID_mutation_lymph_no ~ "lymph met (-)",
              TRUE ~ "Unknown"
            )
          )
        Data_drug_TTF_met = Data_drug_TTF %>%
          dplyr::filter(Regimen_type == paste(input$drug, collapse = ";")) %>%
          dplyr::filter(met != "Unknown")
        if(length(unique(Data_drug_TTF_met$met)) > 1){
          survfit_t <- survfit(Surv(Drug_length, TTF_censor)~met, data=Data_drug_TTF_met,type = "kaplan-meier", conf.type = "log-log")
          if(length(Data_drug_TTF_met[,1]) != survfit_t[[1]][[1]]){
            diff_0 = survdiff(Surv(Drug_length, TTF_censor)~met,
                              data=Data_drug_TTF_met, rho=0)
            diff_1 = survdiff(Surv(Drug_length, TTF_censor)~met,
                              data=Data_drug_TTF_met, rho=1)
            diff_2 = coxph(Surv(Drug_length, TTF_censor)~met,
                              data=Data_drug_TTF_met)
            gb[[kk]] = surv_curv_drug(survfit_t, Data_drug_TTF_met, paste("Lymph node metastasis, time on treatment with ", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
            kk = kk + 1
          }
        }
        if(kk != 9){
          for(kkk in kk:8){
            gb[[kkk]] = ggsurvplot_empty()
          }
        }
        Data_drug_TTF$diagnosis = ""
        for(target_diseases in unique(Data_case_target$症例.基本情報.がん種.OncoTree.)){
          Data_drug_TTF = Data_drug_TTF %>% dplyr::mutate(diagnosis = case_when(
            ID %in% unique((Data_case_target %>% dplyr::filter(症例.基本情報.がん種.OncoTree. == target_diseases))$C.CAT調査結果.基本項目.ハッシュID) ~ target_diseases,
            TRUE ~ diagnosis))
        }
  
        output$figure_drug_1 = renderPlot({
          if(dim(Data_TTF_compare_tmp)[1]>0){
            grid.arrange(ga[[1]], ga[[2]], nrow=2,  ncol = 1)
          }
        })
        incProgress(1 / 13)
        
        output$figure_drug_2 = renderPlot({
          arrange_ggsurvplots(gb,print=TRUE,ncol=2,nrow=4)
        })
        k_3 = 1
        if(length(unique(Data_drug_TTF$diagnosis)) >= 1){
          diagnosis_list = (Data_drug_TTF %>%
                              dplyr::distinct(ID, .keep_all = T))$diagnosis
          diagnosis_list = names(sort(table(diagnosis_list),decreasing = T))[1:7]
          Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::mutate(
            diagnosis = case_when(
              diagnosis %in% diagnosis_list ~ diagnosis,
              TRUE ~ "Other histology"
            )
          )
          Data_drug_TTF_tmp2 = Data_drug_TTF_tmp
          Data_drug_TTF_tmp2$diagnosis = " ALL"
          Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::filter(
            diagnosis %in% diagnosis_list)
          Data_drug_TTF_tmp2 = rbind(Data_drug_TTF_tmp, Data_drug_TTF_tmp2)
          Data_drug_TTF_tmp$diagnosis = as.factor(Data_drug_TTF_tmp$diagnosis)
          Data_drug_TTF_tmp2$diagnosis = as.factor(Data_drug_TTF_tmp2$diagnosis)
          Data_drug_TTF_tmp$diagnosis = relevel(Data_drug_TTF_tmp$diagnosis, ref=names(sort(table(Data_drug_TTF_tmp$diagnosis),decreasing = T))[[1]]) 
          Data_drug_TTF_tmp2$diagnosis = relevel(Data_drug_TTF_tmp2$diagnosis, ref=names(sort(table(Data_drug_TTF_tmp2$diagnosis),decreasing = T))[[1]]) 
          if(length(Data_drug_TTF_tmp$TTF_censor) > 0){
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~diagnosis, data=Data_drug_TTF_tmp,type = "kaplan-meier", conf.type = "log-log")
            if(length(unique(Data_drug_TTF_tmp$diagnosis)) > 1){
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~diagnosis,
                                data=Data_drug_TTF_tmp, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~diagnosis,
                                data=Data_drug_TTF_tmp, rho=1)
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~diagnosis,
                                data=Data_drug_TTF_tmp)
              gc[[k_3]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp, paste0("Time on treatment, diagnosis, all drugs"), diff_0, diff_1, diff_2)
              k_3 = k_3 + 1
            } else{
              gc[[k_3]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp, paste0("Time on treatment, diagnosis, all drugs"), NULL, NULL, NULL)
              k_3 = k_3 + 1
            }
          }
          if(length(Data_drug_TTF_tmp2$TTF_censor) > 0){
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~diagnosis, data=Data_drug_TTF_tmp2,type = "kaplan-meier", conf.type = "log-log")
            if(length(unique(Data_drug_TTF_tmp2$diagnosis)) > 1){
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~diagnosis,
                                data=Data_drug_TTF_tmp2, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~diagnosis,
                                data=Data_drug_TTF_tmp2, rho=1)
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~diagnosis,
                                data=Data_drug_TTF_tmp2)
              gc[[k_3]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp2, paste0("Time on treatment, diagnosis, all drugs"), diff_0, diff_1, diff_2)
              k_3 = k_3 + 1
            } else{
              gc[[k_3]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp2, paste0("Time on treatment, diagnosis, all drugs"), NULL, NULL,NULL)
              k_3 = k_3 + 1
            }
          }
          
          Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::filter(Drug %in% input$drug)
          Data_drug_TTF_tmp = Data_drug_TTF_tmp %>% dplyr::filter(
            diagnosis %in% diagnosis_list)
          if(length(unique(Data_drug_TTF_tmp$diagnosis)) >= 1){
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~diagnosis, data=Data_drug_TTF_tmp,type = "kaplan-meier", conf.type = "log-log")
            if(length(unique(Data_drug_TTF_tmp$diagnosis)) > 1){
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~diagnosis,
                                data=Data_drug_TTF_tmp, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~diagnosis,
                                data=Data_drug_TTF_tmp, rho=1)
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~diagnosis,
                                data=Data_drug_TTF_tmp)
              gc[[k_3]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp, paste0("Time on treatment, diagnosis, treated with ", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
              k_3 = k_3 + 1
            } else {
              gc[[k_3]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp, paste0("Time on treatment, diagnosis, treated with ", paste(input$drug, collapse = ";")), NULL, NULL)
              k_3 = k_3 + 1
            }
          }
          
          if(!is.null(input$gene_group_1)){
            if(!is.null(input$gene_group_2)){
              Data_MAF_target_tmp = Data_MAF_target %>%
                dplyr::filter(Tumor_Sample_Barcode %in%
                                Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
                dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
              
              ID_special_gene_mutation_1 = (Data_MAF_target_tmp %>%
                                              dplyr::filter(Hugo_Symbol %in% input$gene_group_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_1, "_"),collapse ="|")) &
                                                              amino.acid.change %in% input$special_gene_mutation_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_1, "_"),collapse ="|")) &
                                                              amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
              ID_special_gene_mutation_2 = (Data_MAF_target_tmp %>%
                                              dplyr::filter(Hugo_Symbol %in% input$gene_group_2 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_2, "_"),collapse ="|")) &
                                                              amino.acid.change %in% input$special_gene_mutation_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_2, "_"),collapse ="|")) &
                                                              amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
              Data_drug_TTF_tmp = Data_drug_TTF_tmp %>% dplyr::mutate(
                mutation_type = case_when(
                  ID %in% ID_special_gene_mutation_1 &
                    ID %in% ID_special_gene_mutation_2 ~ paste0(paste(input$gene_group_1, sep="/"), " mut, ", paste0(paste(input$gene_group_2, sep="/"), " mut")),
                  ID %in% ID_special_gene_mutation_1 ~ paste0(paste(input$gene_group_1, sep="/"), " mut, ", paste0(paste(input$gene_group_2, sep="/"), " WT")),
                  ID %in% ID_special_gene_mutation_2 ~ paste0(paste(input$gene_group_1, sep="/"), " WT, ", paste0(paste(input$gene_group_2, sep="/"), " mut")),
                  TRUE ~ paste0("No mut in ",  paste(input$gene_group_1, sep="/"), "/", paste(input$gene_group_2, sep="/"))
                )
              )
            } else {
              Data_MAF_target_tmp = Data_MAF_target %>%
                dplyr::filter(Tumor_Sample_Barcode %in%
                                Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
                dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
              
              ID_special_gene_mutation_1 = (Data_MAF_target_tmp %>%
                                              dplyr::filter(Hugo_Symbol %in% input$gene_group_1 | str_detect(Hugo_Symbol, paste(paste0(input$gene_group_1, "_"),collapse ="|")) &
                                                              amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
              Drug_summary_line_renamed = Drug_summary_renamed %>% dplyr::mutate(
                mutation_type = case_when(
                  ID %in% ID_special_gene_mutation_1 ~ paste0(paste(input$gene_group_1, sep="/"), " mut"),
                  TRUE ~ paste0("No mut in ", paste(input$gene_group_1, sep="/"))
                )
              )
            }
          } else if(!input$special_gene == ""){
            Data_MAF_target_tmp = Data_MAF_target %>%
              dplyr::filter(Tumor_Sample_Barcode %in%
                              Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
              dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
            
            ID_special_gene = (Data_MAF_target_tmp %>%
                                 dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_"))))$Tumor_Sample_Barcode
            ID_special_gene_mutation_1 = (Data_MAF_target_tmp %>%
                                            dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                            amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
            ID_special_gene_mutation_2 = (Data_MAF_target_tmp %>%
                                            dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                            amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
            Data_drug_TTF_tmp = Data_drug_TTF_tmp %>% dplyr::mutate(
              mutation_type = case_when(
                ID %in% ID_special_gene_mutation_1 ~ paste0(input$special_gene_mutation_1_name, " in ", input$special_gene),
                ID %in% ID_special_gene_mutation_2 ~ paste0(input$special_gene_mutation_2_name, " in ", input$special_gene),
                ID %in% ID_special_gene ~ paste0("Other mut in ", input$special_gene),
                TRUE ~ paste0("No mut in ", input$special_gene)
              )
            )
          } else {
            Data_drug_TTF_tmp = Data_drug_TTF_tmp %>% dplyr::mutate(
              mutation_type = "No mutation pattern")
          }
          
          Data_age_sex = Data_case_target %>%
            dplyr::select(C.CAT調査結果.基本項目.ハッシュID, YoungOld, 症例.基本情報.性別.名称.) %>%
            dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = T)
          Data_drug_TTF_PS = Data_drug_TTF_tmp %>%
            dplyr::left_join(Data_age_sex, by=c("ID" = "C.CAT調査結果.基本項目.ハッシュID"))
          Data_drug_TTF_PS$mutation_type <- as.factor(Data_drug_TTF_PS$mutation_type)
          freq_table <- table(Data_drug_TTF_PS$mutation_type)
          most_common_level <- names(freq_table)[which.max(freq_table)]
          Data_drug_TTF_PS$mutation_type <- relevel(Data_drug_TTF_PS$mutation_type, ref = most_common_level)
          # propensity score matching considering sex, age, histology
          if(length(unique(Data_drug_TTF_PS$mutation_type)) >= 2){
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~mutation_type, data=Data_drug_TTF_PS,type = "kaplan-meier", conf.type = "log-log")
            if(length(Data_drug_TTF_PS[,1]) != survfit_t[[1]][[1]]){
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~mutation_type,
                                data=Data_drug_TTF_PS, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~mutation_type,
                                data=Data_drug_TTF_PS, rho=1)
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~mutation_type,
                             data=Data_drug_TTF_PS)
              gc[[k_3]] = surv_curv_drug(survfit_t, Data_drug_TTF_PS, paste("Non-PS-matched. Time on treatment, treated with", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
              k_3 = k_3 + 1
            }
            Factor_PS = NULL
            if(length(unique(Data_drug_TTF_PS$diagnosis)) > 1){
              Factor_PS = c(Factor_PS, "diagnosis")
              Data_drug_TTF_PS$diagnosis <- as.factor(Data_drug_TTF_PS$diagnosis)
            }
            if(length(unique(Data_drug_TTF_PS$症例.基本情報.性別.名称.)) > 1){
              Factor_PS = c(Factor_PS, "症例.基本情報.性別.名称.")
              Data_drug_TTF_PS$症例.基本情報.性別.名称. <- as.factor(Data_drug_TTF_PS$症例.基本情報.性別.名称.)
            }
            if(length(unique(Data_drug_TTF_PS$YoungOld)) > 1){
              Factor_PS = c(Factor_PS, "YoungOld")
              Data_drug_TTF_PS$YoungOld <- as.factor(Data_drug_TTF_PS$YoungOld)
            }
  
            formula_PS = formula(
              paste0("mutation_type ~", paste(Factor_PS, collapse = "+"))
            )
            
            if(length(unique(Data_drug_TTF_PS$mutation_type)) >= 3){
              gps_model <- mnps(mutation_type ~ YoungOld + 症例.基本情報.性別.名称. + diagnosis,
                                data = Data_drug_TTF_PS,
                                estimand = "ATE",
                                stop.method = "es.mean",
                                n.trees = 3000,
                                verbose = FALSE)
            } else {
              mut_type_name = unique(Data_drug_TTF_PS$mutation_type)
              Data_drug_TTF_PS$mutation_type_binary = ifelse(Data_drug_TTF_PS$mutation_type == mut_type_name[1], 0, 1)
              gps_model <- ps(mutation_type_binary ~ YoungOld + 症例.基本情報.性別.名称. + diagnosis,
                                data = Data_drug_TTF_PS,
                                estimand = "ATE",
                                stop.method = "es.mean",
                                n.trees = 3000,
                                verbose = FALSE)
            }
            Data_drug_TTF_PS$weights <- get.weights(gps_model, stop.method = "es.mean", estimand = "ATE")
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~mutation_type, data=Data_drug_TTF_PS,type = "kaplan-meier", conf.type = "log-log", weights = weights)
            if(length(unique(Data_drug_TTF_PS$mutation_type)) > 1){
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~mutation_type,
                             data=Data_drug_TTF_PS, weights = weights)
              gc[[k_3]] = surv_curv_drug_IPTW(survfit_t, Data_drug_TTF_PS, paste0("PS matched: age/sex/diagnosis. Time on treatment, treated with ", paste(input$drug, collapse = ";"), "\ncompare to 1st strata"), diff_2)
              summary_data = surv_summary(survfit_t, data = Data_drug_TTF_PS)
              risk_table <- summary_data %>%
                group_by(time, strata) %>%
                summarise(n_risk = sum(n.risk), .groups = 'drop') %>%
                pivot_wider(names_from = strata, values_from = n_risk)
              risk_table_long <- risk_table %>%
                pivot_longer(-time, names_to = "Group", values_to = "n_risk")
              k_3 = k_3 + 1
              }
            }
          if(k_3 < 9){
            for(kkkk in k_3:8){
              gc[[kkkk]] = ggsurvplot_empty()
            }
          }
        } else {
          for(kkkk in 1:8){
            gc[[kkkk]] = ggsurvplot_empty()
          }
        }
        output$figure_drug_3 = renderPlot({
          arrange_ggsurvplots(gc,print=TRUE,ncol=2,nrow=4)
        })
        if(length(unique(Data_drug_TTF$diagnosis)) >= 1){
          diagnosis_list = (Data_drug_TTF %>%
                              dplyr::distinct(ID, .keep_all = T))$diagnosis
          diagnosis_list = names(sort(table(diagnosis_list),decreasing = T))[1:8]
          Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::mutate(
            diagnosis = case_when(
              diagnosis %in% diagnosis_list ~ diagnosis,
              TRUE ~ "Other histology"
            )
          )
          Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::filter(
            diagnosis %in% diagnosis_list)
          if(length(Data_drug_TTF_tmp$TTF_censor) > 0){
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~diagnosis, data=Data_drug_TTF_tmp,type = "kaplan-meier", conf.type = "log-log")
            table_ToT <- 
              data.frame(Cohort = as.character(),
                         Duration_months = as.character(),
                         No_total = as.numeric(),
                         No_at_risk = as.numeric(),
                         No_event = as.numeric(),
                         No_censor = as.numeric(),
                         Survival_rate = as.numeric(),
                         Survival_rate_95LL = as.numeric(),
                         Survival_rate_95UL = as.numeric())
            row_no = 1
            for(cohort in 1:length(summary(survfit_t, times=365.25 / 12, extend=TRUE)$time)){
              for(mths in (0:12)*3){
                dataset = summary(survfit_t, times=365.25 / 12* mths, extend=TRUE)
                if(length(unique(Data_drug_TTF$diagnosis)) > 1){
                  table_ToT[row_no,"Cohort"] = rownames(dataset$table)[cohort]
                } else{
                  table_ToT[row_no,"Cohort"] = unique(Data_drug_TTF_tmp$diagnosis)
                }
                table_ToT[row_no,"Duration_months"] = mths
                table_ToT[row_no,"No_total"] = dataset$n[cohort]
                table_ToT[row_no,"No_at_risk"] = dataset$n.risk[cohort]
                table_ToT[row_no,"No_event"] = dataset$n.event[cohort]
                table_ToT[row_no,"No_censor"] = dataset$n.censor[cohort]
                table_ToT[row_no,"Survival_rate"] = dataset$surv[cohort]
                table_ToT[row_no,"Survival_rate_95LL"] = dataset$lower[cohort]
                table_ToT[row_no,"Survival_rate_95UL"] = dataset$upper[cohort]
                row_no = row_no + 1
              }
            }
          }
        }
        output$figure_drug_3_table = DT::renderDataTable(table_ToT,
                                                         server = FALSE,
                                                         filter = 'top', 
                                                         extensions = c('Buttons'), 
                                                         options = list(pageLength = 100, 
                                                                        scrollX = TRUE,
                                                                        scrollY = "1000px",
                                                                        scrollCollapse = TRUE,
                                                                        dom="Blfrtip",
                                                                        buttons = c('csv', 'excel', 'copy')))
        
        
        Data_drug_volcano_all_ToT = Data_drug_TTF
        drugs_volcano = sort(table(Data_drug_volcano_all_ToT$Drug),decreasing = T)
        drugs_volcano = drugs_volcano[!names(drugs_volcano) %in% c("Others", "Unknown", "")]
        drugs_volcano = drugs_volcano[!is.na(names(drugs_volcano))]
        
        df_volcano_ToT = data.frame(Regimen = as.character(),
                                Gene = as.character(),
                                Total_treated_patients = as.numeric(),
                                Mut_patients = as.numeric(),
                                HazardRatio_considering_pathology = as.numeric(),
                                p_value = as.numeric()
        )
        df_volcano_tmp = data.frame(Regimen = "",
                                    Gene = "",
                                    Total_treated_patients = 0,
                                    Mut_patients = 0,
                                    HazardRatio_considering_pathology = 1,
                                    p_value = 1
        )
        figure_no = 1
        g_volcano_ToT = list()
        options(warn = -1)
        withProgress(message = sample(nietzsche)[1], {
          for(drug_name in names(drugs_volcano)){
            Data_drug_volcano = Data_drug_volcano_all_ToT %>%
              dplyr::filter(Drug == drug_name) %>%
              dplyr::select(ID, Drug_length, TTF_censor, diagnosis) %>%
              dplyr::distinct(ID, .keep_all = T)
            if(length(Data_drug_volcano$ID) >= 1){
              Data_MAF_target_volcano = Data_MAF_target_tmp %>%
                dplyr::filter(Tumor_Sample_Barcode %in%
                                Data_drug_volcano$ID) %>%
                dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol)

              colnames(Data_drug_volcano) = c("ID", "Drug_length", "TTF_censor", "Diagnosis")
              colnames(Data_MAF_target_volcano) = c("ID", "Gene")
              Data_MAF_target_volcano = dplyr::left_join(Data_MAF_target_volcano, Data_drug_volcano, by="ID")

              df_volcano_tmp$Regimen = drug_name
              df_volcano_tmp$Total_treated_patients = length(Data_drug_volcano$ID)
              gene_table_volcano = sort(table(Data_MAF_target_volcano$Gene),decreasing = T)
              gene_table_volcano = gene_table_volcano[gene_table_volcano>=8]
              if(length(gene_table_volcano) > 0){
                for(gene_volcano in names(gene_table_volcano)){
                  Data_drug_volcano_pos = Data_MAF_target_volcano %>%
                    dplyr::filter(Gene == gene_volcano) %>%
                    dplyr::select(-Gene)
                  Data_drug_volcano_neg = Data_MAF_target_volcano %>%
                    dplyr::select(-Gene) %>%
                    dplyr::distinct()
                  Data_drug_volcano_neg = setdiff(Data_drug_volcano_neg, Data_drug_volcano_pos)
                  if(length(Data_drug_volcano_pos$ID) > 0){
                    Data_drug_volcano_pos$mutation = 1
                  }
                  if(length(Data_drug_volcano_neg$ID) > 0){
                    Data_drug_volcano_neg$mutation = 0
                  }
                  if(length(Data_drug_volcano_pos$ID) > 0 & length(Data_drug_volcano_neg$ID) > 0){
                    Data_drug_volcano_neg_pos = rbind(Data_drug_volcano_pos, Data_drug_volcano_neg)
                  } else if(length(Data_drug_volcano_pos$ID) > 0){
                    Data_drug_volcano_neg_pos = Data_drug_volcano_pos
                  } else{
                    Data_drug_volcano_neg_pos = Data_drug_volcano_neg
                  }
                  df_volcano_tmp$Gene = gene_volcano
                  df_volcano_tmp$Mut_patients = length(Data_drug_volcano_pos$ID)
                  Data_drug_volcano_neg_pos$Diagnosis = factor(Data_drug_volcano_neg_pos$Diagnosis)
                  Data_drug_volcano_neg_pos$Diagnosis = relevel(Data_drug_volcano_neg_pos$Diagnosis, ref=names(sort(table(Data_drug_volcano_neg_pos$Diagnosis),decreasing = T))[[1]])
                  if(input$volcano_pathology == "Yes" &
                     min(table(Data_drug_volcano_neg_pos$Diagnosis))>1 &
                     max(table(Data_drug_volcano_neg_pos$Diagnosis))<(length(Data_drug_volcano_neg_pos$Diagnosis)-1) &
                     sum(Data_drug_volcano_neg_pos$mutation)>1 &
                     sum(Data_drug_volcano_neg_pos$mutation)<(length(Data_drug_volcano_neg_pos$mutation)-1) &
                     sum(Data_drug_volcano_neg_pos$TTF_censor)>1 &
                     sum(Data_drug_volcano_neg_pos$TTF_censor)<(length(Data_drug_volcano_neg_pos$TTF_censor)-1)){
                    Data_ORR = coxph(Surv(Drug_length, TTF_censor)~mutation+Diagnosis,
                                     data=Data_drug_volcano_neg_pos)
                    data_tmp = tidy(Data_ORR, exponentiate=TRUE, conf.int=TRUE)[1,]
                    df_volcano_tmp$HazardRatio_considering_pathology = data_tmp$estimate
                    df_volcano_tmp$p_value = data_tmp$p.value
                  } else {
                    major_pathology = names(table(Data_drug_volcano_neg_pos$Diagnosis)[table(Data_drug_volcano_neg_pos$Diagnosis) > 1])
                    Data_drug_volcano_neg_pos_major = Data_drug_volcano_neg_pos %>%
                      dplyr::filter(Diagnosis %in% major_pathology)
                    Data_drug_volcano_neg_pos_major$Diagnosis = factor(Data_drug_volcano_neg_pos_major$Diagnosis)
                    Data_drug_volcano_neg_pos_major$Diagnosis = relevel(Data_drug_volcano_neg_pos_major$Diagnosis, ref=names(sort(table(Data_drug_volcano_neg_pos$Diagnosis),decreasing = T))[[1]])
                    if(input$volcano_pathology == "Yes" &
                       min(table(Data_drug_volcano_neg_pos_major$Diagnosis))>1 &
                       max(table(Data_drug_volcano_neg_pos_major$Diagnosis))<(length(Data_drug_volcano_neg_pos_major$Diagnosis)-1) &
                       sum(Data_drug_volcano_neg_pos_major$mutation)>1 &
                       sum(Data_drug_volcano_neg_pos_major$mutation)<(length(Data_drug_volcano_neg_pos_major$mutation)-1) &
                       sum(Data_drug_volcano_neg_pos_major$TTF_censor)>1 &
                       sum(Data_drug_volcano_neg_pos_major$TTF_censor)<(length(Data_drug_volcano_neg_pos_major$TTF_censor)-1)){
                      Data_ORR = coxph(Surv(Drug_length, TTF_censor)~mutation+Diagnosis,
                                       data=Data_drug_volcano_neg_pos)
                      data_tmp = tidy(Data_ORR, exponentiate=TRUE, conf.int=TRUE)[1,]
                      df_volcano_tmp$HazardRatio_considering_pathology = data_tmp$estimate
                      df_volcano_tmp$p_value = data_tmp$p.value
                    } else if(sum(Data_drug_volcano_neg_pos$mutation)>1 &
                              sum(Data_drug_volcano_neg_pos$mutation)<(length(Data_drug_volcano_neg_pos$mutation)-1) &
                              sum(Data_drug_volcano_neg_pos$TTF_censor)>1 &
                              sum(Data_drug_volcano_neg_pos$TTF_censor)<(length(Data_drug_volcano_neg_pos$TTF_censor)-1)){
                      Data_ORR = coxph(Surv(Drug_length, TTF_censor)~mutation,
                                       data=Data_drug_volcano_neg_pos)
                      data_tmp = tidy(Data_ORR, exponentiate=TRUE, conf.int=TRUE)[1,]
                      df_volcano_tmp$HazardRatio_considering_pathology = data_tmp$estimate
                      df_volcano_tmp$p_value = data_tmp$p.value
                    } else {
                      df_volcano_tmp$HazardRatio_considering_pathology = 1
                      df_volcano_tmp$p_value = 1
                    }
                  }
                  df_volcano_ToT = rbind(df_volcano_ToT, df_volcano_tmp)
                }
                
                if(figure_no<200){
                  Figure_volcano_tmp = df_volcano_ToT %>%
                    dplyr::filter(Regimen == drug_name) %>%
                    dplyr::mutate(Gene_effect = ifelse(p_value >= 0.05 , "Not significant",
                                                       ifelse(HazardRatio_considering_pathology > 1, "Poor",
                                                              ifelse(HazardRatio_considering_pathology < 1, "Good","Not significant"))))
                  Figure_volcano_tmp$HazardRatio_considering_pathology[Figure_volcano_tmp$HazardRatio_considering_pathology>16]=16
                  Figure_volcano_tmp$HazardRatio_considering_pathology[Figure_volcano_tmp$HazardRatio_considering_pathology<1/16]=1/16
                  Figure_volcano_tmp$Gene_effect = factor(Figure_volcano_tmp$Gene_effect, levels = c("Good", "Not significant", "Poor"))
                  Figure_volcano_tmp$label = ifelse((Figure_volcano_tmp$p_value < 0.05), Figure_volcano_tmp$Gene, NA_character_)
                  
                  Figure_volcano_tmp$log2OR = log2(Figure_volcano_tmp$HazardRatio_considering_pathology)
                  Figure_volcano_tmp$log10pval = -log10(Figure_volcano_tmp$p_value)
                  if(length(Figure_volcano_tmp$label>0)){
                    g_volcano_ToT[[figure_no]] =
                      ggplot(data = Figure_volcano_tmp, aes(x = log2OR, y = log10pval, col = Gene_effect, label = label)) +
                      geom_point(size = 2) +
                      labs(x = expression("log"[2]*"hazard ratio of time on treatment"), y = expression("-log"[10]*"p-value")) +
                      scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"), 
                                         labels = c("Good", "Not significant", "Poor"),
                                         drop = FALSE) +
                      geom_hline(yintercept = -log10(0.05), size = 0.2, color = "dark green") + 
                      geom_vline(xintercept = -1, size = 0.2, color = "dark green", linetype = 'dashed') +
                      geom_vline(xintercept = 1, size = 0.2, color = "dark green", linetype = 'dashed') +
                      theme(legend.position = "none", panel.grid = element_blank()) +
                      geom_text_repel(max.overlaps = Inf) +
                      ggtitle(paste0(drug_name, " ",
                                     df_volcano_tmp$Total_treated_patients, " patients")) +
                      theme_classic()
                  }
                  figure_no = figure_no + 1
                }
              }
            }
            incProgress(1 / length(names(drugs_volcano)))
          }
        })
        options(warn = 0)
        if(!is.null(input$drug)){
          Data_drug_volcano = Data_drug_volcano_all_ToT %>%
            dplyr::filter(Drug %in% input$drug) %>%
            dplyr::select(ID, Drug_length, TTF_censor, diagnosis) %>%
            dplyr::distinct(ID, .keep_all = T)
          if(length(Data_drug_volcano$ID) >= 1){
            Data_MAF_target_volcano = Data_MAF_target_tmp %>%
              dplyr::filter(Tumor_Sample_Barcode %in%
                              Data_drug_volcano$ID) %>%
              dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol)
            colnames(Data_drug_volcano) = c("ID", "Drug_length", "TTF_censor", "Diagnosis")
            colnames(Data_MAF_target_volcano) = c("ID", "Gene")
            Data_MAF_target_volcano = dplyr::left_join(Data_MAF_target_volcano, Data_drug_volcano, by="ID")
            df_volcano_tmp$Regimen = paste(input$drug, collapse = ";")
            df_volcano_tmp$Total_treated_patients = length(Data_drug_volcano$ID)
            gene_table_volcano = sort(table(Data_MAF_target_volcano$Gene),decreasing = T)
            gene_table_volcano = gene_table_volcano[gene_table_volcano>=8]
            if(length(gene_table_volcano) > 0){
              for(gene_volcano in names(gene_table_volcano)){
                Data_drug_volcano_pos = Data_MAF_target_volcano %>%
                  dplyr::filter(Gene == gene_volcano) %>%
                  dplyr::select(-Gene)
                Data_drug_volcano_neg = Data_MAF_target_volcano %>%
                  dplyr::select(-Gene) %>%
                  dplyr::distinct()
                Data_drug_volcano_neg = setdiff(Data_drug_volcano_neg, Data_drug_volcano_pos)
                if(length(Data_drug_volcano_pos$ID) > 0){
                  Data_drug_volcano_pos$mutation = 1
                }
                if(length(Data_drug_volcano_neg$ID) > 0){
                  Data_drug_volcano_neg$mutation = 0
                }
                if(length(Data_drug_volcano_pos$ID) > 0 & length(Data_drug_volcano_neg$ID) > 0){
                  Data_drug_volcano_neg_pos = rbind(Data_drug_volcano_pos, Data_drug_volcano_neg)
                } else if(length(Data_drug_volcano_pos$ID) > 0){
                  Data_drug_volcano_neg_pos = Data_drug_volcano_pos
                } else {
                  Data_drug_volcano_neg_pos = Data_drug_volcano_neg
                }

                df_volcano_tmp$Gene = gene_volcano
                df_volcano_tmp$Mut_patients = length(Data_drug_volcano_pos$ID)
                Data_drug_volcano_neg_pos$Diagnosis = factor(Data_drug_volcano_neg_pos$Diagnosis)
                Data_drug_volcano_neg_pos$Diagnosis = relevel(Data_drug_volcano_neg_pos$Diagnosis, ref=names(sort(table(Data_drug_volcano_neg_pos$Diagnosis),decreasing = T))[[1]])
                if(input$volcano_pathology == "Yes" &
                   min(table(Data_drug_volcano_neg_pos$Diagnosis))>1 &
                   max(table(Data_drug_volcano_neg_pos$Diagnosis))<(length(Data_drug_volcano_neg_pos$Diagnosis)-1) &
                   sum(Data_drug_volcano_neg_pos$mutation)>1 &
                   sum(Data_drug_volcano_neg_pos$mutation)<(length(Data_drug_volcano_neg_pos$mutation)-1) &
                   sum(Data_drug_volcano_neg_pos$TTF_censor)>1 &
                   sum(Data_drug_volcano_neg_pos$TTF_censor)<(length(Data_drug_volcano_neg_pos$TTF_censor)-1)){
                  Data_ORR = coxph(Surv(Drug_length, TTF_censor)~mutation+Diagnosis,
                                   data=Data_drug_volcano_neg_pos)
                  data_tmp = tidy(Data_ORR, exponentiate=TRUE, conf.int=TRUE)[1,]
                  df_volcano_tmp$HazardRatio_considering_pathology = data_tmp$estimate
                  df_volcano_tmp$p_value = data_tmp$p.value
                } else {
                  major_pathology = names(table(Data_drug_volcano_neg_pos$Diagnosis)[table(Data_drug_volcano_neg_pos$Diagnosis) > 1])
                  Data_drug_volcano_neg_pos_major = Data_drug_volcano_neg_pos %>%
                    dplyr::filter(Diagnosis %in% major_pathology)
                  Data_drug_volcano_neg_pos_major$Diagnosis = factor(Data_drug_volcano_neg_pos_major$Diagnosis)
                  Data_drug_volcano_neg_pos_major$Diagnosis = relevel(Data_drug_volcano_neg_pos_major$Diagnosis, ref=names(sort(table(Data_drug_volcano_neg_pos$Diagnosis),decreasing = T))[[1]])
                  if(input$volcano_pathology == "Yes" &
                     min(table(Data_drug_volcano_neg_pos_major$Diagnosis))>1 &
                     max(table(Data_drug_volcano_neg_pos_major$Diagnosis))<(length(Data_drug_volcano_neg_pos_major$Diagnosis)-1) &
                     sum(Data_drug_volcano_neg_pos_major$mutation)>1 &
                     sum(Data_drug_volcano_neg_pos_major$mutation)<(length(Data_drug_volcano_neg_pos_major$mutation)-1) &
                     sum(Data_drug_volcano_neg_pos_major$TTF_censor)>1 &
                     sum(Data_drug_volcano_neg_pos_major$TTF_censor)<(length(Data_drug_volcano_neg_pos_major$TTF_censor)-1)){
                    Data_ORR = coxph(Surv(Drug_length, TTF_censor)~mutation+Diagnosis,
                                     data=Data_drug_volcano_neg_pos)
                    data_tmp = tidy(Data_ORR, exponentiate=TRUE, conf.int=TRUE)[1,]
                    df_volcano_tmp$HazardRatio_considering_pathology = data_tmp$estimate
                    df_volcano_tmp$p_value = data_tmp$p.value
                  } else if(sum(Data_drug_volcano_neg_pos$mutation)>1 &
                            sum(Data_drug_volcano_neg_pos$mutation)<(length(Data_drug_volcano_neg_pos$mutation)-1) &
                            sum(Data_drug_volcano_neg_pos$TTF_censor)>1 &
                            sum(Data_drug_volcano_neg_pos$TTF_censor)<(length(Data_drug_volcano_neg_pos$TTF_censor)-1)){
                    Data_ORR = coxph(Surv(Drug_length, TTF_censor)~mutation,
                                     data=Data_drug_volcano_neg_pos)
                    data_tmp = tidy(Data_ORR, exponentiate=TRUE, conf.int=TRUE)[1,]
                    df_volcano_tmp$HazardRatio_considering_pathology = data_tmp$estimate
                    df_volcano_tmp$p_value = data_tmp$p.value
                  } else {
                    df_volcano_tmp$HazardRatio_considering_pathology = 1
                    df_volcano_tmp$p_value = 1
                  }
                }
                df_volcano_ToT = rbind(df_volcano_ToT, df_volcano_tmp)
              }
              Figure_volcano = df_volcano_ToT  %>%
                dplyr::filter(Regimen == paste(input$drug, collapse = ";")) %>%
                dplyr::mutate(Gene_effect = ifelse(p_value >= 0.05 , "Not significant",
                                                   ifelse(HazardRatio_considering_pathology > 1, "Poor",
                                                          ifelse(HazardRatio_considering_pathology < 1, "Good","Not significant"))))
              Figure_volcano$HazardRatio_considering_pathology[Figure_volcano$HazardRatio_considering_pathology>16]=16
              Figure_volcano$HazardRatio_considering_pathology[Figure_volcano$HazardRatio_considering_pathology<1/16]=1/16
              Figure_volcano$Gene_effect = factor(Figure_volcano$Gene_effect, levels = c("Good", "Not significant", "Poor"))
              Figure_volcano$label = ifelse((Figure_volcano$p_value < 0.05), Figure_volcano$Gene, NA_character_)
              Figure_volcano$log2OR = log2(Figure_volcano$HazardRatio_considering_pathology)
              Figure_volcano$log10pval = -log10(Figure_volcano$p_value)
              if(length(Figure_volcano$label>0)){
                g_volcano_ToT[[figure_no]] =
                  ggplot(data = Figure_volcano, aes(x = log2OR, y = log10pval, col = Gene_effect, label = label)) +
                  geom_point(size = 2) +
                  labs(x = expression("log"[2]*"hazard ratio of time on treatment"), y = expression("-log"[10]*"p-value")) +
                  scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"), 
                                     labels = c("Good", "Not significant", "Poor"),
                                     drop = FALSE) +
                  geom_hline(yintercept = -log10(0.05), size = 0.2, color = "dark green") + 
                  geom_vline(xintercept = -1, size = 0.2, color = "dark green", linetype = 'dashed') +
                  geom_vline(xintercept = 1, size = 0.2, color = "dark green", linetype = 'dashed') +
                  theme(legend.position = "none", panel.grid = element_blank()) +
                  ggtitle(paste0(paste(input$drug, collapse = ";"), " ",
                                 Figure_volcano$Total_treated_patients[[1]], " patients")) +
                  geom_text_repel(max.overlaps = Inf) +
                  theme_classic()
                figure_no = figure_no + 1
              }
              significant_genes = (Figure_volcano %>% dplyr::arrange(p_value))$label
              significant_genes = significant_genes[!is.na(significant_genes)]
            }
          }
        }
        
        if(figure_no<=200){
          for(figure_no_tmp in figure_no:200){
            g_volcano_ToT[[figure_no_tmp]] = gg_empty()
          }
        }
        
        output$figure_volcano_ToT_1 = renderPlot({
          grid.arrange(grobs = g_volcano_ToT[1:50], nrow = 10, ncol = 5)
        })
        output$figure_volcano_ToT_2 = renderPlot({
          grid.arrange(grobs = g_volcano_ToT[51:100], nrow = 10, ncol = 5)
        })
        output$figure_volcano_ToT_3 = renderPlot({
          grid.arrange(grobs = g_volcano_ToT[101:150], nrow = 10, ncol = 5)
        })
        output$figure_volcano_ToT_4 = renderPlot({
          grid.arrange(grobs = g_volcano_ToT[151:200], nrow = 10, ncol = 5)
        })
        
        
        output$table_volcano_ToT = DT::renderDataTable(df_volcano_ToT,
                                                   server = FALSE,
                                                   filter = 'top', 
                                                   extensions = c('Buttons'), 
                                                   options = list(pageLength = 100, 
                                                                  scrollX = TRUE,
                                                                  scrollY = "1000px",
                                                                  scrollCollapse = TRUE,
                                                                  dom="Blfrtip",
                                                                  buttons = c('csv', 'excel', 'copy')))
        
                
        has = list()
        if(length(unique(Data_drug_TTF$diagnosis)) >= 1){      
          Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::filter(Drug %in% input$drug)
          oncogenic_genes = Data_drug_TTF_tmp %>%
            dplyr::select(ID, diagnosis) %>%
            dplyr::distinct() %>%
            dplyr::count(diagnosis) %>%
            dplyr::arrange(-n)
          colnames(oncogenic_genes) = c("gene_mutation", "all_patients")
          oncogenic_genes = oncogenic_genes[1:min(length(oncogenic_genes$all_patients), 10),]
          
          gene_table = data.frame(oncogenic_genes$gene_mutation)
          colnames(gene_table) = c("Gene")
          gene_table$positive_patients = oncogenic_genes$all_patients
          gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(unique(Data_drug_TTF_tmp$ID))) / 10
          gene_table$positive_median = 0
          gene_table$positive_LL = 0
          gene_table$positive_UL = 0
          gene_table$negative_median = 0
          gene_table$negative_LL = 0
          gene_table$negative_UL = 0
          gene_table$diff_median = 0
          gene_table$diff_LL = 0
          gene_table$diff_UL = 0
          k = 1
          for(i in 1:length(oncogenic_genes$all_patients)){
            ID_mutation = (Data_drug_TTF %>% dplyr::filter(diagnosis %in% oncogenic_genes$gene_mutation[i]))$ID
            Data_drug_TTF_tmp = Data_drug_TTF_tmp %>% dplyr::mutate(
              gene_mut = case_when(
                ID %in% ID_mutation ~ 1,
                TRUE ~ 0
              )
            )
            traditional_fit = survfit(Surv(event = TTF_censor,
                                           time = Drug_length) ~ gene_mut, 
                                      data = Data_drug_TTF_tmp)
            if(length(Data_drug_TTF_tmp[,1]) != traditional_fit[[1]][[1]]){
              tau0 = ((max((Data_drug_TTF_tmp %>% dplyr::filter(gene_mut == 0))$Drug_length) - 1)/365.25*12)
              tau1 = ((max((Data_drug_TTF_tmp %>% dplyr::filter(gene_mut == 1))$Drug_length) - 1)/365.25*12)
              tau = floor(min(tau0, tau1, input$RMST_drug) * 10) / 10
              verify <- survRM2::rmst2(
                time = Data_drug_TTF_tmp$Drug_length,
                status = Data_drug_TTF_tmp$TTF_censor,
                arm = Data_drug_TTF_tmp$gene_mut,
                tau = tau * 365.25 / 12,
                alpha = 0.05
              )
              
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~gene_mut,
                                data=Data_drug_TTF_tmp, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~gene_mut,
                                data=Data_drug_TTF_tmp, rho=1)
              mut_gene = oncogenic_genes$gene_mutation[i]
              tmp = data.frame(summary(traditional_fit)$table)
              gene_table$positive_median[i] = round(tmp$median[2] / 365.25 * 12, digits = 1)
              gene_table$positive_LL[i] = round(tmp$X0.95LCL[2] / 365.25 * 12, digits = 1)
              gene_table$positive_UL[i] = round(tmp$X0.95UCL[2] / 365.25 * 12, digits = 1)
              gene_table$negative_median[i] = round(tmp$median[1] / 365.25 * 12, digits = 1)
              gene_table$negative_LL[i] = round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1)
              gene_table$negative_UL[i] = round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1)
              gene_table$diff_median[i] = round(verify$unadjusted.result[1] / 365.25 * 12, digits = 1)
              gene_table$diff_UL[i] = round(verify$unadjusted.result[7] / 365.25 * 12, digits = 1)
              gene_table$diff_LL[i] = round(verify$unadjusted.result[4] / 365.25 * 12, digits = 1)
            } else{
              tmp = summary(traditional_fit)$table
              legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (",
                               round(tmp[[8]] / 365.25 * 12, digits = 1), "-",
                               round(tmp[[9]] / 365.25 * 12, digits = 1),")")
              gene_table$negative_median[i] = round(tmp[[7]] / 365.25 * 12, digits = 1)
              gene_table$negative_LL[i] = round(tmp[[8]] / 365.25 * 12, digits = 1)
              gene_table$negative_UL[i] = round(tmp[[9]] / 365.25 * 12, digits = 1)
            }
          }
          Gene_arrange = gene_table$Gene
          Gene_data = gene_table
          gene_table = gene_table %>% dplyr::mutate(
            Gene = factor(gene_table$Gene, levels = Gene_arrange))
          
          p <- 
            gene_table |>
            ggplot(aes(y = fct_rev(Gene))) + 
            theme_classic()
          p <- p +
            geom_point(aes(x=diff_median), shape=15, size=3) +
            geom_linerange(aes(xmin=diff_LL, xmax=diff_UL))
          p <- p +
            geom_vline(xintercept = 0, linetype="dashed") +
            labs(x=paste0("Median time on treatment and difference in restricted mean survival time in ", input$RMST_drug, " months with ", paste(input$drug, collapse = ";"), " (months)"), y="Diagnosis")
          p <- p +
            coord_cartesian(ylim=c(1,length(Gene_arrange) + 1),
                            xlim=c( - max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) - 0.5,
                                    max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) + 0.5))
          p <- p +
            annotate("text",
                     x = (-max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) - 0.5)/2,
                     y = length(Gene_arrange) + 1,
                     label = "diagnosis=short survival") +
            annotate("text",
                     x = (max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) + 0.5)/2,
                     y = length(Gene_arrange) + 1,
                     label = "diagnosis=long survival")
          p_mid <- p + 
            theme(legend.position = "none",
                  axis.line.y = element_blank(),
                  axis.ticks.y= element_blank(),
                  axis.text.y= element_blank(),
                  axis.title.y= element_blank())
          gene_table <- gene_table %>%
            dplyr::mutate(
              estimate_lab_1 = paste0(positive_median, " (", positive_LL, "-", positive_UL, ")")) %>%
            dplyr::mutate(
              estimate_lab_2 = paste0(negative_median, " (", negative_LL, "-", negative_UL, ")")) %>%
            dplyr::mutate(
              estimate_lab_3 = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
            dplyr::mutate(
              patients = paste0(positive_patients, " (", positive_freq, "%)"))
          gene_table = 
            dplyr::bind_rows(
              data.frame(
                Gene = "Dignosis",
                estimate_lab_1 = "ToT, diagnosis",
                estimate_lab_2 = "ToT, others",
                estimate_lab_3 = "RMST difference",
                patients = "Patients"
              ), gene_table)
          Gene_arrange = gene_table$Gene
          gene_table = gene_table %>% dplyr::mutate(
            Gene = factor(gene_table$Gene, levels = Gene_arrange))
          
          p_left <- gene_table  %>% ggplot(aes(y = fct_rev(Gene)))
          p_left <- p_left + geom_text(aes(x = 0, label = Gene), hjust = 0,
                                       fontface = ifelse(gene_table$Gene == "Dignosis", "bold", "bold.italic"))
          p_left <- p_left + geom_text(aes(x = 1.8, label = estimate_lab_1), hjust = 0,
                                       fontface = ifelse(gene_table$estimate_lab_1 == "ToT, diagnosis", "bold", "plain"))
          p_left <- p_left + geom_text(aes(x = 3.2, label = estimate_lab_2), hjust = 0,
                                       fontface = ifelse(gene_table$estimate_lab_2 == "ToT, others", "bold", "plain"))
          p_left <- p_left + geom_text(aes(x = 4.6, label = estimate_lab_3), hjust = 0,
                                       fontface = ifelse(gene_table$estimate_lab_3 == "RMST difference", "bold", "plain"))
          p_left <- p_left + theme_void() + coord_cartesian(xlim = c(0, 7.2))
          
          # right side of plot - pvalues
          p_right <- gene_table  |> ggplot() +
            geom_text(aes(x = 0, y = fct_rev(Gene), label = patients), hjust = 0,
                      fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")) +
            theme_void()
          
          # final plot arrangement
          layout <- c(patchwork::area(t = 0, l = 0, b = 30, r = 7.2),
                      patchwork::area(t = 1, l = 7.2, b = 30, r = 12.5),
                      patchwork::area(t = 0, l = 12.5, b = 30, r = 14))
          has[[1]] = p_left + p_mid + p_right + plot_layout(design = layout)
          gene_table_3 = gene_table
          output$figure_drug_3_forest = renderPlot({
            has[[1]]
          })
          output$figure_drug_3_forest_table = DT::renderDataTable(gene_table_3,
                                                           server = FALSE,
                                                           filter = 'top', 
                                                           extensions = c('Buttons'), 
                                                           options = list(pageLength = 100, 
                                                                          scrollX = TRUE,
                                                                          scrollY = "1000px",
                                                                          scrollCollapse = TRUE,
                                                                          dom="Blfrtip",
                                                                          buttons = c('csv', 'excel', 'copy')))
        }
        
        incProgress(1 / 13)
        
        # analysis for common oncogenic mutations
        Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::filter(Drug %in% input$drug)
        h = list()
        hs = list()
        hsc = list()
        
        oncogenic_genes = Data_MAF_target %>%
          dplyr::select(Tumor_Sample_Barcode, Hugo_Symbol) %>%
          dplyr::filter(Tumor_Sample_Barcode %in% Data_drug_TTF_tmp$ID) %>%
          dplyr::distinct() %>%
          dplyr::count(Hugo_Symbol) %>%
          dplyr::arrange(-n)
        colnames(oncogenic_genes) = c("gene_mutation", "all_patients")
        oncogenic_genes = rbind(oncogenic_genes %>% dplyr::filter(gene_mutation %in% input$gene),
                                oncogenic_genes %>% dplyr::filter(!gene_mutation %in% input$gene))
        oncogenic_genes = oncogenic_genes[1:min(length(oncogenic_genes$all_patients), input$gene_no),]
        
        gene_table = data.frame(oncogenic_genes$gene_mutation)
        colnames(gene_table) = c("Gene")
        gene_table$positive_patients = oncogenic_genes$all_patients
        gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(unique(Data_drug_TTF_tmp$ID))) / 10
        gene_table$positive_median = 0
        gene_table$positive_LL = 0
        gene_table$positive_UL = 0
        gene_table$negative_median = 0
        gene_table$negative_LL = 0
        gene_table$negative_UL = 0
        gene_table$diff_median = 0
        gene_table$diff_LL = 0
        gene_table$diff_UL = 0
        k = 1
        if(length(Data_drug_TTF_tmp$TTF_censor) > 0){
          for(i in 1:length(oncogenic_genes$all_patients)){
            ID_mutation = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% oncogenic_genes$gene_mutation[i]))$Tumor_Sample_Barcode
            Data_drug_TTF_tmp = Data_drug_TTF_tmp %>% dplyr::mutate(
              gene_mut = case_when(
                ID %in% ID_mutation ~ 1,
                TRUE ~ 0
              )
            )
            traditional_fit = survfit(Surv(event = TTF_censor,
                                           time = Drug_length) ~ gene_mut, 
                                      data = Data_drug_TTF_tmp)
            if(length(Data_drug_TTF_tmp[,1]) != traditional_fit[[1]][[1]]){
              tau0 = ((max((Data_drug_TTF_tmp %>% dplyr::filter(gene_mut == 0))$Drug_length) - 1)/365.25*12)
              tau1 = ((max((Data_drug_TTF_tmp %>% dplyr::filter(gene_mut == 1))$Drug_length) - 1)/365.25*12)
              tau = floor(min(tau0, tau1, input$RMST_drug) * 10) / 10
              verify <- survRM2::rmst2(
                time = Data_drug_TTF_tmp$Drug_length,
                status = Data_drug_TTF_tmp$TTF_censor,
                arm = Data_drug_TTF_tmp$gene_mut,
                tau = tau * 365.25 / 12,
                alpha = 0.05
              )
              
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~gene_mut,
                                data=Data_drug_TTF_tmp, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~gene_mut,
                                data=Data_drug_TTF_tmp, rho=1)
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~gene_mut,
                                data=Data_drug_TTF_tmp)
              mut_gene = paste(oncogenic_genes$gene_mutation[i],
                               ", top ", i, " gene", sep="")
              tmp = data.frame(summary(traditional_fit)$table)
              gene_table$positive_median[i] = round(tmp$median[2] / 365.25 * 12, digits = 1)
              gene_table$positive_LL[i] = round(tmp$X0.95LCL[2] / 365.25 * 12, digits = 1)
              gene_table$positive_UL[i] = round(tmp$X0.95UCL[2] / 365.25 * 12, digits = 1)
              gene_table$negative_median[i] = round(tmp$median[1] / 365.25 * 12, digits = 1)
              gene_table$negative_LL[i] = round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1)
              gene_table$negative_UL[i] = round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1)
              gene_table$diff_median[i] = round(verify$unadjusted.result[1] / 365.25 * 12, digits = 1)
              gene_table$diff_UL[i] = round(verify$unadjusted.result[7] / 365.25 * 12, digits = 1)
              gene_table$diff_LL[i] = round(verify$unadjusted.result[4] / 365.25 * 12, digits = 1)
              hs[[k]] = surv_curv_drug(traditional_fit, Data_drug_TTF_tmp, paste(oncogenic_genes$gene_mutation[i], "mut, treated with", paste(input$drug, collapse = ";")),
                                        diff_0, diff_1, diff_2)
              k = k + 1
            } else{
              tmp = summary(traditional_fit)$table
              legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (",
                               round(tmp[[8]] / 365.25 * 12, digits = 1), "-",
                               round(tmp[[9]] / 365.25 * 12, digits = 1),")")
              gene_table$negative_median[i] = round(tmp[[7]] / 365.25 * 12, digits = 1)
              gene_table$negative_LL[i] = round(tmp[[8]] / 365.25 * 12, digits = 1)
              gene_table$negative_UL[i] = round(tmp[[9]] / 365.25 * 12, digits = 1)
            }
          }
        }
        Gene_arrange = gene_table$Gene
        Gene_data = gene_table
        gene_table = gene_table %>% dplyr::mutate(
          Gene = factor(gene_table$Gene, levels = Gene_arrange))
  
        p <- 
          gene_table |>
          ggplot(aes(y = fct_rev(Gene))) + 
          theme_classic()
        p <- p +
          geom_point(aes(x=diff_median), shape=15, size=3) +
          geom_linerange(aes(xmin=diff_LL, xmax=diff_UL))
        p <- p +
          geom_vline(xintercept = 0, linetype="dashed") +
          labs(x=paste0("Median time on treatment and difference in restricted mean survival time in ", input$RMST_drug, " months with ", paste(input$drug, collapse = ";"), " (months)"), y="Genes with oncogenic alteration")
        p <- p +
          coord_cartesian(ylim=c(1,length(Gene_arrange) + 1),
                          xlim=c( - max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) - 0.5,
                                  max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) + 0.5))
        p <- p +
          annotate("text",
                   x = (-max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) - 0.5)/2,
                   y = length(Gene_arrange) + 1,
                   label = "mut=short survival") +
          annotate("text",
                   x = (max(abs(gene_table$diff_LL), abs(gene_table$diff_UL)) + 0.5)/2,
                   y = length(Gene_arrange) + 1,
                   label = "mut=long survival")
        p_mid <- p + 
          theme(legend.position = "none",
                axis.line.y = element_blank(),
                axis.ticks.y= element_blank(),
                axis.text.y= element_blank(),
                axis.title.y= element_blank())
        gene_table <- gene_table %>%
          dplyr::mutate(
            estimate_lab_1 = paste0(positive_median, " (", positive_LL, "-", positive_UL, ")")) %>%
          dplyr::mutate(
            estimate_lab_2 = paste0(negative_median, " (", negative_LL, "-", negative_UL, ")")) %>%
          dplyr::mutate(
            estimate_lab_3 = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
          dplyr::mutate(
            patients = paste0(positive_patients, " (", positive_freq, "%)"))
        gene_table = 
          dplyr::bind_rows(
            data.frame(
              Gene = "Gene",
              estimate_lab_1 = "ToT, mut (+)",
              estimate_lab_2 = "ToT, mut (-)",
              estimate_lab_3 = "RMST difference",
              patients = "Patients"
            ), gene_table)
        Gene_arrange = gene_table$Gene
        gene_table = gene_table %>% dplyr::mutate(
          Gene = factor(gene_table$Gene, levels = Gene_arrange))
        
        p_left <- gene_table  %>% ggplot(aes(y = fct_rev(Gene)))
        p_left <- p_left + geom_text(aes(x = 0, label = Gene), hjust = 0,
                                     fontface = ifelse(gene_table$Gene == "Gene", "bold", "bold.italic"))
        p_left <- p_left + geom_text(aes(x = 2.3, label = estimate_lab_1), hjust = 0,
                                     fontface = ifelse(gene_table$estimate_lab_1 == "ToT, mut (+)", "bold", "plain"))
        p_left <- p_left + geom_text(aes(x = 4.0, label = estimate_lab_2), hjust = 0,
                                     fontface = ifelse(gene_table$estimate_lab_2 == "ToT, mut (-)", "bold", "plain"))
        p_left <- p_left + geom_text(aes(x = 5.0, label = estimate_lab_3), hjust = 0,
                                     fontface = ifelse(gene_table$estimate_lab_3 == "RMST difference", "bold", "plain"))
        p_left <- p_left + theme_void() + coord_cartesian(xlim = c(0, 7.5))
        
        # right side of plot - pvalues
        p_right <- gene_table  |> ggplot() +
          geom_text(aes(x = 0, y = fct_rev(Gene), label = patients), hjust = 0,
                    fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")) +
          theme_void()
        # final plot arrangement
        layout <- c(patchwork::area(t = 0, l = 0, b = 30, r = 7.5),
                    patchwork::area(t = 1, l = 7.5, b = 30, r = 12.5),
                    patchwork::area(t = 0, l = 12.5, b = 30, r = 14))
        h[[1]] = p_left + p_mid + p_right + plot_layout(design = layout)
        
        for(kk in k:32){
          hs[[kk]] = ggsurvplot_empty()
        }
        gene_table_gene = gene_table
        ####################
        # analysis for clusters
        Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::filter(Drug %in% input$drug)
        h_clust = list()
        oncogenic_genes = Data_case_target %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID, cluster) %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% Data_drug_TTF_tmp$ID) %>%
          dplyr::distinct() %>%
          dplyr::count(cluster) %>%
          dplyr::arrange(cluster)
        colnames(oncogenic_genes) = c("cluster", "all_patients")

        gene_table = data.frame(oncogenic_genes$cluster)
        colnames(gene_table) = c("Gene")
        gene_table$positive_patients = oncogenic_genes$all_patients
        gene_table$positive_freq = round(1000 * gene_table$positive_patients / length(unique(Data_drug_TTF_tmp$ID))) / 10
        gene_table$positive_median = 0
        gene_table$positive_LL = 0
        gene_table$positive_UL = 0
        gene_table$negative_median = 0
        gene_table$negative_LL = 0
        gene_table$negative_UL = 0
        gene_table$diff_median = 0
        gene_table$diff_LL = 0
        gene_table$diff_UL = 0
        k = 1
        if(length(Data_drug_TTF_tmp$TTF_censor) > 0){
          for(i in 1:length(oncogenic_genes$cluster)){
            ID_mutation = (Data_case_target %>% dplyr::filter(cluster == i))$C.CAT調査結果.基本項目.ハッシュID
            Data_drug_TTF_tmp = Data_drug_TTF_tmp %>% dplyr::mutate(
              gene_mut = case_when(
                ID %in% ID_mutation ~ 1,
                TRUE ~ 0
              )
            )
            traditional_fit = survfit(Surv(event = TTF_censor,
                                           time = Drug_length) ~ gene_mut, 
                                      data = Data_drug_TTF_tmp)
            if(length(Data_drug_TTF_tmp[,1]) != traditional_fit[[1]][[1]]){
              tau0 = ((max((Data_drug_TTF_tmp %>% dplyr::filter(gene_mut == 0))$Drug_length) - 1)/365.25*12)
              tau1 = ((max((Data_drug_TTF_tmp %>% dplyr::filter(gene_mut == 1))$Drug_length) - 1)/365.25*12)
              tau = floor(min(tau0, tau1, input$RMST_drug) * 10) / 10
              verify <- survRM2::rmst2(
                time = Data_drug_TTF_tmp$Drug_length,
                status = Data_drug_TTF_tmp$TTF_censor,
                arm = Data_drug_TTF_tmp$gene_mut,
                tau = tau * 365.25 / 12,
                alpha = 0.05
              )
              
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~gene_mut,
                                data=Data_drug_TTF_tmp, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~gene_mut,
                                data=Data_drug_TTF_tmp, rho=1)
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~gene_mut,
                             data=Data_drug_TTF_tmp)
              
              mut_gene = paste(oncogenic_genes$cluster[i],
                               ", top ", i, " gene", sep="")
              tmp = data.frame(summary(traditional_fit)$table)
              gene_table$positive_median[i] = round(tmp$median[2] / 365.25 * 12, digits = 1)
              gene_table$positive_LL[i] = round(tmp$X0.95LCL[2] / 365.25 * 12, digits = 1)
              gene_table$positive_UL[i] = round(tmp$X0.95UCL[2] / 365.25 * 12, digits = 1)
              gene_table$negative_median[i] = round(tmp$median[1] / 365.25 * 12, digits = 1)
              gene_table$negative_LL[i] = round(tmp$X0.95LCL[1] / 365.25 * 12, digits = 1)
              gene_table$negative_UL[i] = round(tmp$X0.95UCL[1] / 365.25 * 12, digits = 1)
              data_tmp = tidy(diff_2, exponentiate=TRUE, conf.int=TRUE)
              gene_table$diff_median[i] = round(1/data_tmp[[2]], digits = 3)
              gene_table$diff_LL[i] = round(1/data_tmp[[7]], digits = 3)
              gene_table$diff_UL[i] = round(1/data_tmp[[6]], digits = 3)
              # gene_table$diff_median[i] = round(verify$unadjusted.result[1] / 365.25 * 12, digits = 1)
              # gene_table$diff_UL[i] = round(verify$unadjusted.result[7] / 365.25 * 12, digits = 1)
              # gene_table$diff_LL[i] = round(verify$unadjusted.result[4] / 365.25 * 12, digits = 1)
              hsc[[k]] = surv_curv_drug(traditional_fit, Data_drug_TTF_tmp, paste0("Cluster ", i, ", treated with", paste(input$drug, collapse = ";")),
                                            diff_0, diff_1, diff_2)
              k = k + 1
            } else{
              tmp = summary(traditional_fit)$table
              legends = paste0(round(tmp[[7]] / 365.25 * 12, digits = 1), " (",
                               round(tmp[[8]] / 365.25 * 12, digits = 1), "-",
                               round(tmp[[9]] / 365.25 * 12, digits = 1),")")
              gene_table$negative_median[i] = round(tmp[[7]] / 365.25 * 12, digits = 1)
              gene_table$negative_LL[i] = round(tmp[[8]] / 365.25 * 12, digits = 1)
              gene_table$negative_UL[i] = round(tmp[[9]] / 365.25 * 12, digits = 1)
            }
          }
        }
        Gene_arrange = gene_table$Gene
        Gene_data = gene_table
        # gene_table = gene_table %>% dplyr::mutate(
        #   Gene = factor(gene_table$Gene, levels = Gene_arrange))
        gene_table = gene_table %>% dplyr::mutate(
          Gene = factor(gene_table$Gene))
        
        p <- 
          gene_table |>
          ggplot(aes(y = fct_rev(Gene))) + 
          theme_classic() +  scale_x_log10()
        p <- p +
          geom_point(aes(x=diff_median), shape=15, size=3) +
          geom_linerange(aes(xmin=diff_LL, xmax=diff_UL))
        p <- p +
          geom_vline(xintercept = 1, linetype="dashed") +
          labs(x=paste0("Hazard ratio with ", paste(input$drug, collapse = ";")),
               y="Mutation based cluster")
        p <- p +
          coord_cartesian(ylim=c(1,length(Gene_arrange) + 1),
                          xlim=c( min(gene_table$diff_LL) ^ 1.1,
                                  max(gene_table$diff_UL) ^ 1.1))
        p <- p +
          annotate("text",
                   x = sqrt(max(gene_table$diff_UL) ^ 1.1),
                   y = length(Gene_arrange) + 1,
                   label = "cluster=short survival") +
          annotate("text",
                   x = sqrt(min(gene_table$diff_LL) ^ 1.1),
                   y = length(Gene_arrange) + 1,
                   label = "cluster=long survival")
        p_mid <- p + 
          theme(legend.position = "none",
                axis.line.y = element_blank(),
                axis.ticks.y= element_blank(),
                axis.text.y= element_blank(),
                axis.title.y= element_blank())
        gene_table <- gene_table %>%
          dplyr::mutate(
            estimate_lab_1 = paste0(positive_median, " (", positive_LL, "-", positive_UL, ")")) %>%
          dplyr::mutate(
            estimate_lab_2 = paste0(negative_median, " (", negative_LL, "-", negative_UL, ")")) %>%
          dplyr::mutate(
            estimate_lab_3 = paste0(diff_median, " (", diff_LL, "-", diff_UL, ")")) %>%
          dplyr::mutate(
            patients = paste0(positive_patients, " (", positive_freq, "%)"))
        gene_table = 
          dplyr::bind_rows(
            data.frame(
              Gene = "Cluster",
              estimate_lab_1 = "Survival, mut (+)",
              estimate_lab_2 = "Survival, mut (-)",
              estimate_lab_3 = "Hazard ratio",
              patients = "Patients"
            ), gene_table)
        
        
        Gene_arrange = gene_table$Gene
        gene_table = gene_table %>% dplyr::mutate(
          Gene = factor(gene_table$Gene, levels = Gene_arrange))
        
        p_left <- gene_table  %>% ggplot(aes(y = fct_rev(Gene)))
        p_left <- p_left + geom_text(aes(x = 0, label = Gene), hjust = 0,
                                     fontface = ifelse(gene_table$Gene == "Cluster", "bold", "bold.italic"))
        p_left <- p_left + geom_text(aes(x = 1.8, label = estimate_lab_1), hjust = 0,
                                     fontface = ifelse(gene_table$estimate_lab_1 == "Survival, mut (+)", "bold", "plain"))
        p_left <- p_left + geom_text(aes(x = 3.3, label = estimate_lab_2), hjust = 0,
                                     fontface = ifelse(gene_table$estimate_lab_2 == "Survival, mut (-)", "bold", "plain"))
        p_left <- p_left + geom_text(aes(x = 4.8, label = estimate_lab_3), hjust = 0,
                                     fontface = ifelse(gene_table$estimate_lab_3 == "Hazard ratio", "bold", "plain"))
        p_left <- p_left + theme_void() + coord_cartesian(xlim = c(0, 7.5))
        
        
        # right side of plot - pvalues
        p_right <- gene_table  |> ggplot() +
          geom_text(aes(x = 0, y = fct_rev(Gene), label = patients), hjust = 0,
                    fontface = ifelse(gene_table$patients == "Patients", "bold", "plain")) +
          theme_void()
        
        
        # final plot arrangement
        layout <- c(patchwork::area(t = 0, l = 0, b = 30, r = 7.5),
                    patchwork::area(t = 1, l = 7.5, b = 30, r = 12.5),
                    patchwork::area(t = 0, l = 13.5, b = 30, r = 14))
        h_clust[[1]] = p_left + p_mid + p_right + plot_layout(design = layout)
        for(kk in k:32){
          hsc[[kk]] = ggsurvplot_empty()
        }
        ########################
        
        output$figure_drug_cluster = renderPlot({
          arrange_ggsurvplots(hsc,print=TRUE,ncol=4,nrow=8,surv.plot.height = 0.8,risk.table.height = 0.2)
        })
        output$figure_drug_cluster_2 = renderPlot({
          h_clust[[1]]
        })
        gene_table_cluster = gene_table
        output$figure_drug_cluster_table = DT::renderDataTable(gene_table_cluster,
                                                         server = FALSE,
                                                         filter = 'top', 
                                                         extensions = c('Buttons'), 
                                                         options = list(pageLength = 100, 
                                                                        scrollX = TRUE,
                                                                        scrollY = "1000px",
                                                                        scrollCollapse = TRUE,
                                                                        dom="Blfrtip",
                                                                        buttons = c('csv', 'excel', 'copy')))
        output$figure_drug_4 = renderPlot({
          h[[1]]
        })
        output$figure_drug_4_table = DT::renderDataTable(gene_table_gene,
                                                                server = FALSE,
                                                                filter = 'top', 
                                                                extensions = c('Buttons'), 
                                                                options = list(pageLength = 100, 
                                                                               scrollX = TRUE,
                                                                               scrollY = "1000px",
                                                                               scrollCollapse = TRUE,
                                                                               dom="Blfrtip",
                                                                               buttons = c('csv', 'excel', 'copy')))
        output$figure_drug_5 = renderPlot({
          arrange_ggsurvplots(hs,print=TRUE,ncol=4,nrow=8,surv.plot.height = 0.8,risk.table.height = 0.2)
        })
        incProgress(1 / 13)
        
        output$figure_drug_pattern = renderPlot({
          hsa = list()
          mut_gene_ = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
          if(length(mut_gene_)==0){
            mut_gene_ = "TP53"
            ID_mutation_ = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene_))$Tumor_Sample_Barcode
            Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::mutate(
              mutation = case_when(
                ID %in% ID_mutation_ ~ "TP53 mut(+)",
                TRUE ~ "TP53 mut(-)"
              )
            )
          } else{
            if(is.null(input$gene_group_2)){
              ID_mutation_ = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene_))$Tumor_Sample_Barcode
              Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::mutate(
                mutation = case_when(
                  ID %in% ID_mutation_ ~ paste0(paste(mut_gene_, collapse = "/"), " mut(+)"),
                  TRUE ~ paste0(paste(mut_gene_, collapse = "/"), " mut(-)")
                )
              )
            } else{
              ID_mutation_1 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_1))$Tumor_Sample_Barcode
              ID_mutation_2 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_2))$Tumor_Sample_Barcode
              Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::mutate(
                mutation = case_when(
                  ID %in% ID_mutation_1 & ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+) & ", paste(input$gene_group_2, collapse = "/"), " mut(+)"),
                  ID %in% ID_mutation_1 & !ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+) & ", paste(input$gene_group_2, collapse = "/"), " mut(-)"),
                  !ID %in% ID_mutation_1 & ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-) & ", paste(input$gene_group_2, collapse = "/"), " mut(+)"),
                  !ID %in% ID_mutation_1 & !ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-) & ", paste(input$gene_group_2, collapse = "/"), " mut(-)"),
                )
              )
            }
          }
          if(!input$special_gene == ""){
            ID_special_gene = (Data_MAF_target %>%
                                 dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_"))))$Tumor_Sample_Barcode
            ID_special_gene_mutation_1 = (Data_MAF_target %>%
                                            dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                            amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
            ID_special_gene_mutation_2 = (Data_MAF_target %>%
                                            dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                            amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
            Data_drug_TTF_tmp = Data_drug_TTF_tmp %>% dplyr::mutate(
              separation_value = case_when(
                ID %in% ID_special_gene_mutation_1 ~ paste0(input$special_gene_mutation_1_name, " in ", input$special_gene),
                ID %in% ID_special_gene_mutation_2 ~ paste0(input$special_gene_mutation_2_name, " in ", input$special_gene),
                ID %in% ID_special_gene ~ paste0("Other mut in ", input$special_gene),
                TRUE ~ paste0("No mut in ", input$special_gene)
              )
            )
          } else{
            Data_drug_TTF_tmp = Data_drug_TTF_tmp %>% dplyr::mutate(
              separation_value = "No mutation pattern")
          }
          kkk = 1
          if(length(unique(Data_drug_TTF_tmp$mutation)) > 1){
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~mutation, data=Data_drug_TTF_tmp,type = "kaplan-meier", conf.type = "log-log")
            diff_0 = survdiff(Surv(Drug_length, TTF_censor)~mutation,
                              data=Data_drug_TTF_tmp, rho=0)
            diff_1 = survdiff(Surv(Drug_length, TTF_censor)~mutation,
                              data=Data_drug_TTF_tmp, rho=1)
            diff_2 = coxph(Surv(Drug_length, TTF_censor)~mutation,
                              data=Data_drug_TTF_tmp)
            hsa[[kkk]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp, paste0("Time on treatment, mutation, all drugs"), diff_0, diff_1, diff_2)
            kkk = kkk + 1
          }
          if(length(unique(Data_drug_TTF_tmp$separation_value)) > 1){
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~separation_value, data=Data_drug_TTF_tmp,type = "kaplan-meier", conf.type = "log-log")
            diff_0 = survdiff(Surv(Drug_length, TTF_censor)~separation_value,
                              data=Data_drug_TTF_tmp, rho=0)
            diff_1 = survdiff(Surv(Drug_length, TTF_censor)~separation_value,
                              data=Data_drug_TTF_tmp, rho=1)
            diff_2 = coxph(Surv(Drug_length, TTF_censor)~separation_value,
                              data=Data_drug_TTF_tmp)
            hsa[[kkk]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp, paste0("Time on treatment, mutation pattern, all drugs"), diff_0, diff_1, diff_2)
            kkk = kkk + 1
          }
          Data_drug_TTF_tmp = Data_drug_TTF_tmp %>% dplyr::filter(Drug %in% input$drug)
          if(length(unique(Data_drug_TTF_tmp$mutation)) > 1){
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~mutation, data=Data_drug_TTF_tmp,type = "kaplan-meier", conf.type = "log-log")
            diff_0 = survdiff(Surv(Drug_length, TTF_censor)~mutation,
                              data=Data_drug_TTF_tmp, rho=0)
            diff_1 = survdiff(Surv(Drug_length, TTF_censor)~mutation,
                              data=Data_drug_TTF_tmp, rho=1)
            diff_2 = coxph(Surv(Drug_length, TTF_censor)~mutation,
                              data=Data_drug_TTF_tmp)
            hsa[[kkk]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp, paste0("Time on treatment, mutation, treated with ", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
            kkk = kkk + 1
          }
          if(length(unique(Data_drug_TTF_tmp$separation_value)) > 1){
            survfit_t <- survfit(Surv(Drug_length, TTF_censor)~separation_value, data=Data_drug_TTF_tmp,type = "kaplan-meier", conf.type = "log-log")
            diff_0 = survdiff(Surv(Drug_length, TTF_censor)~separation_value,
                              data=Data_drug_TTF_tmp, rho=0)
            diff_1 = survdiff(Surv(Drug_length, TTF_censor)~separation_value,
                              data=Data_drug_TTF_tmp, rho=1)
            diff_2 = coxph(Surv(Drug_length, TTF_censor)~separation_value,
                              data=Data_drug_TTF_tmp)
            hsa[[kkk]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp, paste0("Time on treatment, mutation pattern, treated with ", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
            kkk = kkk + 1
          }
          if(input$HER2 != "No"){
            ID_mutation_P = (Data_case_target %>% dplyr::filter(HER2_IHC == "Positive"))$C.CAT調査結果.基本項目.ハッシュID
            ID_mutation_N = (Data_case_target %>% dplyr::filter(HER2_IHC == "Negative"))$C.CAT調査結果.基本項目.ハッシュID
            Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::mutate(
              HER2 = case_when(
                ID %in% ID_mutation_P ~ "HER2 IHC (+)",
                ID %in% ID_mutation_N ~ "HER2 IHC (-)",
                TRUE ~ "Unknown"
              )
            )
            Data_drug_TTF_tmp_ = Data_drug_TTF_tmp %>%
              dplyr::filter(HER2 != "Unknown")
              if(length(unique(Data_drug_TTF_tmp_$HER2)) > 1){
              survfit_t <- survfit(Surv(Drug_length, TTF_censor)~HER2, data=Data_drug_TTF_tmp_,type = "kaplan-meier", conf.type = "log-log")
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~HER2,
                                data=Data_drug_TTF_tmp_, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~HER2,
                                data=Data_drug_TTF_tmp_, rho=1)
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~HER2,
                                data=Data_drug_TTF_tmp_)
              hsa[[kkk]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp_, paste0("Time on treatment, mutation pattern, treated with ", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
              kkk = kkk + 1
            }
          }
          if(input$MSI != "No"){
            ID_mutation_P = (Data_case_target %>% dplyr::filter(MSI_PCR == "dMMR"))$C.CAT調査結果.基本項目.ハッシュID
            ID_mutation_N = (Data_case_target %>% dplyr::filter(MSI_PCR == "pMMR"))$C.CAT調査結果.基本項目.ハッシュID
            Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::mutate(
              MSI = case_when(
                ID %in% ID_mutation_P ~ "MSI PCR dMMR",
                ID %in% ID_mutation_N ~ "MSI PCR pMMR",
                TRUE ~ "Unknown"
              )
            )
            Data_drug_TTF_tmp_ = Data_drug_TTF_tmp %>%
              dplyr::filter(MSI != "Unknown")
            if(length(unique(Data_drug_TTF_tmp_$MSI)) > 1){
              survfit_t <- survfit(Surv(Drug_length, TTF_censor)~HER2, data=Data_drug_TTF_tmp_,type = "kaplan-meier", conf.type = "log-log")
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~MSI,
                                data=Data_drug_TTF_tmp_, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~MSI,
                                data=Data_drug_TTF_tmp_, rho=1)
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~MSI,
                                data=Data_drug_TTF_tmp_)
              hsa[[kkk]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp_, paste0("Time on treatment, mutation pattern, treated with ", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
              kkk = kkk + 1
            }
          }
          if(input$MMR != "No"){
            ID_mutation_P = (Data_case_target %>% dplyr::filter(MMR_IHC == "Positive"))$C.CAT調査結果.基本項目.ハッシュID
            ID_mutation_N = (Data_case_target %>% dplyr::filter(MMR_IHC == "Negative"))$C.CAT調査結果.基本項目.ハッシュID
            Data_drug_TTF_tmp = Data_drug_TTF %>% dplyr::mutate(
              MMR = case_when(
                ID %in% ID_mutation_P ~ "MMR IHC (+)",
                ID %in% ID_mutation_N ~ "MMR IHC (-)",
                TRUE ~ "Unknown"
              )
            )
            Data_drug_TTF_tmp_ = Data_drug_TTF_tmp %>%
              dplyr::filter(MMR != "Unknown")
            if(length(unique(Data_drug_TTF_tmp_$MMR)) > 1){
              survfit_t <- survfit(Surv(Drug_length, TTF_censor)~HER2, data=Data_drug_TTF_tmp_,type = "kaplan-meier", conf.type = "log-log")
              diff_0 = survdiff(Surv(Drug_length, TTF_censor)~MMR,
                                data=Data_drug_TTF_tmp_, rho=0)
              diff_1 = survdiff(Surv(Drug_length, TTF_censor)~MMR,
                                data=Data_drug_TTF_tmp_, rho=1)
              diff_2 = coxph(Surv(Drug_length, TTF_censor)~MMR,
                                data=Data_drug_TTF_tmp_)
              hsa[[kkk]] = surv_curv_drug(survfit_t, Data_drug_TTF_tmp_, paste0("Time on treatment, mutation pattern, treated with ", paste(input$drug, collapse = ";")), diff_0, diff_1, diff_2)
              kkk = kkk + 1
            }
          }
          if(kkk<=18){
            for(kkkk in kkk:18){
              hsa[[kkkk]] = ggsurvplot_empty()
            }
          }
          arrange_ggsurvplots(hsa,print=TRUE,ncol=min(3, length(hsa)), nrow=ceiling(length(hsa)/3),surv.plot.height = 0.8,risk.table.height = 0.2)
        })
        incProgress(1 / 13)
      }
      
      Data_drug_volcano_all = Data_drug_RECIST %>%
        dplyr::filter(治療ライン %in% input$target_line) %>%
        dplyr::filter(!最良総合効果 %in% c("NE", "Unknown")) %>%
        dplyr::mutate(最良総合効果 = case_when(
          最良総合効果 %in% c("CR", "PR") ~ "1",
          TRUE ~ "0"
        )) %>%
        dplyr::arrange(ID, 治療ライン) %>%
        dplyr::select(ID, Drug, 最良総合効果) %>%
        dplyr::distinct(ID, Drug, .keep_all = T)
      Data_drug_volcano_all$最良総合効果 = as.integer(Data_drug_volcano_all$最良総合効果)
      drugs_volcano = sort(table(Data_drug_volcano_all$Drug),decreasing = T)
      drugs_volcano = drugs_volcano[!names(drugs_volcano) %in% c("Others", "Unknown", "")]
      drugs_volcano = drugs_volcano[!is.na(names(drugs_volcano))]
      
      df_volcano = data.frame(Regimen = as.character(),
                              Gene = as.character(),
                              Total_treated_patients = as.numeric(),
                              Mut_respond_patients = as.numeric(),
                              WT_respond_patients = as.numeric(),
                              Mut_nonrespond_patients = as.numeric(),
                              WT_nonrespond_patients = as.numeric(),
                              OddsRatio_considering_pathology = as.numeric(),
                              p_value_logistic_regression = as.numeric()
                              )
      df_volcano_tmp = data.frame(Regimen = "",
                                  Gene = "",
                                  Total_treated_patients = 0,
                                  Mut_respond_patients = 0,
                                  WT_respond_patients = 0,
                                  Mut_nonrespond_patients = 0,
                                  WT_nonrespond_patients = 0,
                                  OddsRatio_considering_pathology = 1,
                                  p_value_logistic_regression = 1
                                  )
      figure_no = 1
      g_volcano = list()
      options(warn = -1)
      withProgress(message = sample(nietzsche)[1], {
        for(drug_name in names(drugs_volcano)){
          Data_drug_volcano = Data_drug_volcano_all %>%
            dplyr::filter(Drug == drug_name) %>%
            dplyr::select(ID, 最良総合効果) %>%
            dplyr::distinct(ID, .keep_all = T)
          if(length(Data_drug_volcano$ID) >= 1){
            Data_MAF_target_volcano = Data_MAF_target_tmp %>%
              dplyr::filter(Tumor_Sample_Barcode %in%
                              Data_drug_volcano$ID) %>%
              dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol)
            Data_case_target_volcano = Data_case_target %>%
              dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                              Data_drug_volcano$ID) %>%
              dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, 症例.基本情報.がん種.OncoTree..名称.)
            
            colnames(Data_drug_volcano) = c("ID", "Response")
            colnames(Data_MAF_target_volcano) = c("ID", "Gene")
            colnames(Data_case_target_volcano) = c("ID", "Diagnosis")
            Data_MAF_target_volcano = dplyr::left_join(Data_MAF_target_volcano, Data_drug_volcano, by="ID")
            Data_MAF_target_volcano = dplyr::left_join(Data_MAF_target_volcano, Data_case_target_volcano, by="ID")
            
            df_volcano_tmp$Regimen = drug_name
            df_volcano_tmp$Total_treated_patients = length(Data_drug_volcano$ID)
            gene_table_volcano = sort(table(Data_MAF_target_volcano$Gene),decreasing = T)
            gene_table_volcano = gene_table_volcano[gene_table_volcano>=8]
            if(length(gene_table_volcano) > 0){
              for(gene_volcano in names(gene_table_volcano)){
                Data_drug_volcano_pos = Data_MAF_target_volcano %>%
                  dplyr::filter(Gene == gene_volcano) %>%
                  dplyr::select(-Gene)
                Data_drug_volcano_neg = Data_MAF_target_volcano %>%
                  dplyr::select(-Gene) %>%
                  dplyr::distinct()
                Data_drug_volcano_neg = setdiff(Data_drug_volcano_neg, Data_drug_volcano_pos)
                if(length(Data_drug_volcano_pos$Response) > 0){
                  Data_drug_volcano_pos$mutation = 1
                }
                if(length(Data_drug_volcano_neg$Response) > 0){
                  Data_drug_volcano_neg$mutation = 0
                }
                if(length(Data_drug_volcano_pos$Response) > 0 & length(Data_drug_volcano_neg$Response) > 0){
                  Data_drug_volcano_neg_pos = rbind(Data_drug_volcano_pos, Data_drug_volcano_neg)
                } else if(length(Data_drug_volcano_pos$Response) > 0){
                  Data_drug_volcano_neg_pos = Data_drug_volcano_pos
                } else{
                  Data_drug_volcano_neg_pos = Data_drug_volcano_neg
                }
                df_volcano_tmp$Gene = gene_volcano
                df_volcano_tmp$Mut_respond_patients = sum(Data_drug_volcano_pos$Response)
                df_volcano_tmp$Mut_nonrespond_patients = length(Data_drug_volcano_pos$Response) - df_volcano_tmp$Mut_respond_patients
                df_volcano_tmp$WT_respond_patients = sum(Data_drug_volcano_neg$Response)
                df_volcano_tmp$WT_nonrespond_patients = length(Data_drug_volcano_neg$Response) - df_volcano_tmp$WT_respond_patients
                # print(gene_volcano)
                # print(drug_name)
                Data_drug_volcano_neg_pos$Diagnosis = factor(Data_drug_volcano_neg_pos$Diagnosis)
                Data_drug_volcano_neg_pos$Diagnosis = relevel(Data_drug_volcano_neg_pos$Diagnosis, ref=names(sort(table(Data_drug_volcano_neg_pos$Diagnosis),decreasing = T))[[1]])
                if(input$volcano_pathology == "Yes" &
                   min(table(Data_drug_volcano_neg_pos$Diagnosis))>1 &
                   max(table(Data_drug_volcano_neg_pos$Diagnosis))<(length(Data_drug_volcano_neg_pos$Diagnosis)-1) &
                   sum(Data_drug_volcano_neg_pos$mutation)>1 &
                   sum(Data_drug_volcano_neg_pos$mutation)<(length(Data_drug_volcano_neg_pos$mutation)-1) &
                   sum(Data_drug_volcano_neg_pos$Response)>1 &
                   sum(Data_drug_volcano_neg_pos$Response)<(length(Data_drug_volcano_neg_pos$Response)-1)){
                  Data_ORR = glm(Response ~ mutation + Diagnosis, data=Data_drug_volcano_neg_pos, family=binomial)
                  df_volcano_tmp$OddsRatio_considering_pathology = exp(coef(Data_ORR))["mutation"]
                  df_volcano_tmp$p_value_logistic_regression = summary(Data_ORR)$coefficients["mutation",4]
                } else {
                  major_pathology = names(table(Data_drug_volcano_neg_pos$Diagnosis)[table(Data_drug_volcano_neg_pos$Diagnosis) > 1])
                  Data_drug_volcano_neg_pos_major = Data_drug_volcano_neg_pos %>%
                    dplyr::filter(Diagnosis %in% major_pathology)
                  if(input$volcano_pathology == "Yes" &
                     min(table(Data_drug_volcano_neg_pos_major$Diagnosis))>1 &
                     max(table(Data_drug_volcano_neg_pos_major$Diagnosis))<(length(Data_drug_volcano_neg_pos_major$Diagnosis)-1) &
                     sum(Data_drug_volcano_neg_pos_major$mutation)>1 &
                     sum(Data_drug_volcano_neg_pos_major$mutation)<(length(Data_drug_volcano_neg_pos_major$mutation)-1) &
                     sum(Data_drug_volcano_neg_pos_major$Response)>1 &
                     sum(Data_drug_volcano_neg_pos_major$Response)<(length(Data_drug_volcano_neg_pos_major$Response)-1)){
                    Data_ORR = glm(Response ~ mutation + Diagnosis, data=Data_drug_volcano_neg_pos_major, family=binomial)
                    df_volcano_tmp$OddsRatio_considering_pathology = exp(coef(Data_ORR))["mutation"]
                    df_volcano_tmp$p_value_logistic_regression = summary(Data_ORR)$coefficients["mutation",4]
                  } else if(sum(Data_drug_volcano_neg_pos$mutation)>1 &
                            sum(Data_drug_volcano_neg_pos$mutation)<(length(Data_drug_volcano_neg_pos$mutation)-1) &
                            sum(Data_drug_volcano_neg_pos$Response)>1 &
                            sum(Data_drug_volcano_neg_pos$Response)<(length(Data_drug_volcano_neg_pos$mutation)-1)){
                    Data_ORR = glm(Response ~ mutation, data=Data_drug_volcano_neg_pos, family=binomial)
                    df_volcano_tmp$OddsRatio_considering_pathology = exp(coef(Data_ORR))["mutation"]
                    df_volcano_tmp$p_value_logistic_regression = summary(Data_ORR)$coefficients["mutation",4]
                  } else {
                  df_volcano_tmp$OddsRatio_considering_pathology = 1
                  df_volcano_tmp$p_value_logistic_regression = 1
                  }
                  if((df_volcano_tmp$OddsRatio_considering_pathology > 16 | df_volcano_tmp$OddsRatio_considering_pathology < 1/16) &
                     df_volcano_tmp$Mut_respond_patients == 0 |
                     df_volcano_tmp$Mut_nonrespond_patients == 0 |
                     df_volcano_tmp$WT_respond_patients == 0 |
                     df_volcano_tmp$WT_nonrespond_patients == 0){
                    fisher_tmp = matrix(c(df_volcano_tmp$Mut_respond_patients,
                                          df_volcano_tmp$Mut_nonrespond_patients,
                                          df_volcano_tmp$WT_respond_patients,
                                          df_volcano_tmp$WT_nonrespond_patients),
                                        ncol = 2
                    )
                    df_volcano_tmp$p_value_logistic_regression = fisher.test(fisher_tmp)$p.value
                  }
                }
                df_volcano = rbind(df_volcano, df_volcano_tmp)
              }
  
              if(figure_no<200){
                Figure_volcano_tmp = df_volcano %>%
                  dplyr::filter(Regimen == drug_name) %>%
                  dplyr::mutate(Gene_effect = ifelse(p_value_logistic_regression >= 0.05 , "Not significant",
                                                 ifelse(OddsRatio_considering_pathology >= 2, "Up",
                                                        ifelse(OddsRatio_considering_pathology <= 0.5, "Down","Not significant"))))
                Figure_volcano_tmp$OddsRatio_considering_pathology[Figure_volcano_tmp$OddsRatio_considering_pathology>16]=16
                Figure_volcano_tmp$OddsRatio_considering_pathology[Figure_volcano_tmp$OddsRatio_considering_pathology<1/16]=1/16
                Figure_volcano_tmp$Gene_effect = factor(Figure_volcano_tmp$Gene_effect, levels = c("Down", "Not significant", "Up"))
                # Figure_volcano_tmp$label = ifelse((Figure_volcano_tmp$OddsRatio_considering_pathology >=2 &
                #                                      Figure_volcano_tmp$p_value_logistic_regression < 0.05) |
                #                             (Figure_volcano_tmp$OddsRatio_considering_pathology <= 0.5 &
                #                                Figure_volcano_tmp$p_value_logistic_regression < 0.05), Figure_volcano_tmp$Gene, NA_character_)
                Figure_volcano_tmp$label = ifelse((Figure_volcano_tmp$OddsRatio_considering_pathology >=2 |
                                                   Figure_volcano_tmp$p_value_logistic_regression < 0.05 |
                                                   Figure_volcano_tmp$OddsRatio_considering_pathology <= 0.5), Figure_volcano_tmp$Gene, NA_character_)
                
                Figure_volcano_tmp$log2OR = log2(Figure_volcano_tmp$OddsRatio_considering_pathology)
                Figure_volcano_tmp$log10pval = -log10(Figure_volcano_tmp$p_value_logistic_regression)
                if(length(Figure_volcano_tmp$label>0)){
                  g_volcano[[figure_no]] =
                    ggplot(data = Figure_volcano_tmp, aes(x = log2OR, y = log10pval, col = Gene_effect, label = label)) +
                    geom_point(size = 2) +
                    labs(x = expression("log"[2]*"odds ratio of objective response"), y = expression("-log"[10]*"p-value")) +
                    scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"), 
                                       labels = c("Down", "Not significant", "Up"),
                                       drop = FALSE) +
                    geom_hline(yintercept = -log10(0.05), size = 0.2, color = "dark green") + 
                    geom_vline(xintercept = -1, size = 0.2, color = "dark green", linetype = 'dashed') +
                    geom_vline(xintercept = 1, size = 0.2, color = "dark green", linetype = 'dashed') +
                    theme(legend.position = "none", panel.grid = element_blank()) +
                    geom_text_repel(max.overlaps = Inf) +
                    ggtitle(paste0(drug_name, " ",
                                   df_volcano_tmp$Total_treated_patients, " patients")) +
                    theme_classic()
                }
                figure_no = figure_no + 1
              }
            }
          }
          incProgress(1 / length(names(drugs_volcano)))
        }
      })
      options(warn = 0)
      if(!is.null(input$drug)){
        Data_drug_volcano = Data_drug_volcano_all %>%
          dplyr::filter(Drug %in% input$drug) %>%
          dplyr::select(ID, 最良総合効果) %>%
          dplyr::distinct(ID, .keep_all = T)
        if(length(Data_drug_volcano$ID) >= 1){
          Data_MAF_target_volcano = Data_MAF_target_tmp %>%
            dplyr::filter(Tumor_Sample_Barcode %in%
                            Data_drug_volcano$ID) %>%
            dplyr::distinct(Tumor_Sample_Barcode, Hugo_Symbol)
          Data_case_target_volcano = Data_case_target %>%
            dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in%
                            Data_drug_volcano$ID) %>%
            dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, 症例.基本情報.がん種.OncoTree..名称.)
          
          colnames(Data_drug_volcano) = c("ID", "Response")
          colnames(Data_MAF_target_volcano) = c("ID", "Gene")
          colnames(Data_case_target_volcano) = c("ID", "Diagnosis")
          Data_MAF_target_volcano = dplyr::left_join(Data_MAF_target_volcano, Data_drug_volcano, by="ID")
          Data_MAF_target_volcano = dplyr::left_join(Data_MAF_target_volcano, Data_case_target_volcano, by="ID")
          
          df_volcano_tmp$Regimen = paste(input$drug, collapse = ";")
          df_volcano_tmp$Total_treated_patients = length(Data_drug_volcano$ID)
          gene_table_volcano = sort(table(Data_MAF_target_volcano$Gene),decreasing = T)
          gene_table_volcano = gene_table_volcano[gene_table_volcano>=8]
          if(length(gene_table_volcano) > 0){
            for(gene_volcano in names(gene_table_volcano)){
              Data_drug_volcano_pos = Data_MAF_target_volcano %>%
                dplyr::filter(Gene == gene_volcano) %>%
                dplyr::select(-Gene)
              Data_drug_volcano_neg = Data_MAF_target_volcano %>%
                dplyr::select(-Gene) %>%
                dplyr::distinct()
              Data_drug_volcano_neg = setdiff(Data_drug_volcano_neg, Data_drug_volcano_pos)
              if(length(Data_drug_volcano_pos$Response) > 0){
                Data_drug_volcano_pos$mutation = 1
              }
              if(length(Data_drug_volcano_neg$Response) > 0){
                Data_drug_volcano_neg$mutation = 0
              }
              if(length(Data_drug_volcano_pos$Response) > 0 & length(Data_drug_volcano_neg$Response) > 0){
                Data_drug_volcano_neg_pos = rbind(Data_drug_volcano_pos, Data_drug_volcano_neg)
              } else if(length(Data_drug_volcano_pos$Response) > 0){
                Data_drug_volcano_neg_pos = Data_drug_volcano_pos
              } else {
                Data_drug_volcano_neg_pos = Data_drug_volcano_neg
              }
              df_volcano_tmp$Gene = gene_volcano
              df_volcano_tmp$Mut_respond_patients = sum(Data_drug_volcano_pos$Response)
              df_volcano_tmp$Mut_nonrespond_patients = length(Data_drug_volcano_pos$Response) - df_volcano_tmp$Mut_respond_patients
              df_volcano_tmp$WT_respond_patients = sum(Data_drug_volcano_neg$Response)
              df_volcano_tmp$WT_nonrespond_patients = length(Data_drug_volcano_neg$Response) - df_volcano_tmp$WT_respond_patients
              # print(gene_volcano)
              # print(drug_name)
              Data_drug_volcano_neg_pos$Diagnosis = factor(Data_drug_volcano_neg_pos$Diagnosis)
              Data_drug_volcano_neg_pos$Diagnosis = relevel(Data_drug_volcano_neg_pos$Diagnosis, ref=names(sort(table(Data_drug_volcano_neg_pos$Diagnosis),decreasing = T))[[1]])
              if(input$volcano_pathology == "Yes" &
                 min(table(Data_drug_volcano_neg_pos$Diagnosis))>1 &
                 max(table(Data_drug_volcano_neg_pos$Diagnosis))<(length(Data_drug_volcano_neg_pos$Diagnosis)-1) &
                 sum(Data_drug_volcano_neg_pos$mutation)>1 &
                 sum(Data_drug_volcano_neg_pos$mutation)<(length(Data_drug_volcano_neg_pos$mutation)-1) &
                 sum(Data_drug_volcano_neg_pos$Response)>1 &
                 sum(Data_drug_volcano_neg_pos$Response)<(length(Data_drug_volcano_neg_pos$Response)-1)){
                Data_ORR = glm(Response ~ mutation + Diagnosis, data=Data_drug_volcano_neg_pos, family=binomial)
                df_volcano_tmp$OddsRatio_considering_pathology = exp(coef(Data_ORR))["mutation"]
                df_volcano_tmp$p_value_logistic_regression = summary(Data_ORR)$coefficients["mutation",4]
              } else {
                major_pathology = names(table(Data_drug_volcano_neg_pos$Diagnosis)[table(Data_drug_volcano_neg_pos$Diagnosis) > 1])
                Data_drug_volcano_neg_pos_major = Data_drug_volcano_neg_pos %>%
                  dplyr::filter(Diagnosis %in% major_pathology)
                if(input$volcano_pathology == "Yes" &
                   min(table(Data_drug_volcano_neg_pos_major$Diagnosis))>1 &
                   max(table(Data_drug_volcano_neg_pos_major$Diagnosis))<(length(Data_drug_volcano_neg_pos_major$Diagnosis)-1) &
                   sum(Data_drug_volcano_neg_pos_major$mutation)>1 &
                   sum(Data_drug_volcano_neg_pos_major$mutation)<(length(Data_drug_volcano_neg_pos_major$mutation)-1) &
                   sum(Data_drug_volcano_neg_pos_major$Response)>1 &
                   sum(Data_drug_volcano_neg_pos_major$Response)<(length(Data_drug_volcano_neg_pos_major$Response)-1)){
                  Data_ORR = glm(Response ~ mutation + Diagnosis, data=Data_drug_volcano_neg_pos_major, family=binomial)
                  df_volcano_tmp$OddsRatio_considering_pathology = exp(coef(Data_ORR))["mutation"]
                  df_volcano_tmp$p_value_logistic_regression = summary(Data_ORR)$coefficients["mutation",4]
                } else if(sum(Data_drug_volcano_neg_pos$mutation)>1 &
                          sum(Data_drug_volcano_neg_pos$mutation)<(length(Data_drug_volcano_neg_pos$mutation)-1) &
                          sum(Data_drug_volcano_neg_pos$Response)>1 &
                          sum(Data_drug_volcano_neg_pos$Response)<(length(Data_drug_volcano_neg_pos$mutation)-1)){
                  Data_ORR = glm(Response ~ mutation, data=Data_drug_volcano_neg_pos, family=binomial)
                  df_volcano_tmp$OddsRatio_considering_pathology = exp(coef(Data_ORR))["mutation"]
                  df_volcano_tmp$p_value_logistic_regression = summary(Data_ORR)$coefficients["mutation",4]
                } else {
                  df_volcano_tmp$OddsRatio_considering_pathology = 1
                  df_volcano_tmp$p_value_logistic_regression = 1
                }
              }
              df_volcano = rbind(df_volcano, df_volcano_tmp)
            }
            Figure_volcano = df_volcano %>%
              dplyr::filter(Regimen == paste(input$drug, collapse = ";")) %>%
              dplyr::mutate(Gene_effect = ifelse(p_value_logistic_regression >= 0.05 , "Not significant",
                                                 ifelse(OddsRatio_considering_pathology >= 2, "Up",
                                                        ifelse(OddsRatio_considering_pathology <= 0.5, "Down","Not significant"))))
            Figure_volcano$label = ifelse((Figure_volcano$OddsRatio_considering_pathology >=2 |
                                             Figure_volcano$p_value_logistic_regression < 0.05 |
                                             Figure_volcano$OddsRatio_considering_pathology <= 0.5), Figure_volcano$Gene, NA_character_)
            Figure_volcano$Gene_effect = factor(Figure_volcano$Gene_effect, levels = c("Down", "Not significant", "Up"))
            Figure_volcano$OddsRatio_considering_pathology[Figure_volcano$OddsRatio_considering_pathology>16]=16
            Figure_volcano$OddsRatio_considering_pathology[Figure_volcano$OddsRatio_considering_pathology<1/16]=1/16
            
            Figure_volcano$log2OR = log2(Figure_volcano$OddsRatio_considering_pathology)
            Figure_volcano$log10pval = -log10(Figure_volcano$p_value_logistic_regression)
            if(length(Figure_volcano$label>0)){
              g_volcano[[figure_no]] =
                ggplot(data = Figure_volcano, aes(x = log2OR, y = log10pval, col = Gene_effect, label = label)) +
                geom_point(size = 2) +
                labs(x = expression("log"[2]*"odds ratio of objective response"), y = expression("-log"[10]*"p-value")) +
                scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"), 
                                   labels = c("Down", "Not significant", "Up"),
                                   drop = FALSE) +
                geom_hline(yintercept = -log10(0.05), size = 0.2, color = "dark green") + 
                geom_vline(xintercept = -1, size = 0.2, color = "dark green", linetype = 'dashed') +
                geom_vline(xintercept = 1, size = 0.2, color = "dark green", linetype = 'dashed') +
                theme(legend.position = "none", panel.grid = element_blank()) +
                ggtitle(paste0(paste(input$drug, collapse = ";"), " ",
                               Figure_volcano$Total_treated_patients[[1]], " patients")) +
                geom_text_repel(max.overlaps = Inf) +
                theme_classic()
              figure_no = figure_no + 1
            }
            significant_genes = c(significant_genes, (Figure_volcano %>% dplyr::arrange(p_value_logistic_regression))$label)
            significant_genes = significant_genes[!is.na(significant_genes)]
          }
        }
      }
      
      if(figure_no<=200){
        for(figure_no_tmp in figure_no:200){
          g_volcano[[figure_no_tmp]] = gg_empty()
        }
      }

      output$figure_volcano_1 = renderPlot({
        grid.arrange(grobs = g_volcano[1:50], nrow = 10, ncol = 5)
      })
      output$figure_volcano_2 = renderPlot({
        grid.arrange(grobs = g_volcano[51:100], nrow = 10, ncol = 5)
      })
      output$figure_volcano_3 = renderPlot({
        grid.arrange(grobs = g_volcano[101:150], nrow = 10, ncol = 5)
      })
      output$figure_volcano_4 = renderPlot({
        grid.arrange(grobs = g_volcano[151:200], nrow = 10, ncol = 5)
      })
      

      output$table_volcano = DT::renderDataTable(df_volcano,
                                                 server = FALSE,
                                                 filter = 'top', 
                                                 extensions = c('Buttons'), 
                                                 options = list(pageLength = 100, 
                                                                scrollX = TRUE,
                                                                scrollY = "1000px",
                                                                scrollCollapse = TRUE,
                                                                dom="Blfrtip",
                                                                buttons = c('csv', 'excel', 'copy')))
      if(input$drug_volcano != "Yes"){

        Data_drug_RECIST = Data_drug_RECIST %>%
          dplyr::filter(Drug %in% input$drug) %>%
          dplyr::filter(治療ライン %in% input$target_line) %>%
          dplyr::arrange(ID, 治療ライン) %>%
          dplyr::distinct(ID, Drug, .keep_all = T)
  
        Drug_summary_RECIST = Data_drug_RECIST %>%
          dplyr::filter(!最良総合効果 %in% c("NE", "Unknown")) %>%
          dplyr::select(
            ID,
            Drug,
            治療ライン,
            最良総合効果,
            終了理由,
            Drug_length,
            TTF_censor
          )
        colnames(Drug_summary_RECIST)=
          c("ID", "治療薬", "治療ライン", "RECIST", "終了理由", "Time on treatment", "打ち切り0,終了1")
        Drug_summary_RECIST$終了理由[Drug_summary_RECIST$終了理由 == ""] = "入力なし"
        if(!input$special_gene == ""){
          Data_MAF_target_tmp = Data_MAF_target_tmp %>%
            dplyr::filter(Tumor_Sample_Barcode %in%
                            Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
            dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
          
          ID_special_gene = (Data_MAF_target_tmp %>%
                               dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_"))))$Tumor_Sample_Barcode
          ID_special_gene_mutation_1 = (Data_MAF_target_tmp %>%
                                          dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                          amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
          ID_special_gene_mutation_2 = (Data_MAF_target_tmp %>%
                                          dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                          amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
          Drug_summary_RECIST_line = Drug_summary_RECIST %>% dplyr::mutate(
            separation_value = case_when(
              ID %in% ID_special_gene_mutation_1 ~ paste0(input$special_gene_mutation_1_name, " in ", input$special_gene),
              ID %in% ID_special_gene_mutation_2 ~ paste0(input$special_gene_mutation_2_name, " in ", input$special_gene),
              ID %in% ID_special_gene ~ paste0("Other mut in ", input$special_gene),
              TRUE ~ paste0("No mut in ", input$special_gene)
            )
          )
        } else{
          Drug_summary_RECIST_line = Drug_summary_RECIST %>% dplyr::mutate(
            separation_value = "No mutation pattern")
        }
        colnames(Drug_summary_RECIST_line)=
          c("ID", "Drugs", "CTx line", "RECIST", "Reason to finish treatment", "Time on treatment", "Treatment finished or censored", "separation_value")
        Drug_summary_RECIST_line$`Reason to finish treatment`[is.na(Drug_summary_RECIST_line$`Reason to finish treatment`)] = "Unknown"
        Drug_summary_RECIST_line$`Reason to finish treatment`[Drug_summary_RECIST_line$`Reason to finish treatment` == "その他理由で中止"] = "Other reason"
        Drug_summary_RECIST_line$`Reason to finish treatment`[Drug_summary_RECIST_line$`Reason to finish treatment` == "不明"] = "Unknown"
        Drug_summary_RECIST_line$`Reason to finish treatment`[Drug_summary_RECIST_line$`Reason to finish treatment` == "副作用等で中止"] = "Side effect"
        Drug_summary_RECIST_line$`Reason to finish treatment`[Drug_summary_RECIST_line$`Reason to finish treatment` == "本人希望により中止"] = "Patient will"
        Drug_summary_RECIST_line$`Reason to finish treatment`[Drug_summary_RECIST_line$`Reason to finish treatment` == "無効中止"] = "Not effective"
        Drug_summary_RECIST_line$`Reason to finish treatment`[Drug_summary_RECIST_line$`Reason to finish treatment` == "死亡中止"] = "Death"
        Drug_summary_RECIST_line$`Reason to finish treatment`[Drug_summary_RECIST_line$`Reason to finish treatment` == "計画通り終了"] = "As planned"
        Drug_summary_RECIST_line$`Treatment finished or censored` = as.character(Drug_summary_RECIST_line$`Treatment finished or censored`)
        Drug_summary_RECIST_line$`Treatment finished or censored`[is.na(Drug_summary_RECIST_line$`Treatment finished or censored`)] = "Unknown"
        Drug_summary_RECIST_line$`Treatment finished or censored`[Drug_summary_RECIST_line$`Treatment finished or censored` == "1"] = "Finished"
        Drug_summary_RECIST_line$`Treatment finished or censored`[Drug_summary_RECIST_line$`Treatment finished or censored` == "0"] = "Censored"
        
        output$table_drug_RECIST_1 <- render_gt({
          Drug_summary_RECIST_line %>%
            dplyr::select(-ID) %>%
            tbl_summary(
              by = "separation_value",
              statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                    "{mean} ({sd})",
                                                    "{median} ({p25}, {p75})", 
                                                    "{min}, {max}"),
                               all_categorical() ~ "{n} / {N} ({p}%)"),
              type = list(all_continuous() ~ "continuous2",
                          all_dichotomous() ~ "categorical"),
              digits = all_continuous() ~ 2,
              missing = "ifany") %>%
            modify_caption("**Palliative CTx with RECIST evaluation, selected lines, by mutation pattern**") %>%
            add_n() %>% bold_labels() %>% as_gt()
        })
        
        Drug_summary_RECIST_diagnosis = Drug_summary_RECIST_line
        ID_diagnosis_list = Data_case_target %>% dplyr::select(C.CAT調査結果.基本項目.ハッシュID,症例.基本情報.がん種.OncoTree.)
        Drug_summary_RECIST_diagnosis$diagnosis = unlist(lapply(list(Drug_summary_RECIST_diagnosis$ID), function(x) {
          as.vector(ID_diagnosis_list$症例.基本情報.がん種.OncoTree.[match(x, ID_diagnosis_list$C.CAT調査結果.基本項目.ハッシュID)])}))
        output$table_drug_RECIST_2 <- render_gt({
          Drug_summary_RECIST_diagnosis %>%
            dplyr::select(-ID, -separation_value) %>%
            tbl_summary(
              by = "diagnosis",
              statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                    "{mean} ({sd})",
                                                    "{median} ({p25}, {p75})", 
                                                    "{min}, {max}"),
                               all_categorical() ~ "{n} / {N} ({p}%)"),
              type = list(all_continuous() ~ "continuous2",
                          all_dichotomous() ~ "categorical"),
              digits = all_continuous() ~ 2,
              missing = "ifany") %>%
            modify_caption("**Palliative CTx with RECIST evaluation, selected lines, by diagnosis**") %>%
            add_n() %>% bold_labels() %>% as_gt()
        })
        
        mut_gene_ = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
        if(length(mut_gene_)==0){
          mut_gene_ = "TP53"
          ID_mutation_ = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene_))$Tumor_Sample_Barcode
          Drug_summary_RECIST_mutation = Drug_summary_RECIST_diagnosis %>% dplyr::mutate(
            mutation = case_when(
              ID %in% ID_mutation_ ~ "TP53 mut(+)",
              TRUE ~ "TP53 mut(-)"
            )
          )
        } else{
          if(is.null(input$gene_group_2)){
            ID_mutation_ = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene_))$Tumor_Sample_Barcode
            Drug_summary_RECIST_mutation = Drug_summary_RECIST_diagnosis %>% dplyr::mutate(
              mutation = case_when(
                ID %in% ID_mutation_ ~ paste0(paste(mut_gene_, collapse = "/"), " mut(+)"),
                TRUE ~ paste0(paste(mut_gene_, collapse = "/"), " mut(-)")
              )
            )
          } else{
            mut_gene_ = input$gene
            ID_mutation_1 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_1))$Tumor_Sample_Barcode
            ID_mutation_2 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_2))$Tumor_Sample_Barcode
            Drug_summary_RECIST_mutation = Drug_summary_RECIST_diagnosis %>% dplyr::mutate(
              mutation = case_when(
                ID %in% ID_mutation_1 & ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+) & ", paste(input$gene_group_2, collapse = "/"), " mut(+)"),
                ID %in% ID_mutation_1 & !ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+) & ", paste(input$gene_group_2, collapse = "/"), " mut(-)"),
                !ID %in% ID_mutation_1 & ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-) & ", paste(input$gene_group_2, collapse = "/"), " mut(+)"),
                !ID %in% ID_mutation_1 & !ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-) & ", paste(input$gene_group_2, collapse = "/"), " mut(-)"),
              )
            )
          }
        }
        
        output$table_drug_RECIST_3 <- render_gt({
          Drug_summary_RECIST_mutation %>%
            dplyr::select(-ID, -separation_value, -diagnosis) %>%
            tbl_summary(
              by = "mutation",
              statistic = list(all_continuous() ~ c("{N_nonmiss}",
                                                    "{mean} ({sd})",
                                                    "{median} ({p25}, {p75})", 
                                                    "{min}, {max}"),
                               all_categorical() ~ "{n} / {N} ({p}%)"),
              type = list(all_continuous() ~ "continuous2",
                          all_dichotomous() ~ "categorical"),
              digits = all_continuous() ~ 2,
              missing = "ifany") %>%
            modify_caption("**Palliative CTx with RECIST evaluation, selected lines, by mutation**") %>%
            add_n() %>% bold_labels() %>% as_gt()
        })
        
        Data_forest = Data_drug_RECIST %>%
          dplyr::select(ID, Drug, Drug_length, TTF_censor, 最良総合効果, 治療ライン)
        colnames(Data_forest) = c("ID", "Drug", "Drug_length", "TTF_censor", "最良総合効果", "Lines")
        Data_survival_tmp = Data_survival %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID,
                        YoungOld,
                        Lymph_met,
                        Brain_met,
                        Lung_met,
                        Bone_met,
                        Liver_met,
                        症例.基本情報.性別.名称.,
                        症例.基本情報.がん種.OncoTree.,
                        症例.基本情報.がん種.OncoTree.LEVEL1.,
                        HER2_IHC,
                        MSI_PCR,
                        MMR_IHC,
                        cluster#,
                        #症例.背景情報.喫煙歴有無.名称.,
                        #症例.背景情報.アルコール多飲有無.名称.
                        )
        col_added = 0
        if(input$HER2 != "No"){
          col_added = col_added + 1
        } else {
          Data_survival_tmp = Data_survival_tmp %>%
            dplyr::select(-HER2_IHC)
        }
        if(input$MSI != "No"){
          col_added = col_added + 1
        } else {
          Data_survival_tmp = Data_survival_tmp %>%
            dplyr::select(-MSI_PCR)
        }
        if(input$MMR != "No"){
          col_added = col_added + 1
        } else {
          Data_survival_tmp = Data_survival_tmp %>%
            dplyr::select(-MMR_IHC)
        }
   
        Data_survival_tmp = Data_survival_tmp %>%
          dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = TRUE) %>%
          dplyr::filter(C.CAT調査結果.基本項目.ハッシュID %in% Data_forest$ID) #%>%
          # dplyr::mutate(
          #   症例.背景情報.喫煙歴有無.名称. = case_when(
          #     症例.背景情報.喫煙歴有無.名称. == "あり" ~ "Yes",
          #     症例.背景情報.喫煙歴有無.名称. == "なし" ~ "No",
          #     TRUE ~ "Unknown"
          #   ),
          #   症例.背景情報.アルコール多飲有無.名称. = case_when(
          #     症例.背景情報.アルコール多飲有無.名称. == "あり" ~ "Yes",
          #     症例.背景情報.アルコール多飲有無.名称. == "なし" ~ "No",
          #     TRUE ~ "Unknown"
          #   )
          # )
        colnames_tmp = colnames(Data_survival_tmp)
        colnames_tmp[colnames_tmp == "C.CAT調査結果.基本項目.ハッシュID"] = "ID"
        colnames_tmp[colnames_tmp == "YoungOld"] = "Age"
        colnames_tmp[colnames_tmp == "症例.基本情報.性別.名称."] = "Sex"
        colnames_tmp[colnames_tmp == "症例.基本情報.がん種.OncoTree."] = "Histology"
        colnames_tmp[colnames_tmp == "cluster"] = "Cluster"
        colnames_tmp[colnames_tmp == "症例.基本情報.がん種.OncoTree.LEVEL1."] = "Histology_dummy"
        colnames(Data_survival_tmp) = colnames_tmp
        
        #Data_survival_tmp$Histology_dummy = paste0(Data_survival_tmp$Histology_dummy,"_NOS")
        Data_survival_tmp$Age[is.na(Data_survival_tmp$Age)] <- "不明"
        Data_forest = left_join(Data_forest, Data_survival_tmp, by="ID")
        Data_forest = Data_forest %>% dplyr::mutate(
          Lines = case_when(
            Lines == "1" ~ "1",
            Lines == "2" ~ "2",
            TRUE ~ "3~"
          ),
          ORR_data = case_when(
            最良総合効果 %in% c("CR", "PR") ~ 1,
            最良総合効果 %in% c("LongSD", "SD", "PD") ~ 0,
            最良総合効果 %in% c("NE", "Unknown") ~ -1,
            TRUE ~ -1
          ),
          DCR_data = case_when(
            最良総合効果 %in% c("CR", "PR", "LongSD", "SD") ~ 1,
            最良総合効果 %in% c("PD") ~ 0,
            最良総合効果 %in% c("NE", "Unknown") ~ -1,
            TRUE ~ -1
          )
        )
        
        Data_forest$Cluster = factor(Data_forest$Cluster)
        Gene_data$differentiation=ifelse(Gene_data$diff_UL<0, Gene_data$diff_UL,
                                         ifelse(Gene_data$diff_LL>0, -Gene_data$diff_LL, 0))
        Gene_data_significant = Gene_data %>%
          dplyr::arrange(differentiation) %>%
          dplyr::filter(differentiation > 0)
        gene_names = unique(c(input$gene,
                              as.character(Gene_data$Gene[1:min(5, length(Gene_data$Gene))]),
                              significant_genes,
                              as.character(Gene_data_significant$Gene)
                              ))
        gene_names = gene_names[gene_names %in% unique(Data_MAF_target$Hugo_Symbol)]
        for(i in 1:length(gene_names)){
          ID_mutation = unique((Data_MAF_target %>% dplyr::filter(Hugo_Symbol == gene_names[i]))$Tumor_Sample_Barcode)
          Data_forest_tmp = Data_forest %>%
            dplyr::select(ID) %>%
            dplyr::distinct() %>%
            dplyr::mutate(
              XXX = case_when(
                ID %in% ID_mutation ~ "mut(+)",
                TRUE ~ "mut(-)"
              )
            )
          colnames(Data_forest_tmp) = c("ID", gene_names[i])
          Data_forest = left_join(Data_forest, Data_forest_tmp, by="ID")
        }

        Factor_names = colnames(Data_forest)
        Factor_names = Factor_names[!Factor_names %in% c('ID', 'Drug',
                                                         "TTF_censor", "Drug_length",
                                                         "最良総合効果", "ORR_data", "DCR_data")]
        Data_forest_all_pts = Data_forest
        Data_forest = Data_forest %>%
          dplyr::filter(Sex != "未入力・不明") %>%
          dplyr::filter(Age != "不明") #%>%
          #dplyr::filter(Smoking_history != "Unknown" & Alcoholic_history != "Unknown")
        if(input$HER2 != "No"){
          Data_forest = Data_forest %>%
            dplyr::filter(HER2_IHC != "Unknown")
        }
        if(input$MSI != "No"){
          Data_forest = Data_forest %>%
            dplyr::filter(MSI_PCR != "Unknown")
        }
        if(input$MMR != "No"){
          Data_forest = Data_forest %>%
            dplyr::filter(MMR_IHC != "Unknown")
        }
        
        Data_forest_tmp = Data_forest %>% dplyr::filter(!is.na(TTF_censor))
        Disease_tmp = unique(sort(Data_forest_tmp$Histology))
        Disease_tmp2 = unique(sort(Data_forest_tmp$Histology_dummy))
        Disease_tmp = c(Disease_tmp[!Disease_tmp %in% Disease_tmp2], Disease_tmp2)
        for(i in length(gene_names):1){
          if(sum(Data_forest_tmp[,18+col_added+i] == "mut(+)" & Data_forest_tmp$TTF_censor == 1) < 3){
            Factor_names = Factor_names[-(11+col_added+i)]
          } else if(sum(Data_forest_tmp[,18+col_added+i] == "mut(-)" & Data_forest_tmp$TTF_censor == 1) < 3){
            Factor_names = Factor_names[-(11+col_added+i)]
          }
        }
        for(i in 1:length(Disease_tmp)){
          Data_forest_tmp2 = Data_forest_tmp %>% dplyr::filter(Histology == Disease_tmp[i])
          if(sum(Data_forest_tmp2$TTF_censor == 1,na.rm = T) < 3){
            Data_forest_tmp = Data_forest_tmp %>%
              dplyr::mutate(Histology = case_when(
                Histology == Disease_tmp[i] ~ Histology_dummy,
                TRUE ~ Histology
              )
            )
          } else {
            Data_forest_tmp2 = Data_forest_tmp %>% dplyr::filter(Histology != Disease_tmp[i])
            if(sum(Data_forest_tmp2$TTF_censor == 1,na.rm = T) < 3){
              Data_forest_tmp = Data_forest_tmp %>%
                dplyr::mutate(Histology = case_when(
                  Histology == Disease_tmp[i] ~ Histology_dummy,
                  TRUE ~ Histology
                )
                )
            }
          }
        }
        Data_forest_tmp$Histology = factor(Data_forest_tmp$Histology)
        if(length(unique(Data_forest_tmp$Histology))<2){
          Factor_names = Factor_names[!Factor_names %in% c('Histology')]
        } else {
          Data_forest_tmp$Histology = relevel(Data_forest_tmp$Histology,
                                               ref=names(sort(table(Data_forest_tmp$Histology),decreasing = T))[[1]])
        }
        if(length(unique(Data_forest_tmp$Cluster))<2){
          Factor_names = Factor_names[!Factor_names %in% c('Cluster')]
        } else {
          Data_forest_tmp$Cluster = relevel(Data_forest_tmp$Cluster,
                                              ref=names(sort(table(Data_forest_tmp$Cluster),decreasing = T))[[1]])
        }
        if(sum(Data_forest_tmp$Age == "Younger" & Data_forest_tmp$TTF_censor == 1,na.rm = T) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Age')]
        } else if(sum(Data_forest_tmp$Age == "Older" & Data_forest_tmp$TTF_censor == 1,na.rm = T) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Age')]
        }
        if(sum(Data_forest_tmp$Sex == "男" & Data_forest_tmp$TTF_censor == 1,na.rm = T) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Sex')]
        } else if(sum(Data_forest_tmp$Sex != "男" & Data_forest_tmp$TTF_censor == 1,na.rm = T) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Sex')]
        }
        # if(sum(Data_forest_tmp$Smoking_history == "Yes" & Data_forest_tmp$TTF_censor == 1,na.rm = T) < 3){
        #   Factor_names = Factor_names[!Factor_names %in% c('Smoking_history')]
        # } else if(sum(Data_forest_tmp$Smoking_history != "Yes" & Data_forest_tmp$TTF_censor == 1,na.rm = T) < 3){
        #   Factor_names = Factor_names[!Factor_names %in% c('Smoking_history')]
        # }
        # if(sum(Data_forest_tmp$Alcoholic_history == "Yes" & Data_forest_tmp$TTF_censor == 1,na.rm = T) < 3){
        #   Factor_names = Factor_names[!Factor_names %in% c('Alcoholic_history')]
        # } else if(sum(Data_forest_tmp$Alcoholic_history != "Yes" & Data_forest_tmp$TTF_censor == 1,na.rm = T) < 3){
        #   Factor_names = Factor_names[!Factor_names %in% c('Alcoholic_history')]
        # }
        if(sum(Data_forest_tmp$Lines == "1" & Data_forest_tmp$TTF_censor == 1) < 3 &
           sum(Data_forest_tmp$Lines == "2" & Data_forest_tmp$TTF_censor == 1) < 3 &
           sum(Data_forest_tmp$Lines == "3~" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Lines')]
        } else if((sum(Data_forest_tmp$Lines == "1" & Data_forest_tmp$TTF_censor == 1) >= 3 &
                    sum(Data_forest_tmp$Lines != "1" & Data_forest_tmp$TTF_censor == 1) < 3) |
                  (sum(Data_forest_tmp$Lines == "2" & Data_forest_tmp$TTF_censor == 1) >= 3 &
                   sum(Data_forest_tmp$Lines != "2" & Data_forest_tmp$TTF_censor == 1) < 3) |
                  (sum(Data_forest_tmp$Lines == "3~" & Data_forest_tmp$TTF_censor == 1) >= 3 &
                   sum(Data_forest_tmp$Lines != "3~" & Data_forest_tmp$TTF_censor == 1) < 3)){
          Factor_names = Factor_names[!Factor_names %in% c('Lines')]
        } else if(sum(Data_forest_tmp$Lines == "1" & Data_forest_tmp$TTF_censor == 1) < 3){
          Data_forest_tmp$Lines[Data_forest_tmp$Lines == "1"] = "1~2"
          Data_forest_tmp$Lines[Data_forest_tmp$Lines == "2"] = "1~2"
        } else if(sum(Data_forest_tmp$Lines == "3~" & Data_forest_tmp$TTF_censor == 1) < 3){
          Data_forest_tmp$Lines[Data_forest_tmp$Lines == "3~"] = "2~"
          Data_forest_tmp$Lines[Data_forest_tmp$Lines == "2"] = "2~"
        }
        if(sum(Data_forest_tmp$Lymph_met == "Yes" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Lymph_met')]
        } else if(sum(Data_forest_tmp$Lymph_met != "Yes" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Lymph_met')]
        }
        if(sum(Data_forest_tmp$Lung_met == "Yes" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Lung_met')]
        } else if(sum(Data_forest_tmp$Lung_met != "Yes" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Lung_met')]
        }
        if(sum(Data_forest_tmp$Brain_met == "Yes" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Brain_met')]
        } else if(sum(Data_forest_tmp$Brain_met != "Yes" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Brain_met')]
        }
        if(sum(Data_forest_tmp$Bone_met == "Yes" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Bone_met')]
        } else if(sum(Data_forest_tmp$Bone_met != "Yes" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Bone_met')]
        }
        if(sum(Data_forest_tmp$Liver_met == "Yes" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Liver_met')]
        } else if(sum(Data_forest_tmp$Liver_met != "Yes" & Data_forest_tmp$TTF_censor == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Liver_met')]
        }
        if(input$HER2 != "No"){
          if(sum(Data_forest_tmp$HER2_IHC == "Positive" & Data_forest_tmp$TTF_censor == 1) < 3){
            Factor_names = Factor_names[!Factor_names %in% c('HER2_IHC')]
          } else if(sum(Data_forest_tmp$HER2_IHC != "Positive" & Data_forest_tmp$TTF_censor == 1) < 3){
            Factor_names = Factor_names[!Factor_names %in% c('HER2_IHC')]
          }
        }
        if(input$MSI != "No"){
          if(sum(Data_forest_tmp$MSI_PCR == "Positive" & Data_forest_tmp$TTF_censor == 1) < 3){
            Factor_names = Factor_names[!Factor_names %in% c('MSI_PCR')]
          } else if(sum(Data_forest_tmp$MSI_PCR != "Positive" & Data_forest_tmp$TTF_censor == 1) < 3){
            Factor_names = Factor_names[!Factor_names %in% c('MSI_PCR')]
          }
        }
        if(input$MMR != "No"){
          if(sum(Data_forest_tmp$MMR_IHC == "dMMR" & Data_forest_tmp$TTF_censor == 1) < 3){
            Factor_names = Factor_names[!Factor_names %in% c('MMR_IHC')]
          } else if(sum(Data_forest_tmp$MMR_IHC != "dMMR" & Data_forest_tmp$TTF_censor == 1) < 3){
            Factor_names = Factor_names[!Factor_names %in% c('MMR_IHC')]
          }
        }
        if(length(Factor_names) > 1){
          for(i in length(Factor_names):1){
            if(all(Data_forest_tmp[,colnames(Data_forest_tmp) == Factor_names[i]] ==
                   Data_forest_tmp[1,colnames(Data_forest_tmp) == Factor_names[i]])){
              Factor_names = Factor_names[Factor_names != Factor_names[i]]
            }
          }
        }
        Factor_names = Factor_names[!Factor_names %in% c('Histology_dummy')]
        Factor_names_univariant = Factor_names
        if(length(Factor_names) > 1){
          Factor_names_tmp = Factor_names
          for(i in 1:(length(Factor_names_tmp) - 1)){
            for(j in (i+1):length(Factor_names_tmp)){
              if(Factor_names_tmp[i] %in% colnames(Data_forest_tmp) &
                 Factor_names_tmp[j] %in% colnames(Data_forest_tmp)){
                if(abs(cor(as.numeric(as.factor(Data_forest_tmp[,colnames(Data_forest_tmp) == Factor_names_tmp[i]])),
                           as.numeric(as.factor(Data_forest_tmp[,colnames(Data_forest_tmp) == Factor_names_tmp[j]])))) > 0.95){
                  Factor_names = Factor_names[Factor_names != Factor_names_tmp[j]]
                }
              }
            }
          }
        }
        Data_forest_tmp_x = Data_forest_tmp
        Factor_names_x = Factor_names
        Factor_names_x_univariant = Factor_names_univariant
        
        output$figure_drug_6_1 = render_gt({
          Factor_names_x = Factor_names_x[Factor_names_x != "Cluster"]
          Factor_names_x_univariant = Factor_names_x_univariant[Factor_names_x_univariant != "Cluster"]
          if(length(Factor_names_x) > 0){
            if("Sex" %in% colnames(Data_forest_tmp_x)){
              Data_forest_tmp_x = Data_forest_tmp_x %>%
                dplyr::mutate(Sex=case_when(
                  Sex == "女" ~ "Female",
                  TRUE ~ "Male"
                ))
            }
            colnames_tmp = colnames(Data_forest_tmp_x)
            colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
            colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
            colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
            colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
            colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
            colnames_tmp[colnames_tmp == "EP_option"] = "Treatment recommended"
            colnames_tmp[colnames_tmp == "PS"] = "Performance status"
            colnames_tmp[colnames_tmp == "Lines"] = "CTx line"
            Factor_names_x[Factor_names_x == "Lymph_met"] = "Lymphatic metastasis"
            Factor_names_x[Factor_names_x == "Liver_met"] = "Liver metastasis"
            Factor_names_x[Factor_names_x == "Lung_met"] = "Lung metastasis"
            Factor_names_x[Factor_names_x == "Bone_met"] = "Bone metastasis"
            Factor_names_x[Factor_names_x == "Brain_met"] = "Brain metastasis"
            Factor_names_x[Factor_names_x == "EP_option"] = "Treatment recommended"
            Factor_names_x[Factor_names_x == "PS"] = "Performance status"
            Factor_names_x[Factor_names_x == "Lines"] = "CTx line"
            Factor_names_x_univariant[Factor_names_x_univariant == "Lymph_met"] = "Lymphatic metastasis"
            Factor_names_x_univariant[Factor_names_x_univariant == "Liver_met"] = "Liver metastasis"
            Factor_names_x_univariant[Factor_names_x_univariant == "Lung_met"] = "Lung metastasis"
            Factor_names_x_univariant[Factor_names_x_univariant == "Bone_met"] = "Bone metastasis"
            Factor_names_x_univariant[Factor_names_x_univariant == "Brain_met"] = "Brain metastasis"
            Factor_names_x_univariant[Factor_names_x_univariant == "EP_option"] = "Treatment recommended"
            Factor_names_x_univariant[Factor_names_x_univariant == "PS"] = "Performance status"
            Factor_names_x_univariant[Factor_names_x_univariant == "Lines"] = "CTx line"
            colnames(Data_forest_tmp_x) = colnames_tmp
            
            Formula = as.formula(paste0("Surv(Drug_length, TTF_censor) ~ ",
                                        paste( paste0("`",Factor_names_x,"`"), collapse = "+")))
            linelistsurv_cox <-  coxph(formula = Formula,
                                       data = Data_forest_tmp_x,
                                       control = coxph.control(iter.max = 50))
            univ_tab <- Data_forest_tmp_x %>% 
              tbl_uvregression(
                method = coxph,
                y = Surv(time = Drug_length, event = TTF_censor),
                include = Factor_names_x_univariant,
                exponentiate = TRUE
              ) |>
              add_global_p() |> # add global p-value
              add_n(location = "level") |>
              add_nevent(location = "level") |> # add number of events of the outcome
              add_q() |> # adjusts global p-values for multiple testing
              bold_p() |> # bold p-values under a given threshold (default 0.05)
              modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              bold_labels()
             final_mv_reg <- linelistsurv_cox %>%
              stats::step(direction = "backward", trace = FALSE)
            if(!is.null(final_mv_reg$xlevels)){
              mv_tab = final_mv_reg |>
                tbl_regression(
                  exponentiate = TRUE
                ) |>
                add_global_p() |> # add global p-value
                add_q() |> # adjusts global p-values for multiple testing
                bold_p() |> # bold p-values under a given threshold (default 0.05)
                modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels()
              tbl_merge(
                tbls = list(univ_tab, mv_tab),
                tab_spanner = c("**Univariate**", "**Multivariable**")) |>
                modify_caption(paste0("Hazard ratio for Treatment discontinuation with ", paste(input$drug, collapse = ";"), " (Only factors with >2 events observed)")) |> as_gt()
            } else {
              univ_tab |>
                modify_caption(paste0("Hazard ratio for Treatment discontinuation with ", paste(input$drug, collapse = ";"), ", no significant factor in multivariable analysis (Only factors with >2 events observed)")) |> as_gt()
            }
          }
        })
  
        output$figure_drug_6_2 = render_gt({
          Factor_names_x = Factor_names_x[!Factor_names_x %in% gene_names]
          Factor_names_x_univariant = Factor_names_x_univariant[!Factor_names_x_univariant %in% gene_names]
          if(length(Factor_names_x) > 0){
            if("Sex" %in% colnames(Data_forest_tmp_x)){
              Data_forest_tmp_x = Data_forest_tmp_x %>%
                dplyr::mutate(Sex=case_when(
                  Sex == "女" ~ "Female",
                  TRUE ~ "Male"
                ))
            }
            colnames_tmp = colnames(Data_forest_tmp_x)
            
            
            colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
            colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
            colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
            colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
            colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
            colnames_tmp[colnames_tmp == "EP_option"] = "Treatment recommended"
            colnames_tmp[colnames_tmp == "PS"] = "Performance status"
            colnames_tmp[colnames_tmp == "Lines"] = "CTx line"
            Factor_names_x[Factor_names_x == "Lymph_met"] = "Lymphatic metastasis"
            Factor_names_x[Factor_names_x == "Liver_met"] = "Liver metastasis"
            Factor_names_x[Factor_names_x == "Lung_met"] = "Lung metastasis"
            Factor_names_x[Factor_names_x == "Bone_met"] = "Bone metastasis"
            Factor_names_x[Factor_names_x == "Brain_met"] = "Brain metastasis"
            Factor_names_x[Factor_names_x == "EP_option"] = "Treatment recommended"
            Factor_names_x[Factor_names_x == "PS"] = "Performance status"
            Factor_names_x[Factor_names_x == "Lines"] = "CTx line"
            Factor_names_x_univariant[Factor_names_x_univariant == "Lymph_met"] = "Lymphatic metastasis"
            Factor_names_x_univariant[Factor_names_x_univariant == "Liver_met"] = "Liver metastasis"
            Factor_names_x_univariant[Factor_names_x_univariant == "Lung_met"] = "Lung metastasis"
            Factor_names_x_univariant[Factor_names_x_univariant == "Bone_met"] = "Bone metastasis"
            Factor_names_x_univariant[Factor_names_x_univariant == "Brain_met"] = "Brain metastasis"
            Factor_names_x_univariant[Factor_names_x_univariant == "EP_option"] = "Treatment recommended"
            Factor_names_x_univariant[Factor_names_x_univariant == "PS"] = "Performance status"
            Factor_names_x_univariant[Factor_names_x_univariant == "Lines"] = "CTx line"
            colnames(Data_forest_tmp_x) = colnames_tmp
            Formula = as.formula(paste0("Surv(Drug_length, TTF_censor) ~ ",
                                        paste( paste0("`",Factor_names_x,"`"), collapse = "+")))
            linelistsurv_cox <-  coxph(formula = Formula,
                                       data = Data_forest_tmp_x,
                                       control = coxph.control(iter.max = 50))
            univ_tab <- Data_forest_tmp_x %>% 
              tbl_uvregression(
                method = coxph,
                y = Surv(time = Drug_length, event = TTF_censor),
                include = Factor_names_x_univariant,
                exponentiate = TRUE
              ) |>
              add_global_p() |> # add global p-value
              add_n(location = "level") |>
              add_nevent(location = "level") |> # add number of events of the outcome
              add_q() |> # adjusts global p-values for multiple testing
              bold_p() |> # bold p-values under a given threshold (default 0.05)
              modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              bold_labels()
            final_mv_reg <- linelistsurv_cox %>%
              stats::step(direction = "backward", trace = FALSE)
            if(!is.null(final_mv_reg$xlevels)){
              mv_tab = final_mv_reg |>
                tbl_regression(
                  exponentiate = TRUE
                ) |>
                add_global_p() |> # add global p-value
                add_q() |> # adjusts global p-values for multiple testing
                bold_p() |> # bold p-values under a given threshold (default 0.05)
                modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels()
              tbl_merge(
                tbls = list(univ_tab, mv_tab),
                tab_spanner = c("**Univariate**", "**Multivariable**")) |>
                modify_caption(paste0("Hazard ratio for Treatment discontinuation with ", paste(input$drug, collapse = ";"), " (Only factors with >2 events observed)")) |> as_gt()
            } else {
              univ_tab |>
                modify_caption(paste0("Hazard ratio for Treatment discontinuation with ", paste(input$drug, collapse = ";"), ", no significant factor in multivariable analysis (Only factors with >2 events observed)")) |> as_gt()
            }
          }
        })
        
        incProgress(1 / 13)
        
        Data_forest_tmp_1 = Data_forest  %>%
          dplyr::filter(ORR_data != -1) %>%
          dplyr::select(-ID, -Drug,
                        -TTF_censor, -Drug_length,
                        -最良総合効果, -DCR_data)
        Disease_tmp_1 = unique(sort(Data_forest_tmp_1$Histology))
        Disease_tmp_12 = unique(sort(Data_forest_tmp_1$Histology_dummy))
        Disease_tmp_1 = c(Disease_tmp_1[!Disease_tmp_1 %in% Disease_tmp_12], Disease_tmp_12)
        for(i in 1:length(Disease_tmp_1)){
          Data_forest_tmp_12 = Data_forest_tmp_1 %>% dplyr::filter(Histology == Disease_tmp_1[i])
          if(sum(Data_forest_tmp_12$ORR_data == 1,na.rm = T) < 3){
            Data_forest_tmp_1 = Data_forest_tmp_1 %>%
              dplyr::mutate(Histology = case_when(
                Histology == Disease_tmp_1[i] ~ Histology_dummy,
                TRUE ~ Histology
              )
              )
          } else {
            Data_forest_tmp_12 = Data_forest_tmp_1 %>% dplyr::filter(Histology == Disease_tmp_1[i])
            if(sum(Data_forest_tmp_12$ORR_data == 1,na.rm = T) < 3){
              Data_forest_tmp_1 = Data_forest_tmp_1 %>%
                dplyr::mutate(Histology = case_when(
                  Histology == Disease_tmp_1[i] ~ Histology_dummy,
                  TRUE ~ Histology
                )
                )
            }
          }
        }
        for(i in length(gene_names):1){
          if(sum(Data_forest_tmp_1[,12+col_added+i] == "mut(+)" & Data_forest_tmp_1$ORR_data == 1,na.rm = T) < 3){
            Data_forest_tmp_1 = Data_forest_tmp_1[,-(12+col_added+i)]
          } else if(sum(Data_forest_tmp_1[,12+col_added+i] == "mut(-)" & Data_forest_tmp_1$ORR_data == 1,na.rm = T) < 3){
            Data_forest_tmp_1 = Data_forest_tmp_1[,-(12+col_added+i)]
          }
        }
        if(sum(Data_forest_tmp_1$Age == "Younger" & Data_forest_tmp_1$ORR_data == 1,na.rm = T) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Age)
        } else if(sum(Data_forest_tmp_1$Age == "Older" & Data_forest_tmp_1$ORR_data == 1,na.rm = T) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Age)
        }
        if(sum(Data_forest_tmp_1$Sex == "男" & Data_forest_tmp_1$ORR_data == 1,na.rm = T) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Sex)
        } else if(sum(Data_forest_tmp_1$Sex != "男" & Data_forest_tmp_1$ORR_data == 1,na.rm = T) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Sex)
        }
        # if(sum(Data_forest_tmp_1$Smoking_history == "Yes" & Data_forest_tmp_1$ORR_data == 1,na.rm = T) < 3){
        #   Data_forest_tmp_1 = Data_forest_tmp_1 %>%
        #     dplyr::select(-Smoking_history)
        # } else if(sum(Data_forest_tmp_1$Smoking_history != "Yes" & Data_forest_tmp_1$ORR_data == 1,na.rm = T) < 3){
        #   Data_forest_tmp_1 = Data_forest_tmp_1 %>%
        #     dplyr::select(-Smoking_history)
        # }
        # if(sum(Data_forest_tmp_1$Alcoholic_history == "Yes" & Data_forest_tmp_1$ORR_data == 1,na.rm = T) < 3){
        #   Data_forest_tmp_1 = Data_forest_tmp_1 %>%
        #     dplyr::select(-Alcoholic_history)
        # } else if(sum(Data_forest_tmp_1$Alcoholic_history != "Yes" & Data_forest_tmp_1$ORR_data == 1,na.rm = T) < 3){
        #   Data_forest_tmp_1 = Data_forest_tmp_1 %>%
        #     dplyr::select(-Alcoholic_history)
        # }
        if(sum(Data_forest_tmp_1$Lines == "1" & Data_forest_tmp_1$ORR_data == 1) < 3 &
           sum(Data_forest_tmp_1$Lines == "2" & Data_forest_tmp_1$ORR_data == 1) < 3 &
           sum(Data_forest_tmp_1$Lines == "3~" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Lines')]
        } else if((sum(Data_forest_tmp_1$Lines == "1" & Data_forest_tmp_1$ORR_data == 1) >= 3 &
                   sum(Data_forest_tmp_1$Lines != "1" & Data_forest_tmp_1$ORR_data == 1) < 3) |
                  (sum(Data_forest_tmp_1$Lines == "2" & Data_forest_tmp_1$ORR_data == 1) >= 3 &
                   sum(Data_forest_tmp_1$Lines != "2" & Data_forest_tmp_1$ORR_data == 1) < 3) |
                  (sum(Data_forest_tmp_1$Lines == "3~" & Data_forest_tmp_1$ORR_data == 1) >= 3 &
                   sum(Data_forest_tmp_1$Lines != "3~" & Data_forest_tmp_1$ORR_data == 1) < 3)){
          Factor_names = Factor_names[!Factor_names %in% c('Lines')]
        } else if(sum(Data_forest_tmp_1$Lines == "1" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1$Lines[Data_forest_tmp_1$Lines == "1"] = "1~2"
          Data_forest_tmp_1$Lines[Data_forest_tmp_1$Lines == "2"] = "1~2"
        } else if(sum(Data_forest_tmp_1$Lines == "3~" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1$Lines[Data_forest_tmp_1$Lines == "3~"] = "2~"
          Data_forest_tmp_1$Lines[Data_forest_tmp_1$Lines == "2"] = "2~"
        }
        if(sum(Data_forest_tmp_1$Lymph_met == "Yes" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Lymph_met)
        } else if(sum(Data_forest_tmp_1$Lymph_met != "Yes" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Lymph_met)
        }
        if(sum(Data_forest_tmp_1$Lung_met == "Yes" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Lung_met)
        } else if(sum(Data_forest_tmp_1$Lung_met != "Yes" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Lung_met)
        }
        if(sum(Data_forest_tmp_1$Brain_met == "Yes" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Brain_met)
        } else if(sum(Data_forest_tmp_1$Brain_met != "Yes" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Brain_met)
        }
        if(sum(Data_forest_tmp_1$Bone_met == "Yes" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Bone_met)
        } else if(sum(Data_forest_tmp_1$Bone_met != "Yes" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Bone_met)
        }
        if(sum(Data_forest_tmp_1$Liver_met == "Yes" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Liver_met)
        } else if(sum(Data_forest_tmp_1$Liver_met != "Yes" & Data_forest_tmp_1$ORR_data == 1) < 3){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Liver_met)
        }
        if(input$HER2 != "No"){
          if(sum(Data_forest_tmp_1$HER2_IHC == "Positive" & Data_forest_tmp_1$ORR_data == 1) < 3){
            Data_forest_tmp_1 = Data_forest_tmp_1 %>%
              dplyr::select(-HER2_IHC)
          } else if(sum(Data_forest_tmp_1$HER2_IHC != "Positive" & Data_forest_tmp_1$ORR_data == 1) < 3){
            Data_forest_tmp_1 = Data_forest_tmp_1 %>%
              dplyr::select(-HER2_IHC)
          }
        }
        if(input$MSI != "No"){
          if(sum(Data_forest_tmp_1$MSI_PCR == "Positive" & Data_forest_tmp_1$ORR_data == 1) < 3){
            Data_forest_tmp_1 = Data_forest_tmp_1 %>%
              dplyr::select(-MSI_PCR)
          } else if(sum(Data_forest_tmp_1$MSI_PCR != "Positive" & Data_forest_tmp_1$ORR_data == 1) < 3){
            Data_forest_tmp_1 = Data_forest_tmp_1 %>%
              dplyr::select(-MSI_PCR)
          }
        }
        if(input$MMR != "No"){
          if(sum(Data_forest_tmp_1$MMR_IHC == "dMMR" & Data_forest_tmp_1$ORR_data == 1) < 3){
            Data_forest_tmp_1 = Data_forest_tmp_1 %>%
              dplyr::select(-MMR_IHC)
          } else if(sum(Data_forest_tmp_1$MMR_IHC != "dMMR" & Data_forest_tmp_1$ORR_data == 1) < 3){
            Data_forest_tmp_1 = Data_forest_tmp_1 %>%
              dplyr::select(-MMR_IHC)
          }
        }
        Data_forest_tmp_1$Histology = factor(Data_forest_tmp_1$Histology)
        if(length(unique(Data_forest_tmp_1$Histology))<2){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Histology)
        } else {
          Data_forest_tmp_1$Histology = relevel(Data_forest_tmp_1$Histology,
                                               ref=names(sort(table(Data_forest_tmp_1$Histology),decreasing = T))[[1]])
        }
        if(length(unique(Data_forest_tmp_1$Cluster))<2){
          Data_forest_tmp_1 = Data_forest_tmp_1 %>%
            dplyr::select(-Cluster)
        } else {
          Data_forest_tmp_1$Cluster = relevel(Data_forest_tmp_1$Cluster,
                                                ref=names(sort(table(Data_forest_tmp_1$Cluster),decreasing = T))[[1]])
        }
        Data_forest_tmp_1 = Data_forest_tmp_1 %>%
          dplyr::select(-Histology_dummy)
        if(length(colnames(Data_forest_tmp_1)) > 1){
          Factor_names_tmp_1 = colnames(Data_forest_tmp_1)
          for(i in length(Factor_names_tmp_1):1){
            if(all(Data_forest_tmp_1[,colnames(Data_forest_tmp_1) == Factor_names_tmp_1[i]] ==
                   Data_forest_tmp_1[1,colnames(Data_forest_tmp_1) == Factor_names_tmp_1[i]])){
              Data_forest_tmp_1 = Data_forest_tmp_1 %>%
                dplyr::select(-Factor_names_tmp_1[i])
            }
          }
        }
  
        Data_forest_tmp_1_univariant = Data_forest_tmp_1
        if(length(colnames(Data_forest_tmp_1)) > 1){
          Factor_names_tmp_1 = colnames(Data_forest_tmp_1)
          for(i in 1:(length(Factor_names_tmp_1) - 1)){
            for(j in (i+1):length(Factor_names_tmp_1)){
              if(Factor_names_tmp_1[i] %in% colnames(Data_forest_tmp_1) &
                 Factor_names_tmp_1[j] %in% colnames(Data_forest_tmp_1)){
                if(abs(cor(as.numeric(as.factor(Data_forest_tmp_1[,colnames(Data_forest_tmp_1) == Factor_names_tmp_1[i]])),
                        as.numeric(as.factor(Data_forest_tmp_1[,colnames(Data_forest_tmp_1) == Factor_names_tmp_1[j]])))) > 0.95){
                  Data_forest_tmp_1 = Data_forest_tmp_1 %>%
                    dplyr::select(-Factor_names_tmp_1[j])
                }
              }
            }
          }
        }
        Data_forest_tmp_7 = Data_forest_tmp_1
        Data_forest_tmp_7_univariant = Data_forest_tmp_1_univariant
        output$figure_drug_7_1 = render_gt({
          Data_forest_tmp_7 = Data_forest_tmp_7 %>%
            dplyr::select(-Cluster)
          Data_forest_tmp_7_univariant = Data_forest_tmp_7_univariant %>%
            dplyr::select(-Cluster)
          if(length(colnames(Data_forest_tmp_7)) > 1){
            Data_forest_tmp_7 = data.frame( lapply(Data_forest_tmp_7, as.factor) )
            Data_forest_tmp_7_univariant = data.frame( lapply(Data_forest_tmp_7_univariant, as.factor) )
            
            if("Histology" %in% colnames(Data_forest_tmp_7)){
              Data_forest_tmp_7$Histology <- relevel(Data_forest_tmp_7$Histology, ref=names(sort(table(Data_forest_tmp_7$Histology),decreasing = T))[[1]]) 
            }
            if("Histology" %in% colnames(Data_forest_tmp_7_univariant)){
              Data_forest_tmp_7_univariant$Histology <- relevel(Data_forest_tmp_7_univariant$Histology, ref=names(sort(table(Data_forest_tmp_7_univariant$Histology),decreasing = T))[[1]]) 
            }
            if("Sex" %in% colnames(Data_forest_tmp_7)){
              Data_forest_tmp_7 = Data_forest_tmp_7 %>%
                dplyr::mutate(Sex=case_when(
                  Sex == "女" ~ "Female",
                  TRUE ~ "Male"
                ))
            }
            if("Sex" %in% colnames(Data_forest_tmp_7_univariant)){
              Data_forest_tmp_7_univariant = Data_forest_tmp_7_univariant %>%
                dplyr::mutate(Sex=case_when(
                  Sex == "女" ~ "Female",
                  TRUE ~ "Male"
                ))
            }
            colnames_tmp = colnames(Data_forest_tmp_7)
            colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
            colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
            colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
            colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
            colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
            colnames_tmp[colnames_tmp == "PS"] = "Performance status"
            colnames_tmp[colnames_tmp == "Lines"] = "CTx line"
            colnames(Data_forest_tmp_7) = colnames_tmp
            colnames_tmp = colnames(Data_forest_tmp_7_univariant)
            colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
            colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
            colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
            colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
            colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
            colnames_tmp[colnames_tmp == "PS"] = "Performance status"
            colnames_tmp[colnames_tmp == "Lines"] = "CTx line"
            colnames(Data_forest_tmp_7_univariant) = colnames_tmp
            univ_tab <- Data_forest_tmp_7_univariant %>% 
              tbl_uvregression(                         ## 単変量解析の表を生成
                method = glm,                           ## 実行したい回帰（一般化線形モデル）を定義
                y = ORR_data,
                method.args = list(family = binomial),  ## 実行したい glm のタイプを定義（ここではロジスティック）
                exponentiate = TRUE                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
              ) |>
              add_global_p() |> # add global p-value
              add_n(location = "level") |>
              add_nevent(location = "level") |> # add number of events of the outcome
              add_q() |> # adjusts global p-values for multiple testing
              bold_p() |> # bold p-values under a given threshold (default 0.05)
              modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              bold_labels() 
            m1 <- glm(ORR_data ~., data = Data_forest_tmp_7, family = binomial(link = "logit"))
            final_mv_reg <- m1 %>%
              stats::step(direction = "backward", trace = FALSE)
            if(!is.null(final_mv_reg$xlevels)){
              mv_tab = final_mv_reg |>
                tbl_regression(                         ## 多変量解析の表を生成
                  exponentiate = TRUE                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
                ) |>
                add_global_p() |> # add global p-value
                add_q() |> # adjusts global p-values for multiple testing
                bold_p() |> # bold p-values under a given threshold (default 0.05)
                 modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels()
              tbl_merge(
                tbls = list(univ_tab, mv_tab),
                tab_spanner = c("**Univariate**", "**Multivariable**")) |>
                modify_caption(paste("Factors for objective response rate, ", paste(input$drug, collapse = ";"), "multivariable regression (Only patients with >2 events observed)")) |> as_gt()
            } else {
              univ_tab |>
                modify_caption(paste("Factors for objective response rate,", paste(input$drug, collapse = ";"), "(Only patients with >2 events observed), no significant factor in multivariable analysis")) |> as_gt()
            }
          }
        })

        output$figure_drug_7_2 = render_gt({
          Data_forest_tmp_7 = Data_forest_tmp_7[,!colnames(Data_forest_tmp_7) %in% gene_names]
          Data_forest_tmp_7_univariant = Data_forest_tmp_7_univariant[,!colnames(Data_forest_tmp_7_univariant) %in% gene_names]
          if(length(colnames(Data_forest_tmp_7)) > 1){
            Data_forest_tmp_7 = data.frame( lapply(Data_forest_tmp_7, as.factor) )
            Data_forest_tmp_7_univariant = data.frame( lapply(Data_forest_tmp_7_univariant, as.factor) )
            
            if("Histology" %in% colnames(Data_forest_tmp_7)){
              Data_forest_tmp_7$Histology <- relevel(Data_forest_tmp_7$Histology, ref=names(sort(table(Data_forest_tmp_7$Histology),decreasing = T))[[1]]) 
            }
            if("Histology" %in% colnames(Data_forest_tmp_7_univariant)){
              Data_forest_tmp_7_univariant$Histology <- relevel(Data_forest_tmp_7_univariant$Histology, ref=names(sort(table(Data_forest_tmp_7_univariant$Histology),decreasing = T))[[1]]) 
            }
            if("Sex" %in% colnames(Data_forest_tmp_7)){
              Data_forest_tmp_7 = Data_forest_tmp_7 %>%
                dplyr::mutate(Sex=case_when(
                  Sex == "女" ~ "Female",
                  TRUE ~ "Male"
                ))
            }
            if("Sex" %in% colnames(Data_forest_tmp_7_univariant)){
              Data_forest_tmp_7_univariant = Data_forest_tmp_7_univariant %>%
                dplyr::mutate(Sex=case_when(
                  Sex == "女" ~ "Female",
                  TRUE ~ "Male"
                ))
            }
            colnames_tmp = colnames(Data_forest_tmp_7)
            colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
            colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
            colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
            colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
            colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
            colnames_tmp[colnames_tmp == "PS"] = "Performance status"
            colnames_tmp[colnames_tmp == "Lines"] = "CTx line"
            colnames(Data_forest_tmp_7) = colnames_tmp
            colnames_tmp = colnames(Data_forest_tmp_7_univariant)
            colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
            colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
            colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
            colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
            colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
            colnames_tmp[colnames_tmp == "PS"] = "Performance status"
            colnames_tmp[colnames_tmp == "Lines"] = "CTx line"
            colnames(Data_forest_tmp_7_univariant) = colnames_tmp
            univ_tab <- Data_forest_tmp_7_univariant %>% 
              tbl_uvregression(                         ## 単変量解析の表を生成
                method = glm,                           ## 実行したい回帰（一般化線形モデル）を定義
                y = ORR_data,
                method.args = list(family = binomial),  ## 実行したい glm のタイプを定義（ここではロジスティック）
                exponentiate = TRUE                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
              ) |>
              add_global_p() |> # add global p-value
              add_n(location = "level") |>
              add_nevent(location = "level") |> # add number of events of the outcome
              add_q() |> # adjusts global p-values for multiple testing
              bold_p() |> # bold p-values under a given threshold (default 0.05)
              modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              bold_labels() 
            m1 <- glm(ORR_data ~., data = Data_forest_tmp_7, family = binomial(link = "logit"))
            final_mv_reg <- m1 %>%
              stats::step(direction = "backward", trace = FALSE)
            if(!is.null(final_mv_reg$xlevels)){
              mv_tab = final_mv_reg |>
                tbl_regression(                         ## 多変量解析の表を生成
                  exponentiate = TRUE                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
                ) |>
                add_global_p() |> # add global p-value
                add_q() |> # adjusts global p-values for multiple testing
                bold_p() |> # bold p-values under a given threshold (default 0.05)
                modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels()
              tbl_merge(
                tbls = list(univ_tab, mv_tab),
                tab_spanner = c("**Univariate**", "**Multivariable**")) |>
                modify_caption(paste("Factors for objective response rate, ", paste(input$drug, collapse = ";"), "multivariable regression (Only patients with >2 events observed)")) |> as_gt()
            } else {
              univ_tab |>
                modify_caption(paste("Factors for objective response rate,", paste(input$drug, collapse = ";"), "(Only patients with >2 events observed), no significant factor in multivariable analysis")) |> as_gt()
            }
          }
        })

        incProgress(1 / 13)
  
        Data_forest_tmp_2 = Data_forest  %>%
          dplyr::filter(DCR_data != -1) %>%
          dplyr::select(-ID, -Drug,
                        -TTF_censor, -Drug_length,
                        -最良総合効果, -ORR_data, -Cluster)
        Disease_tmp_2 = unique(sort(Data_forest_tmp_2$Histology))
        Disease_tmp_22 = unique(sort(Data_forest_tmp_2$Histology_dummy))
        Disease_tmp_2 = c(Disease_tmp_2[!Disease_tmp_2 %in% Disease_tmp_22], Disease_tmp_22)
        for(i in 1:length(Disease_tmp_2)){
          Data_forest_tmp_22 = Data_forest_tmp_2 %>% dplyr::filter(Histology == Disease_tmp_2[i])
          if(sum(Data_forest_tmp_22$DCR_data == 1,na.rm = T) < 3){
            Data_forest_tmp_2 = Data_forest_tmp_2 %>%
              dplyr::mutate(Histology = case_when(
                Histology == Disease_tmp_2[i] ~ Histology_dummy,
                TRUE ~ Histology
              )
              )
          } else{
            Data_forest_tmp_22 = Data_forest_tmp_2 %>% dplyr::filter(Histology != Disease_tmp_2[i])
            if(sum(Data_forest_tmp_22$DCR_data == 1,na.rm = T) < 3){
              Data_forest_tmp_2 = Data_forest_tmp_2 %>%
                dplyr::mutate(Histology = case_when(
                  Histology == Disease_tmp_2[i] ~ Histology_dummy,
                  TRUE ~ Histology
                )
                )
            }
          }
        }
        for(i in length(gene_names):1){
          if(sum(Data_forest_tmp_2[,11+col_added+i] == "mut(+)" & Data_forest_tmp_2$DCR_data == 1,na.rm = T) < 3){
            Data_forest_tmp_2 = Data_forest_tmp_2[,-(11+col_added+i)]
          } else if(sum(Data_forest_tmp_2[,11+col_added+i] == "mut(-)" & Data_forest_tmp_2$DCR_data == 1,na.rm = T) < 3){
            Data_forest_tmp_2 = Data_forest_tmp_2[,-(11+col_added+i)]
          }
        }
        if(sum(Data_forest_tmp_2$Age == "Younger" & Data_forest_tmp_2$DCR_data == 1,na.rm = T) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Age)
        } else if(sum(Data_forest_tmp_2$Age == "Older" & Data_forest_tmp_2$DCR_data == 1,na.rm = T) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Age)
        }
        if(sum(Data_forest_tmp_2$Sex == "男" & Data_forest_tmp_2$DCR_data == 1,na.rm = T) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Sex)
        } else if(sum(Data_forest_tmp_2$Sex != "男" & Data_forest_tmp_2$DCR_data == 1,na.rm = T) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Sex)
        }
        if(sum(Data_forest_tmp_2$Lines == "1" & Data_forest_tmp_2$DCR_data == 1) < 3 &
           sum(Data_forest_tmp_2$Lines == "2" & Data_forest_tmp_2$DCR_data == 1) < 3 &
           sum(Data_forest_tmp_2$Lines == "3~" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Factor_names = Factor_names[!Factor_names %in% c('Lines')]
        } else if((sum(Data_forest_tmp_2$Lines == "1" & Data_forest_tmp_2$DCR_data == 1) >= 3 &
                   sum(Data_forest_tmp_2$Lines != "1" & Data_forest_tmp_2$DCR_data == 1) < 3) |
                  (sum(Data_forest_tmp_2$Lines == "2" & Data_forest_tmp_2$DCR_data == 1) >= 3 &
                   sum(Data_forest_tmp_2$Lines != "2" & Data_forest_tmp_2$DCR_data == 1) < 3) |
                  (sum(Data_forest_tmp_2$Lines == "3~" & Data_forest_tmp_2$DCR_data == 1) >= 3 &
                   sum(Data_forest_tmp_2$Lines != "3~" & Data_forest_tmp_2$DCR_data == 1) < 3)){
          Factor_names = Factor_names[!Factor_names %in% c('Lines')]
        } else if(sum(Data_forest_tmp_2$Lines == "1" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2$Lines[Data_forest_tmp_2$Lines == "1"] = "1~2"
          Data_forest_tmp_2$Lines[Data_forest_tmp_2$Lines == "2"] = "1~2"
        } else if(sum(Data_forest_tmp_2$Lines == "3~" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2$Lines[Data_forest_tmp_2$Lines == "3~"] = "2~"
          Data_forest_tmp_2$Lines[Data_forest_tmp_2$Lines == "2"] = "2~"
        }
        if(sum(Data_forest_tmp_2$Lymph_met == "Yes" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Lymph_met)
        } else if(sum(Data_forest_tmp_2$Lymph_met != "Yes" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Lymph_met)
        }
        if(sum(Data_forest_tmp_2$Lung_met == "Yes" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Lung_met)
        } else if(sum(Data_forest_tmp_2$Lung_met != "Yes" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Lung_met)
        }
        if(sum(Data_forest_tmp_2$Brain_met == "Yes" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Brain_met)
        } else if(sum(Data_forest_tmp_2$Brain_met != "Yes" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Brain_met)
        }
        if(sum(Data_forest_tmp_2$Bone_met == "Yes" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Bone_met)
        } else if(sum(Data_forest_tmp_2$Bone_met != "Yes" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Bone_met)
        }
        if(sum(Data_forest_tmp_2$Liver_met == "Yes" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Liver_met)
        } else if(sum(Data_forest_tmp_2$Liver_met != "Yes" & Data_forest_tmp_2$DCR_data == 1) < 3){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Liver_met)
        }
        if(input$HER2 != "No"){
          if(sum(Data_forest_tmp_2$HER2_IHC == "Positive" & Data_forest_tmp_2$DCR_data == 1) < 3){
            Data_forest_tmp_2 = Data_forest_tmp_2 %>%
              dplyr::select(-HER2_IHC)
          } else if(sum(Data_forest_tmp_2$HER2_IHC != "Positive" & Data_forest_tmp_2$DCR_data == 1) < 3){
            Data_forest_tmp_2 = Data_forest_tmp_2 %>%
              dplyr::select(-HER2_IHC)
          }
        }
        if(input$MSI != "No"){
          if(sum(Data_forest_tmp_2$MSI_PCR == "Positive" & Data_forest_tmp_2$DCR_data == 1) < 3){
            Data_forest_tmp_2 = Data_forest_tmp_2 %>%
              dplyr::select(-MSI_PCR)
          } else if(sum(Data_forest_tmp_2$MSI_PCR != "Positive" & Data_forest_tmp_2$DCR_data == 1) < 3){
            Data_forest_tmp_2 = Data_forest_tmp_2 %>%
              dplyr::select(-MSI_PCR)
          }
        }
        if(input$MMR != "No"){
          if(sum(Data_forest_tmp_2$MMR_IHC == "dMMR" & Data_forest_tmp_2$DCR_data == 1) < 3){
            Data_forest_tmp_2 = Data_forest_tmp_2 %>%
              dplyr::select(-MMR_IHC)
          } else if(sum(Data_forest_tmp_2$MMR_IHC != "dMMR" & Data_forest_tmp_2$DCR_data == 1) < 3){
            Data_forest_tmp_2 = Data_forest_tmp_2 %>%
              dplyr::select(-MMR_IHC)
          }
        }
        Data_forest_tmp_2$Histology = factor(Data_forest_tmp_2$Histology)
        if(length(unique(Data_forest_tmp_2$Histology))<2){
          Data_forest_tmp_2 = Data_forest_tmp_2 %>%
            dplyr::select(-Histology)
        }
        Data_forest_tmp_2 = Data_forest_tmp_2 %>%
          dplyr::select(-Histology_dummy)
        if(length(colnames(Data_forest_tmp_2)) > 1){
          Factor_names_tmp_2 = colnames(Data_forest_tmp_2)
          for(i in length(Factor_names_tmp_2):1){
            if(all(Data_forest_tmp_2[,colnames(Data_forest_tmp_2) == Factor_names_tmp_2[i]] ==
                   Data_forest_tmp_2[1,colnames(Data_forest_tmp_2) == Factor_names_tmp_2[i]])){
              Data_forest_tmp_2 = Data_forest_tmp_2 %>%
                dplyr::select(-Factor_names_tmp_2[i])
            }
          }
        }
        Data_forest_tmp_2_univariant = Data_forest_tmp_2
        if(length(colnames(Data_forest_tmp_2)) > 1){
          Factor_names_tmp_2 = colnames(Data_forest_tmp_2)
          for(i in 1:(length(Factor_names_tmp_2) - 1)){
            for(j in (i+1):length(Factor_names_tmp_2)){
              if(Factor_names_tmp_2[i] %in% colnames(Data_forest_tmp_2) &
                 Factor_names_tmp_2[j] %in% colnames(Data_forest_tmp_2)){
                if(abs(cor(as.numeric(as.factor(Data_forest_tmp_2[,colnames(Data_forest_tmp_2) == Factor_names_tmp_2[i]])),
                           as.numeric(as.factor(Data_forest_tmp_2[,colnames(Data_forest_tmp_2) == Factor_names_tmp_2[j]])))) > 0.95){
                  Data_forest_tmp_2 = Data_forest_tmp_2 %>%
                    dplyr::select(-Factor_names_tmp_2[j])
                }
              }
            }
          }
        }
  
        Data_forest_tmp_8 = Data_forest_tmp_2
        Data_forest_tmp_8_univariant = Data_forest_tmp_2_univariant
        output$figure_drug_8 = render_gt({
          if(length(colnames(Data_forest_tmp_8)) > 1){
            Data_forest_tmp_8 = data.frame( lapply(Data_forest_tmp_8, as.factor) )
            Data_forest_tmp_8_univariant = data.frame( lapply(Data_forest_tmp_8_univariant, as.factor) )
            if("Histology" %in% colnames(Data_forest_tmp_8)){
              Data_forest_tmp_8$Histology <- relevel(Data_forest_tmp_8$Histology, ref=names(sort(table(Data_forest_tmp_8$Histology),decreasing = T))[[1]]) 
            }
            if("Histology" %in% colnames(Data_forest_tmp_8_univariant)){
              Data_forest_tmp_8_univariant$Histology <- relevel(Data_forest_tmp_8_univariant$Histology, ref=names(sort(table(Data_forest_tmp_8_univariant$Histology),decreasing = T))[[1]]) 
            }
            if("Sex" %in% colnames(Data_forest_tmp_8)){
              Data_forest_tmp_8 = Data_forest_tmp_8 %>%
                dplyr::mutate(Sex=case_when(
                  Sex == "女" ~ "Female",
                  TRUE ~ "Male"
                ))
            }
            if("Sex" %in% colnames(Data_forest_tmp_8_univariant)){
              Data_forest_tmp_8_univariant = Data_forest_tmp_8_univariant %>%
                dplyr::mutate(Sex=case_when(
                  Sex == "女" ~ "Female",
                  TRUE ~ "Male"
                ))
            }
            colnames_tmp = colnames(Data_forest_tmp_8)
            colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
            colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
            colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
            colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
            colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
            colnames_tmp[colnames_tmp == "PS"] = "Performance status"
            colnames_tmp[colnames_tmp == "Lines"] = "CTx line"
            colnames(Data_forest_tmp_8) = colnames_tmp
            colnames_tmp = colnames(Data_forest_tmp_8_univariant)
            colnames_tmp[colnames_tmp == "Lymph_met"] = "Lymphatic metastasis"
            colnames_tmp[colnames_tmp == "Liver_met"] = "Liver metastasis"
            colnames_tmp[colnames_tmp == "Lung_met"] = "Lung metastasis"
            colnames_tmp[colnames_tmp == "Bone_met"] = "Bone metastasis"
            colnames_tmp[colnames_tmp == "Brain_met"] = "Brain metastasis"
            colnames_tmp[colnames_tmp == "PS"] = "Performance status"
            colnames_tmp[colnames_tmp == "Lines"] = "CTx line"
            colnames(Data_forest_tmp_8_univariant) = colnames_tmp
            univ_tab <- Data_forest_tmp_8_univariant %>% 
              tbl_uvregression(                         ## 単変量解析の表を生成
                method = glm,                           ## 実行したい回帰（一般化線形モデル）を定義
                y = DCR_data,
                method.args = list(family = binomial),  ## 実行したい glm のタイプを定義（ここではロジスティック）
                exponentiate = TRUE                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
              ) |>
              add_global_p() |> # add global p-value
              add_n(location = "level") |>
              add_nevent(location = "level") |> # add number of events of the outcome
              add_q() |> # adjusts global p-values for multiple testing
              bold_p() |> # bold p-values under a given threshold (default 0.05)
               modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
              bold_labels() 
            m1 <- glm(DCR_data ~., data = Data_forest_tmp_8, family = binomial(link = "logit"))
            final_mv_reg <- m1 %>%
              stats::step(direction = "backward", trace = FALSE)
            if(!is.null(final_mv_reg$xlevels)){
              mv_tab = final_mv_reg |>
                tbl_regression(                         ## 多変量解析の表を生成
                  exponentiate = TRUE                     ## 対数オッズ比ではなくオッズ比を得るために指数変換を指定
                ) |>
                add_global_p() |> # add global p-value
                add_q() |> # adjusts global p-values for multiple testing
                bold_p() |> # bold p-values under a given threshold (default 0.05)
                modify_fmt_fun(conf.low ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(conf.high ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                modify_fmt_fun(estimate ~ function(x) ifelse(Abs(x)>100 | Abs(x)<0.01, format(x, digits = 3, scientific = TRUE), ifelse(Abs(x) != 1, format(x, digits = 3), x))) |>
                bold_labels()
              tbl_merge(
                tbls = list(univ_tab, mv_tab),
                tab_spanner = c("**Univariate**", "**Multivariable**")) |>
                modify_caption(paste("Factors for disease control rate, ", paste(input$drug, collapse = ";"), "multivariable regression (Only patients with >2 events observed)")) |> as_gt()
            } else {
              univ_tab |>
                modify_caption(paste("Factors for disease control rate,", paste(input$drug, collapse = ";"), "(Only patients with >2 events observed), no significant factor in multivariable analysis")) |> as_gt()
            }
          }
        })
        incProgress(1 / 13)
      }
      
      if(input$drug_volcano != "Yes"){
        if(!input$special_gene == ""){
          Data_MAF_target_tmp = Data_MAF_target_tmp %>%
            dplyr::filter(Tumor_Sample_Barcode %in%
                            Data_case_target$C.CAT調査結果.基本項目.ハッシュID) %>%
            dplyr::arrange(Tumor_Sample_Barcode, Hugo_Symbol, amino.acid.change)
          
          ID_special_gene_ = (Data_MAF_target_tmp %>%
                               dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_"))))$Tumor_Sample_Barcode
          ID_special_gene_mutation_1_ = (Data_MAF_target_tmp %>%
                                          dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                          amino.acid.change %in% input$special_gene_mutation_1))$Tumor_Sample_Barcode
          ID_special_gene_mutation_2_ = (Data_MAF_target_tmp %>%
                                          dplyr::filter(Hugo_Symbol == input$special_gene | str_detect(Hugo_Symbol, paste0(input$special_gene, "_")) &
                                                          amino.acid.change %in% input$special_gene_mutation_2))$Tumor_Sample_Barcode
          Data_forest_all_pts = Data_forest_all_pts %>% dplyr::mutate(
            separation_value = case_when(
              ID %in% ID_special_gene_mutation_1_ & ID %in% ID_special_gene_mutation_2_ ~
                paste0(input$special_gene_mutation_1_name, " & ", input$special_gene_mutation_2_name, " in ", input$special_gene),
              ID %in% ID_special_gene_mutation_1_ ~ paste0(input$special_gene_mutation_1_name, " in ", input$special_gene),
              ID %in% ID_special_gene_mutation_2_ ~ paste0(input$special_gene_mutation_2_name, " in ", input$special_gene),
              ID %in% ID_special_gene_ ~ paste0("Other mut in ", input$special_gene),
              TRUE ~ paste0("No mut in ", input$special_gene)
            )
          )
        } else{
          Data_forest_all_pts = Data_forest_all_pts %>% dplyr::mutate(
            separation_value = "No mutation pattern")
        }
        Table_outcome_pattern_total = NULL
        drug_list = list(input$drug, input$drug_group_1, input$drug_group_2, input$drug_group_3, input$drug_group_4)
        for(drug_count in 1:5){
          drug_name_2 = case_when(
            drug_count == 1 ~ paste(input$drug, collapse = ";"),
            drug_count == 2 ~ input$drug_group_1_name,
            drug_count == 3 ~ input$drug_group_2_name,
            drug_count == 4 ~ input$drug_group_3_name,
            drug_count == 5 ~ input$drug_group_4_name
          )
          Data_summary_pattern = Data_forest_all_pts %>% dplyr::filter(!最良総合効果 %in% c("NE", "Unknown")) %>% dplyr::filter(Drug %in% drug_list[[drug_count]]) %>% dplyr::arrange(-ORR_data, -DCR_data) %>% dplyr::distinct(ID, .keep_all = T)
          if(length(Data_summary_pattern$Drug)>0){
            Table_outcome_pattern = data.frame(c("Total", as.character((Data_summary_pattern %>% group_by(separation_value) %>% tally())$separation_value)))
            colnames(Table_outcome_pattern) = "Pattern"
            Table_outcome_pattern$Drug = drug_name_2
            Table_outcome_pattern$N = 0
            Table_outcome_pattern$OR = 0
            Table_outcome_pattern$ORR = 0
            Table_outcome_pattern$ORR_95CI_LL = 0
            Table_outcome_pattern$ORR_95CI_UL = 0
            Table_outcome_pattern$DC = 0
            Table_outcome_pattern$DCR = 0
            Table_outcome_pattern$DCR_95CI_LL = 0
            Table_outcome_pattern$DCR_95CI_UL = 0
            Table_outcome_pattern$OR_OddsRatio = 1
            Table_outcome_pattern$OR_Odds_pValue = 1
            Table_outcome_pattern$DC_OddsRatio = 1
            Table_outcome_pattern$DC_Odds_pValue = 1
            
            tmp2 = Data_summary_pattern %>% group_by(separation_value) %>% tally()
            Table_outcome_pattern$N[1] = sum(tmp2$n)
            if(length(Table_outcome_pattern$Pattern)>1){
              for(i in 1:(length(Table_outcome_pattern$Pattern)-1)){
                Table_outcome_pattern$N[i+1] = tmp2$n[i]
              }
              
              tmp2 = Data_summary_pattern %>% group_by(separation_value, ORR_data) %>% tally()
              Table_outcome_pattern$OR[1] = sum((tmp2 %>% dplyr::filter(ORR_data == 1))$n)
              for(i in 1:(length(Table_outcome_pattern$Pattern)-1)){
                if(length((tmp2 %>% dplyr::filter(ORR_data == 1 & separation_value == Table_outcome_pattern$Pattern[i+1]))$n)>0){
                  Table_outcome_pattern$OR[i+1] = (tmp2 %>% dplyr::filter(ORR_data == 1 & separation_value == Table_outcome_pattern$Pattern[i+1]))$n
                }
              }
              tmp2 = Data_summary_pattern %>% group_by(separation_value, DCR_data) %>% tally()
              Table_outcome_pattern$DC[1] = sum((tmp2 %>% dplyr::filter(DCR_data == 1))$n)
              for(i in 1:(length(Table_outcome_pattern$Pattern)-1)){
                if(length((tmp2 %>% dplyr::filter(DCR_data == 1 & separation_value == Table_outcome_pattern$Pattern[i+1]))$n)>0){
                  Table_outcome_pattern$DC[i+1] = (tmp2 %>% dplyr::filter(DCR_data == 1 & separation_value == Table_outcome_pattern$Pattern[i+1]))$n
                }
              }
              
              for(i in 1:length(Table_outcome_pattern$Pattern)){
                Table_outcome_pattern$ORR[i] = round(Table_outcome_pattern$OR[i] / Table_outcome_pattern$N[i], digits = 3)
                Table_outcome_pattern$DCR[i] = round(Table_outcome_pattern$DC[i] / Table_outcome_pattern$N[i], digits = 3)
                Table_outcome_pattern$ORR_95CI_LL[i] = round(exactci(Table_outcome_pattern$OR[i], Table_outcome_pattern$N[i], conf.level=0.95)$conf.int[1], digits = 3)
                Table_outcome_pattern$ORR_95CI_UL[i] = round(exactci(Table_outcome_pattern$OR[i], Table_outcome_pattern$N[i], conf.level=0.95)$conf.int[2], digits = 3)
                Table_outcome_pattern$DCR_95CI_LL[i] = round(exactci(Table_outcome_pattern$DC[i], Table_outcome_pattern$N[i], conf.level=0.95)$conf.int[1], digits = 3)
                Table_outcome_pattern$DCR_95CI_UL[i] = round(exactci(Table_outcome_pattern$DC[i], Table_outcome_pattern$N[i], conf.level=0.95)$conf.int[2], digits = 3)
                test_result <- fisher.test(matrix(c(Table_outcome_pattern$OR[i],
                                                    Table_outcome_pattern$N[i] - Table_outcome_pattern$OR[i],
                                                    Table_outcome_pattern$OR[1] - Table_outcome_pattern$OR[i],
                                                    Table_outcome_pattern$N[1] - Table_outcome_pattern$N[i] -Table_outcome_pattern$OR[1] + Table_outcome_pattern$OR[i]), nrow = 2, byrow = TRUE))
                Table_outcome_pattern$OR_OddsRatio[i] = round(min(unname(test_result$estimate), 99999), digits = 3)
                Table_outcome_pattern$OR_Odds_pValue[i] = round(unname(test_result$p.value), digits = 3)
                test_result <- fisher.test(matrix(c(Table_outcome_pattern$DC[i],
                                                    Table_outcome_pattern$N[i] - Table_outcome_pattern$DC[i],
                                                    Table_outcome_pattern$DC[1] - Table_outcome_pattern$DC[i],
                                                    Table_outcome_pattern$N[1] - Table_outcome_pattern$N[i] -Table_outcome_pattern$DC[1] + Table_outcome_pattern$DC[i]), nrow = 2, byrow = TRUE))
                Table_outcome_pattern$DC_OddsRatio[i] = round(min(unname(test_result$estimate), 99999), digits = 3)
                Table_outcome_pattern$DC_Odds_pValue[i] = round(unname(test_result$p.value), digits = 3)
              }
              Table_outcome_pattern_total = rbind(Table_outcome_pattern_total, Table_outcome_pattern)
            }
          }
        }
        output$table_outcome_1 = DT::renderDataTable(Table_outcome_pattern_total,
                                                     server = FALSE,
                                                     filter = 'top', 
                                                     extensions = c('Buttons'), 
                                                     options = list(pageLength = 100, 
                                                                    scrollX = TRUE,
                                                                    scrollY = "1000px",
                                                                    scrollCollapse = TRUE,
                                                                    dom="Blfrtip",
                                                                    buttons = c('csv', 'copy')))  
        
        Data_forest_all_pts$Cluster = ""
          for(i in unique(Data_case_target$cluster)){
            ID_mutation = (Data_case_target %>% dplyr::filter(cluster == i))$C.CAT調査結果.基本項目.ハッシュID
            if(length(Data_forest_all_pts[Data_forest_all_pts$ID %in% ID_mutation,]$Cluster)>0){
              Data_forest_all_pts[Data_forest_all_pts$ID %in% ID_mutation,]$Cluster = paste0("Cluster ", i)
            }
          }
          Table_outcome_cluster_total = NULL
          drug_list = list(input$drug, input$drug_group_1, input$drug_group_2, input$drug_group_3, input$drug_group_4)
          for(drug_count in 1:5){
            drug_name_1 = case_when(
              drug_count == 1 ~ paste(input$drug, collapse = ";"),
              drug_count == 2 ~ input$drug_group_1_name,
              drug_count == 3 ~ input$drug_group_2_name,
              drug_count == 4 ~ input$drug_group_3_name,
              drug_count == 5 ~ input$drug_group_4_name
            )
            Data_summary_cluster = Data_forest_all_pts %>% dplyr::filter(!最良総合効果 %in% c("NE", "Unknown")) %>% dplyr::filter(Drug %in% drug_list[[drug_count]]) %>% dplyr::arrange(-ORR_data, -DCR_data) %>% dplyr::distinct(ID, .keep_all = T)
            if(length(Data_summary_cluster$Drug)>0){
              Table_outcome_cluster = data.frame(c("Total", as.character((Data_summary_cluster %>% group_by(Cluster) %>% tally())$Cluster)))
              colnames(Table_outcome_cluster) = "Cluster"
              Table_outcome_cluster$Drug = drug_name_1
              Table_outcome_cluster$N = 0
              Table_outcome_cluster$OR = 0
              Table_outcome_cluster$ORR = 0
              Table_outcome_cluster$ORR_95CI_LL = 0
              Table_outcome_cluster$ORR_95CI_UL = 0
              Table_outcome_cluster$DC = 0
              Table_outcome_cluster$DCR = 0
              Table_outcome_cluster$DCR_95CI_LL = 0
              Table_outcome_cluster$DCR_95CI_UL = 0
              Table_outcome_cluster$OR_OddsRatio = 1
              Table_outcome_cluster$OR_Odds_pValue = 1
              Table_outcome_cluster$DC_OddsRatio = 1
              Table_outcome_cluster$DC_Odds_pValue = 1
              tmp = Data_summary_cluster %>% group_by(Cluster) %>% tally()
              Table_outcome_cluster$N[1] = sum(tmp$n)
              if(sum(tmp$n)>0){
                for(i in 1:(length(Table_outcome_cluster$Cluster)-1)){
                  Table_outcome_cluster$N[i+1] = tmp$n[i]
                }
                
                tmp = Data_summary_cluster %>% group_by(Cluster, ORR_data) %>% tally()
                Table_outcome_cluster$OR[1] = sum((tmp %>% dplyr::filter(ORR_data == 1))$n)
                for(i in 1:(length(Table_outcome_cluster$Cluster)-1)){
                  if(length((tmp %>% dplyr::filter(ORR_data == 1 & Cluster == Table_outcome_cluster$Cluster[i+1]))$n)>0){
                    Table_outcome_cluster$OR[i+1] = (tmp %>% dplyr::filter(ORR_data == 1 & Cluster == Table_outcome_cluster$Cluster[i+1]))$n
                  }
                }
                tmp = Data_summary_cluster %>% group_by(Cluster, DCR_data) %>% tally()
                Table_outcome_cluster$DC[1] = sum((tmp %>% dplyr::filter(DCR_data == 1))$n)
                for(i in 1:(length(Table_outcome_cluster$Cluster)-1)){
                  if(length((tmp %>% dplyr::filter(DCR_data == 1 & Cluster == Table_outcome_cluster$Cluster[i+1]))$n)>0){
                    Table_outcome_cluster$DC[i+1] = (tmp %>% dplyr::filter(DCR_data == 1 & Cluster == Table_outcome_cluster$Cluster[i+1]))$n
                  }
                }
                
                for(i in 1:length(Table_outcome_cluster$Cluster)){
                  Table_outcome_cluster$ORR[i] = round(Table_outcome_cluster$OR[i] / Table_outcome_cluster$N[i], digits = 3)
                  Table_outcome_cluster$DCR[i] = round(Table_outcome_cluster$DC[i] / Table_outcome_cluster$N[i], digits = 3)
                  Table_outcome_cluster$ORR_95CI_LL[i] = round(exactci(Table_outcome_cluster$OR[i], Table_outcome_cluster$N[i], conf.level=0.95)$conf.int[1], digits = 3)
                  Table_outcome_cluster$ORR_95CI_UL[i] = round(exactci(Table_outcome_cluster$OR[i], Table_outcome_cluster$N[i], conf.level=0.95)$conf.int[2], digits = 3)
                  Table_outcome_cluster$DCR_95CI_LL[i] = round(exactci(Table_outcome_cluster$DC[i], Table_outcome_cluster$N[i], conf.level=0.95)$conf.int[1], digits = 3)
                  Table_outcome_cluster$DCR_95CI_UL[i] = round(exactci(Table_outcome_cluster$DC[i], Table_outcome_cluster$N[i], conf.level=0.95)$conf.int[2], digits = 3)
                  test_result <- fisher.test(matrix(c(Table_outcome_cluster$OR[i],
                                                      Table_outcome_cluster$N[i] - Table_outcome_cluster$OR[i],
                                                      Table_outcome_cluster$OR[1] - Table_outcome_cluster$OR[i],
                                                      Table_outcome_cluster$N[1] - Table_outcome_cluster$N[i] -Table_outcome_cluster$OR[1] + Table_outcome_cluster$OR[i]), nrow = 2, byrow = TRUE))
                  Table_outcome_cluster$OR_OddsRatio[i] = round(min(unname(test_result$estimate), 99999), digits = 3)
                  Table_outcome_cluster$OR_Odds_pValue[i] = round(unname(test_result$p.value), digits = 3)
                  test_result <- fisher.test(matrix(c(Table_outcome_cluster$DC[i],
                                                      Table_outcome_cluster$N[i] - Table_outcome_cluster$DC[i],
                                                      Table_outcome_cluster$DC[1] - Table_outcome_cluster$DC[i],
                                                      Table_outcome_cluster$N[1] - Table_outcome_cluster$N[i] -Table_outcome_cluster$DC[1] + Table_outcome_cluster$DC[i]), nrow = 2, byrow = TRUE))
                  Table_outcome_cluster$DC_OddsRatio[i] = round(min(unname(test_result$estimate), 99999), digits = 3)
                  Table_outcome_cluster$DC_Odds_pValue[i] = round(unname(test_result$p.value), digits = 3)
                }
              }
              Table_outcome_cluster_total = rbind(Table_outcome_cluster_total, Table_outcome_cluster)
            }
          }
          output$table_outcome_cluster = DT::renderDataTable(Table_outcome_cluster_total,
                                                       server = FALSE,
                                                       filter = 'top', 
                                                       extensions = c('Buttons'), 
                                                       options = list(pageLength = 100, 
                                                                      scrollX = TRUE,
                                                                      scrollY = "1000px",
                                                                      scrollCollapse = TRUE,
                                                                      dom="Blfrtip",
                                                                      buttons = c('csv', 'copy')))
        
      
        Table_outcome_mutation_histology_total = NULL
        drug_list = list(input$drug, input$drug_group_1, input$drug_group_2, input$drug_group_3, input$drug_group_4)
        for(drug_count in 1:5){
          drug_name_1 = case_when(
            drug_count == 1 ~ paste(input$drug, collapse = ";"),
            drug_count == 2 ~ input$drug_group_1_name,
            drug_count == 3 ~ input$drug_group_2_name,
            drug_count == 4 ~ input$drug_group_3_name,
            drug_count == 5 ~ input$drug_group_4_name
          )
          Data_summary_histology = Data_forest_all_pts %>% dplyr::filter(!最良総合効果 %in% c("NE", "Unknown")) %>% dplyr::filter(Drug %in% drug_list[[drug_count]]) %>% dplyr::arrange(-ORR_data, -DCR_data) %>% dplyr::distinct(ID, .keep_all = T)
          if(length(Data_summary_histology$Drug)>0){
            Table_outcome_mutation_histology = data.frame(c("Total", as.character((Data_summary_histology %>% group_by(Histology) %>% tally())$Histology)))
            colnames(Table_outcome_mutation_histology) = "Histology"
            Table_outcome_mutation_histology$Drug = drug_name_1
            Table_outcome_mutation_histology$N = 0
            Table_outcome_mutation_histology$OR = 0
            Table_outcome_mutation_histology$ORR = 0
            Table_outcome_mutation_histology$ORR_95CI_LL = 0
            Table_outcome_mutation_histology$ORR_95CI_UL = 0
            Table_outcome_mutation_histology$DC = 0
            Table_outcome_mutation_histology$DCR = 0
            Table_outcome_mutation_histology$DCR_95CI_LL = 0
            Table_outcome_mutation_histology$DCR_95CI_UL = 0
            Table_outcome_mutation_histology$OR_OddsRatio = 1
            Table_outcome_mutation_histology$OR_Odds_pValue = 1
            Table_outcome_mutation_histology$DC_OddsRatio = 1
            Table_outcome_mutation_histology$DC_Odds_pValue = 1
            tmp = Data_summary_histology %>% group_by(Histology) %>% tally()
            Table_outcome_mutation_histology$N[1] = sum(tmp$n)
            if(sum(tmp$n)>0){
              for(i in 1:(length(Table_outcome_mutation_histology$Histology)-1)){
                Table_outcome_mutation_histology$N[i+1] = tmp$n[i]
              }
              
              tmp = Data_summary_histology %>% group_by(Histology, ORR_data) %>% tally()
              Table_outcome_mutation_histology$OR[1] = sum((tmp %>% dplyr::filter(ORR_data == 1))$n)
              for(i in 1:(length(Table_outcome_mutation_histology$Histology)-1)){
                if(length((tmp %>% dplyr::filter(ORR_data == 1 & Histology == Table_outcome_mutation_histology$Histology[i+1]))$n)>0){
                  Table_outcome_mutation_histology$OR[i+1] = (tmp %>% dplyr::filter(ORR_data == 1 & Histology == Table_outcome_mutation_histology$Histology[i+1]))$n
                }
              }
              tmp = Data_summary_histology %>% group_by(Histology, DCR_data) %>% tally()
              Table_outcome_mutation_histology$DC[1] = sum((tmp %>% dplyr::filter(DCR_data == 1))$n)
              for(i in 1:(length(Table_outcome_mutation_histology$Histology)-1)){
                if(length((tmp %>% dplyr::filter(DCR_data == 1 & Histology == Table_outcome_mutation_histology$Histology[i+1]))$n)>0){
                  Table_outcome_mutation_histology$DC[i+1] = (tmp %>% dplyr::filter(DCR_data == 1 & Histology == Table_outcome_mutation_histology$Histology[i+1]))$n
                }
              }
              
              for(i in 1:length(Table_outcome_mutation_histology$Histology)){
                Table_outcome_mutation_histology$ORR[i] = round(Table_outcome_mutation_histology$OR[i] / Table_outcome_mutation_histology$N[i], digits = 3)
                Table_outcome_mutation_histology$DCR[i] = round(Table_outcome_mutation_histology$DC[i] / Table_outcome_mutation_histology$N[i], digits = 3)
                Table_outcome_mutation_histology$ORR_95CI_LL[i] = round(exactci(Table_outcome_mutation_histology$OR[i], Table_outcome_mutation_histology$N[i], conf.level=0.95)$conf.int[1], digits = 3)
                Table_outcome_mutation_histology$ORR_95CI_UL[i] = round(exactci(Table_outcome_mutation_histology$OR[i], Table_outcome_mutation_histology$N[i], conf.level=0.95)$conf.int[2], digits = 3)
                Table_outcome_mutation_histology$DCR_95CI_LL[i] = round(exactci(Table_outcome_mutation_histology$DC[i], Table_outcome_mutation_histology$N[i], conf.level=0.95)$conf.int[1], digits = 3)
                Table_outcome_mutation_histology$DCR_95CI_UL[i] = round(exactci(Table_outcome_mutation_histology$DC[i], Table_outcome_mutation_histology$N[i], conf.level=0.95)$conf.int[2], digits = 3)
                test_result <- fisher.test(matrix(c(Table_outcome_mutation_histology$OR[i],
                                                    Table_outcome_mutation_histology$N[i] - Table_outcome_mutation_histology$OR[i],
                                                    Table_outcome_mutation_histology$OR[1] - Table_outcome_mutation_histology$OR[i],
                                                    Table_outcome_mutation_histology$N[1] - Table_outcome_mutation_histology$N[i] -Table_outcome_mutation_histology$OR[1] + Table_outcome_mutation_histology$OR[i]), nrow = 2, byrow = TRUE))
                Table_outcome_mutation_histology$OR_OddsRatio[i] = round(min(unname(test_result$estimate), 99999), digits = 3)
                Table_outcome_mutation_histology$OR_Odds_pValue[i] = round(unname(test_result$p.value), digits = 3)
                test_result <- fisher.test(matrix(c(Table_outcome_mutation_histology$DC[i],
                                                    Table_outcome_mutation_histology$N[i] - Table_outcome_mutation_histology$DC[i],
                                                    Table_outcome_mutation_histology$DC[1] - Table_outcome_mutation_histology$DC[i],
                                                    Table_outcome_mutation_histology$N[1] - Table_outcome_mutation_histology$N[i] -Table_outcome_mutation_histology$DC[1] + Table_outcome_mutation_histology$DC[i]), nrow = 2, byrow = TRUE))
                Table_outcome_mutation_histology$DC_OddsRatio[i] = round(min(unname(test_result$estimate), 99999), digits = 3)
                Table_outcome_mutation_histology$DC_Odds_pValue[i] = round(unname(test_result$p.value), digits = 3)
              }
            }
            Table_outcome_mutation_histology_total = rbind(Table_outcome_mutation_histology_total, Table_outcome_mutation_histology)
          }
        }
    
    
        output$table_outcome_2 = DT::renderDataTable(Table_outcome_mutation_histology_total,
                                                    server = FALSE,
                                                    filter = 'top', 
                                                    extensions = c('Buttons'), 
                                                    options = list(pageLength = 100, 
                                                                   scrollX = TRUE,
                                                                   scrollY = "1000px",
                                                                   scrollCollapse = TRUE,
                                                                   dom="Blfrtip",
                                                                   buttons = c('csv', 'copy')))
        
        mut_gene_ = input$gene[input$gene %in% unique(Data_MAF_target$Hugo_Symbol)]
        if(length(mut_gene_)==0){
          mut_gene_ = "TP53"
          ID_mutation_ = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene_))$Tumor_Sample_Barcode
          Data_forest_all_pts = Data_forest_all_pts %>% dplyr::mutate(
            mutation = case_when(
              ID %in% ID_mutation_ ~ "TP53 mut(+)",
              TRUE ~ "TP53 mut(-)"
            )
          )
        } else{
          if(is.null(input$gene_group_2)){
            ID_mutation_ = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% mut_gene_))$Tumor_Sample_Barcode
            Data_forest_all_pts = Data_forest_all_pts %>% dplyr::mutate(
              mutation = case_when(
                ID %in% ID_mutation_ ~ paste0(paste(mut_gene_, collapse = "/"), " mut(+)"),
                TRUE ~ paste0(paste(mut_gene_, collapse = "/"), " mut(-)")
              )
            )
          } else{
            ID_mutation_1 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_1))$Tumor_Sample_Barcode
            ID_mutation_2 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_2))$Tumor_Sample_Barcode
            Data_forest_all_pts = Data_forest_all_pts %>% dplyr::mutate(
              mutation = case_when(
                ID %in% ID_mutation_1 & ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+) & ", paste(input$gene_group_2, collapse = "/"), " mut(+)"),
                ID %in% ID_mutation_1 & !ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+) & ", paste(input$gene_group_2, collapse = "/"), " mut(-)"),
                !ID %in% ID_mutation_1 & ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-) & ", paste(input$gene_group_2, collapse = "/"), " mut(+)"),
                !ID %in% ID_mutation_1 & !ID %in% ID_mutation_2 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-) & ", paste(input$gene_group_2, collapse = "/"), " mut(-)"),
              )
            )
          }
        }
        
        Table_outcome_mutation_total = NULL
        drug_list = list(input$drug, input$drug_group_1, input$drug_group_2, input$drug_group_3, input$drug_group_4)
        for(drug_count in 1:5){
          drug_name_2 = case_when(
            drug_count == 1 ~ paste(input$drug, collapse = ";"),
            drug_count == 2 ~ input$drug_group_1_name,
            drug_count == 3 ~ input$drug_group_2_name,
            drug_count == 4 ~ input$drug_group_3_name,
            drug_count == 5 ~ input$drug_group_4_name
          )
          Data_summary_mutation = Data_forest_all_pts %>% dplyr::filter(!最良総合効果 %in% c("NE", "Unknown")) %>% dplyr::filter(Drug %in% drug_list[[drug_count]])  %>% dplyr::arrange(-ORR_data, -DCR_data) %>% dplyr::distinct(ID, .keep_all = T)
          if(length(Data_summary_mutation$Drug)>0){
            Table_outcome_mutation = data.frame(c("Total", as.character((Data_summary_mutation %>% group_by(mutation) %>% tally())$mutation)))
            colnames(Table_outcome_mutation) = "Mutation"
            Table_outcome_mutation$Drug = drug_name_2
            Table_outcome_mutation$N = 0
            Table_outcome_mutation$OR = 0
            Table_outcome_mutation$ORR = 0
            Table_outcome_mutation$ORR_95CI_LL = 0
            Table_outcome_mutation$ORR_95CI_UL = 0
            Table_outcome_mutation$DC = 0
            Table_outcome_mutation$DCR = 0
            Table_outcome_mutation$DCR_95CI_LL = 0
            Table_outcome_mutation$DCR_95CI_UL = 0
            Table_outcome_mutation$OR_OddsRatio = 1
            Table_outcome_mutation$OR_Odds_pValue = 1
            Table_outcome_mutation$DC_OddsRatio = 1
            Table_outcome_mutation$DC_Odds_pValue = 1
            tmp2 = Data_summary_mutation %>% group_by(mutation) %>% tally()
            Table_outcome_mutation$N[1] = sum(tmp2$n)
            if(length(Table_outcome_mutation$Mutation)>1){
              for(i in 1:(length(Table_outcome_mutation$Mutation)-1)){
                Table_outcome_mutation$N[i+1] = tmp2$n[i]
              }
              
              tmp2 = Data_summary_mutation %>% group_by(mutation, ORR_data) %>% tally()
              Table_outcome_mutation$OR[1] = sum((tmp2 %>% dplyr::filter(ORR_data == 1))$n)
              for(i in 1:(length(Table_outcome_mutation$Mutation)-1)){
                if(length((tmp2 %>% dplyr::filter(ORR_data == 1 & mutation == Table_outcome_mutation$Mutation[i+1]))$n)>0){
                  Table_outcome_mutation$OR[i+1] = (tmp2 %>% dplyr::filter(ORR_data == 1 & mutation == Table_outcome_mutation$Mutation[i+1]))$n
                }
              }
              tmp2 = Data_summary_mutation %>% group_by(mutation, DCR_data) %>% tally()
              Table_outcome_mutation$DC[1] = sum((tmp2 %>% dplyr::filter(DCR_data == 1))$n)
              for(i in 1:(length(Table_outcome_mutation$Mutation)-1)){
                if(length((tmp2 %>% dplyr::filter(DCR_data == 1 & mutation == Table_outcome_mutation$Mutation[i+1]))$n)>0){
                  Table_outcome_mutation$DC[i+1] = (tmp2 %>% dplyr::filter(DCR_data == 1 & mutation == Table_outcome_mutation$Mutation[i+1]))$n
                }
              }
              
              for(i in 1:length(Table_outcome_mutation$Mutation)){
                Table_outcome_mutation$ORR[i] = round(Table_outcome_mutation$OR[i] / Table_outcome_mutation$N[i], digits = 3)
                Table_outcome_mutation$DCR[i] = round(Table_outcome_mutation$DC[i] / Table_outcome_mutation$N[i], digits = 3)
                Table_outcome_mutation$ORR_95CI_LL[i] = round(exactci(Table_outcome_mutation$OR[i], Table_outcome_mutation$N[i], conf.level=0.95)$conf.int[1], digits = 3)
                Table_outcome_mutation$ORR_95CI_UL[i] = round(exactci(Table_outcome_mutation$OR[i], Table_outcome_mutation$N[i], conf.level=0.95)$conf.int[2], digits = 3)
                Table_outcome_mutation$DCR_95CI_LL[i] = round(exactci(Table_outcome_mutation$DC[i], Table_outcome_mutation$N[i], conf.level=0.95)$conf.int[1], digits = 3)
                Table_outcome_mutation$DCR_95CI_UL[i] = round(exactci(Table_outcome_mutation$DC[i], Table_outcome_mutation$N[i], conf.level=0.95)$conf.int[2], digits = 3)
                test_result <- fisher.test(matrix(c(Table_outcome_mutation$OR[i],
                                                    Table_outcome_mutation$N[i] - Table_outcome_mutation$OR[i],
                                                    Table_outcome_mutation$OR[1] - Table_outcome_mutation$OR[i],
                                                    Table_outcome_mutation$N[1] - Table_outcome_mutation$N[i] -Table_outcome_mutation$OR[1] + Table_outcome_mutation$OR[i]), nrow = 2, byrow = TRUE))
                Table_outcome_mutation$OR_OddsRatio[i] = round(min(unname(test_result$estimate), 99999), digits = 3)
                Table_outcome_mutation$OR_Odds_pValue[i] = round(unname(test_result$p.value), digits = 3)
                test_result <- fisher.test(matrix(c(Table_outcome_mutation$DC[i],
                                                    Table_outcome_mutation$N[i] - Table_outcome_mutation$DC[i],
                                                    Table_outcome_mutation$DC[1] - Table_outcome_mutation$DC[i],
                                                    Table_outcome_mutation$N[1] - Table_outcome_mutation$N[i] -Table_outcome_mutation$DC[1] + Table_outcome_mutation$DC[i]), nrow = 2, byrow = TRUE))
                Table_outcome_mutation$DC_OddsRatio[i] = round(min(unname(test_result$estimate), 99999), digits = 3)
                Table_outcome_mutation$DC_Odds_pValue[i] = round(unname(test_result$p.value), digits = 3)
              }
              Table_outcome_mutation_total = rbind(Table_outcome_mutation_total, Table_outcome_mutation)
            }
          }
        }
        output$table_outcome_3 = DT::renderDataTable(Table_outcome_mutation_total,
                                                            server = FALSE,
                                                            filter = 'top', 
                                                            extensions = c('Buttons'), 
                                                            options = list(pageLength = 100, 
                                                                           scrollX = TRUE,
                                                                           scrollY = "1000px",
                                                                           scrollCollapse = TRUE,
                                                                           dom="Blfrtip",
                                                                           buttons = c('csv', 'copy')))  
      }
      incProgress(1 / 13)
      
      g_surv_drug = list()
      kkkkk=1
      if(input$drug_volcano != "Yes"){
        Data_drug_survival = Data_survival %>%
          dplyr::select(C.CAT調査結果.基本項目.ハッシュID, 
                        time_enroll_final,
                        time_palliative_enroll,
                        time_palliative_final,
                        censor)
        Data_drug_post_CGP = Data_drug_original %>%
          dplyr::filter(治療ライン %in% input$target_line) %>%
          dplyr::filter(TxCGP == "Post")
        Data_ID_target_drug_used = unique((Data_drug_post_CGP %>% dplyr::filter(Drug %in% input$drug))$ID)
        Data_ID_other_drug_used = setdiff(unique(Data_drug_post_CGP$ID), Data_ID_target_drug_used)
  
        Data_drug_survival = Data_drug_survival %>%
          dplyr::mutate(Regimen = case_when(
            C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_target_drug_used ~ 2,
            C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_other_drug_used ~ 1,
            TRUE ~ 0
          )) %>%
          dplyr::arrange(desc(Regimen)) %>%
          dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = T) %>%
          dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
        Data_drug_survival = Data_drug_survival %>%
          dplyr::mutate(CTx_after_CGP = case_when(
            Regimen == 2 ~ paste0("Including ", paste(input$drug, collapse = ";")),
            Regimen == 1 ~ "Other drugs",
            TRUE ~ "Treatment (-)"
          ))
  
        Data_drug_survival_entry = Data_drug_survival %>% dplyr::filter(!is.na(time_enroll_final))
        survival_simple = survfit(Surv(time_enroll_final, censor)~CTx_after_CGP,
                                  data=Data_drug_survival_entry)

        if(length(Data_drug_survival_entry[,1][[1]]) != survival_simple[[1]][[1]]){
          diff_0 = survdiff(Surv(time_enroll_final, censor)~CTx_after_CGP,
                            data=Data_drug_survival_entry, rho=0)
          diff_1 = survdiff(Surv(time_enroll_final, censor)~CTx_after_CGP,
                            data=Data_drug_survival_entry, rho=1)
          g_surv_drug[[kkkkk]] = surv_curv_entry(survival_simple, Data_drug_survival_entry, "Treatment and survival after CGP",
                                    NULL, diff_0, diff_1)
          kkkkk = kkkkk + 1
        }
        Data_drug_survival_treated_only = Data_drug_survival_entry %>% dplyr::filter(CTx_after_CGP != "Treatment (-)")
        if(length(Data_drug_survival_treated_only$C.CAT調査結果.基本項目.ハッシュID > 0)){
          survival_simple = survfit(Surv(time_enroll_final, censor)~CTx_after_CGP,
                                    data=Data_drug_survival_treated_only)
          if(length(Data_drug_survival_treated_only[,1][[1]]) != survival_simple[[1]][[1]]){
            diff_0 = survdiff(Surv(time_enroll_final, censor)~CTx_after_CGP,
                              data=Data_drug_survival_treated_only, rho=0)
            diff_1 = survdiff(Surv(time_enroll_final, censor)~CTx_after_CGP,
                              data=Data_drug_survival_treated_only, rho=1)
            g_surv_drug[[kkkkk]] = surv_curv_entry(survival_simple, Data_drug_survival_treated_only, "Treatment and survival after CGP",
                                                   NULL, diff_0, diff_1)
            kkkkk = kkkkk + 1
          }
        }
        if(!is.null(input$gene_group_1)){
          ID_mutation_1 = (Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_1))$Tumor_Sample_Barcode
          Data_drug_survival_target_treated_only = Data_drug_survival_treated_only %>%
            dplyr::filter(CTx_after_CGP != "Other drugs") %>%
            dplyr::mutate(
            mutation = case_when(
              C.CAT調査結果.基本項目.ハッシュID %in% ID_mutation_1 ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+)"),
              TRUE ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-)"),
            )
          )
          if(length(Data_drug_survival_target_treated_only$C.CAT調査結果.基本項目.ハッシュID > 0)){
            survival_simple = survfit(Surv(time_enroll_final, censor)~mutation,
                                      data=Data_drug_survival_target_treated_only)
            if(length(Data_drug_survival_target_treated_only[,1][[1]]) != survival_simple[[1]][[1]]){
              diff_0 = survdiff(Surv(time_enroll_final, censor)~mutation,
                                data=Data_drug_survival_target_treated_only, rho=0)
              diff_1 = survdiff(Surv(time_enroll_final, censor)~mutation,
                                data=Data_drug_survival_target_treated_only, rho=1)
              g_surv_drug[[kkkkk]] = surv_curv_entry(survival_simple, Data_drug_survival_target_treated_only, paste0("Treated regimens including ", paste(input$drug, collapse = ";")," after CGP\nSurvival after CGP"),
                                                     NULL, diff_0, diff_1)
              kkkkk = kkkkk + 1
            }
          }
        }
        if(!is.null(input$drug_group_1) & !is.null(input$drug_group_2)){
          Data_ID_drug_1 = unique((Data_drug_post_CGP %>% dplyr::filter(Drug %in% input$drug_group_1))$ID)
          Data_ID_drug_2 = unique((Data_drug_post_CGP %>% dplyr::filter(Drug %in% input$drug_group_2))$ID)
          Data_ID_drug_1and2 = intersect(Data_ID_drug_1, Data_ID_drug_2)
          Data_drug_survival_1vs2 = Data_drug_survival %>%
            dplyr::mutate(CTx_after_CGP = case_when(
              C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_drug_1and2 ~ paste0("both ", input$drug_group_1_name, "&", input$drug_group_2_name),
              C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_drug_1 ~ input$drug_group_1_name,
              C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_drug_2 ~ input$drug_group_2_name,
              TRUE ~ "Other"
            ))
          Data_drug_survival_1vs2 = Data_drug_survival_1vs2 %>% dplyr::filter(CTx_after_CGP != "Other")
          Data_drug_survival_1vs2 = Data_drug_survival_1vs2 %>% dplyr::filter(!is.na(time_enroll_final))
          if(length(Data_drug_survival_1vs2$C.CAT調査結果.基本項目.ハッシュID > 0)){
            survival_simple = survfit(Surv(time_enroll_final, censor)~CTx_after_CGP,
                                      data=Data_drug_survival_1vs2)
            if(length(Data_drug_survival_1vs2[,1][[1]]) != survival_simple[[1]][[1]]){
              diff_0 = survdiff(Surv(time_enroll_final, censor)~CTx_after_CGP,
                                data=Data_drug_survival_1vs2, rho=0)
              diff_1 = survdiff(Surv(time_enroll_final, censor)~CTx_after_CGP,
                                data=Data_drug_survival_1vs2, rho=1)
              g_surv_drug[[kkkkk]] = surv_curv_entry(survival_simple, Data_drug_survival_1vs2, "Treatment and survival after CGP",
                                                     NULL, diff_0, diff_1)
              kkkkk = kkkkk + 1
            }
          }
        }
      }
        
      if(input$drug_volcano != "Yes"){
        # Survival analysis for drug, #1
        Data_drug_survival_save = Data_drug_survival
        Data_ID_target_drug_used_original = unique((Data_drug_original %>%
                                                      dplyr::filter(治療ライン %in% input$target_line) %>%
                                                      dplyr::filter(Drug %in% input$drug))$ID)
        Data_ID_other_drug_used_original = setdiff(unique((Data_drug_original %>%
                                                             dplyr::filter(治療ライン %in% input$target_line))$ID),
                                                   Data_ID_target_drug_used_original)
        
        Data_drug_survival_1 = Data_drug_survival_save %>%
          dplyr::mutate(Regimen_original = case_when(
            C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_target_drug_used_original ~ 2,
            TRUE ~ 0
          )) %>%
          dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID)
        Data_drug_survival_1 = Data_drug_survival_1 %>% dplyr::mutate(
          CTx = case_when(
            Regimen_original == 2 ~ paste0("Including ", paste(input$drug, collapse = ";")),
            TRUE ~ "Other drugs"
          )) %>%
          dplyr::filter(!is.na(time_palliative_enroll) & !is.na(time_enroll_final))
        Data_drug_survival_1_save = Data_drug_survival_1
        
        if(sum(Data_drug_survival_1$censor) > 0){
          traditional_fit = survfit(Surv(event = censor,
                                         time = time_palliative_final) ~ CTx, 
                                    data = Data_drug_survival_1)
          if(length(Data_drug_survival_1[,1]) != traditional_fit[[1]][[1]]){
            name_1 = paste0("Including ", paste(input$drug, collapse = ";"))
            name_2 = "Other drugs"
            name_1_short = paste(input$drug, collapse = ";")
            name_2_short = "Other drugs"
            
            Data_drug_survival_1$gene_mut = Data_drug_survival_1$CTx == "Other drugs"
            Data_drug_survival_1_curve = Data_drug_survival_1
            Data_survival_tmp_pos = Data_drug_survival_1_curve %>%
              dplyr::filter(gene_mut == TRUE)
            if(sum(Data_survival_tmp_pos$censor) > 0){
              fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_enroll_final)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_enroll_final)*0.9)
            Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
              time_enroll_final >= time_10 &
                time_enroll_final <= time_90)
            Data_survival_tmp$time_enroll_final =
              Data_survival_tmp$time_enroll_final - (time_10 - 1)
            Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
              time_enroll_final > time_90) %>%
              dplyr::arrange(time_enroll_final)
            
            Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp2)
            Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
            
            Data_survival_tmp_neg = Data_drug_survival_1_curve %>%
              dplyr::filter(gene_mut == FALSE)
            if(sum(Data_survival_tmp_neg$censor) > 0){
              fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_neg)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_enroll_final)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_enroll_final)*0.9)
            Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
              time_enroll_final >= time_10 &
                time_enroll_final <= time_90)
            Data_survival_tmp3$time_enroll_final =
              Data_survival_tmp3$time_enroll_final - (time_10 - 1)
            Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
              time_enroll_final > time_90) %>%
              dplyr::arrange(time_enroll_final)
            
            Data_survival_tmp4$time_enroll_final = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp4)
            Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx < max(idx)*(1-survival_rate_10)],])
            Data_drug_survival_1 = rbind(Data_survival_tmp, Data_survival_tmp3)  
            
            Data_survival_tmp_pos = Data_drug_survival_1_curve %>%
              dplyr::filter(gene_mut == TRUE)
            if(sum(Data_survival_tmp_pos$censor) > 0){
              fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_pos)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)
            Data_survival_tmp5 = Data_survival_tmp_pos %>% dplyr::filter(
              time_palliative_enroll >= time_10 &
                time_palliative_enroll <= time_90)
            Data_survival_tmp5$time_palliative_enroll =
              Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
            Data_survival_tmp6 = Data_survival_tmp_pos %>% dplyr::filter(
              time_palliative_enroll > time_90) %>%
              dplyr::arrange(time_palliative_enroll)
            Data_survival_tmp6$time_palliative_enroll = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp6)
            Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
            
            Data_survival_tmp_neg = Data_drug_survival_1_curve %>%
              dplyr::filter(gene_mut == FALSE)
            
            if(sum(Data_survival_tmp_neg$censor) > 0){
              fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)
            Data_survival_tmp7 = Data_survival_tmp_neg %>% dplyr::filter(
              time_palliative_enroll >= time_10 &
                time_palliative_enroll <= time_90)
            Data_survival_tmp7$time_palliative_enroll =
              Data_survival_tmp7$time_palliative_enroll - (time_10 - 1)
            Data_survival_tmp8 = Data_survival_tmp_neg %>% dplyr::filter(
              time_palliative_enroll > time_90) %>%
              dplyr::arrange(time_palliative_enroll)
            Data_survival_tmp8$time_palliative_enroll = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp8)
            Data_survival_tmp7 = rbind(Data_survival_tmp7, Data_survival_tmp8[idx[idx < max(idx)*(1-survival_rate_10)],])
            
            Median_entry_pos = median(Data_survival_tmp5$time_palliative_enroll)
            Median_entry_neg = median(Data_survival_tmp7$time_palliative_enroll)
            Data_drug_survival_1_2 = rbind(Data_survival_tmp5, Data_survival_tmp7)  
            
            Data_drug_survival_1 = Data_drug_survival_1 %>%
              dplyr::mutate(delay_1 = case_when(
                time_palliative_enroll <= Median_entry_pos &
                  gene_mut == TRUE ~ "SPT to entry early",
                time_palliative_enroll > Median_entry_pos &
                  gene_mut == TRUE ~ "SPT to entry later",
                time_palliative_enroll <= Median_entry_neg &
                  gene_mut == FALSE ~ "SPT to entry early",
                time_palliative_enroll > Median_entry_neg &
                  gene_mut == FALSE ~ "SPT to entry later"))
            
            if(sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
               sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
               sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
               sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
              
              if(input$drug_survuval_CTx == "Yes"){
                library(cmdstanr)
                stan_weibull_survival_model_data <-
                  list(
                    Median_pos = as.integer(Median_entry_pos),
                    Median_neg = as.integer(Median_entry_neg),
                    Nobs_early_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Nobs_late_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Ncen_early_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Ncen_late_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Nobs_early_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Nobs_late_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Ncen_early_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Ncen_late_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Nexp_pos = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE]),
                    Nexp_neg = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE]),
                    ybef_exp_pos = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE],
                    ybef_exp_neg = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE],
                    yobs_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    yobs_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    ycen_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    ycen_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    yobs_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                    yobs_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                    ycen_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                    ycen_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE]
                  )
                
                stan_model_path <- "source/Stan_code_loglog_weibull_factor.stan"
                stan_model_compiled <- cmdstan_model(stan_model_path)
                stan_weibull_survival_model_fit <- stan_model_compiled$sample(
                  data = stan_weibull_survival_model_data,
                  seed = 1234,
                  chains = 4,
                  parallel_chains = min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE)),
                  iter_sampling = 2000,  # (iter-warmup)
                  iter_warmup = 1000,
                  thin = 1,
                  refresh = 500
                )
                
                stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
                stan_weibull_survival_model_draws_total <-
                  stan_weibull_survival_model_draws %>%
                  dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
                  tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
                  dplyr::select(-key) 
                stan_weibull_survival_model_draws_total_exp <-
                  stan_weibull_survival_model_draws %>%
                  dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
                  tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
                  dplyr::select(-key) 
                
                median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
                median_os_list_weibull_total = matrix(stan_weibull_survival_model_draws_total$yhat_total, ncol=4000)
                median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
                median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
                median_os_list_weibull_total = colMedians(median_os_list_weibull_total,na.rm = T)
                median_os_list_weibull_bias = colMedians(median_os_list_weibull_bias,na.rm = T)
                
                median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
                median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
                median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
                yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
                yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
                yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_drug_survival_1_curve$censor) * as.integer(length(yhat_weibull_sort_total)/length(Data_drug_survival_1_curve$censor))]
                yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_drug_survival_1_curve$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_drug_survival_1_curve$censor))]
                
                Data_drug_survival_1_curve$factor_neg = yhat_weibull_sort_average_total
                Data_drug_survival_1_curve$factor_pos = yhat_weibull_sort_average_total_exp
                Data_drug_survival_1_curve$factor_neg[Data_drug_survival_1_curve$factor_neg > 10000] = 10000
                Data_drug_survival_1_curve$factor_pos[Data_drug_survival_1_curve$factor_pos > 10000] = 10000
                
                traditional_fit = survfit(Surv(event = censor,
                                               time = time_palliative_final) ~ gene_mut,
                                          data = Data_drug_survival_1_curve)
                simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_drug_survival_1_curve$factor_neg)),
                                                  time = factor_neg) ~ 1, 
                                             data = Data_drug_survival_1_curve)
                simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_drug_survival_1_curve$factor_pos)),
                                                  time = factor_pos) ~ 1, 
                                             data = Data_drug_survival_1_curve)
                
                g_surv_drug[[kkkkk]] = ggsurvplot(
                  fit = list(traditional_fit = traditional_fit,
                             simulation_fit_neg = simulation_fit_neg,
                             simulation_fit_pos = simulation_fit_pos),
                  combine = TRUE,
                  data = Data_drug_survival_1_curve,
                  xlab = "Time from chemotherapy induction to final observation (months)",
                  ylab = "Survival Probability",
                  censor = TRUE,
                  font.title = 8,
                  font.subtitle = 8,
                  font.main = 8,
                  font.submain = 8,
                  font.caption = 8,
                  font.legend = 8,
                  surv.median.line = "hv",
                  conf.int = FALSE,
                  pval = FALSE,
                  risk.table = TRUE,
                  risk.table.y.text = FALSE,
                  tables.theme = clean_theme(), 
                  legend = c(0.8,0.8),
                  xlim = c(0, max(Data_drug_survival_1_curve$time_palliative_final) * 1.05),
                  cumevents = FALSE,
                  cumcensor = FALSE,
                  break.x.by = 365.25 * 2.5,
                  xscale = "d_m",
                  #break.x.by = ceiling(max(Data_drug_survival_1_curve$time_palliative_final) / 500) * 100,
                  legend.labs = c(paste0(name_1, ", Unadjusted"),
                                  paste0(name_2, ", Unadjusted"),
                                  paste0(name_1, ", Adjusted for Delayed Entry"),
                                  paste0(name_2, ", Adjusted for Delayed Entry"))
                ) +   
                  labs(title = paste(sum(traditional_fit[[1]]), " patients ",
                                     "Median OS ",
                                     round(summary(traditional_fit)$table[[13]] / 365.25 * 12, digits = 1)," (",
                                     round(summary(traditional_fit)$table[[15]] / 365.25 * 12, digits = 1),"-",
                                     round(summary(traditional_fit)$table[[17]] / 365.25 * 12, digits = 1),") (",
                                     name_1_short,", unadj.)/",
                                     round(summary(traditional_fit)$table[[14]] / 365.25 * 12, digits = 1), " (",
                                     round(summary(traditional_fit)$table[[16]] / 365.25 * 12, digits = 1),"-",
                                     round(summary(traditional_fit)$table[[18]] / 365.25 * 12, digits = 1),") (",
                                     name_2_short,", unadj.)/",
                                     round(median_os_summary_weibull_total[[2]] / 365.25 * 12, digits = 1), 
                                     " (", round(median_os_summary_weibull_total[[1]] / 365.25 * 12, digits = 1), "-",
                                     round(median_os_summary_weibull_total[[3]] / 365.25 * 12, digits = 1), ") (",
                                     name_1_short,", adj.)/",
                                     round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                                     round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                                     round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1),
                                     ") (", name_2_short,", adj.) months",
                                     sep=""),
                       subtitle = paste("Survival difference, ",name_1_short, "-", name_2_short, ": ",
                                        round(median_os_summary_weibull_bias[[2]] / 365.25 * 12, digits = 1), " (",
                                        round(median_os_summary_weibull_bias[[1]] / 365.25 * 12, digits = 1), "-", 
                                        round(median_os_summary_weibull_bias[[3]] / 365.25 * 12, digits = 1), 
                                        ") months, median (95% CI)",
                                        sep=""))
                g_surv_drug[[kkkkk]]$table <- g_surv_drug[[kkkkk]]$table + theme(plot.title = element_blank(),
                                                                                 plot.subtitle = element_blank())
                kkkkk = kkkkk + 1
              }
            }
            if(length(traditional_fit[[1]])>1){
              delayed_entry_fit = survfit(Surv(event = censor,
                                               time = time_palliative_enroll,
                                               time2 = time_palliative_final) ~ gene_mut,
                                          data = Data_drug_survival_1_curve)
              if(length(unique(Data_drug_survival_1_curve$gene_mut)) > 1){
                diff_0 = coxph(Surv(time = time_palliative_enroll,
                                    time2 = time_palliative_final, censor)~gene_mut,
                               data=Data_drug_survival_1_curve)
                g_surv_drug[[kkkkk]] = surv_curv_CTx(delayed_entry_fit, Data_drug_survival_1_curve, paste0(paste(traditional_fit[[1]],collapse = "/"), " patients"),
                                                     c(paste0(name_1, ", Adjusted for Delayed Entry"),
                                                       paste0(name_2, ", Adjusted for Delayed Entry")),
                                                      diff_0)
                kkkkk = kkkkk + 1
              }
            }
          }
        }
  
        # Survival analysis for drug, concerning mutation, #2
        Data_ID_target_mutation_original = unique((Data_MAF_target %>% dplyr::filter(Hugo_Symbol %in% input$gene_group_1))$Tumor_Sample_Barcode)
        Data_drug_survival_1 = Data_drug_survival_1_save %>%
          dplyr::mutate(mutation = case_when(
            C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_target_mutation_original ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(+)"),
            TRUE ~ paste0(paste(input$gene_group_1, collapse = "/"), " mut(-)")
          )) %>%
          dplyr::arrange(C.CAT調査結果.基本項目.ハッシュID) %>%
          dplyr::filter(Regimen_original == 2)

        if(sum(Data_drug_survival_1$censor) > 0){
          traditional_fit = survfit(Surv(event = censor,
                                         time = time_palliative_final) ~ mutation, 
                                    data = Data_drug_survival_1)
          if(length(Data_drug_survival_1[,1]) != traditional_fit[[1]][[1]]){
            name_1 = paste0(paste(input$gene_group_1, collapse = "/"), " mut(+)")
            name_2 = paste0(paste(input$gene_group_1, collapse = "/"), " mut(-)")
            name_1_short = "mut(+)"
            name_2_short = "mut(-)"
            
            Data_drug_survival_1$gene_mut = Data_drug_survival_1$mutation == paste0(paste(input$gene_group_1, collapse = "/"), " mut(-)")
            Data_drug_survival_1_curve = Data_drug_survival_1
            Data_survival_tmp_pos = Data_drug_survival_1_curve %>%
              dplyr::filter(gene_mut == TRUE)
            if(sum(Data_survival_tmp_pos$censor) > 0){
              fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_enroll_final)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_enroll_final)*0.9)
            Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
              time_enroll_final >= time_10 &
                time_enroll_final <= time_90)
            Data_survival_tmp$time_enroll_final =
              Data_survival_tmp$time_enroll_final - (time_10 - 1)
            Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
              time_enroll_final > time_90) %>%
              dplyr::arrange(time_enroll_final)
            
            Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp2)
            Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
            
            Data_survival_tmp_neg = Data_drug_survival_1_curve %>%
              dplyr::filter(gene_mut == FALSE)
            if(sum(Data_survival_tmp_neg$censor) > 0){
              fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_neg)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_enroll_final)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_enroll_final)*0.9)
            Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
              time_enroll_final >= time_10 &
                time_enroll_final <= time_90)
            Data_survival_tmp3$time_enroll_final =
              Data_survival_tmp3$time_enroll_final - (time_10 - 1)
            Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
              time_enroll_final > time_90) %>%
              dplyr::arrange(time_enroll_final)
            
            Data_survival_tmp4$time_enroll_final = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp4)
            Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx < max(idx)*(1-survival_rate_10)],])
            Data_drug_survival_1 = rbind(Data_survival_tmp, Data_survival_tmp3)  
            
            Data_survival_tmp_pos = Data_drug_survival_1_curve %>%
              dplyr::filter(gene_mut == TRUE)
            if(sum(Data_survival_tmp_pos$censor) > 0){
              fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_pos)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)
            Data_survival_tmp5 = Data_survival_tmp_pos %>% dplyr::filter(
              time_palliative_enroll >= time_10 &
                time_palliative_enroll <= time_90)
            Data_survival_tmp5$time_palliative_enroll =
              Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
            Data_survival_tmp6 = Data_survival_tmp_pos %>% dplyr::filter(
              time_palliative_enroll > time_90) %>%
              dplyr::arrange(time_palliative_enroll)
            Data_survival_tmp6$time_palliative_enroll = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp6)
            Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
            
            Data_survival_tmp_neg = Data_drug_survival_1_curve %>%
              dplyr::filter(gene_mut == FALSE)
            
            if(sum(Data_survival_tmp_neg$censor) > 0){
              fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg)
              time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
              survival_rate_10 = summary(fit,times = time_10)[[6]]
              survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)[[6]]
            } else {
              time_10 = 90
              survival_rate_10 = 0
              survival_rate_90 = 1
            }
            time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)
            Data_survival_tmp7 = Data_survival_tmp_neg %>% dplyr::filter(
              time_palliative_enroll >= time_10 &
                time_palliative_enroll <= time_90)
            Data_survival_tmp7$time_palliative_enroll =
              Data_survival_tmp7$time_palliative_enroll - (time_10 - 1)
            Data_survival_tmp8 = Data_survival_tmp_neg %>% dplyr::filter(
              time_palliative_enroll > time_90) %>%
              dplyr::arrange(time_palliative_enroll)
            Data_survival_tmp8$time_palliative_enroll = time_90 - (time_10 - 1)
            idx <- 1:nrow(Data_survival_tmp8)
            Data_survival_tmp7 = rbind(Data_survival_tmp7, Data_survival_tmp8[idx[idx < max(idx)*(1-survival_rate_10)],])
            
            Median_entry_pos = median(Data_survival_tmp5$time_palliative_enroll)
            Median_entry_neg = median(Data_survival_tmp7$time_palliative_enroll)
            Data_drug_survival_1_2 = rbind(Data_survival_tmp5, Data_survival_tmp7)  
            
            Data_drug_survival_1 = Data_drug_survival_1 %>%
              dplyr::mutate(delay_1 = case_when(
                time_palliative_enroll <= Median_entry_pos &
                  gene_mut == TRUE ~ "SPT to entry early",
                time_palliative_enroll > Median_entry_pos &
                  gene_mut == TRUE ~ "SPT to entry later",
                time_palliative_enroll <= Median_entry_neg &
                  gene_mut == FALSE ~ "SPT to entry early",
                time_palliative_enroll > Median_entry_neg &
                  gene_mut == FALSE ~ "SPT to entry later"))
            
            if(sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
               sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
               sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
               sum((Data_drug_survival_1 %>% dplyr::filter(gene_mut == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
              
              if(input$drug_survuval_CTx == "Yes"){
                stan_weibull_survival_model_data <-
                  list(
                    Median_pos = as.integer(Median_entry_pos),
                    Median_neg = as.integer(Median_entry_neg),
                    Nobs_early_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Nobs_late_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Ncen_early_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Ncen_late_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                    Nobs_early_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Nobs_late_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Ncen_early_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Ncen_late_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                    Nexp_pos = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE]),
                    Nexp_neg = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE]),
                    ybef_exp_pos = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE],
                    ybef_exp_neg = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE],
                    yobs_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    yobs_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    ycen_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    ycen_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                    yobs_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                    yobs_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                    ycen_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                    ycen_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE]
                  )
                
                stan_model_path <- "source/Stan_code_loglog_weibull_factor.stan"
                stan_model_compiled <- cmdstan_model(stan_model_path)
                stan_weibull_survival_model_fit <- stan_model_compiled$sample(
                  data = stan_weibull_survival_model_data,
                  seed = 1234,
                  chains = 4,
                  parallel_chains = min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE)),
                  iter_sampling = 2000,  # (iter-warmup)
                  iter_warmup = 1000,
                  thin = 1,
                  refresh = 500
                )
                
                stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
                stan_weibull_survival_model_draws_total <-
                  stan_weibull_survival_model_draws %>%
                  dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
                  tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
                  dplyr::select(-key) 
                stan_weibull_survival_model_draws_total_exp <-
                  stan_weibull_survival_model_draws %>%
                  dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
                  tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
                  dplyr::select(-key) 
                
                median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
                median_os_list_weibull_total = matrix(stan_weibull_survival_model_draws_total$yhat_total, ncol=4000)
                median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
                median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
                median_os_list_weibull_total = colMedians(median_os_list_weibull_total,na.rm = T)
                median_os_list_weibull_bias = colMedians(median_os_list_weibull_bias,na.rm = T)
                
                median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
                median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
                median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
                yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
                yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
                yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_drug_survival_1_curve$censor) * as.integer(length(yhat_weibull_sort_total)/length(Data_drug_survival_1_curve$censor))]
                yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_drug_survival_1_curve$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_drug_survival_1_curve$censor))]
                
                Data_drug_survival_1_curve$factor_neg = yhat_weibull_sort_average_total
                Data_drug_survival_1_curve$factor_pos = yhat_weibull_sort_average_total_exp
                Data_drug_survival_1_curve$factor_neg[Data_drug_survival_1_curve$factor_neg > 10000] = 10000
                Data_drug_survival_1_curve$factor_pos[Data_drug_survival_1_curve$factor_pos > 10000] = 10000
                
                traditional_fit = survfit(Surv(event = censor,
                                               time = time_palliative_final) ~ gene_mut,
                                          data = Data_drug_survival_1_curve)
                simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_drug_survival_1_curve$factor_neg)),
                                                  time = factor_neg) ~ 1, 
                                             data = Data_drug_survival_1_curve)
                simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_drug_survival_1_curve$factor_pos)),
                                                  time = factor_pos) ~ 1, 
                                             data = Data_drug_survival_1_curve)
                
                g_surv_drug[[kkkkk]] = ggsurvplot(
                  fit = list(traditional_fit = traditional_fit,
                             simulation_fit_neg = simulation_fit_neg,
                             simulation_fit_pos = simulation_fit_pos),
                  combine = TRUE,
                  data = Data_drug_survival_1_curve,
                  xlab = "Time from chemotherapy induction to final observation (months)",
                  ylab = "Survival Probability",
                  censor = TRUE,
                  font.title = 8,
                  font.subtitle = 8,
                  font.main = 8,
                  font.submain = 8,
                  font.caption = 8,
                  font.legend = 8,
                  surv.median.line = "hv",
                  conf.int = FALSE,
                  pval = FALSE,
                  risk.table = TRUE,
                  risk.table.y.text = FALSE,
                  tables.theme = clean_theme(), 
                  legend = c(0.8,0.8),
                  xlim = c(0, max(Data_drug_survival_1_curve$time_palliative_final) * 1.05),
                  cumevents = FALSE,
                  cumcensor = FALSE,
                  break.x.by = 365.25 * 2.5,
                  xscale = "d_m",
                  #break.x.by = ceiling(max(Data_drug_survival_1_curve$time_palliative_final) / 500) * 100,
                  legend.labs = c(paste0(name_1, ", Unadjusted"),
                                  paste0(name_2, ", Unadjusted"),
                                  paste0(name_1, ", Adjusted for Delayed Entry"),
                                  paste0(name_2, ", Adjusted for Delayed Entry"))
                ) +   
                  labs(title = paste(paste0("Treated regimens including ", paste(input$drug, collapse = ";")), "/n", sum(traditional_fit[[1]]), " patients ",
                                     "Median OS ",
                                     round(summary(traditional_fit)$table[[13]] / 365.25 * 12, digits = 1)," (",
                                     round(summary(traditional_fit)$table[[15]] / 365.25 * 12, digits = 1),"-",
                                     round(summary(traditional_fit)$table[[17]] / 365.25 * 12, digits = 1),") (",
                                     name_1_short,", unadj.)/",
                                     round(summary(traditional_fit)$table[[14]] / 365.25 * 12, digits = 1), " (",
                                     round(summary(traditional_fit)$table[[16]] / 365.25 * 12, digits = 1),"-",
                                     round(summary(traditional_fit)$table[[18]] / 365.25 * 12, digits = 1),") (",
                                     name_2_short,", unadj.)/",
                                     round(median_os_summary_weibull_total[[2]] / 365.25 * 12, digits = 1), 
                                     " (", round(median_os_summary_weibull_total[[1]] / 365.25 * 12, digits = 1), "-",
                                     round(median_os_summary_weibull_total[[3]] / 365.25 * 12, digits = 1), ") (",
                                     name_1_short,", adj.)/",
                                     round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                                     round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                                     round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1),
                                     ") (", name_2_short,", adj.) months",
                                     sep=""),
                       subtitle = paste("Survival difference, ",name_1_short, "-", name_2_short, ": ",
                                        round(median_os_summary_weibull_bias[[2]] / 365.25 * 12, digits = 1), " (",
                                        round(median_os_summary_weibull_bias[[1]] / 365.25 * 12, digits = 1), "-", 
                                        round(median_os_summary_weibull_bias[[3]] / 365.25 * 12, digits = 1), 
                                        ") months, median (95% CI)",
                                        sep=""))
                g_surv_drug[[kkkkk]]$table <- g_surv_drug[[kkkkk]]$table + theme(plot.title = element_blank(),
                                                                                 plot.subtitle = element_blank())
                kkkkk = kkkkk + 1
              }
            }
            if(length(traditional_fit[[1]])>1){
              delayed_entry_fit = survfit(Surv(event = censor,
                                               time = time_palliative_enroll,
                                               time2 = time_palliative_final) ~ gene_mut,
                                          data = Data_drug_survival_1_curve)
              if(length(unique(Data_drug_survival_1_curve$gene_mut)) > 1){
                diff_0 = coxph(Surv(time = time_palliative_enroll,
                                    time2 = time_palliative_final, censor)~gene_mut,
                               data=Data_drug_survival_1_curve)
                g_surv_drug[[kkkkk]] = surv_curv_CTx(delayed_entry_fit, Data_drug_survival_1_curve, paste0(paste(traditional_fit[[1]],collapse = "/"), " patients"),
                                                     c(paste0(name_1, ", Adjusted for Delayed Entry"),
                                                       paste0(name_2, ", Adjusted for Delayed Entry")),
                                                     diff_0)
                kkkkk = kkkkk + 1
              }
            }
          }
        }
        

        # Survival analysis for drug, #3
        if(!is.null(input$drug_group_1) & !is.null(input$drug_group_2)){
          Data_ID_drug_1_original = unique((Data_drug_original %>%
                                              dplyr::filter(治療ライン %in% input$target_line) %>%
                                              dplyr::filter(Drug %in% input$drug_group_1))$ID)
          Data_ID_drug_2_original = unique((Data_drug_original %>%
                                              dplyr::filter(治療ライン %in% input$target_line) %>%
                                              dplyr::filter(Drug %in% input$drug_group_2))$ID)
          Data_ID_drug_1and2_original = intersect(Data_ID_drug_1_original, Data_ID_drug_2_original)
          Data_ID_drug_1_original = setdiff(Data_ID_drug_1_original, Data_ID_drug_1and2_original)
          Data_ID_drug_2_original = setdiff(Data_ID_drug_2_original, Data_ID_drug_1and2_original)
          
          Data_drug_survival_2 = Data_drug_survival_save %>%
            dplyr::mutate(CTx = case_when(
              C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_drug_1_original ~ input$drug_group_1_name,
              C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_drug_2_original ~ input$drug_group_2_name,
              TRUE ~ "Other"
            ))
          Data_drug_survival_2 = Data_drug_survival_2 %>%
            dplyr::filter(CTx != "Other") %>%
            dplyr::filter(!is.na(time_palliative_enroll) & !is.na(time_enroll_final))
          
          if(sum(Data_drug_survival_2$censor) > 0){
            traditional_fit = survfit(Surv(event = censor,
                                           time = time_palliative_final) ~ CTx, 
                                      data = Data_drug_survival_2)
            if(length(Data_drug_survival_2[,1]) != traditional_fit[[1]][[1]]){
              name_1 = input$drug_group_1_name
              name_2 = input$drug_group_2_name
              name_1_short = input$drug_group_1_name
              name_2_short = input$drug_group_2_name
              
              Data_drug_survival_2$gene_mut = Data_drug_survival_2$CTx == input$drug_group_2_name
              Data_drug_survival_2_curve = Data_drug_survival_2
              Data_survival_tmp_pos = Data_drug_survival_2_curve %>%
                dplyr::filter(gene_mut == TRUE)
              if(sum(Data_survival_tmp_pos$censor) > 0){
                fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_enroll_final)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
               time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_enroll_final)*0.9)
              Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
                time_enroll_final >= time_10 &
                  time_enroll_final <= time_90)
              Data_survival_tmp$time_enroll_final =
                Data_survival_tmp$time_enroll_final - (time_10 - 1)
              Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
                time_enroll_final > time_90) %>%
                dplyr::arrange(time_enroll_final)
              
              Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp2)
              Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
              
              Data_survival_tmp_neg = Data_drug_survival_2_curve %>%
                dplyr::filter(gene_mut == FALSE)
              if(sum(Data_survival_tmp_neg$censor) > 0){
                fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_neg)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_enroll_final)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_enroll_final)*0.9)
              Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
                time_enroll_final >= time_10 &
                  time_enroll_final <= time_90)
              Data_survival_tmp3$time_enroll_final =
                Data_survival_tmp3$time_enroll_final - (time_10 - 1)
              Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
                time_enroll_final > time_90) %>%
                dplyr::arrange(time_enroll_final)
              
              Data_survival_tmp4$time_enroll_final = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp4)
              Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx < max(idx)*(1-survival_rate_10)],])
              Data_drug_survival_2 = rbind(Data_survival_tmp, Data_survival_tmp3)  
              
              Data_survival_tmp_pos = Data_drug_survival_2_curve %>%
                dplyr::filter(gene_mut == TRUE)
              if(sum(Data_survival_tmp_pos$censor) > 0){
                fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_pos)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)
              Data_survival_tmp5 = Data_survival_tmp_pos %>% dplyr::filter(
                time_palliative_enroll >= time_10 &
                  time_palliative_enroll <= time_90)
              Data_survival_tmp5$time_palliative_enroll =
                Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
              Data_survival_tmp6 = Data_survival_tmp_pos %>% dplyr::filter(
                time_palliative_enroll > time_90) %>%
                dplyr::arrange(time_palliative_enroll)
              Data_survival_tmp6$time_palliative_enroll = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp6)
              Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
              
              Data_survival_tmp_neg = Data_drug_survival_2_curve %>%
                dplyr::filter(gene_mut == FALSE)
              
              if(sum(Data_survival_tmp_neg$censor) > 0){
                fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)
              Data_survival_tmp7 = Data_survival_tmp_neg %>% dplyr::filter(
                time_palliative_enroll >= time_10 &
                  time_palliative_enroll <= time_90)
              Data_survival_tmp7$time_palliative_enroll =
                Data_survival_tmp7$time_palliative_enroll - (time_10 - 1)
              Data_survival_tmp8 = Data_survival_tmp_neg %>% dplyr::filter(
                time_palliative_enroll > time_90) %>%
                dplyr::arrange(time_palliative_enroll)
              Data_survival_tmp8$time_palliative_enroll = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp8)
              Data_survival_tmp7 = rbind(Data_survival_tmp7, Data_survival_tmp8[idx[idx < max(idx)*(1-survival_rate_10)],])
              
              Median_entry_pos = median(Data_survival_tmp5$time_palliative_enroll)
              Median_entry_neg = median(Data_survival_tmp7$time_palliative_enroll)
              Data_drug_survival_2_2 = rbind(Data_survival_tmp5, Data_survival_tmp7)  
              
              Data_drug_survival_2 = Data_drug_survival_2 %>%
                dplyr::mutate(delay_1 = case_when(
                  time_palliative_enroll <= Median_entry_pos &
                    gene_mut == TRUE ~ "SPT to entry early",
                  time_palliative_enroll > Median_entry_pos &
                    gene_mut == TRUE ~ "SPT to entry later",
                  time_palliative_enroll <= Median_entry_neg &
                    gene_mut == FALSE ~ "SPT to entry early",
                  time_palliative_enroll > Median_entry_neg &
                    gene_mut == FALSE ~ "SPT to entry later"))
              
              if(sum((Data_drug_survival_2 %>% dplyr::filter(gene_mut == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
                 sum((Data_drug_survival_2 %>% dplyr::filter(gene_mut == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
                 sum((Data_drug_survival_2 %>% dplyr::filter(gene_mut == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
                 sum((Data_drug_survival_2 %>% dplyr::filter(gene_mut == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
                
                if(input$drug_survuval_CTx == "Yes"){
                  stan_weibull_survival_model_data <-
                    list(
                      Median_pos = as.integer(Median_entry_pos),
                      Median_neg = as.integer(Median_entry_neg),
                      Nobs_early_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                      Nobs_late_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                      Ncen_early_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                      Ncen_late_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                      Nobs_early_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                      Nobs_late_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                      Ncen_early_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                      Ncen_late_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                      Nexp_pos = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE]),
                      Nexp_neg = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE]),
                      ybef_exp_pos = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE],
                      ybef_exp_neg = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE],
                      yobs_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                      yobs_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                      ycen_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                      ycen_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                      yobs_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                      yobs_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                      ycen_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                      ycen_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE]
                    )
                  
                  stan_model_path <- "source/Stan_code_loglog_weibull_factor.stan"
                  stan_model_compiled <- cmdstan_model(stan_model_path)
                  stan_weibull_survival_model_fit <- stan_model_compiled$sample(
                    data = stan_weibull_survival_model_data,
                    seed = 1234,
                    chains = 4,
                    parallel_chains = min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE)),
                    iter_sampling = 2000,  # (iter-warmup)
                    iter_warmup = 1000,
                    thin = 1,
                    refresh = 500
                  )
                  
                  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
                  stan_weibull_survival_model_draws_total <-
                    stan_weibull_survival_model_draws %>%
                    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
                    tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
                    dplyr::select(-key) 
                  stan_weibull_survival_model_draws_total_exp <-
                    stan_weibull_survival_model_draws %>%
                    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
                    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
                    dplyr::select(-key) 
                  
                  median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
                  median_os_list_weibull_total = matrix(stan_weibull_survival_model_draws_total$yhat_total, ncol=4000)
                  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
                  median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
                  median_os_list_weibull_total = colMedians(median_os_list_weibull_total,na.rm = T)
                  median_os_list_weibull_bias = colMedians(median_os_list_weibull_bias,na.rm = T)
                  
                  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
                  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
                  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
                  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
                  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
                  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_drug_survival_2_curve$censor) * as.integer(length(yhat_weibull_sort_total)/length(Data_drug_survival_2_curve$censor))]
                  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_drug_survival_2_curve$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_drug_survival_2_curve$censor))]
                  
                  Data_drug_survival_2_curve$factor_neg = yhat_weibull_sort_average_total
                  Data_drug_survival_2_curve$factor_pos = yhat_weibull_sort_average_total_exp
                  Data_drug_survival_2_curve$factor_neg[Data_drug_survival_2_curve$factor_neg > 10000] = 10000
                  Data_drug_survival_2_curve$factor_pos[Data_drug_survival_2_curve$factor_pos > 10000] = 10000
                  
                  traditional_fit = survfit(Surv(event = censor,
                                                 time = time_palliative_final) ~ gene_mut,
                                            data = Data_drug_survival_2_curve)
                  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_drug_survival_2_curve$factor_neg)),
                                                    time = factor_neg) ~ 1, 
                                               data = Data_drug_survival_2_curve)
                  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_drug_survival_2_curve$factor_pos)),
                                                    time = factor_pos) ~ 1, 
                                               data = Data_drug_survival_2_curve)
                  
                  g_surv_drug[[kkkkk]] = ggsurvplot(
                    fit = list(traditional_fit = traditional_fit,
                               simulation_fit_neg = simulation_fit_neg,
                               simulation_fit_pos = simulation_fit_pos),
                    combine = TRUE,
                    data = Data_drug_survival_2_curve,
                    xlab = "Time from chemotherapy induction to final observation (months)",
                    ylab = "Survival Probability",
                    censor = TRUE,
                    font.title = 8,
                    font.subtitle = 8,
                    font.main = 8,
                    font.submain = 8,
                    font.caption = 8,
                    font.legend = 8,
                    surv.median.line = "hv",
                    conf.int = FALSE,
                    pval = FALSE,
                    risk.table = TRUE,
                    risk.table.y.text = FALSE,
                    tables.theme = clean_theme(), 
                    legend = c(0.8,0.8),
                    xlim = c(0, max(Data_drug_survival_2_curve$time_palliative_final) * 1.05),
                    cumevents = FALSE,
                    cumcensor = FALSE,
                    break.x.by = 365.25 * 2.5,
                    xscale = "d_m",
                    #break.x.by = ceiling(max(Data_drug_survival_2_curve$time_palliative_final) / 500) * 100,
                    legend.labs = c(paste0(name_1, ", Unadjusted"),
                                    paste0(name_2, ", Unadjusted"),
                                    paste0(name_1, ", Adjusted for Delayed Entry"),
                                    paste0(name_2, ", Adjusted for Delayed Entry"))
                  ) +   
                    labs(title = paste(sum(traditional_fit[[1]]), " patients ",
                                       "Median OS ",
                                       round(summary(traditional_fit)$table[[13]] / 365.25 * 12, digits = 1)," (",
                                       round(summary(traditional_fit)$table[[15]] / 365.25 * 12, digits = 1),"-",
                                       round(summary(traditional_fit)$table[[17]] / 365.25 * 12, digits = 1),") (",
                                       name_1_short,", unadj.)/",
                                       round(summary(traditional_fit)$table[[14]] / 365.25 * 12, digits = 1), " (",
                                       round(summary(traditional_fit)$table[[16]] / 365.25 * 12, digits = 1),"-",
                                       round(summary(traditional_fit)$table[[18]] / 365.25 * 12, digits = 1),") (",
                                       name_2_short,", unadj.)/",
                                       round(median_os_summary_weibull_total[[2]] / 365.25 * 12, digits = 1), 
                                       " (", round(median_os_summary_weibull_total[[1]] / 365.25 * 12, digits = 1), "-",
                                       round(median_os_summary_weibull_total[[3]] / 365.25 * 12, digits = 1), ") (",
                                       name_1_short,", adj.)/",
                                       round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                                       round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                                       round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1),
                                       ") (", name_2_short,", adj.) months",
                                       sep=""),
                         subtitle = paste("Survival difference, ",name_1_short, "-", name_2_short, ": ",
                                          round(median_os_summary_weibull_bias[[2]] / 365.25 * 12, digits = 1), " (",
                                          round(median_os_summary_weibull_bias[[1]] / 365.25 * 12, digits = 1), "-", 
                                          round(median_os_summary_weibull_bias[[3]] / 365.25 * 12, digits = 1), 
                                          ") months, median (95% CI)",
                                          sep=""))
                  g_surv_drug[[kkkkk]]$table <- g_surv_drug[[kkkkk]]$table + theme(plot.title = element_blank(),
                                                                                   plot.subtitle = element_blank())
                  kkkkk = kkkkk + 1
                }
              }
              if(length(traditional_fit[[1]])>1){
                delayed_entry_fit = survfit(Surv(event = censor,
                                                 time = time_palliative_enroll,
                                                 time2 = time_palliative_final) ~ gene_mut,
                                            data = Data_drug_survival_2_curve)
                if(length(unique(Data_drug_survival_2_curve$gene_mut)) > 1){
                  diff_0 = coxph(Surv(time = time_palliative_enroll,
                                      time2 = time_palliative_final, censor)~gene_mut,
                                 data=Data_drug_survival_2_curve)
                  g_surv_drug[[kkkkk]] = surv_curv_CTx(delayed_entry_fit, Data_drug_survival_2_curve, paste0(paste(traditional_fit[[1]],collapse = "/"), " patients"),
                                                       c(paste0(name_1, ", Adjusted for Delayed Entry"),
                                                         paste0(name_2, ", Adjusted for Delayed Entry")),
                                                       diff_0)
                  kkkkk = kkkkk + 1
                }
              }
            }
          }
        }
  
        # Survival analysis for drug with propensity score matching, #4
        if(!is.null(input$drug_group_1) & !is.null(input$drug_group_2)){
          Data_ID_drug_1_original = unique((Data_drug_original %>%
                                              dplyr::filter(治療ライン %in% input$target_line) %>%
                                              dplyr::filter(Drug %in% input$drug_group_1))$ID)
          Data_ID_drug_2_original = unique((Data_drug_original %>%
                                              dplyr::filter(治療ライン %in% input$target_line) %>%
                                              dplyr::filter(Drug %in% input$drug_group_2))$ID)
          Data_ID_drug_1and2_original = intersect(Data_ID_drug_1_original, Data_ID_drug_2_original)
          Data_ID_drug_1_original = setdiff(Data_ID_drug_1_original, Data_ID_drug_1and2_original)
          Data_ID_drug_2_original = setdiff(Data_ID_drug_2_original, Data_ID_drug_1and2_original)
          
          Data_drug_survival_2 = Data_drug_survival_save %>%
            dplyr::mutate(CTx = case_when(
              C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_drug_1_original ~ input$drug_group_1_name,
              C.CAT調査結果.基本項目.ハッシュID %in% Data_ID_drug_2_original ~ input$drug_group_2_name,
              TRUE ~ "Other"
            ))
          Data_drug_survival_2 = Data_drug_survival_2 %>%
            dplyr::filter(CTx != "Other") %>%
            dplyr::filter(!is.na(time_palliative_enroll) & !is.na(time_enroll_final))
          
          ID_CTx_start_data = Data_case_target %>%
            dplyr::select(C.CAT調査結果.基本項目.ハッシュID, CTx_start_date) %>%
            dplyr::distinct(C.CAT調査結果.基本項目.ハッシュID, .keep_all = T)
          
          Data_drug_survival_2 = dplyr::left_join(Data_drug_survival_2, ID_CTx_start_data, by = "C.CAT調査結果.基本項目.ハッシュID")
          Data_drug_survival_2$CTx_start_date = as.integer(as.Date(Data_drug_survival_2$CTx_start_date))
          Data_drug_survival_2$CTx_binary = as.integer(as.factor(Data_drug_survival_2$CTx)) - 1
          # propensity score matching considering CTx start date
          formula_PS = formula("CTx_binary ~ CTx_start_date")
          propensityScore = glm(formula_PS,
                                family  = binomial(link = "logit"),
                                data    = Data_drug_survival_2)
          Data_drug_survival_2$PScore = propensityScore$fitted.values
          matching = Matching::Match(Y = Data_drug_survival_2$time_palliative_final,
                                     Tr = Data_drug_survival_2$CTx_binary,
                                     X = Data_drug_survival_2$PScore,
                                     caliper=0.25,
                                     ties=F,
                                     replace=F)
          if(!is.na(matching)[[1]]){
            Data_drug_survival_2 = Data_drug_survival_2[c(matching$index.treated, matching$index.control),]
          }
          
          if(sum(Data_drug_survival_2$censor) > 0){
            traditional_fit = survfit(Surv(event = censor,
                                           time = time_palliative_final) ~ CTx, 
                                      data = Data_drug_survival_2)
            if(length(Data_drug_survival_2[,1]) != traditional_fit[[1]][[1]]){
              name_1 = input$drug_group_1_name
              name_2 = input$drug_group_2_name
              name_1_short = input$drug_group_1_name
              name_2_short = input$drug_group_2_name
              
              Data_drug_survival_2$gene_mut = Data_drug_survival_2$CTx == input$drug_group_2_name
              Data_drug_survival_2_curve = Data_drug_survival_2
              Data_survival_tmp_pos = Data_drug_survival_2_curve %>%
                dplyr::filter(gene_mut == TRUE)
              if(sum(Data_survival_tmp_pos$censor) > 0){
                fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_pos)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_enroll_final)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_enroll_final)*0.9)
              Data_survival_tmp = Data_survival_tmp_pos %>% dplyr::filter(
                time_enroll_final >= time_10 &
                  time_enroll_final <= time_90)
              Data_survival_tmp$time_enroll_final =
                Data_survival_tmp$time_enroll_final - (time_10 - 1)
              Data_survival_tmp2 = Data_survival_tmp_pos %>% dplyr::filter(
                time_enroll_final > time_90) %>%
                dplyr::arrange(time_enroll_final)
              
              Data_survival_tmp2$time_enroll_final = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp2)
              Data_survival_tmp = rbind(Data_survival_tmp, Data_survival_tmp2[idx[idx < max(idx)*(1-survival_rate_10)],])
              
              Data_survival_tmp_neg = Data_drug_survival_2_curve %>%
                dplyr::filter(gene_mut == FALSE)
              if(sum(Data_survival_tmp_neg$censor) > 0){
                fit = survfit(Surv(time_enroll_final, censor) ~ 1, data = Data_survival_tmp_neg)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_enroll_final)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_enroll_final)*0.9)
              Data_survival_tmp3 = Data_survival_tmp_neg %>% dplyr::filter(
                time_enroll_final >= time_10 &
                  time_enroll_final <= time_90)
              Data_survival_tmp3$time_enroll_final =
                Data_survival_tmp3$time_enroll_final - (time_10 - 1)
              Data_survival_tmp4 = Data_survival_tmp_neg %>% dplyr::filter(
                time_enroll_final > time_90) %>%
                dplyr::arrange(time_enroll_final)
              
              Data_survival_tmp4$time_enroll_final = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp4)
              Data_survival_tmp3 = rbind(Data_survival_tmp3, Data_survival_tmp4[idx[idx < max(idx)*(1-survival_rate_10)],])
              Data_drug_survival_2 = rbind(Data_survival_tmp, Data_survival_tmp3)  
              
              Data_survival_tmp_pos = Data_drug_survival_2_curve %>%
                dplyr::filter(gene_mut == TRUE)
              if(sum(Data_survival_tmp_pos$censor) > 0){
                fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_pos)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_pos$time_palliative_enroll)*0.9)
              Data_survival_tmp5 = Data_survival_tmp_pos %>% dplyr::filter(
                time_palliative_enroll >= time_10 &
                  time_palliative_enroll <= time_90)
              Data_survival_tmp5$time_palliative_enroll =
                Data_survival_tmp5$time_palliative_enroll - (time_10 - 1)
              Data_survival_tmp6 = Data_survival_tmp_pos %>% dplyr::filter(
                time_palliative_enroll > time_90) %>%
                dplyr::arrange(time_palliative_enroll)
              Data_survival_tmp6$time_palliative_enroll = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp6)
              Data_survival_tmp5 = rbind(Data_survival_tmp5, Data_survival_tmp6[idx[idx < max(idx)*(1-survival_rate_10)],])
              
              Data_survival_tmp_neg = Data_drug_survival_2_curve %>%
                dplyr::filter(gene_mut == FALSE)
              
              if(sum(Data_survival_tmp_neg$censor) > 0){
                fit = survfit(Surv(time_palliative_enroll, censor) ~ 1, data = Data_survival_tmp_neg)
                time_10 = min(90, quantile(fit, 0.10)$quantile[[1]],na.rm = T)
                survival_rate_10 = summary(fit,times = time_10)[[6]]
                survival_rate_90 = summary(fit,times = max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)[[6]]
              } else {
                time_10 = 90
                survival_rate_10 = 0
                survival_rate_90 = 1
              }
              time_90 = max(time_10 + 1, max(Data_survival_tmp_neg$time_palliative_enroll)*0.9)
              Data_survival_tmp7 = Data_survival_tmp_neg %>% dplyr::filter(
                time_palliative_enroll >= time_10 &
                  time_palliative_enroll <= time_90)
              Data_survival_tmp7$time_palliative_enroll =
                Data_survival_tmp7$time_palliative_enroll - (time_10 - 1)
              Data_survival_tmp8 = Data_survival_tmp_neg %>% dplyr::filter(
                time_palliative_enroll > time_90) %>%
                dplyr::arrange(time_palliative_enroll)
              Data_survival_tmp8$time_palliative_enroll = time_90 - (time_10 - 1)
              idx <- 1:nrow(Data_survival_tmp8)
              Data_survival_tmp7 = rbind(Data_survival_tmp7, Data_survival_tmp8[idx[idx < max(idx)*(1-survival_rate_10)],])
              
              Median_entry_pos = median(Data_survival_tmp5$time_palliative_enroll)
              Median_entry_neg = median(Data_survival_tmp7$time_palliative_enroll)
              Data_drug_survival_2_2 = rbind(Data_survival_tmp5, Data_survival_tmp7)  
              
              Data_drug_survival_2 = Data_drug_survival_2 %>%
                dplyr::mutate(delay_1 = case_when(
                  time_palliative_enroll <= Median_entry_pos &
                    gene_mut == TRUE ~ "SPT to entry early",
                  time_palliative_enroll > Median_entry_pos &
                    gene_mut == TRUE ~ "SPT to entry later",
                  time_palliative_enroll <= Median_entry_neg &
                    gene_mut == FALSE ~ "SPT to entry early",
                  time_palliative_enroll > Median_entry_neg &
                    gene_mut == FALSE ~ "SPT to entry later"))
              
              if(sum((Data_drug_survival_2 %>% dplyr::filter(gene_mut == TRUE & delay_1 == "SPT to entry later"))$censor) >= 1 &
                 sum((Data_drug_survival_2 %>% dplyr::filter(gene_mut == FALSE & delay_1 == "SPT to entry later"))$censor) >= 1 &
                 sum((Data_drug_survival_2 %>% dplyr::filter(gene_mut == TRUE & delay_1 != "SPT to entry later"))$censor) >= 1 &
                 sum((Data_drug_survival_2 %>% dplyr::filter(gene_mut == FALSE & delay_1 != "SPT to entry later"))$censor) >= 1){
                
                if(input$drug_survuval_CTx == "Yes"){  
                  stan_weibull_survival_model_data <-
                    list(
                      Median_pos = as.integer(Median_entry_pos),
                      Median_neg = as.integer(Median_entry_neg),
                      Nobs_early_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                      Nobs_late_pos = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                      Ncen_early_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                      Ncen_late_pos = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE),
                      Nobs_early_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                      Nobs_late_neg = sum(Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                      Ncen_early_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                      Ncen_late_neg = sum(Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE),
                      Nexp_pos = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE]),
                      Nexp_neg = length(Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE]),
                      ybef_exp_pos = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut != FALSE],
                      ybef_exp_neg = Data_drug_survival_1_2$time_palliative_enroll[Data_drug_survival_1_2$gene_mut == FALSE],
                      yobs_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                      yobs_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                      ycen_early_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                      ycen_late_pos = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut != FALSE],
                      yobs_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                      yobs_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 1 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                      ycen_early_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 != "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE],
                      ycen_late_neg = Data_drug_survival_1$time_enroll_final[Data_drug_survival_1$censor == 0 & Data_drug_survival_1$delay_1 == "SPT to entry later" & Data_drug_survival_1$gene_mut == FALSE]
                    )
                  
                  stan_model_path <- "source/Stan_code_loglog_weibull_factor.stan"
                  stan_model_compiled <- cmdstan_model(stan_model_path)
                  stan_weibull_survival_model_fit <- stan_model_compiled$sample(
                    data = stan_weibull_survival_model_data,
                    seed = 1234,
                    chains = 4,
                    parallel_chains = min(4, max(1, parallel::detectCores() - 1, na.rm = TRUE)),
                    iter_sampling = 2000,  # (iter-warmup)
                    iter_warmup = 1000,
                    thin = 1,
                    refresh = 500
                  )
                  
                  stan_weibull_survival_model_draws <- tidybayes::tidy_draws(stan_weibull_survival_model_fit)
                  stan_weibull_survival_model_draws_total <-
                    stan_weibull_survival_model_draws %>%
                    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_total")) %>%
                    tidyr::gather(key = key, value = yhat_total, starts_with("yhat_total")) %>%
                    dplyr::select(-key) 
                  stan_weibull_survival_model_draws_total_exp <-
                    stan_weibull_survival_model_draws %>%
                    dplyr::select(.chain, .iteration, .draw, starts_with("yhat_factor_total")) %>%
                    tidyr::gather(key = key, value = yhat_total_exp, starts_with("yhat_factor_total")) %>%
                    dplyr::select(-key) 
                  
                  median_os_list_weibull_total_exp = matrix(stan_weibull_survival_model_draws_total_exp$yhat_total_exp, ncol=4000)
                  median_os_list_weibull_total = matrix(stan_weibull_survival_model_draws_total$yhat_total, ncol=4000)
                  median_os_list_weibull_bias = median_os_list_weibull_total - median_os_list_weibull_total_exp
                  median_os_list_weibull_total_exp = colMedians(median_os_list_weibull_total_exp,na.rm = T)
                  median_os_list_weibull_total = colMedians(median_os_list_weibull_total,na.rm = T)
                  median_os_list_weibull_bias = colMedians(median_os_list_weibull_bias,na.rm = T)
                  
                  median_os_summary_weibull_total = quantile(median_os_list_weibull_total, probs = c(0.025, 0.5, 0.975))
                  median_os_summary_weibull_total_exp = quantile(median_os_list_weibull_total_exp, probs = c(0.025, 0.5, 0.975))
                  median_os_summary_weibull_bias = quantile(median_os_list_weibull_bias, probs = c(0.025, 0.5, 0.975))
                  yhat_weibull_sort_total = sort(stan_weibull_survival_model_draws_total$yhat_total)
                  yhat_weibull_sort_total_exp = sort(stan_weibull_survival_model_draws_total_exp$yhat_total_exp)
                  yhat_weibull_sort_average_total = yhat_weibull_sort_total[1:length(Data_drug_survival_2_curve$censor) * as.integer(length(yhat_weibull_sort_total)/length(Data_drug_survival_2_curve$censor))]
                  yhat_weibull_sort_average_total_exp = yhat_weibull_sort_total_exp[1:length(Data_drug_survival_2_curve$censor) * as.integer(length(yhat_weibull_sort_total_exp)/length(Data_drug_survival_2_curve$censor))]
                  
                  Data_drug_survival_2_curve$factor_neg = yhat_weibull_sort_average_total
                  Data_drug_survival_2_curve$factor_pos = yhat_weibull_sort_average_total_exp
                  Data_drug_survival_2_curve$factor_neg[Data_drug_survival_2_curve$factor_neg > 10000] = 10000
                  Data_drug_survival_2_curve$factor_pos[Data_drug_survival_2_curve$factor_pos > 10000] = 10000
                  
                  traditional_fit = survfit(Surv(event = censor,
                                                 time = time_palliative_final) ~ gene_mut,
                                            data = Data_drug_survival_2_curve)
                  simulation_fit_neg = survfit(Surv(event = rep(1,length(Data_drug_survival_2_curve$factor_neg)),
                                                    time = factor_neg) ~ 1, 
                                               data = Data_drug_survival_2_curve)
                  simulation_fit_pos = survfit(Surv(event = rep(1,length(Data_drug_survival_2_curve$factor_pos)),
                                                    time = factor_pos) ~ 1, 
                                               data = Data_drug_survival_2_curve)
                  
                  g_surv_drug[[kkkkk]] = ggsurvplot(
                    fit = list(traditional_fit = traditional_fit,
                               simulation_fit_neg = simulation_fit_neg,
                               simulation_fit_pos = simulation_fit_pos),
                    combine = TRUE,
                    data = Data_drug_survival_2_curve,
                    xlab = "Time from chemotherapy induction to final observation (months)",
                    ylab = "Survival Probability",
                    censor = TRUE,
                    font.title = 8,
                    font.subtitle = 8,
                    font.main = 8,
                    font.submain = 8,
                    font.caption = 8,
                    font.legend = 8,
                    surv.median.line = "hv",
                    conf.int = FALSE,
                    pval = FALSE,
                    risk.table = TRUE,
                    risk.table.y.text = FALSE,
                    tables.theme = clean_theme(), 
                    legend = c(0.8,0.8),
                    xlim = c(0, max(Data_drug_survival_2_curve$time_palliative_final) * 1.05),
                    cumevents = FALSE,
                    cumcensor = FALSE,
                    break.x.by = 365.25 * 2.5,
                    xscale = "d_m",
                    #break.x.by = ceiling(max(Data_drug_survival_2_curve$time_palliative_final) / 500) * 100,
                    legend.labs = c(paste0(name_1, ", Unadjusted"),
                                    paste0(name_2, ", Unadjusted"),
                                    paste0(name_1, ", Adjusted for Delayed Entry"),
                                    paste0(name_2, ", Adjusted for Delayed Entry"))
                  ) +   
                    labs(title = paste("CTx start date matched, ", sum(traditional_fit[[1]]), " patients ",
                                       "Median OS ",
                                       round(summary(traditional_fit)$table[[13]] / 365.25 * 12, digits = 1)," (",
                                       round(summary(traditional_fit)$table[[15]] / 365.25 * 12, digits = 1),"-",
                                       round(summary(traditional_fit)$table[[17]] / 365.25 * 12, digits = 1),") (",
                                       name_1_short,", unadj.)/",
                                       round(summary(traditional_fit)$table[[14]] / 365.25 * 12, digits = 1), " (",
                                       round(summary(traditional_fit)$table[[16]] / 365.25 * 12, digits = 1),"-",
                                       round(summary(traditional_fit)$table[[18]] / 365.25 * 12, digits = 1),") (",
                                       name_2_short,", unadj.)/",
                                       round(median_os_summary_weibull_total[[2]] / 365.25 * 12, digits = 1), 
                                       " (", round(median_os_summary_weibull_total[[1]] / 365.25 * 12, digits = 1), "-",
                                       round(median_os_summary_weibull_total[[3]] / 365.25 * 12, digits = 1), ") (",
                                       name_1_short,", adj.)/",
                                       round(median_os_summary_weibull_total_exp[[2]] / 365.25 * 12, digits = 1), " (",
                                       round(median_os_summary_weibull_total_exp[[1]] / 365.25 * 12, digits = 1), "-",
                                       round(median_os_summary_weibull_total_exp[[3]] / 365.25 * 12, digits = 1),
                                       ") (", name_2_short,", adj.) months",
                                       sep=""),
                         subtitle = paste("Survival difference, ",name_1_short, "-", name_2_short, ": ",
                                          round(median_os_summary_weibull_bias[[2]] / 365.25 * 12, digits = 1), " (",
                                          round(median_os_summary_weibull_bias[[1]] / 365.25 * 12, digits = 1), "-", 
                                          round(median_os_summary_weibull_bias[[3]] / 365.25 * 12, digits = 1), 
                                          ") months, median (95% CI)",
                                          sep=""))
                  g_surv_drug[[kkkkk]]$table <- g_surv_drug[[kkkkk]]$table + theme(plot.title = element_blank(),
                                                                                   plot.subtitle = element_blank())
                  kkkkk = kkkkk + 1
                }
              }
              if(length(traditional_fit[[1]])>1){
                delayed_entry_fit = survfit(Surv(event = censor,
                                                 time = time_palliative_enroll,
                                                 time2 = time_palliative_final) ~ gene_mut,
                                            data = Data_drug_survival_2_curve)
                if(length(unique(Data_drug_survival_2_curve$gene_mut)) > 1){
                  diff_0 = coxph(Surv(time = time_palliative_enroll,
                                      time2 = time_palliative_final, censor)~gene_mut,
                                 data=Data_drug_survival_2_curve)
                  g_surv_drug[[kkkkk]] = surv_curv_CTx(delayed_entry_fit, Data_drug_survival_2_curve, paste0("CTx start date matched, ",paste(traditional_fit[[1]],collapse = "/"), " patients"),
                                                       c(paste0(name_1, ", Adjusted for Delayed Entry"),
                                                         paste0(name_2, ", Adjusted for Delayed Entry")),
                                                       diff_0)
                  kkkkk = kkkkk + 1
                }
              }
            }
          }
        }
      }
      if(kkkkk != 13){
        for(kkkkkk in kkkkk:12){
          g_surv_drug[[kkkkkk]] = ggsurvplot_empty()
        }
      }
      output$figure_survival_drug = renderPlot({
        arrange_ggsurvplots(g_surv_drug,print=TRUE,ncol=3,nrow=4,surv.plot.height = 0.8,risk.table.height = 0.2)
      })
      
    })
  })
  observeEvent(input$logout_button, {
    # JavaScript confirm を使ってYes/No確認
    shinyjs::runjs("
      if (confirm('Want to log out?')) {
        Shiny.setInputValue('confirm_logout', true, {priority: 'event'});
      }
    ")
  })
  observeEvent(input$confirm_logout, {
    stopApp()
  })
  session$onSessionEnded(function() {
    res <- tryCatch({
      GET("https://hogehoge.com/logout_notify?user=USERNAME")
    }, error = function(e) {
      NULL
    })
    if (!is.null(res) && status_code(res) == 200) {
      print("Log out.")
      stopApp()
    } else {
      print("Log-out error.")
      stopApp()
    }
  })
}
# Run the application 
shinyApp(ui = ui, server = server)
